---
output: html_document
editor_options: 
  chunk_output_type: console
---

Project description: to investigate the associations between DNA methylation and birthweight, and how genotypes could confound this association.

This paper is under minor revision in the journal of clinical epigenetics.

```{r Preparing files for ewas to run}

library(tidyverse)
library(data.table)

setwd("~/mQTL/birthweight")

winsorize <- function(methylation,pct=winsorize.pct) {
  quantiles <- matrixStats::rowQuantiles(methylation, probs=c(pct,1-pct), na.rm=T)
  low <- quantiles[,1]
  upper <- quantiles[,2]

  outliers.lower <- rowSums(methylation < low, na.rm=T)
  outliers.upper <- rowSums(methylation > upper, na.rm=T)
  
  idx <- which(methylation < low, arr.ind=T)
  methylation[idx] <- low[idx[,1]]
  
  idx <- which(methylation > upper, arr.ind=T)
  methylation[idx] <- upper[idx[,1]]

  n <- rowSums(!is.na(methylation))
  log <- data.frame(outliers.lower, outliers.upper, n)
  return(list(methylation=methylation, log=log))
}

for(set in c(1,2,3,4)){

 if(set==1){
   var = "birthWt"
   clinical0 = read.csv(paste0("~/sets/set",set,"/clinical_variables.csv")) %>%
     filter(Trisomy21.cnm==0)
    }
  
 if(set==2){
  var = "birthWt"
  clinical0 = read.csv(paste0("~/sets/set",set,"/clinical_variables.csv")) %>%
      filter(Trisomy21.cnm==0)
  }
  
 if(set==3){
   var = "Y_birthweight"
   clinical0 = read.csv(paste0("~/sets/set",set,"/clinical_variables.csv")) %>% 
     filter(Trisomy21==0)%>%
     mutate(gestage=Y_gestation_week)
   }
  
 if(set==4){
   var = "ch_birthwt_bc_full"
   clinical0 = read.csv(paste0("~/sets/set",set,"/clinical_variables.csv")) %>%
     filter(smp_type=="bg",Trisomy21.cnm==0)
   }
  
  ###################################
  # Determine subjects to involve
  
  ## Methylation data
  ####noob
  if(set!=4){methy_tm_0 = read_rds(paste0("~/sets/set",set,"/beta_ritu_imputed_set",set,".rds"))}
  if(set==4){methy_tm_0 = read_rds(paste0("~/sets/set",set,"/beta_funnorm_bmiq_imputed_set",set,".rds"))}
  rownames(methy_tm_0)=c()
  metMemDF = data.frame(beadPosition=colnames(methy_tm_0)[-1]) 
  

  ## Cov data
  clinical = clinical0[,c("beadPosition","subjectId","sex","Batch","gestage","CaCo",var)] %>%
    filter(gestage>=30)%>%
    na.omit() %>%
    mutate(ID = paste0("0_",subjectId))%>%
    distinct(ID,.keep_all = TRUE)
  
  # clinical0 %>% filter(gestage<30) %>% nrow()
  # set1 : 2
  # set2 : 4
  # set3 : 1
  # set4 : 2
  
  dcp = fread(paste0("~/sets/set",set,"/idol_deconvolution.csv")) %>%
    as.data.frame() %>%
    dplyr::select(beadPosition,CD8T,CD4T,NK,Bcell,Mono,nRBC) %>%
    na.omit()
  
  genePC = read.table(paste0("~/mQTL/ccls.set",set,".pc.eigenvec"))
  colnames(genePC) = c("FID","IID","PC1","PC2","PC3","PC4","PC5","PC6","PC7","PC8","PC9","PC10")
  genePC1 = genePC %>%
    dplyr::mutate(ID = paste0(FID,"_",IID)) %>%
    dplyr::select(ID,PC1,PC2,PC3,PC4,PC5,PC6,PC7,PC8,PC9,PC10)
  
  clinical1 = clinical %>%
    inner_join(metMemDF,by="beadPosition") %>%
    inner_join(dcp,by="beadPosition") %>%
    inner_join(genePC1,by="ID") %>%
    dplyr::mutate(Batch=as.factor(Batch)) 
  covMem = clinical1$ID
  
  ## Genome data
  genMem = read.table(paste0("/ccls/proj_circle2_p3/users/sebastian/bySET/ccls.gwas.set",set,".fam")) %>%
      dplyr::mutate(ID = paste0(V1,"_",V2)) %>%
      dplyr::select(ID)
  
  genMems = genMem$ID
  
  ## Methylation data processing
  methy_tm =  methy_tm_0 %>%
    column_to_rownames("probeId") %>%
    as.data.frame() %>% 
    data.matrix() 

  replace.outliers <- winsorize(methy_tm,0.005)
  methy <- replace.outliers$methylation %>% 
    as.data.frame() %>%
    rownames_to_column(var="probeId")
  
  ## Choosing overlaps, respecting the sequence of genetics, also outputting a list of subjects dropped from genetics
  miInd = which(!genMems%in%covMem) %>% unique()
  
  if(length(miInd)!=0){
    finalMems = data.frame(ID = genMems[-miInd])
  }else{
    finalMems = data.frame(ID = genMems)}
  
  ###################################
  # Preparing covariates
  clinical2 = clinical1 %>%
    dplyr::select(-beadPosition, -subjectId) %>%
    dplyr::select(ID, one_of(var),everything()) 
  
  write.table(clinical2,paste0("covsSet",set,".txt"),sep='\t',quote=FALSE,row.names = FALSE,col.names = TRUE)
  
  ###################################
  # Preparing methylation files
  subs = clinical1 %>% column_to_rownames(var="ID")
  subs1 = subs[as.character(finalMems$ID),"beadPosition"] %>% as.character()
  methy1 = methy[,c("probeId",subs1)]
  colnames(methy1) = c("probeId",as.character(finalMems$ID))
  
  write.table(methy1,paste0("methSet",set,".txt"),sep='\t',quote=FALSE,row.names = FALSE,col.names = TRUE)
  
  print(set)
  print((dim(methy1)[2]-1)==(dim(clinical2)[1]))

}

# output sample size for each set
setwd("~/mQTL/birthweight")
library(tidyverse)
library(data.table)

for (set in c(1:4)){
  
  covs0 = fread(paste0("covsSet",set,".txt"))
  
  print(paste("set",set,nrow(covs0)))
}

# [1] "set 1 378"
# [1] "set 2 354"
# [1] "set 3 554"
# [1] "set 4 464"
```

```{r adjusting  for scanned mqtls with QTLtools}

setwd("~/mQTL/birthweight")
library(tidyverse)
library(data.table)
library(MASS)
library(lmtest)
library(sandwich)

library(parallel) 
cl <- makeCluster(40, type="FORK")
clusterEvalQ(cl,library(MASS))
clusterEvalQ(cl,library(lmtest))

linear <- function(meth,cov){

  #    write_rds(meth,"meth.test.rds")
  #}
  
  # meth = read_rds("meth.test.rds")
  # cov = covs

  probe <- unname(meth[1])
  methDF = meth[-1] %>% as.numeric()
  
  dat <- data.frame(y = methDF, cov)
  ff  <- formula(paste("y ~ ", paste0(colnames(cov),collapse="+")))
  #mod <- try(rlm(ff, data = dat, maxit = 200),silent = TRUE)
  mod <- try(lm(ff, data = dat),silent = TRUE)
  
  #cf = try(coeftest(mod),silent = TRUE)
  N <- sum(!is.na(dat[,1]))
  
  if (class(mod)=="lm") {
    
    coefs = summary(mod)$coefficients
    coef = coefs[2,"Estimate"]
    SE = coefs[2,"Std. Error"]
    p.value = coefs[2,"Pr(>|t|)"]
    out = c(coef, SE, p.value, N) 
  } else {
    out = c(rep(NA,3), N)
  }
  names(out) = c("Beta","SE", "P", "N"); return(out) 
}
gene.linear <- function(meth,cov,gene){
  
  #    write_rds(meth,"meth.test.rds")
  #}
  
  # meth = read_rds("meth.test.rds")
  # cov = covs
  # gene = mtqls
  
  probe <- unname(meth[1])
  
  #print(probe)
  
  methDF = meth[-1] %>% as.numeric()
  
  SNPname <- gene[which(gene$CpG == probe), 1]  %>% as.character()  
  SNP <-  gene[which(gene$CpG == probe), -c(1:3)] %>% t() %>% as.data.frame()
  colnames(SNP) <- "SNP"
  
  if(length(table(SNP$SNP))==1){
    dat <- data.frame(y = methDF, cov)
    ff  <- formula(paste("y ~ ", paste0(colnames(cov),collapse="+")))
  }else{
    dat <- data.frame(y = methDF, cov,SNP)
    ff  <- formula(paste("y ~ ", paste0(c(colnames(cov),"SNP"),collapse="+")))
  }
  
  mod <- try(lm(ff, data = dat),silent = TRUE)
  
  #cf = try(coeftest(mod),silent = TRUE)
  N <- sum(!is.na(dat[,1]))
  
  if (class(mod)=="lm") {
    
    SNPname=SNPname
    #pvals = try(f.robftest(mod, var = colnames(cov)[1]),silent = TRUE)
    #coef = cf[2,"Estimate"]
    #F.value = ifelse(class(pvals)=="htest",pvals$statistic,NA)
    #p.value = ifelse(class(pvals)=="htest",pvals$p.value,NA)
    #out = c(SNPname,coef, F.value, p.value, N) 

    coefs = summary(mod)$coefficients
    coef = coefs[2,"Estimate"]
    SE = coefs[2,"Std. Error"]
    p.value = coefs[2,"Pr(>|t|)"]
    out = c(SNPname,coef, SE, p.value, N) 
  } else {
    out = c(SNPname,rep(NA,3), N)
  }
  names(out) = c("snp","Beta","SE", "P", "N"); return(out) 
}

for (set in c(1:4)){
  
  if(set==1){var = "birthWt"}
  if(set==2){var = "birthWt"}
  if(set==3){var = "Y_birthweight"}
  if(set==4){var = "ch_birthwt_bc_full"}
  
  methyl0 = fread(paste0("methSet",set,".txt")) %>%
    column_to_rownames(var="probeId")%>%
    as.data.frame()
  
  covs0 = fread(paste0("covsSet",set,".txt"))%>%
    as.data.frame()%>%
    column_to_rownames(var="ID") %>%
    mutate(Batch = as.factor(Batch))
  
  mtqls0 = fread(paste0("~/mQTL/ccls.set",set,".mqtls.hd")) %>% 
    as.data.frame()
  
  subs.list = intersect(intersect(rownames(covs0),colnames(methyl0)), colnames(mtqls0)) 
  
  methyl = methyl0[,subs.list]
  covs = covs0[subs.list,]
  mtqls = mtqls0[,c("SNP","CpG","P-value",subs.list)]
  
  probes = rownames(methyl)[which(rownames(methyl)%in%mtqls$CpG)] 
  
  # no correction for SNP
  methyl = methyl %>% rownames_to_column(var="cpg")
  rownames(methyl) <- methyl$cpg

  #methylt = methyl[1:100,] 
  #results <- t(apply(methylt, 1, linear, covs)) 
  
  results <- t(parApply(cl, methyl, 1, linear, covs)) 
  results <- as.data.frame(results) %>%
    rownames_to_column(var="CpG") %>%
    mutate(N=as.numeric(N))%>%
    mutate(P=as.numeric(P))%>%
    filter(N>max(N)/2,
           !is.na(P))
  
  write.table(results,paste0("ccls.set",set,".",var,".Nogene.ewas.txt"),
              quote=FALSE,
              row.names = FALSE,
              sep='\t')  
  
  Nbeta <- sum(!is.na(results$P))
  tmp <- results[!is.na(results$P),]
  print(paste0("lambda of set ",set," without SNP control:"))
  lambda = median(qchisq(1- tmp$P,1))/qchisq(0.5,1)
  print(lambda)
  
  # correcting for SNPs for CpGs with mQTLs only
  methyl.nosnp = methyl[setdiff(rownames(methyl),probes),] 

  #methyl.nosnp.t = methyl.nosnp[1:100,]
  #results.nosnp <- t(apply(methyl.nosnp.t, 1, linear, covs)) 
  results.nosnp <- t(parApply(cl, methyl.nosnp, 1, linear, covs)) 
  
  results.nosnp <- as.data.frame(results.nosnp) %>%
    rownames_to_column(var="CpG")%>%
    mutate(snp="none")%>%
    mutate(N=as.numeric(N))%>%
    mutate(P=as.numeric(P))%>%
    filter(N>max(N)/2,
           !is.na(P)) %>%
    dplyr::select(CpG,snp,Beta,SE,P,N)
  
  methyl.cpg = methyl[probes,]
  
  #methyl.cpg.t = methyl.cpg[1:100,]
  #results.snp <- t(apply(methyl.cpg.t, 1, gene.linear, covs, mtqls)) 
  results.snp <- t(parApply(cl, methyl.cpg, 1, gene.linear, covs, mtqls)) 
  results.snp <- as.data.frame(results.snp) %>%
    rownames_to_column(var="CpG")%>%
    mutate(N=as.numeric(N))%>%
    mutate(P=as.numeric(P))%>%
    filter(N>max(N)/2,
           !is.na(P))
  
  results.snp.combind = rbind(results.nosnp,results.snp) %>%
    as.data.frame()%>%
    arrange(P)

  write.table(results.snp.combind,
              paste0("ccls.set",set,".",var,".geneCon.ewas.txt"),
              quote=FALSE,
              row.names = FALSE,
              sep='\t')  
  
  tmp.snp <- results.snp.combind %>% 
    dplyr::select(P) %>% mutate(P=as.numeric(P)) %>% na.omit() 
  lambda.snp = median(qchisq(1- tmp.snp$P,1))/qchisq(0.5,1)
  print(paste0("lambda of set ",set," with SNP control:"))
  print(round(lambda.snp, 2))

}

stopCluster(cl)
rm(list=ls())

#[1] "lambda of set 1 without SNP control:"
#[1] 1.370214
#[1] "lambda of set 1 with SNP control:"
#[1] 1.39
#[1] "lambda of set 2 without SNP control:"
#[1] 1.117313
#[1] "lambda of set 2 with SNP control:"
#[1] 1.12
#[1] "lambda of set 3 without SNP control:"
#[1] 1.495706
#[1] "lambda of set 3 with SNP control:"
#[1] 1.53
#[1] "lambda of set 4 without SNP control:"
#[1] 0.9170985
#[1] "lambda of set 4 with SNP control:"
#[1] 0.93

```

```{r EWAS using mQTLs from all four sets AND EPIC probes}

setwd("~/mQTL/birthweight")
library(tidyverse)
library(data.table)
library(MASS)
library(lmtest)
library(sandwich)

library(parallel) 
cl <- makeCluster(40, type="FORK")
clusterEvalQ(cl,library(MASS))
clusterEvalQ(cl,library(lmtest))

linear <- function(meth,cov){

  #    write_rds(meth,"meth.test.rds")
  #}
  
  # meth = read_rds("meth.test.rds")
  # cov = covs

  probe <- unname(meth[1])
  methDF = meth[-1] %>% as.numeric()
  
  dat <- data.frame(y = methDF, cov)
  ff  <- formula(paste("y ~ ", paste0(colnames(cov),collapse="+")))
  #mod <- try(rlm(ff, data = dat, maxit = 200),silent = TRUE)
  mod <- try(lm(ff, data = dat),silent = TRUE)
  
  #cf = try(coeftest(mod),silent = TRUE)
  N <- sum(!is.na(dat[,1]))
  
  if (class(mod)=="lm") {
    
    coefs = summary(mod)$coefficients
    coef = coefs[2,"Estimate"]
    SE = coefs[2,"Std. Error"]
    p.value = coefs[2,"Pr(>|t|)"]
    out = c(coef, SE, p.value, N) 
  } else {
    out = c(rep(NA,3), N)
  }
  names(out) = c("Beta","SE", "P", "N"); return(out) 
}
gene.linear <- function(meth,cov,gene){
  
  #    write_rds(meth,"meth.test.rds")
  #}
  
  # meth = read_rds("meth.test.rds")
  # cov = covs
  # gene = mtqls
  
  probe <- unname(meth[1])
  
  #print(probe)
  
  methDF = meth[-1] %>% as.numeric()
  
  SNPname <- gene[which(gene$CpG == probe), 1]  %>% as.character()  
  SNP <-  gene[which(gene$CpG == probe), -c(1:3)] %>% t() %>% as.data.frame()
  colnames(SNP) <- "SNP"
  
  if(length(table(SNP$SNP))==1){
    dat <- data.frame(y = methDF, cov)
    ff  <- formula(paste("y ~ ", paste0(colnames(cov),collapse="+")))
  }else{
    dat <- data.frame(y = methDF, cov,SNP)
    ff  <- formula(paste("y ~ ", paste0(c(colnames(cov),"SNP"),collapse="+")))
  }
  
  mod <- try(lm(ff, data = dat),silent = TRUE)
  
  #cf = try(coeftest(mod),silent = TRUE)
  N <- sum(!is.na(dat[,1]))
  
  if (class(mod)=="lm") {
    
    SNPname=SNPname
    #pvals = try(f.robftest(mod, var = colnames(cov)[1]),silent = TRUE)
    #coef = cf[2,"Estimate"]
    #F.value = ifelse(class(pvals)=="htest",pvals$statistic,NA)
    #p.value = ifelse(class(pvals)=="htest",pvals$p.value,NA)
    #out = c(SNPname,coef, F.value, p.value, N) 

    coefs = summary(mod)$coefficients
    coef = coefs[2,"Estimate"]
    SE = coefs[2,"Std. Error"]
    p.value = coefs[2,"Pr(>|t|)"]
    out = c(SNPname,coef, SE, p.value, N) 
  } else {
    out = c(SNPname,rep(NA,3), N)
  }
  names(out) = c("snp","Beta","SE", "P", "N"); return(out) 
}


library(IlluminaHumanMethylation450kanno.ilmn12.hg19)
data("IlluminaHumanMethylation450kanno.ilmn12.hg19")
anno       <- IlluminaHumanMethylation450kanno.ilmn12.hg19 %>%
  getAnnotation %>%
  as.data.frame %>%
  dplyr::select(Name)

library(IlluminaHumanMethylationEPICanno.ilm10b4.hg19)
data("IlluminaHumanMethylationEPICanno.ilm10b4.hg19")
annoEPIC       <- IlluminaHumanMethylationEPICanno.ilm10b4.hg19 %>%
  getAnnotation %>%
  as.data.frame %>%
  dplyr::select(Name)


commonCpGs = intersect(anno$Name,annoEPIC$Name)

epicCpGs = anti_join(annoEPIC,anno,by="Name")
epicCpGs = epicCpGs$Name

for (set in c(1:4)){
  
  if(set==1){var = "birthWt"}
  if(set==2){var = "birthWt"}
  if(set==3){var = "Y_birthweight"}
  if(set==4){var = "ch_birthwt_bc_full"}
  
  methyl0 = fread(paste0("methSet",set,".txt")) %>%
    column_to_rownames(var="probeId")%>%
    as.data.frame()
  
  covs0 = fread(paste0("covsSet",set,".txt"))%>%
    as.data.frame()%>%
    column_to_rownames(var="ID") %>%
    mutate(Batch = as.factor(Batch))
  
  mtqls0 = fread(paste0("~/mQTL/ccls.set",set,".all.mqtls.hd")) %>% 
    as.data.frame()
  
  subs.list = intersect(intersect(rownames(covs0),colnames(methyl0)), colnames(mtqls0)) 
  
  if(set %in% c(1,2)){
    
  methyl = methyl0[,subs.list]
  covs = covs0[subs.list,]
  mtqls = mtqls0[,c("SNP","CpG","P-value",subs.list)]
  
  probes = rownames(methyl)[which(rownames(methyl)%in%mtqls$CpG)] 
  
  methyl = methyl %>% rownames_to_column(var="cpg")
  rownames(methyl) <- methyl$cpg
  
  # no correction for SNP is the same as previous block results so don't need to be done again
  # #methylt = methyl[1:100,] 
  # #results <- t(apply(methylt, 1, linear, covs)) 
  # 
  # results <- t(parApply(cl, methyl, 1, linear, covs)) 
  # results <- as.data.frame(results) %>%
  #   rownames_to_column(var="CpG") %>%
  #   mutate(N=as.numeric(N))%>%
  #   mutate(P=as.numeric(P))%>%
  #   filter(N>max(N)/2,
  #          !is.na(P))
  # 
  # write.table(results,paste0("ccls.set",set,".",var,".Nogene.ewas.txt"),
  #             quote=FALSE,
  #             row.names = FALSE,
  #             sep='\t')  
  # 
  # Nbeta <- sum(!is.na(results$P))
  # tmp <- results[!is.na(results$P),]
  # print(paste0("lambda of set ",set," without SNP control:"))
  # lambda = median(qchisq(1- tmp$P,1))/qchisq(0.5,1)
  # print(lambda)
  
  # correcting for SNPs for CpGs with mQTLs only
  methyl.nosnp = methyl[setdiff(rownames(methyl),probes),] 

  #methyl.nosnp.t = methyl.nosnp[1:100,]
  #results.nosnp <- t(apply(methyl.nosnp.t, 1, linear, covs)) 
  results.nosnp <- t(parApply(cl, methyl.nosnp, 1, linear, covs)) 
  
  results.nosnp <- as.data.frame(results.nosnp) %>%
    rownames_to_column(var="CpG")%>%
    mutate(snp="none")%>%
    mutate(N=as.numeric(N))%>%
    mutate(P=as.numeric(P))%>%
    filter(N>max(N)/2,
           !is.na(P)) %>%
    dplyr::select(CpG,snp,Beta,SE,P,N)
  
  methyl.cpg = methyl[probes,]
  
  #methyl.cpg.t = methyl.cpg[1:100,]
  #results.snp <- t(apply(methyl.cpg.t, 1, gene.linear, covs, mtqls)) 
  results.snp <- t(parApply(cl, methyl.cpg, 1, gene.linear, covs, mtqls)) 
  results.snp <- as.data.frame(results.snp) %>%
    rownames_to_column(var="CpG")%>%
    mutate(N=as.numeric(N))%>%
    mutate(P=as.numeric(P))%>%
    filter(N>max(N)/2,
           !is.na(P))
  
  results.snp.combind = rbind(results.nosnp,results.snp) %>%
    as.data.frame()%>%
    arrange(P)

  write.table(results.snp.combind,
              paste0("ccls.set",set,".",var,".geneCon.ewas.commonCpGs.allMQTLs.txt"),
              quote=FALSE,
              row.names = FALSE,
              sep='\t')  
  
  tmp.snp <- results.snp.combind %>% 
    dplyr::select(P) %>% mutate(P=as.numeric(P)) %>% na.omit() 
  lambda.snp = median(qchisq(1- tmp.snp$P,1))/qchisq(0.5,1)
  print(paste0("lambda of set ",set," with SNP control using ALL MQTLs:"))
  print(round(lambda.snp, 2))
  }

  if(set %in% c(3,4)){
    
  covs = covs0[subs.list,]
  mtqls = mtqls0[,c("SNP","CpG","P-value",subs.list)]
  
  ########################################################
  # common CpGs  
  methyl_common = methyl0[which(rownames(methyl0)%in%commonCpGs),subs.list]
  probes = rownames(methyl_common)[which(rownames(methyl_common)%in%mtqls$CpG)] 
  
  #no correction for SNP
  methyl_common = methyl_common %>% rownames_to_column(var="cpg")
  rownames(methyl_common) <- methyl_common$cpg

  #methylt = methyl_common[1:100,]
  #results <- t(apply(methylt, 1, linear, covs))

  results <- t(parApply(cl, methyl_common, 1, linear, covs))
  results <- as.data.frame(results) %>%
    rownames_to_column(var="CpG") %>%
    mutate(N=as.numeric(N))%>%
    mutate(P=as.numeric(P))%>%
    filter(N>max(N)/2,
           !is.na(P))

  write.table(results,paste0("ccls.set",set,".",var,".Nogene.ewas.commonCpGs.allMQTLs.txt"),
              quote=FALSE,
              row.names = FALSE,
              sep='\t')

  Nbeta <- sum(!is.na(results$P))
  tmp <- results[!is.na(results$P),]
  print(paste0("lambda of set ",set," without SNP control (common CpGs):"))
  lambda = median(qchisq(1- tmp$P,1))/qchisq(0.5,1)
  print(lambda)

  # correcting for SNPs for CpGs with mQTLs only
  # selecting CpGs not matched to mQTLs first and do non corrected model
  methyl_common.nosnp = methyl_common[setdiff(rownames(methyl_common),probes),] 

  #methyl.nosnp.t = methyl_common.nosnp[1:100,]
  #results.nosnp <- t(apply(methyl.nosnp.t, 1, linear, covs)) 
  results.nosnp <- t(parApply(cl, methyl_common.nosnp, 1, linear, covs)) 
  
  results.nosnp <- as.data.frame(results.nosnp) %>%
    rownames_to_column(var="CpG")%>%
    mutate(snp="none")%>%
    mutate(N=as.numeric(N))%>%
    mutate(P=as.numeric(P))%>%
    filter(N>max(N)/2,
           !is.na(P)) %>%
    dplyr::select(CpG,snp,Beta,SE,P,N)
  
  methyl_common.cpg = methyl_common[probes,]
  
  #methyl.cpg.t = methyl_common.cpg[1:100,]
  #results.snp <- t(apply(methyl.cpg.t, 1, gene.linear, covs, mtqls)) 
  results.snp <- t(parApply(cl, methyl_common.cpg, 1, gene.linear, covs, mtqls)) 
  results.snp <- as.data.frame(results.snp) %>%
    rownames_to_column(var="CpG")%>%
    mutate(N=as.numeric(N))%>%
    mutate(P=as.numeric(P))%>%
    filter(N>max(N)/2,
           !is.na(P))
  
  results.snp.combind = rbind(results.nosnp,results.snp) %>%
    as.data.frame()%>%
    arrange(P)

  write.table(results.snp.combind,
              paste0("ccls.set",set,".",var,".geneCon.ewas.commonCpGs.allMQTLs.txt"),
              quote=FALSE,
              row.names = FALSE,
              sep='\t')  
  
  tmp.snp <- results.snp.combind %>% 
    dplyr::select(P) %>% mutate(P=as.numeric(P)) %>% na.omit() 
  lambda.snp = median(qchisq(1- tmp.snp$P,1))/qchisq(0.5,1)
  print(paste0("lambda of set ",set," with SNP control using ALL MQTLs (common CpGs):"))
  print(round(lambda.snp, 2))


  ########################################################
  # EPIC CpGs  
  methyl_epic = methyl0[which(rownames(methyl0)%in%epicCpGs),subs.list]
  probes = rownames(methyl_epic)[which(rownames(methyl_epic)%in%mtqls$CpG)] 
  
  # no correction for SNP
  methyl_epic = methyl_epic %>% rownames_to_column(var="cpg")
  rownames(methyl_epic) <- methyl_epic$cpg

  #methylt = methyl_epic[1:100,] 
  #results <- t(apply(methylt, 1, linear, covs)) 
  
  results <- t(parApply(cl, methyl_epic, 1, linear, covs)) 
  results <- as.data.frame(results) %>%
    rownames_to_column(var="CpG") %>%
    mutate(N=as.numeric(N))%>%
    mutate(P=as.numeric(P))%>%
    filter(N>max(N)/2,
           !is.na(P))
  
  write.table(results,paste0("ccls.set",set,".",var,".Nogene.ewas.epicCpGs.allMQTLs.txt"),
              quote=FALSE,
              row.names = FALSE,
              sep='\t')  
  
  Nbeta <- sum(!is.na(results$P))
  tmp <- results[!is.na(results$P),]
  print(paste0("lambda of set ",set," without SNP control (epic CpGs):"))
  lambda = median(qchisq(1- tmp$P,1))/qchisq(0.5,1)
  print(lambda)
  
  # correcting for SNPs for CpGs with mQTLs only
  # selecting CpGs not matched to mQTLs first and do non corrected model
  methyl_epic.nosnp = methyl_epic[setdiff(rownames(methyl_epic),probes),] 

  #methyl.nosnp.t = methyl_epic.nosnp[1:100,]
  #results.nosnp <- t(apply(methyl.nosnp.t, 1, linear, covs)) 
  results.nosnp <- t(parApply(cl, methyl_epic.nosnp, 1, linear, covs)) 
  
  results.nosnp <- as.data.frame(results.nosnp) %>%
    rownames_to_column(var="CpG")%>%
    mutate(snp="none")%>%
    mutate(N=as.numeric(N))%>%
    mutate(P=as.numeric(P))%>%
    filter(N>max(N)/2,
           !is.na(P)) %>%
    dplyr::select(CpG,snp,Beta,SE,P,N)
  
  methyl_epic.cpg = methyl_epic[probes,]
  
  #methyl.cpg.t = methyl_epic.cpg[1:100,]
  #results.snp <- t(apply(methyl.cpg.t, 1, gene.linear, covs, mtqls)) 
  results.snp <- t(parApply(cl, methyl_epic.cpg, 1, gene.linear, covs, mtqls)) 
  results.snp <- as.data.frame(results.snp) %>%
    rownames_to_column(var="CpG")%>%
    mutate(N=as.numeric(N))%>%
    mutate(P=as.numeric(P))%>%
    filter(N>max(N)/2,
           !is.na(P))
  
  results.snp.combind = rbind(results.nosnp,results.snp) %>%
    as.data.frame()%>%
    arrange(P)

  write.table(results.snp.combind,
              paste0("ccls.set",set,".",var,".geneCon.ewas.epicCpGs.allMQTLs.txt"),
              quote=FALSE,
              row.names = FALSE,
              sep='\t')  
  
  tmp.snp <- results.snp.combind %>% 
    dplyr::select(P) %>% mutate(P=as.numeric(P)) %>% na.omit() 
  lambda.snp = median(qchisq(1- tmp.snp$P,1))/qchisq(0.5,1)
  print(paste0("lambda of set ",set," with SNP control using ALL MQTLs (epic CpGs):"))
  print(round(lambda.snp, 2))
     
      
    
  }  
  
  

}

stopCluster(cl)
rm(list=ls())


#[1] "lambda of set 1 with SNP control using ALL MQTLs:"
#[1] 1.39
#[1] "lambda of set 2 with SNP control using ALL MQTLs:"
#[1] 1.12
#[1] "lambda of set 3 without SNP control (common CpGs):"
#[1] 1.218287
#[1] "lambda of set 3 with SNP control using ALL MQTLs (common CpGs):"
#[1] 1.24
#[1] "lambda of set 3 without SNP control (epic CpGs):"
#[1] 1.888519
#[1] "lambda of set 3 with SNP control using ALL MQTLs (epic CpGs):"
#[1] 1.95
#[1] "lambda of set 4 without SNP control (common CpGs):"
#[1] 0.92146
#[1] "lambda of set 4 with SNP control using ALL MQTLs (common CpGs):"
#[1] 0.93
#[1] "lambda of set 4 without SNP control (epic CpGs):"
#[1] 0.9120701
#[1] "lambda of set 4 with SNP control using ALL MQTLs (epic CpGs):"
#[1] 0.92

```

```{r reviewer: GOF test for EWAS models}

setwd("~/mQTL/birthweight")
library(tidyverse)
library(data.table)
library(MASS)
library(lmtest)
library(sandwich)

library(parallel) 
cl <- makeCluster(40, type="FORK")
clusterEvalQ(cl,library(MASS))
clusterEvalQ(cl,library(lmtest))

gofModel <- function(meth,cov,gene){
  
  #     write_rds(meth,"meth.test.rds")
  # }
  
  # meth = read_rds("meth.test.rds")
  # cov = covs
  # gene = mtqls
  
  probe <- unname(meth[1])
  
  #print(probe)
  
  methDF = meth[-1] %>% as.numeric()
  
  SNPname <- gene[which(gene$CpG == probe), 1]  %>% as.character()  
  SNP <-  gene[which(gene$CpG == probe), -c(1:3)] %>% t() %>% as.data.frame()
  colnames(SNP) <- "SNP"
  
  if(length(table(SNP$SNP))>1){
    dat <- data.frame(y = methDF, cov,SNP)
    ff  <- formula(paste("y ~ ", paste0(c(colnames(cov),"SNP"),collapse="+")))
    ff_red <- formula(paste("y ~ ", paste0(colnames(cov),collapse="+")))
  }

  mod <- try(lm(ff, data = dat),silent = TRUE)
  mod_red <- try(lm(ff_red, data = dat),silent = TRUE)
  ano <- anova(mod_red,mod)
  
  N <- sum(!is.na(dat[,1]))
  
  if (class(ano)[1]=="anova") {
    
    SNPname=SNPname
    RSS_full = ano$RSS[1]
    RSS_red = ano$RSS[2]
    F_val = ano$F[2]
    F_p_val = ano$`Pr(>F)`[2]
    out = c(SNPname,RSS_full, RSS_red, F_val, F_p_val,N) 
    
  } else {
    out = c(SNPname,rep(NA,4), N)
  }
  names(out) = c("snp","RSS_full","RSS_red", "F_val","F_p_val", "N"); return(out) 
}

  set=4
  # for (set in c(1:4)){
  #   
  #   if(set==1){var = "birthWt"}
  #   if(set==2){var = "birthWt"}
  #   if(set==3){var = "Y_birthweight"}
  if(set==4){var = "ch_birthwt_bc_full"}
  
  methyl0 = fread(paste0("methSet",set,".txt")) %>%
    column_to_rownames(var="probeId")%>%
    as.data.frame()
  
  covs0 = fread(paste0("covsSet",set,".txt"))%>%
    as.data.frame()%>%
    column_to_rownames(var="ID") %>%
    mutate(Batch = as.factor(Batch))
  
  mtqls0 = fread(paste0("~/mQTL/ccls.set",set,".mqtls.hd")) %>% 
    as.data.frame()
  
  subs.list = intersect(intersect(rownames(covs0),colnames(methyl0)), colnames(mtqls0)) 
  
  methyl = methyl0[,subs.list]
  covs = covs0[subs.list,]
  mtqls = mtqls0[,c("SNP","CpG","P-value",subs.list)]
  
  probes = rownames(methyl)[which(rownames(methyl)%in%mtqls$CpG)] 
  methyl = methyl %>% rownames_to_column(var="cpg")
  rownames(methyl) <- methyl$cpg

  # correcting for SNPs for CpGs with mQTLs only
  methyl.cpg = methyl[probes,]

  #methyl.cpg.t = methyl.cpg[1:10,]
  #results.snp <- t(apply(methyl.cpg.t, 1, gofModel, covs, mtqls)) 
  
  results.snp <- t(parApply(cl, methyl.cpg, 1, gofModel, covs, mtqls)) 
  results.snp <- as.data.frame(results.snp) %>%
    rownames_to_column(var="CpG")%>%
    mutate(N=as.numeric(N))%>%
    mutate(F_p_val=as.numeric(F_p_val))%>%
    filter(N>max(N)/2,
           !is.na(F_p_val))
  
  write.table(results.snp,
              "gof.test.set.4.txt",
              quote=FALSE,
              row.names = FALSE,
              sep='\t')  
  #}

stopCluster(cl)
rm(list=ls())

################################
# look at the results
setwd("~/mQTL/birthweight")
library(tidyverse)
library(data.table)

# library(IlluminaHumanMethylation450kanno.ilmn12.hg19)
# data("IlluminaHumanMethylation450kanno.ilmn12.hg19")
# anno       <- IlluminaHumanMethylation450kanno.ilmn12.hg19 %>%
#   getAnnotation %>%
#   as.data.frame %>%
#   rownames_to_column(var="cpg") %>%
#   dplyr::select(cpg,chr,pos,Relation_to_Island,UCSC_RefGene_Name,UCSC_RefGene_Group,Probe_maf,CpG_maf)

library(IlluminaHumanMethylationEPICanno.ilm10b4.hg19)
data("IlluminaHumanMethylationEPICanno.ilm10b4.hg19")
annoEPIC       <- IlluminaHumanMethylationEPICanno.ilm10b4.hg19 %>%
  getAnnotation %>%
  as.data.frame %>%
  rownames_to_column(var="cpg") %>%
  dplyr::select(cpg,chr,pos,Relation_to_Island,UCSC_RefGene_Name,UCSC_RefGene_Group,Methyl450_Loci,CpG_maf,Probe_maf)

results.snp = fread("gof.test.set.4.txt") %>%
    dplyr::filter(grepl('cg', CpG)) %>%
    left_join(annoEPIC,by=c("CpG"="cpg")) %>%
    dplyr::filter(Probe_maf<0.05|is.na(Probe_maf))%>%
    dplyr::filter(CpG_maf<0.05|is.na(CpG_maf)) %>%
    dplyr::filter(chr!="chrX" & chr!="chrY")

table(results.snp$F_p_val<0.05)
#FALSE   TRUE 
#96532 199625 

# 199625/(199625+96532) = 0.6740513

hist(-log10(results.snp$F_p_val))
```

```{bash meta-analysis}

## Meta-analysis by platform

cd ~/mQTL/birthweight

### 450K
metal
MARKER CpG
EFFECT Beta
PVALUE P
SCHEME STDERR
STDERR SE
VERBOSE OFF
GENOMICCONTROL OFF
STDERR_PRINT_PRECISION 20
EFFECT_PRINT_PRECISION 20
PROCESS ccls.set1.birthWt.geneCon.ewas.txt
PROCESS ccls.set2.birthWt.geneCon.ewas.txt
ANALYZE
QUIT
mv METAANALYSIS1.TBL meta.450K.bw.geneCon.txt

### 450K, noGene
metal
MARKER CpG
EFFECT Beta
PVALUE P
SCHEME STDERR
STDERR SE
VERBOSE OFF
GENOMICCONTROL OFF
STDERR_PRINT_PRECISION 20
EFFECT_PRINT_PRECISION 20
PROCESS ccls.set1.birthWt.Nogene.ewas.txt
PROCESS ccls.set1.birthWt.Nogene.ewas.txt
ANALYZE
QUIT
mv METAANALYSIS1.TBL meta.450K.bw.Nogene.txt

### EPIC
metal
MARKER CpG
EFFECT Beta
PVALUE P
SCHEME STDERR
STDERR SE
VERBOSE OFF
GENOMICCONTROL OFF
STDERR_PRINT_PRECISION 20
EFFECT_PRINT_PRECISION 20
PROCESS ccls.set3.Y_birthweight.geneCon.ewas.txt
PROCESS ccls.set4.ch_birthwt_bc_full.geneCon.ewas.txt
ANALYZE
QUIT
mv METAANALYSIS1.TBL meta.EPIC.bw.geneCon.txt

### EPIC, noGene
metal
MARKER CpG
EFFECT Beta
PVALUE P
SCHEME STDERR
STDERR SE
VERBOSE OFF
GENOMICCONTROL OFF
STDERR_PRINT_PRECISION 20
EFFECT_PRINT_PRECISION 20
PROCESS ccls.set3.Y_birthweight.Nogene.ewas.txt
PROCESS ccls.set4.ch_birthwt_bc_full.Nogene.ewas.txt
ANALYZE
QUIT
mv METAANALYSIS1.TBL meta.EPIC.bw.Nogene.txt

### 450K, with mQTLs from all platforms
metal
MARKER CpG
EFFECT Beta
PVALUE P
SCHEME STDERR
STDERR SE
VERBOSE OFF
GENOMICCONTROL OFF
STDERR_PRINT_PRECISION 20
EFFECT_PRINT_PRECISION 20
PROCESS ccls.set1.birthWt.geneCon.ewas.commonCpGs.allMQTLs.txt
PROCESS ccls.set2.birthWt.geneCon.ewas.commonCpGs.allMQTLs.txt
ANALYZE
QUIT
mv METAANALYSIS1.TBL meta.450K.all.Platforms.bw.geneCon.txt

metal
MARKER CpG
EFFECT Beta
PVALUE P
SCHEME STDERR
STDERR SE
VERBOSE OFF
GENOMICCONTROL OFF
STDERR_PRINT_PRECISION 20
EFFECT_PRINT_PRECISION 20
PROCESS ccls.set1.birthWt.geneCon.ewas.commonCpGs.allMQTLs.txt
PROCESS ccls.set2.birthWt.geneCon.ewas.commonCpGs.allMQTLs.txt
PROCESS ccls.set3.Y_birthweight.geneCon.ewas.commonCpGs.allMQTLs.txt
PROCESS ccls.set4.ch_birthwt_bc_full.geneCon.ewas.commonCpGs.allMQTLs.txt
ANALYZE
QUIT
mv METAANALYSIS1.TBL meta.450K.all.allsubs.Platforms.bw.geneCon.txt

### 450K, allsubs, noGene
metal
MARKER CpG
EFFECT Beta
PVALUE P
SCHEME STDERR
STDERR SE
VERBOSE OFF
GENOMICCONTROL OFF
STDERR_PRINT_PRECISION 20
EFFECT_PRINT_PRECISION 20
PROCESS ccls.set1.birthWt.Nogene.ewas.commonCpGs.allMQTLs.txt
PROCESS ccls.set2.birthWt.Nogene.ewas.commonCpGs.allMQTLs.txt
PROCESS ccls.set3.Y_birthweight.Nogene.ewas.commonCpGs.allMQTLs.txt
PROCESS ccls.set4.ch_birthwt_bc_full.Nogene.ewas.commonCpGs.allMQTLs.txt
ANALYZE
QUIT
mv METAANALYSIS1.TBL meta.450K.all.allsubs.Platforms.bw.Nogene.txt


```

```{r looking at results after meta}

setwd("~/mQTL/birthweight")
library(tidyverse)
library(data.table)
library(qvalue)
library(grid)
library(bacon)

library(IlluminaHumanMethylation450kanno.ilmn12.hg19)
data("IlluminaHumanMethylation450kanno.ilmn12.hg19")
anno       <- IlluminaHumanMethylation450kanno.ilmn12.hg19 %>%
  getAnnotation %>%
  as.data.frame %>%
  rownames_to_column(var="cpg") %>%
  dplyr::select(cpg,chr,pos,Relation_to_Island,UCSC_RefGene_Name,UCSC_RefGene_Group,Probe_maf,CpG_maf,Regulatory_Feature_Group)

library(IlluminaHumanMethylationEPICanno.ilm10b4.hg19)
data("IlluminaHumanMethylationEPICanno.ilm10b4.hg19")
annoEPIC       <- IlluminaHumanMethylationEPICanno.ilm10b4.hg19 %>%
  getAnnotation %>%
  as.data.frame %>%
  rownames_to_column(var="cpg") %>%
  dplyr::select(cpg,chr,pos,Relation_to_Island,UCSC_RefGene_Name,UCSC_RefGene_Group,Methyl450_Loci,CpG_maf,Probe_maf,Regulatory_Feature_Group)

gold = read.csv("bw_meta_paper.csv") %>%
  as.data.frame() %>%
  dplyr::select(CpG)
  
set1.snp = fread("ccls.set1.birthWt.geneCon.ewas.txt") %>%
  mutate(set1_snp=snp)%>%
  dplyr::select(CpG,set1_snp)
set2.snp = fread("ccls.set2.birthWt.geneCon.ewas.txt")%>%
  mutate(set2_snp=snp)%>%
  dplyr::select(CpG,set2_snp)
set3.snp = fread("ccls.set3.Y_birthweight.geneCon.ewas.txt")%>%
  mutate(set3_snp=snp)%>%
  dplyr::select(CpG,set3_snp)
set4.snp = fread("ccls.set4.ch_birthwt_bc_full.geneCon.ewas.txt")%>%
  mutate(set4_snp=snp)%>%
  dplyr::select(CpG,set4_snp)
  
#########################################################################################
##### Counting number of significant CpGs with/without SNP controls (same CpGs with mQTL)


for(platform in c("450K","EPIC")){
  
  if(platform=="450K")annoFile = anno
  if(platform=="EPIC")annoFile = annoEPIC
  
  if(platform=="450K"){
    meta = fread(paste0("meta.",platform,".bw.geneCon.txt")) %>%
      dplyr::select(-Allele1, -Allele2) %>%
      dplyr::filter(grepl('cg', MarkerName)) %>%
      dplyr::rename("CpG" = "MarkerName") %>%
      left_join(set1.snp,by="CpG")%>%
      left_join(set2.snp,by="CpG") %>%
      left_join(annoFile,by=c("CpG"="cpg")) %>%
      dplyr::filter(Probe_maf<0.05|is.na(Probe_maf))%>%
      dplyr::filter(CpG_maf<0.05|is.na(CpG_maf)) %>%
      dplyr::filter(chr!="chrX" & chr!="chrY")%>%
      mutate(`P-value`=as.numeric(`P-value`))%>%
      dplyr::filter(!is.na(`P-value`))%>%
      arrange(by=`P-value`) %>%
      mutate(tss=0)
    meta$tss[grep("TSS",meta$UCSC_RefGene_Group)]=1
    
    bacon_P = pval(bacon((meta$Effect)/(meta$StdErr),na.exclude = TRUE))

    # raw P values
    Nbeta <- sum(!is.na(meta$`P-value`))
    tmp <- meta[!is.na(meta$`P-value`),]
    lambda = median(qchisq(1- tmp$`P-value`,1))/qchisq(0.5,1)
    print(round(lambda, 2))
    #1.3
    
    # bacon P values
    Nbeta <- sum(!is.na(bacon_P))
    tmp <- bacon_P[!is.na(bacon_P)]
    lambda = median(qchisq(1- bacon_P,1))/qchisq(0.5,1)
    print(round(lambda, 2))
    #1.03
  
    meta$bacon_P = bacon_P
    meta = meta %>%
      mutate(fdr=p.adjust(meta$`P-value`, method = "fdr", n=length(meta$`P-value`))) %>%
      mutate(bonferroni=p.adjust(meta$`P-value`, method = "bonferroni", n=length(meta$`P-value`))) 

    0.05/nrow(meta) # bonferroni threshold
    # 1.192299e-07
  }

  if(platform=="EPIC"){
    
    meta = fread(paste0("meta.",platform,".bw.geneCon.txt")) %>%
      dplyr::select(-Allele1, -Allele2) %>%
      dplyr::filter(grepl('cg', MarkerName)) %>%
      dplyr::rename("CpG" = "MarkerName") %>%
      left_join(set3.snp,by="CpG")%>%
      left_join(set4.snp,by="CpG") %>%
      left_join(annoFile,by=c("CpG"="cpg")) %>%
      dplyr::filter(Probe_maf<0.05|is.na(Probe_maf))%>%
      dplyr::filter(CpG_maf<0.05|is.na(CpG_maf)) %>%
      dplyr::filter(chr!="chrX" & chr!="chrY")%>%
      mutate(`P-value`=as.numeric(`P-value`))%>%
      dplyr::filter(!is.na(`P-value`))%>%
      arrange(by=`P-value`) %>%
      mutate(tss=0)
    meta$tss[grep("TSS",meta$UCSC_RefGene_Group)]=1
    
    bacon_P = pval(bacon((meta$Effect)/(meta$StdErr),na.exclude = TRUE))

    # raw P values
    Nbeta <- sum(!is.na(meta$`P-value`))
    tmp <- meta[!is.na(meta$`P-value`),]
    lambda = median(qchisq(1- tmp$`P-value`,1))/qchisq(0.5,1)
    print(round(lambda, 2))
    #1.64
    
    # bacon P values
    Nbeta <- sum(!is.na(bacon_P))
    tmp <- bacon_P[!is.na(bacon_P)]
    lambda = median(qchisq(1- bacon_P,1))/qchisq(0.5,1)
    print(round(lambda, 2))
    #1.15
  
    meta$bacon_P = bacon_P
    meta = meta %>%
      mutate(fdr=p.adjust(meta$`P-value`, method = "fdr", n=length(meta$`P-value`))) %>%
      mutate(bonferroni=p.adjust(meta$`P-value`, method = "bonferroni", n=length(meta$`P-value`))) 

    0.05/nrow(meta) # bonferroni threshold
    # 6.628399e-08
   
  }
      
  metaNG = fread(paste0("meta.",platform,".bw.Nogene.txt")) %>%
    dplyr::select(-Allele1, -Allele2) %>%
    dplyr::filter(grepl('cg', MarkerName)) %>%
    dplyr::rename("CpG" = "MarkerName") %>%
    left_join(annoFile,by=c("CpG"="cpg")) %>%
    dplyr::filter(Probe_maf<0.05|is.na(Probe_maf))%>%
    dplyr::filter(CpG_maf<0.05|is.na(CpG_maf)) %>%
    dplyr::filter(chr!="chrX" & chr!="chrY")%>%
    mutate(`P-value`=as.numeric(`P-value`))%>%
    dplyr::filter(!is.na(`P-value`))%>%
    arrange(by=`P-value`) %>%
    mutate(tss=0)
  metaNG$tss[grep("TSS",metaNG$UCSC_RefGene_Group)]=1
  
  bacon_P = pval(bacon((metaNG$Effect)/(metaNG$StdErr),na.exclude = TRUE))

  # raw P values
  Nbeta <- sum(!is.na(metaNG$`P-value`))
  tmp <- metaNG[!is.na(metaNG$`P-value`),]
  lambda = median(qchisq(1- tmp$`P-value`,1))/qchisq(0.5,1)
  print(round(lambda, 2))
  #2.77 for 450K
  #1.6 for EPIC
  
  # bacon P values
  Nbeta <- sum(!is.na(bacon_P))
  tmp <- bacon_P[!is.na(bacon_P)]
  lambda = median(qchisq(1- bacon_P,1))/qchisq(0.5,1)
  print(round(lambda, 2))
  #1.4 for 450K
  #1.14 for EPIC
  
  metaNG$bacon_P = bacon_P
  metaNG = metaNG %>%
    mutate(fdr=p.adjust(metaNG$bacon_P, method = "fdr", n=length(metaNG$bacon_P))) %>%
    mutate(bonferroni=p.adjust(metaNG$bacon_P, method = "bonferroni", n=length(metaNG$bacon_P))) 

  0.05/nrow(metaNG) # bonferroni threshold
  # 1.194515e-07 for 450K
  # 6.628399e-08 for EPIC
  
  write.csv(metaNG,paste0("bw.Nomqtl.",platform,".meta.csv"))
  write.csv(meta,paste0("bw.mqtl.",platform,".meta.csv"))

  meta.sig = meta%>%filter(bonferroni<0.05)
  metaNG.sig = metaNG%>%filter(bonferroni<0.05)
  
  write.csv(metaNG.sig,paste0("bw.sig.Nomqtl.",platform,".meta.csv"))
  write.csv(meta.sig,paste0("bw.sig.mqtl.",platform,".meta.csv"))

  print(paste0("Results of ", platform))
  print(paste0(nrow(meta.sig)," probes are significant after SNP correction for platform ",platform))
  
  if(platform=="450K"){
  a1=sum(meta.sig$set1_snp!="none")
  a2=sum(meta.sig$set2_snp!="none")    
  }

  if(platform=="EPIC"){
  a1=sum(meta.sig$set3_snp!="none" & (!is.na(meta.sig$set4_snp)))
  a2=sum(meta.sig$set4_snp!="none" & (!is.na(meta.sig$set4_snp)))
  
  # number of probes not on 450K
  sum(meta.sig$Methyl450_Loci!="TRUE")
  
  }
  
  preGold = length(which(meta.sig$CpG %in% gold$CpG))
  novelCpGs = nrow(meta.sig) - preGold
  print(paste0(preGold, " of CpGs were previously reported to be associated with birthweight"))
  print(paste0(novelCpGs, " of CpGs were novel CpGS with birthweight"))
  
  
  print(paste0(a1," (or ",a2,") probes are matched to a significant mQTL for platform ",platform))

  print("CpGs that would not have been identified if not corrected for mQTLs")
  meta.sig[which(!meta.sig$CpG %in% metaNG.sig$CpG),]
  print(nrow(meta.sig[which(!meta.sig$CpG %in% metaNG.sig$CpG),]))
            
  print(paste0("distributiono of TSS",platform,", with mQTL"))
  print(table(meta.sig$tss))
  print(table(meta.sig$tss)[2]/(table(meta.sig$tss)[1]+table(meta.sig$tss)[2]))
  
  print(paste0("distributiono of TSS",platform,", without mQTL"))
  print(table(metaNG.sig$tss))
  print(table(metaNG.sig$tss)[2]/(table(metaNG.sig$tss)[1]+table(metaNG.sig$tss)[2]))
  
  }

#[1] 1.3
#[1] 1.03
#[1] 2.77
#[1] 1.4
#[1] "Results of 450K"
#[1] "33 probes are significant after SNP correction for platform 450K"
#[1] "19 of CpGs were previously reported to be associated with birthweight"
#[1] "14 of CpGs were novel CpGS with birthweight"
#[1] "15 (or 15) probes are matched to a significant mQTL for platform 450K"
#[1] "CpGs that would not have been identified if not corrected for mQTLs"
#[1] 33
#[1] "distributiono of TSS450K, with mQTL"
# 0  1 
#30  3 
#         1 
#0.09090909 
#[1] "distributiono of TSS450K, without mQTL"
#0 1 
#2 1 
#1 
#0.3333333 

#[1] 1.64
#[1] 1.15
#[1] 1.6
#[1] 1.13
#[1] "3294 probes are significant after SNP correction for platform EPIC"
#[1] "266 of CpGs were previously reported to be associated with birthweight"
#[1] "3028 of CpGs were novel CpGS with birthweight"
#[1] "2004 (or 2004) probes are matched to a significant mQTL for platform EPIC"
#[1] "CpGs that would not have been identified if not corrected for mQTLs"
#[1] 2216
#[1] "distributiono of TSSEPIC, with mQTL"
#   0    1 
#2869  425 
#        1 
#0.1290225 
#[1] "distributiono of TSSEPIC, without mQTL"
#  0   1 
#958 121 
#        1 
#0.1121409 



#########################################################################################
#top Hits with SNP control from Boydyweight EWAS and results of GSEA pathway analysis, both platforms

#platform 450K
platform="450K"
meta = fread(paste0("bw.mqtl.",platform,".meta.csv"))%>%
  filter(bonferroni < 0.05) 

meta %>% 
  filter(UCSC_RefGene_Name!="") %>% 
  mutate(geneName=gsub(";.*$","",UCSC_RefGene_Name)) %>% 
  select(geneName)

#NIN
#LRCH1
#TMEM30A
#EPSTI1
#PLD2
#TGFB2
#C3orf67
#ARID3A
#SLC9A3R2
#ATP8B1
#TP53INP1
#GABBR1
#ARID5B
#PSMB9
#CCDC33
#ATP6V0A1
#LY6G6E
#MCF2L
#GNG7
#GCNT2
#C10orf140
#RAI1
#TRAF5
#IGF2BP1
#SCHIP1
#LOC100128542
# 'Early-TGFB1 signature': genes overexpressed in primary hepatocytes at an early phase of TGFB1 [GeneID=7040] treatment; is associated with a less invasive phenotype. (fdr 1.92 e-2)

#platform EPIC
platform="EPIC"
meta = fread(paste0("bw.mqtl.",platform,".meta.csv"))%>%
  filter(bonferroni < 0.05) 

meta %>% 
  filter(UCSC_RefGene_Name!="") %>% 
  mutate(geneName=gsub(";.*$","",UCSC_RefGene_Name)) %>% 
  select(geneName) %>%
  unique() %>%
  head(50)

#TMEM30A
#EXOSC10
#ITGA9-AS1
#CCR6
#ARID5B
#NDUFA13
#DNTT
#MSI2
#ARID3A
#ST3GAL6
#RPL39L
#MED15
#SLC25A13
#HECA
#STARD13
#TBL1XR1
#VPS45
#RARA
#FBLN7
#DNASE1L3
#SMG7
#TRAF5
#TG
#TOX
#IGF2BP1
#SETBP1
#VWF
#MMRN1
#ARHGEF7
#NRP1
#ITGB2
#SLC8A1-AS1
#SLC9A3R2
#ENPP2
#PLD2
#IL21R
#RAD51B
#DFNB31
#PTH2R
#TGFB2
#PYGM
#COBL
#APOBEC3A
#THEMIS2
#KLF7
#RAPGEF3
#MLLT3
#XYLT1
#ANKRD11
#BSN

# GSEA: 
#The process whose specific outcome is the progression of a tissue over time, from its formation to the mature structure. [ISBN:0471245208] (fdr q value = 3.87 e-4)
# Abnormal inflammatory response	(fdr q value = 3.87 e-4)

##################################################################################
###### comparing SNP corrected vs none

## without regression line
### platform="450K" is used in the manuscript
for(platform in c("450K","EPIC")){
  if(platform=="450K"){
  meta = fread(paste0("bw.mqtl.",platform,".meta.csv"))%>%
    dplyr::rename("effect_mqtl"="Effect","p_mqtl"="P-value","direction_mqtl"="Direction","mqtl_bonferroni"="bonferroni") %>%
    dplyr::select(effect_mqtl,p_mqtl,mqtl_bonferroni,direction_mqtl,CpG,set1_snp,set2_snp) %>%
    dplyr::filter(!(set1_snp=="none"&set2_snp=="none"))
  }
  if(platform=="EPIC"){
  meta = fread(paste0("bw.mqtl.",platform,".meta.csv"))%>%
    dplyr::rename("effect_mqtl"="Effect","p_mqtl"="P-value","direction_mqtl"="Direction","mqtl_bonferroni"="bonferroni") %>%
    dplyr::select(effect_mqtl,p_mqtl,mqtl_bonferroni,direction_mqtl,CpG,set3_snp,set4_snp)%>%
    dplyr::filter(!(set3_snp=="none"&set4_snp=="none"))
  }  
  
  metaNG = fread(paste0("bw.Nomqtl.",platform,".meta.csv"))%>%
    dplyr::rename("effect_nomqtl"="Effect","p_nomqtl"="P-value","direction_nomqtl"="Direction",
                  "nomqtl_bonferroni"="bonferroni") %>%
    dplyr::select(CpG,effect_nomqtl,p_nomqtl,nomqtl_bonferroni,direction_nomqtl)

  # select probes that are significant either with or without SNP correction
  metaCompare = meta %>%
    inner_join(metaNG,by="CpG")%>%
    mutate(sigBi = factor(ifelse(mqtl_bonferroni < 0.05,"sig","non_sig"))) %>%
    mutate(logP_nomqtl=-log10(p_nomqtl),logP_mqtl=-log10(p_mqtl))
  nrow(metaCompare)

  ###############
  ggplot(metaCompare)+
    geom_point(data = subset(metaCompare,sigBi == 'non_sig'),
             aes(x = effect_nomqtl, y = effect_mqtl), alpha = 1/8,color='turquoise')+
    geom_point(data = subset(metaCompare,sigBi == 'sig'),
             aes(x = effect_nomqtl, y = effect_mqtl), alpha = 0.8,color='salmon')+
    geom_abline(aes(intercept=0,slope=1),alpha=0.4,color="red")+
    geom_hline(aes(yintercept=0),color="red",alpha=0.4)+
    geom_vline(aes(xintercept=0),color="red",alpha=0.4)+
    xlab("regression coefficient values, without mQTL") + ylab("regression coefficient values, with mQTL")+
    theme_bw()+
    theme(axis.text.x = element_text(size=12),
          axis.text.y = element_text(size=12),
          axis.title.x = element_text(size=14),
          axis.title.y = element_text(size=14),
          plot.margin=unit(c(5.5, 19, 5.5, 5.5), "points"))
  ggsave(paste0(platform,".effectComp.png"),width = 8.5,height = 6)
  
  ###############
  ggplot(metaCompare)+
    geom_point(data = subset(metaCompare,sigBi == 'non_sig'),
             aes(x = logP_nomqtl, y = logP_mqtl), alpha = 1/8,color='turquoise')+
    geom_point(data = subset(metaCompare,sigBi == 'sig'),
             aes(x = logP_nomqtl, y = logP_mqtl), alpha = 1,color='salmon')+
    geom_abline(aes(intercept=0,slope=1),alpha=0.4,color="red")+
    xlab("竏値og10 (P-value), without mQTL") + ylab("竏値og10 (P-value), with mQTL")+
    theme_bw()+
    theme(axis.text.x = element_text(size=12),
          axis.text.y = element_text(size=12),
          axis.title.x = element_text(size=14),
          axis.title.y = element_text(size=14))
  ggsave(paste0(platform,".pCompAll.png"),width = 8,height = 6)
}

## with regression line
### platform="EPIC" is used in the manuscript
for(platform in c("450K","EPIC")){

  if(platform=="450K"){
  meta = fread(paste0("bw.mqtl.",platform,".meta.csv"))%>%
    dplyr::rename("effect_mqtl"="Effect","p_mqtl"="P-value","direction_mqtl"="Direction","mqtl_bonferroni"="bonferroni") %>%
    dplyr::select(effect_mqtl,p_mqtl,mqtl_bonferroni,direction_mqtl,CpG,set1_snp,set2_snp) %>%
    dplyr::filter(!(set1_snp=="none"&set2_snp=="none"))
  }

  if(platform=="EPIC"){
  meta = fread(paste0("bw.mqtl.",platform,".meta.csv"))%>%
    dplyr::rename("effect_mqtl"="Effect","p_mqtl"="P-value","direction_mqtl"="Direction","mqtl_bonferroni"="bonferroni") %>%
    dplyr::select(effect_mqtl,p_mqtl,mqtl_bonferroni,direction_mqtl,CpG,set3_snp,set4_snp)%>%
    dplyr::filter(!(set3_snp=="none"&set4_snp=="none"))
  }  
  
  #meta = meta %>%filter(!grepl("\\?",direction_mqtl)) %>%filter(p_mqtl>0)
  
  metaNG = fread(paste0("bw.Nomqtl.",platform,".meta.csv"))%>%
    dplyr::rename("effect_nomqtl"="Effect","p_nomqtl"="P-value","direction_nomqtl"="Direction",
                  "nomqtl_bonferroni"="bonferroni") %>%
    dplyr::select(CpG,effect_nomqtl,p_nomqtl,nomqtl_bonferroni,direction_nomqtl)

  #metaNG = metaNG %>%filter(!grepl("\\?",direction_nomqtl)) %>%filter(p_nomqtl>0)
  
  # select probes that are significant either with or without SNP correction
  metaCompare = meta %>%
    inner_join(metaNG,by="CpG")%>%
    mutate(sigBi = factor(ifelse(mqtl_bonferroni < 0.05,"sig","non_sig"))) %>%
    mutate(logP_nomqtl=-log10(p_nomqtl),logP_mqtl=-log10(p_mqtl))
  nrow(metaCompare)
  # 123012 for 450K
  # 296450 for EPIC
  
  # metaCompareSig = metaCompare %>%
  #   filter(mqtl_bonferroni < 0.05 | nomqtl_bonferroni<0.05) 
  
  ###############
  max = max(c(metaCompare$effect_nomqtl,metaCompare$effect_mqtl))
  min = min(c(metaCompare$effect_nomqtl,metaCompare$effect_mqtl))
  
  ggplot(metaCompare)+
    geom_point(data = subset(metaCompare,sigBi == 'non_sig'),
             aes(x = effect_nomqtl, y = effect_mqtl), alpha = 1/8,color='turquoise')+
    geom_point(data = subset(metaCompare,sigBi == 'sig'),
             aes(x = effect_nomqtl, y = effect_mqtl), alpha = 0.8,color='salmon')+
    geom_abline(aes(intercept=0,slope=1),alpha=0.4,color="red")+
    geom_hline(aes(yintercept=0),color="red",alpha=0.4)+
    geom_vline(aes(xintercept=0),color="red",alpha=0.4)+
    xlim(min,max)+ylim(min,max)+
    stat_smooth(aes(x=effect_nomqtl,y=effect_mqtl),
                geom = 'line',method = "lm",se=FALSE,color='blue',alpha=0.3)+
    xlab("regression coefficient values, without mQTL") + ylab("regression coefficient values, with mQTL")+
    theme_bw()+
    theme(axis.text.x = element_text(size=12),
          axis.text.y = element_text(size=12),
          axis.title.x = element_text(size=14),
          axis.title.y = element_text(size=14),
          plot.margin=unit(c(5.5, 19, 5.5, 5.5), "points"))

  ggsave(paste0(platform,".effectComp.png"),width = 8.5,height = 6)
  
  ###############
  max = max(c(-log10(metaCompare$p_nomqtl),-log10(metaCompare$p_mqtl)))
  min = min(c(-log10(metaCompare$p_nomqtl),-log10(metaCompare$p_mqtl)))
  
  ggplot(metaCompare)+
    geom_point(data = subset(metaCompare,sigBi == 'non_sig'),
             aes(x = logP_nomqtl, y = logP_mqtl), alpha = 1/8,color='turquoise')+
    geom_point(data = subset(metaCompare,sigBi == 'sig'),
             aes(x = logP_nomqtl, y = logP_mqtl), alpha = 1,color='salmon')+
    geom_abline(aes(intercept=0,slope=1),alpha=0.4,color="red")+
      stat_smooth(aes(x=logP_nomqtl,y=logP_mqtl),
                  geom = 'line',method = "lm",se=FALSE,color='blue',alpha=0.3)+
    xlab("竏値og10 (P-value), without mQTL") + ylab("竏値og10 (P-value), with mQTL")+
    theme_bw()+
    theme(axis.text.x = element_text(size=12),
          axis.text.y = element_text(size=12),
          axis.title.x = element_text(size=14),
          axis.title.y = element_text(size=14))
  ggsave(paste0(platform,".pCompAll.png"),width = 8,height = 6)
  
  # reviwer question: What is the level of significance? Please clarify.
  metaCompare_nonsig = metaCompare %>%
    filter(sigBi=="non_sig") %>%
    mutate(diff = logP_nomqtl- logP_mqtl)
  #mean(metaCompare_nonsig$diff)
  #median(metaCompare_nonsig$diff)
  table(metaCompare_nonsig$diff>0)
}

#### suggestion from Joe
# USEFUL_regs = c("Gene_Associated","Gene_Associated_Cell_type_specific",
#                 "Promoter_Associated","Promoter_Associated_Cell_type_specific",
#                 "NonGene_Associated","NonGene_Associated_Cell_type_specific")
USEFUL_regs = c("Gene_Associated","Gene_Associated_Cell_type_specific",
                 "Promoter_Associated","Promoter_Associated_Cell_type_specific")

## 1. check the regulatory regions (that are )CPG island, shores, shelves (and if there is any other annotation) on EPIC array that are NOT on the 450K and see if the trend towards increased significance is stronger 
platform = "EPIC"
meta = fread(paste0("bw.mqtl.",platform,".meta.csv"))%>%
  dplyr::rename("effect_mqtl"="Effect","p_mqtl"="P-value","direction_mqtl"="Direction","mqtl_bonferroni"="bonferroni") %>%
  dplyr::select(effect_mqtl,p_mqtl,mqtl_bonferroni,direction_mqtl,CpG,set3_snp,set4_snp)%>%
  dplyr::filter(!(set3_snp=="none"&set4_snp=="none"))

metaNG = fread(paste0("bw.Nomqtl.",platform,".meta.csv"))%>%
  dplyr::rename("effect_nomqtl"="Effect","p_nomqtl"="P-value",
                "direction_nomqtl"="Direction",
                "nomqtl_bonferroni"="bonferroni") %>%
  dplyr::select(CpG,effect_nomqtl,p_nomqtl,nomqtl_bonferroni,direction_nomqtl)

metaCompare = meta %>%
  inner_join(metaNG,by="CpG")%>%
  mutate(sigBi = factor(ifelse(mqtl_bonferroni < 0.05,"sig","non_sig"))) %>%
  mutate(logP_nomqtl=-log10(p_nomqtl),logP_mqtl=-log10(p_mqtl)) %>%
  inner_join(annoEPIC,by=c("CpG"="cpg")) %>%
  mutate(epic_reg = ifelse( (Methyl450_Loci!="TRUE")&(Regulatory_Feature_Group %in% USEFUL_regs),1,0 ))
ggplot(metaCompare)+
  geom_point(data = subset(metaCompare,epic_reg == 0 ),
           aes(x = logP_nomqtl, y = logP_mqtl), alpha = 1/8,color='salmon')+
  geom_point(data = subset(metaCompare,epic_reg == 1),
           aes(x = logP_nomqtl, y = logP_mqtl), alpha = 1, shape= 4, color='turquoise')+
  geom_abline(aes(intercept=0,slope=1),alpha=0.4,color="red")+
  xlab("竏値og10 (P-value),, without mQTL") + ylab("竏値og10 (P-value),, with mQTL")+
  theme_bw()+
  theme(axis.text.x = element_text(size=12),
        axis.text.y = element_text(size=12),
        axis.title.x = element_text(size=14),
        axis.title.y = element_text(size=14))
ggsave(paste0(platform,".pCompAll.joe1.png"),width = 8,height = 6)

## 2. take regulatory regions in the 450K and see if those CpGs as a subset get stronger P values
platform = "450K"
meta = fread(paste0("bw.mqtl.",platform,".meta.csv"))%>%
  dplyr::rename("effect_mqtl"="Effect","p_mqtl"="P-value","direction_mqtl"="Direction","mqtl_bonferroni"="bonferroni") %>%
  dplyr::select(effect_mqtl,p_mqtl,mqtl_bonferroni,direction_mqtl,CpG,set1_snp,set2_snp) %>%
  dplyr::filter(!(set1_snp=="none"&set2_snp=="none"))
metaNG = fread(paste0("bw.Nomqtl.",platform,".meta.csv"))%>%
  dplyr::rename("effect_nomqtl"="Effect","p_nomqtl"="P-value","direction_nomqtl"="Direction",
                "nomqtl_bonferroni"="bonferroni") %>%
  dplyr::select(CpG,effect_nomqtl,p_nomqtl,nomqtl_bonferroni,direction_nomqtl)
metaCompare = meta %>%
  inner_join(metaNG,by="CpG")%>%
  mutate(sigBi = factor(ifelse(mqtl_bonferroni < 0.05,"sig","non_sig"))) %>%
  mutate(logP_nomqtl=-log10(p_nomqtl),logP_mqtl=-log10(p_mqtl)) %>%
  inner_join(anno,by=c("CpG"="cpg")) %>%
  mutate(reg = ifelse( Regulatory_Feature_Group %in% USEFUL_regs,1,0 ))
ggplot(metaCompare)+
  geom_point(data = subset(metaCompare,reg == 0 ),
           aes(x = logP_nomqtl, y = logP_mqtl), alpha = 1/8,color='salmon')+
  geom_point(data = subset(metaCompare,reg == 1),
           aes(x = logP_nomqtl, y = logP_mqtl), alpha = 1, shape= 4, color='turquoise')+
  geom_abline(aes(intercept=0,slope=1),alpha=0.4,color="red")+
  xlab("竏値og10 (P-value), without mQTL") + ylab("竏値og10 (P-value), with mQTL")+
  theme_bw()+
  theme(axis.text.x = element_text(size=12),
        axis.text.y = element_text(size=12),
        axis.title.x = element_text(size=14),
        axis.title.y = element_text(size=14))
ggsave(paste0(platform,".pCompAll.joe2.png"),width = 8,height = 6)

### 3. 450K regulatory CpGs that are on the EPIC array
platform = "EPIC"
meta = fread(paste0("bw.mqtl.",platform,".meta.csv"))%>%
  dplyr::rename("effect_mqtl"="Effect","p_mqtl"="P-value","direction_mqtl"="Direction","mqtl_bonferroni"="bonferroni") %>%
  dplyr::select(effect_mqtl,p_mqtl,mqtl_bonferroni,direction_mqtl,CpG,set3_snp,set4_snp)%>%
  dplyr::filter(!(set3_snp=="none"&set4_snp=="none"))

metaNG = fread(paste0("bw.Nomqtl.",platform,".meta.csv"))%>%
  dplyr::rename("effect_nomqtl"="Effect","p_nomqtl"="P-value",
                "direction_nomqtl"="Direction",
                "nomqtl_bonferroni"="bonferroni") %>%
  dplyr::select(CpG,effect_nomqtl,p_nomqtl,nomqtl_bonferroni,direction_nomqtl)

metaCompare = meta %>%
  inner_join(metaNG,by="CpG")%>%
  mutate(sigBi = factor(ifelse(mqtl_bonferroni < 0.05,"sig","non_sig"))) %>%
  mutate(logP_nomqtl=-log10(p_nomqtl),logP_mqtl=-log10(p_mqtl)) %>%
  inner_join(annoEPIC,by=c("CpG"="cpg")) %>%
  mutate(epic_reg = ifelse( (Methyl450_Loci=="TRUE")&(Regulatory_Feature_Group %in% USEFUL_regs),1,0 ))
ggplot(metaCompare)+
  geom_point(data = subset(metaCompare,epic_reg == 0 ),
           aes(x = logP_nomqtl, y = logP_mqtl), alpha = 1/8,color='salmon')+
  geom_point(data = subset(metaCompare,epic_reg == 1),
           aes(x = logP_nomqtl, y = logP_mqtl), alpha = 1, shape= 4, color='turquoise')+
  geom_abline(aes(intercept=0,slope=1),alpha=0.4,color="red")+
  xlab("竏値og10 (P-value),, without mQTL") + ylab("竏値og10 (P-value),, with mQTL")+
  theme_bw()+
  theme(axis.text.x = element_text(size=12),
        axis.text.y = element_text(size=12),
        axis.title.x = element_text(size=14),
        axis.title.y = element_text(size=14))
ggsave(paste0(platform,".pCompAll.joe3.png"),width = 8,height = 6)

### 4. narrow the EPIC down to 450K corner to see if it's the same trend
#### meta preparation is the same as 3
metaCompare = meta %>%
  inner_join(metaNG,by="CpG")%>%
  mutate(sigBi = factor(ifelse(mqtl_bonferroni < 0.05,"sig","non_sig"))) %>%
  mutate(logP_nomqtl=-log10(p_nomqtl),logP_mqtl=-log10(p_mqtl)) %>%
  inner_join(annoEPIC,by=c("CpG"="cpg")) %>%
  filter(logP_mqtl<=15) 

ggplot(metaCompare)+
  geom_point(data = subset(metaCompare,sigBi == "sig" ),
           aes(x = logP_nomqtl, y = logP_mqtl), alpha = 1/8,color='salmon')+
  geom_point(data = subset(metaCompare,sigBi == "non_sig"),
           aes(x = logP_nomqtl, y = logP_mqtl), alpha = 1, shape= 4, color='turquoise')+
  geom_abline(aes(intercept=0,slope=1),alpha=0.4,color="red")+
  xlab("竏値og10 (P-value),, without mQTL") + ylab("竏値og10 (P-value),, with mQTL")+
  theme_bw()+
  theme(axis.text.x = element_text(size=12),
        axis.text.y = element_text(size=12),
        axis.title.x = element_text(size=14),
        axis.title.y = element_text(size=14))
ggsave(paste0(platform,".pCompAll.joe4.png"),width = 8,height = 6)

### 5. only look at set 4 results
set=4
var = "ch_birthwt_bc_full"

results =fread(paste0("ccls.set",set,".",var,".geneCon.ewas.txt")) %>%
  inner_join(annoEPIC,by=c("CpG"="cpg")) %>%
  dplyr::filter(grepl('cg', CpG)) %>%
  dplyr::filter(Probe_maf<0.05|is.na(Probe_maf))%>%
  dplyr::filter(CpG_maf<0.05|is.na(CpG_maf)) %>%
  dplyr::filter(chr!="chrX" & chr!="chrY") 
results$bonferroni = p.adjust(results$P, method = "bonferroni",
                              n=length(results$P))
results = results %>%
  filter(snp!='none') %>%
  dplyr::rename("effect_mqtl"="Beta",
                "p_mqtl"="P",
                "mqtl_bonferroni"="bonferroni")

results_NG = fread(paste0("ccls.set",set,".",var,".Nogene.ewas.txt")) %>%
  inner_join(annoEPIC,by=c("CpG"="cpg")) %>%
  dplyr::filter(grepl('cg', CpG)) %>%
  dplyr::filter(Probe_maf<0.05|is.na(Probe_maf))%>%
  dplyr::filter(CpG_maf<0.05|is.na(CpG_maf)) %>%
  dplyr::filter(chr!="chrX" & chr!="chrY") 
results_NG$bonferroni = p.adjust(results_NG$P, method = "bonferroni",
                              n=length(results_NG$P))
results_NG = results_NG %>%
  dplyr::rename("effect_nomqtl"="Beta",
                "p_nomqtl"="P",
                "nomqtl_bonferroni"="bonferroni")

resCompare = results %>%
  inner_join(results_NG,by="CpG")%>%
  mutate(sigBi = factor(ifelse(mqtl_bonferroni < 0.05,"sig","non_sig"))) %>%
  mutate(logP_nomqtl=-log10(p_nomqtl),
         logP_mqtl=-log10(p_mqtl)) %>%
  inner_join(annoEPIC,by=c("CpG"="cpg")) 

ggplot(resCompare)+
  geom_point(data = subset(resCompare,sigBi == "sig" ),
           aes(x = logP_nomqtl, y = logP_mqtl), alpha = 1,color='salmon')+
  geom_point(data = subset(resCompare,sigBi == "non_sig"),
           aes(x = logP_nomqtl, y = logP_mqtl), alpha = 0.3, shape= 4, color='turquoise')+
  geom_abline(aes(intercept=0,slope=1),alpha=0.4,color="red")+
  xlab("竏値og10 (P-value), without mQTL") + ylab("竏値og10 (P-value), with mQTL")+
  theme_bw()+
  theme(axis.text.x = element_text(size=12),
        axis.text.y = element_text(size=12),
        axis.title.x = element_text(size=14),
        axis.title.y = element_text(size=14))
ggsave(paste0(platform,".pCompAll.joe5.png"),width = 8,height = 6)

#################################################################################
###################  (gene*birthweight interaction)

setwd("~/mQTL/birthweight")
library(tidyverse)
library(data.table)
library(qvalue)
library(IlluminaHumanMethylation450kanno.ilmn12.hg19)
library(IlluminaHumanMethylationEPICanno.ilm10b4.hg19)
data("IlluminaHumanMethylation450kanno.ilmn12.hg19")
anno       <- IlluminaHumanMethylation450kanno.ilmn12.hg19 %>%
  getAnnotation %>%
  as.data.frame %>%
  rownames_to_column(var="cpg") %>%
  dplyr::select(cpg,chr,pos,Relation_to_Island,UCSC_RefGene_Name,UCSC_RefGene_Group,Probe_maf)

library(IlluminaHumanMethylationEPICanno.ilm10b4.hg19)
data("IlluminaHumanMethylationEPICanno.ilm10b4.hg19")
annoEPIC       <- IlluminaHumanMethylationEPICanno.ilm10b4.hg19 %>%
  getAnnotation %>%
  as.data.frame %>%
  rownames_to_column(var="cpg") %>%
  dplyr::select(cpg,chr,pos,Relation_to_Island,UCSC_RefGene_Name,UCSC_RefGene_Group,Methyl450_Loci,Probe_maf)

################### Pathway analysis of probes whose beta changes 
library(methylGSA)

for(platform in c("450K","EPIC")){
  
  if(platform=="450K"){
 meta = fread(paste0("bw.mqtl.",platform,".meta.csv"))%>%
    dplyr::rename("effect_mqtl"="Effect","p_mqtl"="P-value","direction_mqtl"="Direction",
                  "mqtl_fdr"="fdr") %>%
    dplyr::select(CpG,set1_snp,set2_snp,effect_mqtl,p_mqtl,mqtl_fdr,direction_mqtl)%>%
    dplyr::filter(!(set1_snp=="none"&set2_snp=="none"))
  }
  
  if(platform=="EPIC"){
 meta = fread(paste0("bw.mqtl.",platform,".meta.csv"))%>%
    dplyr::rename("effect_mqtl"="Effect","p_mqtl"="P-value","direction_mqtl"="Direction",
                  "mqtl_fdr"="fdr") %>%
    dplyr::select(CpG,set3_snp,set4_snp,effect_mqtl,p_mqtl,mqtl_fdr,direction_mqtl)%>%
    dplyr::filter(!(set3_snp=="none"&set4_snp=="none"))
  }
 
 metaNG = fread(paste0("bw.Nomqtl.",platform,".meta.csv"))%>%
    dplyr::rename("effect_nomqtl"="Effect","p_nomqtl"="P-value","direction_nomqtl"="Direction",
                  "nomqtl_fdr"="fdr") %>%
    dplyr::select(CpG,effect_nomqtl,p_nomqtl,nomqtl_fdr,direction_nomqtl)
 
  # select probes that are significant either with or without SNP correction
   metaCompare = meta %>%
    inner_join(metaNG,by="CpG")%>%
    mutate(delta=abs((effect_mqtl-effect_nomqtl)/effect_nomqtl))%>%
    distinct(CpG,.keep_all = TRUE) %>%
    filter(mqtl_fdr< 0.05)%>%
    filter(delta>0.2) %>%
    arrange(mqtl_fdr)%>%
    dplyr::select(CpG,p_mqtl) 
  # metaCompare = metaCompare[1:1000,] %>% na.omit()
  nrow(metaCompare)
  
  # take a look at the direction of change
   DC = meta %>%
    inner_join(metaNG,by="CpG")%>%
    mutate(delta=abs((effect_mqtl-effect_nomqtl)/effect_nomqtl))%>%
    mutate(product = effect_nomqtl*effect_mqtl) %>% 
    mutate(absDiff = abs(effect_mqtl)-abs(effect_nomqtl)) %>% 
    distinct(CpG,.keep_all = TRUE) %>%
    filter(mqtl_fdr< 0.05)%>%
    filter(delta>0.2) %>%
    arrange(mqtl_fdr)
   
   print(platform)
   print("same direction?")
   print(table(DC$product>0))
   print("same direction and stronger?")
   print(table(DC[DC$product>0 ,'absDiff']>0))
   
  #[1] "450K"
  #[1] "same direction?"
  # FALSE  TRUE 
  #   1    92 
  #[1] "same direction and stronger?"
  # FALSE  TRUE 
  # 21    71 
   
  #[1] "EPIC"
  #[1] "same direction?"
  # FALSE  TRUE 
  # 24  1687 
  #[1] "same direction and stronger?"
  # FALSE  TRUE 
  # 571  1116  
     
  cpg.pval = c(metaCompare$p_mqtl)
  # P=0 is not accepted. so replace 0 into min/100
  if(length(which(cpg.pval==0))!=0){
    cpg.pval[which(cpg.pval==0)] = min(cpg.pval[-which(cpg.pval==0)])/100
  }
  names(cpg.pval) = metaCompare$CpG
  
  if(platform=="450K"){
  res1 = methylglm(cpg.pval = cpg.pval, minsize = 50, GS.type = "KEGG")
  res2 = methylglm(cpg.pval = cpg.pval, minsize = 50, GS.type = "GO")
  }
  
  if(platform=="EPIC"){
  res1 = methylglm(cpg.pval = cpg.pval, minsize = 50, GS.type = "KEGG",array.type = "EPIC")

  res2 = methylglm(cpg.pval = cpg.pval, minsize = 50, GS.type = "GO", array.type = "EPIC")
   }
  
  write.table(res1,paste0(platform,".interaction.kegg.txt"),
              sep = "\t", 
              row.names = FALSE, 
              quote = FALSE)
  write.table(res2,paste0(platform,".interaction.go.txt"),
              sep = "\t", 
              row.names = FALSE, 
              quote = FALSE)
  
  
  if(platform=="450K"){
    metaInt = metaCompare %>% left_join(anno,by=c("CpG"="cpg")) 
  }else{
    metaInt = metaCompare %>% left_join(annoEPIC,by=c("CpG"="cpg"))
  }
  
  write.table(metaInt,paste0(platform,".interaction.topGenes.txt"),sep = "\t", row.names = FALSE, quote = FALSE)
  
}  
######## Interaction terms P value summary
gene.linear.interaction <- function(meth,cov,gene){
  
  probe <- unname(meth[1])
  methDF = meth[-1] %>% as.numeric()
  SNPname <- gene[which(gene$CpG == probe), 1]  %>% as.character()  
  SNP <-  gene[which(gene$CpG == probe), -c(1:3)] %>% t() %>% as.data.frame()
  colnames(SNP) <- "SNP"
  
  dat <- data.frame(y = methDF, cov,SNP)
  ff  <- formula(paste("y ~ birthWt*SNP+", paste0(c(colnames(cov),"SNP"),collapse="+")))
  mod <- try(lm(ff, data = dat),silent = TRUE)
  N <- sum(!is.na(dat[,1]))
  
  if (class(mod)=="lm") {
    SNPname=SNPname
    coefs = summary(mod)$coefficients
    coef = coefs[2,"Estimate"]
    SE = coefs[2,"Std. Error"]
    p.value = coefs[2,"Pr(>|t|)"]
    out = c(SNPname,coef, SE, p.value, N) 
  } else {
    out = c(SNPname,rep(NA,3), N)
  }
  names(out) = c("snp","interationBeta","interationSE", "interationP", "N"); return(out) 
}


for(platform in c("450K","EPIC")){

  if(platform=="450K"){
 meta = fread(paste0("bw.mqtl.",platform,".meta.csv"))%>%
    dplyr::rename("effect_mqtl"="Effect","p_mqtl"="P-value","direction_mqtl"="Direction",
                  "mqtl_fdr"="fdr") %>%
    dplyr::select(CpG,set1_snp,set2_snp,effect_mqtl,p_mqtl,mqtl_fdr,direction_mqtl)%>%
    dplyr::filter(!(set1_snp=="none"&set2_snp=="none"))
  }
  
  if(platform=="EPIC"){
 meta = fread(paste0("bw.mqtl.",platform,".meta.csv"))%>%
    dplyr::rename("effect_mqtl"="Effect","p_mqtl"="P-value","direction_mqtl"="Direction",
                  "mqtl_fdr"="fdr") %>%
    dplyr::select(CpG,set3_snp,set4_snp,effect_mqtl,p_mqtl,mqtl_fdr,direction_mqtl)%>%
    dplyr::filter(!(set3_snp=="none"&set4_snp=="none"))
  }

  metaNG = fread(paste0("bw.Nomqtl.",platform,".meta.csv"))%>%
    dplyr::rename("effect_nomqtl"="Effect","p_nomqtl"="P-value","direction_nomqtl"="Direction",
                  "nomqtl_fdr"="fdr") %>%
    dplyr::select(CpG,effect_nomqtl,p_nomqtl,nomqtl_fdr,direction_nomqtl)
  
   metaCompare = meta %>%
    inner_join(metaNG,by="CpG")%>%
    mutate(delta=abs((effect_mqtl-effect_nomqtl)/effect_nomqtl))%>%
    distinct(CpG,.keep_all = TRUE) %>%
    filter(mqtl_fdr< 0.05)%>%
    filter(delta>0.2) %>%
    arrange(mqtl_fdr)%>%
    dplyr::select(CpG,p_mqtl) 
# nrow(metaCompare)

     if(platform=="450K") sets=c(1,2)
     if(platform=="EPIC") sets=c(3,4)

for (set in sets){
  
  if(set==1){var = "birthWt"}
  if(set==2){var = "birthWt"}
  if(set==3){var = "Y_birthweight"}
  if(set==4){var = "ch_birthwt_bc_full"}
  
  methyl0 = fread(paste0("methSet",set,".txt")) %>%
    column_to_rownames(var="probeId")%>%
    as.data.frame()

  covs0 = fread(paste0("covsSet",set,".txt"))%>%
    as.data.frame()%>%
    dplyr::rename("birthWt"=var)%>%
    column_to_rownames(var="ID") %>%
    mutate(Batch = as.factor(Batch))
  
  mtqls0 = fread(paste0("~/mQTL/ccls.set",set,".mqtls.hd")) %>% 
    as.data.frame()
  
  subs.list = intersect(intersect(rownames(covs0),colnames(methyl0)), colnames(mtqls0)) 
  
  methyl = methyl0[,subs.list]
  covs = covs0[subs.list,]
  mtqls = mtqls0[,c("SNP","CpG","P-value",subs.list)]
  probes = intersect(rownames(methyl)[which(rownames(methyl)%in%mtqls$CpG)],metaCompare$CpG) 
  
  methyl = methyl[probes,] %>% rownames_to_column(var="cpg")
  rownames(methyl) <- methyl$cpg

  results.snp <- t(apply(methyl, 1, gene.linear.interaction, covs, mtqls)) 
  
  results.snp <- as.data.frame(results.snp) %>%
    rownames_to_column(var="CpG")%>%
    mutate(N=as.numeric(N))%>%
    mutate(interationP=as.numeric(interationP))
  
  write.table(results.snp,
              paste0("ccls.set",set,".",var,".interaction.P.ewas.txt"),
              quote=FALSE,
              row.names = FALSE,
              sep='\t')  
  
  results.snp %>%
    ggplot(aes(x=-log(interationP))) +
    geom_density()+
    geom_vline(xintercept = -log(0.05),linetype="dotted")+
    theme_bw()
  ggsave(paste0("ccls.set",set,".",var,".interaction.P.density.png"),
         width = 8,height = 6)

  }}
 
################## 
#######(not included in the manuscript)
#comparing SNP corrected vs none (sig results with interaction)
for(platform in c("450K","EPIC")){

 meta = fread(paste0("bw.mqtl.",platform,".meta.csv"))%>%
    dplyr::rename("effect_mqtl"="Effect","p_mqtl"="P-value","direction_mqtl"="Direction",
                  "mqtl_fdr"="fdr") %>%
    dplyr::select(CpG,effect_mqtl,p_mqtl,mqtl_fdr,direction_mqtl)
 
 metaNG = fread(paste0("bw.Nomqtl.",platform,".meta.csv"))%>%
    dplyr::rename("effect_nomqtl"="Effect","p_nomqtl"="P-value","direction_nomqtl"="Direction",
                  "nomqtl_fdr"="fdr") %>%
    dplyr::select(CpG,effect_nomqtl,p_nomqtl,nomqtl_fdr,direction_nomqtl)
  
  # select probes that are significant either with or without SNP correction
  metaCompare = meta %>%
    inner_join(metaNG,by="CpG")%>%
    mutate(delta=abs((effect_mqtl-effect_nomqtl)/effect_nomqtl))%>%
    distinct(CpG,.keep_all = TRUE) %>%
    filter(mqtl_fdr < 0.05 )%>%
    filter(delta>0.2) %>%
    select(CpG,effect_nomqtl,effect_mqtl,p_nomqtl,p_mqtl) 

  ###############
  max = max(c(metaCompare$effect_nomqtl,metaCompare$effect_mqtl)*1.2)
  min = min(c(metaCompare$effect_nomqtl,metaCompare$effect_mqtl)*1.2)
  
  metaCompare %>%
    ggplot(aes(x=effect_nomqtl,y=effect_mqtl))+
    geom_point()+
    geom_abline(aes(intercept=0,slope=1),alpha=0.4,color="red")+
    geom_hline(aes(yintercept=0),color="red")+geom_vline(aes(xintercept=0),color="red")+
    xlim(min,max)+ylim(min,max)+
    xlab("beta value, without mQTL") + ylab("beta value, with mQTL")+
    ggtitle(paste0("Effects with or without mQTL, sig results with interaction, platform ",platform))+
    theme_bw()+
    theme(plot.title = element_text(hjust = 0.5))
  ggsave(paste0(platform,".interaction.effectCompSig.png"),width = 8,height = 6)
  
  ###############
  #metaCompare = metaCompare %>% filter(p_nomqtl!=0 & p_mqtl!=0)
  max = max(c(-log10(metaCompare$p_nomqtl),-log10(metaCompare$p_mqtl)))*1.1
  min = min(c(-log10(metaCompare$p_nomqtl),-log10(metaCompare$p_mqtl)))*0.9
  
  metaCompare %>%
    mutate(logP_nomqtl=-log10(p_nomqtl),logP_mqtl=-log10(p_mqtl))%>%
    ggplot(aes(x=logP_nomqtl,y=logP_mqtl))+
    geom_point()+
    geom_abline(aes(intercept=0,slope=1),alpha=0.4,color="red")+
    geom_hline(aes(yintercept=0),color="red")+geom_vline(aes(xintercept=0),color="red")+
    xlim(0,max)+ylim(0,max)+
    xlab("log P value, without mQTL") + ylab("log P value, with mQTL")+
    ggtitle(paste0("Significance with or without mQTL, mqtl sig results with interaction, platform ",platform))+
    theme_bw()+
    theme(plot.title = element_text(hjust = 0.5))
  ggsave(paste0(platform,".interaction.pCompSig.png"),width = 8,height = 6)
}

#######((not included in the manuscript)
## comparing with gold standard
for(platform in c("450K","EPIC")){

  meta = fread(paste0("bw.mqtl.",platform,".meta.csv"))
  
  metaNG = fread(paste0("bw.Nomqtl.",platform,".meta.csv"))%>%
    dplyr::rename("effect_noMQTL"="Effect",
                  "P_noMQTL"="P-value",
                  "Direction_noMQTL"="Direction")
  
  gold = read.csv("bw_meta_paper.csv") %>%
    left_join(meta,by="CpG")%>%
    left_join(metaNG,by="CpG")%>%
    na.omit()%>%
    mutate(dis_mqtl = abs(Effect*10^6 -Coef),
           rate_mqtl =`P-value`/`P.value`,
           dis_nomqtl = abs(effect_noMQTL*10^6 -Coef),
           rate_nomqtl =P_noMQTL/`P.value`) %>%
    mutate(p_gold = -log10(`P.value`),
           p_mqtl = -log10(`P-value`),
           p_nomqtl = -log10(P_noMQTL))%>%
    dplyr::select(CpG,Coef,Effect,effect_noMQTL,p_gold,p_mqtl,p_nomqtl) 

 write.csv(gold,paste0("1095.gold.",platform,".meta.csv"))

  max = max(c(-log(gold$p_mqtl),-log(gold$p_nomqtl)))
  min = min(c(-log(gold$p_mqtl),-log(gold$p_nomqtl)))

  gold %>%
    ggplot(aes(x=-log(p_nomqtl), y=-log(p_mqtl)))+
    geom_point()+
    geom_abline(aes(intercept=0,slope=1),alpha=0.4,color="red")+
    xlim(min,max)+ylim(min,max)+
    xlab("log P value, without mQTL") + ylab("log P value, with mQTL")+
    ggtitle(paste0("p value comparison of 1095 probes in ",platform))+
    theme_bw()+
    theme(plot.title = element_text(hjust = 0.5))
  ggsave(paste0(platform,".metapaper.p.pairwise.compare.png"),width = 8,height = 6)

  
  a=cor.test(gold$Coef,gold$Effect)
  b=cor.test(gold$Coef,gold$effect_noMQTL)
  c=cor.test(gold$p_gold,gold$p_mqtl)
  d=cor.test(gold$p_gold,gold$p_nomqtl)

  lab.beta = as_labeller(c( "Effect" = paste0("With mQTL correction
                                              \n correlation=",round(a$estimate,5),
                                              "\n p=",round(a$p.value,5)),
                            "effect_noMQTL" = paste0("Without mQTL correction 
                                              \n correlation=",round(b$estimate,5),
                                              "\n p=",round(b$p.value,5))))
  
  gold %>%
    select(CpG,Coef,Effect,effect_noMQTL)%>%
    reshape::melt(id=c("CpG","Coef"))%>%
    na.omit() %>%
    ggplot(aes(x=Coef,y=value))+
    geom_point()+
    geom_smooth(method = "lm",se=TRUE)+
    facet_wrap(~variable, ncol=2, scales="free_y",
               labeller = lab.beta)+
    ggtitle(paste0("Beta value comparison with gold standard in ",platform))+
    xlab("Beta Values from gold standard") + ylab("Beta Values from CCLS")+
    theme_bw()+
    theme(plot.title = element_text(hjust = 0.5))
  ggsave(paste0(platform,".metapaper.beta.compare.png"),width = 8,height = 6)

  lab.p = as_labeller(c( "p_mqtl" = paste0("With mQTL correction
                                            \n correlation=",round(c$estimate,5),
                                            "\n p=",round(c$p.value,5)),
                          "p_nomqtl" = paste0("Without mQTL correction 
                                            \n correlation=",round(d$estimate,5),
                                            "\n p=",round(d$p.value,5))))
  gold %>%
    select(CpG,p_gold,p_mqtl,p_nomqtl)%>%
    reshape::melt(id=c("CpG","p_gold"))%>%
    na.omit() %>%
    ggplot(aes(x=p_gold,y=value))+
    geom_point()+
    geom_abline(aes(intercept=0,slope=1),alpha=0.4,color="red")+
    facet_wrap(~variable, ncol=2, 
               labeller = lab.p)+
    ggtitle(paste0("-log10(P) comparison with gold standard in ",platform))+
    xlab("-log10(P) Values from gold standard") + ylab("-log10(P) Values from CCLS")+
    theme_bw()+
    theme(plot.title = element_text(hjust = 0.5))
ggsave(paste0(platform,".metapaper.p.compare.png"),width = 8,height = 6)

}
```

450K CpGs (Sets 1, 2 subjects) using mQTLs from all sets

```{r looking at results after meta}

setwd("~/mQTL/birthweight")
library(tidyverse)
library(data.table)
library(qvalue)
library(grid)
library(bacon)

library(IlluminaHumanMethylation450kanno.ilmn12.hg19)
data("IlluminaHumanMethylation450kanno.ilmn12.hg19")
anno       <- IlluminaHumanMethylation450kanno.ilmn12.hg19 %>%
  getAnnotation %>%
  as.data.frame %>%
  rownames_to_column(var="cpg") %>%
  dplyr::select(cpg,chr,pos,Relation_to_Island,UCSC_RefGene_Name,UCSC_RefGene_Group,Probe_maf,CpG_maf)


gold = read.csv("bw_meta_paper.csv") %>%
  as.data.frame() %>%
  dplyr::select(CpG)
  
#########################################################################################
##### Counting number of significant CpGs with/without SNP controls (same CpGs with mQTL)

set1.snp = fread("ccls.set1.birthWt.geneCon.ewas.txt") %>%
  mutate(set1_snp=snp)%>%
  dplyr::select(CpG,set1_snp)
set2.snp = fread("ccls.set2.birthWt.geneCon.ewas.txt")%>%
  mutate(set2_snp=snp)%>%
  dplyr::select(CpG,set2_snp)
set3.snp = fread("ccls.set3.Y_birthweight.geneCon.ewas.txt")%>%
  mutate(set3_snp=snp)%>%
  dplyr::select(CpG,set3_snp)
set4.snp = fread("ccls.set4.ch_birthwt_bc_full.geneCon.ewas.txt")%>%
  mutate(set4_snp=snp)%>%
  dplyr::select(CpG,set4_snp)
  

platform= "450K"
annoFile = anno

meta = fread("meta.450K.all.Platforms.bw.geneCon.txt") %>%
  dplyr::select(-Allele1, -Allele2) %>%
  dplyr::filter(grepl('cg', MarkerName)) %>%
  dplyr::rename("CpG" = "MarkerName") %>%
  left_join(set1.snp,by="CpG")%>%
  left_join(set2.snp,by="CpG") %>%
  inner_join(annoFile,by=c("CpG"="cpg")) %>%
  arrange(by=`P-value`) %>%
  mutate(`P-value`=as.numeric(`P-value`))%>%
  dplyr::filter(Probe_maf<0.05|is.na(Probe_maf))%>%
  dplyr::filter(CpG_maf<0.05|is.na(CpG_maf)) %>%
  dplyr::filter(chr!="chrX" & chr!="chrY")%>%
  dplyr::filter(!is.na(`P-value`))%>%
  mutate(tss=0)
meta$tss[grep("TSS",meta$UCSC_RefGene_Group)]=1

bacon_P = pval(bacon((meta$Effect)/(meta$StdErr),na.exclude = TRUE))

# raw P values
Nbeta <- sum(!is.na(meta$`P-value`))
tmp <- meta[!is.na(meta$`P-value`),]
lambda = median(qchisq(1- tmp$`P-value`,1))/qchisq(0.5,1)
print(round(lambda, 2))
#1.3

# bacon P values
Nbeta <- sum(!is.na(bacon_P))
tmp <- bacon_P[!is.na(bacon_P)]
lambda = median(qchisq(1- bacon_P,1))/qchisq(0.5,1)
print(round(lambda, 2))
#1.03

meta$bacon_P = bacon_P
meta = meta %>%
  mutate(fdr=p.adjust(meta$`P-value`, method = "fdr", n=length(meta$`P-value`))) %>%
  mutate(bonferroni=p.adjust(meta$`P-value`, method = "bonferroni", n=length(meta$`P-value`))) 

0.05/nrow(meta) # bonferroni threshold
# 1.192299e-07    

metaNG = fread(paste0("meta.",platform,".bw.Nogene.txt")) %>%
  dplyr::select(-Allele1, -Allele2) %>%
  dplyr::filter(grepl('cg', MarkerName)) %>%
  dplyr::rename("CpG" = "MarkerName") %>%
  inner_join(annoFile,by=c("CpG"="cpg")) %>%
  dplyr::filter(Probe_maf<0.05|is.na(Probe_maf))%>%
  dplyr::filter(CpG_maf<0.05|is.na(CpG_maf)) %>%
  dplyr::filter(chr!="chrX" & chr!="chrY")%>%
  mutate(`P-value`=as.numeric(`P-value`))%>%
  dplyr::filter(!is.na(`P-value`))%>%
  arrange(by=`P-value`) %>%
  mutate(tss=0)
metaNG$tss[grep("TSS",metaNG$UCSC_RefGene_Group)]=1

bacon_P = pval(bacon((metaNG$Effect)/(metaNG$StdErr),na.exclude = TRUE))

# raw P values
Nbeta <- sum(!is.na(metaNG$`P-value`))
tmp <- metaNG[!is.na(metaNG$`P-value`),]
lambda = median(qchisq(1- tmp$`P-value`,1))/qchisq(0.5,1)
print(round(lambda, 2))
#2.77

# bacon P values
Nbeta <- sum(!is.na(bacon_P))
tmp <- bacon_P[!is.na(bacon_P)]
lambda = median(qchisq(1- bacon_P,1))/qchisq(0.5,1)
print(round(lambda, 2))
#1.37

metaNG$bacon_P = bacon_P
metaNG = metaNG %>%
  mutate(fdr=p.adjust(metaNG$bacon_P, method = "fdr", n=length(metaNG$bacon_P))) %>%
  mutate(bonferroni=p.adjust(metaNG$bacon_P, method = "bonferroni", n=length(metaNG$bacon_P))) 

0.05/nrow(metaNG) # bonferroni threshold
# 1.194515e-07 for 450K

write.csv(meta,paste0("bw.mqtl.",platform,".meta.commonCpGs.allMQTLs.csv"))
write.csv(metaNG,paste0("bw.Nomqtl.",platform,".meta.commonCpGs.allMQTLs.csv"))

meta.sig = meta%>%filter(bonferroni<0.05)
# dim(meta.sig)
#[1] 36   20

metaNG.sig = metaNG%>%filter(bonferroni<0.05)
# dim(metaNG.sig)
#[1] 3   16

write.csv(metaNG.sig,paste0("bw.sig.Nomqtl.",platform,".meta.commonCpGs.allMQTLs.csv"))
write.csv(meta.sig,paste0("bw.sig.mqtl.",platform,".meta.commonCpGs.allMQTLs.csv"))

print(paste0("Results of ", platform," using all mQTLs"))
print(paste0(nrow(meta.sig)," probes are significant after SNP correction for platform ",platform))

a1=sum(meta.sig$set1_snp!="none")
a2=sum(meta.sig$set2_snp!="none")    

preGold = length(which(meta.sig$CpG %in% gold$CpG))
novelCpGs = nrow(meta.sig) - preGold
print(paste0(preGold, " of CpGs were previously reported to be associated with birthweight"))
print(paste0(novelCpGs, " of CpGs were novel CpGS with birthweight"))

print(paste0(a1," (or ",a2,") probes are matched to a significant mQTL for platform ",platform))

print("CpGs that would not have been identified if not corrected for mQTLs")
meta.sig[which(!meta.sig$CpG %in% metaNG.sig$CpG),]
print(nrow(meta.sig[which(!meta.sig$CpG %in% metaNG.sig$CpG),]))
          
print(paste0("distributiono of TSS",platform,", with mQTL"))
print(table(meta.sig$tss))
print(table(meta.sig$tss)[2]/(table(meta.sig$tss)[1]+table(meta.sig$tss)[2]))

print(paste0("distributiono of TSS",platform,", without mQTL"))
print(table(metaNG.sig$tss))
print(table(metaNG.sig$tss)[2]/(table(metaNG.sig$tss)[1]+table(metaNG.sig$tss)[2]))


#[1] 2.77
#[1] 1.4
#[1] 1.194515e-07

#[1] "Results of 450K using all mQTLs"
#[1] "36 probes are significant after SNP correction for platform 450K"
#[1] "Results of 450K using all mQTLs"
#[1] "36 probes are significant after SNP correction for platform 450K"
#[1] "20 of CpGs were previously reported to be associated with birthweight"
#[1] "16 of CpGs were novel CpGS with birthweight"

# [1] "17 (or 17) probes are matched to a significant mQTL for platform 450K"

# [1] "CpGs that would not have been identified if not corrected for mQTLs"
# [1] 35

#[1] "distributiono of TSS450K, with mQTL"
# 0  1 
#32  4 
#        1 
#0.1111111 

#[1] "distributiono of TSS450K, without mQTL"
#0 1 
#2 1 
#        1 
#0.3333333 

##################################################################################
###### comparing SNP corrected vs none
platform = "450K"

meta = fread(paste0("bw.mqtl.",platform,".meta.commonCpGs.allMQTLs.csv"))%>%
  dplyr::rename("effect_mqtl"="Effect","p_mqtl"="P-value","direction_mqtl"="Direction","mqtl_bonferroni"="bonferroni") %>%
  dplyr::select(effect_mqtl,p_mqtl,mqtl_bonferroni,direction_mqtl,CpG,set1_snp,set2_snp) %>%
  dplyr::filter(!(set1_snp=="none"&set2_snp=="none"))

metaNG = fread(paste0("bw.Nomqtl.",platform,".meta.csv"))%>%
  dplyr::rename("effect_nomqtl"="Effect","p_nomqtl"="P-value","direction_nomqtl"="Direction",
                "nomqtl_bonferroni"="bonferroni") %>%
  dplyr::select(CpG,effect_nomqtl,p_nomqtl,nomqtl_bonferroni,direction_nomqtl)

# select probes that are significant either with or without SNP correction
metaCompare = meta %>%
  inner_join(metaNG,by="CpG")%>%
  mutate(sigBi = factor(ifelse(mqtl_bonferroni < 0.05,"sig","non_sig"))) %>%
  mutate(logP_nomqtl=-log10(p_nomqtl),logP_mqtl=-log10(p_mqtl))
nrow(metaCompare)

ggplot(metaCompare)+
  geom_point(data = subset(metaCompare,sigBi == 'non_sig'),
           aes(x = effect_nomqtl, y = effect_mqtl), alpha = 1/8,color='turquoise')+
  geom_point(data = subset(metaCompare,sigBi == 'sig'),
           aes(x = effect_nomqtl, y = effect_mqtl), alpha = 0.8,color='salmon')+
  geom_abline(aes(intercept=0,slope=1),alpha=0.4,color="red")+
  geom_hline(aes(yintercept=0),color="red",alpha=0.4)+
  geom_vline(aes(xintercept=0),color="red",alpha=0.4)+
  #xlim(min,max)+ylim(min,max)+
  xlab("regression coefficient values, without mQTL") + ylab("regression coefficient values, with mQTL")+
  theme_bw()+
  theme(axis.text.x = element_text(size=12),
        axis.text.y = element_text(size=12),
        axis.title.x = element_text(size=14),
        axis.title.y = element_text(size=14),
        plot.margin=unit(c(5.5, 19, 5.5, 5.5), "points"))

ggsave(paste0(platform,".effectComp.commonCpGs.allMQTLs.png"),width = 8.5,height = 6)


ggplot(metaCompare)+
  geom_point(data = subset(metaCompare,sigBi == 'non_sig'),
           aes(x = logP_nomqtl, y = logP_mqtl), alpha = 1/8,color='turquoise')+
  geom_point(data = subset(metaCompare,sigBi == 'sig'),
           aes(x = logP_nomqtl, y = logP_mqtl), alpha = 1,color='salmon')+
  geom_abline(aes(intercept=0,slope=1),alpha=0.4,color="red")+
  xlab("log P values, without mQTL") + ylab("log P values, with mQTL")+
  theme_bw()+
  theme(axis.text.x = element_text(size=12),
        axis.text.y = element_text(size=12),
        axis.title.x = element_text(size=14),
        axis.title.y = element_text(size=14))
ggsave(paste0(platform,".pCompAll.commonCpGs.allMQTLs.png"),width = 8,height = 6)

```

450K CpGs (Sets 1, 2, 3, 4 subjects) using mQTLs from all sets

```{r looking at results after meta}

setwd("~/mQTL/birthweight")
library(tidyverse)
library(data.table)
library(qvalue)
library(grid)
library(bacon)

library(IlluminaHumanMethylation450kanno.ilmn12.hg19)
data("IlluminaHumanMethylation450kanno.ilmn12.hg19")
anno       <- IlluminaHumanMethylation450kanno.ilmn12.hg19 %>%
  getAnnotation %>%
  as.data.frame %>%
  rownames_to_column(var="cpg") %>%
  dplyr::select(cpg,chr,pos,Relation_to_Island,UCSC_RefGene_Name,UCSC_RefGene_Group,Probe_maf,CpG_maf)


gold = read.csv("bw_meta_paper.csv") %>%
  as.data.frame() %>%
  dplyr::select(CpG)
  
#########################################################################################
##### Counting number of significant CpGs with/without SNP controls (same CpGs with mQTL)

set1.snp = fread("ccls.set1.birthWt.geneCon.ewas.txt") %>%
  mutate(set1_snp=snp)%>%
  dplyr::select(CpG,set1_snp)
set2.snp = fread("ccls.set2.birthWt.geneCon.ewas.txt")%>%
  mutate(set2_snp=snp)%>%
  dplyr::select(CpG,set2_snp)
set3.snp = fread("ccls.set3.Y_birthweight.geneCon.ewas.txt")%>%
  mutate(set3_snp=snp)%>%
  dplyr::select(CpG,set3_snp)
set4.snp = fread("ccls.set4.ch_birthwt_bc_full.geneCon.ewas.txt")%>%
  mutate(set4_snp=snp)%>%
  dplyr::select(CpG,set4_snp)
  

platform= "450K"
annoFile = anno

meta = fread("meta.450K.all.allsubs.Platforms.bw.geneCon.txt") %>%
  dplyr::select(-Allele1, -Allele2) %>%
  dplyr::filter(grepl('cg', MarkerName)) %>%
  dplyr::rename("CpG" = "MarkerName") %>%
  left_join(set1.snp,by="CpG")%>%
  left_join(set2.snp,by="CpG") %>%
  left_join(set3.snp,by="CpG")%>%
  left_join(set4.snp,by="CpG") %>%
  inner_join(annoFile,by=c("CpG"="cpg")) %>%
  arrange(by=`P-value`) %>%
  mutate(`P-value`=as.numeric(`P-value`))%>%
  dplyr::filter(Probe_maf<0.05|is.na(Probe_maf))%>%
  dplyr::filter(CpG_maf<0.05|is.na(CpG_maf)) %>%
  dplyr::filter(chr!="chrX" & chr!="chrY")%>%
  dplyr::filter(!is.na(`P-value`))%>%
  mutate(tss=0)
meta$tss[grep("TSS",meta$UCSC_RefGene_Group)]=1

bacon_P = pval(bacon((meta$Effect)/(meta$StdErr),na.exclude = TRUE))

# raw P values
Nbeta <- sum(!is.na(meta$`P-value`))
tmp <- meta[!is.na(meta$`P-value`),]
lambda = median(qchisq(1- tmp$`P-value`,1))/qchisq(0.5,1)
print(round(lambda, 2))
#1.55

# bacon P values
Nbeta <- sum(!is.na(bacon_P))
tmp <- bacon_P[!is.na(bacon_P)]
lambda = median(qchisq(1- bacon_P,1))/qchisq(0.5,1)
print(round(lambda, 2))
#1.16

meta$bacon_P = bacon_P
meta = meta %>%
  mutate(fdr=p.adjust(meta$`P-value`, method = "fdr", n=length(meta$`P-value`))) %>%
  mutate(bonferroni=p.adjust(meta$`P-value`, method = "bonferroni", n=length(meta$`P-value`))) 

0.05/nrow(meta) # bonferroni threshold
#  1.191784e-07

metaNG = fread("meta.450K.all.allsubs.Platforms.bw.Nogene.txt") %>%
  dplyr::select(-Allele1, -Allele2) %>%
  dplyr::filter(grepl('cg', MarkerName)) %>%
  dplyr::rename("CpG" = "MarkerName") %>%
  inner_join(annoFile,by=c("CpG"="cpg")) %>%
  dplyr::filter(Probe_maf<0.05|is.na(Probe_maf))%>%
  dplyr::filter(CpG_maf<0.05|is.na(CpG_maf)) %>%
  dplyr::filter(chr!="chrX" & chr!="chrY")%>%
  mutate(`P-value`=as.numeric(`P-value`))%>%
  dplyr::filter(!is.na(`P-value`))%>%
  arrange(by=`P-value`) %>%
  mutate(tss=0)
metaNG$tss[grep("TSS",metaNG$UCSC_RefGene_Group)]=1

bacon_P = pval(bacon((metaNG$Effect)/(metaNG$StdErr),na.exclude = TRUE))

# raw P values
Nbeta <- sum(!is.na(metaNG$`P-value`))
tmp <- metaNG[!is.na(metaNG$`P-value`),]
lambda = median(qchisq(1- tmp$`P-value`,1))/qchisq(0.5,1)
print(round(lambda, 2))
#1.3

# bacon P values
Nbeta <- sum(!is.na(bacon_P))
tmp <- bacon_P[!is.na(bacon_P)]
lambda = median(qchisq(1- bacon_P,1))/qchisq(0.5,1)
print(round(lambda, 2))
#1.1

metaNG$bacon_P = bacon_P
metaNG = metaNG %>%
  mutate(fdr=p.adjust(metaNG$bacon_P, method = "fdr", n=length(metaNG$bacon_P))) %>%
  mutate(bonferroni=p.adjust(metaNG$bacon_P, method = "bonferroni", n=length(metaNG$bacon_P))) 

0.05/nrow(metaNG) # bonferroni threshold
# 1.277625e-07

write.csv(meta,paste0("bw.mqtl.",platform,".allsubs.meta.commonCpGs.allMQTLs.csv"))
write.csv(metaNG,paste0("bw.Nomqtl.",platform,".allsubs.meta.commonCpGs.allMQTLs.csv"))

meta.sig = meta%>%filter(bonferroni<0.05)
# dim(meta.sig)
#[1] 1928    20

metaNG.sig = metaNG%>%filter(bonferroni<0.05)
# dim(metaNG.sig)
#[1] 698   16

write.csv(metaNG.sig,paste0("bw.sig.Nomqtl.",platform,".allsubs.meta.commonCpGs.allMQTLs.csv"))
write.csv(meta.sig,paste0("bw.sig.mqtl.",platform,".allsubs.meta.commonCpGs.allMQTLs.csv"))

print(paste0("Results of ", platform," using all mQTLs, all subs"))
print(paste0(nrow(meta.sig)," probes are significant after SNP correction for platform ",platform))

a1=sum(meta.sig$set1_snp!="none",na.rm=TRUE)
a2=sum(meta.sig$set2_snp!="none",na.rm=TRUE)    
a3=sum(meta.sig$set3_snp!="none",na.rm=TRUE)
a4=sum(meta.sig$set4_snp!="none",na.rm=TRUE)    

preGold = length(which(meta.sig$CpG %in% gold$CpG))
novelCpGs = nrow(meta.sig) - preGold
print(paste0(preGold, " of CpGs were previously reported to be associated with birthweight"))
print(paste0(novelCpGs, " of CpGs were novel CpGS with birthweight"))

print(paste0(a1," (or ",a2,") probes are matched to a significant mQTL for platform ",platform))
print(paste0(a3," (or ",a4,") probes are matched to a significant mQTL for platform ",platform))

print("CpGs that would not have been identified if not corrected for mQTLs")
meta.sig[which(!meta.sig$CpG %in% metaNG.sig$CpG),]
print(nrow(meta.sig[which(!meta.sig$CpG %in% metaNG.sig$CpG),]))
          
print(paste0("distributiono of TSS",platform,", with mQTL"))
print(table(meta.sig$tss))
print(table(meta.sig$tss)[2]/(table(meta.sig$tss)[1]+table(meta.sig$tss)[2]))

print(paste0("distributiono of TSS",platform,", without mQTL"))
print(table(metaNG.sig$tss))
print(table(metaNG.sig$tss)[2]/(table(metaNG.sig$tss)[1]+table(metaNG.sig$tss)[2]))


#Meta-analysis of sets 1, 2, 3, 4 subjects, limited to 450K CpGs.
#1928 probes are significant after SNP correction for platform 450K.
#366 of CpGs were previously reported to be associated with birthweight, and 1562 of CpGs were novel #CpGS with birthweight.
#968 probes on 450K and 1222 probes on EPIC are matched to SNPs.
#1240 probes would not have been identified without mQTL controlling.


##################################################################################
###### comparing SNP corrected vs none
platform = "450K"

meta = fread(paste0("bw.mqtl.",platform,".allsubs.meta.commonCpGs.allMQTLs.csv"))%>%
  dplyr::rename("effect_mqtl"="Effect","p_mqtl"="P-value","direction_mqtl"="Direction","mqtl_bonferroni"="bonferroni") %>%
  dplyr::select(effect_mqtl,p_mqtl,mqtl_bonferroni,direction_mqtl,CpG,set1_snp,set2_snp) %>%
  dplyr::filter(!(set1_snp=="none"&set2_snp=="none"))

metaNG = fread(paste0("bw.Nomqtl.",platform,".allsubs.meta.commonCpGs.allMQTLs.csv"))%>%
  dplyr::rename("effect_nomqtl"="Effect","p_nomqtl"="P-value","direction_nomqtl"="Direction",
                "nomqtl_bonferroni"="bonferroni") %>%
  dplyr::select(CpG,effect_nomqtl,p_nomqtl,nomqtl_bonferroni,direction_nomqtl)

# select probes that are significant either with or without SNP correction
metaCompare = meta %>%
  inner_join(metaNG,by="CpG")%>%
  mutate(sigBi = factor(ifelse(mqtl_bonferroni < 0.05,"sig","non_sig"))) %>%
  mutate(logP_nomqtl=-log10(p_nomqtl),logP_mqtl=-log10(p_mqtl))
nrow(metaCompare)

# figure S6
ggplot(metaCompare)+
  geom_point(data = subset(metaCompare,sigBi == 'non_sig'),
           aes(x = effect_nomqtl, y = effect_mqtl), alpha = 1/8,color='turquoise')+
  geom_point(data = subset(metaCompare,sigBi == 'sig'),
           aes(x = effect_nomqtl, y = effect_mqtl), alpha = 0.8,color='salmon')+
  geom_abline(aes(intercept=0,slope=1),alpha=0.4,color="red")+
  geom_hline(aes(yintercept=0),color="red",alpha=0.4)+
  geom_vline(aes(xintercept=0),color="red",alpha=0.4)+
  #xlim(min,max)+ylim(min,max)+
  xlab("regression coefficient values, without mQTL") + ylab("regression coefficient values, with mQTL")+
  theme_bw()+
  theme(axis.text.x = element_text(size=17),
        axis.text.y = element_text(size=17),
        axis.title.x = element_text(size=19),
        axis.title.y = element_text(size=19),
        plot.margin=unit(c(5.5, 19, 5.5, 5.5), "points"))

ggsave(paste0(platform,".allsubs.effectComp.commonCpGs.allMQTLs.png"),width = 8.5,height = 6)


ggplot(metaCompare)+
  geom_point(data = subset(metaCompare,sigBi == 'non_sig'),
           aes(x = logP_nomqtl, y = logP_mqtl), alpha = 1/8,color='turquoise')+
  geom_point(data = subset(metaCompare,sigBi == 'sig'),
           aes(x = logP_nomqtl, y = logP_mqtl), alpha = 1,color='salmon')+
  geom_abline(aes(intercept=0,slope=1),alpha=0.4,color="red")+
  xlab("竏値og10 (P-value), without mQTL") + 
  ylab("竏値og10 (P-value), with mQTL")+
  theme_bw()+
  theme(axis.text.x = element_text(size=17),
        axis.text.y = element_text(size=17),
        axis.title.x = element_text(size=19),
        axis.title.y = element_text(size=19))
ggsave(paste0(platform,".allsubs.pCompAll.commonCpGs.allMQTLs.png"),width = 8,height = 6)

```

>> Sensitivity analysis: including maternal weight gain in the model

```{r preparing data}

library(tidyverse)
library(data.table)

setwd("~/mQTL/birthweight")

winsorize <- function(methylation,pct=winsorize.pct) {
  quantiles <- matrixStats::rowQuantiles(methylation, probs=c(pct,1-pct), na.rm=T)
  low <- quantiles[,1]
  upper <- quantiles[,2]

  outliers.lower <- rowSums(methylation < low, na.rm=T)
  outliers.upper <- rowSums(methylation > upper, na.rm=T)
  
  idx <- which(methylation < low, arr.ind=T)
  methylation[idx] <- low[idx[,1]]
  
  idx <- which(methylation > upper, arr.ind=T)
  methylation[idx] <- upper[idx[,1]]

  n <- rowSums(!is.na(methylation))
  log <- data.frame(outliers.lower, outliers.upper, n)
  return(list(methylation=methylation, log=log))
}

wg = read.csv("~/ewas/birthorder/sourceF/set124_weightgain.csv") %>%
    mutate(setN=set) %>%
    dplyr::select(beadPosition,mo_weightgain,setN)


#for(set in c(1,2,4)){
set=4

 wg1 = wg %>% 
    as.data.frame()%>%
    dplyr::filter(setN==paste0("set",set)) %>% 
    dplyr::select(-setN) %>%
    na.omit()


 if(set==1){
   var = "birthWt"
   clinical0 = read.csv(paste0("~/sets/set",set,"/clinical_variables.csv")) %>%
     filter(Trisomy21.cnm==0)%>%
      inner_join(wg1, by="beadPosition")
    }
  
 if(set==2){
  var = "birthWt"
  clinical0 = read.csv(paste0("~/sets/set",set,"/clinical_variables.csv")) %>%
      filter(Trisomy21.cnm==0)%>%
      inner_join(wg1, by="beadPosition")
  }
    
 if(set==4){
   var = "ch_birthwt_bc_full"
   clinical0 = read.csv(paste0("~/sets/set",set,"/clinical_variables.csv")) %>%
     filter(smp_type=="bg",Trisomy21.cnm==0)%>%
     inner_join(wg1, by="beadPosition")
   }
  
###################################
# Determine subjects to involve
## Methylation data
####noob
if(set!=4){methy_tm_0 = read_rds(paste0("~/sets/set",set,"/beta_ritu_imputed_set",set,".rds"))}
if(set==4){methy_tm_0 = read_rds(paste0("~/sets/set",set,"/beta_funnorm_bmiq_imputed_set",set,".rds"))}
rownames(methy_tm_0)=c()
metMemDF = data.frame(beadPosition=colnames(methy_tm_0)[-1]) 
 
## Cov data

clinical = clinical0[,c("beadPosition","subjectId","sex","Batch","gestage","CaCo","mo_weightgain",var)] %>%
  filter(gestage>=30)%>%
  na.omit() %>%
  mutate(ID = paste0("0_",subjectId))%>%
  distinct(ID,.keep_all = TRUE)

dcp = fread(paste0("~/sets/set",set,"/idol_deconvolution.csv")) %>%
  as.data.frame() %>%
  dplyr::select(beadPosition,CD8T,CD4T,NK,Bcell,Mono,Gran,nRBC) %>%
  na.omit()

genePC = read.table(paste0("~/mQTL/ccls.set",set,".pc.eigenvec"))
colnames(genePC) = c("FID","IID","PC1","PC2","PC3","PC4","PC5","PC6","PC7","PC8","PC9","PC10")
genePC1 = genePC %>%
  dplyr::mutate(ID = paste0(FID,"_",IID)) %>%
  dplyr::select(ID,PC1,PC2,PC3,PC4,PC5,PC6,PC7,PC8,PC9,PC10)

clinical1 = clinical %>%
  inner_join(metMemDF,by="beadPosition") %>%
  inner_join(dcp,by="beadPosition") %>%
  inner_join(genePC1,by="ID") %>%
  dplyr::mutate(Batch=as.factor(Batch)) 
covMem = clinical1$ID

## Genome data
genMem = read.table(paste0("/ccls/proj_circle2_p3/users/sebastian/bySET/ccls.gwas.set",set,".fam")) %>%
    dplyr::mutate(ID = paste0(V1,"_",V2)) %>%
    dplyr::select(ID)
genMems = genMem$ID

## Methylation data processing
methy_tm =  methy_tm_0 %>%
  as.data.frame() %>% 
  column_to_rownames("probeId") %>%
  data.matrix() 
replace.outliers <- winsorize(methy_tm,0.005)
methy <- replace.outliers$methylation %>% 
  as.data.frame() %>%
  rownames_to_column(var="probeId")

## Choosing overlaps, respecting the sequence of genetics, also outputting a list of subjects dropped from genetics
miInd = which(!genMems%in%covMem) %>% unique()

if(length(miInd)!=0){
  finalMems = data.frame(ID = genMems[-miInd])
}else{
  finalMems = data.frame(ID = genMems)}

###################################
# Preparing covariates
clinical2 = clinical1 %>%
  dplyr::select(-beadPosition, -subjectId) %>%
  dplyr::select(ID, one_of(var),everything()) 

write.table(clinical2,paste0("covsSet",set,".s1.txt"),sep='\t',quote=FALSE,row.names = FALSE,col.names = TRUE)

###################################
# Preparing methylation files
subs = clinical1 %>% column_to_rownames(var="ID")
subs1 = subs[as.character(finalMems$ID),"beadPosition"] %>% as.character()
methy1 = methy[,c("probeId",subs1)]
colnames(methy1) = c("probeId",as.character(finalMems$ID))

write.table(methy1,paste0("methSet",set,".s1.txt"),sep='\t',quote=FALSE,row.names = FALSE,col.names = TRUE)

print(set)
print((dim(methy1)[2]-1)==(dim(clinical2)[1]))

#}

```

```{r adjusting  for scanned mqtls with QTLtools}

setwd("~/mQTL/birthweight")
library(tidyverse)
library(data.table)
library(MASS)
library(lmtest)
library(sandwich)

library(parallel) 
cl <- makeCluster(40, type="FORK")
clusterEvalQ(cl,library(MASS))
clusterEvalQ(cl,library(lmtest))

linear <- function(meth,cov){

  #    write_rds(meth,"meth.test.rds")
  #}
  
  # meth = read_rds("meth.test.rds")
  # cov = covs

  probe <- unname(meth[1])
  methDF = meth[-1] %>% as.numeric()
  
  dat <- data.frame(y = methDF, cov)
  ff  <- formula(paste("y ~ ", paste0(colnames(cov),collapse="+")))
  #mod <- try(rlm(ff, data = dat, maxit = 200),silent = TRUE)
  mod <- try(lm(ff, data = dat),silent = TRUE)
  
  #cf = try(coeftest(mod),silent = TRUE)
  N <- sum(!is.na(dat[,1]))
  
  if (class(mod)=="lm") {
    
    coefs = summary(mod)$coefficients
    coef = coefs[2,"Estimate"]
    SE = coefs[2,"Std. Error"]
    p.value = coefs[2,"Pr(>|t|)"]
    out = c(coef, SE, p.value, N) 
  } else {
    out = c(rep(NA,3), N)
  }
  names(out) = c("Beta","SE", "P", "N"); return(out) 
}
gene.linear <- function(meth,cov,gene){
  
  #    write_rds(meth,"meth.test.rds")
  #}
  
  # meth = read_rds("meth.test.rds")
  # cov = covs
  # gene = mtqls
  
  probe <- unname(meth[1])
  
  #print(probe)
  
  methDF = meth[-1] %>% as.numeric()
  
  SNPname <- gene[which(gene$CpG == probe), 1]  %>% as.character()  
  SNP <-  gene[which(gene$CpG == probe), -c(1:3)] %>% t() %>% as.data.frame()
  colnames(SNP) <- "SNP"
  
  if(length(table(SNP$SNP))==1){
    dat <- data.frame(y = methDF, cov)
    ff  <- formula(paste("y ~ ", paste0(colnames(cov),collapse="+")))
  }else{
    dat <- data.frame(y = methDF, cov,SNP)
    ff  <- formula(paste("y ~ ", paste0(c(colnames(cov),"SNP"),collapse="+")))
  }
  
  mod <- try(lm(ff, data = dat),silent = TRUE)
  
  #cf = try(coeftest(mod),silent = TRUE)
  N <- sum(!is.na(dat[,1]))
  
  if (class(mod)=="lm") {
    
    SNPname=SNPname
    #pvals = try(f.robftest(mod, var = colnames(cov)[1]),silent = TRUE)
    #coef = cf[2,"Estimate"]
    #F.value = ifelse(class(pvals)=="htest",pvals$statistic,NA)
    #p.value = ifelse(class(pvals)=="htest",pvals$p.value,NA)
    #out = c(SNPname,coef, F.value, p.value, N) 

    coefs = summary(mod)$coefficients
    coef = coefs[2,"Estimate"]
    SE = coefs[2,"Std. Error"]
    p.value = coefs[2,"Pr(>|t|)"]
    out = c(SNPname,coef, SE, p.value, N) 
  } else {
    out = c(SNPname,rep(NA,3), N)
  }
  names(out) = c("snp","Beta","SE", "P", "N"); return(out) 
}

#for (set in c(1,2,4)){
set=4
  
 if(set==1){var = "birthWt"}
 if(set==2){var = "birthWt"}
 if(set==4){var = "ch_birthwt_bc_full"}
  
  methyl0 = fread(paste0("methSet",set,".s1.txt")) %>%
    column_to_rownames(var="probeId")%>%
    as.data.frame()

  covs0 = fread(paste0("covsSet",set,".s1.txt"))%>%
    as.data.frame()%>%
    column_to_rownames(var="ID") %>%
    mutate(Batch = as.factor(Batch))

  mtqls0 = fread(paste0("~/mQTL/ccls.set",set,".mqtls.hd")) %>% 
    as.data.frame()
  
  subs.list = intersect(intersect(rownames(covs0),colnames(methyl0)), colnames(mtqls0)) 

  methyl = methyl0[,subs.list]%>% rownames_to_column(var="cpg")
  rownames(methyl) <- methyl$cpg
  
  covs = covs0[subs.list,]
  mtqls = mtqls0[,c("SNP","CpG","P-value",subs.list)]
  
  probes = rownames(methyl)[which(rownames(methyl)%in%mtqls$CpG)] 

# no correction for SNP
    
  results <- t(parApply(cl, methyl, 1, linear, covs)) 
  results <- as.data.frame(results) %>%
    rownames_to_column(var="CpG") %>%
    mutate(N=as.numeric(N))%>%
    mutate(P=as.numeric(P))%>%
    filter(N>max(N)/2,
           !is.na(P))

  write.table(results,paste0("ccls.set",set,".",var,".Nogene.s1.ewas.txt"),
          quote=FALSE,
          row.names = FALSE,
          sep='\t')  

  Nbeta <- sum(!is.na(results$P))
  tmp <- results[!is.na(results$P),]
  
  print(paste0("lambda of set ",set," without SNP control:"))
  lambda = median(qchisq(1- tmp$P,1))/qchisq(0.5,1)
  print(round(lambda, 2))
  
# correcting for SNPs for CpGs with mQTLs only
  
  methyl.nosnp = methyl[setdiff(rownames(methyl),probes),] 

  #methyl.nosnp.t = methyl.nosnp[1:100,]
  #results.nosnp <- t(apply(methyl.nosnp.t, 1, linear, covs)) 
  results.nosnp <- t(parApply(cl, methyl.nosnp, 1, linear, covs)) 
  
  results.nosnp <- as.data.frame(results.nosnp) %>%
    rownames_to_column(var="CpG")%>%
    mutate(snp="none")%>%
    mutate(N=as.numeric(N))%>%
    mutate(P=as.numeric(P))%>%
    filter(N>max(N)/2,
           !is.na(P)) %>%
    dplyr::select(CpG,snp,Beta,SE,P,N)
  
  methyl.cpg = methyl[probes,]
  
  #methyl.cpg.t = methyl.cpg[1:100,]
  #results.snp <- t(apply(methyl.cpg.t, 1, gene.linear, covs, mtqls)) 
  results.snp <- t(parApply(cl, methyl.cpg, 1, gene.linear, covs, mtqls)) 
  results.snp <- as.data.frame(results.snp) %>%
    rownames_to_column(var="CpG")%>%
    mutate(N=as.numeric(N))%>%
    mutate(P=as.numeric(P))%>%
    filter(N>max(N)/2,
           !is.na(P))
  
  results.snp.combind = rbind(results.nosnp,results.snp) %>%
    as.data.frame()%>%
    arrange(P)

  write.table(results.snp.combind,
              paste0("ccls.set",set,".",var,".geneCon.s1.ewas.txt"),
              quote=FALSE,
              row.names = FALSE,
              sep='\t')  
  
  tmp.snp <- results.snp.combind %>% 
    dplyr::select(P) %>% mutate(P=as.numeric(P)) %>% na.omit() 
  lambda.snp = median(qchisq(1- tmp.snp$P,1))/qchisq(0.5,1)
  print(paste0("lambda of set ",set," with SNP control:"))
  print(round(lambda.snp, 2))
  
#  }

stopCluster(cl)
rm(list=ls())
  

```

```{bash cleaning files}

cd ~/mQTL/birthweight

### 450K
metal
MARKER CpG
EFFECT Beta
PVALUE P
SCHEME STDERR
STDERR SE
VERBOSE OFF
GENOMICCONTROL OFF
STDERR_PRINT_PRECISION 20
EFFECT_PRINT_PRECISION 20
PROCESS ccls.set1.birthWt.geneCon.s1.ewas.txt
PROCESS ccls.set2.birthWt.geneCon.s1.ewas.txt
ANALYZE
QUIT
mv METAANALYSIS1.TBL meta.450K.bw.geneCon.s1.txt

### 450K, noGene
metal
MARKER CpG
EFFECT Beta
PVALUE P
SCHEME STDERR
STDERR SE
VERBOSE OFF
GENOMICCONTROL OFF
STDERR_PRINT_PRECISION 20
EFFECT_PRINT_PRECISION 20
PROCESS ccls.set1.birthWt.Nogene.s1.ewas.txt
PROCESS ccls.set2.birthWt.Nogene.s1.ewas.txt
ANALYZE
QUIT
mv METAANALYSIS1.TBL meta.450K.bw.Nogene.s1.txt

```

```{r plot the beta values and P values in analysis and sensitivity analysis}
setwd("~/mQTL/birthweight")
library(tidyverse)
library(data.table)

library(IlluminaHumanMethylation450kanno.ilmn12.hg19)
data("IlluminaHumanMethylation450kanno.ilmn12.hg19")
anno       <- IlluminaHumanMethylation450kanno.ilmn12.hg19 %>%
  getAnnotation %>%
  as.data.frame %>%
  rownames_to_column(var="cpg") %>%
  dplyr::select(cpg,chr,pos,Relation_to_Island,UCSC_RefGene_Name,UCSC_RefGene_Group,Probe_maf)

library(IlluminaHumanMethylationEPICanno.ilm10b4.hg19)
data("IlluminaHumanMethylationEPICanno.ilm10b4.hg19")
annoEPIC       <- IlluminaHumanMethylationEPICanno.ilm10b4.hg19 %>%
  getAnnotation %>%
  as.data.frame %>%
  rownames_to_column(var="cpg") %>%
  dplyr::select(cpg,chr,pos,Relation_to_Island,UCSC_RefGene_Name,UCSC_RefGene_Group,Methyl450_Loci,Probe_maf)


set="set4"
########################
## with SNP correction
set4G = fread("ccls.set4.ch_birthwt_bc_full.geneCon.ewas.txt") %>%
  dplyr::filter(grepl('cg', CpG)) %>%
  dplyr::rename("Beta_mqtl"="Beta",
                "P_mqtl"="P") %>%
  dplyr::select(CpG,Beta_mqtl,P_mqtl)  
set4G_s = fread("ccls.set4.ch_birthwt_bc_full.geneCon.s1.ewas.txt") %>%
  dplyr::filter(grepl('cg', CpG)) %>%
  dplyr::rename("Beta_mqtl_s"="Beta",
                "P_mqtl_s"="P") %>%
  dplyr::select(CpG,Beta_mqtl_s,P_mqtl_s)  

metaCompare = set4G %>%
  inner_join(set4G_s,by="CpG")

metaCompare %>%
  ggplot(aes(x=Beta_mqtl,y=Beta_mqtl_s))+
  geom_point(alpha = 1/5,color='turquoise',shape=4)+
  geom_abline(aes(intercept=0,slope=1),alpha=0.4,color="red")+
  geom_hline(aes(yintercept=0),color="red",alpha=0.4)+
  geom_vline(aes(xintercept=0),color="red",alpha=0.4)+
  xlab("regression coefficient values") + 
  ylab("regression coefficient values, \nsensitivity analysis")+
  theme_bw()+
  theme(axis.text.x = element_text(size=17),
        axis.text.y = element_text(size=17),
        axis.title.x = element_text(size=19),
        axis.title.y = element_text(size=19),
        plot.margin=unit(c(5.5, 19, 5.5, 5.5), "points"))
ggsave(paste0(set,".sensitivity.beta.png"), width = 8,height = 6)

metaCompare %>%
  mutate(logP_mqtl=-log10(P_mqtl),logP_mqtl_s=-log10(P_mqtl_s))%>%
  ggplot(aes(x=logP_mqtl,y=logP_mqtl_s))+
  geom_point(alpha = 1/5,color='turquoise',shape=4)+
  geom_abline(aes(intercept=0,slope=1),alpha=0.4,color="red")+
  xlab("竏値og10 (P-value)") + 
  ylab("竏値og10 (P-value), \nsensitivity analysis")+
  theme_bw()+
  theme(axis.text.x = element_text(size=17),
        axis.text.y = element_text(size=17),
        axis.title.x = element_text(size=19),
        axis.title.y = element_text(size=19),
        plot.margin=unit(c(5.5, 19, 5.5, 5.5), "points"))
ggsave(paste0(set,".sensitivity.logP.png"), width = 8,height = 6)


# platform="450K"
# ########################
# ## without SNP correction
# metaNG = fread(paste0("meta.",platform,".bw.Nogene.txt")) %>%
#   dplyr::select(-Allele1, -Allele2) %>%
#   mutate(`P-value`=as.numeric(`P-value`))%>%
#   dplyr::filter(!is.na(`P-value`))%>%
#   dplyr::filter(grepl('cg', MarkerName)) %>%
#   dplyr::rename("cpg"="MarkerName","effect_nomqtl"="Effect",
#                 "p_nomqtl"="P-value","direction_nomqtl"="Direction") %>%
#   left_join(anno,by="cpg") %>%
#   dplyr::filter(Probe_maf<0.05|is.na(Probe_maf))%>%
#   mutate(adjP_nomqtl = p.adjust(p_nomqtl, method = "bonferroni", n = length(p_nomqtl)))%>%
#   dplyr::select(cpg,effect_nomqtl,p_nomqtl,adjP_nomqtl,direction_nomqtl)
# 
# metaNG_s = fread(paste0("meta.",platform,".bw.Nogene.s1.txt")) %>%
#   dplyr::select(-Allele1, -Allele2) %>%
#   mutate(`P-value`=as.numeric(`P-value`))%>%
#   dplyr::filter(!is.na(`P-value`))%>%
#   dplyr::filter(grepl('cg', MarkerName)) %>%
#   dplyr::rename("cpg"="MarkerName","effect_nomqtl_s"="Effect",
#                 "p_nomqtll_s"="P-value","direction_nomqtll_s"="Direction") %>%
#   left_join(anno,by="cpg") %>%
#   dplyr::filter(Probe_maf<0.05|is.na(Probe_maf))%>%
#   mutate(adjP_nomqtl_s = p.adjust(p_nomqtll_s, method = "bonferroni", n = length(p_nomqtll_s)))%>%
#   dplyr::select(cpg,effect_nomqtl_s,p_nomqtll_s,adjP_nomqtl_s,direction_nomqtll_s)
#     
# metaCompare = metaNG %>%
#   inner_join(metaNG_s,by="cpg")%>%
#   distinct(cpg,.keep_all = TRUE) %>%
#   filter(p_nomqtl < 0.05 | p_nomqtll_s<0.05)
#  
# max = max(c(metaCompare$effect_nomqtl,metaCompare$effect_nomqtl_s)*1.2)
# min = min(c(metaCompare$effect_nomqtl,metaCompare$effect_nomqtl_s)*1.2)
# 
# metaCompare %>%
#   ggplot(aes(x=effect_nomqtl,y=effect_nomqtl_s))+
#   geom_point()+
#   geom_abline(aes(intercept=0,slope=1),alpha=0.4,color="red")+
#   geom_hline(aes(yintercept=0),color="red")+geom_vline(aes(xintercept=0),color="red")+
#   xlim(min,max)+ylim(min,max)+
#   xlab("beta value") + ylab("beta value, sensitivity analysis")+
#   ggtitle("comparison of beta coefficients in sensitivity analysis")+
#   theme_bw()+
#   theme(plot.title = element_text(hjust = 0.5))
# ggsave(paste0(platform,"sensitivity.beta.png"), width = 8,height = 6)
# 
# ###############
# max = max(c(-log10(metaCompare$p_nomqtl),-log10(metaCompare$p_nomqtll_s)))*1.1
# min = min(c(-log10(metaCompare$p_nomqtl),-log10(metaCompare$p_nomqtll_s)))*0.9
# 
# metaCompare %>%
#   mutate(logP_nomqtl=-log10(p_nomqtl),logP_nomqtl_s=-log10(p_nomqtll_s))%>%
#   ggplot(aes(x=logP_nomqtl,y=logP_nomqtl_s))+
#   geom_point()+
#   geom_abline(aes(intercept=0,slope=1),alpha=0.4,color="red")+
#   geom_hline(aes(yintercept=0),color="red")+geom_vline(aes(xintercept=0),color="red")+
#   xlim(0,max)+ylim(0,max)+
#   xlab("log P value") + ylab("log P value, sensitivity analysis")+
#   ggtitle("comparison of P values in sensitivity analysis")+
#   theme_bw()+
#   theme(plot.title = element_text(hjust = 0.5))
# ggsave(paste0(platform,"sensitivity.logP.png"), width = 8,height = 6)
# 
# 
# ########################
# ## with SNP correction
# meta = fread(paste0("meta.",platform,".bw.geneCon.txt")) %>%
#   dplyr::select(-Allele1, -Allele2) %>%
#   mutate(`P-value`=as.numeric(`P-value`))%>%
#   dplyr::filter(!is.na(`P-value`))%>%
#   dplyr::filter(grepl('cg', MarkerName)) %>%
#   dplyr::rename("cpg"="MarkerName","effect_mqtl"="Effect",
#                 "p_mqtl"="P-value","direction_mqtl"="Direction") %>%
#   left_join(anno,by="cpg") %>%
#   dplyr::filter(Probe_maf<0.05|is.na(Probe_maf))%>%
#   mutate(adjP_mqtl = p.adjust(p_mqtl, method = "bonferroni", n = length(p_mqtl)))%>%
#   dplyr::select(cpg,effect_mqtl,p_mqtl,adjP_mqtl,direction_mqtl)
# 
# meta_s = fread(paste0("meta.",platform,".bw.geneCon.s1.txt")) %>%
#   dplyr::select(-Allele1, -Allele2) %>%
#   mutate(`P-value`=as.numeric(`P-value`))%>%
#   dplyr::filter(!is.na(`P-value`))%>%
#   dplyr::filter(grepl('cg', MarkerName)) %>%
#   dplyr::rename("cpg"="MarkerName","effect_mqtl_s"="Effect",
#                 "p_mqtl_s"="P-value","direction_mqtl_s"="Direction") %>%
#   left_join(anno,by="cpg") %>%
#   dplyr::filter(Probe_maf<0.05|is.na(Probe_maf))%>%
#   mutate(adjP_mqtl_s = p.adjust(p_mqtl_s, method = "bonferroni", n = length(p_mqtl_s)))%>%
#   dplyr::select(cpg,effect_mqtl_s,p_mqtl_s,adjP_mqtl_s,direction_mqtl_s)
# 
# metaCompare = meta %>%
#   inner_join(meta_s,by="cpg")%>%
#   distinct(cpg,.keep_all = TRUE) %>%
#   filter(p_mqtl < 0.05 | p_mqtl_s<0.05)
#  
# max = max(c(metaCompare$effect_mqtl,metaCompare$effect_mqtl_s)*1.2)
# min = min(c(metaCompare$effect_mqtl,metaCompare$effect_mqtl_s)*1.2)
# 
# metaCompare %>%
#   ggplot(aes(x=effect_mqtl,y=effect_mqtl_s))+
#   geom_point()+
#   geom_abline(aes(intercept=0,slope=1),alpha=0.4,color="red")+
#   geom_hline(aes(yintercept=0),color="red")+geom_vline(aes(xintercept=0),color="red")+
#   xlim(min,max)+ylim(min,max)+
#   xlab("beta value, with SNP control") + ylab("beta value, sensitivity analysis, with SNP control")+
#   ggtitle("comparison of beta coefficients in sensitivity analysis, with SNP control")+
#   theme_bw()+
#   theme(plot.title = element_text(hjust = 0.5))
# ggsave(paste0(platform,"sensitivity.mqtl.beta.png"), width = 8,height = 6)
# 
# ###############
# max = max(c(-log10(metaCompare$p_mqtl),-log10(metaCompare$p_mqtl_s)))*1.1
# min = min(c(-log10(metaCompare$p_mqtl),-log10(metaCompare$p_mqtl_s)))*0.9
# 
# metaCompare %>%
#   mutate(logP_mqtl=-log10(p_mqtl),logP_mqtl_s=-log10(p_mqtl_s))%>%
#   ggplot(aes(x=logP_mqtl,y=logP_mqtl_s))+
#   geom_point()+
#   geom_abline(aes(intercept=0,slope=1),alpha=0.4,color="red")+
#   geom_hline(aes(yintercept=0),color="red")+geom_vline(aes(xintercept=0),color="red")+
#   xlim(0,max)+ylim(0,max)+
#   xlab("log P value, with SNP control") + ylab("log P value, sensitivity analysis, with SNP control")+
#   ggtitle("comparison of P values in sensitivity analysis, with SNP control")+
#   theme_bw()+
#   theme(plot.title = element_text(hjust = 0.5))
# ggsave(paste0(platform,"sensitivity.mqtl.logP.png"), width = 8,height = 6)

```

# check SNPs that are significant in birthweight GWAS 

```{r check SNPs and their relationship with CpGs}

setwd("~/mQTL/birthweight")
library(tidyverse)
library(data.table)

snps = fread("gwas-association-downloaded_2021-07-20-EFO_0004344.tsv") %>%
  dplyr::distinct(SNPS,.keep_all=TRUE) %>%
  dplyr::mutate(id = paste0("chr",CHR_ID,":",CHR_POS,":"))%>%
  dplyr::arrange(CHR_ID,CHR_POS) %>%
  dplyr::select(SNPS,CHR_ID,CHR_POS,id)
#write.table(snps,"bw.gwas.hits.txt",quote=FALSE,sep='\t')

# number of SNPs that are identified as mQTLs
all1SNPs = fread("~/mQTL/mqtl.k.EPIC.sim")
all2SNPs = fread("~/mQTL/mqtl.k.450K.sim")
allSNPs = unique(c(all1SNPs$SNP,all2SNPs$SNP)) %>%
  gsub("\\w:\\w$","",.,perl=TRUE)
length(intersect(snps$id,allSNPs))

p1NG = fread("bw.Nomqtl.450K.meta.csv")%>%
    dplyr::rename("Effect_noMQTL"="Effect","P_noMQTL"="P-value","Direction_noMQTL"="Direction",
                  "nomqtl_fdr"="fdr") %>%
    dplyr::select(CpG,Effect_noMQTL,P_noMQTL,Direction_noMQTL)

p1 = fread("bw.mqtl.450K.meta.csv")%>%
    dplyr::rename("Effect_MQTL"="Effect","P_MQTL"="P-value","Direction_MQTL"="Direction") %>%
    dplyr::mutate(set1_snpID=set1_snp,
                  set2_snpID=set2_snp) %>%
    separate(set1_snp,c("set1_chr","set1_pos","set1_ref","set1_alt"),sep = ":") %>%
    separate(set2_snp,c("set2_chr","set2_pos","set2_ref","set2_alt"),sep = ":") %>%
    dplyr::mutate(set1_id=paste0(set1_chr,":",set1_pos,":"),
                  set2_id=paste0(set2_chr,":",set2_pos,":"))%>%
    dplyr::filter(set1_id %in% snps$id |set2_id %in% snps$id ) %>%
    inner_join(p1NG,by="CpG")%>%
    dplyr::select(CpG,Effect_MQTL,P_MQTL,Direction_MQTL,Effect_noMQTL,P_noMQTL)

p2NG = fread("bw.Nomqtl.EPIC.meta.csv")%>%
    dplyr::rename("Effect_noMQTL"="Effect","P_noMQTL"="P-value","Direction_noMQTL"="Direction",
                  "nomqtl_fdr"="fdr") %>%
    dplyr::select(CpG,Effect_noMQTL,P_noMQTL,Direction_noMQTL)
  
p2 = fread("bw.mqtl.EPIC.meta.csv")%>%
    dplyr::rename("Effect_MQTL"="Effect","P_MQTL"="P-value","Direction_MQTL"="Direction") %>%
    dplyr::mutate(set3_snpID=set3_snp,
                  set4_snpID=set4_snp) %>%
    separate(set3_snp,c("set3_chr","set3_pos","set3_ref","set3_alt"),sep = ":") %>%
    separate(set4_snp,c("set4_chr","set4_pos","set4_ref","set4_alt"),sep = ":") %>%
    dplyr::mutate(set3_id=paste0(set3_chr,":",set3_pos,":"),
                  set4_id=paste0(set4_chr,":",set4_pos,":"))%>%
    dplyr::filter(set3_id %in% snps$id |set4_id %in% snps$id ) %>%
    inner_join(p2NG,by="CpG")%>%
    dplyr::select(CpG,Effect_MQTL,P_MQTL,Direction_MQTL,Effect_noMQTL,P_noMQTL)

nrow(p2)

###############
platform = "450K"
#max = max(c(p1$Effect_MQTL,p1$Effect_noMQTL)*1.2)
#min = min(c(p1$Effect_MQTL,p1$Effect_noMQTL)*1.2)
p1 %>%
  ggplot(aes(x=Effect_noMQTL,y=Effect_MQTL))+
  geom_point(alpha = 1,color='Turquoise',shape=17,size=2)+
  geom_abline(aes(intercept=0,slope=1),alpha=0.4,color="red")+
  geom_hline(aes(yintercept=0),alpha=0.4,color="red")+
  geom_vline(aes(xintercept=0),alpha=0.4,color="red")+
  #xlim(min,max)+
  #ylim(min,max)+
  xlab("regression coefficient values, without mQTL") + 
  ylab("regression coefficient values, with mQTL")+
  theme_bw()+
  theme(axis.text.x = element_text(size=12),
          axis.text.y = element_text(size=12),
          axis.title.x = element_text(size=14),
          axis.title.y = element_text(size=14),
          plot.margin=unit(c(5.5, 19, 5.5, 5.5), "points"))
ggsave(paste0(platform,".GWASHitseffectCompSig.png"),width = 8,height = 6)

#max = max(c(-log10(p1$P_MQTL),-log10(p1$P_noMQTL)))*1.1
#min = min(c(-log10(p1$P_MQTL),-log10(p1$P_noMQTL)))*0.9

p1 %>%
  mutate(logP_nomqtl=-log10(P_noMQTL),logP_mqtl=-log10(P_MQTL))%>%
  ggplot(aes(x=logP_nomqtl,y=logP_mqtl))+
  geom_point(alpha = 1,color='Turquoise',shape=17,size=2)+
  geom_abline(aes(intercept=0,slope=1),alpha=0.4,color="red")+
  geom_hline(aes(yintercept=0),color="red",alpha=0.4)+
  geom_vline(aes(xintercept=0),color="red",alpha=0.4)+
  #xlim(0,max)+ylim(0,max)+
  xlab("-log10 (P-value), without mQTL") + 
  ylab("-log10 (P-value), with mQTL")+
  #ggtitle(paste0("Significance with or without mQTL, BW GWAS hits, platform ",platform))+
  theme_bw()+
  theme(axis.text.x = element_text(size=12),
          axis.text.y = element_text(size=12),
          axis.title.x = element_text(size=14),
          axis.title.y = element_text(size=14),
          plot.margin=unit(c(5.5, 19, 5.5, 5.5), "points"))
ggsave(paste0(platform,".GWASHitspCompSig.png"),width = 8,height = 6)


###############
platform = "EPIC"
#max = max(c(p2$Effect_MQTL,p2$Effect_noMQTL)*1.2)
#min = min(c(p2$Effect_MQTL,p2$Effect_noMQTL)*1.2)

p2 %>%
  ggplot(aes(x=Effect_noMQTL,y=Effect_MQTL))+
  geom_point(alpha = 1,color='Turquoise',shape=17,size=2)+
  geom_abline(aes(intercept=0,slope=1),alpha=0.4,color="red")+
  geom_hline(aes(yintercept=0),color="red",alpha=0.4)+
  geom_vline(aes(xintercept=0),color="red",alpha=0.4)+
  #xlim(min,max)+
  #ylim(min,max)+
  xlab("regression coefficient values, without mQTL") + 
  ylab("regression coefficient values, with mQTL")+
  #ggtitle(paste0("Effects with or without mQTL, BW GWAS hits, platform ",platform))+
  theme_bw()+
  theme(axis.text.x = element_text(size=12),
          axis.text.y = element_text(size=12),
          axis.title.x = element_text(size=14),
          axis.title.y = element_text(size=14),
          plot.margin=unit(c(5.5, 19, 5.5, 5.5), "points"))
ggsave(paste0(platform,".GWASHitseffectCompSig.png"),width = 8,height = 6)

#max = max(c(-log10(p2$P_MQTL),-log10(p2$P_noMQTL)))*1.1
#min = min(c(-log10(p2$P_MQTL),-log10(p2$P_noMQTL)))*0.9

p2 %>%
  mutate(logP_nomqtl=-log10(P_noMQTL),logP_mqtl=-log10(P_MQTL))%>%
  ggplot(aes(x=logP_nomqtl,y=logP_mqtl))+
  geom_point(alpha = 1,color='Turquoise',shape=17,size=2)+
  geom_abline(aes(intercept=0,slope=1),alpha=0.4,color="red")+
  geom_hline(aes(yintercept=0),color="red",alpha=0.4)+
  geom_vline(aes(xintercept=0),color="red",alpha=0.4)+
  #xlim(0,max)+ylim(0,max)+
  xlab("-log10 (P-value), without mQTL") + 
  ylab("-log10 (P-value), with mQTL")+
  #ggtitle(paste0("Significance with or without mQTL, BW GWAS hits, platform ",platform))+
  theme_bw()+
  theme(axis.text.x = element_text(size=12),
          axis.text.y = element_text(size=12),
          axis.title.x = element_text(size=14),
          axis.title.y = element_text(size=14),
          plot.margin=unit(c(5.5, 19, 5.5, 5.5), "points"))
ggsave(paste0(platform,".GWASHitspCompSig.png"),width = 8,height = 6)

##################################################################################
###### comparing SNP corrected vs none
platform = "all"

meta = fread(paste0("bw.mqtl.",platform,".meta.csv")) %>%
  dplyr::rename("effect_mqtl"="Effect","p_mqtl"="P-value",
            "direction_mqtl"="Direction","mqtl_fdr"="fdr") %>%
  dplyr::select(CpG,effect_mqtl,p_mqtl,mqtl_fdr,direction_mqtl)
metaNG = fread(paste0("bw.Nomqtl.",platform,".meta.csv")) %>%
  dplyr::rename("effect_nomqtl"="Effect","p_nomqtl"="P-value",
            "direction_nomqtl"="Direction","fdr_nomqtl"="fdr") %>%
  dplyr::select(CpG,effect_nomqtl,p_nomqtl,fdr_nomqtl,direction_nomqtl)
  
# select probes that are significant either with or without SNP correction
metaCompare = meta %>%
  inner_join(metaNG,by="CpG")
  #filter(p_mqtl < 0.05 | p_nomqtl<0.05) %>%
  #filter(p_mqtl>0 & p_nomqtl>0 )

###############
#max = max(c(metaCompare$ffect_nomqtl,metaCompare$effect_mqtl)*1.2)
#min = min(c(metaCompare$ffect_nomqtl,metaCompare$effect_mqtl)*1.2)

metaCompare %>%
  ggplot(aes(x=effect_nomqtl,y=effect_mqtl))+
  geom_point()+
  geom_abline(aes(intercept=0,slope=1),alpha=0.4,color="red")+
  geom_hline(aes(yintercept=0),color="red")+geom_vline(aes(xintercept=0),color="red")+
  #xlim(min,max)+ylim(min,max)+
  xlab("beta value, without mQTL") + ylab("beta value, with mQTL")+
  #ggtitle(paste0("Effects with or without mQTL, sig results, platform ",platform))+
  theme_bw()+
  theme(plot.title = element_text(hjust = 0.5))
ggsave(paste0(platform,".GWAS.effectCompSig.png"),width = 8,height = 6)

  ###############
  #max = max(c(-log10(metaCompare$p_nomqtl),-log10(metaCompare$p_mqtl)))*1.1
  #min = min(c(-log10(metaCompare$p_nomqtl),-log10(metaCompare$p_mqtl)))*0.9
  
  metaCompare %>%
    mutate(logP_nomqtl=-log10(p_nomqtl),logP_mqtl=-log10(p_mqtl))%>%
    ggplot(aes(x=logP_nomqtl,y=logP_mqtl))+
    geom_point()+
    geom_abline(aes(intercept=0,slope=1),alpha=0.4,color="red")+
    geom_hline(aes(yintercept=0),color="red")+geom_vline(aes(xintercept=0),color="red")+
    #xlim(0,max)+ylim(0,max)+
    xlab("log P value, without mQTL") + ylab("log P value, with mQTL")+
    ggtitle(paste0("Significance with or without mQTL, sig results, platform ",platform))+
    theme_bw()+
    theme(plot.title = element_text(hjust = 0.5))
  ggsave(paste0(platform,"GWAS.pCompSig.png"),width = 8,height = 6)

  ### genes that overlaap CpGs with GWAS hits as mQTLs:
  genes_cpg = meta %>%
    dplyr::select(CpG) %>%
    inner_join(annoEPIC,by=c("CpG"="cpg")) %>%
    mutate(geneName = gsub(";.*$","",UCSC_RefGene_Name))%>%
    filter(geneName!="") %>%
    dplyr::select(geneName) %>%
    as.data.frame() 

  genes_cpg$geneName %>% unique() %>% as.data.frame()
  
  
#GRB10
#MIER1
#GNA12
#LINC01483
#TOLLIP
#AS3MT
#MAD1L1
#C10orf32-ASMT
#FAM162B
#SMAD3
#FES
#SLC2A1-AS1
#RBCK1
#PHF20
#H19
#LINCR-0003
#WDR78
#WT1
#SHBG
#TFDP1
#CDC37L1
#ZNF512B
#APOLD1
#NCOA6
#CAMK2B
#TRIM8
#HKDC1
#MIR423
#EDEM2
#WDR3
#LOC100507346
#LOC338799
#COL27A1
#C9
#SOX15
#SAPS2
#C6orf48
#PLCB3
#NT5C2
#C19orf10
#C14orf169
#ZBTB38
#DLG4
#KDM4B
#KIAA1407
#ASPSCR1
#MAPK8IP2
#IGF1
#PTCH1
#CLDN7
#UGT3A2
#PAPD5
#FAM46C
#HMGA1
#TFDP2
#ZBTB9
#FAM184B
#ASGR1
#EPAS1
#MIR4513
#MED18
#NF1
#ARRDC4
#EHMT2
#HLA-DQB2
#TEF
#XKR8
#GATA6
#LRRC28
#WNT3A
#FURIN
#RGS3
#IGF1R
#LOC100131801
#PACSIN1
#RANBP3L
#CHST1
#CEP55
#DUS2
#IGF2BP3
#ADCY5
#NKX6-3
#NEURL
#ZNF169
#AIG1
#HCCA2
#SDPR
#HSPA1L
#AMFR
#BAT2
#CHRD
#PRDX5
#CCND2
  
```

(in europeans only)

```{r Preparing files for ewas to run}

library(tidyverse)
library(data.table)

setwd("~/mQTL/birthweight")

winsorize <- function(methylation,pct=winsorize.pct) {
  quantiles <- matrixStats::rowQuantiles(methylation, probs=c(pct,1-pct), na.rm=T)
  low <- quantiles[,1]
  upper <- quantiles[,2]

  outliers.lower <- rowSums(methylation < low, na.rm=T)
  outliers.upper <- rowSums(methylation > upper, na.rm=T)
  
  idx <- which(methylation < low, arr.ind=T)
  methylation[idx] <- low[idx[,1]]
  
  idx <- which(methylation > upper, arr.ind=T)
  methylation[idx] <- upper[idx[,1]]

  n <- rowSums(!is.na(methylation))
  log <- data.frame(outliers.lower, outliers.upper, n)
  return(list(methylation=methylation, log=log))
}

gwas.hits = fread("gwas.snps.bw",header=FALSE) %>%
  dplyr::select(V1)

race="eur"
for(set in c(1,2,3,4)){

 if(set==1){
   var = "birthWt"
   clinical0 = read.csv(paste0("~/sets/set",set,"/clinical_variables.csv")) %>%
     filter(Trisomy21.cnm==0)%>%
     filter(race==1) 
    }
  
 if(set==2){
  var = "birthWt"
  clinical0 = read.csv(paste0("~/sets/set",set,"/clinical_variables.csv")) %>%
      filter(Trisomy21.cnm==0)%>%
      filter(race==1) 
  }
  
 if(set==3){
   var = "Y_birthweight"
   clinical0 = read.csv(paste0("~/sets/set",set,"/clinical_variables.csv")) %>% 
     filter(Trisomy21==0)%>%
     filter(race==1) %>%
     mutate(gestage=Y_gestation_week)
   }
  
 if(set==4){
   var = "ch_birthwt_bc_full"
   clinical0 = read.csv(paste0("~/sets/set",set,"/clinical_variables.csv")) %>%
     filter(smp_type=="bg",Trisomy21.cnm==0)%>%
     filter(race==1) 
   }
  
###################################
# Determine subjects to involve

## Methylation data
####noob
if(set!=4){methy_tm_0 = read_rds(paste0("~/sets/set",set,"/beta_ritu_imputed_set",set,".rds"))}
if(set==4){methy_tm_0 = read_rds(paste0("~/sets/set",set,"/beta_funnorm_bmiq_imputed_set",set,".rds"))}
rownames(methy_tm_0)=c()
metMemDF = data.frame(beadPosition=colnames(methy_tm_0)[-1]) 
  
## Cov data

clinical = clinical0[,c("beadPosition","subjectId","sex","Batch","gestage","CaCo",var)] %>%
  filter(gestage>=30)%>%
  na.omit() %>%
  mutate(ID = paste0("0_",subjectId))%>%
  distinct(ID,.keep_all = TRUE)

dcp = fread(paste0("~/sets/set",set,"/idol_deconvolution.csv")) %>%
  as.data.frame() %>%
  dplyr::select(beadPosition,CD8T,CD4T,NK,Bcell,Mono,nRBC) %>%
  na.omit()

genePC = read.table(paste0("~/mQTL/ccls.set",set,".pc.eigenvec"))
colnames(genePC) = c("FID","IID","PC1","PC2","PC3","PC4","PC5","PC6","PC7","PC8","PC9","PC10")
genePC1 = genePC %>%
  dplyr::mutate(ID = paste0(FID,"_",IID)) %>%
  dplyr::select(ID,PC1,PC2,PC3,PC4,PC5,PC6,PC7,PC8,PC9,PC10)

clinical1 = clinical %>%
  inner_join(metMemDF,by="beadPosition") %>%
  inner_join(dcp,by="beadPosition") %>%
  inner_join(genePC1,by="ID") %>%
  dplyr::mutate(Batch=as.factor(Batch)) 
covMem = clinical1$ID

## Genome data
genMem = read.table(paste0("/ccls/proj_circle2_p3/users/sebastian/bySET/ccls.gwas.set",set,".fam")) %>%
    dplyr::mutate(ID = paste0(V1,"_",V2)) %>%
    dplyr::select(ID)
genMems = genMem$ID

## Methylation data
mtqls0.CpG = fread(paste0("~/mQTL/ccls.set",set,".mqtls.hd")) %>% 
  as.data.frame() %>%
  mutate(snp_id = SNP) %>%
  separate(snp_id,c("chr","pos","alt","ref"),sep=":") %>%
  dplyr::mutate(id = paste0(chr,":",pos,":"))%>%
  dplyr::select(-alt,-ref,-pos,-chr) %>%
  inner_join(gwas.hits,by=c("id"="V1"))%>%
  dplyr::select(CpG)

methy_tm =  methy_tm_0 %>%
  as.data.frame() %>% 
  column_to_rownames("probeId") %>%
  data.matrix() 
replace.outliers <- winsorize(methy_tm,0.005)

methy <- replace.outliers$methylation %>% 
  as.data.frame() %>%
  rownames_to_column(var="probeId") %>%
  dplyr::filter(probeId%in%mtqls0.CpG$CpG)

## Choosing overlaps, respecting the sequence of genetics, also outputting a list of subjects dropped from genetics
miInd = which(!genMems%in%covMem) %>% unique()

if(length(miInd)!=0){
  finalMems = data.frame(ID = genMems[-miInd])
}else{
  finalMems = data.frame(ID = genMems)}

###################################
# Preparing covariates
clinical2 = clinical1 %>%
  dplyr::select(-beadPosition, -subjectId) %>%
  dplyr::select(ID, one_of(var),everything()) 

write.table(clinical2,paste0("covsSet.",race,".",set,".txt"),
            sep='\t',quote=FALSE,row.names = FALSE,col.names = TRUE)

###################################
# Preparing methylation files
subs = clinical1 %>% column_to_rownames(var="ID")
subs1 = subs[as.character(finalMems$ID),"beadPosition"] %>% as.character()
methy1 = methy[,c("probeId",subs1)]
colnames(methy1) = c("probeId",as.character(finalMems$ID))

write.table(methy1,paste0("methSet",race,".",set,".txt"),sep='\t',quote=FALSE,row.names = FALSE,col.names = TRUE)


print(set)
print((dim(methy1)[2]-1)==(dim(clinical2)[1]))

}




```

```{r adjusting  for scanned mqtls with QTLtools}

setwd("~/mQTL/birthweight")
library(tidyverse)
library(data.table)
library(MASS)
library(lmtest)
library(sandwich)

linear <- function(meth,cov){

  #    write_rds(meth,"meth.test.rds")
  #}
  
  # meth = read_rds("meth.test.rds")
  # cov = covs

  probe <- unname(meth[1])
  methDF = meth[-1] %>% as.numeric()
  
  dat <- data.frame(y = methDF, cov)
  ff  <- formula(paste("y ~ ", paste0(colnames(cov),collapse="+")))
  #mod <- try(rlm(ff, data = dat, maxit = 200),silent = TRUE)
  mod <- try(lm(ff, data = dat),silent = TRUE)
  
  #cf = try(coeftest(mod),silent = TRUE)
  N <- sum(!is.na(dat[,1]))
  
  if (class(mod)=="lm") {
    
    coefs = summary(mod)$coefficients
    coef = coefs[2,"Estimate"]
    SE = coefs[2,"Std. Error"]
    p.value = coefs[2,"Pr(>|t|)"]
    out = c(coef, SE, p.value, N) 
  } else {
    out = c(rep(NA,3), N)
  }
  names(out) = c("Beta","SE", "P", "N"); return(out) 
}
gene.linear <- function(meth,cov,gene){
  
  #    write_rds(meth,"meth.test.rds")
  #}
  
  # meth = read_rds("meth.test.rds")
  # cov = covs
  # gene = mtqls
  
  probe <- unname(meth[1])
  
  #print(probe)
  
  methDF = meth[-1] %>% as.numeric()
  
  SNPname <- gene[which(gene$CpG == probe), 1]  %>% as.character()  
  SNP <-  gene[which(gene$CpG == probe), -c(1:3)] %>% t() %>% as.data.frame()
  colnames(SNP) <- "SNP"
  
  if(length(table(SNP$SNP))==1){
    dat <- data.frame(y = methDF, cov)
    ff  <- formula(paste("y ~ ", paste0(colnames(cov),collapse="+")))
  }else{
    dat <- data.frame(y = methDF, cov,SNP)
    ff  <- formula(paste("y ~ ", paste0(c(colnames(cov),"SNP"),collapse="+")))
  }
  
  mod <- try(lm(ff, data = dat),silent = TRUE)
  
  #cf = try(coeftest(mod),silent = TRUE)
  N <- sum(!is.na(dat[,1]))
  
  if (class(mod)=="lm") {
    
    SNPname=SNPname
    #pvals = try(f.robftest(mod, var = colnames(cov)[1]),silent = TRUE)
    #coef = cf[2,"Estimate"]
    #F.value = ifelse(class(pvals)=="htest",pvals$statistic,NA)
    #p.value = ifelse(class(pvals)=="htest",pvals$p.value,NA)
    #out = c(SNPname,coef, F.value, p.value, N) 

    coefs = summary(mod)$coefficients
    coef = coefs[2,"Estimate"]
    SE = coefs[2,"Std. Error"]
    p.value = coefs[2,"Pr(>|t|)"]
    out = c(SNPname,coef, SE, p.value, N) 
  } else {
    out = c(SNPname,rep(NA,3), N)
  }
  names(out) = c("snp","Beta","SE", "P", "N"); return(out) 
}

race="eur"

gwas.hits = fread("gwas.snps.bw",header=FALSE) %>%
  dplyr::select(V1)

for (set in c(1:4)){
  
 if(set==1){var = "birthWt"}
 if(set==2){var = "birthWt"}
 if(set==3){var = "Y_birthweight"}
 if(set==4){var = "ch_birthwt_bc_full"}
  

  mtqls0 = fread(paste0("~/mQTL/ccls.set",set,".mqtls.hd")) %>% 
    as.data.frame() %>%
    mutate(snp_id = SNP) %>%
    separate(snp_id,c("chr","pos","alt","ref"),sep=":") %>%
    dplyr::mutate(id = paste0(chr,":",pos,":"))%>%
    dplyr::select(-alt,-ref,-pos,-chr) %>%
    inner_join(gwas.hits,by=c("id"="V1"))%>%
    dplyr::select(-id)
  
  methyl0 = fread(paste0("methSet",race,".",set,".txt")) %>%
    column_to_rownames(var="probeId")%>%
    as.data.frame()

  covs0 = fread(paste0("covsSet",".",race,".",set,".txt"))%>%
    as.data.frame()%>%
    column_to_rownames(var="ID") %>%
    mutate(Batch = as.factor(Batch))
  
  subs.list = intersect(intersect(rownames(covs0),colnames(methyl0)), colnames(mtqls0)) 
  
  methyl = methyl0[,subs.list] %>% rownames_to_column(var="cpg")
  rownames(methyl) <- methyl$cpg
  
  covs = covs0[subs.list,]
  mtqls = mtqls0[,c("SNP","CpG","P-value",subs.list)]

# no correction for SNP
  
  results <- t(apply(methyl, 1, linear, covs)) 
  results <- as.data.frame(results) %>%
    rownames_to_column(var="CpG") %>%
    mutate(N=as.numeric(N))%>%
    mutate(P=as.numeric(P))%>%
    filter(N>max(N)/2,
           !is.na(P))

  write.table(results,paste0("ccls.set",set,".",var,".",race,".bwGWAS.Nogene.txt"),
          row.names = FALSE,
          quote=FALSE,
          sep='\t')  

# correction for SNP
  
  probes = rownames(methyl)[which(rownames(methyl)%in%mtqls$CpG)] 

  methyl.cpg = methyl[probes,] 
    
  results.snp <- t(apply(methyl.cpg, 1, gene.linear, covs, mtqls)) 
  results.snp <- as.data.frame(results.snp)%>%
    rownames_to_column(var="CpG")%>%
    mutate(N=as.numeric(N))%>%
    mutate(P=as.numeric(P))%>%
    filter(N>max(N)/2,
           !is.na(P))
  
  write.table(results.snp,
              paste0("ccls.set",set,".",var,".",race,".bwGWAS.geneCon.txt"),
              row.names = FALSE,
              quote=FALSE,
              sep='\t')  


      }

```

```{bash meta-analysis}

## Meta-analysis of all four datasets

cd ~/mQTL/birthweight

### all four sets
metal
MARKER CpG
EFFECT Beta
PVALUE P
SCHEME STDERR
STDERR SE
VERBOSE OFF
GENOMICCONTROL OFF
STDERR_PRINT_PRECISION 20
EFFECT_PRINT_PRECISION 20
PROCESS ccls.set1.birthWt.eur.bwGWAS.geneCon.txt
PROCESS ccls.set2.birthWt.eur.bwGWAS.geneCon.txt
PROCESS ccls.set3.Y_birthweight.eur.bwGWAS.geneCon.txt
PROCESS ccls.set4.ch_birthwt_bc_full.eur.bwGWAS.geneCon.txt
ANALYZE
QUIT
mv METAANALYSIS1.TBL meta.all.bw.geneCon.txt

### all four sets, noGene
metal
MARKER CpG
EFFECT Beta
PVALUE P
SCHEME STDERR
STDERR SE
VERBOSE OFF
GENOMICCONTROL OFF
STDERR_PRINT_PRECISION 20
EFFECT_PRINT_PRECISION 20
PROCESS ccls.set1.birthWt.eur.bwGWAS.Nogene.txt
PROCESS ccls.set2.birthWt.eur.bwGWAS.Nogene.txt
PROCESS ccls.set3.Y_birthweight.eur.bwGWAS.Nogene.txt
PROCESS ccls.set4.ch_birthwt_bc_full.eur.bwGWAS.Nogene.txt
ANALYZE
QUIT
mv METAANALYSIS1.TBL meta.all.bw.Nogene.txt

```

```{r looking at results after meta}

setwd("~/mQTL/birthweight")
library(tidyverse)
library(data.table)
library(qvalue)

library(IlluminaHumanMethylation450kanno.ilmn12.hg19)
data("IlluminaHumanMethylation450kanno.ilmn12.hg19")
anno       <- IlluminaHumanMethylation450kanno.ilmn12.hg19 %>%
  getAnnotation %>%
  as.data.frame %>%
  rownames_to_column(var="cpg") %>%
  dplyr::select(cpg,chr,pos,Relation_to_Island,UCSC_RefGene_Name,UCSC_RefGene_Group)

library(IlluminaHumanMethylationEPICanno.ilm10b4.hg19)
data("IlluminaHumanMethylationEPICanno.ilm10b4.hg19")
annoEPIC       <- IlluminaHumanMethylationEPICanno.ilm10b4.hg19 %>%
  getAnnotation %>%
  as.data.frame %>%
  rownames_to_column(var="cpg") %>%
  dplyr::select(cpg,chr,pos,Relation_to_Island,UCSC_RefGene_Name,UCSC_RefGene_Group)

############################################################################################################## Counting number of significant CpGs with/without SNP controls (same CpGs with mQTL)

##platform all
platform="all"

meta = fread(paste0("meta.",platform,".bw.geneCon.txt")) %>%
  dplyr::select(-Allele1, -Allele2) %>%
  dplyr::rename("CpG" = "MarkerName") %>%
  left_join(annoEPIC,by=c("CpG"="cpg")) %>%
  mutate(`P-value`=as.numeric(`P-value`))%>%
  dplyr::filter(!is.na(`P-value`))%>%
  arrange(by=`P-value`) %>%
  mutate(qvalue = qvalue(`P-value`)$qvalues,
         bonferroni = p.adjust(`P-value`, method = "bonferroni", n = length(`P-value`)),
         fdr = p.adjust(`P-value`, method = "fdr", n = length(`P-value`)))%>%
  mutate(tss=0)
meta$tss[grep("TSS",meta$UCSC_RefGene_Group)]=1

metaNG = fread(paste0("meta.",platform,".bw.Nogene.txt")) %>%
  dplyr::select(-Allele1, -Allele2) %>%
  dplyr::rename("CpG" = "MarkerName") %>%
  left_join(annoEPIC,by=c("CpG"="cpg")) %>%
  mutate(`P-value`=as.numeric(`P-value`))%>%
  dplyr::filter(!is.na(`P-value`))%>%
  arrange(by=`P-value`) %>%
  mutate(qvalue = qvalue(`P-value`)$qvalues,
         bonferroni = p.adjust(`P-value`, method = "bonferroni", n = length(`P-value`)),
         fdr = p.adjust(`P-value`, method = "fdr", n = length(`P-value`)))%>%
  mutate(tss=0)
metaNG$tss[grep("TSS",metaNG$UCSC_RefGene_Group)]=1

#print(paste0(nrow(meta)," probes are significant after SNP correction for platform ",platform))
#print(paste0(nrow(metaNG)," probes are significant without SNP correction for platform ",platform))
#print(paste0("distributiono of TSS",platform,", with mQTL"))
#print(table(meta$tss))
#print(table(meta$tss)[2]/(table(meta$tss)[1]+table(meta$tss)[2]))
#print(paste0("distributiono of TSS",platform,", without mQTL"))
#print(table(metaNG$tss))
#print(table(metaNG$tss)[2]/(table(metaNG$tss)[1]+table(metaNG$tss)[2]))

#write.csv(metaNG,paste0("bw.Nomqtl.",platform,".meta.csv"))
#write.csv(meta,paste0("bw.mqtl.",platform,".meta.csv"))
  
# [1] "139 probes are significant after SNP correction for platform all"
# [1] "139 probes are significant without SNP correction for platform all"
# [1] "distributiono of TSSall, with mQTL"
#  0  1 
# 95 44 
# 0.3165468 
# 
# [1] "distributiono of TSSall, without mQTL"
#  0  1 
# 95 44 
#         1 
# 0.3165468 


```








# NOT INCLUDED IN THE PAPER
>> Sensitivity analysis: including POBW in the model

```{r preparing data}

library(tidyverse)
library(data.table)

setwd("~/mQTL/birthweight")

pobw = read.csv("sourceF/clin_var_set4_POBW.csv") %>%
    dplyr::mutate(setN=set) %>%
    dplyr::select(beadPosition,POBW,setN)

var = "POBW"

#for(set in c(1,2,4)){
 set=4
 
  pobw1 = pobw %>% 
    as.data.frame()%>%
    dplyr::filter(setN==paste0("set",set)) %>% 
    dplyr::select(-setN) %>%
    na.omit()

 if(set==1){
   var = "birthWt"
   clinical0 = read.csv(paste0("~/sets/set",set,"/clinical_variables.csv")) %>%
     filter(Trisomy21.cnm==0)%>%
      mutate(plate = as.factor(gsub("_.*$","",beadPosition))) %>%
      inner_join(pobw1, by="beadPosition")
    }
  
 if(set==2){
  var = "birthWt"
  clinical0 = read.csv(paste0("~/sets/set",set,"/clinical_variables.csv")) %>%
      filter(Trisomy21.cnm==0)%>%
      mutate(plate = as.factor(gsub("_.*$","",beadPosition))) %>%
      inner_join(pobw1, by="beadPosition")
  }
  
 if(set==4){
   clinical0 = read.csv(paste0("~/sets/set",set,"/clinical_variables.csv")) %>%
     filter(smp_type=="bg",Trisomy21.cnm==0)%>%
     mutate(plate = as.factor(gsub("_.*$","",beadPosition))) %>%
      inner_join(pobw1, by="beadPosition")
   }
  
 
###################################
# Determine subjects to involve
## Cov data
clinical = clinical0[,c("beadPosition","subjectId","sex","plate","gestage","CaCo",var)] %>%
  dplyr::filter(gestage>=30)%>%
  na.omit() %>%
  dplyr::mutate(ID = paste0("0_",subjectId))%>%
  distinct(ID,.keep_all = TRUE)
refPC = read.csv(paste0("~/sets/set",set,"/glintControl.csv")) %>%
  dplyr::select(beadPosition,rc1,rc2,rc3,rc4,rc5,rc6,rc7,rc8,rc9,rc10)
genePC = read.table(paste0("~/mQTL/ccls.set",set,".pc.eigenvec"))
colnames(genePC) = c("FID","IID","PC1","PC2","PC3","PC4","PC5","PC6","PC7","PC8","PC9","PC10")
genePC1 = genePC %>%
  dplyr::mutate(ID = paste0(FID,"_",IID)) %>%
  dplyr::select(ID,PC1,PC2,PC3,PC4,PC5,PC6,PC7,PC8,PC9,PC10)
clinical1 = clinical %>%
  inner_join(refPC,by="beadPosition") %>%
  inner_join(genePC1,by="ID") %>%
  mutate(plate=as.factor(plate)) 
covMem = clinical1$ID

## Genome data
genMem = read.table(paste0("/ccls/proj_circle2_p3/users/sebastian/bySET/ccls.gwas.set",set,".fam")) %>%
  dplyr::mutate(ID = paste0(V1,"_",V2)) %>%
  dplyr::select(ID)

genMems = genMem$ID
## Methylation data
methy = read_rds(paste0("~/sets/set",set,"/beta_noob_imputed_set",set,".rds"))
metMemDF = data.frame(beadPosition=colnames(methy)[-1]) %>% 
  dplyr::left_join(clinical1,by="beadPosition") %>% 
  dplyr::select(ID)
metMem = metMemDF$ID
## Choosing overlaps, respecting the sequence of genetics, also outputting a list of subjects dropped from genetics
miInd = c( which(!genMems%in%metMem),
          which(!genMems%in%covMem) ) %>% unique()

if(length(miInd)!=0){
  finalMems = data.frame(ID = genMems[-miInd])
}else{
  finalMems = data.frame(ID = genMems)}
dropped = setdiff(as.character(genMem$ID),as.character(finalMems$ID))
write.table(dropped,paste0("dropped.from.Set",set,".s2.txt"), sep='\t', quote=FALSE,row.names = FALSE,col.names = FALSE)

###################################
# Preparing covariates

clinical2 = clinical1 %>%
  dplyr::select(-beadPosition, -subjectId) %>%
  dplyr::select(one_of(var),everything()) %>%
  column_to_rownames(var="ID")

clinical3 = clinical2[as.character(finalMems$ID),] %>%
  t()%>%
  as.data.frame() %>%
  rownames_to_column(var="id")

write.table(clinical3,paste0("covsSet",set,".s2.txt"),sep='\t',quote=FALSE,row.names = FALSE,col.names = TRUE)


###################################
# Preparing methylation files

subs = clinical1 %>%
  column_to_rownames(var="ID")
subs1 = subs[as.character(finalMems$ID),"beadPosition"] %>% as.character()
methy1 = methy[,c("probeId",subs1)]
colnames(methy1) = c("probeId",as.character(finalMems$ID))

write.table(methy1,paste0("methSet",set,".s2.txt"),sep='\t',quote=FALSE,row.names = FALSE,col.names = TRUE)

print(set)
dim(methy1)[2]-1
dim(clinical3)[2]-1

#}
```

```{bash preparing generic data}

ssh sli@n0001

mkdir /tmp/sli/s2
cd /tmp/sli/s2

scp -v /ccls/home/sli/mQTL/dumm/ccls.set4.chr*.dum.raw .
scp -v /ccls/home/sli/mQTL/mqtl.k.EPIC.sim .
scp -v /ccls/home/sli/mQTL/birthweight/dropped.from.Set4.s2.txt .

# cd ~/mQTL/birthweight

set=4
# Exclude PAT MAT SEX PHENOTYPE columns
  for chr in {1..22}; do
    paste <(awk '{print $1"_"$2}' ccls.set$set.chr$chr.dum.raw | sed 's/FID_IID/snpid/') \
      <(cut -d' ' -f7- ccls.set$set.chr$chr.dum.raw) | \
      grep -v -w -f dropped.from.Set$set.s2.txt \
      > ccls.set$set.chr$chr.dum.int
  done
  
  # Use -W to treat one or more consecutive whitespace characters as field delimiters
  for chr in {1..22}; do
    datamash -W transpose <ccls.set$set.chr$chr.dum.int |\
      grep -v 'NA' > ccls.set$set.chr$chr.dum.tsp
  done

# cd ~/mQTL/birthweight

# Combining all dummy genotype data into 1 dataframe
set=4
  head -1 ccls.set$set.chr1.dum.tsp > ccls.set$set.dum.tsp
  tail -n +2 -q ccls.set$set.chr*.dum.tsp >> ccls.set$set.dum.tsp

# Changing the name of SNPs to make it concordant with mqtl file
set=4
  awk '{gsub(/_.*$/,"",$1); print $0}' ccls.set$set.dum.tsp > ccls.set$set.dum.tsp.nm

set=4
  join -1 1 -2 1 <(sort -k1,1 mqtl.k.EPIC.sim)\
      <(sort -k1,1 ccls.set$set.dum.tsp.nm) |\
      sort -k2,2 -k3,3g |\
      sort -u -k2,2 \
      > ccls.set$set.mqtls
  hdr=$(echo -n  $(head -1 mqtl.k.EPIC.sim) $(head -1 ccls.set$set.dum.tsp.nm) | sed 's/snpid//g') 
  cat <(echo $hdr) ccls.set$set.mqtls > ccls.set$set.mqtls.s2.hd

cp ccls.set4.mqtls.s2.hd /ccls/home/sli/mQTL/birthweight


```

```{r adjusting  for scanned mqtls with QTLtools}

setwd("~/mQTL/birthweight")
library(tidyverse)
library(data.table)
library(MASS)
library(lmtest)
library(sandwich)
library(parallel) 
cl <- makeCluster(40, type="FORK")
clusterEvalQ(cl,library(MASS))
clusterEvalQ(cl,library(lmtest))

RLM.Robust.nogene <- function(meth, cov){
  
  # meth = methyl[,1]
  # cov = covs
  
    dat <- data.frame(y = meth, cov)
    ff  <- formula(paste("y ~ ", paste0(colnames(cov), collapse="+")))
    mod <- try(rlm(ff, data = dat, maxit = 200))
    #cf = try(coeftest(mod, vcov=vcovHC(mod, type="HC0")))
    cf = try(coeftest(mod))

  	if (exists(class(cf)) == FALSE) {
  		out = rep(NA,3)
  	} else if (exists(class(cf)) == TRUE) {
  		if( class(cf)[1] == "try-error") {
  		cat("error throw by CpG", set, ";")
  		out = rep(NA,3)
  		print(paste0(cov, " - chol2inv could not be computed"))
  		} 
  		else {
  			coef = cf[2,"Estimate"]
  			se = cf[2,"Std. Error"]
  			p.value = cf[2,"Pr(>|z|)"]
  			out = c(coef, se, p.value) 
  		}
    }
  names(out) = c("Beta","SE", "P"); return(out) 
  }


RLM.Robust.gene <- function(meth,cov,gene){
  
  # meth = read_rds("meth.test.rds")
  # cov = covs
  # gene = mtqls
  
    probe <- unname(meth[1])
    methDF = meth[-1] %>% as.numeric()
    
    SNPname <- gene[which(gene$CpG == probe), 1]  %>% as.character()  
    SNP <-  gene[which(gene$CpG == probe), -c(1:3)] %>% t() %>% as.data.frame()
    colnames(SNP) <- "SNP"

    dat <- data.frame(y = methDF, cov,SNP)
    ff  <- formula(paste("y ~ ", paste0(c(colnames(cov),"SNP"),collapse="+")))
    mod <- try(rlm(ff, data = dat, maxit = 200))
    #cf = try(coeftest(mod, vcov=vcovHC(mod, type="HC0")))
    cf = try(coeftest(mod))
    
  	if (exists(class(cf)) == FALSE) {
  		out = rep(NA,4)
  	} else if (exists(class(cf)) == TRUE) {
  		if( class(cf)[1] == "try-error") {
  		cat("error throw by CpG", set, ";")
  		out = rep(NA,4)
  		print(paste0(cov, " - chol2inv could not be computed"))
  		} 
  		else {
  		  SNP=SNPname
  			coef = cf[2,"Estimate"]
  			se = cf[2,"Std. Error"]
  			p.value = cf[2,"Pr(>|z|)"]
  			out = c(SNP,coef, se, p.value) 
  		}
  	}
  names(out) <- c("snp","Beta","SE","P"); return(out) 
  }

#for (set in c(1,2,4)){
set=4
  
var = "POBW"

  methyl = fread(paste0("methSet",set,".s2.txt")) %>%
    column_to_rownames(var="probeId")%>%
    as.data.frame()

  covs = fread(paste0("covsSet",set,".s2.txt"))%>%
    column_to_rownames(var="id")%>%
    t()%>%
    as.data.frame()%>%
    mutate(plate = as.factor(plate))

  mtqls = fread(paste0("ccls.set",set,".mqtls.s2.hd"))
  
  probes = rownames(methyl)[which(rownames(methyl)%in%mtqls$CpG)] 

# no correction for SNP
  
  results <- t(parApply(cl, methyl, 1, RLM.Robust.nogene, covs)) 
  results <- as.data.frame(results)
  
    
  write.table(results,paste0("ccls.set",set,".",var,".Nogene.s2.ewas.txt"),
            quote=FALSE,
            sep='\t')  

  Nbeta <- sum(!is.na(results$P))

  tmp <- results[!is.na(results$P),]
  #print(dim(tmp))
  
  print(paste0("lambda of set ",set," without SNP control:"))
  lambda = median(qchisq(1- tmp$P,1))/qchisq(0.5,1)
  print(round(lambda, 2))

# correction for SNP
  methyl.cpg = methyl[probes,] %>% rownames_to_column(var="cpg")
  rownames(methyl.cpg) <- methyl.cpg$cpg
  results.snp <- t(parApply(cl, methyl.cpg, 1, RLM.Robust.gene, covs, mtqls)) 

  results.snp <- as.data.frame(results.snp)
  
  write.table(results.snp,paste0("ccls.set",set,".",var,".geneCon.s2.ewas.txt"),
          quote=FALSE,
          sep='\t')  

  Nbeta.snp <- sum(!is.na(results.snp$P))
  tmp.snp <- results.snp[!is.na(results.snp$P),]
  tmp.snp$P <- as.numeric(tmp.snp$P)
  print(paste0("lambda of set ",set," with SNP control:"))
  lambda.snp = median(qchisq(1- tmp.snp$P,1))/qchisq(0.5,1)
  print(round(lambda.snp, 2))
  
#  }

  
stopCluster(cl)
rm(list=ls())


# testing covariates
meth = methyl[1,] %>% t() %>% as.data.frame
cov = covs
dat <- data.frame(y = meth, cov)
plot(dat$POBW,dat$ch_birthwt_bc_full)

```

```{bash cleaning files}

cd ~/mQTL/birthweight

awk '{print $1"_none",$2,$3,$4}' <(tail -n +2 ccls.set4.POBW.Nogene.s2.ewas.txt) |\
sed '1i CpG_snp Estimate  SE P-value' > set4.POBW.Nogene.s2.ewas.txt
awk '{print $1"_"$2,$3,$4,$5}' <(tail -n +2 ccls.set4.POBW.geneCon.s2.ewas.txt) |\
sed '1i CpG_snp Estimate  SE P-value' > set4.POBW.geneCon.s2.ewas.txt

```

```{r checking the results of sensitivity analysis}
setwd("~/mQTL/birthweight")
library(tidyverse)
library(data.table)
library(qvalue)

platform = "set4"

library(IlluminaHumanMethylationEPICanno.ilm10b4.hg19)
data("IlluminaHumanMethylationEPICanno.ilm10b4.hg19")
annoEPIC       <- IlluminaHumanMethylationEPICanno.ilm10b4.hg19 %>%
  getAnnotation %>%
  as.data.frame %>%
  rownames_to_column(var="cpg") %>%
  select(cpg,chr,pos,Relation_to_Island,UCSC_RefGene_Name,UCSC_RefGene_Group)

#########################
##### Counting number of significant CpGs with/without SNP controls (same CpGs with mQTL)

set4 = fread("set4.POBW.geneCon.s2.ewas.txt") %>%
  mutate(cpg = gsub("_.*$","",CpG_snp),
         snp = gsub("^.*_","",CpG_snp)) %>%
  arrange(by=`P-value`) %>%
  mutate(bonferroni = p.adjust(`P-value`, method = "bonferroni", n = length(`P-value`)),
         fdr = p.adjust(`P-value`, method = "fdr", n = length(`P-value`))) %>%
  filter(bonferroni < 0.05) %>%
  left_join(annoEPIC,by="cpg") %>%
  mutate(tss=0)
set4$tss[grep("TSS",set4$UCSC_RefGene_Group)]=1

set4NG = fread("set4.POBW.Nogene.s2.ewas.txt") %>%
  mutate(cpg = gsub("_.*$","",CpG_snp)) %>%
  arrange(by=`P-value`) %>%
  mutate(bonferroni = p.adjust(`P-value`, method = "bonferroni", n = length(`P-value`)),
         fdr = p.adjust(`P-value`, method = "fdr", n = length(`P-value`))) %>%
  filter(bonferroni < 0.05) %>%
  left_join(annoEPIC,by="cpg") %>%
  mutate(tss=0)
set4NG$tss[grep("TSS",set4NG$UCSC_RefGene_Group)]=1
  
print(paste0(nrow(set4)," probes are significant after SNP correction for platform ",platform))
print(paste0(nrow(set4NG)," probes are significant without SNP correction for platform ",platform))
print(paste0("distributiono of TSS",platform,", with mQTL"))
print(table(set4$tss))
print(table(set4$tss)[2]/(table(set4$tss)[1]+table(set4$tss)[2]))
print(paste0("distributiono of TSS",platform,", without mQTL"))
print(table(set4NG$tss))
print(table(set4NG$tss)[2]/(table(set4NG$tss)[1]+table(set4NG$tss)[2]))

write.csv(set4NG,paste0("POBW.Nomqtl.s2.",platform,".csv"))
write.csv(set4,paste0("POBW.mqtl.s2.",platform,".csv"))

# [1] "206 probes are significant after SNP correction for platform set4"
# [1] "193 probes are significant without SNP correction for platform set4"

# [1] "distributiono of TSS450K, with mQTL"
#  0   1 
# 167  39 
# 0.1893204 

# [1] "distributiono of TSS450K, without mQTL"
#   0   1 
# 154  39 
# 0.2020725 

###############################################################################
####top Hits with SNP control from Boydyweight EWAS and results of GSEA pathway analysis, both platforms

platform="set4"

set4 = fread("set4.POBW.geneCon.s2.ewas.txt") %>%
  mutate(cpg = gsub("_.*$","",CpG_snp),
         snp = gsub("^.*_","",CpG_snp)) %>%
  arrange(by=`P-value`) %>%
  mutate(bonferroni = p.adjust(`P-value`, method = "bonferroni", n = length(`P-value`)),
         fdr = p.adjust(`P-value`, method = "fdr", n = length(`P-value`))) %>%
  filter(bonferroni < 0.05) %>%
  left_join(annoEPIC,by="cpg") 

set4 %>% 
  filter(UCSC_RefGene_Name!="") %>% 
  head(20) %>%
  mutate(geneName=gsub(";.*$","",UCSC_RefGene_Name)) %>% 
  select(geneName)
# 1:    WDR19
# 2:      EED
# 3:  PLEKHA5
# 4:   LRRC27
# 5:    FBXL7
# 6: C1RL-AS1
# 7:     TPPP
# 8:     COBL
# 9:    RPTOR
# 10:     TEN1
# 11:    RASA3
# 12:   DLGAP3
# 13:    ESRRG
# 14:  RPS6KC1
# 15:  ARFGEF2
# 16:   RASSF3
# 17:   SPPL2B
# 18:   BAHCC1
# 19:     RELT
# 20:     VAV2
#GSEA: No overlap

set4NG = fread("set4.POBW.Nogene.s2.ewas.txt") %>%
  mutate(cpg = gsub("_.*$","",CpG_snp)) %>%
  arrange(by=`P-value`) %>%
  mutate(bonferroni = p.adjust(`P-value`, method = "bonferroni", n = length(`P-value`)),
         fdr = p.adjust(`P-value`, method = "fdr", n = length(`P-value`))) %>%
  filter(bonferroni < 0.05) %>%
  left_join(annoEPIC,by="cpg") 
set4NG %>% 
  filter(UCSC_RefGene_Name!="") %>% 
  head(20) %>%
  mutate(geneName=gsub(";.*$","",UCSC_RefGene_Name)) %>% 
  select(geneName)
#  1:    BCYRN1
#  2:   PLEKHA5
#  3:     ESRRG
#  4:      KPTN
#  5:      USP7
#  6:    DFNB31
#  7:      TEN1
#  8:    TMIGD2
#  9:    TFAP2B
# 10:     IL21R
# 11:     RASA3
# 12:   C7orf34
# 13:     WDR19
# 14: LINC00624
# 15:    IGSF9B
# 16:    ZNF37B
# 17:    OR51B6
# 18:      VAV2
# 19:     ENDOV
# 20:     COX8C
# GSEA: Genes up-regulated in bone marrow-derived dendritic cellstreated by poly(IC): 3h versus 24h.

##################################################################################
###### comparing SNP corrected vs none

platform="set4"

set4 = fread("set4.POBW.geneCon.s2.ewas.txt") %>%
  mutate(cpg = gsub("_.*$","",CpG_snp),
         snp = gsub("^.*_","",CpG_snp)) %>%
  arrange(by=`P-value`) %>%
  mutate(mqtl_bonferroni = p.adjust(`P-value`, 
                                    method = "bonferroni", 
                                    n = length(`P-value`))) %>%
  dplyr::rename("effect_mqtl"="Estimate",
            "p_mqtl"="P-value") %>%
  select(cpg,effect_mqtl,p_mqtl,mqtl_bonferroni,snp)

set4NG = fread("set4.POBW.Nogene.s2.ewas.txt") %>%
  mutate(cpg = gsub("_.*$","",CpG_snp)) %>%
  mutate(nomqtl_bonferroni = p.adjust(`P-value`, 
                                      method = "bonferroni", 
                                      n = length(`P-value`))) %>%
  dplyr::rename("effect_nomqtl"="Estimate",
                "p_nomqtl"="P-value")  %>%
  select(cpg,effect_nomqtl,p_nomqtl,nomqtl_bonferroni)

# select probes that are significant either with or without SNP correction
set4Compare = set4 %>%
  inner_join(set4NG,by="cpg")%>%
  filter(mqtl_bonferroni < 0.05 | nomqtl_bonferroni<0.05) 


###############
max = max(c(set4Compare$effect_nomqtl,set4Compare$effect_mqtl)*1.2)
min = min(c(set4Compare$effect_nomqtl,set4Compare$effect_mqtl)*1.2)

set4Compare %>%
  ggplot(aes(x=effect_nomqtl,y=effect_mqtl))+
  geom_point()+
  geom_abline(aes(intercept=0,slope=1),alpha=0.4,color="red")+
  geom_hline(aes(yintercept=0),color="red")+
  geom_vline(aes(xintercept=0),color="red")+
  xlim(min,max)+ylim(min,max)+
  xlab("beta value, without mQTL") + ylab("beta value, with mQTL")+
  ggtitle(paste0("POBW Effects with or without mQTL, sig results, platform ",platform))+
  theme_bw()+
  theme(plot.title = element_text(hjust = 0.5))
ggsave(paste0(platform,".POBW.effectCompSig.png"),width = 8,height = 6)

###############
max = max(c(-log10(set4Compare$p_nomqtl),-log10(set4Compare$p_mqtl)))*1.1
min = min(c(-log10(set4Compare$p_nomqtl),-log10(set4Compare$p_mqtl)))*0.9

set4Compare %>%
  mutate(logP_nomqtl=-log10(p_nomqtl),logP_mqtl=-log10(p_mqtl))%>%
  ggplot(aes(x=logP_nomqtl,y=logP_mqtl))+
  geom_point()+
  geom_abline(aes(intercept=0,slope=1),alpha=0.4,color="red")+
  geom_hline(aes(yintercept=0),color="red")+
  geom_vline(aes(xintercept=0),color="red")+
  xlim(0,max)+ylim(0,max)+
  xlab("log P value, without mQTL") + ylab("log P value, with mQTL")+
  ggtitle(paste0("POBW Significance with or without mQTL, sig results, platform ",platform))+
  theme_bw()+
  theme(plot.title = element_text(hjust = 0.5))
ggsave(paste0(platform,".POBW.pCompSig.png"),width = 8,height = 6)

#################################################################################
#################################################################################
###################  (gene*birthweight interaction)

################### Pathway analysis of probes whose beta changes 
library(methylGSA)

platform="set4"
  
set4 = fread("set4.POBW.geneCon.s2.ewas.txt") %>%
  mutate(cpg = gsub("_.*$","",CpG_snp),
         snp = gsub("^.*_","",CpG_snp)) %>%
  arrange(by=`P-value`) %>%
  mutate(mqtl_fdr = p.adjust(`P-value`, 
                                    method = "fdr", 
                                    n = length(`P-value`))) %>%
  dplyr::rename("effect_mqtl"="Estimate",
            "p_mqtl"="P-value") %>%
  select(cpg,effect_mqtl,p_mqtl,mqtl_fdr,snp)

set4NG = fread("set4.POBW.Nogene.s2.ewas.txt") %>%
  mutate(cpg = gsub("_.*$","",CpG_snp)) %>%
  mutate(nomqtl_fdr = p.adjust(`P-value`, 
                                      method = "fdr", 
                                      n = length(`P-value`))) %>%
  dplyr::rename("effect_nomqtl"="Estimate",
                "p_nomqtl"="P-value")  %>%
  select(cpg,effect_nomqtl,p_nomqtl,nomqtl_fdr)

# select probes that are significant either with or without SNP correction
set4Compare = set4 %>%
  inner_join(set4NG,by="cpg")%>%
  mutate(delta=abs((effect_mqtl-effect_nomqtl)/effect_nomqtl))%>%
  distinct(cpg,.keep_all = TRUE) %>%
  filter(mqtl_fdr < 0.05)%>%
  filter(delta>0.2) %>%
  select(cpg,p_mqtl) 

cpg.pval = c(set4Compare$p_mqtl)
names(cpg.pval) = set4Compare$cpg
  
res1 = methylglm(cpg.pval = cpg.pval, minsize = 10, GS.type = "KEGG",array.type = "EPIC")
res2 = methylglm(cpg.pval = cpg.pval, minsize = 50, GS.type = "GO",array.type = "EPIC")
  
write.table(res1,paste0(platform,".powb.interaction.kegg.txt"),sep = "\t", row.names = FALSE, quote = FALSE)
write.table(res2,paste0(platform,".powb.interaction.go.txt"),sep = "\t", row.names = FALSE, quote = FALSE)
  
set4Int = set4Compare %>% left_join(annoEPIC,by="cpg") 

write.table(set4Int,paste0(platform,".powb.interaction.topGenes.txt"),sep = "\t", row.names = FALSE, quote = FALSE)
  
################## comparing SNP corrected vs none
platform="set4"

set4 = fread("set4.POBW.geneCon.s2.ewas.txt") %>%
  mutate(cpg = gsub("_.*$","",CpG_snp),
         snp = gsub("^.*_","",CpG_snp)) %>%
  arrange(by=`P-value`) %>%
  mutate(mqtl_fdr = p.adjust(`P-value`, 
                                    method = "fdr", 
                                    n = length(`P-value`))) %>%
  dplyr::rename("effect_mqtl"="Estimate",
            "p_mqtl"="P-value") %>%
  select(cpg,effect_mqtl,p_mqtl,mqtl_fdr,snp)

set4NG = fread("set4.POBW.Nogene.s2.ewas.txt") %>%
  mutate(cpg = gsub("_.*$","",CpG_snp)) %>%
  mutate(nomqtl_fdr = p.adjust(`P-value`, 
                                      method = "fdr", 
                                      n = length(`P-value`))) %>%
  dplyr::rename("effect_nomqtl"="Estimate",
                "p_nomqtl"="P-value")  %>%
  select(cpg,effect_nomqtl,p_nomqtl,nomqtl_fdr)

# select probes that are significant either with or without SNP correction
set4Compare = set4 %>%
  inner_join(set4NG,by="cpg")%>%
  mutate(delta=abs((effect_mqtl-effect_nomqtl)/effect_nomqtl))%>%
  filter(mqtl_fdr < 0.05)%>%
  filter(delta>0.2) %>%
  select(cpg,p_nomqtl,effect_nomqtl,effect_mqtl,p_mqtl) 

###############
max = max(c(set4Compare$effect_nomqtl,set4Compare$effect_mqtl)*1.2)
min = min(c(set4Compare$effect_nomqtl,set4Compare$effect_mqtl)*1.2)

set4Compare %>%
  ggplot(aes(x=effect_nomqtl,y=effect_mqtl))+
  geom_point()+
  geom_abline(aes(intercept=0,slope=1),alpha=0.4,color="red")+
  geom_hline(aes(yintercept=0),color="red")+
  geom_vline(aes(xintercept=0),color="red")+
  xlim(min,max)+ylim(min,max)+
  xlab("beta value, without mQTL") + ylab("beta value, with mQTL")+
  ggtitle(paste0("POBW Effects with or without mQTL, sig results with interaction, platform ",platform))+
  theme_bw()+
  theme(plot.title = element_text(hjust = 0.5))
ggsave(paste0(platform,".POBW.interaction.effectCompSig.png"),width = 8,height = 6)

###############
max = max(c(-log10(set4Compare$p_nomqtl),-log10(set4Compare$p_mqtl)))*1.1
min = min(c(-log10(set4Compare$p_nomqtl),-log10(set4Compare$p_mqtl)))*0.9

set4Compare %>%
  mutate(logP_nomqtl=-log10(p_nomqtl),logP_mqtl=-log10(p_mqtl))%>%
  ggplot(aes(x=logP_nomqtl,y=logP_mqtl))+
  geom_point()+
  geom_abline(aes(intercept=0,slope=1),alpha=0.4,color="red")+
  geom_hline(aes(yintercept=0),color="red")+
  geom_vline(aes(xintercept=0),color="red")+
  xlim(0,max)+ylim(0,max)+
  xlab("log P value, without mQTL") + ylab("log P value, with mQTL")+
  ggtitle(paste0("POBW Significance with or without mQTL, sig results with interaction, platform ",platform))+
  theme_bw()+
  theme(plot.title = element_text(hjust = 0.5))
ggsave(paste0(platform,".POBW.interaction.pCompSig.png"),width = 8,height = 6)


########################################################################
## sensitivity analysis
## without SNP correction

set4NG = fread("set4.bw.Nogene.ewas.txt") %>%
  mutate(cpg = gsub("_.*$","",CpG_snp)) %>%
  mutate(effect_norm_nomqtl = BBmisc::normalize(Estimate,range = c(-1, 1))) %>%
  mutate(nomqtl_bonferroni = p.adjust(`P-value`, 
                                      method = "bonferroni", 
                                      n = length(`P-value`))) %>%
  dplyr::rename("p_nomqtl"="P-value")  %>%
  select(cpg,effect_norm_nomqtl,p_nomqtl,nomqtl_bonferroni)

set4NG_s = fread("set4.POBW.Nogene.s2.ewas.txt") %>%
  mutate(cpg = gsub("_.*$","",CpG_snp)) %>%
  mutate(effect_norm_nomqtl_s = BBmisc::normalize(Estimate,range = c(-1, 1))) %>%
  mutate(nomqtl_bonferroni_s = p.adjust(`P-value`, 
                                      method = "bonferroni", 
                                      n = length(`P-value`))) %>%
  dplyr::rename("p_nomqtl_s"="P-value")  %>%
  select(cpg,effect_norm_nomqtl_s,p_nomqtl_s,nomqtl_bonferroni_s)

set4Compare = set4NG %>%
  inner_join(set4NG_s,by="cpg")%>%
  filter(nomqtl_bonferroni < 0.05 | nomqtl_bonferroni_s<0.05)
 
max = max(c(set4Compare$effect_norm_nomqtl,set4Compare$effect_norm_nomqtl_s)*1.2)
min = min(c(set4Compare$effect_norm_nomqtl,set4Compare$effect_norm_nomqtl_s)*1.2)

set4Compare %>%
  ggplot(aes(x=effect_norm_nomqtl,y=effect_norm_nomqtl_s))+
  geom_point()+
  geom_abline(aes(intercept=0,slope=1),alpha=0.4,color="red")+
  geom_hline(aes(yintercept=0),color="red")+
  geom_vline(aes(xintercept=0),color="red")+
  xlim(min,max)+
  ylim(min,max)+
  xlab("normalized beta value") + ylab("normalized beta value, sensitivity analysis")+
  ggtitle("comparison of beta coefficients in 2nd sensitivity analysis")+
  theme_bw()+
  theme(plot.title = element_text(hjust = 0.5))
ggsave(paste0(platform,"s2.sensitivity.beta.png"), width = 8,height = 6)

###############
max = max(c(-log10(set4Compare$p_nomqtl),-log10(set4Compare$p_nomqtl_s)))*1.1
min = min(c(-log10(set4Compare$p_nomqtl),-log10(set4Compare$p_nomqtl_s)))*0.9

set4Compare %>%
  mutate(logP_nomqtl=-log10(p_nomqtl),logP_nomqtl_s=-log10(p_nomqtl_s))%>%
  ggplot(aes(x=logP_nomqtl,y=logP_nomqtl_s))+
  geom_point()+
  geom_abline(aes(intercept=0,slope=1),alpha=0.4,color="red")+
  geom_hline(aes(yintercept=0),color="red")+
  geom_vline(aes(xintercept=0),color="red")+
  xlim(0,max)+ylim(0,max)+
  xlab("log P value") + ylab("log P value, sensitivity analysis")+
  ggtitle("comparison of P values in 2nd sensitivity analysis")+
  theme_bw()+
  theme(plot.title = element_text(hjust = 0.5))
ggsave(paste0(platform,".s2.sensitivity.logP.png"), width = 8,height = 6)


########################
## with SNP correction
set4 = fread("set4.bw.geneCon.ewas.txt") %>%
  mutate(cpg = gsub("_.*$","",CpG_snp)) %>%
  mutate(effect_norm_mqtl = BBmisc::normalize(Estimate,range = c(-1, 1))) %>%
  mutate(mqtl_bonferroni = p.adjust(`P-value`, 
                                    method = "bonferroni", 
                                    n = length(`P-value`))) %>%
  dplyr::rename("p_mqtl"="P-value")  %>%
  select(cpg,effect_norm_mqtl,p_mqtl,mqtl_bonferroni)

set4_s = fread("set4.POBW.geneCon.s2.ewas.txt") %>%
  mutate(cpg = gsub("_.*$","",CpG_snp)) %>%
  mutate(effect_norm_mqtl_s = BBmisc::normalize(Estimate,range = c(-1, 1))) %>%
  mutate(mqtl_bonferroni_s = p.adjust(`P-value`, 
                                      method = "bonferroni", 
                                      n = length(`P-value`))) %>%
  dplyr::rename("p_mqtl_s"="P-value")  %>%
  select(cpg,effect_norm_mqtl_s,p_mqtl_s,mqtl_bonferroni_s)

set4Compare = set4 %>%
  inner_join(set4_s,by="cpg")%>%
  filter(mqtl_bonferroni < 0.05 | mqtl_bonferroni_s<0.05)
 
max = max(c(set4Compare$effect_norm_mqtl,set4Compare$effect_norm_mqtl_s)*1.2)
min = min(c(set4Compare$effect_norm_mqtl,set4Compare$effect_norm_mqtl_s)*1.2)

set4Compare %>%
  ggplot(aes(x=effect_norm_mqtl,y=effect_norm_mqtl_s))+
  geom_point()+
  geom_abline(aes(intercept=0,slope=1),alpha=0.4,color="red")+
  geom_hline(aes(yintercept=0),color="red")+
  geom_vline(aes(xintercept=0),color="red")+
  xlim(min,max)+
  ylim(min,max)+
  xlab("normalized beta value") + 
  ylab("normalized beta value, sensitivity analysis, with SNP control")+
  ggtitle("comparison of beta coefficients in 2nd sensitivity analysis,\n with SNP control")+
  theme_bw()+
  theme(plot.title = element_text(hjust = 0.5))
ggsave(paste0(platform,"s2.sensitivity.mqtl.beta.png"), width = 8,height = 6)

###############
max = max(c(-log10(set4Compare$p_mqtl),-log10(set4Compare$p_mqtl_s)))*1.1
min = min(c(-log10(set4Compare$p_mqtl),-log10(set4Compare$p_mqtl_s)))*0.9

set4Compare %>%
  mutate(logP_nomqtl=-log10(p_mqtl),logP_nomqtl_s=-log10(p_mqtl_s))%>%
  ggplot(aes(x=logP_nomqtl,y=logP_nomqtl_s))+
  geom_point()+
  geom_abline(aes(intercept=0,slope=1),alpha=0.4,color="red")+
  geom_hline(aes(yintercept=0),color="red")+
  geom_vline(aes(xintercept=0),color="red")+
  xlim(0,max)+ylim(0,max)+
  xlab("log P value") + ylab("log P value, sensitivity analysis, with SNP control")+
  ggtitle("comparison of P values in 2nd sensitivity analysis,\n with SNP control")+
  theme_bw()+
  theme(plot.title = element_text(hjust = 0.5))
ggsave(paste0(platform,".s2.sensitivity.mqtl.logP.png"), width = 8,height = 6)

```

>> Select probes that are significant after mqtl multiple correct and compare that subgroup when the analysis is done.

```{bash choosing used SNP-cpg pair and their P value}

for set in 1 2 3 4; do
  awk 'NR>1 {print $1,$2,$3}' ccls.set$set.mqtls.hd > set$set.cpgsnp.pairs
done

```

```{r select significant SNP-cpg pairs and check their results}

setwd("~/mQTL/birthweight")
library(tidyverse)
library(data.table)
library(qvalue)

library(IlluminaHumanMethylation450kanno.ilmn12.hg19)
data("IlluminaHumanMethylation450kanno.ilmn12.hg19")
anno       <- IlluminaHumanMethylation450kanno.ilmn12.hg19 %>%
  getAnnotation %>%
  as.data.frame %>%
  rownames_to_column(var="cpg") %>%
  select(cpg,chr,pos,Relation_to_Island,UCSC_RefGene_Name,UCSC_RefGene_Group)

library(IlluminaHumanMethylationEPICanno.ilm10b4.hg19)
data("IlluminaHumanMethylationEPICanno.ilm10b4.hg19")
annoEPIC       <- IlluminaHumanMethylationEPICanno.ilm10b4.hg19 %>%
  getAnnotation %>%
  as.data.frame %>%
  rownames_to_column(var="cpg") %>%
  select(cpg,chr,pos,Relation_to_Island,UCSC_RefGene_Name,UCSC_RefGene_Group)

for(platform in c("450K","EPIC","all")){
  
  if(platform!="450K"){
    pairs = fread("~/mQTL/mqtl.k.450K.sim")
    }else{
    pairs = fread("~/mQTL/mqtl.k.EPIC.sim")}
  
  pairs_conc = pairs %>%
   mutate(p_fdr=p.adjust(`P-value`,method="bonferroni"))%>%
   filter(p_fdr<0.05)
  nrow(pairs_conc)/nrow(pairs)
  sig_CpGs = pairs_conc$CpG
  
  # checking results that are significant with/withou mqtl
  
  geneCon = fread(paste0("cp.meta.",platform,".bw.geneCon.txt")) %>%
    separate(MarkerName,c("cpg","SNP"),sep="_") %>%
    mutate(Estimate.mqtl=Effect, P.fdr.mqtl=p.adjust(`P-value`,method="fdr"))%>%
    select(cpg,SNP,Estimate.mqtl,P.fdr.mqtl) %>%
    filter(SNP!="none")
  
  NogeneCon = fread(paste0("cp.meta.",platform,".bw.Nogene.txt")) %>%
    separate(MarkerName,c("cpg","SNP"),sep="_") %>%
    mutate(Estimate.nomqtl=Effect, P.fdr.nomqtl=p.adjust(`P-value`,method="fdr"))%>%
    select(cpg,Estimate.nomqtl,P.fdr.nomqtl) 
  
  totGeneCon = geneCon %>% inner_join(NogeneCon,by = "cpg") 
  totGeneConSig = totGeneCon %>% 
    filter(cpg %in% sig_CpGs) %>%
    na.omit() %>%
    filter(P.fdr.mqtl<0.05 | P.fdr.nomqtl<0.05)
  

  if(platform=="450K"){
    totGeneConSigs_anno = totGeneConSig  %>%  left_join(anno,by=c("cpg")) %>%
    mutate(tss=0)
    totGeneConSigs_anno$tss[grep("TSS",totGeneConSigs_anno$UCSC_RefGene_Group)]=1
  }else{
    totGeneConSigs_anno = totGeneConSig  %>%  left_join(annoEPIC,by="cpg") %>%
    mutate(tss=0)
    totGeneConSigs_anno$tss[grep("TSS",totGeneConSigs_anno$UCSC_RefGene_Group)]=1
  }

  
  print(paste0("Significant probes for platform ",platform, "among significant mQTLs"))
  print("with SNP adjustment")
  print(table(totGeneConSigs_anno$P.fdr.mqtl<0.05))
  print("without SNP adjustment")
  print(table(totGeneConSigs_anno$P.fdr.nomqtl<0.05))
  
  print(paste0("distributiono of TSS ",platform))
  print("with SNP adjustment")
  print(table(totGeneConSigs_anno %>% filter(P.fdr.mqtl<0.05) %>% select(tss)))
  print("without SNP adjustment")
  print(table(totGeneConSigs_anno %>% filter(P.fdr.nomqtl<0.05) %>% select(tss)))
  
  write.csv(totGeneConSigs_anno,paste0("cp.bw.mqtl.",platform,".meta.sigmqtls.csv"))

   totGeneConSig %>%
    na.omit() %>%
    ggplot(aes(x=Estimate.nomqtl,y=Estimate.mqtl))+
    geom_point()+
    geom_abline(aes(intercept=0,slope=1),alpha=0.4,color="red")+
    ggtitle(paste0("Beta value comparison in ",platform, "\nSignificant mqtls"))+
    theme_bw()+
    theme(plot.title = element_text(hjust = 0.5))
  ggsave(paste0("cp.",platform,".sig.bw.beta.comp.png"),width = 8,height = 6)

  totGeneConSig %>%
    mutate(logP.fdr.NoMqtl = -log10(P.fdr.nomqtl),
           logP.fdr.mqtl = -log10(P.fdr.mqtl))%>%
    na.omit() %>%
    ggplot(aes(x=logP.fdr.NoMqtl,y=logP.fdr.mqtl))+
    geom_point()+
    geom_abline(aes(intercept=0,slope=1),alpha=0.4,color="red")+
    ggtitle(paste0("P value comparison in ",platform, "\nSignificant mqtls"))+
    theme_bw()+
    theme(plot.title = element_text(hjust = 0.5))
  ggsave(paste0("cp.",platform,".sig.bw.pv.comp.sigmqtls.png"),width = 8,height = 6)
  
#################################################################################################### Pathway analysis of probes whose beta changes (gene*birthweight interaction)
  
  library(methylGSA)
  
  # select probes that are significant either with or without SNP correction
  metaCompare = totGeneConSig %>%
    mutate(delta=abs((Estimate.mqtl-Estimate.nomqtl)/Estimate.nomqtl))%>%
    distinct(cpg,.keep_all = TRUE) %>%
    filter(delta>0.2) %>%
    select(cpg,P.fdr.mqtl) 

  cpg.pval = c(metaCompare$P.fdr.mqtl)
  names(cpg.pval) = metaCompare$cpg
  
  res1 = methylglm(cpg.pval = cpg.pval, minsize = 2, GS.type = "KEGG")
  res2 = methylglm(cpg.pval = cpg.pval, minsize = 10, GS.type = "GO")
  
  write.table(res1,paste0(platform,".interaction.kegg.sigmqtls.txt"),sep = "\t", row.names = FALSE, quote = FALSE)
  write.table(res2,paste0(platform,".interaction.go.sigmqtls.txt"),sep = "\t", row.names = FALSE, quote = FALSE)
  
  if(platform=="450K"){
    metaInt = metaCompare %>% left_join(anno,by="cpg") 
  }else{
    metaInt = metaCompare %>% left_join(annoEPIC,by="cpg") 
  }
  
  write.table(metaInt,paste0("cp.",platform,".interaction.topGenes.sigmqtls.txt"),sep = "\t", row.names = FALSE, quote = FALSE)
  
}


# [1] "Significant probes for platform 450Kamong significant mQTLs"
# [1] "with SNP adjustment"
# 
# FALSE  TRUE 
#    46   308 
# [1] "without SNP adjustment"
# 
# FALSE  TRUE 
#   146   208 
# [1] "distributiono of TSS 450K"
# [1] "with SNP adjustment"
# 
#   0   1 
# 250  58 
# [1] "without SNP adjustment"
# 
#   0   1 
# 168  40 
# 
# [1] "Significant probes for platform EPICamong significant mQTLs"
# [1] "with SNP adjustment"
# 
# FALSE  TRUE 
#  1125  1228 
# [1] "without SNP adjustment"
# 
# FALSE  TRUE 
#   511  1842 
# [1] "distributiono of TSS EPIC"
# [1] "with SNP adjustment"
# 
#   0   1 
# 982 246 
# [1] "without SNP adjustment"
# 
#    0    1 
# 1422  420 
# 
# [1] "Significant probes for platform allamong significant mQTLs"
# [1] "with SNP adjustment"
# 
# FALSE  TRUE 
#  3295  1626 
# [1] "without SNP adjustment"
# 
# FALSE  TRUE 
#   484  4437 
# [1] "distributiono of TSS all"
# [1] "with SNP adjustment"
# 
#    0    1 
# 1304  322 
# [1] "without SNP adjustment"
# 
#    0    1 
# 3506  931 
# 

```

>> rerun birth weight in different ancestries

```{r adjusting  for scanned mqtls with QTLtools}


setwd("~/mQTL/birthweight")
library(tidyverse)
library(data.table)

for (set in c(1:4)){
  
 if(set==1){var = "birthWt"}
 if(set==2){var = "birthWt"}
 if(set==3){var = "Y_birthweight"}
 if(set==4){var = "ch_birthwt_bc_full"}
  
  subs.anc = fread(paste0("~/sets/set",set,"/clinical_variables.csv")) %>%
    mutate(id=paste0("0_",subjectId))%>%
    filter(race%in%c(1,3))%>%
    select(id,race)
  
  methyl = fread(paste0("cp.methSet",set,".txt")) %>%
  column_to_rownames(var="probeId")%>%
  t()%>%
  as.data.frame()

  covs = fread(paste0("cp.covsSet",set,".txt"))%>%
  column_to_rownames(var="id")%>%
  t()%>%
  as.data.frame()%>%
  mutate(plate = as.factor(plate))

  mtqls = fread(paste0("cp.ccls.set",set,".mqtls.hd")) %>%as.data.frame()
  
### select European subjects  
  
  subs.anc.eur = subs.anc %>% filter(race==1)
  eur.ind = which(rownames(methyl)%in%subs.anc.eur$id)
  methyl.eur = methyl[eur.ind,]
  covs.eur = covs[eur.ind,]
  mtqls.eur <- mtqls[,c("SNP","CpG","P-value",rownames(methyl.eur))]
  
  race="eur"
  
  probe_gN = colnames(methyl.eur)[which(!colnames(methyl.eur)%in%mtqls.eur$CpG)] 
  probe_gY = colnames(methyl.eur)[which(colnames(methyl.eur)%in%mtqls.eur$CpG)] 

  # Group 1 of CpGs: those without mqtls

  probes <- probe_gN

  reg <- lapply(probes,function(probe){
    c(data.frame(CpG=probe, snp="none"),summary(
      lm( methyl.eur[,probe] ~  covs.eur[ ,var] + covs.eur$gestage + covs.eur$sex + covs.eur$plate + covs.eur$CD8T + 
            covs.eur$CD4T + covs.eur$Mono + covs.eur$Bcell + covs.eur$nRBC + covs.eur$PC1 + covs.eur$PC2 + covs.eur$PC3 + covs.eur$PC4 + covs.eur$PC5 +
            covs.eur$PC6 + covs.eur$PC7 + covs.eur$PC8 + covs.eur$PC8 + covs.eur$PC9 + covs.eur$PC10 ))$coefficients[2,c(1,2,4)])})
  reg_1 <- bind_rows(reg) %>% as.data.frame() 

  
# Group 2 of CpGs: those with mqtls
  
    probes <- probe_gY
    reg <- lapply(probes,function(probe){
    
    SNPname =  mtqls.eur[which(mtqls.eur$CpG == probe), 1]  %>% as.character()  
    SNP =  mtqls.eur[which(mtqls.eur$CpG == probe), -c(1:3)] %>% t() %>% as.data.frame()
      
    c(data.frame(CpG=probe, snp=SNPname),summary(
      lm( methyl.eur[,probe] ~  covs.eur[ ,var] + SNP[,1] + covs.eur$gestage + covs.eur$sex + covs.eur$plate + covs.eur$CD8T + 
            covs.eur$CD4T + covs.eur$Mono + covs.eur$Bcell + covs.eur$nRBC + covs.eur$PC1 + covs.eur$PC2 + covs.eur$PC3 + covs.eur$PC4 + covs.eur$PC5 +
            covs.eur$PC6 + covs.eur$PC7 + covs.eur$PC8 + covs.eur$PC8 + covs.eur$PC9 + covs.eur$PC10 ))$coefficients[2,c(1,2,4)]) })
    reg_2 <- bind_rows(reg) %>% as.data.frame() 
    
    resStore = rbind(reg_1,reg_2) %>% as.data.frame() %>% arrange(by=`Pr(>|t|)`)

    write.table(resStore,paste0("cp.ccls.set",set,".",var,".",race,".geneCon.ewas.txt"),
                quote=FALSE,
                sep='\t',
                row.names = FALSE, 
                col.names = TRUE)  
    
    
### select Latino subjects  
  
  subs.anc.lat = subs.anc %>% filter(race==3)
  lat.ind = which(rownames(methyl)%in%subs.anc.lat$id)
  methyl.lat = methyl[lat.ind,]
  covs.lat = covs[lat.ind,]
  mtqls.lat <- mtqls[,c("SNP","CpG","P-value",rownames(methyl.lat))]
  
  race="lat"
  
  probe_gN = colnames(methyl.lat)[which(!colnames(methyl.lat)%in%mtqls.lat$CpG)] 
  probe_gY = colnames(methyl.lat)[which(colnames(methyl.lat)%in%mtqls.lat$CpG)] 

  # Group 1 of CpGs: those without mqtls

  probes <- probe_gN

  reg <- lapply(probes,function(probe){
    c(data.frame(CpG=probe, snp="none"),summary(
      lm( methyl.lat[,probe] ~  covs.lat[ ,var] + covs.lat$gestage + covs.lat$sex + covs.lat$plate + covs.lat$CD8T + 
            covs.lat$CD4T + covs.lat$Mono + covs.lat$Bcell + covs.lat$nRBC + covs.lat$PC1 + covs.lat$PC2 + covs.lat$PC3 + covs.lat$PC4 + covs.lat$PC5 +
            covs.lat$PC6 + covs.lat$PC7 + covs.lat$PC8 + covs.lat$PC8 + covs.lat$PC9 + covs.lat$PC10 ))$coefficients[2,c(1,2,4)])})
  reg_1 <- bind_rows(reg) %>% as.data.frame() 

  
# Group 2 of CpGs: those with mqtls
  
    probes <- probe_gY
    reg <- lapply(probes,function(probe){
    
    SNPname =  mtqls.lat[which(mtqls.lat$CpG == probe), 1]  %>% as.character()  
    SNP =  mtqls.lat[which(mtqls.lat$CpG == probe), -c(1:3)] %>% t() %>% as.data.frame()
      
    c(data.frame(CpG=probe, snp=SNPname),summary(
      lm( methyl.lat[,probe] ~  covs.lat[ ,var] + SNP[,1] + covs.lat$gestage + covs.lat$sex + covs.lat$plate + covs.lat$CD8T + 
            covs.lat$CD4T + covs.lat$Mono + covs.lat$Bcell + covs.lat$nRBC + covs.lat$PC1 + covs.lat$PC2 + covs.lat$PC3 + covs.lat$PC4 + covs.lat$PC5 +
            covs.lat$PC6 + covs.lat$PC7 + covs.lat$PC8 + covs.lat$PC8 + covs.lat$PC9 + covs.lat$PC10 ))$coefficients[2,c(1,2,4)]) })
    reg_2 <- bind_rows(reg) %>% as.data.frame() 
    
    resStore = rbind(reg_1,reg_2) %>% as.data.frame() %>% arrange(by=`Pr(>|t|)`)

    write.table(resStore,paste0("cp.ccls.set",set,".",var,".",race,".geneCon.ewas.txt"),
                quote=FALSE,
                sep='\t',
                row.names = FALSE, 
                col.names = TRUE)      
    
    }


```

```{r Method 2: comparison, no SNP at all}


setwd("~/mQTL/birthweight")
library(tidyverse)
library(data.table)

for (set in c(1:4)){
  
 if(set==1){var = "birthWt"}
 if(set==2){var = "birthWt"}
 if(set==3){var = "Y_birthweight"}
 if(set==4){var = "ch_birthwt_bc_full"}

  subs.anc = fread(paste0("~/sets/set",set,"/clinical_variables.csv")) %>%
    mutate(id=paste0("0_",subjectId))%>%
    filter(race%in%c(1,3))%>%
    select(id,race)
  
  methyl = fread(paste0("cp.methSet",set,".txt")) %>%
  column_to_rownames(var="probeId")%>%
  t()%>%
  as.data.frame()

  covs = fread(paste0("cp.covsSet",set,".txt"))%>%
  column_to_rownames(var="id")%>%
  t()%>%
  as.data.frame()%>%
  mutate(plate = as.factor(plate))


### select European subjects  

  subs.anc.eur = subs.anc %>% filter(race==1)
  eur.ind = which(rownames(methyl)%in%subs.anc.eur$id)
  methyl.eur = methyl[eur.ind,]
  covs.eur = covs[eur.ind,]

  race="eur"

  probes = colnames(methyl.eur)

  reg <- lapply(probes,function(probe){
    c(data.frame(CpG=probe, snp="none"),summary(
      lm( methyl.eur[,probe] ~  covs.eur[ ,var] + covs.eur$sex + covs.eur$gestage + covs.eur$plate + covs.eur$CD8T + 
            covs.eur$CD4T + covs.eur$Mono + covs.eur$Bcell + covs.eur$nRBC + covs.eur$PC1 + covs.eur$PC2 + covs.eur$PC3 + covs.eur$PC4 + covs.eur$PC5 +
            covs.eur$PC6 + covs.eur$PC7 + covs.eur$PC8 + covs.eur$PC8 + covs.eur$PC9 + covs.eur$PC10 ))$coefficients[2,c(1,2,4)])})

  reg_1 <- bind_rows(reg) %>% as.data.frame() %>% arrange(by=`Pr(>|t|)`)
  write.table(reg_1,paste0("cp.ccls.set",set,".",var,".",race,".Nogene.ewas.txt"),
            quote=FALSE,
            sep='\t',
            row.names = FALSE, 
            col.names = TRUE)  


### select Latino subjects  
  
  subs.anc.lat = subs.anc %>% filter(race==3)
  lat.ind = which(rownames(methyl)%in%subs.anc.lat$id)
  methyl.lat = methyl[lat.ind,]
  covs.lat = covs[lat.ind,]

  race="lat"

  probes = colnames(methyl.lat)

  reg <- lapply(probes,function(probe){
    c(data.frame(CpG=probe, snp="none"),summary(
      lm( methyl.lat[,probe] ~  covs.lat[ ,var] + covs.lat$sex + covs.lat$gestage + covs.lat$plate + covs.lat$CD8T + 
            covs.lat$CD4T + covs.lat$Mono + covs.lat$Bcell + covs.lat$nRBC + covs.lat$PC1 + covs.lat$PC2 + covs.lat$PC3 + covs.lat$PC4 + covs.lat$PC5 +
            covs.lat$PC6 + covs.lat$PC7 + covs.lat$PC8 + covs.lat$PC8 + covs.lat$PC9 + covs.lat$PC10 ))$coefficients[2,c(1,2,4)])})

  reg_1 <- bind_rows(reg) %>% as.data.frame() %>% arrange(by=`Pr(>|t|)`)
  write.table(reg_1,paste0("cp.ccls.set",set,".",var,".",race,".Nogene.ewas.txt"),
            quote=FALSE,
            sep='\t',
            row.names = FALSE, 
            col.names = TRUE)  


      }

```

```{bash  grab data back from usc}

lcd ~/mQTL/birthweight
cd /scratch/lishaobo/temps/berkeley/bw

get *eur*
get *lat*

```

```{bash meta-analysis}

cd ~/mQTL/birthweight

## Meta-analysis by platform
awk '{print $1"_"$2,$3,$4,$5}' <(tail -n +2 cp.ccls.set1.birthWt.eur.geneCon.ewas.txt) |\
sed '1i CpG_snp Estimate  SE P-value' > set1.bw.eur.geneCon.ewas.txt
awk '{print $1"_"$2,$3,$4,$5}' <(tail -n +2 cp.ccls.set2.birthWt.eur.geneCon.ewas.txt) |\
sed '1i CpG_snp Estimate  SE P-value' > set2.bw.eur.geneCon.ewas.txt
awk '{print $1"_"$2,$3,$4,$5}' <(tail -n +2 cp.ccls.set3.Y_birthweight.eur.geneCon.ewas.txt) |\
sed '1i CpG_snp Estimate  SE P-value' > set3.bw.eur.geneCon.ewas.txt
awk '{print $1"_"$2,$3,$4,$5}' <(tail -n +2 cp.ccls.set4.ch_birthwt_bc_full.eur.geneCon.ewas.txt) |\
sed '1i CpG_snp Estimate  SE P-value' > set4.bw.eur.geneCon.ewas.txt
awk '{print $1"_"$2,$3,$4,$5}' <(tail -n +2 cp.ccls.set1.birthWt.lat.geneCon.ewas.txt) |\
sed '1i CpG_snp Estimate  SE P-value' > set1.bw.lat.geneCon.ewas.txt
awk '{print $1"_"$2,$3,$4,$5}' <(tail -n +2 cp.ccls.set2.birthWt.lat.geneCon.ewas.txt) |\
sed '1i CpG_snp Estimate  SE P-value' > set2.bw.lat.geneCon.ewas.txt
awk '{print $1"_"$2,$3,$4,$5}' <(tail -n +2 cp.ccls.set3.Y_birthweight.lat.geneCon.ewas.txt) |\
sed '1i CpG_snp Estimate  SE P-value' > set3.bw.lat.geneCon.ewas.txt
awk '{print $1"_"$2,$3,$4,$5}' <(tail -n +2 cp.ccls.set4.ch_birthwt_bc_full.lat.geneCon.ewas.txt) |\
sed '1i CpG_snp Estimate  SE P-value' > set4.bw.lat.geneCon.ewas.txt

## No Gnene
awk '{print $1"_"$2,$3,$4,$5}' <(tail -n +2 cp.ccls.set1.birthWt.eur.Nogene.ewas.txt) |\
sed '1i CpG_snp Estimate  SE P-value' > set1.bw.eur.Nogene.ewas.txt
awk '{print $1"_"$2,$3,$4,$5}' <(tail -n +2 cp.ccls.set2.birthWt.eur.Nogene.ewas.txt) |\
sed '1i CpG_snp Estimate  SE P-value' > set2.bw.eur.Nogene.ewas.txt
awk '{print $1"_"$2,$3,$4,$5}' <(tail -n +2 cp.ccls.set3.Y_birthweight.eur.Nogene.ewas.txt) |\
sed '1i CpG_snp Estimate  SE P-value' > set3.bw.eur.Nogene.ewas.txt
awk '{print $1"_"$2,$3,$4,$5}' <(tail -n +2 cp.ccls.set4.ch_birthwt_bc_full.eur.Nogene.ewas.txt) |\
sed '1i CpG_snp Estimate  SE P-value' > set4.bw.eur.Nogene.ewas.txt
awk '{print $1"_"$2,$3,$4,$5}' <(tail -n +2 cp.ccls.set1.birthWt.lat.Nogene.ewas.txt) |\
sed '1i CpG_snp Estimate  SE P-value' > set1.bw.lat.Nogene.ewas.txt
awk '{print $1"_"$2,$3,$4,$5}' <(tail -n +2 cp.ccls.set2.birthWt.lat.Nogene.ewas.txt) |\
sed '1i CpG_snp Estimate  SE P-value' > set2.bw.lat.Nogene.ewas.txt
awk '{print $1"_"$2,$3,$4,$5}' <(tail -n +2 cp.ccls.set3.Y_birthweight.lat.Nogene.ewas.txt) |\
sed '1i CpG_snp Estimate  SE P-value' > set3.bw.lat.Nogene.ewas.txt
awk '{print $1"_"$2,$3,$4,$5}' <(tail -n +2 cp.ccls.set4.ch_birthwt_bc_full.lat.Nogene.ewas.txt) |\
sed '1i CpG_snp Estimate  SE P-value' > set4.bw.lat.Nogene.ewas.txt

### 450K, Gene, eur
metal
MARKER CpG_snp
EFFECT Estimate
PVALUE `P-value`
SCHEME STDERR
STDERR SE
VERBOSE OFF
GENOMICCONTROL OFF
STDERR_PRINT_PRECISION 20
EFFECT_PRINT_PRECISION 20
PROCESS set1.bw.eur.geneCon.ewas.txt
PROCESS set2.bw.eur.geneCon.ewas.txt
ANALYZE
QUIT
mv METAANALYSIS1.TBL meta.450K.bw.eur.geneCon.ewas.TBL

### 450K, Gene, lat
metal
MARKER CpG_snp
EFFECT Estimate
PVALUE `P-value`
SCHEME STDERR
STDERR SE
VERBOSE OFF
GENOMICCONTROL OFF
STDERR_PRINT_PRECISION 20
EFFECT_PRINT_PRECISION 20
PROCESS set1.bw.lat.geneCon.ewas.txt
PROCESS set2.bw.lat.geneCon.ewas.txt
ANALYZE
QUIT
mv METAANALYSIS1.TBL meta.450K.bw.lat.geneCon.ewas.TBL

### 450K, noGene, eur
metal
MARKER CpG_snp
EFFECT Estimate
PVALUE `P-value`
SCHEME STDERR
STDERR SE
VERBOSE OFF
GENOMICCONTROL OFF
STDERR_PRINT_PRECISION 20
EFFECT_PRINT_PRECISION 20
PROCESS set1.bw.eur.Nogene.ewas.txt
PROCESS set2.bw.eur.Nogene.ewas.txt
ANALYZE
QUIT
mv METAANALYSIS1.TBL meta.450K.bw.eur.Nogene.ewas.TBL

### 450K, noGene, lat
metal
MARKER CpG_snp
EFFECT Estimate
PVALUE `P-value`
SCHEME STDERR
STDERR SE
VERBOSE OFF
GENOMICCONTROL OFF
STDERR_PRINT_PRECISION 20
EFFECT_PRINT_PRECISION 20
PROCESS set1.bw.lat.Nogene.ewas.txt
PROCESS set2.bw.lat.Nogene.ewas.txt
ANALYZE
QUIT
mv METAANALYSIS1.TBL meta.450K.bw.lat.Nogene.ewas.TBL

### EPIC, Gene, eur
metal
MARKER CpG_snp
EFFECT Estimate
PVALUE `P-value`
SCHEME STDERR
STDERR SE
VERBOSE OFF
GENOMICCONTROL OFF
STDERR_PRINT_PRECISION 20
EFFECT_PRINT_PRECISION 20
PROCESS set3.bw.eur.geneCon.ewas.txt
PROCESS set4.bw.eur.geneCon.ewas.txt
ANALYZE
QUIT
mv METAANALYSIS1.TBL meta.EPIC.bw.eur.geneCon.ewas.TBL

### EPIC, Gene, lat
metal
MARKER CpG_snp
EFFECT Estimate
PVALUE `P-value`
SCHEME STDERR
STDERR SE
VERBOSE OFF
GENOMICCONTROL OFF
STDERR_PRINT_PRECISION 20
EFFECT_PRINT_PRECISION 20
PROCESS set3.bw.lat.geneCon.ewas.txt
PROCESS set4.bw.lat.geneCon.ewas.txt
ANALYZE
QUIT
mv METAANALYSIS1.TBL meta.EPIC.bw.lat.geneCon.ewas.TBL

### EPIC, noGene, eur
metal
MARKER CpG_snp
EFFECT Estimate
PVALUE `P-value`
SCHEME STDERR
STDERR SE
VERBOSE OFF
GENOMICCONTROL OFF
STDERR_PRINT_PRECISION 20
EFFECT_PRINT_PRECISION 20
PROCESS set3.bw.eur.Nogene.ewas.txt
PROCESS set4.bw.eur.Nogene.ewas.txt
ANALYZE
QUIT
mv METAANALYSIS1.TBL meta.EPIC.bw.eur.Nogene.ewas.TBL

### EPIC, noGene, lat
metal
MARKER CpG_snp
EFFECT Estimate
PVALUE `P-value`
SCHEME STDERR
STDERR SE
VERBOSE OFF
GENOMICCONTROL OFF
STDERR_PRINT_PRECISION 20
EFFECT_PRINT_PRECISION 20
PROCESS set3.bw.lat.Nogene.ewas.txt
PROCESS set4.bw.lat.Nogene.ewas.txt
ANALYZE
QUIT
mv METAANALYSIS1.TBL meta.EPIC.bw.lat.Nogene.ewas.TBL

### all, Gene, eur
metal
MARKER CpG_snp
EFFECT Estimate
PVALUE `P-value`
SCHEME STDERR
STDERR SE
VERBOSE OFF
GENOMICCONTROL OFF
STDERR_PRINT_PRECISION 20
EFFECT_PRINT_PRECISION 20
PROCESS set1.bw.eur.geneCon.ewas.txt
PROCESS set2.bw.eur.geneCon.ewas.txt
PROCESS set3.bw.eur.geneCon.ewas.txt
PROCESS set4.bw.eur.geneCon.ewas.txt
ANALYZE
QUIT
mv METAANALYSIS1.TBL meta.all.bw.eur.geneCon.ewas.TBL

### all, Gene, lat
metal
MARKER CpG_snp
EFFECT Estimate
PVALUE `P-value`
SCHEME STDERR
STDERR SE
VERBOSE OFF
GENOMICCONTROL OFF
STDERR_PRINT_PRECISION 20
EFFECT_PRINT_PRECISION 20
PROCESS set1.bw.lat.geneCon.ewas.txt
PROCESS set2.bw.lat.geneCon.ewas.txt
PROCESS set3.bw.lat.geneCon.ewas.txt
PROCESS set4.bw.lat.geneCon.ewas.txt
ANALYZE
QUIT
mv METAANALYSIS1.TBL meta.all.bw.lat.geneCon.ewas.TBL

### all, noGene, eur
metal
MARKER CpG_snp
EFFECT Estimate
PVALUE `P-value`
SCHEME STDERR
STDERR SE
VERBOSE OFF
GENOMICCONTROL OFF
STDERR_PRINT_PRECISION 20
EFFECT_PRINT_PRECISION 20
PROCESS set1.bw.eur.Nogene.ewas.txt
PROCESS set2.bw.eur.Nogene.ewas.txt
PROCESS set3.bw.eur.Nogene.ewas.txt
PROCESS set4.bw.eur.Nogene.ewas.txt
ANALYZE
QUIT
mv METAANALYSIS1.TBL meta.all.bw.eur.Nogene.ewas.TBL

### all, noGene, lat
metal
MARKER CpG_snp
EFFECT Estimate
PVALUE `P-value`
SCHEME STDERR
STDERR SE
VERBOSE OFF
GENOMICCONTROL OFF
STDERR_PRINT_PRECISION 20
EFFECT_PRINT_PRECISION 20
PROCESS set1.bw.lat.Nogene.ewas.txt
PROCESS set2.bw.lat.Nogene.ewas.txt
PROCESS set3.bw.lat.Nogene.ewas.txt
PROCESS set4.bw.lat.Nogene.ewas.txt
ANALYZE
QUIT
mv METAANALYSIS1.TBL meta.all.bw.lat.Nogene.ewas.TBL

```

```{r Method 2: looking at results after meta}

setwd("~/mQTL/birthweight")
library(tidyverse)
library(data.table)
library(qvalue)

library(IlluminaHumanMethylation450kanno.ilmn12.hg19)
data("IlluminaHumanMethylation450kanno.ilmn12.hg19")
anno       <- IlluminaHumanMethylation450kanno.ilmn12.hg19 %>%
  getAnnotation %>%
  as.data.frame %>%
  rownames_to_column(var="cpg") %>%
  select(cpg,chr,pos,Relation_to_Island,UCSC_RefGene_Name,UCSC_RefGene_Group)

library(IlluminaHumanMethylationEPICanno.ilm10b4.hg19)
data("IlluminaHumanMethylationEPICanno.ilm10b4.hg19")
annoEPIC       <- IlluminaHumanMethylationEPICanno.ilm10b4.hg19 %>%
  getAnnotation %>%
  as.data.frame %>%
  rownames_to_column(var="cpg") %>%
  select(cpg,chr,pos,Relation_to_Island,UCSC_RefGene_Name,UCSC_RefGene_Group)

for(race in c("eur","lat")){
  for(platform in c("450K","EPIC","all")){

  ## For results with mqtl correction
  meta = fread(paste0("meta.",platform,".bw.",race,".geneCon.ewas.TBL")) %>%
    separate(MarkerName,c("cpg","SNP"),sep="_") %>%
    mutate(Estimate.mqtl=Effect, P.fdr.mqtl=p.adjust(`P-value`,method="fdr"))%>%
    select(cpg,SNP,Estimate.mqtl,P.fdr.mqtl) %>%
    filter(SNP!="none")

  metaNG = fread(paste0("meta.",platform,".bw.",race,".Nogene.ewas.TBL")) %>%
    separate(MarkerName,c("cpg","SNP"),sep="_") %>%
    mutate(Estimate.nomqtl=Effect, P.fdr.nomqtl=p.adjust(`P-value`,method="fdr"))%>%
    select(cpg,Estimate.nomqtl,P.fdr.nomqtl) 

  totGeneCon = meta %>% inner_join(metaNG,by = "cpg") %>% 
    filter(P.fdr.mqtl<0.05 | P.fdr.nomqtl<0.05) 


  if(platform=="450K"){
    totGeneConSigs_anno = totGeneCon  %>%  left_join(anno,by=c("cpg")) %>%
    mutate(tss=0)
    totGeneConSigs_anno$tss[grep("TSS",totGeneConSigs_anno$UCSC_RefGene_Group)]=1
  }else{
    totGeneConSigs_anno = totGeneCon  %>%  left_join(annoEPIC,by="cpg") %>%
    mutate(tss=0)
    totGeneConSigs_anno$tss[grep("TSS",totGeneConSigs_anno$UCSC_RefGene_Group)]=1
  }
    
  print(paste0("Significant probes for platform ",platform, "among significant mQTLs in ",race))
  print("with SNP adjustment")
  print(table(totGeneConSigs_anno$P.fdr.mqtl<0.05))
  print("without SNP adjustment")
  print(table(totGeneConSigs_anno$P.fdr.nomqtl<0.05))
  
  print(paste0("distributiono of TSS ",platform, " in ",race))
  print("with SNP adjustment")
  print(table(totGeneConSigs_anno %>% filter(P.fdr.mqtl<0.05) %>% select(tss)))
  print("without SNP adjustment")
  print(table(totGeneConSigs_anno %>% filter(P.fdr.nomqtl<0.05) %>% select(tss)))
  
  write.csv(totGeneConSigs_anno,paste0("bw.",platform,".",race,".meta.csv"))
# 
# [1] "Significant probes for platform 450Kamong significant mQTLs in eur"
# [1] "with SNP adjustment"
# 
# TRUE 
#    3 
# [1] "without SNP adjustment"
# 
# FALSE  TRUE 
#     1     2 
# [1] "distributiono of TSS 450K in eur"
# [1] "with SNP adjustment"
# 
# 0 
# 3 
# [1] "without SNP adjustment"
# 
# 0 
# 2 
# [1] "Significant probes for platform EPICamong significant mQTLs in eur"
# [1] "with SNP adjustment"
# 
# FALSE  TRUE 
# 24112   864 
# [1] "without SNP adjustment"
# 
# FALSE  TRUE 
#   285 24691 
# [1] "distributiono of TSS EPIC in eur"
# [1] "with SNP adjustment"
# 
#   0   1 
# 709 155 
# [1] "without SNP adjustment"
# 
#     0     1 
# 21270  3421 
# [1] "Significant probes for platform allamong significant mQTLs in eur"
# [1] "with SNP adjustment"
# 
# FALSE  TRUE 
# 27731   538 
# [1] "without SNP adjustment"
# 
# FALSE  TRUE 
#   221 28048 
# [1] "distributiono of TSS all in eur"
# [1] "with SNP adjustment"
# 
#   0   1 
# 448  90 
# [1] "without SNP adjustment"
# 
#     0     1 
# 24126  3922 
# [1] "Significant probes for platform 450Kamong significant mQTLs in lat"
# [1] "with SNP adjustment"
# 
# TRUE 
#    7 
# [1] "without SNP adjustment"
# 
# FALSE  TRUE 
#     6     1 
# [1] "distributiono of TSS 450K in lat"
# [1] "with SNP adjustment"
# 
# 0 1 
# 6 1 
# [1] "without SNP adjustment"
# 
# 0 
# 1 
# [1] "Significant probes for platform EPICamong significant mQTLs in lat"
# [1] "with SNP adjustment"
# 
# FALSE  TRUE 
#  5840  2586 
# [1] "without SNP adjustment"
# 
# FALSE  TRUE 
#   830  7596 
# [1] "distributiono of TSS EPIC in lat"
# [1] "with SNP adjustment"
# 
#    0    1 
# 2254  332 
# [1] "without SNP adjustment"
# 
#    0    1 
# 6608  988 
# [1] "Significant probes for platform allamong significant mQTLs in lat"
# [1] "with SNP adjustment"
# 
# FALSE  TRUE 
#  8097  2070 
# [1] "without SNP adjustment"
# 
# FALSE  TRUE 
#   678  9489 
# [1] "distributiono of TSS all in lat"
# [1] "with SNP adjustment"
# 
#    0    1 
# 1808  262 
# [1] "without SNP adjustment"
# 
#    0    1 
# 8170 1319 

###############
  
  max = max(c(totGeneConSigs_anno$Estimate.nomqtl,totGeneConSigs_anno$Estimate.mqtl ))
  min = min(c(totGeneConSigs_anno$Estimate.nomqtl,totGeneConSigs_anno$Estimate.mqtl ))
  
  totGeneConSigs_anno %>%
    ggplot(aes(x=Estimate.nomqtl,y=Estimate.mqtl))+
    geom_point()+
    geom_abline(aes(intercept=0,slope=1),alpha=0.4,color="red")+
    geom_hline(aes(yintercept=0),color="red",alpha=0.4)+geom_vline(aes(xintercept=0),color="red",alpha=0.4)+
    xlab("beta value, without mQTL") + ylab("beta value, with mQTL")+
    ggtitle(paste0("Effects with or without mQTL, sig results, platform ",platform," in ",race))+
    theme_bw()+
    theme(plot.title = element_text(hjust = 0.5))
  ggsave(paste0(platform,".",race,".effectCompSig.png"),width = 8,height = 6)
  
  ###############
  max = max(c(-log10(totGeneConSigs_anno$P.fdr.nomqtl),-log10(totGeneConSigs_anno$P.fdr.mqtl)))*1.1
  min = min(c(-log10(totGeneConSigs_anno$P.fdr.nomqtl),-log10(totGeneConSigs_anno$P.fdr.mqtl)))*0.9
  
  totGeneConSigs_anno %>%
    mutate(logP_noMQTL=-log10(P.fdr.nomqtl),logP_mQTL=-log10(P.fdr.mqtl))%>%
    ggplot(aes(x=logP_noMQTL,y=logP_mQTL))+
    geom_point()+
    geom_abline(aes(intercept=0,slope=1),alpha=0.4,color="red")+
    geom_hline(aes(yintercept=0),color="red",alpha=0.4)+geom_vline(aes(xintercept=0),color="red",alpha=0.4)+
    xlim(0,max)+ylim(0,max)+
    xlab("log P value, without mQTL") + ylab("log P value, with mQTL")+
    ggtitle(paste0("Significance with or without mQTL, sig results, platform ",platform," in ",race))+
    theme_bw()+
    theme(plot.title = element_text(hjust = 0.5))
ggsave(paste0(platform,".",race,".pCompSig.png"),width = 8,height = 6)

}}

#################################################################################################### mqtl significant results after fdr correction, comparing with gold standard
for(race in c("eur","lat")){
  for(platform in c("450K","EPIC","all")){

  meta = fread(paste0("meta.",platform,".bw.",race,".geneCon.ewas.TBL")) %>%
    select(-Allele1, -Allele2) %>%
    mutate(CpG = gsub("_.*$","",MarkerName),
           snp = gsub("^.*_","",MarkerName)) %>%
    select(-MarkerName)
    
  metaNG = fread(paste0("meta.",platform,".bw.",race,".Nogene.ewas.TBL")) %>%
    select(-Allele1, -Allele2) %>%
    mutate(CpG = gsub("_.*$","",MarkerName),
           snp = gsub("^.*_","",MarkerName)) %>%
    select(-MarkerName,-snp,-StdErr) %>% 
    dplyr::rename("effect_noMQTL"="Effect",
                  "P_noMQTL"="P-value",
                  "Direction_noMQTL"="Direction")

  gold = read.csv("bw_meta_paper.csv") %>%
    left_join(meta,by="CpG")%>%
    left_join(metaNG,by="CpG")%>%
    mutate(dis_mqtl = abs(Effect*10^6 -Coef),
           rate_mqtl =`P-value`/`P.value`,
           dis_nomqtl = abs(effect_noMQTL*10^6 -Coef),
           rate_nomqtl =P_noMQTL/`P.value`) 

 write.csv(gold,paste0("1095.gold.",platform,".",race,".meta.csv"))

 gold1 = gold %>%
    mutate(p_gold = -log10(`P.value`),
         p_mqtl = -log10(`P-value`),
         p_nomqtl = -log10(P_noMQTL))%>%
    select(CpG,Coef,Effect,effect_noMQTL,p_gold,p_mqtl,p_nomqtl)

  max = max(c(gold$Effect,gold$effect_noMQTL))
  min = min(c(gold$Effect,gold$effect_noMQTL))

  gold %>%
    na.omit() %>%
    ggplot(aes(x=effect_noMQTL, y=Effect))+
    geom_point()+
    geom_abline(aes(intercept=0,slope=1),alpha=0.4,color="red")+
    geom_hline(aes(yintercept=0),alpha=0.4,color="red")+
    geom_vline(aes(xintercept=0),alpha=0.4,color="red")+
    xlim(min,max)+ylim(min,max)+
    xlab("Effect, without mQTL") + ylab("Effect, with mQTL")+
    ggtitle(paste0("beta value comparison of 1095 probes in ",platform," in ",race))+
    theme_bw()+
    theme(plot.title = element_text(hjust = 0.5))
  ggsave(paste0(platform,".",race,".metapaper.beta.pairwise.compare.png"),width = 8,height = 6)

  
  max = max(c(-log(gold$`P-value`),-log(gold$P_noMQTL)))
  min = min(c(-log(gold$`P-value`),-log(gold$P_noMQTL)))

  gold %>%
    na.omit() %>%
    ggplot(aes(x=-log(P_noMQTL), y=-log(`P-value`)))+
    geom_point()+
    geom_abline(aes(intercept=0,slope=1),alpha=0.4,color="red")+
    xlim(min,max)+ylim(min,max)+
    xlab("log P value, without mQTL") + ylab("log P value, with mQTL")+
    ggtitle(paste0("p value comparison of 1095 probes in ",platform," in ",race))+
    theme_bw()+
    theme(plot.title = element_text(hjust = 0.5))
  ggsave(paste0(platform,".",race,".metapaper.p.pairwise.compare.png"),width = 8,height = 6)

  a=cor.test(gold1$Coef,gold1$Effect)
  b=cor.test(gold1$Coef,gold1$effect_noMQTL)
  c=cor.test(gold1$p_gold,gold1$p_mqtl)
  d=cor.test(gold1$p_gold,gold1$p_nomqtl)

  lab.beta = as_labeller(c( "Effect" = paste0("With mQTL correction
                                              \n correlation=",round(a$estimate,5),
                                              "\n p=",round(a$p.value,5)),
                            "effect_noMQTL" = paste0("Without mQTL correction 
                                              \n correlation=",round(b$estimate,5),
                                              "\n p=",round(b$p.value,5))))
  
  gold %>%
    select(CpG,Coef,Effect,effect_noMQTL)%>%
    reshape::melt(id=c("CpG","Coef"))%>%
    na.omit() %>%
    ggplot(aes(x=Coef,y=value))+
    geom_point()+
    geom_smooth(method = "lm",se=TRUE)+
    facet_wrap(~variable, ncol=2, scales="free_y",
               labeller = lab.beta)+
    ggtitle(paste0("Beta value comparison with gold standard in ",platform," in ",race))+
    xlab("Beta Values from gold standard") + ylab("Beta Values from CCLS")+
    theme_bw()+
    theme(plot.title = element_text(hjust = 0.5))
  ggsave(paste0(platform,".",race,".metapaper.beta.compare.png"),width = 8,height = 6)

  lab.p = as_labeller(c( "p_mqtl" = paste0("With mQTL correction
                                            \n correlation=",round(c$estimate,5),
                                            "\n p=",round(c$p.value,5)),
                          "p_nomqtl" = paste0("Without mQTL correction 
                                            \n correlation=",round(d$estimate,5),
                                            "\n p=",round(d$p.value,5))))
  gold %>%
    mutate(p_gold = -log10(`P.value`),
           p_mqtl = -log10(`P-value`),
           p_nomqtl = -log10(P_noMQTL))%>%
    select(CpG,p_gold,p_mqtl,p_nomqtl)%>%
    reshape::melt(id=c("CpG","p_gold"))%>%
    na.omit() %>%
    ggplot(aes(x=p_gold,y=value))+
    geom_point()+
    geom_smooth(method = "lm",se=TRUE)+
    facet_wrap(~variable, ncol=2, 
               labeller = lab.p)+
    ggtitle(paste0("-log10(P) comparison with gold standard in ",platform," in ",race))+
    xlab("-log10(P) Values from gold standard") + ylab("-log10(P) Values from CCLS")+
    theme_bw()+
    theme(plot.title = element_text(hjust = 0.5))
ggsave(paste0(platform,".",race,".metapaper.p.compare.png"),width = 8,height = 6)

}}


#################################################################################################### Pathway analysis of probes whose beta changes (gene*birthweight interaction)
library(methylGSA)
library(IlluminaHumanMethylation450kanno.ilmn12.hg19)
data("IlluminaHumanMethylation450kanno.ilmn12.hg19")
anno       <- IlluminaHumanMethylation450kanno.ilmn12.hg19 %>%
  getAnnotation %>%
  as.data.frame %>%
  rownames_to_column(var="cpg") %>%
  select(cpg,chr,pos,Relation_to_Island,UCSC_RefGene_Name,UCSC_RefGene_Group)

library(IlluminaHumanMethylationEPICanno.ilm10b4.hg19)
data("IlluminaHumanMethylationEPICanno.ilm10b4.hg19")
annoEPIC       <- IlluminaHumanMethylationEPICanno.ilm10b4.hg19 %>%
  getAnnotation %>%
  as.data.frame %>%
  rownames_to_column(var="cpg") %>%
  select(cpg,chr,pos,Relation_to_Island,UCSC_RefGene_Name,UCSC_RefGene_Group)

for(race in c("eur","lat")){
  for(platform in c("450K","EPIC","all")){
  
  meta = fread(paste0("meta.",platform,".bw.",race,".geneCon.ewas.TBL")) %>%
    separate(MarkerName,c("cpg","SNP"),sep="_") %>%
    mutate(Estimate.mqtl=Effect, P.fdr.mqtl=p.adjust(`P-value`,method="fdr"))%>%
    select(cpg,SNP,Estimate.mqtl,P.fdr.mqtl) %>%
    filter(SNP!="none")
  metaNG = fread(paste0("meta.",platform,".bw.",race,".Nogene.ewas.TBL")) %>%
    separate(MarkerName,c("cpg","SNP"),sep="_") %>%
    mutate(Estimate.nomqtl=Effect, P.fdr.nomqtl=p.adjust(`P-value`,method="fdr"))%>%
    select(cpg,Estimate.nomqtl,P.fdr.nomqtl) 
  totGeneConSig = meta %>% inner_join(metaNG,by = "cpg") %>% 
    filter(P.fdr.mqtl<0.05 | P.fdr.nomqtl<0.05) 
  
  # select probes that are significant either with or without SNP correction
  metaCompare = totGeneConSig %>%
    mutate(delta=abs((Estimate.mqtl-Estimate.nomqtl)/Estimate.nomqtl))%>%
    distinct(cpg,.keep_all = TRUE) %>%
    filter(delta>0.2) %>%
    select(cpg,P.fdr.mqtl) 

  if(nrow(metaCompare)>10){
  cpg.pval = c(metaCompare$P.fdr.mqtl)
  names(cpg.pval) = metaCompare$cpg
  
  res1 = methylglm(cpg.pval = cpg.pval, minsize = 2, GS.type = "KEGG")
  res2 = methylglm(cpg.pval = cpg.pval, minsize = 10, GS.type = "GO")
  
  write.table(res1,paste0(platform,".",race,".interaction.kegg.txt"),sep = "\t", row.names = FALSE, quote = FALSE)
    write.table(res2,paste0(platform,".",race,".interaction.go.txt"),sep = "\t", row.names = FALSE, quote = FALSE)
  
  if(platform=="450K"){
    metaInt = metaCompare %>% left_join(anno,by="cpg") 
  }else{
    metaInt = metaCompare %>% left_join(annoEPIC,by="cpg") 
  }
  
  write.table(metaInt,paste0(platform,".",race,".interaction.topGenes.txt"),sep = "\t", row.names = FALSE, quote = FALSE)
  }
  
  
  }}  

################################################################################
################################################################################
#select significant SNP-cpg pairs and check their results

setwd("~/mQTL/birthweight")
library(tidyverse)
library(data.table)
library(qvalue)

library(IlluminaHumanMethylation450kanno.ilmn12.hg19)
data("IlluminaHumanMethylation450kanno.ilmn12.hg19")
anno       <- IlluminaHumanMethylation450kanno.ilmn12.hg19 %>%
  getAnnotation %>%
  as.data.frame %>%
  rownames_to_column(var="cpg") %>%
  select(cpg,chr,pos,Relation_to_Island,UCSC_RefGene_Name,UCSC_RefGene_Group)

library(IlluminaHumanMethylationEPICanno.ilm10b4.hg19)
data("IlluminaHumanMethylationEPICanno.ilm10b4.hg19")
annoEPIC       <- IlluminaHumanMethylationEPICanno.ilm10b4.hg19 %>%
  getAnnotation %>%
  as.data.frame %>%
  rownames_to_column(var="cpg") %>%
  select(cpg,chr,pos,Relation_to_Island,UCSC_RefGene_Name,UCSC_RefGene_Group)

for(platform in c("450K","EPIC","all")){
  
  if(platform!="450K"){
    pairs = fread("~/mQTL/mqtl.k.450K.sim")
    }else{
    pairs = fread("~/mQTL/mqtl.k.EPIC.sim")}
  
  pairs_conc = pairs %>%
   mutate(p_fdr=p.adjust(`P-value`,method="bonferroni"))%>%
   filter(p_fdr<0.05)
  nrow(pairs_conc)/nrow(pairs)
  sig_CpGs = pairs_conc$CpG
  
  # checking results that are significant with/withou mqtl
  for(race in c("eur","lat")){

  geneCon = fread(paste0("meta.",platform,".bw.",race,".geneCon.ewas.TBL")) %>%
    separate(MarkerName,c("cpg","SNP"),sep="_") %>%
    mutate(Estimate.mqtl=Effect, P.fdr.mqtl=p.adjust(`P-value`,method="fdr"))%>%
    select(cpg,SNP,Estimate.mqtl,P.fdr.mqtl) %>%
    filter(SNP!="none")
  
  NogeneCon = fread(paste0("meta.",platform,".bw.",race,".Nogene.ewas.TBL")) %>%
    separate(MarkerName,c("cpg","SNP"),sep="_") %>%
    mutate(Estimate.nomqtl=Effect, P.fdr.nomqtl=p.adjust(`P-value`,method="fdr"))%>%
    select(cpg,Estimate.nomqtl,P.fdr.nomqtl) 
  
  totGeneCon = geneCon %>% inner_join(NogeneCon,by = "cpg") 
  totGeneConSig = totGeneCon %>% 
    filter(cpg %in% sig_CpGs) %>%
    na.omit() %>%
    filter(P.fdr.mqtl<0.05 | P.fdr.nomqtl<0.05)
  
  if(platform=="450K"){
    totGeneConSigs_anno = totGeneConSig  %>%  left_join(anno,by=c("cpg")) %>%
    mutate(tss=0)
    totGeneConSigs_anno$tss[grep("TSS",totGeneConSigs_anno$UCSC_RefGene_Group)]=1
  }else{
    totGeneConSigs_anno = totGeneConSig  %>%  left_join(annoEPIC,by="cpg") %>%
    mutate(tss=0)
    totGeneConSigs_anno$tss[grep("TSS",totGeneConSigs_anno$UCSC_RefGene_Group)]=1
  }

  
  print(paste0("Significant probes for platform ",platform," in ",race, " among significant mQTLs"))
  print("with SNP adjustment")
  print(table(totGeneConSigs_anno$P.fdr.mqtl<0.05))
  print("without SNP adjustment")
  print(table(totGeneConSigs_anno$P.fdr.nomqtl<0.05))
  
  print(paste0("distributiono of TSS ",platform, "in", race, " among significant mQTLs"))
  print("with SNP adjustment")
  print(table(totGeneConSigs_anno %>% filter(P.fdr.mqtl<0.05) %>% select(tss)))
  print("without SNP adjustment")
  print(table(totGeneConSigs_anno %>% filter(P.fdr.nomqtl<0.05) %>% select(tss)))
  
  write.csv(totGeneConSigs_anno,paste0("bw.mqtl.",platform,".",race,".meta.sigmqtls.csv"))

   totGeneConSig %>%
    na.omit() %>%
    ggplot(aes(x=Estimate.nomqtl,y=Estimate.mqtl))+
    geom_point()+
    geom_abline(aes(intercept=0,slope=1),alpha=0.4,color="red")+
    ggtitle(paste0("Beta value comparison in ",platform," in ",race, "\nSignificant mqtls"))+
    theme_bw()+
    theme(plot.title = element_text(hjust = 0.5))
  ggsave(paste0(platform,".",race,".sig.bw.beta.comp.png"),width = 8,height = 6)

  totGeneConSig %>%
    mutate(logP.fdr.NoMqtl = -log10(P.fdr.nomqtl),
           logP.fdr.mqtl = -log10(P.fdr.mqtl))%>%
    na.omit() %>%
    ggplot(aes(x=logP.fdr.NoMqtl,y=logP.fdr.mqtl))+
    geom_point()+
    geom_abline(aes(intercept=0,slope=1),alpha=0.4,color="red")+
    ggtitle(paste0("P value comparison in ",platform," in ",race, "\nSignificant mqtls"))+
    theme_bw()+
    theme(plot.title = element_text(hjust = 0.5))
  ggsave(paste0(platform,".",race,".sig.bw.pv.comp.png"),width = 8,height = 6)

######################### Pathway analysis of probes whose beta changes (gene*birthweight interaction)
  
  library(methylGSA)
  
  # select probes that are significant either with or without SNP correction
  metaCompare = totGeneConSig %>%
    mutate(delta=abs((Estimate.mqtl-Estimate.nomqtl)/Estimate.nomqtl))%>%
    distinct(cpg,.keep_all = TRUE) %>%
    filter(delta>0.2) %>%
    select(cpg,P.fdr.mqtl) 

  cpg.pval = c(metaCompare$P.fdr.mqtl)
  names(cpg.pval) = metaCompare$cpg
  
  if(length(cpg.pval)>10){
    res1 = methylglm(cpg.pval = cpg.pval, minsize = 2, GS.type = "KEGG")
    res2 = methylglm(cpg.pval = cpg.pval, minsize = 10, GS.type = "GO")
    
    write.table(res1,paste0(platform,".",race,".interaction.kegg.sigmqtls.txt"),sep = "\t", row.names = FALSE, quote = FALSE)
    write.table(res2,paste0(platform,".",race,".interaction.go.sigmqtls.txt"),sep = "\t", row.names = FALSE, quote = FALSE)
  }
  
  if(platform=="450K"){
    metaInt = metaCompare %>% left_join(anno,by="cpg") 
  }else{
    metaInt = metaCompare %>% left_join(annoEPIC,by="cpg") 
  }
  
  write.table(metaInt,paste0(platform,".",race,".interaction.topGenes.sigmqtls.txt"),sep = "\t", row.names = FALSE, quote = FALSE)
  
  }
}
# 
# [1] "Significant probes for platform 450K in eur among significant mQTLs"
# [1] "with SNP adjustment"
# < table of extent 0 >
# [1] "without SNP adjustment"
# < table of extent 0 >
# [1] "distributiono of TSS 450Kineur among significant mQTLs"
# [1] "with SNP adjustment"
# < table of extent 0 >
# [1] "without SNP adjustment"
# < table of extent 0 >
# [1] "Significant probes for platform 450K in lat among significant mQTLs"
# [1] "with SNP adjustment"
# 
# TRUE 
#    3 
# [1] "without SNP adjustment"
# 
# FALSE 
#     3 
# [1] "distributiono of TSS 450Kinlat among significant mQTLs"
# [1] "with SNP adjustment"
# 
# 0 
# 3 
# [1] "without SNP adjustment"
# < table of extent 0 >
# [1] "Significant probes for platform EPIC in eur among significant mQTLs"
# [1] "with SNP adjustment"
# 
# FALSE  TRUE 
#   570    47 
# [1] "without SNP adjustment"
# 
# FALSE  TRUE 
#    19   598 
# [1] "distributiono of TSS EPICineur among significant mQTLs"
# [1] "with SNP adjustment"
# 
#  0  1 
# 38  9 
# [1] "without SNP adjustment"
# 
#   0   1 
# 488 110 
# 
# [1] "Significant probes for platform EPIC in lat among significant mQTLs"
# [1] "with SNP adjustment"
# 
# FALSE  TRUE 
#   283   180 
# [1] "without SNP adjustment"
# 
# FALSE  TRUE 
#   106   357 
# [1] "distributiono of TSS EPICinlat among significant mQTLs"
# [1] "with SNP adjustment"
# 
#   0   1 
# 143  37 
# [1] "without SNP adjustment"
# 
#   0   1 
# 270  87 
# 
# [1] "Significant probes for platform all in eur among significant mQTLs"
# [1] "with SNP adjustment"
# 
# FALSE  TRUE 
#  1008    39 
# [1] "without SNP adjustment"
# 
# FALSE  TRUE 
#    20  1027 
# [1] "distributiono of TSS allineur among significant mQTLs"
# [1] "with SNP adjustment"
# 
#  0  1 
# 29 10 
# [1] "without SNP adjustment"
# 
#   0   1 
# 817 210 
# 
# [1] "Significant probes for platform all in lat among significant mQTLs"
# [1] "with SNP adjustment"
# 
# FALSE  TRUE 
#   783   215 
# [1] "without SNP adjustment"
# 
# FALSE  TRUE 
#   102   896 
# [1] "distributiono of TSS allinlat among significant mQTLs"
# [1] "with SNP adjustment"
# 
#   0   1 
# 176  39 
# [1] "without SNP adjustment"
# 
#   0   1 
# 682 214 
# retrieving KEGG sets...
# 106 gene sets are being tested...
# Done!
# retrieving GO sets...
# 2896 gene sets are being tested...
# Done!
# There were 50 or more warnings (use warnings() to see the first 50)


```
