---
output: html_document
editor_options: 
  chunk_output_type: console
---

Project description: to conduct an EWAS analysis on DNA methylation and Down Syndrome subjects, and identify methylation change.

This paper has been published: 
Shaobo Li et al, Epigenome-wide association study of acute lymphoblastic leukemia in children with Down syndrome
DOI: 10.1182/bloodadvances.2022007098

##### DS_ALL_EWAS analysis

Addressing reviewers' questions 

```{r methods correction}

setwd("~/DSALL/usc")
library(tidyverse)
library(data.table)
library(minfi)
library(factoextra)

# sex check of subjects

cov = fread("ds.all.covaraites.txt") %>% # generated with the next block
  as.data.frame() %>%
  dplyr::select(beadPosition,sex)
RGset <- read_rds("~/sets/set3/RGset.rds")
MSet <- preprocessRaw(RGset)
gmset <- mapToGenome(MSet)
sexMethyl <- getSex(gmset)
#rm(list=c("MSet","gmset"))

sexMethyl <- sexMethyl %>%
  as.data.frame %>%
  dplyr::select(predictedSex) %>%
  rownames_to_column(var="beadPosition") %>%
  inner_join(cov,by="beadPosition")

sexMethyl %>%
  filter((sex==1&predictedSex=="F")|(sex==2&predictedSex=="M"))

# 202148010085_R03C01
# one person has disconcordant sex, and will be dropped

# potential cryptic relatedness among individuals
cov = fread("ds.all.covaraites.postReviewQCs.txt") %>% # generated with the next block
  as.data.frame() %>%
  dplyr::select(starts_with("epi"))

cov.pca = prcomp(cov, scale = FALSE)

fviz_eig(cov.pca)

#options(ggrepel.max.overlaps = Inf)
png("ds.all.pca.png")
fviz_pca_ind(cov.pca, geom="point",col.ind="red",alpha=0.8)
dev.off()

```

----------------

export TMPDIR=/ccls/home/sli/tempos

```{r M value then bacon including the 18 washington cases}

setwd("~/DSALL/usc")

library(tidyverse)
library(data.table)
library(MASS)
library(lmtest)
library(sandwich)
library(parallel) 
library(minfi)
library(wateRmelon)
library(bacon)
library(lumi)

cl <- makeCluster(40, type="FORK")
clusterEvalQ(cl,library(MASS))
clusterEvalQ(cl,library(lmtest))

RLM.Robust <- function(meth, cov){

    # meth = methyl["cg25813447",]
    # cov = covs
  
    dat <- data.frame(y = meth, cov)
    ff  <- formula(paste("y ~ ", paste0(colnames(cov), collapse="+")))
    mod <- try(rlm(ff, data = dat, maxit = 200))
    #cf = try(coeftest(mod, vcov=vcovHC(mod, type="HC0")))
    cf = try(coeftest(mod))
    Ncase = nrow(dat %>% filter(CaCo==1) %>% na.omit)
    Ncontrol = nrow(dat %>% filter(CaCo==0) %>% na.omit)

  	if (exists(class(cf)) == FALSE) {
  		out = c(rep(NA,4),Ncase,Ncontrol)
  	} else if (exists(class(cf)) == TRUE) {
  		if( class(cf)[1] == "try-error") {
  		cat("error throw by CpG", set, ";")
  		out = c(rep(NA,4),Ncase,Ncontrol)
  		print(paste0(cov, " - chol2inv could not be computed"))
  		} 
  		else {
  			coef = cf[2,"Estimate"]
  			se = cf[2,"Std. Error"]
  			z.value = cf[2,"z value"]
  			p.value = cf[2,"Pr(>|z|)"]
  			out = c(coef, se, z.value,p.value,Ncase,Ncontrol) 
  		}
    }
  names(out) = c("Beta","SE", "Z","P","ncase","ncontrol"); 
  return(out) 
  }

# Winsorize beta values
winsorize <- function(methylation,pct=winsorize.pct) {
  quantiles <- matrixStats::rowQuantiles(methylation, probs=c(pct,1-pct), na.rm=T)
  low <- quantiles[,1]
  upper <- quantiles[,2]

  outliers.lower <- rowSums(methylation < low, na.rm=T)
  outliers.upper <- rowSums(methylation > upper, na.rm=T)
  
  idx <- which(methylation < low, arr.ind=T)
  methylation[idx] <- low[idx[,1]]
  
  idx <- which(methylation > upper, arr.ind=T)
  methylation[idx] <- upper[idx[,1]]

  n <- rowSums(!is.na(methylation))
  log <- data.frame(outliers.lower, outliers.upper, n)
  return(list(methylation=methylation, log=log))
}

# prepare a file of subjects (called var) with deconvoluted cell proportions, sex and plate, 10 EpiStructure PCs

dcp = fread("~/sets/set3/idol_deconvolution.csv") %>%
  as.data.frame() %>%
  dplyr::select(-V1,-subjectId) %>%
  na.omit()

glint = fread("~/sets/set3/glintControl.csv") %>%
    dplyr::select(beadPosition, epi1, epi2, epi3, epi4, epi5, epi6, epi7, epi8, epi9, epi10)%>%
    na.omit()

vars = fread("~/sets/set3/clinical_variables.csv") %>%
  as.data.frame() %>%
  mutate(plate=Batch) %>%
  dplyr::filter(Trisomy21==1) %>%
  dplyr::select(beadPosition,subjectId,sex,plate,CaCo)%>%
  inner_join(dcp,by="beadPosition")%>%
  inner_join(glint,by="beadPosition")%>%
  na.omit() %>%
  filter(beadPosition!="202148010085_R03C01") # failed sex check

table(vars$CaCo)
#  0   1 
#198 144 

## without deleting the person who has the wrong sex
#  0   1 
#198 145 

################################################################
betas3_0 = fread("~/sets/set3/set3_Sesame_beta.csv") %>%
  as.data.frame()%>%
  dplyr::select(-V1)
betas3_1 = betas3_0 %>%
  column_to_rownames(var="probeId")
# 865918    944

#sum(is.na(betas3_0))
#115271259

# betas3_0 = read_rds("~/sets/set3/set3_Sesame_beta_imputed.rds") %>%
#   as.data.frame()
# rownames(betas3_0) <- c()
# betas3_1 = betas3_0 %>%
#   column_to_rownames(var="probeId")
# 722664    929

subsInTheStudy = intersect(vars$beadPosition,colnames(betas3_1))

noob.to.include = betas3_1[,subsInTheStudy] %>% as.matrix # 865918    342
noob.to.include = noob.to.include[rowMeans(is.na(noob.to.include))<0.05,] #  702850    342
#noob.to.include = noob.to.include[,colMeans(is.na(noob.to.include))<0.05] #  702850    335
  
noob.to.include.M = beta2m(noob.to.include)

replace.outliers <- winsorize(noob.to.include.M, 0.005)
methyl <- replace.outliers$methylation
outlier.log <- replace.outliers$log

vars = vars %>%
  column_to_rownames(var="beadPosition")

covs = vars[subsInTheStudy,] %>%
  dplyr::select(CaCo,sex,plate,CD8T,CD4T,NK,Bcell,Mono,nRBC,epi1,
                epi2,epi3,epi4,epi5,epi6,
                epi7,epi8,epi9,epi10)%>%
  mutate(plate=as.factor(plate))

head(covs,3)

print(table(covs$CaCo))
#  0   1 
#198 144 

write_rds(covs,"covs.postReviewQCs.rds")
write_rds(methyl,"methyl.postReviewQCs.rds")

#covs1 = vars[subsInTheStudy,] %>% rownames_to_column(var="beadPosition")
#write.table(covs1,"ds.all.covaraites.postReviewQCs.txt",quote=FALSE,row.names=FALSE,sep='\t')

#subsInTheStudyID = vars[subsInTheStudy,]%>% rownames_to_column(var="beadPosition") %>% dplyr::select(beadPosition,subjectId)
# write.table(subsInTheStudyID,"ds.all.final.subs.AfterReview.txt",
#             quote=FALSE,
#             sep='\t',
#             row.names = FALSE)

# this has to be true to continue

if(identical(rownames(covs),colnames(methyl))==TRUE){

  # methylt = methyl[1:100,] 
  # results = t(apply(methylt,1,RLM.Robust,covs))
  
  results <- t(parApply(cl, methyl, 1, RLM.Robust, covs)) 
  results <- as.data.frame(results) %>%
    na.omit()
  
  #results$bacon_P = pval(bacon(results$Z,na.exclude = TRUE))
  
  write.table(results,"ds.all.usc.sesame.m.ewas.txt",
        quote=FALSE,
        sep='\t')  

  
  # raw P values
  # Nbeta <- sum(!is.na(results$P))
  # tmp <- results[!is.na(results$P),]
  # lambda = median(qchisq(1- tmp$P,1))/qchisq(0.5,1)
  # print(round(lambda, 2))
  
  # bacon P values
  # Nbeta <- sum(!is.na(results$bacon_P))
  # tmp <- results[!is.na(results$bacon_P),]
  # lambda = median(qchisq(1- tmp$bacon_P,1))/qchisq(0.5,1)
  # print(round(lambda, 2))
  

}

stopCluster(cl)

#####################################################################

# looking at the results
setwd("~/DSALL/usc")
library(tidyverse)
library(data.table)
library(bacon)
library(IlluminaHumanMethylationEPICanno.ilm10b4.hg19)
data("IlluminaHumanMethylationEPICanno.ilm10b4.hg19")
annoEPIC <- IlluminaHumanMethylationEPICanno.ilm10b4.hg19 %>%
  getAnnotation %>%
  as.data.frame %>%
  rownames_to_column(var="cpg") %>%
  dplyr::select(cpg,chr,pos,Relation_to_Island,UCSC_RefGene_Name,
                Probe_maf,UCSC_RefGene_Group,CpG_maf)

results <- fread("ds.all.usc.sesame.m.ewas.txt") %>% 
  dplyr::rename("cpg"="V1") %>%
  dplyr::filter(grepl('cg', cpg)) %>%
  filter(!is.na(P))%>% 
  left_join(annoEPIC,by="cpg") %>%
  arrange(by=P) %>% # 701293
  dplyr::filter(chr!="chrX"&chr!="chrY") %>%  #687445
  dplyr::filter(Probe_maf<0.05|is.na(Probe_maf)) %>% # 629078
  dplyr::filter(CpG_maf<0.05|is.na(CpG_maf))  # 628764

results$bacon_P = pval(bacon(results$Z,na.exclude = TRUE))

results <- results %>%
  mutate(fdr=p.adjust(bacon_P, method = "fdr", n=length(bacon_P))) %>%
  mutate(bf=p.adjust(bacon_P, method = "bonferroni", n=length(bacon_P))) 

write.csv(results,"ds.all.usc.sesame.m.ewas.anno.csv")

0.05/nrow(results) # bonferroni threshold
# 7.952109e-08


# raw P values
Nbeta <- sum(!is.na(results$P))
tmp <- results[!is.na(results$P),]
lambda = median(qchisq(1- tmp$P,1))/qchisq(0.5,1)
print(round(lambda, 2))

# bacon P values
Nbeta <- sum(!is.na(results$bacon_P))
tmp <- results[!is.na(results$bacon_P),]
lambda = median(qchisq(1- tmp$bacon_P,1))/qchisq(0.5,1)
print(round(lambda, 2))

  
# figures (qq plot and manhattan)
library(CMplot)
library(QCEWAS)
library(qqman)

usc_model_anno <- results %>%
  dplyr::mutate(cpg_name=cpg,estimate=Beta, std_error=SE,p_val=bacon_P)

sig_thresghold = usc_model_anno %>%
  filter(fdr<0.05)%>%
  arrange(desc(bacon_P))
sig_threshold = sig_thresghold[1,"p_val"]

annotation = usc_model_anno%>%
  filter(bf<0.05) %>%
  filter(UCSC_RefGene_Name!="") %>%
  mutate(gene.highlight = sub(";.*$","",UCSC_RefGene_Name))
cpg.highlight = annotation$cpg_name
genes.highlight = annotation$gene.highlight

outcomes_model_CM_plot <- usc_model_anno%>%
  dplyr::mutate(p_val_cmplot = -log10(p_val))%>%
  dplyr::mutate(p_val_cmplot= ifelse(estimate<0,-p_val_cmplot, p_val_cmplot )) %>%
  dplyr::select(cpg_name, chr,pos, p_val_cmplot)
outcomes_model_CM_plot$chr <- gsub("chr","",outcomes_model_CM_plot$chr)

CMplot(outcomes_model_CM_plot,
       plot.type="m",
       band=0, 
       cex=c(0.5,0.5,0.5),
       amplify=FALSE,
       LOG10=FALSE,  
       signal.col=NULL,
       chr.den.col=NULL,
       threshold=c(log10(sig_threshold[[1]]), -log10(sig_threshold[[1]])),
       threshold.lty=2, 
       threshold.lwd=1, 
       threshold.col="black", 
       highlight=cpg.highlight,
       highlight.text=genes.highlight, 
       file="jpg",
       memo="usc.sesame.m",
       #main="USC DS-ALL EWAS (SESAME)",
       dpi=500,file.output=TRUE,verbose=TRUE)

outcomes_model_qq <- usc_model_anno %>% 
  dplyr::select(cpg_name,p_val)%>%
  dplyr::mutate(ID=cpg_name, P=p_val)
lambda <- P_lambda(outcomes_model_qq$P)

#png("usc.sesame.qq.m.png")
tiff("usc.sesame.qq.m.tiff", width = 10, height = 7, units = 'in', res = 300)
qq(outcomes_model_qq$P)
#   main=paste0("USC dataset \n lambda value= ",
#               round(lambda,3)))
dev.off()

## top CpGs
head(results)
dim(results%>%filter(fdr<0.05))
# 31
dim(results%>%filter(bf<0.05))
# 11

results%>%filter(fdr<0.05)

# pathway analysis
library(methylGSA)
library(IlluminaHumanMethylationEPICanno.ilm10b4.hg19)
library(IlluminaHumanMethylation450kanno.ilmn12.hg19)

down_all <- results %>%
  filter(fdr<0.05) %>%
  mutate(cpg_name=cpg,estimate=Beta,std_error=SE,p_val=bacon_P) %>%
  dplyr::select(cpg_name,chr,pos,estimate,std_error,p_val)

down_100_vec = as.numeric(down_all$p_val)
names(down_100_vec) = down_all$cpg_name
res1 = methylglm(cpg.pval = down_100_vec, minsize=100, GS.type = "KEGG")
res2 = methylglm(cpg.pval = down_100_vec, minsize=100, GS.type = "GO")

write.csv(res1,"down_M_100_KEGGpathway.sesame.csv")
write.csv(res2,"down_M_100_GOpathway.sesame.csv")

## Check RUNX1 and FLI1 to see if their methylation was different in DS with vs without ALL
# results_lookup = results %>%
#   filter(grepl("RUNX1|FLI1",UCSC_RefGene_Name)) 
# write.table(results_lookup,"runx1.fli1.txt",
#             quote=FALSE,
#             row.names = FALSE,
#             sep='\t')

## ToppFun within the ToppGene suite
## https://toppgene.cchmc.org/input_enrichment.jsp
# results_lookup = results %>%
#   filter(fdr<0.05) %>%
#   mutate(geneName = gsub(";.*$","",UCSC_RefGene_Name)) %>%
#   filter(geneName!="") %>%
#   distinct(geneName,.keep_all=TRUE)%>%
#   dplyr::select(geneName)
### NO significant disease


## DMR analysis (overlap of comb-p and DMRCATE)
### r run DMRCate in USC dataset}

setwd("~/DSALL/usc")
library(DMRcate)
library(tidyverse)
library(limma)
library(plyr)

MA_DMRcate <- results%>%
  dplyr::mutate(ID=cpg, stat=Beta/SE, CHR=chr, betafc=Beta,raw = bacon_P, indfdr = fdr)%>%
  dplyr::mutate(is.sig=ifelse(indfdr<0.05,TRUE,FALSE))%>%
  dplyr::select(ID, stat,CHR,pos,raw,betafc,indfdr,is.sig)%>%
  filter(!is.na(CHR))
MA_DMRcate_list <- as.list(MA_DMRcate)
class(MA_DMRcate_list) <- "annot"

dmrcate <- function (object, lambda = 1000, C = NULL, p.adjust.method = "BH", 
    pcutoff = "fdr", consec = FALSE, conseclambda = 10, betacutoff = NULL, 
    min.cpgs = 2, mc.cores = 1) {
    stopifnot(is(object, "annot"))
    stopifnot(lambda >= 1)
    stopifnot(pcutoff == "fdr" | (0 <= pcutoff & pcutoff <= 1))
    stopifnot(C >= 0.2)
    if (consec & is.null(conseclambda)) {
        stop("Consecutive CpG bandwidth must be specified")
    }
    object <- data.frame(ID = object$ID, weights = abs(object$stat), 
        CHR = as.character(object$CHR), pos = object$pos, betafc = object$betafc, 
        indfdr = object$indfdr, is.sig = object$is.sig)
    object <- object[order(object$CHR, object$pos), ]
    if (is.null(C) & !consec) {
        if (nrow(object) < 9e+05) {
            C = 2
        }
        else {
            C = 50
        }
    }
    if (consec) {
        lambda = conseclambda
        message(paste("Consecutive mode specified, lambda is now set at", 
            conseclambda, "consecutive CpGs."))
        if (is.null(C)) {
            stop("Error: argument C must be specified (in CpG sites) for consecutive mode.")
        }
        object$realcoordforconsec <- object$pos
        object$pos <- unlist(sapply(as.numeric(table(object$CHR)), 
            function(x) 1:x))
    }
    lag = lambda
    chr.unique <- unique(c(as.character(object$CHR)))
    fitted <- mclapply(chr.unique, fitParallel, object = object, 
        consec = consec, conseclambda = conseclambda, lambda = lambda, 
        C = C, mc.cores = mc.cores)
    object <- rbind.fill(fitted)
    object$fdr <- p.adjust(object$raw, method = p.adjust.method)
    if (pcutoff == "fdr") {
        nsig <- sum(object$is.sig)
        if (nsig == 0) {
            txt <- "The FDR you specified in cpg.annotate() returned no significant CpGs, hence there are no DMRs.\n    Try specifying a value of 'pcutoff' in dmrcate() and/or increasing 'fdr' in cpg.annotate()."
            stop(paste(strwrap(txt, exdent = 2), collapse = "\n"))
        }
        pcutoff <- sort(object$fdr)[nsig]
    }
    object$sig <- (object$fdr <= pcutoff)
    if (nrow(object) == 0) {
        txt <- "No signficant regions found. Try increasing the value of\n    'pcutoff' in dmrcate() and/or 'fdr' in cpg.annotate()."
        stop(paste(strwrap(txt, exdent = 2), collapse = "\n"))
    }
    message("Demarcating regions...")
    chr.N <- as.character(object$CHR)
    pos.N <- object$pos
    sig.N <- object$sig
    N <- length(sig.N)
    n.K <- which(sig.N)
    K <- length(n.K)
    stopifnot(K >= 2)
    pos.K <- pos.N[n.K]
    chr.K <- chr.N[n.K]
    jump_chr.k <- (chr.K[-1] != chr.K[-K])
    jump_pos.k <- (diff(pos.K) > lag)
    jump.k <- (jump_chr.k | jump_pos.k)
    ksegments.A2 <- Segment(jump.k)
    A <- nrow(ksegments.A2)
    kstart.A <- ksegments.A2[, "start"]
    kend.A <- ksegments.A2[, "end"]
    realpos.K <- pos.K
    if (consec) {
        realpos.N <- object$realcoordforconsec
        realpos.K <- realpos.N[n.K]
    }
    start.A <- realpos.K[kstart.A]
    end.A <- realpos.K[kend.A]
    chr.A <- chr.K[kstart.A]
    stopifnot(all(chr.K[kend.A] == chr.A))
    fmt <- "%s:%1d-%1d"
    coord.A <- sprintf(fmt, chr.A, start.A, end.A)
    nstart.A <- n.K[kstart.A]
    nend.A <- n.K[kend.A]
    width.A <- nend.A + 1 - nstart.A
    a.Z <- rep(seq(A), width.A)
    fn <- function(a) seq(from = nstart.A[a], to = nend.A[a])
    l.listA <- lapply(seq(A), fn)
    n.Z <- unlist(l.listA)
    region.N <- rep(NA_integer_, N)
    region.N[n.Z] <- a.Z
    levels <- seq(A)
    region.N <- factor(region.N, levels = levels)
    no_cpg.A <- c(table(region.N))
    REGIONSTAT <- function(field, fn) {
        x.N <- object[[field]]
        x.R <- tapply(x.N, region.N, fn)
        c(x.R)
    }
    fn_Stouffer <- function(x) pnorm(sum(qnorm(x))/sqrt(length(x)))
    fn_max <- function(x) x[which.max(abs(x))]
    results <- data.frame(coord = coord.A, no.cpgs = no_cpg.A, 
        minfdr = REGIONSTAT("fdr", min), Stouffer = REGIONSTAT("indfdr", 
            fn_Stouffer), maxbetafc = REGIONSTAT("betafc", fn_max), 
        meanbetafc = REGIONSTAT("betafc", mean), row.names = seq(A), 
        stringsAsFactors = FALSE)
    keep <- (results$no.cpgs >= min.cpgs)
    results <- results[keep, ]
    if (!is.null(betacutoff)) {
        keep <- (abs(results$meanbetafc) >= betacutoff)
        results <- results[keep, ]
    }
    o <- order(results$Stouffer, -results$no.cpgs)
    results <- results[o, ]
    message("Done!")
    output <- NULL
    output$input <- object
    output$results <- results
    output$cutoff <- pcutoff
    class(output) <- "dmrcate.output"
    output
}

dmrcoutput <- dmrcate(MA_DMRcate_list, lambda=1000, C=2, pcutoff = 0.01,min.cpgs = 2)
outcome_dmrcate <- dmrcoutput$results%>%
  write_rds("DMRCATE_USC_M_VAL_DS_EWAS_MA_output.rds")

### comb-p 
usc <- results %>%
  dplyr::mutate(chrom=chr, end=pos+1, start=pos, p_val=bacon_P)%>%
  dplyr::filter(grepl('cg', cpg)) %>%
  dplyr::select(chrom, start, end, p_val)%>%
  na.omit()%>%
  arrange(chrom, start)

usc%>%
  write_tsv("comb_p_USC_SESAME_M_VAL_ALL_DS_EWAS_MA.txt")

```

```{r PCA analysis of all subjects}

setwd("~/DSALL/usc")

library(tidyverse)
library(data.table)
library(impute)

betas3_0 = fread("~/sets/set3/set3_Sesame_beta.csv") %>%
  as.data.frame()%>%
  dplyr::select(-V1)
betas3_1 = betas3_0 %>%
  column_to_rownames(var="probeId")
betas3_2 = betas3_1[rowMeans(is.na(betas3_1))<0.05,] %>%
  as.matrix()

subs = fread("ds.all.covaraites.postReviewQCs.txt")
clinical3 = fread("~/sets/set3/clinical_variables.csv") %>%
  filter(beadPosition %in% subs$beadPosition) %>%
  mutate(source=ifelse(Sample_Source=="Washington Biobank",'wash','cali')) %>%
  dplyr::select(beadPosition,source,CaCo) %>%
  column_to_rownames(var='beadPosition')

subsInTheStudy = intersect(rownames(clinical3),colnames(betas3_2))

betafile = betas3_2[,subsInTheStudy]
betaFileImp = impute.knn(betafile)$data

betaFileImp_p = betaFileImp %>% t %>% as.data.frame()

clinicalFile = clinical3[subsInTheStudy,] 
#table(clinicalFile$source)

#pca
pca_res = prcomp(betaFileImp_p, scale. = TRUE)

dataPCs = cbind(clinicalFile,pca_res$x) %>% 
  as.data.frame %>%
  mutate(CaCo = as.factor(CaCo),
         source=as.factor(source))

write_rds(dataPCs,"dataPCs.rds")

dataPCs %>%
  ggplot(aes(x=PC2,y=PC18))+
  geom_point(aes(color=source))+
  theme_bw()

#tsne
library(Rtsne)
tSNE_fit <- betaFileImp_p %>%
  as.matrix() %>%
  scale() %>% 
  Rtsne()

tSNE_df <- tSNE_fit$Y %>% 
  as.data.frame() %>%
  dplyr::rename(tSNE1="V1",
         tSNE2="V2") 
rownames(tSNE_df) = rownames(betaFileImp_p)

datatSNEs = cbind(clinicalFile,tSNE_df) %>% 
  as.data.frame %>%
  mutate(CaCo = as.factor(CaCo),
         source=as.factor(source))

datatSNEs %>%
  ggplot(aes(x=tSNE1,y=tSNE2))+
  geom_point(aes(color=source))+
  theme_bw()


```


```{r Repeat DS-ALL EWAS after removing the 18 or so Washington ALL cases and assess inflation (and top hits)}

setwd("~/DSALL/usc")
library(tidyverse)
library(data.table)
library(MASS)
library(lmtest)
library(sandwich)
library(parallel) 
library(minfi)
library(wateRmelon)
library(bacon)
library(lumi)

RLM.Robust <- function(meth, cov){

    dat <- data.frame(y = meth, cov)
    ff  <- formula(paste("y ~ ", paste0(colnames(cov), collapse="+")))
    mod <- try(rlm(ff, data = dat, maxit = 200))
    cf = try(coeftest(mod))
    Ncase = nrow(dat %>% filter(CaCo==1) %>% na.omit)
    Ncontrol = nrow(dat %>% filter(CaCo==0) %>% na.omit)

  	if (exists(class(cf)) == FALSE) {
  		out = c(rep(NA,4),Ncase,Ncontrol)
  	} else if (exists(class(cf)) == TRUE) {
  		if( class(cf)[1] == "try-error") {
  		cat("error throw by CpG", set, ";")
  		out = c(rep(NA,4),Ncase,Ncontrol)
  		print(paste0(cov, " - chol2inv could not be computed"))
  		} 
  		else {
  			coef = cf[2,"Estimate"]
  			se = cf[2,"Std. Error"]
  			z.value = cf[2,"z value"]
  			p.value = cf[2,"Pr(>|z|)"]
  			out = c(coef, se, z.value,p.value,Ncase,Ncontrol) 
  		}
    }
  names(out) = c("Beta","SE", "Z","P","ncase","ncontrol"); 
  return(out) 
  }

clinical3 = fread("~/sets/set3/clinical_variables.csv") %>%
  filter(Trisomy21==1) %>%
  filter(Sample_Source=="Washington Biobank") %>%
  dplyr::select(beadPosition)

covs = read_rds("covs.postReviewQCs.rds") 
covs1 = covs[!rownames(covs)%in%clinical3$beadPosition,]
#write_rds(covs1,"covs.postReviewQCs.dropWashU.rds") 
table(covs1$CaCo)
#   0   1 
# 198 126 

vars1 = fread("~/sets/set3/clinical_variables.csv") %>%
  dplyr::select(beadPosition,subjectId)%>%
  filter(beadPosition %in% rownames(covs1))
#write.table(vars1,"ds.all.final.subs.AfterReview.dropWashU.txt",quote=FALSE,sep='\t',row.names = FALSE)

methyl = read_rds("methyl.postReviewQCs.rds")
methyl1 = methyl[,rownames(covs1)]
#write_rds(methyl1,"methyl.postReviewQCs.dropWashU.rds")

cl <- makeCluster(40, type="FORK")
clusterEvalQ(cl,library(MASS))
clusterEvalQ(cl,library(lmtest))

# this has to be true to continue

if(identical(rownames(covs1),colnames(methyl1))==TRUE){

  # methylt = methyl1[1:100,] 
  # results = t(apply(methylt,1,RLM.Robust,covs1))
  
  results <- t(parApply(cl, methyl1, 1, RLM.Robust, covs1)) 
  results <- as.data.frame(results) %>%
    na.omit()
  
  write.table(results,"ds.all.usc.sesame.m.nowashington.ewas.txt",
        quote=FALSE,
        sep='\t')  
}

stopCluster(cl)

###########################################################################
# looking at the results

setwd("~/DSALL/usc")
library(tidyverse)
library(data.table)
library(bacon)
library(IlluminaHumanMethylationEPICanno.ilm10b4.hg19)
data("IlluminaHumanMethylationEPICanno.ilm10b4.hg19")
annoEPIC <- IlluminaHumanMethylationEPICanno.ilm10b4.hg19 %>%
  getAnnotation %>%
  as.data.frame %>%
  rownames_to_column(var="cpg") %>%
  dplyr::select(cpg,chr,pos,Relation_to_Island,UCSC_RefGene_Name,
                Probe_maf,UCSC_RefGene_Group,CpG_maf,UCSC_RefGene_Group,
                Regulatory_Feature_Group)

results <- fread("ds.all.usc.sesame.m.nowashington.ewas.txt") %>% #702850
  dplyr::rename("cpg"="V1") %>%
  dplyr::filter(grepl('cg', cpg)) %>%
  filter(!is.na(P))%>%  #701293
  left_join(annoEPIC,by="cpg") %>%
  arrange(by=P) %>% 
  dplyr::filter(chr!="chrX"&chr!="chrY") %>% #687445
  dplyr::filter(Probe_maf<0.05|is.na(Probe_maf)) %>%  # 629078
  dplyr::filter(CpG_maf<0.05|is.na(CpG_maf)) #628764

results$bacon_P = pval(bacon(results$Z,na.exclude = TRUE))

results <- results %>%
  mutate(fdr=p.adjust(bacon_P, method = "fdr", n=length(bacon_P))) %>%
  mutate(bf=p.adjust(bacon_P, method = "bonferroni", n=length(bacon_P))) 

write.csv(results,"ds.all.usc.sesame.m.nowashington.ewas.anno.csv")

#[1] "EBF1"                                                         
#[2] "TRIM13;TRIM13;TRIM13;TRIM13;TRIM13;DLEU2;TRIM13;TRIM13;TRIM13"
#[3] "TRIM13;TRIM13;TRIM13;TRIM13;TRIM13;DLEU2;TRIM13;TRIM13;TRIM13"
#[4] "CCK"                                                          
#[5] "CTNNA2;CTNNA2;CTNNA2;CTNNA2"                                  
#[6] "CPLX2;CPLX2;CPLX2"        

0.05/nrow(results) # bonferroni threshold
# 7.952109e-08

# raw P values
Nbeta <- sum(!is.na(results$P))
tmp <- results[!is.na(results$P),]
lambda = median(qchisq(1- tmp$P,1))/qchisq(0.5,1)
print(round(lambda, 2))
# 1.24

# bacon P values
Nbeta <- sum(!is.na(results$bacon_P))
tmp <- results[!is.na(results$bacon_P),]
lambda = median(qchisq(1- tmp$bacon_P,1))/qchisq(0.5,1)
print(round(lambda, 2))
# 1.11

## top CpGs
head(results)
dim(results%>%filter(fdr<0.05))
# 38
dim(results%>%filter(bf<0.05))
# 10

results%>%filter(fdr<0.05)
write.csv(results%>%filter(fdr<0.05),"no.washington.top.csv")

###########################################################################
# figures (qq plot and manhattan)
library(CMplot)
library(QCEWAS)
library(qqman)

usc_model_anno <- results %>%
  arrange(log(fdr))%>%
  dplyr::mutate(cpg_name=cpg,estimate=Beta, std_error=SE,p_val=bacon_P)

sig_thresghold = usc_model_anno %>%
  filter(fdr<0.05)%>%
  arrange(desc(bacon_P))
sig_threshold = sig_thresghold[1,"p_val"]

annotation = usc_model_anno%>%
  filter(bf<0.05) %>%
  filter(UCSC_RefGene_Name!="") %>%
  mutate(gene.highlight = sub(";.*$","",UCSC_RefGene_Name))
cpg.highlight = annotation$cpg_name
genes.highlight = annotation$gene.highlight

outcomes_model_CM_plot <- usc_model_anno%>%
  dplyr::mutate(p_val_cmplot = -log10(p_val))%>%
  dplyr::mutate(p_val_cmplot= ifelse(estimate<0,-p_val_cmplot, p_val_cmplot )) %>%
  dplyr::select(cpg_name, chr,pos, p_val_cmplot)
outcomes_model_CM_plot$chr <- gsub("chr","",outcomes_model_CM_plot$chr)

CMplot(outcomes_model_CM_plot,
       plot.type="m",
       band=0, 
       cex=c(0.5,0.5,0.5),
       amplify=FALSE,
       LOG10=FALSE,  
       signal.col=NULL,
       chr.den.col=NULL,
       threshold=c(log10(sig_threshold[[1]]), -log10(sig_threshold[[1]])),
       threshold.lty=2, 
       threshold.lwd=1, 
       threshold.col="black", 
       highlight=cpg.highlight,
       highlight.text=genes.highlight, 
       file="jpg",
       memo="usc.sesame.m.drop.washU",
       dpi=500,file.output=TRUE,verbose=TRUE)

outcomes_model_qq <- usc_model_anno %>% 
  dplyr::select(cpg_name,p_val)%>%
  dplyr::mutate(ID=cpg_name, P=p_val)
lambda <- P_lambda(outcomes_model_qq$P)

tiff("usc.drop.washu.sesame.qq.m.tiff", width = 10, height = 7, units = 'in', res = 300)
qq(outcomes_model_qq$P)
dev.off()

###########################################################################
# pathway analysis
library(methylGSA)
library(IlluminaHumanMethylationEPICanno.ilm10b4.hg19)

down_all <- results %>%
  filter(fdr<0.05) %>%
  mutate(cpg_name=cpg,estimate=Beta,std_error=SE,p_val=bacon_P) %>%
  dplyr::select(cpg_name,chr,pos,estimate,std_error,p_val)

down_100_vec = as.numeric(down_all$p_val)
names(down_100_vec) = down_all$cpg_name
res1 = methylglm(cpg.pval = down_100_vec, GS.type = "KEGG",array.type = "EPIC")
res2 = methylglm(cpg.pval = down_100_vec, GS.type = "GO", array.type = "EPIC")

write.csv(res1,"down_drop.washu.M_100_KEGGpathway.sesame.csv")
write.csv(res2,"down_drop.washu.M_100_GOpathway.sesame.csv")


###########################################################################
## DMR analysis (overlap of comb-p and DMRCATE)
### r run DMRCate in USC dataset}
setwd("~/DSALL/usc")
library(DMRcate)
library(tidyverse)
library(limma)
library(plyr)

#results = read.csv("ds.all.usc.sesame.m.nowashington.ewas.anno.csv")
MA_DMRcate <- results%>%
  dplyr::mutate(ID=cpg, stat=Beta/SE, CHR=chr, betafc=Beta,raw = bacon_P, indfdr = fdr)%>%
  dplyr::mutate(is.sig=ifelse(indfdr<0.05,TRUE,FALSE))%>%
  dplyr::select(ID, stat,CHR,pos,raw,betafc,indfdr,is.sig)%>%
  filter(!is.na(CHR))
MA_DMRcate_list <- as.list(MA_DMRcate)
class(MA_DMRcate_list) <- "annot"

dmrcate <- function (object, lambda = 1000, C = NULL, p.adjust.method = "BH", 
    pcutoff = "fdr", consec = FALSE, conseclambda = 10, betacutoff = NULL, 
    min.cpgs = 2, mc.cores = 1) {
    stopifnot(is(object, "annot"))
    stopifnot(lambda >= 1)
    stopifnot(pcutoff == "fdr" | (0 <= pcutoff & pcutoff <= 1))
    stopifnot(C >= 0.2)
    if (consec & is.null(conseclambda)) {
        stop("Consecutive CpG bandwidth must be specified")
    }
    object <- data.frame(ID = object$ID, weights = abs(object$stat), 
        CHR = as.character(object$CHR), pos = object$pos, betafc = object$betafc, 
        indfdr = object$indfdr, is.sig = object$is.sig)
    object <- object[order(object$CHR, object$pos), ]
    if (is.null(C) & !consec) {
        if (nrow(object) < 9e+05) {
            C = 2
        }
        else {
            C = 50
        }
    }
    if (consec) {
        lambda = conseclambda
        message(paste("Consecutive mode specified, lambda is now set at", 
            conseclambda, "consecutive CpGs."))
        if (is.null(C)) {
            stop("Error: argument C must be specified (in CpG sites) for consecutive mode.")
        }
        object$realcoordforconsec <- object$pos
        object$pos <- unlist(sapply(as.numeric(table(object$CHR)), 
            function(x) 1:x))
    }
    lag = lambda
    chr.unique <- unique(c(as.character(object$CHR)))
    fitted <- mclapply(chr.unique, fitParallel, object = object, 
        consec = consec, conseclambda = conseclambda, lambda = lambda, 
        C = C, mc.cores = mc.cores)
    object <- rbind.fill(fitted)
    object$fdr <- p.adjust(object$raw, method = p.adjust.method)
    if (pcutoff == "fdr") {
        nsig <- sum(object$is.sig)
        if (nsig == 0) {
            txt <- "The FDR you specified in cpg.annotate() returned no significant CpGs, hence there are no DMRs.\n    Try specifying a value of 'pcutoff' in dmrcate() and/or increasing 'fdr' in cpg.annotate()."
            stop(paste(strwrap(txt, exdent = 2), collapse = "\n"))
        }
        pcutoff <- sort(object$fdr)[nsig]
    }
    object$sig <- (object$fdr <= pcutoff)
    if (nrow(object) == 0) {
        txt <- "No signficant regions found. Try increasing the value of\n    'pcutoff' in dmrcate() and/or 'fdr' in cpg.annotate()."
        stop(paste(strwrap(txt, exdent = 2), collapse = "\n"))
    }
    message("Demarcating regions...")
    chr.N <- as.character(object$CHR)
    pos.N <- object$pos
    sig.N <- object$sig
    N <- length(sig.N)
    n.K <- which(sig.N)
    K <- length(n.K)
    stopifnot(K >= 2)
    pos.K <- pos.N[n.K]
    chr.K <- chr.N[n.K]
    jump_chr.k <- (chr.K[-1] != chr.K[-K])
    jump_pos.k <- (diff(pos.K) > lag)
    jump.k <- (jump_chr.k | jump_pos.k)
    ksegments.A2 <- Segment(jump.k)
    A <- nrow(ksegments.A2)
    kstart.A <- ksegments.A2[, "start"]
    kend.A <- ksegments.A2[, "end"]
    realpos.K <- pos.K
    if (consec) {
        realpos.N <- object$realcoordforconsec
        realpos.K <- realpos.N[n.K]
    }
    start.A <- realpos.K[kstart.A]
    end.A <- realpos.K[kend.A]
    chr.A <- chr.K[kstart.A]
    stopifnot(all(chr.K[kend.A] == chr.A))
    fmt <- "%s:%1d-%1d"
    coord.A <- sprintf(fmt, chr.A, start.A, end.A)
    nstart.A <- n.K[kstart.A]
    nend.A <- n.K[kend.A]
    width.A <- nend.A + 1 - nstart.A
    a.Z <- rep(seq(A), width.A)
    fn <- function(a) seq(from = nstart.A[a], to = nend.A[a])
    l.listA <- lapply(seq(A), fn)
    n.Z <- unlist(l.listA)
    region.N <- rep(NA_integer_, N)
    region.N[n.Z] <- a.Z
    levels <- seq(A)
    region.N <- factor(region.N, levels = levels)
    no_cpg.A <- c(table(region.N))
    REGIONSTAT <- function(field, fn) {
        x.N <- object[[field]]
        x.R <- tapply(x.N, region.N, fn)
        c(x.R)
    }
    fn_Stouffer <- function(x) pnorm(sum(qnorm(x))/sqrt(length(x)))
    fn_max <- function(x) x[which.max(abs(x))]
    results <- data.frame(coord = coord.A, no.cpgs = no_cpg.A, 
        minfdr = REGIONSTAT("fdr", min), Stouffer = REGIONSTAT("indfdr", 
            fn_Stouffer), maxbetafc = REGIONSTAT("betafc", fn_max), 
        meanbetafc = REGIONSTAT("betafc", mean), row.names = seq(A), 
        stringsAsFactors = FALSE)
    keep <- (results$no.cpgs >= min.cpgs)
    results <- results[keep, ]
    if (!is.null(betacutoff)) {
        keep <- (abs(results$meanbetafc) >= betacutoff)
        results <- results[keep, ]
    }
    o <- order(results$Stouffer, -results$no.cpgs)
    results <- results[o, ]
    message("Done!")
    output <- NULL
    output$input <- object
    output$results <- results
    output$cutoff <- pcutoff
    class(output) <- "dmrcate.output"
    output
}

dmrcoutput <- dmrcate(MA_DMRcate_list, lambda=1000, C=2, pcutoff = 0.01,min.cpgs = 2)
outcome_dmrcate <- dmrcoutput$results%>%
  write_rds("DMRCATE_USC_drop.wash_M_VAL_DS_EWAS_MA_output.rds")

### comb-p 
usc <- results %>%
  dplyr::mutate(chrom=chr, end=pos+1, start=pos, p_val=bacon_P)%>%
  dplyr::filter(grepl('cg', cpg)) %>%
  dplyr::select(chrom, start, end, p_val)%>%
  na.omit()%>%
  arrange(chrom, start)

usc%>%
  write_tsv("comb_p_USC_drop.wash_SESAME_M_VAL_ALL_DS_EWAS_MA.txt")

```

```{bash run comb-p in USC dataset}

cd ~/DSALL/usc
#run comb-p and set settings according to preferences
comb-p pipeline -c 4 --seed 0.05 --dist 1000\
  -p comb_p_USC_drop.wash_M_VAL_DS_EWAS_MA_output\
  --region-filter-p 0.01 --region-filter-n 2 comb_p_USC_drop.wash_SESAME_M_VAL_ALL_DS_EWAS_MA.txt

#setting --acf-dist to 0.33 * --dist == 330
#calculated stepsize as: 330
#ACF:
# #lag_min	lag_max	correlation	N	p
#1	331	0.07491	601610	0
#wrote: comb_p_USC_drop.wash_M_VAL_DS_EWAS_MA_output.acf.txt
## original lambda: 1.11
#wrote: comb_p_USC_drop.wash_M_VAL_DS_EWAS_MA_output.slk.bed.gz with lambda: 1.15
#wrote: comb_p_USC_drop.wash_M_VAL_DS_EWAS_MA_output.fdr.bed.gz
#wrote: comb_p_USC_drop.wash_M_VAL_DS_EWAS_MA_output.regions.bed.gz (174 regions)
## read 174 regions from comb_p_USC_drop.wash_M_VAL_DS_EWAS_MA_output.regions.bed.gz
## calculating ACF out to: 784
##           with 4  lags: [1, 331, 661, 991]
## Done with one-time ACF calculation
#628764 bases used as coverage for sidak correction
#wrote: comb_p_USC_drop.wash_M_VAL_DS_EWAS_MA_output.regions-p.bed.gz, (regions with corrected-p < 0.05: 59)
#wrote: comb_p_USC_drop.wash_M_VAL_DS_EWAS_MA_output.regions-t.bed, (regions with region-p < 0.010 and n-probes >= 2: 36)

cd ~/DSALL/usc
cat <(head -1 <(less comb_p_USC_drop.wash_M_VAL_DS_EWAS_MA_output.regions-p.bed.gz)) \
  <(tail -n +2 <(less comb_p_USC_drop.wash_M_VAL_DS_EWAS_MA_output.regions-p.bed.gz) | sort -k7,7g) \
  > comb_p_USC_drop.wash_M_VAL_DS_EWAS_MA_output.txt
```

```{r compare overlap of DMRcate and comb-p in USC datasset}

library(tidyverse)
library(AnnotationHub)
library(TxDb.Hsapiens.UCSC.hg19.knownGene)
library(GenomicRanges)
library(annotatr)
library(methyAnalysis)
setwd("~/DSALL/usc")

comb_p_output <- read_tsv("comb_p_USC_drop.wash_M_VAL_DS_EWAS_MA_output.txt")%>%
  dplyr::mutate(chrom=`#chrom`)%>%
  dplyr::select(-`#chrom`)%>%
  dplyr::select(chrom, everything())%>%
  dplyr::filter(z_sidak_p<0.01)
dmrcate_results <- read_rds("DMRCATE_USC_drop.wash_M_VAL_DS_EWAS_MA_output.rds")%>%
  dplyr::filter(minfdr<0.01)%>%
  separate(coord, c("chrom","coord"), sep =":")%>%
  separate(coord, c("start","end"), sep = "-")%>%
  dplyr::mutate(start=as.numeric(start),end=as.numeric(end))
comb_p_GR <- makeGRangesFromDataFrame(comb_p_output, keep.extra.columns=T,
                         ignore.strand=T,
                         seqinfo=NULL,
                         seqnames.field=c("chrom"),
                         start.field="start",
                         end.field=c("end"),
                         strand.field="strand",
                         starts.in.df.are.0based=FALSE)
DMRCate_GR <- makeGRangesFromDataFrame(dmrcate_results, keep.extra.columns=T,
                         ignore.strand=T,
                         seqinfo=NULL,
                         seqnames.field=c("chrom"),
                         start.field="start",
                         end.field=c("end"),
                         strand.field="strand",
                         starts.in.df.are.0based=FALSE)
overlap_DMRcate_comb_p_GR <- GenomicRanges::intersect(DMRCate_GR,comb_p_GR, ignore.strand=T)
length(overlap_DMRcate_comb_p_GR)
#31

if (require(TxDb.Hsapiens.UCSC.hg19.knownGene)) {
overlap_DMRcate_comb_p_DF <-annotateDMRInfo(overlap_DMRcate_comb_p_GR, 'TxDb.Hsapiens.UCSC.hg19.knownGene')
}

#left join with original files to get relevant meta data columns
overlap_DMRcate_comb_p_DF <- data.frame(overlap_DMRcate_comb_p_DF$sigDMRInfo)%>%
  left_join(comb_p_output, by=c("seqnames"="chrom","start","end"))%>%
  left_join(dmrcate_results,by=c("seqnames"="chrom","start","end"))
geneRanges <-
    function(db, column="ENTREZID"){
    g <- genes(db, columns=column)
    col <- mcols(g)[[column]]
    genes <- granges(g)[rep(seq_along(g), elementLengths(col))]
    mcols(genes)[[column]] <- as.character(unlist(col))
    genes
}
splitColumnByOverlap <-
    function(query, subject, column="ENTREZID", ...){
    olaps <- GenomicRanges::findOverlaps(query, subject, ...)
    f1 <- factor(subjectHits(olaps),
                 levels=seq_len(subjectLength(olaps)))
    splitAsList(mcols(query)[[column]][queryHits(olaps)], f1)
    }
annotate_grange <- function(grange_object, annotation_grange_object, column_name=NULL){
    file <- splitColumnByOverlap(annotation_grange_object,grange_object, column = column_name)
    grange_object$name <- file
    return(grange_object)}
annotate_grange <- function(grange_object, annotation_grange_object, column_name=NULL){
    file <- splitColumnByOverlap(annotation_grange_object,grange_object, column = column_name)
    grange_object$name <- file
    return(grange_object)}
overlap_DMRcate_comb_p_GR <- annotate_grange(overlap_DMRcate_comb_p_GR,DMRCate_GR, column_name="minfdr")
overlap_DMRcate_comb_p_GR$DMRCate_fdr <- overlap_DMRcate_comb_p_GR$name
overlap_DMRcate_comb_p_GR <- annotate_grange(overlap_DMRcate_comb_p_GR,DMRCate_GR, column_name="meanbetafc")
overlap_DMRcate_comb_p_GR$DMRCate_meanbetafc <- overlap_DMRcate_comb_p_GR$name
overlap_DMRcate_comb_p_GR <- annotate_grange(overlap_DMRcate_comb_p_GR,DMRCate_GR, column_name="no.cpgs")
overlap_DMRcate_comb_p_GR$DMRcate_N_cpgs <- overlap_DMRcate_comb_p_GR$name
overlap_DMRcate_comb_p_GR <- annotate_grange(overlap_DMRcate_comb_p_GR,comb_p_GR, column_name="z_sidak_p")
overlap_DMRcate_comb_p_GR$combp_sidak <- overlap_DMRcate_comb_p_GR$name
overlap_DMRcate_comb_p_GR <- annotate_grange(overlap_DMRcate_comb_p_GR,comb_p_GR, column_name="n_probes")
overlap_DMRcate_comb_p_GR$combp_N_cpgs <- overlap_DMRcate_comb_p_GR$name
overlap_DMRcate_comb_p_GR$name <- NULL
overlap_DMRcate_comb_p_GR$Gene_Symbol <- overlap_DMRcate_comb_p_DF$GeneSymbol
overlap_DMRcate_comb_p_GR$Promoter <- overlap_DMRcate_comb_p_DF$PROMOTER
overlap_DMRcate_comb_p_final <- overlap_DMRcate_comb_p_GR%>%
  as.data.frame()%>%
  dplyr::mutate(chrom=seqnames)%>%
  dplyr::select(-seqnames)%>%
  dplyr::select(chrom, start,end,width,Gene_Symbol,Promoter,everything())
colnames(overlap_DMRcate_comb_p_final) = paste0("usc_",colnames(overlap_DMRcate_comb_p_final))
write.csv(overlap_DMRcate_comb_p_final,"compDMARoverlap_comb_p_USC_drop.wash_SESAME_M_VAL_ALL.csv")

```

Check out CpGs within FLI1 and RUNX1 DMRs from NC paper

```{r Check out CpGs within FLI1 and RUNX1 DMRs from NC paper}

library(tidyverse)
library(data.table)
setwd("~/DSALL/usc")

fli1 <- read.csv("ds.all.usc.sesame.m.nowashington.ewas.anno.csv") %>%
  filter(chr=="chr11") %>%
  filter(pos>=128553855 & pos<=128557589) %>%
  mutate(dmr="fli1_promoter") 

# plot beta coefficients
fli1 %>%
  arrange(pos) %>%
  ggplot()+
  geom_point(aes(x=pos,y=Beta)) +
  geom_path(aes(x=pos,y=Beta)) +
  geom_bar(aes(x=pos,y=bacon_P), stat="identity", size=.1, color="black", alpha=.4) + 
  theme_classic() +
  labs(x="DMRs")+
  geom_hline(yintercept = 0,linetype="dotted") +
  scale_y_continuous(
    name="EWAS effect sizes",
    sec.axis = sec_axis(~.*1, name="p values"))+
  theme(axis.text.x = element_text(size=6,angle =30, hjust=1))
ggsave("fli1.ds.alll.png",width = 8,height = 4)
  

runx1 <- read.csv("ds.all.usc.sesame.m.nowashington.ewas.anno.csv") %>%
  filter(chr=="chr21") %>%
  mutate(dmr="none") %>%
  mutate(dmr = ifelse(pos>=36258497 & pos<=36259694, "runx1_1_promoter", 
               ifelse(pos>=36263287 & pos<=36263808, "runx1_2_enhancer", 
               ifelse(pos>=36421857 & pos<=36422800,"runx1_3_enhancer",
               ifelse( pos>=36689229 & pos<=36689434, "runx1_4_geneBody",
               ifelse( pos>=37049727 & pos<=37049788,"runx1_5_geneBody","none")))))) %>%
  filter(dmr!="none") %>%
  mutate(dmr=as.factor(dmr))%>%
  dplyr::select(cpg,pos,dmr,Beta,bacon_P) %>%
  arrange(pos)
dim(runx1)

# plot beta coefficients
runx1 %>%
  ggplot()+
  geom_point(aes(x=pos,y=Beta)) +
  geom_path(aes(x=pos,y=Beta)) +
  geom_bar(aes(x=pos,y=bacon_P), stat="identity", size=.1, color="black", alpha=.4) + 
  facet_wrap(~dmr, ncol=2,scales = "free") +
  theme_classic() +
  labs(x="DMRs")+
  geom_hline(yintercept = 0,linetype="dotted") +
  scale_y_continuous(
    name="EWAS effect sizes",
    sec.axis = sec_axis(~.*1, name="p values"))+
  theme(axis.text.x = element_text(size=6,angle =30, hjust=1))
ggsave("runx1.ds.alll.png",width = 8,height = 8)
  


 
```

----------------

Checking results EWAS M Value from BCM

```{r Checking results EWAS M Value from BCM}
# looking at the results

#lambda = [1] 
#lambda bacon = [1] 

setwd("~/DSALL/bcm")
library(tidyverse)
library(data.table)
library(bacon)
library(IlluminaHumanMethylationEPICanno.ilm10b4.hg19)
data("IlluminaHumanMethylationEPICanno.ilm10b4.hg19")
annoEPIC <- IlluminaHumanMethylationEPICanno.ilm10b4.hg19 %>%
  getAnnotation %>%
  as.data.frame %>%
  rownames_to_column(var="cpg") %>%
  dplyr::select(cpg,chr,pos,Relation_to_Island,UCSC_RefGene_Name,
                Probe_maf,UCSC_RefGene_Group,CpG_maf)

results <- fread("3_ds.all.bcm.ewas_m.txt") %>% 
  dplyr::rename("cpg"="V1") %>%
  mutate(Z=Beta/SE) %>%
  dplyr::filter(grepl('cg', cpg)) %>%
  filter(!is.na(P))%>% 
  left_join(annoEPIC,by="cpg") %>%
  arrange(by=P) %>% # results
  dplyr::filter(chr!="chrX"&chr!="chrY") %>%  #843393
  dplyr::filter(Probe_maf<0.05|is.na(Probe_maf)) %>% # 765704 
  dplyr::filter(CpG_maf<0.05|is.na(CpG_maf))  # 754405

results$bacon_P = pval(bacon(results$Z,na.exclude = TRUE))

results <- results %>%
  mutate(fdr=p.adjust(bacon_P, method = "fdr", n=length(bacon_P))) %>%
  mutate(bf=p.adjust(bacon_P, method = "bonferroni", n=length(bacon_P))) 

write.csv(results,"ds.all.bcm.sesame.m.ewas.anno.csv")

0.05/nrow(results) # bonferroni threshold
# 6.62774e-08


# raw P values
Nbeta <- sum(!is.na(results$P))
tmp <- results[!is.na(results$P),]
lambda = median(qchisq(1- tmp$P,1))/qchisq(0.5,1)
print(round(lambda, 2))
#1.27

# bacon P values
Nbeta <- sum(!is.na(results$bacon_P))
tmp <- results[!is.na(results$bacon_P),]
lambda = median(qchisq(1- tmp$bacon_P,1))/qchisq(0.5,1)
print(round(lambda, 2))
#1.05
  
# figures (qq plot and manhattan)
library(CMplot)
library(QCEWAS)
library(qqman)

bcm_model_anno <- results %>%
  dplyr::mutate(cpg_name=cpg,estimate=Beta, std_error=SE,p_val=bacon_P)

sig_thresghold = bcm_model_anno %>%
  filter(fdr<0.05)%>%
  arrange(desc(bacon_P))
sig_threshold = sig_thresghold[1,"p_val"]

annotation = bcm_model_anno%>%
  filter(bf<0.05) %>%
  filter(UCSC_RefGene_Name!="") %>%
  mutate(gene.highlight = sub(";.*$","",UCSC_RefGene_Name))
cpg.highlight = annotation$cpg_name
genes.highlight = annotation$gene.highlight

outcomes_model_CM_plot <- bcm_model_anno%>%
  dplyr::mutate(p_val_cmplot = -log10(p_val))%>%
  dplyr::mutate(p_val_cmplot= ifelse(estimate<0,-p_val_cmplot, p_val_cmplot )) %>%
  dplyr::select(cpg_name, chr,pos, p_val_cmplot)
outcomes_model_CM_plot$chr <- gsub("chr","",outcomes_model_CM_plot$chr)

CMplot(outcomes_model_CM_plot,
       plot.type="m",
       band=0, 
       cex=c(0.5,0.5,0.5),
       amplify=FALSE,
       LOG10=FALSE,  
       signal.col=NULL,
       chr.den.col=NULL,
       threshold=c(log10(sig_threshold[[1]]), -log10(sig_threshold[[1]])),
       threshold.lty=2, 
       threshold.lwd=1, 
       threshold.col="black", 
       highlight=cpg.highlight,
       highlight.text=genes.highlight, 
       file="jpg",
       memo="bcm.sesame.m",
       #main="USC DS-ALL EWAS (SESAME)",
       dpi=500,file.output=TRUE,verbose=TRUE)

outcomes_model_qq <- usc_model_anno %>% 
  dplyr::select(cpg_name,p_val)%>%
  dplyr::mutate(ID=cpg_name, P=p_val)
lambda <- P_lambda(outcomes_model_qq$P)

#png("usc.sesame.qq.m.png")
tiff("ucb.sesame.qq.m.tiff", width = 10, height = 7, units = 'in', res = 300)
qq(outcomes_model_qq$P)
#   main=paste0("USC dataset \n lambda value= ",
#               round(lambda,3)))
dev.off()

## top CpGs
head(results)
dim(results%>%filter(fdr<0.05))
# 1024
dim(results%>%filter(bf<0.05))
# 73

results%>%filter(fdr<0.05)

```

Overlap of top hits and DMR in usc and bcm

```{r Overlap of top hits and DMR in usc and bcm}

setwd("~/DSALL")

library(tidyverse)
library(data.table)

# beta values
betas3_0 = fread("~/sets/set3/set3_Sesame_beta.csv") %>%
  as.data.frame()%>%
  dplyr::select(-V1)
betas3_1 = betas3_0 %>%
  column_to_rownames(var="probeId")

covs = read_rds("usc/covs.postReviewQCs.dropWashU.rds")
covs.controls = covs %>% filter(CaCo==0)
covs.cases = covs %>% filter(CaCo==1)

results.bcm <- read.csv("bcm/ds.all.bcm.sesame.m.ewas.anno.csv") %>%
  as.data.frame() %>%
  dplyr::rename("P replication"="P",
                "bacon P replication"="bacon_P",
                "effect size replication"="Beta",
                "fdr replication"="fdr",
                "bf replication"="bf") %>%
  dplyr::select(cpg,`effect size replication`,
                `P replication`,
                `fdr replication`,
                `bf replication`,
                `bacon P replication`)

results.usc <- read.csv("usc/ds.all.usc.sesame.m.nowashington.ewas.anno.csv") %>%
  as.data.frame() %>%
  dplyr::select(-SE,-Probe_maf) %>%
  mutate(chr = gsub("chr","",chr))%>%
  dplyr::select(cpg,chr,pos,Relation_to_Island,
                UCSC_RefGene_Name,UCSC_RefGene_Group,
                everything()) %>%
  dplyr::rename("P discovery"="P",
                "bacon P discovery" = "bacon_P",
                "effect size discovery"="Beta",
                "fdr discovery"="fdr",
                "bf discovery"="bf") %>%
  arrange(by=`P discovery`) %>%
  filter(`fdr discovery`<0.05) %>%
  left_join(results.bcm,by="cpg") %>%
  dplyr::select(cpg,chr,pos,UCSC_RefGene_Name,
                Relation_to_Island,
                Regulatory_Feature_Group,
                UCSC_RefGene_Group,
                `effect size discovery`,`bacon P discovery`,
                `fdr discovery`,`bf discovery`,
                `effect size replication`,
                `bacon P replication`, `fdr replication`,
                `bf replication`)

betas_candidate = data.frame(
  controlMean=betas3_1[results.usc$cpg,rownames(covs.controls)] %>%
    rowMeans(na.rm=TRUE),
  caseMean=betas3_1[results.usc$cpg,rownames(covs.cases)] %>% 
    rowMeans(na.rm=TRUE) ) %>%
  mutate(`beta Difference` = caseMean-controlMean) %>%
  rownames_to_column(var="cpg")
  
topDMPs = results.usc %>%
  left_join(betas_candidate, by = "cpg")
  
write.csv(topDMPs,"tableS5.topHits.m.csv")


```

Finding individual CpGs in BCM in USC DMR results 

```{r extract raw beta values for CpGs in USC DMRs}

setwd("~/DSALL/usc")
library(tidyverse)
library(data.table)
library(GenomicRanges)

uscDMR = read.csv("compDMARoverlap_comb_p_USC_drop.wash_SESAME_M_VAL_ALL.csv", row.names=1) %>%
  mutate(chrom=usc_chrom,start=usc_start,end=usc_end)%>%
  dplyr::select(-usc_chrom, -usc_start, -usc_end)

usc = read.csv("ds.all.usc.sesame.m.nowashington.ewas.anno.csv")  %>%
  mutate(cpg_name=cpg,estimate=Beta,chrom=chr,p_val=P,start=pos,end=pos)%>%
  dplyr::select(chrom,start,end,cpg_name,estimate,p_val)%>%
  na.omit()

uscDMR_GR <- makeGRangesFromDataFrame(uscDMR, keep.extra.columns=T,
                         ignore.strand=T,
                         seqinfo=NULL,
                         seqnames.field=c("chrom"),
                         start.field="start",
                         end.field=c("end"),
                         strand.field="strand",
                         starts.in.df.are.0based=FALSE)
uscRes_GR <- makeGRangesFromDataFrame(usc, keep.extra.columns=T,
                         ignore.strand=T,
                         seqinfo=NULL,
                         seqnames.field=c("chrom"),
                         start.field="start",
                         end.field=c("end"),
                         strand.field="strand",
                         starts.in.df.are.0based=FALSE)
overlap_uscDMR_uscRes <- GenomicRanges::findOverlaps(uscRes_GR,uscDMR_GR, ignore.strand=T)

a=uscDMR[subjectHits(overlap_uscDMR_uscRes),]
b=usc[queryHits(overlap_uscDMR_uscRes),] %>%
  dplyr::rename(estimate_usc=estimate, p_val_usc=p_val,cpg.chrom=chrom,cpg.start=start,cpg.end=end)
com = cbind(b,a) %>% as.data.frame() %>%
  arrange(chrom,start)

outcomes3 <- fread("ds.all.covaraites.postReviewQCs.txt") %>% 
  as.data.frame() %>%
  dplyr::mutate(ID=beadPosition,all_case=CaCo) %>%
  dplyr::select(ID,all_case)

betas3_0 = fread("~/sets/set3/set3_Sesame_beta.csv") %>%
  as.data.frame()%>%
  dplyr::select(-V1)

usc_raw = betas3_0 %>%
  column_to_rownames(var="probeId") %>%
  t %>%
  as.data.frame()

usc_raw_1 = usc_raw[,com$cpg_name] %>%
  rownames_to_column(var="ID")%>%
  inner_join(outcomes3,by="ID")
  
write.table(com,"ds.all.cpg.pos.m.val.txt",sep = "\t", row.names = FALSE, quote = FALSE)
write.table(usc_raw_1,"ds.all.raw.beta.file.m.val.txt",sep = "\t", row.names = FALSE, quote = FALSE)

```

```{r looking up CpGs in BCM EWAS results from USC DMR results}

setwd("~/DSALL")
library(tidyverse)
library(data.table)
library(GenomicRanges)

uscDMR = read.csv("usc/compDMARoverlap_comb_p_USC_drop.wash_SESAME_M_VAL_ALL.csv", row.names=1) %>%
  mutate(chrom=usc_chrom,start=usc_start,end=usc_end,width=usc_width,promoter=usc_Promoter)%>%
  dplyr::select(-usc_chrom, -usc_start, -usc_end,-usc_width,-usc_Promoter)

bcmRes = fread("bcm/ds.all.bcm.sesame.m.ewas.anno.csv") %>%
  mutate(chrom=chr,start=pos,end=pos,cpg_name=cpg,estimate=Beta,p_val=bacon_P)%>%
  dplyr::select(chrom,start,end,cpg_name,estimate,p_val)%>%
  na.omit()
  
uscDMR_GR <- makeGRangesFromDataFrame(uscDMR, keep.extra.columns=T,
                         ignore.strand=T,
                         seqinfo=NULL,
                         seqnames.field=c("chrom"),
                         start.field="start",
                         end.field=c("end"),
                         strand.field="strand",
                         starts.in.df.are.0based=FALSE)
bcmRes_GR <- makeGRangesFromDataFrame(bcmRes, keep.extra.columns=T,
                         ignore.strand=T,
                         seqinfo=NULL,
                         seqnames.field=c("chrom"),
                         start.field="start",
                         end.field=c("end"),
                         strand.field="strand",
                         starts.in.df.are.0based=FALSE)
overlap_uscDMR_bcmRes <- GenomicRanges::findOverlaps(bcmRes_GR,uscDMR_GR, ignore.strand=T)

a=uscDMR[subjectHits(overlap_uscDMR_bcmRes),]
b=bcmRes[queryHits(overlap_uscDMR_bcmRes),] %>%
  dplyr::select(-chrom,-start,-end)%>%
  dplyr::rename(estimate_bcm=estimate, p_val_bcm=p_val)
com = cbind(b,a) %>% as.data.frame() %>%
  arrange(chrom,start)

table(com$usc_Gene_Symbol)

write.csv(com,"CpGswithUSCdmr.m.val.csv")
write.csv(com$cpg_name,"bcm.raw.m.val.csv")

## Cacculating consistency
com.cpg = com %>% dplyr::select(cpg_name)

consisDF = list()
j=1

for(i in unique(com$usc_Gene_Symbol)){
  
  bcm1.avails = com %>%
    filter(usc_Gene_Symbol==i) 
  
  bcm1.sigs = bcm1.avails %>%
    filter(p_val_bcm < 0.05)
  
  sigs = dim(bcm1.sigs)[1]
  sig.same = sum(bcm1.sigs$estimate_bcm*bcm1.sigs$usc_DMRCate_meanbetafc>0)

  consisDF[[j]] = c(i,sigs,sig.same)
    
  j=j+1
}
  
consisDF.ls = do.call(rbind,consisDF) %>%
  as.data.frame() %>%
  arrange(V1)
colnames(consisDF.ls) = c("usc_Gene_Symbol","sig.CpGs.bcm","sig.CpGs.same.bcm")  

uscDMR.lookup = uscDMR %>%
  inner_join(consisDF.ls,by="usc_Gene_Symbol") %>%
  mutate(chrom=gsub("chr","",chrom),
         Gene_Symbol=usc_Gene_Symbol) %>%
  dplyr::select(chrom,start,end,width,Gene_Symbol,promoter,
                usc_DMRcate_N_cpgs,usc_DMRCate_meanbetafc,usc_DMRCate_fdr,
                usc_combp_N_cpgs,usc_combp_sidak,
                sig.CpGs.bcm, sig.CpGs.same.bcm)

sum(uscDMR.lookup$sig.CpGs.same.bcm!=0)

write.csv(uscDMR.lookup,"uscDMR.lookup.inBCM.m.val.csv")


```

----------------

>  table 1: descriptives

```{r table 1: descriptives}

### USC descriptive
setwd("~/DSALL/usc")
library(tidyverse)
library(data.table)


########################################################################
# DS ALL and control

#subsInTheStudy <- fread("ds.all.final.subs.txt")
subsInTheStudy <- fread("ds.all.final.subs.AfterReview.dropWashU.txt")


clinical3 = fread("~/sets/set3/clinical_variables.csv") %>%
  filter(Trisomy21==1) %>%
  mutate(gestage=Y_gestation_week,birthWt=Y_birthweight) %>%
  dplyr::select(beadPosition,race,collectionAge,gestage,birthWt,sex,CaCo,AGE,Sample_Source) %>%
  mutate(collectionAge=collectionAge/24) %>%
  inner_join(subsInTheStudy,by="beadPosition")

summary(clinical3$AGE)

table(clinical3 %>% filter(CaCo==1) %>% dplyr::select(Sample_Source))

print(nrow(clinical3)==nrow(subsInTheStudy)) # it has to be true

clinical3[which(clinical3$gestage==99),"gestage"] <- NA

table(clinical3$CaCo)

table(clinical3$sex,clinical3$CaCo, useNA = "ifany")
a = table(clinical3$sex,clinical3$CaCo)
fisher.test(x=a, alternative = "two.sided")

table(clinical3$race,clinical3$CaCo, useNA = "ifany")
a = table(clinical3$race,clinical3$CaCo)
fisher.test(x=a, alternative = "two.sided")

a1=clinical3 %>% filter(CaCo==0) %>% dplyr::select(collectionAge) 
a1 %>% summary
sd(a1$collectionAge,na.rm = TRUE)
a2 = clinical3 %>% filter(CaCo==1) %>% dplyr::select(collectionAge)  
a2 %>% summary
sd(a2$collectionAge,na.rm = TRUE)
lm(collectionAge~CaCo,data=clinical3) %>% summary

###
a1=clinical3 %>% filter(CaCo==0) %>% dplyr::select(gestage) 
a1 %>% summary
sd(a1$gestage,na.rm = TRUE)

a2 = clinical3 %>% filter(CaCo==1) %>% dplyr::select(gestage)  
a2 %>% summary
sd(a2$gestage,na.rm = TRUE)
lm(gestage~CaCo,data=clinical3) %>% summary

b = clinical3 %>% 
  mutate(preterm=gestage<37) %>%
  dplyr::select(CaCo,preterm) %>% 
  na.omit 
table(b$preterm,b$CaCo)
fisher.test(x=table(b$preterm,b$CaCo), alternative = "two.sided")

###
a1=clinical3 %>% filter(CaCo==0) %>% dplyr::select(birthWt) 
a1 %>% summary
sd(a1$birthWt,na.rm = TRUE)

a2 = clinical3 %>% filter(CaCo==1) %>% dplyr::select(birthWt)  
a2 %>% summary
sd(a2$birthWt,na.rm = TRUE)
lm(birthWt~CaCo,data=clinical3) %>% summary

```

>  supplement tables

```{r}

setwd("~/DSALL/usc")
library(tidyverse)
library(data.table)


ages = fread("~/sets/set3/clinical_variables.csv") %>%
  as.data.frame() %>%
  dplyr::rename("age_at_diagnosis" = "AGE") %>%
  dplyr::select(beadPosition,age_at_diagnosis)

snps = read.csv('DS_GWAS_ALL_SNPs_14Jul21.csv') %>%
  mutate(subjectId=as.character(StudyID))%>%
  dplyr::select(subjectId, rs7089424_ARID5B,	
                rs11978267_IKZF1, rs3731249_CDKN2A, rs3824662_GATA3) 

GATA1_table = read_tsv("DS_controls_GATA1_results.txt")%>%
  mutate(subjectId=as.character(subjectId)) %>%
  as.data.frame()

clinicls = fread("~/sets/set3/clinical_variables.csv") %>%
  as.data.frame() %>%
  dplyr::rename("age_at_diagnosis" = "AGE") %>%
  dplyr::select(beadPosition,Ethnicity,Y_race,age_at_diagnosis,,Y_gestation_day,
                collectionAge, Y_birthweight)

otherVar = read_rds("covs.postReviewQCs.dropWashU.rds") %>%
  rownames_to_column(var="beadPosition")

gran = fread("~/sets/set3/idol_deconvolution.csv") %>%
  as.data.frame() %>%
  dplyr::select(beadPosition,Gran)

# task 1

supp = readxl::read_excel("04122022  Li_et_al_SupplementalDataset.xlsx") %>%
  as.data.frame() %>%
  mutate(subjectId=as.character(subjectId)) %>%
  dplyr::select(-`age-at-diagnosis`,-ARID5B_rs7089424,
                -IKZF1_rs11978267,-CDKN2A_rs3731249,-GATA3_rs3824662)

supp1 = supp %>%
  left_join(ages,by="beadPosition") %>%
  left_join(snps,by="subjectId")

write.csv(supp1,"04122022_Li_et_al_SupplementalDataset.csv")

# task 2

fill = data.frame(
  beadPosition = c("202172220181_R05C01","202148010158_R04C01","202163110003_R02C01"),
  subjectId = as.character(c("198110","198479","198161"))
) %>%
  left_join(snps,by="subjectId") %>%
  left_join(GATA1_table,by=c("beadPosition","subjectId")) %>%
  left_join(clinicls,by="beadPosition") %>%
  left_join(otherVar,by="beadPosition") %>%
  left_join(gran,by="beadPosition") %>%
  dplyr::select(beadPosition,subjectId,CaCo,sex,Ethnicity,plate,Y_gestation_day,
                collectionAge,Y_birthweight,Mutation,`Mutation VAF`,CD8T,CD4T,NK,
                Bcell,Mono,Gran,nRBC,	epi1,	epi2,	epi3,	epi4,	epi5,	epi6,	
                epi7,	epi8,	epi9,	epi10,age_at_diagnosis,
                rs7089424_ARID5B,rs11978267_IKZF1,rs3731249_CDKN2A,rs3824662_GATA3)
write.csv(fill,"fill.supp.table.csv")
  


```








> a heatmap showing clustering of DS subjects, perhaps also along with non-DS subjects, to show (probably) that we cannot differentiate DS-ALL cases from DS controls using genome-wide DNA methylation data

```{r}

setwd("~/DSALL/usc")
library(tidyverse)
library(data.table)
library(DescTools)
library(ComplexHeatmap)
library(circlize)

library(IlluminaHumanMethylationEPICanno.ilm10b4.hg19)
data("IlluminaHumanMethylationEPICanno.ilm10b4.hg19")
anno       <- IlluminaHumanMethylationEPICanno.ilm10b4.hg19 %>%
              getAnnotation %>%
              as.data.frame %>%
              rownames_to_column(var="cpg")

cellP = fread("~/sets/set3/idol_deconvolution.csv")%>%
  as.data.frame()%>%
  dplyr::select(-V1,-subjectId)

# ds and ds_all samples beta values and variables
subsInTheStudy <- fread("ds.all.final.subs.AfterReview.dropWashU.txt")
outcomes3 = fread("~/sets/set3/clinical_variables.csv") %>%
  inner_join(subsInTheStudy,by=c("beadPosition","subjectId"))%>%
  mutate(ID=beadPosition,Blindid=subjectId,
         female=sex-1,down_syndrome=Trisomy21,
         all_case=CaCo) %>%
  left_join(cellP,by="beadPosition")%>%
  dplyr::select(ID,Blindid,plate,female,Ethnicity,down_syndrome,all_case,
                CD8T,CD4T,NK,Bcell,Mono,Gran,nRBC) %>%
    mutate(`CD8 T-cell`=CD8T,
         `CD4 T-cell`=CD4T,
         `B-cell`=Bcell,
         Monocyte=Mono,
         Granulocyte=Gran)
dim(outcomes3)
#324  19
#write_rds(outcomes3,"covariates_down_ALL_08_08.rds")

# library(impute)
# betas3_0 = fread("~/sets/set3/set3_Sesame_beta.csv") %>%
#   as.data.frame()%>%
#   dplyr::select(-V1)
# rownames(betas3_0) <- c()
# betas3_0 <- betas3_0 %>%
#   column_to_rownames(var="probeId")%>%
#   t%>%
#   as.data.frame()
# betas3_0 <- betas3_0[,colMeans(is.na(betas3_0)) < 0.05] %>% as.matrix()
# betas3_0_imp <- impute.knn(betas3_0)
#write_rds(betas3_0_imp,"sesame.beta.s3.rds")
betas3_0_imp <- read_rds("sesame.beta.s3.rds")
betas3_0 <- betas3_0_imp$data

length(intersect(rownames(betas3_0),outcomes3$ID))
# [1] 324

betas3 = betas3_0[outcomes3$ID,]

dim(betas3)
# [1] 324 722664

# Heatmap of DS-ALL and DS subjects (top 2000, excluding chrX, chrY)
## exclude CpGs on chrX and Y
betas = betas3
outcomes = outcomes3
xyprobbes = anno %>% filter(Probe_maf<0.05|is.na(Probe_maf), chr!="chrX"&chr!="chrY") %>% dplyr::select(cpg)

betas_1 = betas[,colnames(betas)%in%xyprobbes$cpg] %>% 
  t() %>%
  as.data.frame()

## calculate MAD3 for all CpGs
MAD <- apply(betas_1, 1, MeanAD)
# write_rds(MAD,'MAD.rds')
# MAD = read_rds('MAD.rds')
betas_1$MAD = MAD
betas_2 <- betas_1 %>%
  arrange(desc(MAD)) %>%
  dplyr::select(-MAD)
betas_3 <- betas_2[1:2000,] %>% as.matrix()

annot_df <- outcomes%>%
  mutate(`ALL-status`=factor(ifelse(all_case==1,"ALL","Non-ALL"),
                             levels=c("Non-ALL","ALL")),
          Sex = ifelse(female==1,"Female","Male")) %>%
  dplyr::select(Sex,`ALL-status`, `B-cell`, 
                `CD4 T-cell`, `CD8 T-cell`,
                Granulocyte, Monocyte, NK, nRBC)%>%
  as.data.frame()

col = list(`ALL-status` = c("Non-ALL" = "#f8766d", "ALL" = "#00bfc4"),
           Sex = c("Male" = "royalblue2", "Female" =  "plum1"),
            `B-cell`= colorRamp2(c(0, 0.15), c("white", "royalblue4")),
            `CD4 T-cell`= colorRamp2(c(0, 0.4), c("white", "salmon4")),
            `CD8 T-cell`= colorRamp2(c(0, 0.2), c("white", "wheat4")),
            Granulocyte= colorRamp2(c(0, 0.8), c("white", "hotpink4")),
            Monocyte= colorRamp2(c(0, 0.2), c("white", "olivedrab4")),
            NK= colorRamp2(c(0, 0.15), c("white", "yellow2")),
           nRBC= colorRamp2(c(0, 1), c("white", "red")))
ha <- HeatmapAnnotation(df=annot_df, col = col)

# a heatmap showing ds-all and ds
set.seed(11)
hm <- Heatmap(betas_3, name = "beta values",
        top_annotation = ha, 
        show_column_names = FALSE,
        show_row_names = FALSE, 
        clustering_method_rows="ward.D2", 
        clustering_method_columns="ward.D2", 
        clustering_distance_rows="pearson",
        clustering_distance_columns="pearson",
        column_title = "Heatmap with top 2000 most variable CpGs",
        column_title_gp = gpar(fontsize = 18, fontface = "bold"))

tiff(filename = "hm_ds_all.tiff", 
     res=300, units="cm", width = 20, height = 37)
draw(hm, heatmap_legend_side = "right", 
     annotation_legend_side = "right",
     legend_title_gp=gpar(fontsize = 15))
dev.off()


# a heatmap showing ds-all and ds, sorting by B cell proportions
set.seed(11)
hm_bcell_order <- Heatmap(betas_3, name = "beta values",
        top_annotation = ha, 
        show_column_names = FALSE,
        show_row_names = FALSE, 
        clustering_method_rows="ward.D2", 
        column_order = order(outcomes3$Bcell),
        #clustering_method_columns="ward.D2", 
        clustering_distance_rows="pearson",
        #clustering_distance_columns="pearson",
        column_title = "Heatmap with top 2000 most variable CpGs
        ordered by B cell proportions",
        column_title_gp = gpar(fontsize = 18, fontface = "bold"))

tiff(filename = "hm_ds_all_bcell.tiff", 
     res=300, units="cm", width = 20, height = 37)
draw(hm_bcell_order, heatmap_legend_side = "right", 
     annotation_legend_side = "right",
     legend_title_gp=gpar(fontsize = 15))
dev.off()

```

> DMR figures using coMET is in my local code file

##### B cell proportion analysis

>  B cell proportion comparisons in DS-ALL and DS-controls

```{r boxplot for cell proportions}

########################################################################
# plot usc results

setwd("~/DSALL/usc")
library(tidyverse)
library(data.table)

subs = fread("ds.all.final.subs.AfterReview.dropWashU.txt")
cellp <- fread("ds.all.covaraites.txt") %>%
  inner_join(subs,by="beadPosition")


bcell.cases = cellp %>% filter(CaCo==1) %>% dplyr::select(Bcell) 
mean(bcell.cases$Bcell)
# [1] 0.01285246
sd(bcell.cases$Bcell)
# [1] 0.01535381
median(bcell.cases$Bcell)
# [1] 0.007624641

bcell.controls = cellp %>% filter(CaCo==0) %>% dplyr::select(Bcell) 
mean(bcell.controls$Bcell)
# [1] 0.008261422
sd(bcell.controls$Bcell)
# [1] 0.01149786
median(bcell.controls$Bcell)
# [1] 0.003441876

cellp %>%
  dplyr::select(beadPosition,CaCo,CD8T,CD4T,NK,Bcell,Mono,Gran,nRBC)%>%
  reshape::melt(id=c("beadPosition","CaCo"),
                variable = "cellType")%>%
  mutate(cellProportion = value) %>%
  ggplot(aes(x=cellType, y=cellProportion, fill=as.factor(CaCo))) +
  geom_boxplot(outlier.size=0.2)+
  geom_point(colour="darkblue",
             position=position_jitterdodge(),
             alpha=0.5,size=0.01) +
  theme_classic() +
#ggtitle("Cell proportions in Down Syndrome subjects\nDiscovery Study")+
  xlab("Cell Types") +
  ylab("Proportions") +
  theme(plot.title = element_text(hjust = 0.5),
        axis.text.x = element_text(size=12),
        axis.text.y = element_text(size=12),
        axis.title.x = element_text(size=14),
        axis.title.y = element_text(size=14))+
  scale_fill_discrete(name="Disease Status",labels=c("DS","DS-ALL"))
ggsave("Cell_proportions_all_ds.png",width = 8,height = 6)

# boxplot for B cell only
cellp %>%
  dplyr::select(beadPosition,CaCo,Bcell)%>%
  reshape::melt(id=c("beadPosition","CaCo"),
                variable = "cellType")%>%
  mutate(cellProportion = value) %>%
  ggplot(aes(x=cellType, y=cellProportion, fill=as.factor(CaCo))) +
  geom_boxplot(outlier.size=0.2)+
  geom_point(colour="darkblue",
             position=position_jitterdodge(),
             alpha=0.7,size=0.5) +
  theme_classic() +
#ggtitle("B Cell proportions in Down Syndrome subjects\nDiscovery Study")+
  xlab("Case Control Status") +
  ylab("B cell Proportions") +
  theme(plot.title = element_text(hjust = 0.5),
        axis.text.x = element_text(size=12),
        axis.text.y = element_text(size=12),
        axis.title.x = element_text(size=14),
        axis.title.y = element_text(size=14))+
  scale_fill_discrete(name="Disease Status",labels=c("DS","DS-ALL"))
ggsave("B_proportions_all_ds_usc.png",width = 8,height = 6)

########################################################################
# plot bcm results

setwd("~/DSALL/bcm")
library(tidyverse)
library(data.table)

cellp <- fread("11_cellsproportion_idoldeconvolution.csv") %>%
  mutate(CaCo=phenotype)

cellp %>%
  dplyr::select(CaCo,CD8T,CD4T,NK,Bcell,Mono,Gran,nRBC)%>%
  reshape::melt(id=c("CaCo"),
                variable = "cellType")%>%
  mutate(cellProportion = value) %>%
  ggplot(aes(x=cellType, y=cellProportion, fill=as.factor(CaCo))) +
  geom_boxplot(outlier.size=0.2)+
  geom_point(colour="darkblue",
             position=position_jitterdodge(),
             alpha=0.5,size=0.01) +
  theme_classic() +
#ggtitle("Cell proportions in Down Syndrome subjects\nReplication Study") +
  xlab("Cell Types") +
  ylab("Proportions") +
  ylim(NA,1)+
  theme(plot.title = element_text(hjust = 0.5),
        axis.text.x = element_text(size=12),
        axis.text.y = element_text(size=12),
        axis.title.x = element_text(size=14),
        axis.title.y = element_text(size=14))+
  scale_fill_discrete(name="Disease Status",labels=c("DS","DS-ALL"))
ggsave("Cell_proportions_all_ds_bcm.png",width = 8,height = 6)

# boxplot for B cell only
cellp %>%
  dplyr::select(CaCo,Bcell)%>%
  reshape::melt(id=c("CaCo"),
                variable = "cellType")%>%
  mutate(cellProportion = value) %>%
  ggplot(aes(x=cellType, y=cellProportion, fill=as.factor(CaCo))) +
  geom_boxplot(outlier.size=0.2)+
  geom_point(colour="darkblue",
             position=position_jitterdodge(),
             alpha=0.7,size=0.5) +
  theme_classic() +
#ggtitle("B Cell proportions in Down Syndrome subjects\nReplication Study")+
  xlab("Case Control Status") +
  ylab("B cell Proportions") +
  theme(plot.title = element_text(hjust = 0.5),
        axis.text.x = element_text(size=12),
        axis.text.y = element_text(size=12),
        axis.title.x = element_text(size=14),
        axis.title.y = element_text(size=14))+
  scale_fill_discrete(name="Disease Status",labels=c("DS","DS-ALL"))
ggsave("B_proportions_all_ds_bcm.png",width = 8,height = 6)

```

proportion difference between cases and controls
```{r}

setwd("~/DSALL")
library(tidyverse)
library(data.table)

subs = fread("~/DSALL/usc/ds.all.final.subs.AfterReview.dropWashU.txt")
cellp.usc <- fread("~/DSALL/usc/ds.all.covaraites.txt") %>%
  inner_join(subs,by="beadPosition") %>%
  dplyr::select(CaCo,CD8T,CD4T,NK,Bcell,Mono,Gran,nRBC) %>%
  mutate(bcellTotalLym =Bcell / (Bcell+CD8T+CD4T) ) %>%
  group_by(CaCo)%>%
  summarise(CD8T_mean = mean(CD8T),
            CD4T_mean = mean(CD4T),
            NK_mean = mean(NK),
            Bcell_mean = mean(Bcell),
            Mono_mean = mean(Mono),
            Gran_mean = mean(Gran),
            nRBC_mean = mean(nRBC),
            bcellTotalLym=mean(bcellTotalLym)) %>%
  mutate(cohort="usc")

cellp.bcm <- fread("~/DSALL/bcm/11_cellsproportion_idoldeconvolution.csv") %>%
  mutate(CaCo=phenotype)%>%
  dplyr::select(CaCo,CD8T,CD4T,NK,Bcell,Mono,Gran,nRBC) %>%
  mutate(bcellTotalLym =Bcell / (Bcell+CD8T+CD4T) ) %>%
  group_by(CaCo)%>%
  summarise(CD8T_mean = mean(CD8T),
            CD4T_mean = mean(CD4T),
            NK_mean = mean(NK),
            Bcell_mean = mean(Bcell),
            Mono_mean = mean(Mono),
            Gran_mean = mean(Gran),
            nRBC_mean = mean(nRBC),
            bcellTotalLym=mean(bcellTotalLym)) %>%
  mutate(cohort="ucb")

cellp = rbind(cellp.usc,cellp.bcm) %>% as.data.frame()
write.csv(cellp,"cellProportions.additional.info.csv")
```

proportion of B cell in all lymphocytes

```{r cell proportions controlling for epi structures}

library(broom)
library(tidyverse)
library(data.table)

setwd("~/DSALL/usc")

#subs_cellproportion_1 <- fread('ds.all.covaraites.txt') %>% 
#subs_cellproportion_1 <- fread('ds.all.covaraites.postReviewQCs.txt') %>% 
subjects = fread("ds.all.final.subs.AfterReview.dropWashU.txt")
subCP <- fread("ds.all.covaraites.txt") %>% 
  as.data.frame() %>%
  filter(beadPosition%in%subjects$beadPosition)

cellP <- data.frame()
j = 1

for(i in  c("CD8T","CD4T","NK","Bcell","Mono","Gran","nRBC")){
  
  lm1 <- lm(subCP[,i] ~ subCP$CaCo + subCP$sex + subCP$plate + subCP$epi1 + 
              subCP$epi2 + subCP$epi3 + subCP$epi4 + subCP$epi5 + subCP$epi6+ 
              subCP$epi7+ subCP$epi8+ subCP$epi9+ subCP$epi10)%>%
    tidy() %>% as.data.frame()

  cellP[j , 1] <- i
  cellP[j , 2] <- lm1[2,2]
  cellP[j , 3] <- lm1[2,3]
  cellP[j , 4] <- lm1[2,4]
  cellP[j , 5] <- lm1[2,5]
  j = j+1
 
  }

colnames(cellP) <- c("cell_name","estimate_10epi","stderror_10epi","stat_10epi","pvalue_10epi")
write.csv(cellP,"usc_B_reg_lin_epi10.csv")

colnames(cellP) <- c("cellType","beta","se","stat_10epi","pvalue")
write.table(cellP,"usc_cp_10pc.txt",
            row.names = FALSE,
            quote=FALSE,
            sep='\t')
```

```{r stratified by age!}
######################################################
library(broom)
library(tidyverse)
library(metafor)
library(data.table)
setwd("~/DSALL/usc")

subjects = fread("ds.all.final.subs.AfterReview.dropWashU.txt")

subs_cellproportion_1 <- fread('ds.all.covaraites.txt') %>%
  as.data.frame() %>%
  filter(beadPosition%in%subjects$beadPosition)
  
vars = fread("~/sets/set3/clinical_variables.csv") %>%
  as.data.frame() %>%
  dplyr::select(beadPosition,AGE)%>%
  inner_join(subs_cellproportion_1,by="beadPosition")%>%
  filter(!is.na(AGE) | CaCo==0)

vars %>%
  arrange(AGE) %>%
  head()

missing_age = fread("~/sets/set3/clinical_variables.csv")  %>%
  as.data.frame() %>%
  inner_join(subjects, by=c("beadPosition","subjectId")) %>%
  filter(CaCo==1) %>%
  #filter(is.na(AGE)) %>%
  dplyr::select(subjectId,status,Sample_Source,AGE)

i = "Bcell"

smokingDF = vars %>%
  filter( (AGE<=4) | CaCo==0)
table(smokingDF$CaCo)
#   0   1 
# 198  64 

lm1 <- lm(smokingDF[,i] ~ smokingDF$CaCo + smokingDF$sex + smokingDF$plate +
             smokingDF$epi1 + smokingDF$epi2 + smokingDF$epi3 + smokingDF$epi4 +
             smokingDF$epi5 + smokingDF$epi6 + smokingDF$epi7 + smokingDF$epi8 + 
             smokingDF$epi9 + smokingDF$epi10)%>%
  tidy() %>% as.data.frame()

cellP <- data.frame()
cellP[1 , 1] <- i
cellP[1 , 2] <- lm1[2,2]
cellP[1 , 3] <- lm1[2,3]
cellP[1 , 4] <- lm1[2,4]
cellP[1 , 5] <- lm1[2,5]
colnames(cellP) <- c("cell_name","estimate_10epi","stderror_10epi","stat_10epi","pvalue_10epi")
write.table(cellP,"below4_B_reg_lin_epi10.txt",quote=FALSE,sep='\t',row.names = FALSE)

smokingDF = vars %>%
  filter( (AGE>4 ) | CaCo==0)
table(smokingDF$CaCo)
#   0   1 
# 198  62 

lm2 <- lm(smokingDF[,i] ~ smokingDF$CaCo + smokingDF$sex + smokingDF$plate +
             smokingDF$epi1 + smokingDF$epi2 + smokingDF$epi3 + smokingDF$epi4 +
             smokingDF$epi5 + smokingDF$epi6 + smokingDF$epi7 + smokingDF$epi8 + 
             smokingDF$epi9 + smokingDF$epi10)%>%
  tidy() %>% as.data.frame()

cellP <- data.frame()
cellP[1 , 1] <- i
cellP[1 , 2] <- lm2[2,2]
cellP[1 , 3] <- lm2[2,3]
cellP[1 , 4] <- lm2[2,4]
cellP[1 , 5] <- lm2[2,5]
colnames(cellP) <- c("cell_name","estimate_10epi","stderror_10epi","stat_10epi","pvalue_10epi")
write.table(cellP,"above4_B_reg_lin_epi10.txt",quote=FALSE,sep='\t',row.names = FALSE)


# yi=c(lm1[2,2],lm2[2,2])
# vi=c(lm1[2,3]^2,lm2[2,3]^2)
# mtdf = rma(yi, vi,method="FE")
# mtdf

```

```{bash Phet for age groups}

cd ~/DSALL/usc
metal
MARKER cell_name
EFFECT estimate_10epi
PVALUE pvalue_10epi
SCHEME STDERR
STDERR stderror_10epi
VERBOSE OFF
EFFECT_PRINT_PRECISION 20
STDERR_PRINT_PRECISION 20
PROCESS below4_B_reg_lin_epi10.txt
PROCESS above4_B_reg_lin_epi10.txt
ANALYZE HETEROGENEITY
exit
mv METAANALYSIS1.TBL below4.above4.meta.metal.txt


```

```{r excluding the ~30 GATA1 mutation-positive DS controls}

library(broom)
library(tidyverse)

setwd("~/DSALL/usc")

subjects = fread("ds.all.final.subs.AfterReview.dropWashU.txt")

GATA1_table = read_tsv("DS_controls_GATA1_results.txt")%>%
  as.data.frame%>%
  filter(Mutation=="yes")

subs_cellproportion_1 <-  fread('ds.all.covaraites.txt') %>% 
  as.data.frame() %>%
  filter(beadPosition%in%subjects$beadPosition) %>%
  filter(!beadPosition%in%GATA1_table$beadPosition) %>% 
  na.omit()  

table(subs_cellproportion_1$CaCo)
#   0   1 
# 168 126 

cellP <- data.frame()
j = 1
for(i in  c("CD8T","CD4T","NK","Bcell","Mono","Gran","nRBC")){
  
  smokingDF <- subs_cellproportion_1

  lm1 <- lm( smokingDF[,i] ~ smokingDF$CaCo + smokingDF$sex + smokingDF$plate + smokingDF$epi1 + smokingDF$epi2 +
               smokingDF$epi3+smokingDF$epi4+smokingDF$epi5+smokingDF$epi6+
               smokingDF$epi7+smokingDF$epi8+smokingDF$epi9+smokingDF$epi10)%>%
  tidy() %>% as.data.frame()

  # 6 epis, linear model
  cellP[j , 1] <- i
  cellP[j , 2] <- lm1[2,2]
  cellP[j , 3] <- lm1[2,3]
  cellP[j , 4] <- lm1[2,4]
  cellP[j , 5] <- lm1[2,5]
  j = j+1
 
  }

colnames(cellP) <- c("cell_name","estimate_10epi","stderror_10epi","stat_10epi","pvalue_10epi")

write.csv(cellP,"usc_B_reg_lin_excludeGATA.csv")

```

```{r race stratified cell proportions}

library(broom)
library(tidyverse)
library(data.table)
library(metafor)

setwd("~/DSALL/usc")

## cell proportion comparisons in eur and lat
### in eur subjects, without SNP controls

subjects = fread("ds.all.final.subs.AfterReview.dropWashU.txt")

clinical = fread('ds.all.covaraites.txt') %>% 
    as.data.frame() %>%
    filter(beadPosition%in%subjects$beadPosition)

subs_cellproportion_0 <- read.csv('~/sets/set3/clinical_variables.csv') %>% 
  dplyr::select(beadPosition,race) %>%
  inner_join(clinical,by="beadPosition") %>%
  mutate(subjectId=as.character(subjectId))

smokingDF <- subs_cellproportion_0 %>% filter(race==1)

# table(smokingDF$CaCo)
# 
#  0  1 
# 54 32 

cellP <- data.frame()
j = 1
for(i in  c("CD4T","CD8T","Bcell","NK","Gran","Mono","nRBC")){
  lm1 <- lm(smokingDF[,i] ~ smokingDF$CaCo + smokingDF$sex + 
              smokingDF$plate + smokingDF$epi1 + smokingDF$epi2 + 
              smokingDF$epi3 + smokingDF$epi4 + smokingDF$epi5 + smokingDF$epi6+ 
              smokingDF$epi7+ smokingDF$epi8+ smokingDF$epi9+ smokingDF$epi10) %>%
    tidy() %>% as.data.frame()
  cellP[j , 1] <- i
  cellP[j , 2] <- lm1[2,2]
  cellP[j , 3] <- lm1[2,3]
  cellP[j , 4] <- lm1[2,4]
  cellP[j , 5] <- lm1[2,5]
  j = j+1
  }
colnames(cellP) <- c("cell_name","estimate_10epi","stderror_10epi","stat_10epi","pvalue_10epi")
write.csv(cellP,"white_B_reg_lin_epi10.csv")
write.table(cellP,"white_B_reg_lin_epi10.txt",
            quote=FALSE,
            sep='\t',
            row.names = FALSE)

# ## plot CD8T cell for white
# smokingDF %>%
#   dplyr::select(CaCo,CD8T)%>%
#   reshape::melt(id=c("CaCo"),
#                 variable = "cellType")%>%
#   mutate(cellProportion = value) %>%
#   ggplot(aes(x=cellType, y=cellProportion, fill=as.factor(CaCo))) +
#   geom_boxplot(outlier.size=0.2)+
#   geom_point(colour="darkblue",
#              position=position_jitterdodge(),
#              alpha=0.7,size=0.5) +
#   theme_classic() +
#   xlab("Case Control Status") +
#   ylab("CD8T Proportions") +
#   theme(plot.title = element_text(hjust = 0.5),
#         axis.text.x = element_text(size=12),
#         axis.text.y = element_text(size=12),
#         axis.title.x = element_text(size=14),
#         axis.title.y = element_text(size=14))+
#   scale_fill_discrete(name="Disease Status",labels=c("DS","DS-ALL"))
# ggsave("B_proportions_all_ds_bcm.png",width = 8,height = 6)
# 
# a = smokingDF %>% filter(CaCo==1) %>% dplyr::select(CD8T)
# b = smokingDF %>% filter(CaCo==0) %>% dplyr::select(CD8T)
# t.test(a$CD8T,b$CD8T)
# wilcox.test(a$CD8T,b$CD8T)

### in lat subjects, without SNP controls

smokingDF <- subs_cellproportion_0 %>% filter(race==3)

# > table(smokingDF$CaCo)
# 
#  0  1 
# 96 86 

cellP <- data.frame()
j = 1
for(i in  c("CD4T","CD8T","Bcell","NK","Gran","Mono","nRBC")){
  lm1 <- lm(smokingDF[,i] ~ smokingDF$CaCo + smokingDF$sex + 
              smokingDF$plate + smokingDF$epi1 + smokingDF$epi2 + 
              smokingDF$epi3 + smokingDF$epi4 + smokingDF$epi5 + smokingDF$epi6+ 
              smokingDF$epi7+ smokingDF$epi8+ smokingDF$epi9+ smokingDF$epi10) %>%
    tidy() %>% as.data.frame()
  cellP[j , 1] <- i
  cellP[j , 2] <- lm1[2,2]
  cellP[j , 3] <- lm1[2,3]
  cellP[j , 4] <- lm1[2,4]
  cellP[j , 5] <- lm1[2,5]
  j = j+1
  }
colnames(cellP) <- c("cell_name","estimate_10epi","stderror_10epi","stat_10epi","pvalue_10epi")
write.csv(cellP,"latino_B_reg_lin_epi10.csv")
write.table(cellP,"latino_B_reg_lin_epi10.txt",
            quote=FALSE,
            sep='\t',
            row.names = FALSE)

## plot CD8T cell for latino
# smokingDF %>%
#   dplyr::select(CaCo,CD8T)%>%
#   reshape::melt(id=c("CaCo"),
#                 variable = "cellType")%>%
#   mutate(cellProportion = value) %>%
#   ggplot(aes(x=cellType, y=cellProportion, fill=as.factor(CaCo))) +
#   geom_boxplot(outlier.size=0.2)+
#   geom_point(colour="darkblue",
#              position=position_jitterdodge(),
#              alpha=0.7,size=0.5) +
#   theme_classic() +
#   xlab("Case Control Status") +
#   ylab("CD8T Proportions") +
#   theme(plot.title = element_text(hjust = 0.5),
#         axis.text.x = element_text(size=12),
#         axis.text.y = element_text(size=12),
#         axis.title.x = element_text(size=14),
#         axis.title.y = element_text(size=14))+
#   scale_fill_discrete(name="Disease Status",labels=c("DS","DS-ALL"))
# ggsave("B_proportions_all_ds_bcm.png",width = 8,height = 6)

## heterogeneity of eur and lat results
white.cp = fread("white_B_reg_lin_epi10.csv") %>%
  as.data.frame()%>%
  column_to_rownames(var="cell_name")

lat.cp = fread("latino_B_reg_lin_epi10.csv") %>%
  as.data.frame()%>%
  column_to_rownames(var="cell_name")

i=1
mtf = list()
for(ct in rownames(white.cp)){
  yi=c(white.cp[ct,"estimate_10epi"],lat.cp[ct,"estimate_10epi"])
  vi=c(white.cp[ct,"stderror_10epi"]^2,lat.cp[ct,"stderror_10epi"]^2)

  mtdf = rma(yi, vi,method="FE")
  mtf[[i]]=c("cell"=ct,"beta"=mtdf$beta,
             "pval"=mtdf$pval,
             "Heterogeneity"=mtdf$QEp,
             "HeterogeneityP"=mtdf$QEp)
  i=i+1
}

mtfDATA = mtf %>% bind_rows %>% as.data.frame()
write.csv(mtfDATA,"white.lat.meta.csv")
```

```{bash meta-analysis of cell proportions of usc european and latino}

cd ~/DSALL/usc
metal
MARKER cell_name
EFFECT estimate_10epi
PVALUE pvalue_10epi
SCHEME STDERR
STDERR stderror_10epi
VERBOSE OFF
EFFECT_PRINT_PRECISION 20
STDERR_PRINT_PRECISION 20
PROCESS white_B_reg_lin_epi10.txt
PROCESS latino_B_reg_lin_epi10.txt
ANALYZE HETEROGENEITY
exit
mv METAANALYSIS1.TBL white.lat.meta.metal.txt
```

```{r SNP corrected cell proportion comparisons in eur and lat and both}

setwd("~/DSALL/usc")

subjects = fread("ds.all.final.subs.AfterReview.dropWashU.txt")

clinical = fread('ds.all.covaraites.txt') %>% 
    as.data.frame() %>%
    filter(beadPosition%in%subjects$beadPosition)

subs_cellproportion_0 <- read.csv('~/sets/set3/clinical_variables.csv') %>% 
  dplyr::select(beadPosition,race) %>%
  inner_join(clinical,by="beadPosition") %>%
  mutate(subjectId=as.character(subjectId))

snps <- read.csv('DS_GWAS_ALL_SNPs_14Jul21.csv') %>%
  mutate(subjectId=as.character(StudyID))%>%
  dplyr::select(subjectId, rs7089424_ARID5B,	rs11978267_IKZF1, rs3731249_CDKN2A, rs3824662_GATA3) %>%
  inner_join(subs_cellproportion_0,by="subjectId")

### In both
# > table(snps$CaCo)
# 
#   0   1 
# 130 117 

reg <- list()
for(snpsL in c("rs7089424_ARID5B","rs11978267_IKZF1","rs3731249_CDKN2A","rs3824662_GATA3")){
  i <- c("CD8T","CD4T","NK","Bcell","Mono","Gran","nRBC")
  reg1 <- lapply(i,function(cp){
    c(data.frame(SNP=snpsL,cell=cp),summary(
      lm(snps[,cp] ~  snps$CaCo + snps$sex + snps[,snpsL] +
              snps$plate + snps$epi1 + snps$epi2 +
              snps$epi3 + snps$epi4 + snps$epi5 + snps$epi6 + 
              snps$epi7+ snps$epi8+ snps$epi9+ snps$epi10))$coefficients[2,c(1,2,4)])})
  reg <- c(reg,reg1) 
  }

i <- c("CD8T","CD4T","NK","Bcell","Mono","Gran","nRBC")

regall <- lapply(i,function(cp){
     c(data.frame(SNP="allFour",cell=cp),summary(
          lm(snps[,cp] ~  snps$CaCo + snps$sex + 
               snps$rs7089424_ARID5B + 
               snps$rs11978267_IKZF1 + 
               snps$rs3731249_CDKN2A + 
               snps$rs3824662_GATA3 +
               snps$plate + snps$epi1 + snps$epi2 +
               snps$epi3 + snps$epi4 + snps$epi5 + snps$epi6+
               snps$epi7+ snps$epi8+ snps$epi9+ snps$epi10))$coefficients[2,c(1,2,4)])})
regFinal <- c(reg,regall) 
regFinalDF <- bind_rows(regFinal) %>% as.data.frame() 

write.table(regFinalDF,"snp.controlled.cp.allSubs.txt",
            quote=FALSE,
            sep='\t',
            row.names = FALSE, 
            col.names = TRUE)  

### In eur
snpsEur =  snps %>% filter(race==1)

# > table(snpsEur$CaCo)
# 
#  0  1 
# 44 32 

reg <- list()
for(snpsL in c("rs7089424_ARID5B","rs11978267_IKZF1","rs3731249_CDKN2A","rs3824662_GATA3")){
  i <- c("CD8T","CD4T","NK","Bcell","Mono","Gran","nRBC")
  reg1 <- lapply(i,function(cp){
    c(data.frame(SNP=snpsL,cell=cp),summary(
      lm(snpsEur[,cp] ~  snpsEur$CaCo + snpsEur$sex + snpsEur[,snpsL] +
              snpsEur$plate + snpsEur$epi1 + snpsEur$epi2 +
              snpsEur$epi3 + snpsEur$epi4 + snpsEur$epi5 +snpsEur$epi6+
              snpsEur$epi7+ snpsEur$epi8+ snpsEur$epi9+ snpsEur$epi10))$coefficients[2,c(1,2,4)])})
  reg <- c(reg,reg1) 
  }

i <- c("CD8T","CD4T","NK","Bcell","Mono","Gran","nRBC")

regall <- lapply(i,function(cp){
     c(data.frame(SNP="allFour",cell=cp),summary(
          lm(snpsEur[,cp] ~  snpsEur$CaCo + snpsEur$sex + 
               snpsEur$rs7089424_ARID5B + 
               snpsEur$rs11978267_IKZF1 + 
               snpsEur$rs3731249_CDKN2A + 
               snpsEur$rs3824662_GATA3 +
               snpsEur$plate + snpsEur$epi1 + snpsEur$epi2 +
               snpsEur$epi3 + snpsEur$epi4 + snpsEur$epi5 + snpsEur$epi6+
               snpsEur$epi7+ snpsEur$epi8+ snpsEur$epi9+ snpsEur$epi10
             ))$coefficients[2,c(1,2,4)])})
regFinal <- c(reg,regall) 
regFinalDF <- bind_rows(regFinal) %>% as.data.frame() 

write.table(regFinalDF,"snp.controlled.cp.EurSubs.txt",
            quote=FALSE,
            sep='\t',
            row.names = FALSE, 
            col.names = TRUE)  


### In lat
snpsLat =  snps %>% filter(race==3)

# > table(snpsLat$CaCo)
# 
#  0  1 
# 86 85 

reg <- list()
for(snpsL in c("rs7089424_ARID5B","rs11978267_IKZF1","rs3731249_CDKN2A","rs3824662_GATA3")){
  i <- c("CD8T","CD4T","NK","Bcell","Mono","Gran","nRBC")
  reg1 <- lapply(i,function(cp){
    c(data.frame(SNP=snpsL,cell=cp),summary(
      lm(snpsLat[,cp] ~  snpsLat$CaCo + snpsLat$sex + snpsLat[,snpsL] +
              snpsLat$plate + snpsLat$epi1 + snpsLat$epi2 +
              snpsLat$epi3 + snpsLat$epi4 + snpsLat$epi5 +snpsLat$epi6+
              snpsLat$epi7 + snpsLat$epi8 + snpsLat$epi9 +snpsLat$epi10
         ))$coefficients[2,c(1,2,4)])})
  reg <- c(reg,reg1) 
  }

i <- c("CD8T","CD4T","NK","Bcell","Mono","Gran","nRBC")

regall <- lapply(i,function(cp){
     c(data.frame(SNP="allFour",cell=cp),summary(
          lm(snpsLat[,cp] ~  snpsLat$CaCo + snpsLat$sex + 
               snpsLat$rs7089424_ARID5B + 
               snpsLat$rs11978267_IKZF1 + 
               snpsLat$rs3731249_CDKN2A + 
               snpsLat$rs3824662_GATA3 +
               snpsLat$plate + snpsLat$epi1 + snpsLat$epi2 +
               snpsLat$epi3 + snpsLat$epi4 + snpsLat$epi5 + snpsLat$epi6+
              snpsLat$epi7 + snpsLat$epi8 + snpsLat$epi9 +snpsLat$epi10               
             ))$coefficients[2,c(1,2,4)])})
regFinal <- c(reg,regall) 
regFinalDF <- bind_rows(regFinal) %>% as.data.frame() 

write.table(regFinalDF,"snp.controlled.cp.LatSubs.txt",
            quote=FALSE,
            sep='\t',
            row.names = FALSE, 
            col.names = TRUE)  

```

```{r adding clinical variaable}
################################################################################################
## adding clinical variaable

subjects = fread("ds.all.final.subs.AfterReview.dropWashU.txt")

clinical = fread('ds.all.covaraites.txt') %>% 
  as.data.frame() %>%
  filter(beadPosition%in%subjects$beadPosition)%>%
  dplyr::select(epi1,epi2,epi3,epi4,epi5,epi6,epi7,epi8,epi9,epi10,beadPosition,CD8T,CD4T,NK,Bcell,Mono,Gran,nRBC)%>%
  as.data.frame() 

subs_cellproportion_0 <- read.csv('~/sets/set3/clinical_variables.csv') %>% 
  as.data.frame() %>%
  mutate(gestage=Y_gestation_week,birthWt=Y_birthweight) %>%
  dplyr::select(beadPosition,sex,CaCo,plate,collectionAge,gestage,birthWt)%>%
  inner_join(clinical,by="beadPosition") %>%
  na.omit()

table(subs_cellproportion_0$CaCo)
#   0   1 
# 173 116   

cellP <- data.frame()
j = 1

for(i in  c("CD8T","CD4T","NK","Bcell","Mono","Gran","nRBC")){
  
  smokingDF <- subs_cellproportion_0

  lm1 <- lm( smokingDF[,i] ~ smokingDF$CaCo + smokingDF$sex + smokingDF$plate + 
             smokingDF$collectionAge + smokingDF$birthWt + smokingDF$gestage +
             smokingDF$epi1 + smokingDF$epi2 + smokingDF$epi3+
             smokingDF$epi4 + smokingDF$epi5+smokingDF$epi6 + smokingDF$epi7
             +smokingDF$epi8 + smokingDF$epi9 + smokingDF$epi10)%>%
  tidy() %>% as.data.frame()

  # 10 epis, linear model
  cellP[j , 1] <- i
  cellP[j , 2] <- lm1[2,2]
  cellP[j , 3] <- lm1[2,3]
  cellP[j , 4] <- lm1[2,4]
  cellP[j , 5] <- lm1[2,5]
  j = j+1
  }
colnames(cellP) <- c("cell_name","estimate","stderror","stat","pvalue")
write.csv(cellP,"usc_B_reg_lin_clinicalVar.csv")

```

meta-analysis of cell proportions with BCM

```{r meta-analysis of cell proportions with BCM}

setwd("~/DSALL")
library(data.table)
library(tidyverse)
library(metafor)

# try metafor for meta-analysis
bcm = fread("bcm/balor_cp.txt") %>%
  as.data.frame()%>%
  column_to_rownames(var="cellType")

usc_10 = fread("usc/usc_cp_10pc.txt") %>%
  as.data.frame()%>%
  column_to_rownames(var="cellType")
rownames(usc_10)=c("cd8t","cd4t","nk","bcell","mono","gran","nrbc")

# 10pc version
i=1
mtf = list()
for(ct in rownames(bcm)){
  yi=c(bcm[ct,"beta"],usc_10[ct,"estimate_10epi"])
  vi=c(bcm[ct,"se"]^2,usc_10[ct,"stderror_10epi"]^2)
  
  mtdf = rma(yi, vi,method="FE")
  mtf[[i]]=c("cell"=ct,"beta"=mtdf$beta, "pval"=mtdf$pval,"HeterogeneityP"=mtdf$QEp)
  i=i+1
}
mtfDATA = mtf %>% bind_rows %>% as.data.frame() 
write.csv(mtfDATA,"usc_bcm_cp_10pc_metafor.csv")
```

```{bash meta-analysis of cell proportions with BCM}
cd ~/DSALL
metal
MARKER cellType
EFFECT beta
PVALUE pvalue
SCHEME STDERR
STDERR se
VERBOSE OFF
EFFECT_PRINT_PRECISION 20
STDERR_PRINT_PRECISION 20
PROCESS bcm/balor_cp.txt
PROCESS usc/usc_cp_10pc.txt
ANALYZE HETEROGENEITY
exit
mv METAANALYSIS1.TBL usc_bcm_cp_se_meta.txt
```

```{r table S1}


setwd("~/DSALL/usc")
library(tidyverse)
library(data.table)

subjects = fread("ds.all.final.subs.AfterReview.dropWashU.txt")

cellp <- fread("ds.all.covaraites.txt") %>%
  as.data.frame() %>%
  inner_join(subjects,by="beadPosition") %>%
  mutate(bdl = Bcell/(Bcell+CD4T+CD8T))

for(cell in c("CD4T","CD8T", "Bcell","NK","Gran","Mono","nRBC","bdl")){

  print(cell)
  
  controls = cellp %>% filter(CaCo==0) 
  control.num = controls[,cell]
  print("control mean")
  print(mean(control.num))

  cases = cellp %>% filter(CaCo==1)
  case.num = cases[,cell]
  print("case mean")
  print(mean(case.num))
    

  print("change")
  print( (mean(case.num)-mean(control.num))/mean(control.num))
}

# [1] "CD4T"
# [1] "control mean"
# [1] 0.1179047
# [1] "case mean"
# [1] 0.1072581
# [1] "change"
# [1] -0.09029817
# [1] "CD8T"
# [1] "control mean"
# [1] 0.05266318
# [1] "case mean"
# [1] 0.05794602
# [1] "change"
# [1] 0.1003138
# [1] "Bcell"
# [1] "control mean"
# [1] 0.008261422
# [1] "case mean"
# [1] 0.01285246
# [1] "change"
# [1] 0.5557196
# [1] "NK"
# [1] "control mean"
# [1] 0.02615251
# [1] "case mean"
# [1] 0.02633157
# [1] "change"
# [1] 0.006846568
# [1] "Gran"
# [1] "control mean"
# [1] 0.5943656
# [1] "case mean"
# [1] 0.6068791
# [1] "change"
# [1] 0.0210535
# [1] "Mono"
# [1] "control mean"
# [1] 0.06994091
# [1] "case mean"
# [1] 0.0663454
# [1] "change"
# [1] -0.05140774
# [1] "nRBC"
# [1] "control mean"
# [1] 0.1148445
# [1] "case mean"
# [1] 0.1011639
# [1] "change"
# [1] -0.119123
# [1] "bdl"
# [1] "control mean"
# [1] 0.05463124
# [1] "case mean"
# [1] 0.07023946
# [1] "change"
# [1] 0.2857013
```

Addressing reviewers' questions 

```{r Addressing reviewers' questions }

# Include 10 PCs in cell proportion comparison
# Add Phet to see if its’ statistically significant

setwd("~/DSALL")
library(data.table)
library(tidyverse)
library(metafor)

bcm = fread("bcm/balor_cp.txt") %>%
  as.data.frame()%>%
  column_to_rownames(var="cellType")

usc_10 = fread("usc/usc_cp_10pc.txt") %>%
  as.data.frame()%>%
  column_to_rownames(var="cell_name")
rownames(usc_10)=c("cd8t","cd4t","nk","bcell","mono","gran","nrbc")

i=1
mtf = list()
for(ct in rownames(bcm)){
  yi=c(bcm[ct,"beta"],usc_10[ct,"estimate_10epi"])
  vi=c(bcm[ct,"se"]^2,usc_10[ct,"stderror_10epi"]^2)
  
  mtdf = rma(yi, vi,method="FE")
  mtf[[i]]=c("cell"=ct,"beta"=mtdf$beta, "pval"=mtdf$pval,"HeterogeneityP"=mtdf$QEp)
  i=i+1
}
mtfDATA = mtf %>% bind_rows %>% as.data.frame() 
#write.csv(mtfDATA,"usc_bcm_cp_10pc_metafor.csv")

#	p-value for association between batch and cell types
cellp_usc <- fread("~/DSALL/usc/ds.all.covaraites.postReviewQCs.txt") %>%
  as.data.frame() %>%
  mutate(batch=1) %>%
  dplyr::select(batch,CD8T,CD4T,NK,Bcell,Mono,Gran,nRBC)
cellp_ucb <- fread("~/DSALL/bcm/11_cellsproportion_idoldeconvolution.csv") %>%
  mutate(CaCo=phenotype) %>%
  as.data.frame() %>%
  mutate(batch=2) %>%
  dplyr::select(batch,CD8T,CD4T,NK,Bcell,Mono,Gran,nRBC)
  
cellp_tot = rbind(cellp_usc,cellp_ucb) %>%
  as.data.frame()

cor.test(cellp_tot$Bcell,cellp_tot$batch)

cor.test(cellp_tot$CD8T,cellp_tot$batch)

cor.test(cellp_tot$CD4T,cellp_tot$batch)

cor.test(cellp_tot$NK,cellp_tot$batch)

cor.test(cellp_tot$Mono,cellp_tot$batch)

cor.test(cellp_tot$Gran,cellp_tot$batch)

cor.test(cellp_tot$nRBC,cellp_tot$batch)

```

export TMPDIR=/ccls/home/sli/tempos
export EXPERIMENT_HUB_CACHE=/ccls/home/sli/.cache/R/ExperimentHub

```{r}
#THIS IS R 4.12 NOT DEFAULT R
# /ccls/home/sli/dependencies/r_self/R-4.1.2/bin/./R

#options(EXPERIMENT_HUB_CACHE="/ccls/home/sli/.cache/R/ExperimentHub")

#tempfile()
#.libPaths()
#[1] "/ccls/home/sli/dependencies/r_self/R-4.1.2/lib64/R/library"
#tools::R_user_dir("ExperimentHub", which="cache")
#BiocManager::install("FlowSorted.CordBloodCombined.450k")
#BiocManager::install("EpiDISH")
#BiocManager::install("IlluminaHumanMethylation450kmanifest")

#	another cell deconvolution algorithm to confirm our findings (EPIDISH)
setwd("~/DSALL/usc")
library(data.table)
library(tidyverse)
library(EpiDISH)
library(FlowSorted.Blood.EPIC)
library(FlowSorted.CordBloodCombined.450k)
library(IlluminaHumanMethylation450kmanifest)
library(impute)

## LOADING reference CpGs
#a = FlowSorted.CordBloodCombined.450k() #RGSET
#MsetEx <- preprocessNoob(a)
#RSet <- ratioConvert(MsetEx, what = "both", keepCN = TRUE)
#beta <- getBeta(RSet)
# write_rds(beta,"select_beta_ref_bulkulski.rds")
beta = read_rds("select_beta_ref_bulkulski.rds")

# select = colData(a) %>%
#   as.data.frame %>%
#   arrange(CellType) %>%
#   dplyr::select(CellType,Sex,Age,Study,ArrayTypes)

selectedSubs = c("200693480131_R08C01",
                 "200693480050_R01C01",
                 "200693480149_R06C01",
                 "200693480149_R08C01",
                 "200693480050_R08C01",
                 "200693480050_R03C01",
                 "GSM1672189")

cell_types = c("Bcell","CD4T","CD8T","Gran","Mono","NK","nRBC")

select_beta = beta[,selectedSubs]
colnames(select_beta) = cell_types  

#write_rds(select_beta,"epidish.ref.rds")
#select_beta = read_rds("epidish.ref.rds")

## LOADING query CpGs

subjects = fread("ds.all.final.subs.AfterReview.dropWashU.txt")

var = fread('ds.all.covaraites.txt') %>% 
    as.data.frame() %>%
    filter(beadPosition%in%subjects$beadPosition)

# betas3_0 = fread("~/sets/set3/set3_Sesame_beta.csv") %>%
#   as.data.frame()%>%
#   dplyr::select(-V1)
# betas3_1 = betas3_0 %>%
#   column_to_rownames(var="probeId")
# noob.to.include = betas3_1[,var$beadPosition] %>% as.matrix
# noob.to.include1 = noob.to.include[rowMeans(is.na(noob.to.include))<0.15,]
# noob.to.include2 = noob.to.include1[rownames(noob.to.include1) %in% rownames(select_beta),]
# noob.to.include3 <- impute.knn(noob.to.include2)$data
#write_rds(noob.to.include3,"query_beta_epdish.rds")
noob.to.include3 = read_rds("query_beta_epdish.rds")

data(centBloodSub.m)
BloodFrac.m <- epidish(beta.m = noob.to.include3, ref.m = select_beta, method = "CP")$estF
#colnames(BloodFrac.m) = c("CD4T",  "NK" ,   "nRBC" , "Bcell", "Mono",  "CD8T" , "Gran" )
write_rds(BloodFrac.m,"BloodFrac.m.rds")

```

```{r formal statistical comparison}

library(broom)
library(tidyverse)

setwd("~/DSALL/usc")

subs_cellproportion_1 <- fread('ds.all.covaraites.txt') %>% 
  as.data.frame() %>%
  dplyr::select(-CD8T,-CD4T,-NK,-Bcell,-Mono,-Gran,-nRBC )
subs_cellproportion_2 <- read_rds("BloodFrac.m.rds") %>%
  as.data.frame() %>%
  rownames_to_column(var="beadPosition") %>%
  inner_join(subs_cellproportion_1,by="beadPosition")

cellP <- data.frame()
j = 1

for(i in  c("CD8T","CD4T","NK","Bcell","Mono","Gran","nRBC")){
  
  smokingDF <- subs_cellproportion_2

  lm1 <- lm( smokingDF[,i] ~ smokingDF$CaCo + smokingDF$sex + smokingDF$plate + smokingDF$epi1 + smokingDF$epi2 + smokingDF$epi3 + smokingDF$epi4 + smokingDF$epi5 + smokingDF$epi6+ smokingDF$epi7+ smokingDF$epi8+ smokingDF$epi9+ smokingDF$epi10)%>%
  tidy() %>% as.data.frame()

  cellP[j , 1] <- i
  cellP[j , 2] <- lm1[2,2]
  cellP[j , 3] <- lm1[2,3]
  cellP[j , 4] <- lm1[2,4]
  cellP[j , 5] <- lm1[2,5]
  j = j+1
 
  }

colnames(cellP) <- c("cellType","beta","se","stat_10epi","pvalue")

write.table(cellP,
            sep='\t',
            row.names = FALSE,
            quote=FALSE,
            "usc_B_reg_lin_epi10_epdish.txt")

```

```{r compare epdish results}

setwd("~/DSALL/usc")
library(data.table)
library(tidyverse)

var = fread("ds.all.covaraites.txt") %>%
  as.data.frame() %>%
  dplyr::select(beadPosition,CaCo)

BloodFrac.m = read_rds("BloodFrac.m.rds") %>%
  as.data.frame() %>%
  rownames_to_column(var="beadPosition") %>%
  inner_join(var,by="beadPosition")

BloodFrac.m %>%
  dplyr::select(beadPosition,CaCo,CD8T,CD4T,NK,Bcell,Mono,Gran,nRBC)%>%
  reshape::melt(id=c("beadPosition","CaCo"),
                variable = "cellType")%>%
  mutate(cellProportion = value) %>%
  ggplot(aes(x=cellType, y=cellProportion, fill=as.factor(CaCo))) +
  geom_boxplot(outlier.size=0.2)+
  geom_point(colour="darkblue",
             position=position_jitterdodge(),
             alpha=0.5,size=0.01) +
  theme_classic() +
#ggtitle("Cell proportions in Down Syndrome subjects\nDiscovery Study")+
  xlab("Cell Types") +
  ylab("Proportions") +
  theme(plot.title = element_text(hjust = 0.5),
        axis.text.x = element_text(size=12),
        axis.text.y = element_text(size=12),
        axis.title.x = element_text(size=14),
        axis.title.y = element_text(size=14))+
  scale_fill_discrete(name="Disease Status",labels=c("DS","DS-ALL"))
ggsave("epidish_Cell_proportions_all_ds.png",width = 8,height = 6)

```

```{r BCM using epidish results}

library(broom)
library(tidyverse)

setwd("~/DSALL/bcm")

subs_cellproportion_1 <- read_rds('BloodFrac.m.annotated.rds') %>% 
  as.data.frame() %>%
  dplyr::select(-SEX,-epis1,-epis2,-epis3)

subs_cellproportion_1 %>%
  dplyr::select(id,phenotype,CD8T,CD4T,NK,Bcell,Mono,Gran,nRBC)%>%
  reshape::melt(id=c("id","phenotype"),
                variable = "cellType")%>%
  mutate(cellProportion = value) %>%
  ggplot(aes(x=cellType, y=cellProportion, fill=as.factor(phenotype))) +
  geom_boxplot(outlier.size=0.2)+
  geom_point(colour="darkblue",
             position=position_jitterdodge(),
             alpha=0.5,size=0.01) +
  theme_classic() +
  xlab("Cell Types") +
  ylab("Proportions") +
  theme(plot.title = element_text(hjust = 0.5),
        axis.text.x = element_text(size=12),
        axis.text.y = element_text(size=12),
        axis.title.x = element_text(size=14),
        axis.title.y = element_text(size=14))+
  scale_fill_discrete(name="Disease Status",labels=c("DS","DS-ALL"))
ggsave("epidish_Cell_proportions_all_ds_bcm.png",width = 8,height = 6)


subs_cellproportion_2 <- read_rds('BloodFrac.m.annotated.rds') %>% 
  as.data.frame() 
cellP <- data.frame()
j = 1

for(i in  c("CD8T","CD4T","NK","Bcell","Mono","Gran","nRBC")){
  
  smokingDF <- subs_cellproportion_2

  lm1 <- lm(smokingDF[,i] ~ smokingDF$phenotype + smokingDF$SEX + smokingDF$epis1 + smokingDF$epis2 + smokingDF$epis3)%>%
  tidy() %>% as.data.frame()

  # only 6 epis, linear model
  cellP[j , 1] <- i
  cellP[j , 2] <- lm1[2,2]
  cellP[j , 3] <- lm1[2,3]
  cellP[j , 4] <- lm1[2,4]
  cellP[j , 5] <- lm1[2,5]
  j = j+1
 
  }

colnames(cellP) <- c("cellType","beta","se","stat_3epi","pvalue")

write.table(cellP,
            sep='\t',
            row.names = FALSE,
            quote=FALSE,
            "bcm_B_reg_lin_epi3_epdish.txt")

```

```{bash meta-analysis of cell proportions with BCM}
cd ~/DSALL
metal
MARKER cellType
EFFECT beta
PVALUE pvalue
SCHEME STDERR
STDERR se
VERBOSE OFF
EFFECT_PRINT_PRECISION 20
STDERR_PRINT_PRECISION 20
PROCESS bcm/bcm_B_reg_lin_epi3_epdish.txt
PROCESS usc/usc_B_reg_lin_epi10_epdish.txt
ANALYZE HETEROGENEITY
exit
mv METAANALYSIS1.TBL usc_bcm_epidish_cp_se_meta.txt
```

>> TCA for cell specific difference

```{r}

library(tidyverse)
library(data.table)
library(impute)
library(fastDummies)

setwd("~/DSALL/usc")

winsorize <- function(methylation,pct=winsorize.pct) {
  quantiles <- matrixStats::rowQuantiles(methylation, probs=c(pct,1-pct), na.rm=T)
  low <- quantiles[,1]
  upper <- quantiles[,2]

  outliers.lower <- rowSums(methylation < low, na.rm=T)
  outliers.upper <- rowSums(methylation > upper, na.rm=T)
  
  idx <- which(methylation < low, arr.ind=T)
  methylation[idx] <- low[idx[,1]]
  
  idx <- which(methylation > upper, arr.ind=T)
  methylation[idx] <- upper[idx[,1]]

  n <- rowSums(!is.na(methylation))
  log <- data.frame(outliers.lower, outliers.upper, n)
  return(list(methylation=methylation, log=log))
}


var = fread("ds.all.covaraites.postReviewQCs.txt") %>%
  as.data.frame() %>%
  column_to_rownames(var="beadPosition")

betas3_0 = fread("~/sets/set3/set3_Sesame_beta.csv") %>%
  as.data.frame()%>%
  dplyr::select(-V1) %>%
  column_to_rownames(var="probeId")
betas3_1 = betas3_0[rowMeans(is.na(betas3_0))<0.05,rownames(var)] %>% 
  as.matrix
betas3_2 = impute.knn(betas3_1)$data

replace.outliers <- winsorize(betas3_2, 0.005)
betas3_3 <- replace.outliers$methylation
write_rds(betas3_3,"betas3_3_tca.rds")
write_rds(var,"var_tca.rds")





```

export TMPDIR=/ccls/home/sli/tempos
export EXPERIMENT_HUB_CACHE=/ccls/home/sli/.cache/R/ExperimentHub

```{r}
#THIS IS R 4.12 NOT DEFAULT R
# /ccls/home/sli/dependencies/r_self/R-4.1.2/bin/./R

library(TCA)
library(ggpubr)
library(pracma)
library(matrixStats)
library(tidyverse)
library(data.table)
library(impute)
library(fastDummies)
library(hexbin)
#install.packages("DataExplorer")
#library(DataExplorer)

setwd("~/DSALL/usc")

betas3_3 = read_rds("betas3_3_tca.rds")
var = read_rds("var_tca.rds")

X = betas3_3

W = var %>% 
  dplyr::select(CD8T,CD4T,NK,Bcell,Mono,Gran,nRBC) %>%
  as.data.frame()

harmonize = function(x){
   dat = x
   max = which.max(dat)
   dat[max] = 1-sum(dat[-max])
  if(min(dat)<0){
    neg = which(dat<0)
    max = which.max(dat)
    dat[neg] = 0 
    dat[max] = 1-sum(dat[-max])
    }
   return(dat)
}

W1 = apply(W,1,harmonize) %>% t %>% as.data.frame()

C1 = var %>% dplyr::select(CaCo,sex) 

#C2 = var %>% dplyr::select(plate,epi1,epi2,epi3,epi4,epi5,epi6,epi7,epi8,epi9,epi10) %>% 
#failed because of over corrrection: Error in quadprog::solve.QP(Dmat, dvec, t(Amat), bvec, meq = meq) : matrix D in quadratic function is not positive definite!)
C2 = var %>% dplyr::select(epi1,epi2,epi3,epi4,epi5,epi6,epi7,epi8,epi9,epi10) %>%
  rownames_to_column(var="beadPosition") %>%
  #dummy_cols(select_columns="plate") %>%
  #dplyr::select(-plate) %>%
  column_to_rownames(var="beadPosition") 

#C1 - for covariates that may affect methylation at the cell-type level (e.g., age and gender).
#C2 - for covariates that are not expected to affect methylation at the cell-type level but rather may affect tissue-level methylation, regardless of the cell-type composition and the methylation in the different cell types (e.g., batch information and factors that capture technical variation).

# p-values for the effects of caco on methylation at the cell-type level under two statistical tests - a marginal conditional test and a joint test (under the assumption X|Y, caco affects methylation).
 
# tca.mdl.hannum <- tca(X = X,
#                       W = W1,
#                       C1 = C1,
#                       C2 = C2,
#                       vars.mle = TRUE,
#                       parallel=TRUE,
#                       num_cores=20)


#Since you did get a "non-positive definite" error, you either have very low variation in cell-type fractions or one or more of your included covariates (i.e. in 'C1') are redundant as they can be linearly explained by the other covariates and the cell type fractions.

# tca.mdl.hannum <- tca(X = X,
#                       W = W1,
#                       C1 = C1,
#                       C2 = C2)
#write_rds(tca.mdl.hannum,"tca.mdl.hannum.rds")
```

```{r check tca results}

library(tidyverse)
library(data.table)
library(methylGSA)
library(ggpubr)
library(IlluminaHumanMethylationEPICanno.ilm10b4.hg19)
data("IlluminaHumanMethylationEPICanno.ilm10b4.hg19")
annoEPIC <- IlluminaHumanMethylationEPICanno.ilm10b4.hg19 %>%
  getAnnotation %>%
  as.data.frame %>%
  rownames_to_column(var="cpg") %>%
  dplyr::select(cpg,chr,pos,Relation_to_Island,UCSC_RefGene_Name,
                Probe_maf,UCSC_RefGene_Group,CpG_maf)
plot_qq <- function(pvals, labels, ggarrange.nrow = 1, ggarrange.ncol = 1, alpha = 0.05, experiment_wide_line = TRUE){
  significance_th <- list(alpha/length(pvals[[1]]))
  if(length(pvals)-1) significance_th[[2]] <- alpha/(length(pvals)*length(pvals[[1]]))
  qqplots <- lapply(1:length(pvals), function(p){
    df <- data.frame(pvals.obs = -log10(sort(pvals[[p]])), pvals.exp = -log10(sort((1:length(pvals[[1]]))/length(pvals[[1]]))));
    qqplot <- ggplot(df, aes(x = pvals.exp, y = pvals.obs)) +
      stat_binhex(geom = "point", bins=1000, size=1) +
      geom_abline() +
      ggtitle(labels[p]) +
      xlab(expression(Expected~-log[10](P))) + ylab(expression(Observed~-log[10](P))) +
      theme_bw() +
      guides(fill="none") +
      geom_hline(yintercept=-log10(significance_th[[1]]), linetype="dashed", color = "red", size=1)
    if (length(significance_th)-1 & experiment_wide_line) qqplot <- qqplot + geom_hline(yintercept=-log10(significance_th[[2]]), linetype="dashed", color = "red", size=0.5)
    return(qqplot)
  })
  ggarrange(plotlist = qqplots, ncol = ggarrange.ncol, nrow = ggarrange.nrow)
}


# usc
setwd("~/DSALL/usc")

tca.mdl.hannum = read_rds("tca.mdl.hannum.rds")
# Extract CpG-level p-values for the association with age, under a joint test
tca.mdl.hannum.pvals.joint <- tca.mdl.hannum$gammas_hat_pvals.joint[,"CaCo"]

# Extract p-values for each cell type for its association with age, under a marginal conditional test
tca.mdl.hannum.pvals.marg_cond <- 
  tca.mdl.hannum$gammas_hat_pvals[,paste(colnames(W1),"CaCo",sep=".")]  
# qq-plots - for the p-values of the joint test, and for the marginal conditional p-values for B cells

#celltype="Bcell"

for(celltype in c("Bcell","CD8T","CD4T","NK","Mono","Gran","nRBC")){
  
  png(paste0(celltype,".p.png"),width = 800, height = 800, units = "px")
  plot_qq(list(tca.mdl.hannum.pvals.joint,
               tca.mdl.hannum.pvals.marg_cond[,paste0(celltype,".CaCo")]),
          labels = c("Joint test with CaCo", 
                     paste0(celltype," marginal conditional test with CaCo")),
          ggarrange.nrow = 1,
          ggarrange.ncol = 2,
          experiment_wide_line = FALSE)
  dev.off()
  
  
  cell.result = data.frame(cell_gammas_hat_pvals =
                  tca.mdl.hannum$gammas_hat_pvals[,paste0(celltype,".CaCo")],
                cell_gammas_hat =
                  tca.mdl.hannum$gammas_hat[,paste0(celltype,".CaCo")]) %>%
    arrange(cell_gammas_hat_pvals) %>%
    mutate(fdr=p.adjust(cell_gammas_hat_pvals, method = "fdr", 
                        n=length(cell_gammas_hat_pvals))) %>%
    mutate(bf=p.adjust(cell_gammas_hat_pvals, method = "bonferroni",
                        n=length(cell_gammas_hat_pvals))) 
  write_rds(cell.result, paste0(celltype,".cpgs.result.rds"))
  
  # output CpGs that are differentially methylated in B cells
  ## can be default R
  
  cell.cpgs = read_rds(paste0(celltype,".cpgs.result.rds")) %>%
    filter(bf<0.05) %>%
    rownames_to_column(var="cpg") %>%
    left_join(annoEPIC,by="cpg")
  
  write.csv(cell.cpgs,paste0(celltype,".cpgs.csv"))

}

# bcm

setwd("~/DSALL/bcm")

tca.mdl.hannum = read_rds("bcm.tca.mdl.hannum.rds")
tca.mdl.hannum.pvals.joint <- tca.mdl.hannum$gammas_hat_pvals.joint[,"phenotype"]

# Extract p-values for each cell type for its association with age, under a marginal conditional test
tca.mdl.hannum.pvals.marg_cond <- 
  tca.mdl.hannum$gammas_hat_pvals[,paste(c("CD8T","CD4T","NK","Bcell",
                                           "Mono","Gran","nRBC"),"phenotype",sep=".")]  
# qq-plots - for the p-values of the joint test, and for the marginal conditional p-values for B cells

#celltype="Bcell"

for(celltype in c("Bcell","CD8T","CD4T","NK","Mono","Gran","nRBC")){
  
  png(paste0("bcm.",celltype,".p.png"),width = 800, height = 800, units = "px")
  plot_qq(list(tca.mdl.hannum.pvals.joint,
               tca.mdl.hannum.pvals.marg_cond[,paste0(celltype,".phenotype")]),
          labels = c("Joint test with CaCo", 
                     paste0(celltype," marginal conditional test with CaCo")),
          ggarrange.nrow = 1,
          ggarrange.ncol = 2,
          experiment_wide_line = FALSE)
  dev.off()
  
  
  cell.result = data.frame(cell_gammas_hat_pvals =
                  tca.mdl.hannum$gammas_hat_pvals[,paste0(celltype,".phenotype")],
                cell_gammas_hat =
                  tca.mdl.hannum$gammas_hat[,paste0(celltype,".phenotype")]) %>%
    arrange(cell_gammas_hat_pvals) %>%
    mutate(fdr=p.adjust(cell_gammas_hat_pvals, method = "fdr", 
                        n=length(cell_gammas_hat_pvals))) %>%
    mutate(bf=p.adjust(cell_gammas_hat_pvals, method = "bonferroni",
                        n=length(cell_gammas_hat_pvals))) 
  write_rds(cell.result, paste0("bcm.",celltype,".cpgs.result.rds"))
  
  # output CpGs that are differentially methylated in B cells
  ## can be default R
  
  cell.cpgs = read_rds(paste0("bcm.",celltype,".cpgs.result.rds")) %>%
    filter(bf<0.05) %>%
    rownames_to_column(var="cpg") %>%
    left_join(annoEPIC,by="cpg")
  
  write.csv(cell.cpgs,paste0("bcm.",celltype,".cpgs.csv"))

}

```

















################################################################################
################################################################################
################################################################################
################################################################################
##### First submission, results no longer used

Methylation array SNPs to calculate PiHat

```{r}

setwd("~/DSALL/usc")

library(tidyverse)
library(data.table)

beta3 <- read_rds("~/sets/set3/RGset.rds")
snp = getSnpBeta(RGset)
#> dim(snp)
#[1]  59 944
```

```{r Perform case-case EWAS, e.g. Washington cases vs. California cases, followed by removal of significant CpGs and then repeat of DS-ALL EWAS (similar to what we have done for GWAS with Kaiser controls)}

setwd("~/DSALL/usc")
library(tidyverse)
library(data.table)
library(MASS)
library(lmtest)
library(sandwich)
library(parallel) 
library(minfi)
library(wateRmelon)
library(bacon)
library(lumi)

RLM.Robust <- function(meth, cov){

    dat <- data.frame(y = meth, cov)
    ff  <- formula(paste("y ~ ", paste0(colnames(cov), collapse="+")))
    mod <- try(rlm(ff, data = dat, maxit = 200))
    cf = try(coeftest(mod))
    Ncase = nrow(dat %>% filter(CaCo==1) %>% na.omit)
    Ncontrol = nrow(dat %>% filter(CaCo==0) %>% na.omit)

  	if (exists(class(cf)) == FALSE) {
  		out = c(rep(NA,4),Ncase,Ncontrol)
  	} else if (exists(class(cf)) == TRUE) {
  		if( class(cf)[1] == "try-error") {
  		cat("error throw by CpG", set, ";")
  		out = c(rep(NA,4),Ncase,Ncontrol)
  		print(paste0(cov, " - chol2inv could not be computed"))
  		} 
  		else {
  			coef = cf[2,"Estimate"]
  			se = cf[2,"Std. Error"]
  			z.value = cf[2,"z value"]
  			p.value = cf[2,"Pr(>|z|)"]
  			out = c(coef, se, z.value,p.value,Ncase,Ncontrol) 
  		}
    }
  names(out) = c("Beta","SE", "Z","P","ncase","ncontrol"); 
  return(out) 
  }

clinical3 = fread("~/sets/set3/clinical_variables.csv") %>%
  filter(Trisomy21==1) %>%
  filter(Sample_Source=="Washington Biobank") %>%
  dplyr::select(beadPosition)

covs = read_rds("covs.postReviewQCs.rds") %>%
  filter(CaCo==1)
covs$CaCo[rownames(covs)%in%clinical3$beadPosition]=0

methyl = read_rds("methyl.postReviewQCs.rds")
methyl = methyl[,rownames(covs)]

cl <- makeCluster(40, type="FORK")
clusterEvalQ(cl,library(MASS))
clusterEvalQ(cl,library(lmtest))

# this has to be true to continue

if(identical(rownames(covs),colnames(methyl))==TRUE){

  # methylt = methyl[1:100,] 
  # results = t(apply(methylt,1,RLM.Robust,covs))
  
  results <- t(parApply(cl, methyl, 1, RLM.Robust, covs)) 
  results <- as.data.frame(results) %>%
    na.omit()
  
  write.table(results,"ds.all.usc.sesame.m.caseVsCase.ewas.txt",
        quote=FALSE,
        sep='\t')  
}

stopCluster(cl)

library(bacon)
library(IlluminaHumanMethylationEPICanno.ilm10b4.hg19)
data("IlluminaHumanMethylationEPICanno.ilm10b4.hg19")
annoEPIC <- IlluminaHumanMethylationEPICanno.ilm10b4.hg19 %>%
  getAnnotation %>%
  as.data.frame %>%
  rownames_to_column(var="cpg") %>%
  dplyr::select(cpg,chr,pos,Relation_to_Island,UCSC_RefGene_Name,
                Probe_maf,UCSC_RefGene_Group,CpG_maf)

results <- results %>% 
  as.data.frame() %>%
  rownames_to_column(var="cpg") %>%
  dplyr::filter(grepl('cg', cpg)) %>%
  filter(!is.na(P))%>% 
  left_join(annoEPIC,by="cpg") %>%
  arrange(by=P) %>% # 701293
  dplyr::filter(chr!="chrX"&chr!="chrY") %>%  #687445
  dplyr::filter(Probe_maf<0.05|is.na(Probe_maf)) %>% # 629078
  dplyr::filter(CpG_maf<0.05|is.na(CpG_maf))  # 628764

results$bacon_P = pval(bacon(results$Z,na.exclude = TRUE))

results <- results %>%
  mutate(fdr=p.adjust(bacon_P, method = "fdr", n=length(bacon_P))) %>%
  mutate(bf=p.adjust(bacon_P, method = "bonferroni", n=length(bacon_P))) 

write_rds(results,"ds.all.usc.sesame.m.caseVsCase.ewas.anno.rds")

#results = read_rds("ds.all.usc.sesame.m.caseVsCase.ewas.anno.rds")
#nrow(results)
#[1] 628764
#table(results$P<0.05)
#table(results$bacon_P<0.05)

droppCpGs = results %>%
  dplyr::filter(fdr<0.05)%>%
  dplyr::select(cpg)

print("number of dropped CpGs is")
print(nrow(droppCpGs))
print(droppCpGs)
#[1] "number of dropped CpGs is"
#[1] 2645

## check EWAS results without these CpGs
setwd("~/DSALL/usc")
library(tidyverse)
library(data.table)
library(bacon)
library(IlluminaHumanMethylationEPICanno.ilm10b4.hg19)
data("IlluminaHumanMethylationEPICanno.ilm10b4.hg19")
annoEPIC <- IlluminaHumanMethylationEPICanno.ilm10b4.hg19 %>%
  getAnnotation %>%
  as.data.frame %>%
  rownames_to_column(var="cpg") %>%
  dplyr::select(cpg,chr,pos,Relation_to_Island,UCSC_RefGene_Name,
                Probe_maf,UCSC_RefGene_Group,CpG_maf)

results <- fread("ds.all.usc.sesame.m.ewas.txt") %>% 
  dplyr::rename("cpg"="V1") %>%
  dplyr::filter(grepl('cg', cpg)) %>%
  filter(!cpg%in%droppCpGs$cpg) %>%
  filter(!is.na(P))%>% 
  left_join(annoEPIC,by="cpg") %>%
  arrange(by=P) %>% # 701293
  dplyr::filter(chr!="chrX"&chr!="chrY") %>%  #687445
  dplyr::filter(Probe_maf<0.05|is.na(Probe_maf)) %>% # 629078
  dplyr::filter(CpG_maf<0.05|is.na(CpG_maf))  # 628764

results$bacon_P = pval(bacon(results$Z,na.exclude = TRUE))

results <- results %>%
  mutate(fdr=p.adjust(bacon_P, method = "fdr", n=length(bacon_P))) %>%
  mutate(bf=p.adjust(bacon_P, method = "bonferroni", n=length(bacon_P))) 

write_rds(results,"ds.all.usc.sesame.m.ewas.anno.without.CVSs.rds")

ds.all.usc.sesame.m.ewas.anno.without.CVSs = read_rds("ds.all.usc.sesame.m.ewas.anno.without.CVSs.rds")

print(0.05/nrow(results)) # bonferroni threshold

# raw P values
print("raw.lambda.values")
Nbeta <- sum(!is.na(results$P))
tmp <- results[!is.na(results$P),]
lambda = median(qchisq(1- tmp$P,1))/qchisq(0.5,1)
print(round(lambda, 2))

# bacon P values
print("bacon.lambda.values")
Nbeta <- sum(!is.na(results$bacon_P))
tmp <- results[!is.na(results$bacon_P),]
lambda = median(qchisq(1- tmp$bacon_P,1))/qchisq(0.5,1)
print(round(lambda, 2))


#[1] 7.985702e-08
#[1] "raw.lambda.values"
#[1] 1.98
#[1] "bacon.lambda.values"
#[1] 1.41

```

```{r detection p values and missing values for 18 washington cases}

setwd("~/DSALL/usc")
library(tidyverse)
library(data.table)
library(minfi)

# compare missing
ds.all.final.subs.AfterReview.txt = fread("ds.all.final.subs.AfterReview.txt")

betaRaw= fread("~/sets/set3/set3_Sesame_beta.csv")
betaRawMissing = colMeans(is.na(betaRaw[,-c(1,2)]))
rm(betaRaw)

betaRawMissingDF = data.frame(missing=betaRawMissing) %>%
  rownames_to_column(var="beadPosition")

clinical3 = fread("~/sets/set3/clinical_variables.csv") %>%
  filter(Trisomy21==1) %>%
  dplyr::select(beadPosition,Sample_Source) %>%
  inner_join(betaRawMissingDF,by="beadPosition") %>%
  inner_join(ds.all.final.subs.AfterReview.txt,by="beadPosition")
  
clinical3 %>%
  group_by(Sample_Source) %>%
  summarize(mean = mean(missing), 
            n = n()) %>%
  as.data.frame()
#       Sample_Source      mean   n
# 1        Cal_Biobank 0.1409387 196
# 2               CCLS 0.1682953  18
# 3              CCRLP 0.1517387 108
# 4         CCRLP_CTRL 0.1337315   2
# 5 Washington Biobank 0.1592372  18

ggplot(clinical3, aes(x=Sample_Source, y=missing, fill=as.factor(Sample_Source))) +
  geom_boxplot(outlier.size=0.2)+
  theme_classic() +
  xlab("Sample Source") +
  ylab("Missing Percentage") +
  theme(plot.title = element_text(hjust = 0.5)) +
  scale_fill_discrete(name="Sample Source")+
  theme(axis.text.x = element_text(angle = 30, hjust = 1))

# compare detection p values
RGset <- read_rds("~/sets/set3/RGset.rds")
detecP = detectionP(RGset)
# write_rds(detecP,"detecP.set3.rds")
# detecP = read_rds("detecP.set3.rds")

detecPDF = data.frame(detecPMean=colMeans(detecP)) %>%
  rownames_to_column(var="beadPosition")
rm(detecP)

clinical3 = fread("~/sets/set3/clinical_variables.csv") %>%
  filter(Trisomy21==1) %>%
  dplyr::select(beadPosition,Sample_Source) %>%
  inner_join(detecPDF,by="beadPosition") %>%
  inner_join(ds.all.final.subs.AfterReview.txt,by="beadPosition")

clinical3 %>%
  mutate(log_P = -log(detecPMean))%>%
  group_by(Sample_Source) %>%
  summarize(mean = mean(log_P), 
            n = n()) %>%
  as.data.frame()
#        Sample_Source     mean   n
# 1        Cal_Biobank 8.875319 196
# 2               CCLS 8.689142  18
# 3              CCRLP 8.941013 108
# 4         CCRLP_CTRL 9.086503   2
# 5 Washington Biobank 8.747028  18

ggplot(clinical3, aes(x=Sample_Source, y=-log(detecPMean), fill=as.factor(Sample_Source))) +
  geom_boxplot(outlier.size=0.2)+
  theme_classic() +
  xlab("Sample Source") +
  ylab("Detection P value") +
  theme(plot.title = element_text(hjust = 0.5)) +
  scale_fill_discrete(name="Sample Source")+
  theme(axis.text.x = element_text(angle = 30, hjust = 1))

```

> Conducting DS_ALL EWAS in USC dataset (By Ivo)

final imputation
```{r}
# > getwd()
# [1] "/project/wiemels_260/sebastian/down_syndrome/methylation_data"

# #test with knn https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6474284/
library(tidyverse)
setwd("/home/rcf-proj3/ism1/down_methylation/methylation_data")
load("final_beta_s3_934.Rdata")
ses_betas <- as.data.frame(ses_betas)
ses_betas_t <- as.data.frame(t(ses_betas))
#remove cpgs with only missingness
ses_betas_t <- ses_betas_t[!map_lgl(ses_betas_t, ~ all(is.na(.)))]
#107249 cols with just missingness

ses_betas_t_filtered <- ses_betas_t[ -which(rowMeans(is.na(ses_betas_t)) > 0.05),]
ses_betas_t_filtered <- ses_betas_t_filtered[, -which(colMeans(is.na(ses_betas_t_filtered)) > 0.05)]
ses_betas_t_filtered <- t(ses_betas_t_filtered)
library(impute)
ses_betas_t_filtered <- impute.knn(ses_betas_t_filtered)
#ses_betas_t_filtered <- data.frame(t(ses_betas_t_filtered))

ses_betas_t_filtered_rds <- data.frame(t(ses_betas_t_filtered$data))

save(ses_betas_t_filtered_rds, file="ses_betas_t_filtered_no_missing_impute_KNN_glint.rDATA")

write_rds(ses_betas_t_filtered_rds, "ses_betas_t_filtered_no_missing_impute_KNN_glint.rds")
```

```{r}
# > getwd()
# [1] "/project/wiemels_260/sebastian/down_syndrome/methylation_data"

#create phenofile for ALL methylation
library(tidyverse)
#load data
library(readxl)
setwd("/home/rcf-proj3/ism1/down_methylation/methylation_data")
base <- read_xlsx("base3.xlsx")%>%
  dplyr::select(arrayPos, sampleName,beadchip, plate,sex,periDsal, caco )%>%
  dplyr::mutate(ID=as.numeric(sampleName), ID_array=arrayPos) 

setwd("/home/rcf-proj3/ism1/down_methylation/methylation_data")
base2 <- read_xlsx("base.xlsx")%>%
  dplyr::select(Sample_Plate, array_pos, Sample_Name, Chip, Batch)%>%
  dplyr::mutate(ID=as.numeric(Sample_Name), arrayPos = array_pos)%>%
  dplyr::select(-array_pos, -ID)

#ethnicity down
#setwd("/staging/ism1/muskens")
down_ethnicity <- read_xlsx("DS-ALL_Study_Covariates.xlsx")%>%
    dplyr::mutate(Ethnicity_DS=ifelse(Ethnicity=="Mixed race (Asian/White)"|Ethnicity=="Gujarati","Asian",ifelse(Ethnicity=="Mixed race (Black/Hisp)"|Ethnicity=="Mixed race (Black/White)","Black",ifelse(Ethnicity=="Mixed race (Hispanic/Asian)"|Ethnicity=="Mixed race (Hispanic/White)","Hispanic",ifelse(is.na(Ethnicity),"Other",Ethnicity)))))%>%
  dplyr::mutate(Ethnicity_DS=ifelse(is.na(Ethnicity_DS),"Other",Ethnicity_DS))%>%
 # dplyr::mutate(Ethnicity_DS=factor(Ethnicity, levels=unique(Ethnicity)))%>%
  dplyr::mutate(Blindid = `DS-AL Study ID`)%>%
  dplyr::select(Blindid, Ethnicity_DS)


#ethnicity for perinatal
#setwd("/staging/ism1/muskens")
ethnicity_perinatal <- read_csv("perinatal_race_19Jul2019.csv")%>%
  dplyr::mutate(Ethnicity_perinatal=ifelse(Y_race==1,"White", ifelse(Y_race==2,"Black",ifelse(Y_race==3,"Hispanic",ifelse(Y_race==4,"Asian",ifelse(Y_race==5,"Other","Other"))))))%>%
  dplyr::select(Blindid, Ethnicity_perinatal)

setwd("/home/rcf-proj3/ism1/down_methylation/methylation_data")
betas <- read_rds("ses_betas_t_filtered_no_missing_impute_KNN_glint.rds")

base <- base[base$arrayPos%in%row.names(betas),]
#removal of mislabeled ids
base <- base%>%
  dplyr::filter(arrayPos!="202148010085_R03C01"&arrayPos!="202148010040_R08C01"&arrayPos!="202148010121_R08C01"&arrayPos!="202163110001_R01C01"&arrayPos!="202163110045_R01C01"&arrayPos!="202240580156_R06C01")

#901 left 

#merging
step1 <- base%>%
  dplyr::mutate(sampleName=as.numeric(sampleName))%>%
  left_join(base2, by ="arrayPos")%>%
  #left_join(ALL, by ="ID")%>%
  left_join(down_ethnicity, by =c("sampleName"="Blindid"))%>%
  left_join(ethnicity_perinatal, by =c("sampleName"="Blindid"))%>%
  dplyr::mutate(Ethnicity=ifelse(is.na(Ethnicity_DS), Ethnicity_perinatal,Ethnicity_DS))%>%
  dplyr::mutate(Ethnicity=ifelse(is.na(Ethnicity),"Other",Ethnicity))%>%
  dplyr::mutate(female= ifelse(sex=="f",1,0), down_syndrome= ifelse(periDsal=="dsal",1,0)
                ,all_case=ifelse(caco=="case",1,0), ID=arrayPos, Blindid=sampleName, all_case=factor(all_case))

step1 <- step1%>%
  dplyr::select(ID, Blindid, plate, female, Ethnicity, down_syndrome, all_case)%>%
  mutate(plate = as.numeric(as.factor(plate)))%>%
  arrange(desc(ID))%>%
  write_rds("covariate_file_all_analysis_08_08.rds")

betas4 <- read_rds("betas_total_analysis_08_08.rds")%>%
  as.data.frame()
outcomes <- read_rds("covariate_file_all_analysis_08_08.rds")%>%
  dplyr::filter(down_syndrome==1)

#refractor
refractor <- read_delim("set_3_down_all_rf_08_08.refactor.components.txt", col_names=T, delim=" ")

#epistructure
epistructure <- read_tsv("set_3_down_all_rf_epi_08_08.epistructure.pcs.txt", col_names=F)%>%
  dplyr::mutate(ID=X1, epi1=X2, epi2=X3, epi3=X4, epi4=X5, epi5=X6,epi6=X7,epi7=X8,epi8=X9,epi9=X10,epi10=X11)%>%
 dplyr::select( -X1, -X2,-X3,-X4,-X5,-X6,-X7,-X8,-X9,-X10, -X11)

#combine and write output
outcomes <- outcomes%>%
  left_join(refractor, by="ID") %>%
  left_join(epistructure, by="ID")%>%
  dplyr::mutate(female=factor(female))%>%
  write_rds("covariates_down_ALL_08_08.rds")

#write new betas file 
betas <- betas[row.names(betas)%in%step1$ID,]
betas <- betas%>%
  rownames_to_column(var="ID")%>%
  arrange(desc(ID))%>%
  column_to_rownames(var="ID")
write_rds(betas, "betas_total_analysis_08_08.rds")

```

```{r DS_ALL EWAS}
# > getwd()
# [1] "/project/wiemels_260/sebastian/down_syndrome/methylation_data"

setwd("/scratch/lishaobo/downSyndrome")

setwd("/home/rcf-proj3/ism1/down_methylation/methylation_data")
.libPaths( c( .libPaths(), "/auto/rcf-proj3/ism/R_packages") )
library(rstudioapi,lib.loc = "/auto/rcf-proj3/ism/R_packages")
library(rlang,lib.loc = "/auto/rcf-proj3/ism/R_packages")
library(withr, lib.loc = "/auto/rcf-proj3/ism/R_packages")
library(tidyverse,lib.loc = "/auto/rcf-proj3/ism/R_packages" )
library(DMRcate,lib.loc = "/auto/rcf-proj3/ism/R_packages" )

#load the betas (samples are rows and variables are betas)
betas <- read_rds("betas_total_analysis_08_08.rds")%>%
  as.data.frame()

#load covariates including Refactor, epistructure, sex and plate 
outcomes <- read_rds("covariates_down_ALL_08_08.rds")

#filtering step to make sure the betas and outcome file contain the same samples
#also make sure that the betas and outcome file have the same order of smaples
betas <- betas[row.names(betas)%in%outcomes$ID,]

#specific manipulation for out dataset (can ignore)
outcomes$plate <- gsub(" ","",outcomes$plate)

#make Mvalues (with adequate transposing)
Mvalues_prepped <-betas%>%
  t()%>%
  logit2()%>% #mvalue function from DMRcate
  t()%>%
  as.data.frame()

#make sure colums are finite (otherwise errors are introduced)
Mvalues_prepped <- Mvalues_prepped[,is.finite(colSums(Mvalues_prepped))]

#wrapper function to produce convenient output 
beta_pval <- function(x){
  data.frame(estimate=summary(x)$coefficients[2,1],std_error=summary(x)$coefficients[2,2],t_stat=summary(x)$coefficients[2,3],p_val=summary(x)$coefficients[2,4])
}

#write Mvalues (may be usefull when repeating analyses)
setwd("/staging/ism/muskens/methylation/slurmr")
write_rds(Mvalues_prepped, "Mvalues_prepped_ALL_DS.rds")
Mvalues_prepped<- read_rds("Mvalues_prepped_ALL_DS.rds") 

#load slurm
library(slurmR)
#set directory for slurmr
opts_slurmR$set_tmp_path("/staging/ism/muskens/methylation/slurmr")
#set name
opts_slurmR$set_job_name("ALL_DS")
#choose parition (not necessary)
opts_slurmR$set_opts(partition="scavenge")
#set other slurm options
options <- list("time" = "01:00:00", "mem"= "20G")
#set regressions
regressors<-setdiff(names(Mvalues_prepped),"Sscore")

#submit slurmr with 10 rc 10 epi
ALL_DS_output <- Slurm_lapply(regressors,function(cpg){cbind(cpg_name=cpg,beta_pval(lm(Mvalues_prepped[,cpg]~outcomes$all_case+outcomes$female
     +outcomes$plate+outcomes$rc1+outcomes$rc2+outcomes$rc3
     +outcomes$rc4+outcomes$rc5+outcomes$rc6+outcomes$rc7
     +outcomes$rc8+outcomes$rc9+outcomes$rc10+outcomes$epi1
     +outcomes$epi2+outcomes$epi3+outcomes$epi4+outcomes$epi5
     +outcomes$epi6+outcomes$epi7+outcomes$epi8+outcomes$epi9
     +outcomes$epi10 )))}, njobs=100, plan="submit", overwrite=T, compress = F,sbatch_opt = options,export=c("Mvalues_prepped", "outcomes","regressors", "beta_pval"),libPaths= .libPaths( c( .libPaths(), "/auto/rcf-proj3/ism/R_packages") ))

#not slurmr
# Mvalues_prepped_test <- Mvalues_prepped[,1:100]
# regressors<-setdiff(names(Mvalues_prepped_test),"Sscore")
# ALL_DS_output_test <- lapply(regressors,function(cpg){cbind(cpg_name=cpg,beta_pval(lm(Mvalues_prepped_test[,cpg]~outcomes$all_case+outcomes$female
#      +outcomes$plate+outcomes$rc1+outcomes$rc2+outcomes$rc3
#      +outcomes$rc4+outcomes$rc5+outcomes$rc6+outcomes$rc7
#      +outcomes$rc8+outcomes$rc9+outcomes$rc10+outcomes$epi1
#      +outcomes$epi2+outcomes$epi3+outcomes$epi4+outcomes$epi5
#      +outcomes$epi6+outcomes$epi7+outcomes$epi8+outcomes$epi9
#      +outcomes$epi10)))})

#load annotation files
library(IlluminaHumanMethylationEPICanno.ilm10b4.hg19)
data("IlluminaHumanMethylationEPICanno.ilm10b4.hg19")
anno       <- IlluminaHumanMethylationEPICanno.ilm10b4.hg19 %>%
              getAnnotation %>%
              as.data.frame%>%
              rownames_to_column(var="cpg")

#Loadback and annotate
ewas_DS_ALL <- Slurm_collect(ALL_DS_output)%>%
  bind_rows%>%
  left_join(anno, by=c("cpg_name"="cpg"))%>%
  arrange(p_val)

#save data an check lambda
write_rds(ewas_DS_ALL, "EWAS_ALL_DS_Linear_model_10rc_10epi.rds")
library(QCEWAS)
P_lambda(ewas_DS_ALL$p_val)

#cleanup slurmr
Slurm_clean(ALL_DS_output)



################################################
# a list of subjects included in the following study
setwd("/project/wiemels_260/sebastian/down_syndrome/methylation_data")
library(tidyverse)
library(data.table)

outcomes <- read_rds("covariates_down_ALL_08_08.rds") %>% as.data.frame
table(outcomes$all_case)
#   0   1 
# 196 127 = 323

clinical3 <- read.csv('~/sets/set3/clinical_variables.csv') %>%
  filter(Trisomy21==1) 
# > table(clinical3$CaCo)
#   0   1 
# 198 147 =  345

intersect(clinical3$beadPosition,outcomes$ID) %>% length()
# [1] 319

a = which(!clinical3$beadPosition%in%outcomes$ID)
clinical3[a,c("beadPosition","subjectId","Trisomy21","CaCo","Sample_Source")]

b = which(!outcomes$ID%in%clinical3$beadPosition)
outcomes[b,c("ID","Blindid","down_syndrome","all_case")]
# 99  202226320174_R08C01      NA             1        1
# 126 202172220181_R06C01      NA             1        1
# 226 202163110003_R06C01      NA             1        1
# 240 202148010158_R07C01      NA             1        1

## going back to base and find their original sample IDs
# 202226320174_R08C01   198075-2 202226320174   11B   f     dsal    case NA
# 202172220181_R06C01   198110-2 202172220181    8B   f     dsal    case NA
# 202163110003_R06C01   198161-2 202163110003    4A   f     dsal    case NA
# 202148010158_R07C01   198479-2 202148010158    6B   m     dsal    case NA

clinical3 %>% filter(Trisomy21==1) %>% filter(subjectId %in% c(198075,198110,198161,198479))
#        beadPosition   Sample_Plate Sample_Well subjectId
# 202148010158_R04C01 EPIC25_Plate06         D08    198479
# 202163110003_R02C01 EPIC22_Plate04         B02    198161
# 202172220181_R05C01  EPIC27_Plate8         E11    198110
# 202226320174_R06C01 EPIC29_Plate10         F12    198075

clinical3[which(clinical3$beadPosition=="202148010158_R04C01"),"beadPosition"] <- "202148010158_R07C01"
clinical3[which(clinical3$beadPosition=="202163110003_R02C01"),"beadPosition"] <- "202163110003_R06C01"
clinical3[which(clinical3$beadPosition=="202172220181_R05C01"),"beadPosition"] <- "202172220181_R06C01"
clinical3[which(clinical3$beadPosition=="202226320174_R06C01"),"beadPosition"] <- "202226320174_R08C01"

intersect(clinical3$beadPosition,outcomes$ID) %>% length()

which(clinical3$beadPosition=="202148010158_R04C01")
which(clinical3$beadPosition=="202163110003_R02C01")
which(clinical3$beadPosition=="202172220181_R05C01")
which(clinical3$beadPosition=="202226320174_R06C01")
which(clinical3$beadPosition=="202148010158_R07C01")
which(clinical3$beadPosition=="202163110003_R06C01")
which(clinical3$beadPosition=="202172220181_R06C01")
which(clinical3$beadPosition=="202226320174_R08C01")

outcomes[which(!outcomes$ID%in%clinical3$beadPosition),
         c("ID","Blindid","down_syndrome","all_case")]
# 101 202226320174_R06C01      NA             1        1
# 127 202172220181_R05C01      NA             1        1
# 229 202163110003_R02C01      NA             1        1
# 242 202148010158_R04C01      NA             1        1



ids = c(
"202148010158_R04C01",
"202163110003_R02C01",
"202172220181_R05C01",
"202226320174_R06C01",
"202148010158_R07C01",
"202163110003_R06C01",
"202172220181_R06C01",
"202226320174_R08C01")

outcomes %>% filter(ID %in%ids)

base <- read_xlsx("base3.xlsx")%>%
  dplyr::select(arrayPos, sampleName,beadchip, plate,sex,periDsal, caco )%>%
  dplyr::mutate(ID=as.numeric(sampleName), ID_array=arrayPos) %>% 
  filter(ID_array %in%ids)

outcomes %>% filter(ID %in%ids)

### 





```

> Conducting DS_ALL EWAS in USC dataset (By Sebastian)

```{r}

setwd("~/DSALL/usc")

library(tidyverse)
library(data.table)
library(MASS)
library(lmtest)
library(sandwich)
library(parallel) 
library(minfi)
library(wateRmelon)

cl <- makeCluster(40, type="FORK")
clusterEvalQ(cl,library(MASS))
clusterEvalQ(cl,library(lmtest))

RLM.Robust <- function(meth, cov){

    # meth = methyl["cg25813447",]
    # cov = covs
  
    dat <- data.frame(y = meth, cov)
    ff  <- formula(paste("y ~ ", paste0(colnames(cov), collapse="+")))
    mod <- try(rlm(ff, data = dat, maxit = 200))
    #cf = try(coeftest(mod, vcov=vcovHC(mod, type="HC0")))
    cf = try(coeftest(mod))

  	if (exists(class(cf)) == FALSE) {
  		out = rep(NA,3)
  	} else if (exists(class(cf)) == TRUE) {
  		if( class(cf)[1] == "try-error") {
  		cat("error throw by CpG", set, ";")
  		out = rep(NA,3)
  		print(paste0(cov, " - chol2inv could not be computed"))
  		} 
  		else {
  			coef = cf[2,"Estimate"]
  			se = cf[2,"Std. Error"]
  			p.value = cf[2,"Pr(>|z|)"]
  			out = c(coef, se, p.value) 
  		}
    }
  names(out) = c("Beta","SE", "P"); return(out) 
  }

# Winsorize beta values
winsorize <- function(methylation,pct=winsorize.pct) {
  quantiles <- matrixStats::rowQuantiles(methylation, probs=c(pct,1-pct), na.rm=T)
  low <- quantiles[,1]
  upper <- quantiles[,2]

  outliers.lower <- rowSums(methylation < low, na.rm=T)
  outliers.upper <- rowSums(methylation > upper, na.rm=T)
  
  idx <- which(methylation < low, arr.ind=T)
  methylation[idx] <- low[idx[,1]]
  
  idx <- which(methylation > upper, arr.ind=T)
  methylation[idx] <- upper[idx[,1]]

  n <- rowSums(!is.na(methylation))
  log <- data.frame(outliers.lower, outliers.upper, n)
  return(list(methylation=methylation, log=log))
}

# prepare a file of subjects (called var) with deconvoluted cell proportions, sex and plate, 10 EpiStructure PCs

dcp = fread("~/sets/set3/idol_deconvolution.csv") %>%
  as.data.frame() %>%
  dplyr::select(-V1,-subjectId) %>%
  na.omit()

glint = fread("~/sets/set3/glintControl.csv") %>%
    dplyr::select(beadPosition, epi1, epi2, epi3, epi4, epi5, epi6, epi7, epi8, epi9, epi10)%>%
    na.omit()

vars = fread("~/sets/set3/clinical_variables.csv") %>%
  as.data.frame() %>%
  dplyr::filter(Trisomy21==1) %>%
  dplyr::select(beadPosition,subjectId,sex,plate,CaCo)%>%
  inner_join(dcp,by="beadPosition")%>%
  inner_join(glint,by="beadPosition")%>%
  na.omit() 

table(vars$CaCo)
# write.table(vars,"ds.all.covaraites.txt",
#             quote=FALSE,
#             sep='\t',
#             row.names = FALSE)

subsInTheStudy = vars[,c("beadPosition","subjectId")]
# write.table(subsInTheStudy,"ds.all.final.subs.txt",
#             quote=FALSE,
#             sep='\t',
#             row.names = FALSE)

# vars = fread("ds.all.covaraites.txt") %>% as.data.frame()

## the head of the vars file is like this.
## var cannot have any missing values. make sure to do na.omit to drop any subjects that have missing values.

#          beadPosition subjectId sex plate CaCo       CD8T       CD4T
# 1 202053820102_R01C01    198106   1    31    1 0.05803930 0.07774256
# 2 202053820102_R06C01    198048   2    31    0 0.05974364 0.07228877
# 3 202053820102_R08C01    198398   1    31    1 0.05008209 0.11407259
#           NK      Bcell       Mono      Gran nRBC      epi1        epi2
# 1 0.02917945 0.01115651 0.09605564 0.7143988    0 -14.17960  -0.8876962
# 2 0.01991806 0.03983181 0.03068947 0.7787478    0 -17.72907 -10.0481424
# 3 0.02307635 0.00000000 0.05588111 0.7591911    0 -13.54827  -2.4054445
#        epi3      epi4      epi5       epi6      epi7      epi8      epi9
# 1 -11.61660  2.355388 -5.751580 -1.9882530  1.726514 -1.484965  1.958523
# 2  13.47025 -6.027756  1.538562 -0.9044501 -3.496759 -1.146718  6.852522
# 3 -14.95999 -5.870502 -8.209416  4.3085533 -4.059880  2.030065 11.172533
#        epi10
# 1 -0.1103879
# 2  5.5101328
# 3  2.2978038

# prepare a methylaiton file for subjects included in "var" (called methy)

################
##scenario using Noob
## if you already have a NOOB normalized beta values of methylation file, then feel free to skip the following steps. Otherwise here's how to create it from RGSet files. 
## RGSet file is created from idat files and i think you already created it before. Let me know if you need help creating it.
RGset <- read_rds("~/sets/set3/RGset.rds")
noobMethylSet <- preprocessNoob(RGset)
mSetBmiq <- BMIQ(noobMethylSet)
write_rds(mSetBmiq,"~/sets/set3/beta_noob_set3.rds")

noob = read_rds("~/sets/set3/beta_noob_set3.rds") 
noob.to.include = noob[,vars$beadPosition]
#            202053820102_R01C01 202053820102_R06C01 202053820102_R08C01
# cg18478105          0.01669368          0.01215185          0.01185778
# cg09835024          0.03749958          0.16102342          0.02382422
# cg14361672          0.92392927          0.92952634          0.95703239
# cg01763666          0.78974682          0.74906463          0.73299074
# cg12950382          0.85329832          0.91411088          0.84463116
# cg02115394          0.08789014          0.03922020          0.04410196
#            202053820141_R01C01 202053820141_R07C01 202060330085_R01C01
# cg18478105          0.01603176          0.01317428          0.01322269
# cg09835024          0.03035469          0.02579181          0.02902467
# cg14361672          0.92784982          0.93663449          0.88604567
# cg01763666          0.70243675          0.72674590          0.47341155
# cg12950382          0.91845575          0.92160532          0.90096988
# cg02115394          0.06333653          0.11623978          0.04294634


################
##scenario using Sesame
betas3_0 = fread("~/sets/set3/set3_Sesame_beta.csv") %>%
  as.data.frame()%>%
  dplyr::select(-V1)
betas3_1 = betas3_0 %>%
  column_to_rownames(var="probeId")
noob.to.include = betas3_1[,vars$beadPosition] %>% as.matrix

replace.outliers <- winsorize(noob.to.include, 0.005)
methyl <- replace.outliers$methylation
outlier.log <- replace.outliers$log

rownames(vars) <- c()
covs = vars %>% 
  dplyr::select(beadPosition,CaCo,sex,plate,CD8T,CD4T,NK,Bcell,Mono,nRBC,epi1,epi2,epi3,epi4,epi5,epi6,
                epi7,epi8,epi9,epi10)%>%
  mutate(plate=as.factor(plate)) %>%
  column_to_rownames(var="beadPosition")

# this has to be true to continue

if(identical(rownames(covs),colnames(methyl))==TRUE){

  # methylt = methyl[1:100,] 
  # t(apply(methylt,1,RLM.Robust,covs))
  
  results <- t(parApply(cl, methyl, 1, RLM.Robust, covs)) 
  results <- as.data.frame(results)
  
  # write.table(results,"ds.all.usc.ewas.txt",
  #         quote=FALSE,
  #         sep='\t')  
  
  write.table(results,"ds.all.usc.sesame.ewas.txt",
        quote=FALSE,
        sep='\t')  

  
  Nbeta <- sum(!is.na(results$P))
  tmp <- results[!is.na(results$P),]
  
  lambda = median(qchisq(1- tmp$P,1))/qchisq(0.5,1)
  # it will be great if you can also share with me your lambda values
  print(round(lambda, 2))

}

stopCluster(cl)

```

```{r EWAS of probes on Chromosome X}

setwd("~/DSALL/usc")

library(tidyverse)
library(data.table)
library(MASS)
library(lmtest)
library(sandwich)
library(parallel) 
library(minfi)
library(wateRmelon)

cl <- makeCluster(40, type="FORK")
clusterEvalQ(cl,library(MASS))
clusterEvalQ(cl,library(lmtest))

winsorize <- function(methylation,pct=winsorize.pct) {
  quantiles <- matrixStats::rowQuantiles(methylation, probs=c(pct,1-pct), na.rm=T)
  low <- quantiles[,1]
  upper <- quantiles[,2]

  outliers.lower <- rowSums(methylation < low, na.rm=T)
  outliers.upper <- rowSums(methylation > upper, na.rm=T)
  
  idx <- which(methylation < low, arr.ind=T)
  methylation[idx] <- low[idx[,1]]
  
  idx <- which(methylation > upper, arr.ind=T)
  methylation[idx] <- upper[idx[,1]]

  n <- rowSums(!is.na(methylation))
  log <- data.frame(outliers.lower, outliers.upper, n)
  return(list(methylation=methylation, log=log))
}

RLM.Robust <- function(meth, cov){

    # meth = methyl["cg25813447",]
    # cov = covs
  
    dat <- data.frame(y = meth, cov)
    ff  <- formula(paste("y ~ ", paste0(colnames(cov), collapse="+")))
    mod <- try(rlm(ff, data = dat, maxit = 200))
    #cf = try(coeftest(mod, vcov=vcovHC(mod, type="HC0")))
    cf = try(coeftest(mod))

  	if (exists(class(cf)) == FALSE) {
  		out = rep(NA,3)
  	} else if (exists(class(cf)) == TRUE) {
  		if( class(cf)[1] == "try-error") {
  		cat("error throw by CpG", set, ";")
  		out = rep(NA,3)
  		print(paste0(cov, " - chol2inv could not be computed"))
  		} 
  		else {
  			coef = cf[2,"Estimate"]
  			se = cf[2,"Std. Error"]
  			p.value = cf[2,"Pr(>|z|)"]
  			out = c(coef, se, p.value) 
  		}
    }
  names(out) = c("Beta","SE", "P"); return(out) 
  
  }

linear.conv <- function(meth, cov){

    # meth = methyl["cg25813447",]
    # cov = covs
  
    dat <- data.frame(y = meth, cov)
    ff  <- formula(paste("y ~ ", paste0(colnames(cov), collapse="+")))
    mod <- try(lm(ff, data = dat),silent = TRUE)

    
  if (class(mod)=="lm") {
    
    coefs = summary(mod)$coefficients
    coef = coefs[2,"Estimate"]
    SE = coefs[2,"Std. Error"]
    p.value = coefs[2,"Pr(>|t|)"]
    out = c(coef, SE, p.value) 
  } else {
    out = c(rep(NA,3))
  }
  names(out) = c("Beta","SE", "P"); return(out) 
}

    
library(IlluminaHumanMethylationEPICanno.ilm10b4.hg19)
data("IlluminaHumanMethylationEPICanno.ilm10b4.hg19")
annoEPIC <- IlluminaHumanMethylationEPICanno.ilm10b4.hg19 %>%
  getAnnotation %>%
  as.data.frame %>%
  rownames_to_column(var="cpg") %>%
  dplyr::select(cpg,chr,pos,Relation_to_Island,UCSC_RefGene_Name,
                Probe_maf,UCSC_RefGene_Group) %>%
  filter(chr=="chrX")

betas3_0 = fread("~/sets/set3/set3_Sesame_beta.csv") %>%
  as.data.frame()%>%
  dplyr::select(-V1)
betas3_1 = betas3_0 %>%
  filter(probeId%in%annoEPIC$cpg)%>%
  column_to_rownames(var="probeId")

vars = fread("ds.all.covaraites.txt") 
vars.male = vars %>% filter(sex==1)
vars.female = vars %>% filter(sex==2)

noob.to.include.male = betas3_1[,vars.male$beadPosition] %>% as.matrix
noob.to.include.female = betas3_1[,vars.female$beadPosition] %>% as.matrix

replace.outliers.male <- winsorize(noob.to.include.male, 0.005)
methyl.male <- replace.outliers.male$methylation
outlier.log.male <- replace.outliers.male$log

replace.outliers.female <- winsorize(noob.to.include.female, 0.005)
methyl.female <- replace.outliers.female$methylation
outlier.log.female <- replace.outliers.female$log

rownames(vars.male) <- c()
rownames(vars.female) <- c()

covs.male = vars.male %>% 
  dplyr::select(beadPosition,CaCo,plate,CD8T,CD4T,NK,Bcell,Mono,nRBC,epi1,epi2,epi3,epi4,epi5,epi6,
                epi7,epi8,epi9,epi10)%>%
  mutate(plate=as.factor(plate)) %>%
  column_to_rownames(var="beadPosition")

covs.female = vars.female %>% 
  dplyr::select(beadPosition,CaCo,plate,CD8T,CD4T,NK,Bcell,Mono,nRBC,epi1,epi2,epi3,epi4,epi5,epi6,
                epi7,epi8,epi9,epi10)%>%
  mutate(plate=as.factor(plate)) %>%
  column_to_rownames(var="beadPosition")

# this has to be true to continue

if(identical(rownames(covs.male),colnames(methyl.male))==TRUE){

  # methylt = methyl.male[1:100,] 
  # t(apply(methylt,1,RLM.Robust,covs.male))
  # t(apply(methylt,1,linear.conv,covs.male))
  
  results.male <- t(parApply(cl, methyl.male, 1, RLM.Robust, covs.male)) 
  results.male.lm <- t(parApply(cl, methyl.male, 1, linear.conv, covs.male)) 

  #results.male  <- t(apply(methyl.male,1,RLM.Robust,covs.male))

  results.male <- as.data.frame(results.male)
  results.male.lm <- as.data.frame(results.male.lm)
  
  write.table(results.male,"ds.male.X.usc.sesame.ewas.txt",
        quote=FALSE,
        sep='\t')  

  write.table(results.male.lm,"ds.lm.male.X.usc.sesame.ewas.txt",
      quote=FALSE,
      sep='\t')  

  
  Nbeta.male <- sum(!is.na(results.male$P))
  tmp.male <- results.male[!is.na(results.male$P),]
  
  lambda.male = median(qchisq(1- tmp.male$P,1))/qchisq(0.5,1)
  # it will be great if you can also share with me your lambda values
  print(round(lambda.male, 2))
  # [1] 3.24
}

if(identical(rownames(covs.female),colnames(methyl.female))==TRUE){

  # methylt = methyl.female[1:100,] 
  # t(apply(methylt,1,RLM.Robust,covs.female))
  
  results.female <- t(parApply(cl, methyl.female, 1, RLM.Robust, covs.female)) 
  results.female.lm <- t(parApply(cl, methyl.female, 1, linear.conv, covs.female)) 

  #results.female  <- t(apply(methyl.female,1,RLM.Robust,covs.female))

  results.female <- as.data.frame(results.female)
  results.female.lm <- as.data.frame(results.female.lm)

  write.table(results.female,"ds.female.X.usc.sesame.ewas.txt",
        quote=FALSE,
        sep='\t')  

  write.table(results.female.lm,"ds.lm.female.X.usc.sesame.ewas.txt",
        quote=FALSE,
        sep='\t')  
  
  
  Nbeta.female <- sum(!is.na(results.female$P))
  tmp.female <- results.female[!is.na(results.female$P),]
  
  lambda.female = median(qchisq(1- tmp.female$P,1))/qchisq(0.5,1)
  # it will be great if you can also share with me your lambda values
  print(round(lambda.female, 2))
  # [1] 2.36
}


stopCluster(cl)



```

```{r a sneak peak into usc results}

setwd("~/DSALL/usc")

library(tidyverse)
library(data.table)
library(IlluminaHumanMethylationEPICanno.ilm10b4.hg19)

data("IlluminaHumanMethylationEPICanno.ilm10b4.hg19")
annoEPIC <- IlluminaHumanMethylationEPICanno.ilm10b4.hg19 %>%
  getAnnotation %>%
  as.data.frame %>%
  rownames_to_column(var="cpg") %>%
  dplyr::select(cpg,chr,pos,Relation_to_Island,UCSC_RefGene_Name,
                Probe_maf,UCSC_RefGene_Group)

############################################ 
## sesame, lambda=1.38
#### correct method for paper probe number reporting
results <- fread("ds.all.usc.sesame.ewas.txt") %>% # 865918
  dplyr::rename("cpg"="V1") %>%
  filter(!is.na(P))%>% # 691841
  left_join(annoEPIC,by="cpg") %>%
  arrange(by=P) %>%
  mutate(fdr=p.adjust(P, method = "fdr", n=length(P))) %>%
  mutate(bf=p.adjust(P, method = "bonferroni", n=length(P))) %>%  
  dplyr::filter(grepl('cg', cpg)) %>% # 690244
  dplyr::filter(Probe_maf<0.05|is.na(Probe_maf)) %>% # 632480
  dplyr::filter(chr!="chrX"&chr!="chrY") # 619915
                

write.table(results,"ds.all.usc.sesame.ewas.anno.txt",
        quote=FALSE,
        row.names = FALSE,
        sep='\t')  

# 8.0656E-08
dim(results)
dim(results%>%filter(fdr<0.05))
dim(results%>%filter(bf<0.05))

sigs.usc = results%>%
  filter(bf<0.05) %>%
  dplyr::select(cpg,Beta,SE,P,chr,pos,Relation_to_Island,UCSC_RefGene_Name,fdr,bf)
dim(sigs.usc)

write.csv(sigs.usc,"sigs.usc.cpgs.csv")
            
write.table(sigs.usc,"~/DSALL/sigs.usc.cpgs.txt",
            quote=FALSE,
            row.names = FALSE,
            sep='\t')

Nbeta <- sum(!is.na(results$P))
tmp <- results[!is.na(results$P),]
lambda = median(qchisq(1- tmp$P,1))/qchisq(0.5,1)
# it will be great if you can also share with me your lambda values
print(round(lambda, 2))
# [1] 1.38

# expression database
a1 = fread("keren/FHS.TableS6_cis_meta.csv") %>% 
  mutate(file="a1")%>%
  dplyr::select(CpG,file)
a2 = fread("keren/FHS.TableS3_trans_FHS.csv") %>% 
  mutate(file="a2")%>%
  dplyr::select(CpG,file)
a3 = fread("keren/FHS.TableS2_cis_FHS.csv") %>% 
  mutate(file="a3")%>%
  dplyr::select(CpG,file)
a4 = fread("keren/FUSIONSkeletalMuscleStudy.eqtm-1mb-5expr_5dname_peer.pv.le0.05.csv") %>%
  mutate(CpG = x_id) %>%
  mutate(file="a4")%>%
  dplyr::select(CpG,file)
a5 = fread("keren/Hannon.mmc2.methExp.csv")%>%
  mutate(CpG = ProbeID_met) %>%
  mutate(file="a5")%>%
  dplyr::select(CpG,file)

cpgs_all = rbind(a1,a2,a3,a4,a5) %>% 
  as.data.frame()

cpgs_all[which(cpgs_all$CpG %in% sigs.usc$cpg),]

############################################ 
## sesame, X chromosome, robust linear regression (rlm)
#### correct method for paper probe number reporting
results.male <- fread("ds.male.X.usc.sesame.ewas.txt") %>%
  dplyr::rename("cpg"="V1") %>%
  filter(!is.na(P))%>%
  left_join(annoEPIC,by="cpg") %>%
  arrange(by=P) %>%
  mutate(fdr=p.adjust(P, method = "fdr", n=length(P))) %>%
  mutate(bf=p.adjust(P, method = "bonferroni", n=length(P))) %>%  
  dplyr::filter(grepl('cg', cpg)) %>%
  dplyr::filter(Probe_maf<0.05|is.na(Probe_maf))

write.table(results.male,"ds.male.X.usc.sesame.ewas.anno.txt",
        quote=FALSE,
        row.names = FALSE,
        sep='\t')  


results.female <- fread("ds.female.X.usc.sesame.ewas.txt") %>%
  dplyr::rename("cpg"="V1") %>%
  filter(!is.na(P))%>%
  left_join(annoEPIC,by="cpg") %>%
  arrange(by=P) %>%
  mutate(fdr=p.adjust(P, method = "fdr", n=length(P))) %>%
  mutate(bf=p.adjust(P, method = "bonferroni", n=length(P))) %>%  
  dplyr::filter(grepl('cg', cpg)) %>%
  dplyr::filter(Probe_maf<0.05|is.na(Probe_maf))

write.table(results.female,"ds.female.X.usc.sesame.ewas.anno.txt",
        quote=FALSE,
        row.names = FALSE,
        sep='\t')  

############################################ 
## sesame, X chromosome, linear lm
#### correct method for paper probe number reporting
results.male <- fread("ds.lm.male.X.usc.sesame.ewas.txt") %>%
  dplyr::rename("cpg"="V1") %>%
  filter(!is.na(P))%>%
  left_join(annoEPIC,by="cpg") %>%
  arrange(by=P) %>%
  mutate(fdr=p.adjust(P, method = "fdr", n=length(P))) %>%
  mutate(bf=p.adjust(P, method = "bonferroni", n=length(P))) %>%  
  dplyr::filter(grepl('cg', cpg)) %>%
  dplyr::filter(Probe_maf<0.05|is.na(Probe_maf))

write.table(results.male,"ds.lm.male.X.usc.sesame.ewas.anno.txt",
        quote=FALSE,
        row.names = FALSE,
        sep='\t')  


results.female <- fread("ds.lm.female.X.usc.sesame.ewas.txt") %>%
  dplyr::rename("cpg"="V1") %>%
  filter(!is.na(P))%>%
  left_join(annoEPIC,by="cpg") %>%
  arrange(by=P) %>%
  mutate(fdr=p.adjust(P, method = "fdr", n=length(P))) %>%
  mutate(bf=p.adjust(P, method = "bonferroni", n=length(P))) %>%  
  dplyr::filter(grepl('cg', cpg)) %>%
  dplyr::filter(Probe_maf<0.05|is.na(Probe_maf))

write.table(results.female,"ds.lm.female.X.usc.sesame.ewas.anno.txt",
        quote=FALSE,
        row.names = FALSE,
        sep='\t')  
############################################ 
#### without filtering for MAF and SNPs (somatic)
results <- fread("ds.all.usc.sesame.ewas.txt") %>%
  dplyr::rename("cpg"="V1") %>%
  left_join(annoEPIC,by="cpg") %>%
  arrange(by=P) %>%
  mutate(fdr=p.adjust(P, method = "fdr", n=length(P))) %>%
  mutate(bf=p.adjust(P, method = "bonferroni", n=length(P)))  
# write.table(results,"ds.all.usc.sesame.ewas.anno.txt",
#         quote=FALSE,
#         row.names = FALSE,
#         sep='\t')  
dim(results%>%filter(fdr<0.05))
dim(results%>%filter(bf<0.05))

# look for ARID5B
head(results)
arid5b <- results %>%
  filter(grepl("ARID5B",UCSC_RefGene_Name)) %>%
  arrange(by=fdr)

arid5b[arid5b$cpg=="cg13344587",]

#####
## noob, lambda=1.41
results <- fread("ds.all.usc.ewas.txt") %>%
  dplyr::rename("cpg"="V1") %>%
  left_join(annoEPIC,by="cpg") %>%
  arrange(by=P) %>%
  mutate(fdr=p.adjust(P, method = "fdr", n=length(P))) %>%
  mutate(bf=p.adjust(P, method = "bonferroni", n=length(P)))  

write.table(results,"ds.all.usc.ewas.anno.txt",
        quote=FALSE,
        sep='\t')  
  
dim(results%>%filter(fdr<0.05))
dim(results%>%filter(bf<0.05))
# look for ARID5B
head(results)
arid5b <- results %>%
  filter(grepl("ARID5B",UCSC_RefGene_Name)) %>%
  arrange(by=fdr)

arid5b[arid5b$cpg=="cg13344587",]


```

```{bash meta-analysis of males and females}

cd ~/DSALL/usc

metal
MARKER cpg
EFFECT Beta
PVALUE P
SCHEME STDERR
STDERR SE
VERBOSE OFF
GENOMICCONTROL OFF
PROCESS ds.male.X.usc.sesame.ewas.anno.txt
PROCESS ds.female.X.usc.sesame.ewas.anno.txt
ANALYZE HETEROGENEITY
QUIT
mv METAANALYSIS1.TBL usc_chrX_meta_sd.TBL

# results present in both studies
awk -v FS='\t' '$7 !~ /?/ {print $1,$4,$5,$6,$7,$8,$9,$10,$11}' usc_chrX_meta_sd.TBL | sort -k4,4n > usc_chrX_meta_sd_both.txt

## lineaar regression lm
metal
MARKER cpg
EFFECT Beta
PVALUE P
SCHEME STDERR
STDERR SE
VERBOSE OFF
GENOMICCONTROL OFF
PROCESS ds.lm.male.X.usc.sesame.ewas.anno.txt
PROCESS ds.lm.female.X.usc.sesame.ewas.anno.txt
ANALYZE HETEROGENEITY
QUIT
mv METAANALYSIS1.TBL usc_lm_chrX_meta_sd.TBL

# results present in both studies
awk -v FS='\t' '$7 !~ /?/ {print $1,$4,$5,$6,$7,$8,$9,$10,$11}' usc_lm_chrX_meta_sd.TBL | sort -k4,4n > usc_lm_chrX_meta_sd_both.txt

```

```{r annotation of ChrX meta-analysis}

setwd("~/DSALL/usc")

library(tidyverse)
library(data.table)
library(IlluminaHumanMethylationEPICanno.ilm10b4.hg19)

data("IlluminaHumanMethylationEPICanno.ilm10b4.hg19")
annoEPIC <- IlluminaHumanMethylationEPICanno.ilm10b4.hg19 %>%
  getAnnotation %>%
  as.data.frame %>%
  rownames_to_column(var="cpg") %>%
  dplyr::select(cpg,chr,pos,Relation_to_Island,UCSC_RefGene_Name,
                Probe_maf,UCSC_RefGene_Group)


results <- fread("usc_chrX_meta_sd.TBL") %>%
  dplyr::select(-Allele1,-Allele2)%>%
  dplyr::rename("cpg"="MarkerName",
                "P"="P-value") %>%
  left_join(annoEPIC,by="cpg") %>%
  dplyr::select(-Probe_maf)%>%
  arrange(by=P) %>%
  mutate(fdr=p.adjust(P, method = "fdr", n=length(P))) %>%
  mutate(bf=p.adjust(P, method = "bonferroni", n=length(P))) %>%  
  dplyr::filter(grepl('cg', cpg))

write.table(results,"ds.meta.X.usc.sesame.ewas.anno.txt",
        quote=FALSE,
        row.names = FALSE,
        sep='\t')  

write.csv(results,"meta.chrX.robustLinearRegression.csv")

# lm model
results <- fread("usc_lm_chrX_meta_sd.TBL") %>%
  dplyr::select(-Allele1,-Allele2)%>%
  dplyr::rename("cpg"="MarkerName",
                "P"="P-value") %>%
  left_join(annoEPIC,by="cpg") %>%
  dplyr::select(-Probe_maf)%>%
  arrange(by=P) %>%
  mutate(fdr=p.adjust(P, method = "fdr", n=length(P))) %>%
  mutate(bf=p.adjust(P, method = "bonferroni", n=length(P))) %>%  
  dplyr::filter(grepl('cg', cpg))

write.table(results,"ds.lm.meta.X.usc.sesame.ewas.anno.txt",
        quote=FALSE,
        row.names = FALSE,
        sep='\t')  

write.csv(results,"meta.chrX.LinearRegression.csv")

```

```{r association between EPIC PCs and blood proportions}

setwd("~/DSALL/usc")

library(tidyverse)
library(data.table)
library(Hmisc)
library(corrplot)

flattenCorrMatrix <- function(cormat, pmat) {
  ut <- upper.tri(cormat)
  data.frame(
    row = rownames(cormat)[row(cormat)[ut]],
    column = rownames(cormat)[col(cormat)[ut]],
    cor  =(cormat)[ut],
    p = pmat[ut]
  )
}
vars = fread("ds.all.covaraites.txt") %>%
  as.data.frame()%>%
  dplyr::select(CD8T,CD4T,NK,Bcell,Mono,
                Gran,nRBC,epi1,epi2,epi3,epi4,epi5,epi6,epi7,epi8,epi9,epi10) %>%
  as.matrix()

corr_mat <- cor(vars,method="pearson")
corre <- Hmisc::rcorr(as.matrix(vars),type="pearson")

heatmap(x = corr_mat, symm = TRUE,cexCol=0.6,cexRow=0.8)

corrplot::corrplot(corr_mat,type="upper",order = "hclust",tl.cex =0.6,	
 tl.col = "black", tl.srt = 45)



```

> DMRs in the USC dataset

```{r DMRs in the USC dataset}

library(tidyverse)
library(data.table)
setwd("~/DSALL/usc")

library(IlluminaHumanMethylationEPICanno.ilm10b4.hg19)

data("IlluminaHumanMethylationEPICanno.ilm10b4.hg19")
annoEPIC <- IlluminaHumanMethylationEPICanno.ilm10b4.hg19 %>%
  getAnnotation %>%
  as.data.frame %>%
  rownames_to_column(var="cpg") %>%
  select(cpg,chr,pos,Probe_maf)

############SESAME
#prep for comb-p and write out
usc <- fread("ds.all.usc.sesame.ewas.txt") %>%
  left_join(annoEPIC,by=c("V1"= "cpg")) %>%
  dplyr::mutate(chrom=chr, end=pos, start=pos, p_val=P,cpg=V1)%>%
  filter(!is.na(P))%>%
  arrange(by=P) %>%
  mutate(fdr=p.adjust(P, method = "fdr", n=length(P))) %>%
  mutate(bf=p.adjust(P, method = "bonferroni", n=length(P))) %>%  
  dplyr::filter(grepl('cg', cpg)) %>%
  dplyr::filter(Probe_maf<0.05|is.na(Probe_maf),
                chrom!="chrX"&chrom!="chrY") %>%
  dplyr::select(chrom, start, end, p_val)%>%
  na.omit()%>%
  arrange(chrom, start)

usc%>%
  write_tsv("comb_p_USC_SESAME_ALL_DS_EWAS_MA.txt")

############
##NOOB
#prep for comb-p and write out
usc <- fread("ds.all.usc.ewas.txt") %>%
  left_join(annoEPIC,by=c("V1"= "cpg")) %>%
  dplyr::mutate(chrom=chr, end=pos, start=pos, p_val=P,cpg=V1)%>%
  dplyr::select(chrom, start, end, p_val)%>%
  na.omit()%>%
  arrange(chrom, start)

usc%>%
  write_tsv("comb_p_USC_ALL_DS_EWAS_MA.txt")


```

```{bash run comb-p in USC dataset}

# this is done on the usc cluster for comb-p
sftp lishaobo@discovery.usc.edu
cd /scratch/lishaobo/temps/berkeley/dsall
lcd ~/DSALL/usc

##############SESAME
put comb_p_USC_SESAME_ALL_DS_EWAS_MA.txt
#source python 2
#source /usr/usc/python/2.7.8/setup.sh
#run comb-p and set settings according to preferences
comb-p pipeline -c 4 --seed 0.05 --dist 1000\
  -p comb_p_USC_ALL_DS_EWAS_MA_output\
  --region-filter-p 0.01 --region-filter-n 2 comb_p_USC_SESAME_ALL_DS_EWAS_MA.txt
  
#setting --acf-dist to 0.33 * --dist == 330
#calculated stepsize as: 330
#ACF:
# #lag_min	lag_max	correlation	N	p
#1	331	0.07029	710014	0
#wrote: comb_p_USC_SESAME_ALL_DS_EWAS_MA_output.acf.txt
#
# original lambda: 1.38
#wrote: comb_p_USC_SESAME_ALL_DS_EWAS_MA_output.slk.bed.gz with lambda: 1.61
#wrote: comb_p_USC_SESAME_ALL_DS_EWAS_MA_output.fdr.bed.gz
#wrote: comb_p_USC_SESAME_ALL_DS_EWAS_MA_output.regions.bed.gz (1040 regions)
# read 1040 regions from comb_p_USC_SESAME_ALL_DS_EWAS_MA_output.regions.bed.gz
# calculating ACF out to: 2041
#           with 8  lags: [1, 331, 661, 991, 1321, 1651, 1981, 2311]
# Done with one-time ACF calculation
#691791 bases used as coverage for sidak correction
#"warning: 0-length region found.
#does input have 0-length intervals? using length of 1 and not reporting
  
cd ~/DSALL/usc
cat <(head -1 <(less comb_p_USC_ALL_DS_EWAS_MA_output.regions-p.bed.gz)) \
  <(tail -n +2 <(less comb_p_USC_ALL_DS_EWAS_MA_output.regions-p.bed.gz) | sort -k7,7g) > comb_p_USC_SESAME_ALL_DS.txt

```

```{r run DMRCate in USC dataset}

setwd("~/DSALL/usc")
library(DMRcate)
library(tidyverse)
library(limma)
library(plyr)

####################SESAME
MA_outcomes <- fread("ds.all.usc.sesame.ewas.anno.txt")
MA_DMRcate <- MA_outcomes%>%
  dplyr::mutate(ID=cpg, stat=Beta/SE, CHR=chr, betafc=Beta,raw = P, indfdr = fdr)%>%
  dplyr::mutate(is.sig=ifelse(indfdr<0.05,TRUE,FALSE))%>%
  dplyr::select(ID, stat,CHR,pos,raw,betafc,indfdr,is.sig)%>%
  filter(!is.na(CHR))
MA_DMRcate_list <- as.list(MA_DMRcate)
class(MA_DMRcate_list) <- "annot"

dmrcate <- function (object, lambda = 1000, C = NULL, p.adjust.method = "BH", 
    pcutoff = "fdr", consec = FALSE, conseclambda = 10, betacutoff = NULL, 
    min.cpgs = 2, mc.cores = 1) {
    stopifnot(is(object, "annot"))
    stopifnot(lambda >= 1)
    stopifnot(pcutoff == "fdr" | (0 <= pcutoff & pcutoff <= 1))
    stopifnot(C >= 0.2)
    if (consec & is.null(conseclambda)) {
        stop("Consecutive CpG bandwidth must be specified")
    }
    object <- data.frame(ID = object$ID, weights = abs(object$stat), 
        CHR = as.character(object$CHR), pos = object$pos, betafc = object$betafc, 
        indfdr = object$indfdr, is.sig = object$is.sig)
    object <- object[order(object$CHR, object$pos), ]
    if (is.null(C) & !consec) {
        if (nrow(object) < 9e+05) {
            C = 2
        }
        else {
            C = 50
        }
    }
    if (consec) {
        lambda = conseclambda
        message(paste("Consecutive mode specified, lambda is now set at", 
            conseclambda, "consecutive CpGs."))
        if (is.null(C)) {
            stop("Error: argument C must be specified (in CpG sites) for consecutive mode.")
        }
        object$realcoordforconsec <- object$pos
        object$pos <- unlist(sapply(as.numeric(table(object$CHR)), 
            function(x) 1:x))
    }
    lag = lambda
    chr.unique <- unique(c(as.character(object$CHR)))
    fitted <- mclapply(chr.unique, fitParallel, object = object, 
        consec = consec, conseclambda = conseclambda, lambda = lambda, 
        C = C, mc.cores = mc.cores)
    object <- rbind.fill(fitted)
    object$fdr <- p.adjust(object$raw, method = p.adjust.method)
    if (pcutoff == "fdr") {
        nsig <- sum(object$is.sig)
        if (nsig == 0) {
            txt <- "The FDR you specified in cpg.annotate() returned no significant CpGs, hence there are no DMRs.\n    Try specifying a value of 'pcutoff' in dmrcate() and/or increasing 'fdr' in cpg.annotate()."
            stop(paste(strwrap(txt, exdent = 2), collapse = "\n"))
        }
        pcutoff <- sort(object$fdr)[nsig]
    }
    object$sig <- (object$fdr <= pcutoff)
    if (nrow(object) == 0) {
        txt <- "No signficant regions found. Try increasing the value of\n    'pcutoff' in dmrcate() and/or 'fdr' in cpg.annotate()."
        stop(paste(strwrap(txt, exdent = 2), collapse = "\n"))
    }
    message("Demarcating regions...")
    chr.N <- as.character(object$CHR)
    pos.N <- object$pos
    sig.N <- object$sig
    N <- length(sig.N)
    n.K <- which(sig.N)
    K <- length(n.K)
    stopifnot(K >= 2)
    pos.K <- pos.N[n.K]
    chr.K <- chr.N[n.K]
    jump_chr.k <- (chr.K[-1] != chr.K[-K])
    jump_pos.k <- (diff(pos.K) > lag)
    jump.k <- (jump_chr.k | jump_pos.k)
    ksegments.A2 <- Segment(jump.k)
    A <- nrow(ksegments.A2)
    kstart.A <- ksegments.A2[, "start"]
    kend.A <- ksegments.A2[, "end"]
    realpos.K <- pos.K
    if (consec) {
        realpos.N <- object$realcoordforconsec
        realpos.K <- realpos.N[n.K]
    }
    start.A <- realpos.K[kstart.A]
    end.A <- realpos.K[kend.A]
    chr.A <- chr.K[kstart.A]
    stopifnot(all(chr.K[kend.A] == chr.A))
    fmt <- "%s:%1d-%1d"
    coord.A <- sprintf(fmt, chr.A, start.A, end.A)
    nstart.A <- n.K[kstart.A]
    nend.A <- n.K[kend.A]
    width.A <- nend.A + 1 - nstart.A
    a.Z <- rep(seq(A), width.A)
    fn <- function(a) seq(from = nstart.A[a], to = nend.A[a])
    l.listA <- lapply(seq(A), fn)
    n.Z <- unlist(l.listA)
    region.N <- rep(NA_integer_, N)
    region.N[n.Z] <- a.Z
    levels <- seq(A)
    region.N <- factor(region.N, levels = levels)
    no_cpg.A <- c(table(region.N))
    REGIONSTAT <- function(field, fn) {
        x.N <- object[[field]]
        x.R <- tapply(x.N, region.N, fn)
        c(x.R)
    }
    fn_Stouffer <- function(x) pnorm(sum(qnorm(x))/sqrt(length(x)))
    fn_max <- function(x) x[which.max(abs(x))]
    results <- data.frame(coord = coord.A, no.cpgs = no_cpg.A, 
        minfdr = REGIONSTAT("fdr", min), Stouffer = REGIONSTAT("indfdr", 
            fn_Stouffer), maxbetafc = REGIONSTAT("betafc", fn_max), 
        meanbetafc = REGIONSTAT("betafc", mean), row.names = seq(A), 
        stringsAsFactors = FALSE)
    keep <- (results$no.cpgs >= min.cpgs)
    results <- results[keep, ]
    if (!is.null(betacutoff)) {
        keep <- (abs(results$meanbetafc) >= betacutoff)
        results <- results[keep, ]
    }
    o <- order(results$Stouffer, -results$no.cpgs)
    results <- results[o, ]
    message("Done!")
    output <- NULL
    output$input <- object
    output$results <- results
    output$cutoff <- pcutoff
    class(output) <- "dmrcate.output"
    output
}

dmrcoutput <- dmrcate(MA_DMRcate_list, lambda=1000, C=2, pcutoff = 0.01,min.cpgs = 2)
outcome_dmrcate <- dmrcoutput$results%>%
  write_rds("dmrcoutput_DS_ALL_USC_SESAME_MA.rds")

dmrcoutput$results %>%
  filter(grepl("chr11:",coord))
```

```{r compare overlap of DMRcate and comb-p in USC datasset}

library(tidyverse)
library(AnnotationHub)
library(TxDb.Hsapiens.UCSC.hg19.knownGene)
library(GenomicRanges)
library(annotatr)
library(methyAnalysis)
setwd("~/DSALL/usc")


####################SESAME
comb_p_output <- read_tsv("comb_p_USC_SESAME_ALL_DS_EWAS_MA_output.regions-p.bed.gz")%>%
  dplyr::mutate(chrom=`#chrom`)%>%
  dplyr::select(-`#chrom`)%>%
  dplyr::select(chrom, everything())%>%
  dplyr::filter(z_sidak_p<0.01)
dmrcate_results <- read_rds("dmrcoutput_DS_ALL_USC_SESAME_MA.rds")%>%
  dplyr::filter(minfdr<0.01)%>%
  separate(coord, c("chrom","coord"), sep =":")%>%
  separate(coord, c("start","end"), sep = "-")%>%
  dplyr::mutate(start=as.numeric(start),end=as.numeric(end))
comb_p_GR <- makeGRangesFromDataFrame(comb_p_output, keep.extra.columns=T,
                         ignore.strand=T,
                         seqinfo=NULL,
                         seqnames.field=c("chrom"),
                         start.field="start",
                         end.field=c("end"),
                         strand.field="strand",
                         starts.in.df.are.0based=FALSE)
DMRCate_GR <- makeGRangesFromDataFrame(dmrcate_results, keep.extra.columns=T,
                         ignore.strand=T,
                         seqinfo=NULL,
                         seqnames.field=c("chrom"),
                         start.field="start",
                         end.field=c("end"),
                         strand.field="strand",
                         starts.in.df.are.0based=FALSE)
overlap_DMRcate_comb_p_GR <- GenomicRanges::intersect(DMRCate_GR,comb_p_GR, ignore.strand=T)
length(overlap_DMRcate_comb_p_GR)
#79

if (require(TxDb.Hsapiens.UCSC.hg19.knownGene)) {
overlap_DMRcate_comb_p_DF <-annotateDMRInfo(overlap_DMRcate_comb_p_GR, 'TxDb.Hsapiens.UCSC.hg19.knownGene')
}

#left join with original files to get relevant meta data columns
overlap_DMRcate_comb_p_DF <- data.frame(overlap_DMRcate_comb_p_DF$sigDMRInfo)%>%
  left_join(comb_p_output, by=c("seqnames"="chrom","start","end"))%>%
  left_join(dmrcate_results,by=c("seqnames"="chrom","start","end"))
geneRanges <-
    function(db, column="ENTREZID"){
    g <- genes(db, columns=column)
    col <- mcols(g)[[column]]
    genes <- granges(g)[rep(seq_along(g), elementLengths(col))]
    mcols(genes)[[column]] <- as.character(unlist(col))
    genes
}
splitColumnByOverlap <-
    function(query, subject, column="ENTREZID", ...){
    olaps <- GenomicRanges::findOverlaps(query, subject, ...)
    f1 <- factor(subjectHits(olaps),
                 levels=seq_len(subjectLength(olaps)))
    splitAsList(mcols(query)[[column]][queryHits(olaps)], f1)
    }
annotate_grange <- function(grange_object, annotation_grange_object, column_name=NULL){
    file <- splitColumnByOverlap(annotation_grange_object,grange_object, column = column_name)
    grange_object$name <- file
    return(grange_object)}
annotate_grange <- function(grange_object, annotation_grange_object, column_name=NULL){
    file <- splitColumnByOverlap(annotation_grange_object,grange_object, column = column_name)
    grange_object$name <- file
    return(grange_object)}
overlap_DMRcate_comb_p_GR <- annotate_grange(overlap_DMRcate_comb_p_GR,DMRCate_GR, column_name="minfdr")
overlap_DMRcate_comb_p_GR$DMRCate_fdr <- overlap_DMRcate_comb_p_GR$name
overlap_DMRcate_comb_p_GR <- annotate_grange(overlap_DMRcate_comb_p_GR,DMRCate_GR, column_name="meanbetafc")
overlap_DMRcate_comb_p_GR$DMRCate_meanbetafc <- overlap_DMRcate_comb_p_GR$name
overlap_DMRcate_comb_p_GR <- annotate_grange(overlap_DMRcate_comb_p_GR,DMRCate_GR, column_name="no.cpgs")
overlap_DMRcate_comb_p_GR$DMRcate_N_cpgs <- overlap_DMRcate_comb_p_GR$name
overlap_DMRcate_comb_p_GR <- annotate_grange(overlap_DMRcate_comb_p_GR,comb_p_GR, column_name="z_sidak_p")
overlap_DMRcate_comb_p_GR$combp_sidak <- overlap_DMRcate_comb_p_GR$name
overlap_DMRcate_comb_p_GR <- annotate_grange(overlap_DMRcate_comb_p_GR,comb_p_GR, column_name="n_probes")
overlap_DMRcate_comb_p_GR$combp_N_cpgs <- overlap_DMRcate_comb_p_GR$name
overlap_DMRcate_comb_p_GR$name <- NULL
overlap_DMRcate_comb_p_GR$Gene_Symbol <- overlap_DMRcate_comb_p_DF$GeneSymbol
overlap_DMRcate_comb_p_GR$Promoter <- overlap_DMRcate_comb_p_DF$PROMOTER

overlap_DMRcate_comb_p_final <- overlap_DMRcate_comb_p_GR%>%
  as.data.frame()%>%
  dplyr::mutate(chrom=seqnames)%>%
  dplyr::select(-seqnames)%>%
  dplyr::select(chrom, start,end,width,Gene_Symbol,Promoter,everything())
colnames(overlap_DMRcate_comb_p_final) = paste0("usc_",colnames(overlap_DMRcate_comb_p_final))
write.csv(overlap_DMRcate_comb_p_final,"compDMARoverlap_comb_p_USC_SESAME_ALL.csv")

####################
#NOOB
comb_p_output <- read_tsv("comb_p_USC_ALL_DS_EWAS_MA_output.regions-p.bed.gz")%>%
  dplyr::mutate(chrom=`#chrom`)%>%
  dplyr::select(-`#chrom`)%>%
  dplyr::select(chrom, everything())%>%
  dplyr::filter(z_sidak_p<0.01)
dmrcate_results <- read_rds("dmrcoutput_DS_ALL_USC_MA.rds")%>%
  dplyr::filter(minfdr<0.01)%>%
  separate(coord, c("chrom","coord"), sep =":")%>%
  separate(coord, c("start","end"), sep = "-")%>%
  dplyr::mutate(start=as.numeric(start),end=as.numeric(end))
comb_p_GR <- makeGRangesFromDataFrame(comb_p_output, keep.extra.columns=T,
                         ignore.strand=T,
                         seqinfo=NULL,
                         seqnames.field=c("chrom"),
                         start.field="start",
                         end.field=c("end"),
                         strand.field="strand",
                         starts.in.df.are.0based=FALSE)
DMRCate_GR <- makeGRangesFromDataFrame(dmrcate_results, keep.extra.columns=T,
                         ignore.strand=T,
                         seqinfo=NULL,
                         seqnames.field=c("chrom"),
                         start.field="start",
                         end.field=c("end"),
                         strand.field="strand",
                         starts.in.df.are.0based=FALSE)
overlap_DMRcate_comb_p_GR <- GenomicRanges::intersect(DMRCate_GR,comb_p_GR, ignore.strand=T)
length(overlap_DMRcate_comb_p_GR)
#132

if (require(TxDb.Hsapiens.UCSC.hg19.knownGene)) {
overlap_DMRcate_comb_p_DF <-annotateDMRInfo(overlap_DMRcate_comb_p_GR, 'TxDb.Hsapiens.UCSC.hg19.knownGene')
}

#left join with original files to get relevant meta data columns
overlap_DMRcate_comb_p_DF <- data.frame(overlap_DMRcate_comb_p_DF$sigDMRInfo)%>%
  left_join(comb_p_output, by=c("seqnames"="chrom","start","end"))%>%
  left_join(dmrcate_results,by=c("seqnames"="chrom","start","end"))
geneRanges <-
    function(db, column="ENTREZID"){
    g <- genes(db, columns=column)
    col <- mcols(g)[[column]]
    genes <- granges(g)[rep(seq_along(g), elementLengths(col))]
    mcols(genes)[[column]] <- as.character(unlist(col))
    genes
}
splitColumnByOverlap <-
    function(query, subject, column="ENTREZID", ...){
    olaps <- GenomicRanges::findOverlaps(query, subject, ...)
    f1 <- factor(subjectHits(olaps),
                 levels=seq_len(subjectLength(olaps)))
    splitAsList(mcols(query)[[column]][queryHits(olaps)], f1)
    }
annotate_grange <- function(grange_object, annotation_grange_object, column_name=NULL){
    file <- splitColumnByOverlap(annotation_grange_object,grange_object, column = column_name)
    grange_object$name <- file
    return(grange_object)}
annotate_grange <- function(grange_object, annotation_grange_object, column_name=NULL){
    file <- splitColumnByOverlap(annotation_grange_object,grange_object, column = column_name)
    grange_object$name <- file
    return(grange_object)}
overlap_DMRcate_comb_p_GR <- annotate_grange(overlap_DMRcate_comb_p_GR,DMRCate_GR, column_name="minfdr")
overlap_DMRcate_comb_p_GR$DMRCate_fdr <- overlap_DMRcate_comb_p_GR$name
overlap_DMRcate_comb_p_GR <- annotate_grange(overlap_DMRcate_comb_p_GR,DMRCate_GR, column_name="meanbetafc")
overlap_DMRcate_comb_p_GR$DMRCate_meanbetafc <- overlap_DMRcate_comb_p_GR$name
overlap_DMRcate_comb_p_GR <- annotate_grange(overlap_DMRcate_comb_p_GR,DMRCate_GR, column_name="no.cpgs")
overlap_DMRcate_comb_p_GR$DMRcate_N_cpgs <- overlap_DMRcate_comb_p_GR$name
overlap_DMRcate_comb_p_GR <- annotate_grange(overlap_DMRcate_comb_p_GR,comb_p_GR, column_name="z_sidak_p")
overlap_DMRcate_comb_p_GR$combp_sidak <- overlap_DMRcate_comb_p_GR$name
overlap_DMRcate_comb_p_GR <- annotate_grange(overlap_DMRcate_comb_p_GR,comb_p_GR, column_name="n_probes")
overlap_DMRcate_comb_p_GR$combp_N_cpgs <- overlap_DMRcate_comb_p_GR$name
overlap_DMRcate_comb_p_GR$name <- NULL
overlap_DMRcate_comb_p_GR$Gene_Symbol <- overlap_DMRcate_comb_p_DF$GeneSymbol
overlap_DMRcate_comb_p_GR$Promoter <- overlap_DMRcate_comb_p_DF$PROMOTER

overlap_DMRcate_comb_p_final <- overlap_DMRcate_comb_p_GR%>%
  as.data.frame()%>%
  dplyr::mutate(chrom=seqnames)%>%
  dplyr::select(-seqnames)%>%
  dplyr::select(chrom, start,end,width,Gene_Symbol,Promoter,everything())
colnames(overlap_DMRcate_comb_p_final) = paste0("usc_",colnames(overlap_DMRcate_comb_p_final))
write.csv(overlap_DMRcate_comb_p_final,"compDMARoverlap_comb_p_USC_ALL.csv")

```

> DMRs in the BCM dataset

```{r a sneak peak into bcm results}

setwd("~/DSALL/bcm")

library(tidyverse)
library(data.table)
library(IlluminaHumanMethylationEPICanno.ilm10b4.hg19)

data("IlluminaHumanMethylationEPICanno.ilm10b4.hg19")
annoEPIC <- IlluminaHumanMethylationEPICanno.ilm10b4.hg19 %>%
  getAnnotation %>%
  as.data.frame %>%
  rownames_to_column(var="cpg") %>%
  dplyr::select(cpg,chr,pos,Relation_to_Island,UCSC_RefGene_Name,Probe_maf,
                UCSC_RefGene_Group)

results <- fread("5_ds.all.bcm.ewas_lambda1.75.txt") %>%
  dplyr::rename("cpg"="V1") %>%
  left_join(annoEPIC,by="cpg") %>%
  filter(!is.na(P))%>%
  arrange(by=P) %>%
  dplyr::filter(grepl('cg', cpg)) %>%
  dplyr::filter(Probe_maf<0.05|is.na(Probe_maf),
                chr!="chrX"&chr!="chrY")%>%
  mutate(fdr=p.adjust(P, method = "fdr", n=length(P)))  %>%
  mutate(bf=p.adjust(P, method = "bonferroni", n=length(P)))   

{
  Nbeta <- sum(!is.na(results$P))
  tmp <- results[!is.na(results$P),]
  lambda = median(qchisq(1- tmp$P,1))/qchisq(0.5,1)
  print(lambda)
}

# try differnt cut-offs for this analysis
results1 = results %>%
  filter(N.exp>20 & N.unexp>20)

{
  Nbeta <- sum(!is.na(results1$P))
  tmp <- results1[!is.na(results1$P),]
  lambda = median(qchisq(1- tmp$P,1))/qchisq(0.5,1)
  print(lambda)
  head(results1)
}


write.table(results1,"ds.all.bcm.ewas.anno.txt",
        quote=FALSE,
        sep='\t')  

dim(results1%>%filter(fdr<0.05))
dim(results1%>%filter(bf<0.05))

results.check <- head(results1,20)
write.csv(results.check,"bcm.top20.cpgs.csv")

## check the 26 CpGs significant in USC data 
results.sigs.lu = results %>%
  mutate(P.bcm=P,
         bf.bcm=bf,
         Beta.bcm=Beta)%>%
  dplyr::select(cpg,P.bcm,bf.bcm,Beta.bcm,N.exp, N.unexp)
usc.sigs = fread("~/DSALL/sigs.usc.cpgs.txt") %>%
  dplyr::rename("bf.usc"="bf",
         "P.usc"="P",
         "Beta.usc"="Beta")%>%
  dplyr::select(cpg,Beta.usc,P.usc,bf.usc)%>%
  inner_join(results.sigs.lu,by="cpg")
write.csv(usc.sigs,"~/DSALL/usc.sigs.lookup.csv")

```

```{r DMRs in the BCM dataset}

library(tidyverse)
library(data.table)
setwd("~/DSALL/bcm")

library(IlluminaHumanMethylationEPICanno.ilm10b4.hg19)

data("IlluminaHumanMethylationEPICanno.ilm10b4.hg19")
annoEPIC <- IlluminaHumanMethylationEPICanno.ilm10b4.hg19 %>%
  getAnnotation %>%
  as.data.frame %>%
  rownames_to_column(var="cpg") %>%
  select(cpg,chr,pos)

#prep for comb-p and write out
bcm<- fread("5_ds.all.bcm.ewas_lambda1.75.txt") %>%
  left_join(annoEPIC,by=c("V1"= "cpg")) %>%
  filter(N.exp>20 & N.unexp>20) %>%
  dplyr::mutate(chrom=chr, end=pos, start=pos, p_val=P,cpg=V1)%>%
  dplyr::select(chrom, start, end, p_val)%>%
  na.omit()%>%
  arrange(chrom, start)

bcm%>%
  write_tsv("comb_p_BCM_ALL_DS_EWAS_MA.txt")

```

```{bash run comb-p in BCM dataset}

# this is done on the usc cluster for comb-p
sftp lishaobo@discovery.usc.edu
cd /scratch/lishaobo/temps/berkeley/dsall
lcd ~/DSALL/bcm
put comb_p_BCM_ALL_DS_EWAS_MA.txt

#run comb-p and set settings according to preferences
comb-p pipeline -c 4 --seed 0.05 --dist 1000\
  -p comb_p_BCM_ALL_DS_EWAS_MA_output\
  --region-filter-p 0.01 --region-filter-n 2 comb_p_BCM_ALL_DS_EWAS_MA.txt
  
#setting --acf-dist to 0.33 * --dist == 330
#calculated stepsize as: 330
#ACF:
# #lag_min	lag_max	correlation	N	p
#1	331	0.115	567997	0
#wrote: comb_p_BCM_ALL_DS_EWAS_MA_output.acf.txt
# original lambda: 1.66
#wrote: comb_p_BCM_ALL_DS_EWAS_MA_output.slk.bed.gz with lambda: 2.20
#wrote: comb_p_BCM_ALL_DS_EWAS_MA_output.fdr.bed.gz
#wrote: comb_p_BCM_ALL_DS_EWAS_MA_output.regions.bed.gz (4833 regions)
# read 4833 regions from comb_p_BCM_ALL_DS_EWAS_MA_output.regions.bed.gz
# calculating ACF out to: 2145
#           with 8  lags: [1, 331, 661, 991, 1321, 1651, 1981, 2311]
# Done with one-time ACF calculation
#539571 bases used as coverage for sidak correction
#"warning: 0-length region found.
#does input have 0-length intervals? using length of 1 and not reporting
#further 0-length intervalswrote: comb_p_BCM_ALL_DS_EWAS_MA_output.regions-p.bed.gz, (regions #with corrected-p < 0.05: 344)

cd ~/DSALL/usc
cat <(head -1 <(less comb_p_BCM_ALL_DS_EWAS_MA_output.regions-p.bed.gz)) \
  <(tail -n +2 <(less comb_p_BCM_ALL_DS_EWAS_MA_output.regions-p.bed.gz) | sort -k7,7g) > comb_p_BCM_ALL_DS.txt
  
get comb_p_BCM_ALL_DS_EWAS_MA_output.regions-p.bed.gz
get comb_p_BCM_ALL_DS.txt
  
```

```{r run DMRCate in BCM dataset}
setwd("~/DSALL/bcm")
library(DMRcate)
library(tidyverse)
library(limma)
library(plyr)

MA_outcomes <- fread("ds.all.bcm.ewas.anno.txt")
MA_DMRcate <- MA_outcomes%>%
  dplyr::mutate(ID=cpg, stat=Beta/SE, CHR=chr, betafc=Beta,raw = P, indfdr = fdr)%>%
  dplyr::mutate(is.sig=ifelse(indfdr<0.05,TRUE,FALSE))%>%
  dplyr::select(ID, stat,CHR,pos,raw,betafc,indfdr,is.sig)%>%
  filter(!is.na(CHR))
MA_DMRcate_list <- as.list(MA_DMRcate)
class(MA_DMRcate_list) <- "annot"

dmrcate <- function (object, lambda = 1000, C = NULL, p.adjust.method = "BH", 
    pcutoff = "fdr", consec = FALSE, conseclambda = 10, betacutoff = NULL, 
    min.cpgs = 2, mc.cores = 1) {
    stopifnot(is(object, "annot"))
    stopifnot(lambda >= 1)
    stopifnot(pcutoff == "fdr" | (0 <= pcutoff & pcutoff <= 1))
    stopifnot(C >= 0.2)
    if (consec & is.null(conseclambda)) {
        stop("Consecutive CpG bandwidth must be specified")
    }
    object <- data.frame(ID = object$ID, weights = abs(object$stat), 
        CHR = as.character(object$CHR), pos = object$pos, betafc = object$betafc, 
        indfdr = object$indfdr, is.sig = object$is.sig)
    object <- object[order(object$CHR, object$pos), ]
    if (is.null(C) & !consec) {
        if (nrow(object) < 9e+05) {
            C = 2
        }
        else {
            C = 50
        }
    }
    if (consec) {
        lambda = conseclambda
        message(paste("Consecutive mode specified, lambda is now set at", 
            conseclambda, "consecutive CpGs."))
        if (is.null(C)) {
            stop("Error: argument C must be specified (in CpG sites) for consecutive mode.")
        }
        object$realcoordforconsec <- object$pos
        object$pos <- unlist(sapply(as.numeric(table(object$CHR)), 
            function(x) 1:x))
    }
    lag = lambda
    chr.unique <- unique(c(as.character(object$CHR)))
    fitted <- mclapply(chr.unique, fitParallel, object = object, 
        consec = consec, conseclambda = conseclambda, lambda = lambda, 
        C = C, mc.cores = mc.cores)
    object <- rbind.fill(fitted)
    object$fdr <- p.adjust(object$raw, method = p.adjust.method)
    if (pcutoff == "fdr") {
        nsig <- sum(object$is.sig)
        if (nsig == 0) {
            txt <- "The FDR you specified in cpg.annotate() returned no significant CpGs, hence there are no DMRs.\n    Try specifying a value of 'pcutoff' in dmrcate() and/or increasing 'fdr' in cpg.annotate()."
            stop(paste(strwrap(txt, exdent = 2), collapse = "\n"))
        }
        pcutoff <- sort(object$fdr)[nsig]
    }
    object$sig <- (object$fdr <= pcutoff)
    if (nrow(object) == 0) {
        txt <- "No signficant regions found. Try increasing the value of\n    'pcutoff' in dmrcate() and/or 'fdr' in cpg.annotate()."
        stop(paste(strwrap(txt, exdent = 2), collapse = "\n"))
    }
    message("Demarcating regions...")
    chr.N <- as.character(object$CHR)
    pos.N <- object$pos
    sig.N <- object$sig
    N <- length(sig.N)
    n.K <- which(sig.N)
    K <- length(n.K)
    stopifnot(K >= 2)
    pos.K <- pos.N[n.K]
    chr.K <- chr.N[n.K]
    jump_chr.k <- (chr.K[-1] != chr.K[-K])
    jump_pos.k <- (diff(pos.K) > lag)
    jump.k <- (jump_chr.k | jump_pos.k)
    ksegments.A2 <- Segment(jump.k)
    A <- nrow(ksegments.A2)
    kstart.A <- ksegments.A2[, "start"]
    kend.A <- ksegments.A2[, "end"]
    realpos.K <- pos.K
    if (consec) {
        realpos.N <- object$realcoordforconsec
        realpos.K <- realpos.N[n.K]
    }
    start.A <- realpos.K[kstart.A]
    end.A <- realpos.K[kend.A]
    chr.A <- chr.K[kstart.A]
    stopifnot(all(chr.K[kend.A] == chr.A))
    fmt <- "%s:%1d-%1d"
    coord.A <- sprintf(fmt, chr.A, start.A, end.A)
    nstart.A <- n.K[kstart.A]
    nend.A <- n.K[kend.A]
    width.A <- nend.A + 1 - nstart.A
    a.Z <- rep(seq(A), width.A)
    fn <- function(a) seq(from = nstart.A[a], to = nend.A[a])
    l.listA <- lapply(seq(A), fn)
    n.Z <- unlist(l.listA)
    region.N <- rep(NA_integer_, N)
    region.N[n.Z] <- a.Z
    levels <- seq(A)
    region.N <- factor(region.N, levels = levels)
    no_cpg.A <- c(table(region.N))
    REGIONSTAT <- function(field, fn) {
        x.N <- object[[field]]
        x.R <- tapply(x.N, region.N, fn)
        c(x.R)
    }
    fn_Stouffer <- function(x) pnorm(sum(qnorm(x))/sqrt(length(x)))
    fn_max <- function(x) x[which.max(abs(x))]
    results <- data.frame(coord = coord.A, no.cpgs = no_cpg.A, 
        minfdr = REGIONSTAT("fdr", min), Stouffer = REGIONSTAT("indfdr", 
            fn_Stouffer), maxbetafc = REGIONSTAT("betafc", fn_max), 
        meanbetafc = REGIONSTAT("betafc", mean), row.names = seq(A), 
        stringsAsFactors = FALSE)
    keep <- (results$no.cpgs >= min.cpgs)
    results <- results[keep, ]
    if (!is.null(betacutoff)) {
        keep <- (abs(results$meanbetafc) >= betacutoff)
        results <- results[keep, ]
    }
    o <- order(results$Stouffer, -results$no.cpgs)
    results <- results[o, ]
    message("Done!")
    output <- NULL
    output$input <- object
    output$results <- results
    output$cutoff <- pcutoff
    class(output) <- "dmrcate.output"
    output
}

dmrcoutput <- dmrcate(MA_DMRcate_list, lambda=1000, C=2, pcutoff = 0.01,min.cpgs = 2)
outcome_dmrcate <- dmrcoutput$results%>%
  write_rds("dmrcoutput_DS_ALL_BCM_MA.rds")

```

```{r compare overlap of DMRcate and comb-p in BCM datasset}

library(tidyverse)
library(GenomicRanges)
library(AnnotationHub)
library(TxDb.Hsapiens.UCSC.hg19.knownGene)
library(annotatr)
library(methyAnalysis)

setwd("~/DSALL/bcm")

comb_p_output <- read_tsv("comb_p_BCM_ALL_DS_EWAS_MA_output.regions-p.bed.gz")%>%
  dplyr::mutate(chrom=`#chrom`)%>%
  dplyr::select(-`#chrom`)%>%
  dplyr::select(chrom, everything())%>%
  dplyr::filter(z_sidak_p<0.01)
dmrcate_results <- read_rds("dmrcoutput_DS_ALL_BCM_MA.rds")%>%
  dplyr::filter(minfdr<0.01)%>%
  separate(coord, c("chrom","coord"), sep =":")%>%
  separate(coord, c("start","end"), sep = "-")%>%
  dplyr::mutate(start=as.numeric(start),end=as.numeric(end))
comb_p_GR <- makeGRangesFromDataFrame(comb_p_output, keep.extra.columns=T,
                         ignore.strand=T,
                         seqinfo=NULL,
                         seqnames.field=c("chrom"),
                         start.field="start",
                         end.field="end",
                         strand.field="strand",
                         starts.in.df.are.0based=FALSE)
DMRCate_GR <- makeGRangesFromDataFrame(dmrcate_results, keep.extra.columns=T,
                         ignore.strand=T,
                         seqinfo=NULL,
                         seqnames.field=c("chrom"),
                         start.field="start",
                         end.field=c("end"),
                         strand.field="strand",
                         starts.in.df.are.0based=FALSE)
overlap_DMRcate_comb_p_GR <- GenomicRanges::intersect(DMRCate_GR,comb_p_GR, ignore.strand=T)
length(overlap_DMRcate_comb_p_GR)
#140

if (require(TxDb.Hsapiens.UCSC.hg19.knownGene)) {
overlap_DMRcate_comb_p_DF <-annotateDMRInfo(overlap_DMRcate_comb_p_GR, 'TxDb.Hsapiens.UCSC.hg19.knownGene')
}

#left join with original files to get relevant meta data columns
overlap_DMRcate_comb_p_DF <- data.frame(overlap_DMRcate_comb_p_DF$sigDMRInfo)%>%
  left_join(comb_p_output, by=c("seqnames"="chrom","start","end"))%>%
  left_join(dmrcate_results,by=c("seqnames"="chrom","start","end"))
geneRanges <-
    function(db, column="ENTREZID"){
    g <- genes(db, columns=column)
    col <- mcols(g)[[column]]
    genes <- granges(g)[rep(seq_along(g), elementLengths(col))]
    mcols(genes)[[column]] <- as.character(unlist(col))
    genes
}
splitColumnByOverlap <-
    function(query, subject, column="ENTREZID", ...){
    olaps <- GenomicRanges::findOverlaps(query, subject, ...)
    f1 <- factor(subjectHits(olaps),
                 levels=seq_len(subjectLength(olaps)))
    splitAsList(mcols(query)[[column]][queryHits(olaps)], f1)
    }
annotate_grange <- function(grange_object, annotation_grange_object, column_name=NULL){
    file <- splitColumnByOverlap(annotation_grange_object,grange_object, column = column_name)
    grange_object$name <- file
    return(grange_object)}
annotate_grange <- function(grange_object, annotation_grange_object, column_name=NULL){
    file <- splitColumnByOverlap(annotation_grange_object,grange_object, column = column_name)
    grange_object$name <- file
    return(grange_object)}
overlap_DMRcate_comb_p_GR <- annotate_grange(overlap_DMRcate_comb_p_GR,DMRCate_GR, column_name="minfdr")
overlap_DMRcate_comb_p_GR$DMRCate_fdr <- overlap_DMRcate_comb_p_GR$name
overlap_DMRcate_comb_p_GR <- annotate_grange(overlap_DMRcate_comb_p_GR,DMRCate_GR, column_name="meanbetafc")
overlap_DMRcate_comb_p_GR$DMRCate_meanbetafc <- overlap_DMRcate_comb_p_GR$name
overlap_DMRcate_comb_p_GR <- annotate_grange(overlap_DMRcate_comb_p_GR,DMRCate_GR, column_name="no.cpgs")
overlap_DMRcate_comb_p_GR$DMRcate_N_cpgs <- overlap_DMRcate_comb_p_GR$name
overlap_DMRcate_comb_p_GR <- annotate_grange(overlap_DMRcate_comb_p_GR,comb_p_GR, column_name="z_sidak_p")
overlap_DMRcate_comb_p_GR$combp_sidak <- overlap_DMRcate_comb_p_GR$name
overlap_DMRcate_comb_p_GR <- annotate_grange(overlap_DMRcate_comb_p_GR,comb_p_GR, column_name="n_probes")
overlap_DMRcate_comb_p_GR$combp_N_cpgs <- overlap_DMRcate_comb_p_GR$name
overlap_DMRcate_comb_p_GR$name <- NULL
overlap_DMRcate_comb_p_GR$Gene_Symbol <- overlap_DMRcate_comb_p_DF$GeneSymbol
overlap_DMRcate_comb_p_GR$Promoter <- overlap_DMRcate_comb_p_DF$PROMOTER

overlap_DMRcate_comb_p_final <- overlap_DMRcate_comb_p_GR%>%
  as.data.frame()%>%
  dplyr::mutate(chrom=seqnames)%>%
  dplyr::select(-seqnames)%>%
  dplyr::select(chrom, start,end,width,Gene_Symbol,Promoter,everything())
colnames(overlap_DMRcate_comb_p_final) = paste0("bcm_",colnames(overlap_DMRcate_comb_p_final))
write.csv(overlap_DMRcate_comb_p_final,"compDMARoverlap_comb_p_BCM_ALL.csv")
```

>> investigating why Noob and Sesame top variable CpGs are different

```{r}

setwd("~/DSALL/usc")
library(tidyverse)
library(data.table)
library(DescTools)

library(IlluminaHumanMethylationEPICanno.ilm10b4.hg19)
data("IlluminaHumanMethylationEPICanno.ilm10b4.hg19")
anno       <- IlluminaHumanMethylationEPICanno.ilm10b4.hg19 %>%
              getAnnotation %>%
              as.data.frame %>%
              rownames_to_column(var="cpg")

xyprobbes = anno %>% filter(Probe_maf<0.05|is.na(Probe_maf), chr!="chrX"&chr!="chrY") %>% dplyr::select(cpg)
betas3_noob = read_rds("~/sets/set3/beta_noob_set3.rds") %>%
  t %>%
  as.data.frame()
betas3_noob_1 = betas3_noob[,colnames(betas3_noob)%in%xyprobbes$cpg] %>% 
  t() %>%
  as.data.frame()
MAD <- apply(betas3_noob_1, 1, MeanAD)
betas3_noob_1$MAD <- MAD
betas3_noob_2 <- betas3_noob_1 %>%
  arrange(desc(MAD))
betas3_noob_3 <- betas3_noob_2[1:2000,] %>% as.matrix()

sesame.raw = fread("~/sets/set3/set3_Sesame_beta.csv") %>%
  as.data.frame()%>%
  dplyr::select(-V1)%>%
  column_to_rownames(var="probeId")

betas3_0_imp <- read_rds("sesame.beta.s3.rds")
betas3_sesame <- betas3_0_imp$data

betas3_sesame_1 = betas3_sesame[,colnames(betas3_sesame)%in%xyprobbes$cpg] %>% 
  t() %>%
  as.data.frame()

MAD <- apply(betas3_sesame_1, 1, MeanAD)
betas3_sesame_1$MAD <- MAD
betas3_sesame_2 <- betas3_sesame_1 %>%
  arrange(desc(MAD))
betas3_sesame_3 <- betas3_sesame_2[1:2000,] %>% as.matrix()


length(intersect(rownames(betas3_noob_3),rownames(betas3_sesame_3)))
# 110 of the CpGs overlap

# choose 50 CpGs that are common (passed SESAME QC)
goodCpGs = sample(intersect(rownames(betas3_noob_3),rownames(betas3_sesame_3)),50)

# investigating top 2000 CpGs identified by NOOB
length(intersect(rownames(betas3_sesame_2),rownames(betas3_noob_2)))
# 647937

com.cpg <- length(intersect(rownames(betas3_sesame_2),rownames(betas3_noob_3)))
# 110

### MAD score comparison
ovlpd = intersect(rownames(betas3_sesame_2),rownames(betas3_noob_3))

betas3_sesame_2_1 = betas3_sesame_2[ovlpd,]
betas3_noob_2_1 = betas3_noob_2[ovlpd,]

mad_compare = data.frame(mad_sesame=betas3_sesame_2_1$MAD, mad_noob=betas3_noob_2_1$MAD)

mad_compare %>%
  ggplot(aes(x=mad_sesame,y=mad_noob))+
  geom_point()
ggsave("mad.common.png")

# choose 50 cpgs that are filtered out and compare their values
badCpGs = rownames(betas3_noob_3)[sample(which(!rownames(betas3_noob_3) %in% ovlpd),50)]

nob.sel = betas3_noob_3[badCpGs,-ncol(betas3_noob_3)] %>% as.data.frame() 
nob.sel[,1:6]

# sesame.seld = 
  sesame.sel =  sesame.raw[badCpGs,] %>% t %>% as.data.frame() 
  summary(sesame.sel)


head(goodCpGs)  
head(badCpGs)  

allCpGs = c(goodCpGs,badCpGs)
qualityS = c(rep("good",50),rep("bad",50))
allCpGs_1 = data.frame(cpg=allCpGs,qualityS=qualityS) %>%
  mutate(qualityS=as.factor(qualityS))


### dive into the normalization process of SeSame
allSubs = fread("ds.all.final.subs.txt")

s3rg = read_rds("~/sets/set3/RGset.rds")
s3rg_1 = s3rg[,allSubs$beadPosition]
write_rds(s3rg_1,"s3rg_1.rds")


s3rg_ts_all = subsetByLoci(s3rg_1, includeLoci = allCpGs)
detecP <- detectionP(s3rg_ts_all) %>% as.data.frame()

detecP_mean = rowMeans(detecP) %>%
  as.data.frame() 
colnames(detecP_mean) <- "detectionP"
detecP_mean_1 = detecP_mean %>%
  rownames_to_column(var="cpg") %>%
  inner_join(allCpGs_1,by="cpg") %>%
  mutate(rankQ = rank(detectionP))

detecP_mean_1 %>%
  ggplot(aes(x=qualityS,y=detectionP))+
  geom_boxplot()+
  theme_classic() +
  ggtitle("detection p value in 2 groups of probes")+
  theme(plot.title = element_text(hjust = 0.5)) 
ggsave("box.plot.good.bad.probes.png",width = 5,height = 5)

detecP_mean_1 %>%
  ggplot(aes(x=qualityS,y=rankQ))+
  geom_boxplot()+
  theme_classic() +
  ggtitle("detection p value in 2 groups of probes")+
  theme(plot.title = element_text(hjust = 0.5)) 
ggsave("box.plot.good.bad.probes.rank.png",width = 5,height = 5)

```

> meta-analysis with BCM EWAS results

```{r meta-analysis}

setwd("~/DSALL")
library(tidyverse)
library(data.table)
 
usc <- fread("usc/ds.all.usc.sesame.ewas.anno.txt")
baylor <- fread("bcm/ds.all.bcm.ewas.anno.txt")
#prep datasets, add sample size, filter for cpgs with probes overlapping snps, select relevant columns and write out
usc%>%
  dplyr::mutate(N=343)%>%
  dplyr::filter(chr!="chrX"&chr!="chrY")%>%
  dplyr::mutate(cpg_name=cpg,estimate=Beta,std_error=SE,p_val=P)%>%
  dplyr::select(cpg_name,estimate, std_error, p_val, chr, pos, N)%>%
  write_tsv("USC_ALL_DS_MA_prep.tsv")

baylor%>%
  dplyr::mutate(N=48)%>%
  dplyr::filter(chr!="chrX"&chr!="chrY")%>%
  dplyr::mutate(cpg_name=cpg,estimate=Beta,std_error=SE,p_val=P)%>%
  dplyr::select(cpg_name,estimate, std_error, p_val, chr, pos, N)%>%
  write_tsv("BCM_ALL_DS_MA_prep.tsv")
```

```{bash}

cd ~/DSALL

metal
MARKER cpg_name
EFFECT estimate
PVALUE p_val
SCHEME STDERR
STDERR std_error
VERBOSE OFF
GENOMICCONTROL OFF
PROCESS USC_ALL_DS_MA_prep.tsv
PROCESS BCM_ALL_DS_MA_prep.tsv
ANALYZE HETEROGENEITY
QUIT
mv METAANALYSIS1.TBL usc_bcm_meta_sd.TBL

# results present in both studies
awk -v FS='\t' '$7 !~ /?/ {print $1,$4,$5,$6,$7,$8,$9,$10,$11}' usc_bcm_meta_sd.TBL | sort -k4,4n > usc_bcm_meta_sd_both.txt

######## 
#using sample size
metal
MARKER cpg_name
EFFECT estimate
PVALUE p_val
WEIGHTLABEL N
STDERR std_error
VERBOSE OFF
GENOMICCONTROL OFF
PROCESS USC_ALL_DS_MA_prep.tsv
PROCESS BCM_ALL_DS_MA_prep.tsv
ANALYZE HETEROGENEITY
QUIT
mv METAANALYSIS1.TBL usc_bcm_meta.TBL

awk -v FS='\t' '$7 !~ /?/ {print $1,$4,$5,$6,$7,$8,$9,$10,$11}' usc_bcm_meta.TBL | sort -k4,4n > usc_bcm_meta_both.txt


```

```{r annotating the results of meta-analysis}
setwd("~/DSALL")
library(tidyverse)
library(data.table)
library(IlluminaHumanMethylationEPICanno.ilm10b4.hg19)
data("IlluminaHumanMethylationEPICanno.ilm10b4.hg19")
anno       <- IlluminaHumanMethylationEPICanno.ilm10b4.hg19 %>%
              getAnnotation %>%
              as.data.frame %>%
              rownames_to_column(var="MarkerName")

# for results in both sets
MA_outcomes <- fread("usc_bcm_meta_sd_both.txt")%>%
  left_join(anno, by="MarkerName")%>%
  dplyr::mutate(p_val_fdr= p.adjust(`P-value`, method="fdr"))%>%
  dplyr::mutate(cpg=MarkerName)%>%
  dplyr::select(cpg,everything(), -MarkerName)%>%
  arrange(`P-value`)%>%
  write_tsv("BCM_USC_ALL_DS_EWAS_MA.tsv")
##also write as rds
MA_outcomes%>%
  write_rds("BCM_USC_ALL_DS_EWAS_MA.rds")

# for results regardless of appearance in each set
MA_outcomes <- fread("usc_bcm_meta_sd.TBL")%>%
  left_join(anno, by="MarkerName")%>%
  dplyr::mutate(p_val_fdr= p.adjust(`P-value`, method="fdr"))%>%
  dplyr::mutate(cpg=MarkerName)%>%
  dplyr::select(cpg,everything(), -MarkerName)%>%
  arrange(`P-value`)%>%
  write_tsv("BCM_USC_ANY_DS_EWAS_MA.tsv")

## also write as rds
MA_outcomes%>%
  write_rds("BCM_USC_ANY_DS_EWAS_MA.rds")


## A short version for sharing
## Adding columns of effect sizes and P values from single studies
usc <- fread("usc/ds.all.usc.sesame.ewas.anno.txt") %>%
  dplyr::select(cpg, Beta, P)%>%
  dplyr::rename("estimate_usc"="Beta",
                 "p_usc" = "P")
baylor <- fread("bcm/ds.all.bcm.ewas.anno.txt")  %>%
  dplyr::select(cpg, Beta, P)%>%
  dplyr::rename("estimate_bcm"="Beta",
                 "p_bcm" = "P") %>%
  inner_join(usc, by ="cpg")
outcomes_model_f <- read_rds("BCM_USC_ALL_DS_EWAS_MA.rds") %>%
  filter(p_val_fdr<0.05) %>%
  mutate(Gencode_Group = gsub(";.*$","",GencodeCompV12_Group),
         Gencode_Name = gsub(";.*$","",UCSC_RefGene_Name)) %>%
  dplyr::select(cpg, Effect, `P-value`, Direction,p_val_fdr,Gencode_Name,Gencode_Group) %>%
  left_join(baylor, by="cpg")
write_tsv(outcomes_model_f,"BCM_USC_ALL_sim.tsv")
```

> For USC, BCM and meta analyses, Manhattan plot, Q-Q plot and lambda value for the DS-ALL EWAS

```{r}

setwd("~/DSALL")
library(tidyverse)
library(data.table)
library(CMplot)
library(QCEWAS)
library(qqman)

#load p-values from EWAS
## USC

################################SESAME
usc_model_anno <- fread("usc/ds.all.usc.sesame.ewas.anno.txt") %>%
  dplyr::mutate(cpg_name=cpg,estimate=Beta, std_error=SE,p_val=P)

outcomes_model_anno <- usc_model_anno %>% 
  filter(chr!="chrX"&chr!="chrY")

sig_thresghold = outcomes_model_anno %>%
  mutate(adjp = p.adjust(p_val, method="fdr"))%>%
  filter(adjp<0.05)%>%
  arrange(desc(adjp))
sig_threshold = sig_thresghold[1,"p_val"]

annotation = outcomes_model_anno%>%
  filter(bf<0.05) %>%
  filter(UCSC_RefGene_Name!="") %>%
  mutate(gene.highlight = sub(";.*$","",UCSC_RefGene_Name))
cpg.highlight = annotation$cpg_name
genes.highlight = annotation$gene.highlight

outcomes_model_CM_plot <- outcomes_model_anno%>%
  dplyr::mutate(p_val_cmplot = -log10(p_val))%>%
  dplyr::mutate(p_val_cmplot= ifelse(estimate<0,-p_val_cmplot, p_val_cmplot )) %>%
  dplyr::select(cpg_name, chr,pos, p_val_cmplot)
outcomes_model_CM_plot$chr <- gsub("chr","",outcomes_model_CM_plot$chr)

CMplot(outcomes_model_CM_plot,
       plot.type="m",
       band=0, 
       cex=c(0.5,0.5,0.5),
       amplify=FALSE,
       LOG10=FALSE,  
       signal.col=NULL,
       chr.den.col=NULL,
       threshold=c(log10(sig_threshold[[1]]), -log10(sig_threshold[[1]])),
       threshold.lty=2, 
       threshold.lwd=1, 
       threshold.col="black", 
       highlight=cpg.highlight,
       highlight.text=genes.highlight, 
       file="jpg",
       memo="usc.sesame",
       #main="USC DS-ALL EWAS (SESAME)",
       dpi=500,file.output=TRUE,verbose=TRUE)

outcomes_model_qq <- outcomes_model_anno %>% 
  dplyr::select(cpg_name,p_val)%>%
  dplyr::mutate(ID=cpg_name, P=p_val)
lambda <- P_lambda(outcomes_model_qq$P)
tiff("usc.sesame.qq.tiff", width = 10, height = 7, units = 'in', res = 300)
qq(outcomes_model_qq$P, main=paste0("USC dataset \n lambda value= ",round(lambda,3)))
dev.off()

## BCM
bcm_model_anno <-fread("bcm/ds.all.bcm.ewas.anno.txt")  %>%
  dplyr::mutate(cpg_name=cpg,estimate=Beta, std_error=SE,p_val=P)

outcomes_model_anno <- bcm_model_anno %>% 
  filter(chr!="chrX"&chr!="chrY")
sig_threshold = 2.510822e-05
outcomes_model_CM_plot <- outcomes_model_anno%>%
  dplyr::mutate(p_val_cmplot = -log10(p_val))%>%
  dplyr::mutate(p_val_cmplot= ifelse(estimate<0,-p_val_cmplot, p_val_cmplot )) %>%
  dplyr::select(cpg_name, chr,pos, p_val_cmplot)
outcomes_model_CM_plot$chr <- gsub("chr","",outcomes_model_CM_plot$chr)
CMplot(outcomes_model_CM_plot,
       plot.type="m",
       band=0, 
       cex=c(0.5,0.5,0.5),
       amplify=FALSE,
       LOG10=FALSE,  
       signal.col=NULL,
       chr.den.col=NULL,
       threshold=c(log10(sig_threshold[[1]]), -log10(sig_threshold[[1]])),
       threshold.lty=2, 
       threshold.lwd=1, 
       threshold.col="black", 
       file="jpg",
       memo="bcm",
       main="BCM DS-ALL EWAS",
       dpi=500,file.output=TRUE,verbose=TRUE)

outcomes_model_qq <- outcomes_model_anno %>% 
  dplyr::select(cpg_name,p_val)%>%
  dplyr::mutate(ID=cpg_name, P=p_val)
lambda <- P_lambda(outcomes_model_qq$P)
tiff("bcm.qq.tiff", width = 10, height = 7, units = 'in', res = 300)
qq(outcomes_model_qq$P, main=paste0("BCM dataset \n lambda value= ",round(lambda,3)))
dev.off()

# meta
meta_model_anno <- read_rds("BCM_USC_ALL_DS_EWAS_MA.rds")

outcomes_model_anno <- meta_model_anno %>% 
  filter(chr!="chrX"&chr!="chrY",
         !is.na(p_val_fdr),
         is.finite(p_val_fdr)) 

sig_thresghold = outcomes_model_anno %>%
  dplyr::mutate(adjp = p_val_fdr)%>%
  dplyr::filter(adjp<0.05)%>%
  dplyr::arrange(desc(adjp)) %>%
  as.data.frame()
sig_threshold = sig_thresghold[1,"P-value"]

outcomes_model_CM_plot <- outcomes_model_anno%>%
  dplyr::mutate(p_val_cmplot = -log10(`P-value`))%>%
  dplyr::mutate(p_val_cmplot= ifelse(Effect<0,-p_val_cmplot, p_val_cmplot )) %>%
  dplyr::select(cpg, chr,pos, p_val_cmplot)
outcomes_model_CM_plot$chr <- gsub("chr","",outcomes_model_CM_plot$chr)
CMplot(outcomes_model_CM_plot,
       plot.type="m",
       band=0, 
       cex=c(0.5,0.5,0.5),
       amplify=FALSE,
       LOG10=FALSE,  
       signal.col=NULL,
       chr.den.col=NULL,
       threshold=c(log10(sig_threshold[[1]]), -log10(sig_threshold[[1]])),
       threshold.lty=2, 
       threshold.lwd=1, 
       threshold.col="black", 
       file="jpg",
       memo="meta",
       main="Meta DS-ALL EWAS",
       dpi=500,file.output=TRUE,verbose=TRUE)

outcomes_model_qq <- outcomes_model_anno %>% 
  dplyr::select(cpg,`P-value`)%>%
  dplyr::mutate(ID=cpg, P=`P-value`)
lambda <- P_lambda(outcomes_model_qq$P)
tiff("meta.qq.tiff", width = 10, height = 7, units = 'in', res = 300)
qq(outcomes_model_qq$P, main=paste0("Meta dataset \n lambda value= ",round(lambda,3)))
dev.off()

```

>  plot the cg27347265 beta values in a boxplot for DS-ALL cases and DS controls so we can see whether this association is driven by outliers or a general shift in methylation in DS-ALL

```{r}

setwd("~/DSALL/usc")
library(tidyverse)
library(data.table)

# in USC dataset

usc <- fread("ds.all.usc.sesame.ewas.anno.txt")
head(usc)
  
cg="cg27347265"
# ds and ds_all samples beta values and variables
outcomes3 <- fread("ds.all.covaraites.txt") %>%
  as.data.frame()%>%
  mutate(all_case=CaCo,ID=beadPosition)%>%
  dplyr::select(ID,all_case)
betas3_0 <- read_rds("~/sets/set3/beta_noob_set3.rds") %>% 
  t%>%
  as.data.frame()
betas3 = betas3_0[outcomes3$ID,cg,drop=FALSE] %>%
  rownames_to_column(var="ID") %>%
  inner_join(outcomes3,by="ID")

# Boxplot
betas3 %>%
  mutate(all_case=as.factor(all_case))%>%
  ggplot(aes(x=all_case,y=cg27347265))+
  geom_boxplot(outlier.size=0.2)+
  geom_jitter(shape=16, position=position_jitter(0.1))+
  theme_classic() +
  ggtitle("distribution of cg27347265 in ALL-DS and DS")+
  xlab("ALL Status") +
  ylab("Methylation beta values") +
  theme(plot.title = element_text(hjust = 0.5))
ggsave("cg27347265.dist.png")

setwd("~/DSALL/bcm")
library(tidyverse)
library(data.table)

# in BCM dataset
bcm = fread("bcm_cg27347265.txt") %>%
  as.data.frame()%>%
  mutate(all_case=as.factor(phenotype))

bcm%>%
  ggplot(aes(x=all_case,y=cg27347265))+
  geom_boxplot(outlier.size=0.2)+
  geom_jitter(shape=16, position=position_jitter(0.1))+
  theme_classic() +
  ggtitle("distribution of cg27347265 in ALL-DS and DS in baylor dataset")+
  xlab("ALL Status") +
  ylab("Methylation beta values") +
  theme(plot.title = element_text(hjust = 0.5))
ggsave("cg27347265.dist.bcm.png")

# 
# bcm%>%
#   filter(cg27347265<0.04)%>%
#   ggplot(aes(x=all_case,y=cg27347265))+
#   geom_boxplot(outlier.size=0.2)+
#   geom_jitter(shape=16, position=position_jitter(0.1))+
#   theme_classic() +
#   ggtitle("distribution of cg27347265 in ALL-DS and DS in baylor, remove outliers")+
#   xlab("ALL Status") +
#   ylab("Methylation beta values") +
#   theme(plot.title = element_text(hjust = 0.5))
# ggsave("cg27347265.dist.bcm.nooutlier.png")
# 
# # b cell proportions and cg27347265
# cellp <- read.csv("~/sets/set3/idol_deconvolution.csv",row.names = 1) %>%
#   select(beadPosition,Bcell) %>%
#   inner_join(betas3,by=c("beadPosition"="ID"))
# 
# cellp %>%
#   ggplot(aes(x=cg27347265,y=Bcell))+
#   geom_point()+
#   geom_smooth(method=lm, linetype="dashed")+
#   theme_classic() +
#   ggtitle("Correlation between b cell proportions and EBF1 methylation, in all subjects")
# ggsave("b.EBF1.all.dist.png")
# 
# cellp %>%
#   filter(all_case==0)%>%
#   ggplot(aes(x=cg27347265,y=Bcell))+
#   geom_point()+
#   geom_smooth(method=lm, linetype="dashed")+
#   theme_classic() +
#   ggtitle("Correlation between b cell proportions and EBF1 methylation, in DS subjects")
# ggsave("b.EBF1.ds.dist.png")
# 
# cellp %>%
#   filter(all_case==1)%>%
#   ggplot(aes(x=cg27347265,y=Bcell))+
#   geom_point()+
#   geom_smooth(method=lm, linetype="dashed")+
#   theme_classic() +
#   ggtitle("Correlation between b cell proportions and EBF1 methylation, in DS ALL subjects")
# ggsave("b.EBF1.dsall.dist.png")


```

> DMR analysis for USC and BCM meta dataset 

```{r comb-p file preparation}

setwd("~/DSALL")
#prep for comb-p and write out
outcomes_model <- read_rds("BCM_USC_ALL_DS_EWAS_MA.rds")%>%
  mutate(chrom=chr, end=pos, start=pos, p_val= `P-value`)%>%
  dplyr::select(chrom, start, end, p_val)%>%
  drop_na()%>%
  arrange(chrom, start)

outcomes_model%>%
  write_tsv("comb_p_BCM_USC_ALL_DS_EWAS_MA.txt")
```

```{bash run comb-p}

sftp lishaobo@discovery.usc.edu
cd /scratch/lishaobo/temps/berkeley/dsall
lcd ~/DSALL
put comb_p_BCM_USC_ALL_DS_EWAS_MA.txt

#run comb-p and set settings according to preferences
comb-p pipeline -c 4 --seed 0.05 --dist 1000 -p comb_p_BCM_USC_ALL_DS_EWAS_MA_output --region-filter-p 0.01 --region-filter-n 2 comb_p_BCM_USC_ALL_DS_EWAS_MA.txt

cat <(head -1 <(less comb_p_BCM_USC_ALL_DS_EWAS_MA_output.regions-p.bed.gz)) \
  <(tail -n +2 <(less comb_p_BCM_USC_ALL_DS_EWAS_MA_output.regions-p.bed.gz) | sort -k7,7g) > comb_p_BCM_USC_ALL_DS.txt

#setting --acf-dist to 0.33 * --dist == 330
#calculated stepsize as: 330
#ACF:
 #lag_min	lag_max	correlation	N	p
#1	331	0.08763	554509	0
#wrote: comb_p_BCM_USC_ALL_DS_EWAS_MA_output.acf.txt
# original lambda: 1.45
#wrote: comb_p_BCM_USC_ALL_DS_EWAS_MA_output.slk.bed.gz with lambda: 1.78
#wrote: comb_p_BCM_USC_ALL_DS_EWAS_MA_output.fdr.bed.gz
#wrote: comb_p_BCM_USC_ALL_DS_EWAS_MA_output.regions.bed.gz (1057 regions)
# read 1057 regions from comb_p_BCM_USC_ALL_DS_EWAS_MA_output.regions.bed.gz
# calculating ACF out to: 1552
#           with 6  lags: [1, 331, 661, 991, 1321, 1651]
# Done with one-time ACF calculation
#529009 bases used as coverage for sidak correction

get comb_p_BCM_USC_ALL_DS_EWAS_MA_output.regions-p.bed.gz
get comb_p_BCM_USC_ALL_DS.txt
```

```{r dmrcate analysis on the meta dataset}
library(DMRcate)
library(tidyverse)
library(limma)
library(plyr)

#set directories
setwd("~/DSALL")

#load annotation file 
library(IlluminaHumanMethylationEPICanno.ilm10b4.hg19)
data("IlluminaHumanMethylationEPICanno.ilm10b4.hg19")
anno       <- IlluminaHumanMethylationEPICanno.ilm10b4.hg19 %>%
              getAnnotation %>%
              as.data.frame %>%
              rownames_to_column(var="cpg")

#load metal outcomes
MA_outcomes <- fread("usc_bcm_meta_sd_both.txt")%>%
  left_join(anno, by=c("MarkerName"="cpg"))

#manipulate file for usability with DMRcate
MA_DMRcate <- MA_outcomes%>%
  dplyr::mutate(ID=MarkerName, stat=Effect/StdErr, CHR=chr, pos=pos, betafc= Effect,raw =`P-value`, indfdr= p.adjust(`P-value`, method="BH"))%>%
  dplyr::mutate(is.sig=ifelse(indfdr<0.05,TRUE,FALSE))%>%
  dplyr::select(ID, stat,CHR,pos,raw,betafc,indfdr,is.sig)

MA_DMRcate_list <- as.list(MA_DMRcate)
class(MA_DMRcate_list) <- "annot"

#it requires the old DMRcate (yes indeed annoying but it works)
dmrcate <- function (object, lambda = 1000, C = NULL, p.adjust.method = "BH", 
    pcutoff = "fdr", consec = FALSE, conseclambda = 10, betacutoff = NULL, 
    min.cpgs = 2, mc.cores = 1) {
    stopifnot(is(object, "annot"))
    stopifnot(lambda >= 1)
    stopifnot(pcutoff == "fdr" | (0 <= pcutoff & pcutoff <= 1))
    stopifnot(C >= 0.2)
    if (consec & is.null(conseclambda)) {
        stop("Consecutive CpG bandwidth must be specified")
    }
    object <- data.frame(ID = object$ID, weights = abs(object$stat), 
        CHR = as.character(object$CHR), pos = object$pos, betafc = object$betafc, 
        indfdr = object$indfdr, is.sig = object$is.sig)
    object <- object[order(object$CHR, object$pos), ]
    if (is.null(C) & !consec) {
        if (nrow(object) < 9e+05) {
            C = 2
        }
        else {
            C = 50
        }
    }
    if (consec) {
        lambda = conseclambda
        message(paste("Consecutive mode specified, lambda is now set at", 
            conseclambda, "consecutive CpGs."))
        if (is.null(C)) {
            stop("Error: argument C must be specified (in CpG sites) for consecutive mode.")
        }
        object$realcoordforconsec <- object$pos
        object$pos <- unlist(sapply(as.numeric(table(object$CHR)), 
            function(x) 1:x))
    }
    lag = lambda
    chr.unique <- unique(c(as.character(object$CHR)))
    fitted <- mclapply(chr.unique, fitParallel, object = object, 
        consec = consec, conseclambda = conseclambda, lambda = lambda, 
        C = C, mc.cores = mc.cores)
    object <- rbind.fill(fitted)
    object$fdr <- p.adjust(object$raw, method = p.adjust.method)
    if (pcutoff == "fdr") {
        nsig <- sum(object$is.sig)
        if (nsig == 0) {
            txt <- "The FDR you specified in cpg.annotate() returned no significant CpGs, hence there are no DMRs.\n    Try specifying a value of 'pcutoff' in dmrcate() and/or increasing 'fdr' in cpg.annotate()."
            stop(paste(strwrap(txt, exdent = 2), collapse = "\n"))
        }
        pcutoff <- sort(object$fdr)[nsig]
    }
    object$sig <- (object$fdr <= pcutoff)
    if (nrow(object) == 0) {
        txt <- "No signficant regions found. Try increasing the value of\n    'pcutoff' in dmrcate() and/or 'fdr' in cpg.annotate()."
        stop(paste(strwrap(txt, exdent = 2), collapse = "\n"))
    }
    message("Demarcating regions...")
    chr.N <- as.character(object$CHR)
    pos.N <- object$pos
    sig.N <- object$sig
    N <- length(sig.N)
    n.K <- which(sig.N)
    K <- length(n.K)
    stopifnot(K >= 2)
    pos.K <- pos.N[n.K]
    chr.K <- chr.N[n.K]
    jump_chr.k <- (chr.K[-1] != chr.K[-K])
    jump_pos.k <- (diff(pos.K) > lag)
    jump.k <- (jump_chr.k | jump_pos.k)
    ksegments.A2 <- Segment(jump.k)
    A <- nrow(ksegments.A2)
    kstart.A <- ksegments.A2[, "start"]
    kend.A <- ksegments.A2[, "end"]
    realpos.K <- pos.K
    if (consec) {
        realpos.N <- object$realcoordforconsec
        realpos.K <- realpos.N[n.K]
    }
    start.A <- realpos.K[kstart.A]
    end.A <- realpos.K[kend.A]
    chr.A <- chr.K[kstart.A]
    stopifnot(all(chr.K[kend.A] == chr.A))
    fmt <- "%s:%1d-%1d"
    coord.A <- sprintf(fmt, chr.A, start.A, end.A)
    nstart.A <- n.K[kstart.A]
    nend.A <- n.K[kend.A]
    width.A <- nend.A + 1 - nstart.A
    a.Z <- rep(seq(A), width.A)
    fn <- function(a) seq(from = nstart.A[a], to = nend.A[a])
    l.listA <- lapply(seq(A), fn)
    n.Z <- unlist(l.listA)
    region.N <- rep(NA_integer_, N)
    region.N[n.Z] <- a.Z
    levels <- seq(A)
    region.N <- factor(region.N, levels = levels)
    no_cpg.A <- c(table(region.N))
    REGIONSTAT <- function(field, fn) {
        x.N <- object[[field]]
        x.R <- tapply(x.N, region.N, fn)
        c(x.R)
    }
    fn_Stouffer <- function(x) pnorm(sum(qnorm(x))/sqrt(length(x)))
    fn_max <- function(x) x[which.max(abs(x))]
    results <- data.frame(coord = coord.A, no.cpgs = no_cpg.A, 
        minfdr = REGIONSTAT("fdr", min), Stouffer = REGIONSTAT("indfdr", 
            fn_Stouffer), maxbetafc = REGIONSTAT("betafc", fn_max), 
        meanbetafc = REGIONSTAT("betafc", mean), row.names = seq(A), 
        stringsAsFactors = FALSE)
    keep <- (results$no.cpgs >= min.cpgs)
    results <- results[keep, ]
    if (!is.null(betacutoff)) {
        keep <- (abs(results$meanbetafc) >= betacutoff)
        results <- results[keep, ]
    }
    o <- order(results$Stouffer, -results$no.cpgs)
    results <- results[o, ]
    message("Done!")
    output <- NULL
    output$input <- object
    output$results <- results
    output$cutoff <- pcutoff
    class(output) <- "dmrcate.output"
    output
}

#dmrcate 
dmrcoutput <- dmrcate(MA_DMRcate_list, lambda=1000, C=2, pcutoff = 0.01,min.cpgs = 2)

#save data
outcome_dmrcate <- dmrcoutput$results%>%
  write_rds("dmrcoutput_DS_ALL_BCM_USC_MA.rds")

```

```{r Compare overlap of comb-p and DMRcate on usc bcm meta dataset}
library(GenomicRanges)
library(tidyverse)
library(AnnotationHub)
library(TxDb.Hsapiens.UCSC.hg19.knownGene)
library(annotatr)
library(methyAnalysis)sssss
setwd("/ccls/home/sli/DSALL")

comb_p_output <- read_tsv("comb_p_BCM_USC_ALL_DS_EWAS_MA_output.regions-p.bed.gz")%>%
  dplyr::mutate(chrom=`#chrom`)%>%
  dplyr::select(-`#chrom`)%>%
  dplyr::select(chrom, everything())%>%
  dplyr::filter(z_sidak_p<0.01)

dmrcate_results <- read_rds("dmrcoutput_DS_ALL_BCM_USC_MA.rds")%>%
  dplyr::filter(minfdr<0.01)%>%
  separate(coord, c("chrom","coord"), sep =":")%>%
  separate(coord, c("start","end"), sep = "-")%>%
  dplyr::mutate(start=as.numeric(start),end=as.numeric(end))

comb_p_GR <- makeGRangesFromDataFrame(comb_p_output, keep.extra.columns=T,
                         ignore.strand=T,
                         seqinfo=NULL,
                         seqnames.field=c("chrom"),
                         start.field="start",
                         end.field=c("end"),
                         strand.field="strand",
                         starts.in.df.are.0based=FALSE)
DMRCate_GR <- makeGRangesFromDataFrame(dmrcate_results, keep.extra.columns=T,
                         ignore.strand=T,
                         seqinfo=NULL,
                         seqnames.field=c("chrom"),
                         start.field="start",
                         end.field=c("end"),
                         strand.field="strand",
                         starts.in.df.are.0based=FALSE)
overlap_DMRcate_comb_p_GR <- GenomicRanges::intersect(DMRCate_GR,comb_p_GR, ignore.strand=T)
length(overlap_DMRcate_comb_p_GR)
#63

if (require(TxDb.Hsapiens.UCSC.hg19.knownGene)) {
overlap_DMRcate_comb_p_DF <-annotateDMRInfo(overlap_DMRcate_comb_p_GR, 'TxDb.Hsapiens.UCSC.hg19.knownGene')
}

#left join with original files to get relevant meta data columns
overlap_DMRcate_comb_p_DF <- data.frame(overlap_DMRcate_comb_p_DF$sigDMRInfo)%>%
  left_join(comb_p_output, by=c("seqnames"="chrom","start","end"))%>%
  left_join(dmrcate_results,by=c("seqnames"="chrom","start","end"))
geneRanges <-
    function(db, column="ENTREZID"){
    g <- genes(db, columns=column)
    col <- mcols(g)[[column]]
    genes <- granges(g)[rep(seq_along(g), elementLengths(col))]
    mcols(genes)[[column]] <- as.character(unlist(col))
    genes
}
splitColumnByOverlap <-
    function(query, subject, column="ENTREZID", ...){
    olaps <- GenomicRanges::findOverlaps(query, subject, ...)
    f1 <- factor(subjectHits(olaps),
                 levels=seq_len(subjectLength(olaps)))
    splitAsList(mcols(query)[[column]][queryHits(olaps)], f1)
    }
annotate_grange <- function(grange_object, annotation_grange_object, column_name=NULL){
    file <- splitColumnByOverlap(annotation_grange_object,grange_object, column = column_name)
    grange_object$name <- file
    return(grange_object)}
annotate_grange <- function(grange_object, annotation_grange_object, column_name=NULL){
    file <- splitColumnByOverlap(annotation_grange_object,grange_object, column = column_name)
    grange_object$name <- file
    return(grange_object)}
overlap_DMRcate_comb_p_GR <- annotate_grange(overlap_DMRcate_comb_p_GR,DMRCate_GR, column_name="minfdr")
overlap_DMRcate_comb_p_GR$DMRCate_fdr <- overlap_DMRcate_comb_p_GR$name
overlap_DMRcate_comb_p_GR <- annotate_grange(overlap_DMRcate_comb_p_GR,DMRCate_GR, column_name="meanbetafc")
overlap_DMRcate_comb_p_GR$DMRCate_meanbetafc <- overlap_DMRcate_comb_p_GR$name
overlap_DMRcate_comb_p_GR <- annotate_grange(overlap_DMRcate_comb_p_GR,DMRCate_GR, column_name="no.cpgs")
overlap_DMRcate_comb_p_GR$DMRcate_N_cpgs <- overlap_DMRcate_comb_p_GR$name
overlap_DMRcate_comb_p_GR <- annotate_grange(overlap_DMRcate_comb_p_GR,comb_p_GR, column_name="z_sidak_p")
overlap_DMRcate_comb_p_GR$combp_sidak <- overlap_DMRcate_comb_p_GR$name
overlap_DMRcate_comb_p_GR <- annotate_grange(overlap_DMRcate_comb_p_GR,comb_p_GR, column_name="n_probes")
overlap_DMRcate_comb_p_GR$combp_N_cpgs <- overlap_DMRcate_comb_p_GR$name
overlap_DMRcate_comb_p_GR$name <- NULL
overlap_DMRcate_comb_p_GR$Gene_Symbol <- overlap_DMRcate_comb_p_DF$GeneSymbol
overlap_DMRcate_comb_p_GR$Promoter <- overlap_DMRcate_comb_p_DF$PROMOTER

overlap_DMRcate_comb_p_final <- overlap_DMRcate_comb_p_GR%>%
  as.data.frame()%>%
  dplyr::mutate(chrom=seqnames)%>%
  dplyr::select(-seqnames)%>%
  dplyr::select(chrom, start,end,width,Gene_Symbol,Promoter,everything())
write.csv(overlap_DMRcate_comb_p_final,"compDMARoverlap_comb_p_USC_BCM_MA.csv")
write_rds(overlap_DMRcate_comb_p_final,"compDMARoverlap_comb_p_USC_BCM_MA.rds")

```

> quadrant plot

```{r draw a quadrant plot to visualize the consistency of effects across all significant CpGs in USC data}

setwd("~/DSALL")
library(tidyverse)
library(data.table)

res = read_tsv("BCM_USC_ALL_sim.tsv") %>%
  as.data.frame() %>%
  dplyr::select(estimate_bcm,p_bcm,estimate_usc,p_usc)
nrow(res)
#186

res %>%
  ggplot(aes(x=estimate_usc,y=estimate_bcm))+
  geom_point()+
  geom_smooth(method=lm)+
  theme_bw()+
  labs(x="estimate usc",y="estimate bcm")+
  ggtitle("Correlation between estimates in USC and BCM probe beta values among USC significant probes")+
  theme(plot.title = element_text(hjust = 0.5))
ggsave("186.uscquadrantBeta.png",width = 8,height = 6)

cor.test(res$estimate_bcm,res$estimate_usc) 
# 	Pearson's product-moment correlation
# 
# data:  res$estimate_bcm and res$estimate_usc
# t = 5.8381, df = 184, p-value = 2.345e-08
# alternative hypothesis: true correlation is not equal to 0
# 95 percent confidence interval:
#  0.2666155 0.5101897
# sample estimates:
#       cor 
# 0.3953298 

```

----------------
M value, bacon, logistic regression

```{r M value, bacon, logistic regression}

setwd("~/DSALL/usc")

library(tidyverse)
library(data.table)
library(bacon)
library(lumi)

cl <- makeCluster(40, type="FORK")
clusterEvalQ(cl,library(MASS))
clusterEvalQ(cl,library(lmtest))

logisticDS <- function(meth, cov){

    # meth = methyl["cg25813447",]
    # cov = covs
  
    dat <- data.frame(meth = meth, cov) %>% dplyr::select(CaCo,meth,everything())
    ff  <- formula(paste("CaCo ~ ", paste0(colnames(dat)[-1], collapse="+")))
    mod <- try(glm(ff, data = dat, family=binomial))
    cf = try(summary(mod)$coefficients)
    Ncase = nrow(dat %>% filter(CaCo==1) %>% na.omit)
    Ncontrol = nrow(dat %>% filter(CaCo==0) %>% na.omit)

  	if (exists(class(cf)) == FALSE) {
  		out = c(rep(NA,4),Ncase,Ncontrol)
  	} else {
			coef = cf[2,"Estimate"]
			se = cf[2,"Std. Error"]
			z.value = cf[2,"z value"]
			p.value = cf[2,"Pr(>|z|)"]
			out = c(coef, se, z.value,p.value,Ncase,Ncontrol) 
    }
  names(out) = c("Beta","SE", "Z","P","ncase","ncontrol"); 
  return(out) 
  }

# Winsorize beta values
winsorize <- function(methylation,pct=winsorize.pct) {
  quantiles <- matrixStats::rowQuantiles(methylation, probs=c(pct,1-pct), na.rm=T)
  low <- quantiles[,1]
  upper <- quantiles[,2]

  outliers.lower <- rowSums(methylation < low, na.rm=T)
  outliers.upper <- rowSums(methylation > upper, na.rm=T)
  
  idx <- which(methylation < low, arr.ind=T)
  methylation[idx] <- low[idx[,1]]
  
  idx <- which(methylation > upper, arr.ind=T)
  methylation[idx] <- upper[idx[,1]]

  n <- rowSums(!is.na(methylation))
  log <- data.frame(outliers.lower, outliers.upper, n)
  return(list(methylation=methylation, log=log))
}

# prepare a file of subjects (called var) with deconvoluted cell proportions, sex and plate, 10 EpiStructure PCs

dcp = fread("~/sets/set3/idol_deconvolution.csv") %>%
  as.data.frame() %>%
  dplyr::select(-V1,-subjectId) %>%
  na.omit()

glint = fread("~/sets/set3/glintControl.csv") %>%
    dplyr::select(beadPosition, epi1, epi2, epi3, epi4, epi5, epi6, epi7, epi8, epi9, epi10)%>%
    na.omit()

vars = fread("~/sets/set3/clinical_variables.csv") %>%
  as.data.frame() %>%
  mutate(plate=Batch) %>%
  dplyr::filter(Trisomy21==1) %>%
  dplyr::select(beadPosition,subjectId,sex,plate,CaCo)%>%
  inner_join(dcp,by="beadPosition")%>%
  inner_join(glint,by="beadPosition")%>%
  na.omit() %>%
  filter(beadPosition!="202148010085_R03C01") # failed sex check


##scenario using Sesame
################
betas3_0 = fread("~/sets/set3/set3_Sesame_beta.csv") %>%
  as.data.frame()%>%
  dplyr::select(-V1)
betas3_1 = betas3_0 %>%
  column_to_rownames(var="probeId")

subsInTheStudy = intersect(vars$beadPosition,colnames(betas3_1))

noob.to.include = betas3_1[,subsInTheStudy] %>% as.matrix
noob.to.include = noob.to.include[rowMeans(is.na(noob.to.include))<0.05,]

noob.to.include.M = beta2m(noob.to.include)

replace.outliers <- winsorize(noob.to.include.M, 0.005)
methyl <- replace.outliers$methylation
outlier.log <- replace.outliers$log

vars = vars %>%
  column_to_rownames(var="beadPosition")

covs = vars[subsInTheStudy,] %>%
  dplyr::select(CaCo,sex,plate,CD8T,CD4T,NK,Bcell,Mono,nRBC,epi1,
                epi2,epi3,epi4,epi5,epi6,
                epi7,epi8,epi9,epi10)%>%
  mutate(plate=as.factor(plate))

# this has to be true to continue

if(identical(rownames(covs),colnames(methyl))==TRUE){

  # methylt = methyl[1:100,] 
  # results = t(apply(methylt,1,logisticDS,covs))
  
  results <- t(parApply(cl, methyl, 1, logisticDS, covs)) 
  results <- as.data.frame(results) %>%
    na.omit()
  
  results$bacon_P = pval(bacon(results$Z,na.exclude = TRUE))
  
  write.table(results,"ds.all.usc.sesame.m.logistic.ewas.txt",
        quote=FALSE,
        sep='\t')  

  
  # raw P values
  Nbeta <- sum(!is.na(results$P))
  tmp <- results[!is.na(results$P),]
  lambda = median(qchisq(1- tmp$P,1))/qchisq(0.5,1)
  print(round(lambda, 2))
  
  # bacon P values
  Nbeta <- sum(!is.na(results$bacon_P))
  tmp <- results[!is.na(results$bacon_P),]
  lambda = median(qchisq(1- tmp$bacon_P,1))/qchisq(0.5,1)
  print(round(lambda, 2))
}

stopCluster(cl)

# looking at the results
# lambda =  2.7
# lambda =  1.94

##########################################################################################
##########################################################################################
##########################################################################################

# looking at the results


#[1]lambda 2.55
#[1] lamda 1.72

setwd("~/DSALL/usc")
library(tidyverse)
library(data.table)
library(IlluminaHumanMethylationEPICanno.ilm10b4.hg19)
data("IlluminaHumanMethylationEPICanno.ilm10b4.hg19")
annoEPIC <- IlluminaHumanMethylationEPICanno.ilm10b4.hg19 %>%
  getAnnotation %>%
  as.data.frame %>%
  rownames_to_column(var="cpg") %>%
  dplyr::select(cpg,chr,pos,Relation_to_Island,UCSC_RefGene_Name,
                Probe_maf,UCSC_RefGene_Group,CpG_maf)


results <- fread("ds.all.usc.sesame.m.logistic.ewas.txt") %>% 
  dplyr::rename("cpg"="V1") %>%
  dplyr::filter(grepl('cg', cpg)) %>%
  filter(!is.na(P))%>% 
  left_join(annoEPIC,by="cpg") %>%
  arrange(by=P) %>% 
  dplyr::filter(chr!="chrX"&chr!="chrY") %>%  
  dplyr::filter(Probe_maf<0.05|is.na(Probe_maf)) %>% 
  dplyr::filter(CpG_maf<0.05|is.na(CpG_maf))  

results$bacon_P = pval(bacon(results$Z,na.exclude = TRUE))

results <- results %>%
  mutate(fdr=p.adjust(bacon_P, method = "fdr", n=length(bacon_P))) %>%
  mutate(bf=p.adjust(bacon_P, method = "bonferroni", n=length(bacon_P))) 

0.05/nrow(results) # bonferroni threshold
# 7.952109e-08

# raw P values
Nbeta <- sum(!is.na(results$P))
tmp <- results[!is.na(results$P),]
lambda = median(qchisq(1- tmp$P,1))/qchisq(0.5,1)
print(round(lambda, 2))

# bacon P values
Nbeta <- sum(!is.na(results$bacon_P))
tmp <- results[!is.na(results$bacon_P),]
lambda = median(qchisq(1- tmp$bacon_P,1))/qchisq(0.5,1)
print(round(lambda, 2))

# figures (qq plot and manhattan)
library(CMplot)
library(QCEWAS)
library(qqman)

usc_model_anno <- results %>%
  dplyr::mutate(cpg_name=cpg,estimate=Beta, std_error=SE,p_val=bacon_P)

sig_thresghold = usc_model_anno %>%
  filter(fdr<0.05)%>%
  arrange(desc(bacon_P))
sig_threshold = sig_thresghold[1,"p_val"]

annotation = usc_model_anno%>%
  filter(bf<0.05) %>%
  filter(UCSC_RefGene_Name!="") %>%
  mutate(gene.highlight = sub(";.*$","",UCSC_RefGene_Name))
cpg.highlight = annotation$cpg_name
genes.highlight = annotation$gene.highlight

outcomes_model_CM_plot <- usc_model_anno%>%
  dplyr::mutate(p_val_cmplot = -log10(p_val))%>%
  dplyr::mutate(p_val_cmplot= ifelse(estimate<0,-p_val_cmplot, p_val_cmplot )) %>%
  dplyr::select(cpg_name, chr,pos, p_val_cmplot)
outcomes_model_CM_plot$chr <- gsub("chr","",outcomes_model_CM_plot$chr)

CMplot(outcomes_model_CM_plot,
       plot.type="m",
       band=0, 
       cex=c(0.5,0.5,0.5),
       amplify=FALSE,
       LOG10=FALSE,  
       signal.col=NULL,
       chr.den.col=NULL,
       threshold=c(log10(sig_threshold[[1]]), -log10(sig_threshold[[1]])),
       threshold.lty=2, 
       threshold.lwd=1, 
       threshold.col="black", 
       highlight=cpg.highlight,
       highlight.text=genes.highlight, 
       file="jpg",
       memo="usc.sesame.m.logistic",
       #main="USC DS-ALL EWAS (SESAME)",
       dpi=500,file.output=TRUE,verbose=TRUE)

outcomes_model_qq <- usc_model_anno %>% 
  dplyr::select(cpg_name,p_val)%>%
  dplyr::mutate(ID=cpg_name, P=p_val)
lambda <- P_lambda(outcomes_model_qq$P)

png("usc.sesame.qq.m.logistic.png")
qq(outcomes_model_qq$P, main=paste0("USC dataset \n lambda value= ",round(lambda,3)))
dev.off()

## top CpGs
head(results)
dim(results%>%filter(fdr<0.05))
# 16
dim(results%>%filter(bf<0.05))
# 4

a =results%>%filter(fdr<0.05)
write.csv(a,"logi.sig.res.csv")

# pathway analysis
library(methylGSA)
library(IlluminaHumanMethylationEPICanno.ilm10b4.hg19)
library(IlluminaHumanMethylation450kanno.ilmn12.hg19)

down_all <- results %>%
  filter(fdr<0.05) %>%
  mutate(cpg_name=cpg,estimate=Beta,std_error=SE,p_val=bacon_P) %>%
  dplyr::select(cpg_name,chr,pos,estimate,std_error,p_val)

down_100_vec = as.numeric(down_all$p_val)
names(down_100_vec) = down_all$cpg_name
res1 = methylglm(cpg.pval = down_100_vec, GS.type = "KEGG")
res2 = methylglm(cpg.pval = down_100_vec, GS.type = "GO")

write.csv(res1,"down_M_LOGISTIC_100_KEGGpathway.sesame.csv")
write.csv(res2,"down_M_LOGISTIC_100_GOpathway.sesame.csv")

## Check RUNX1 and FLI1 to see if their methylation was different in DS with vs without ALL
results_lookup = results %>%
  filter(grepl("RUNX1|FLI1",UCSC_RefGene_Name)) 
write.table(results_lookup,"runx1.fli1.logistic.m.txt",
            quote=FALSE,
            row.names = FALSE,
            sep='\t')

## ToppFun within the ToppGene suite
## https://toppgene.cchmc.org/input_enrichment.jsp
# results_lookup = results %>%
#   filter(fdr<0.05) %>%
#   mutate(geneName = gsub(";.*$","",UCSC_RefGene_Name)) %>%
#   filter(geneName!="") %>%
#   distinct(geneName,.keep_all=TRUE)%>%
#   dplyr::select(geneName)
### NO significant disease

## DMR analysis (overlap of comb-p and DMRCATE)
### r run DMRCate in USC dataset}

setwd("~/DSALL/usc")
library(DMRcate)
library(tidyverse)
library(limma)
library(plyr)

MA_DMRcate <- results%>%
  dplyr::mutate(ID=cpg, stat=Beta/SE, CHR=chr, betafc=Beta,raw = bacon_P, indfdr = fdr)%>%
  dplyr::mutate(is.sig=ifelse(indfdr<0.05,TRUE,FALSE))%>%
  dplyr::select(ID, stat,CHR,pos,raw,betafc,indfdr,is.sig)%>%
  filter(!is.na(CHR))
MA_DMRcate_list <- as.list(MA_DMRcate)
class(MA_DMRcate_list) <- "annot"

dmrcate <- function (object, lambda = 1000, C = NULL, p.adjust.method = "BH", 
    pcutoff = "fdr", consec = FALSE, conseclambda = 10, betacutoff = NULL, 
    min.cpgs = 2, mc.cores = 1) {
    stopifnot(is(object, "annot"))
    stopifnot(lambda >= 1)
    stopifnot(pcutoff == "fdr" | (0 <= pcutoff & pcutoff <= 1))
    stopifnot(C >= 0.2)
    if (consec & is.null(conseclambda)) {
        stop("Consecutive CpG bandwidth must be specified")
    }
    object <- data.frame(ID = object$ID, weights = abs(object$stat), 
        CHR = as.character(object$CHR), pos = object$pos, betafc = object$betafc, 
        indfdr = object$indfdr, is.sig = object$is.sig)
    object <- object[order(object$CHR, object$pos), ]
    if (is.null(C) & !consec) {
        if (nrow(object) < 9e+05) {
            C = 2
        }
        else {
            C = 50
        }
    }
    if (consec) {
        lambda = conseclambda
        message(paste("Consecutive mode specified, lambda is now set at", 
            conseclambda, "consecutive CpGs."))
        if (is.null(C)) {
            stop("Error: argument C must be specified (in CpG sites) for consecutive mode.")
        }
        object$realcoordforconsec <- object$pos
        object$pos <- unlist(sapply(as.numeric(table(object$CHR)), 
            function(x) 1:x))
    }
    lag = lambda
    chr.unique <- unique(c(as.character(object$CHR)))
    fitted <- mclapply(chr.unique, fitParallel, object = object, 
        consec = consec, conseclambda = conseclambda, lambda = lambda, 
        C = C, mc.cores = mc.cores)
    object <- rbind.fill(fitted)
    object$fdr <- p.adjust(object$raw, method = p.adjust.method)
    if (pcutoff == "fdr") {
        nsig <- sum(object$is.sig)
        if (nsig == 0) {
            txt <- "The FDR you specified in cpg.annotate() returned no significant CpGs, hence there are no DMRs.\n    Try specifying a value of 'pcutoff' in dmrcate() and/or increasing 'fdr' in cpg.annotate()."
            stop(paste(strwrap(txt, exdent = 2), collapse = "\n"))
        }
        pcutoff <- sort(object$fdr)[nsig]
    }
    object$sig <- (object$fdr <= pcutoff)
    if (nrow(object) == 0) {
        txt <- "No signficant regions found. Try increasing the value of\n    'pcutoff' in dmrcate() and/or 'fdr' in cpg.annotate()."
        stop(paste(strwrap(txt, exdent = 2), collapse = "\n"))
    }
    message("Demarcating regions...")
    chr.N <- as.character(object$CHR)
    pos.N <- object$pos
    sig.N <- object$sig
    N <- length(sig.N)
    n.K <- which(sig.N)
    K <- length(n.K)
    stopifnot(K >= 2)
    pos.K <- pos.N[n.K]
    chr.K <- chr.N[n.K]
    jump_chr.k <- (chr.K[-1] != chr.K[-K])
    jump_pos.k <- (diff(pos.K) > lag)
    jump.k <- (jump_chr.k | jump_pos.k)
    ksegments.A2 <- Segment(jump.k)
    A <- nrow(ksegments.A2)
    kstart.A <- ksegments.A2[, "start"]
    kend.A <- ksegments.A2[, "end"]
    realpos.K <- pos.K
    if (consec) {
        realpos.N <- object$realcoordforconsec
        realpos.K <- realpos.N[n.K]
    }
    start.A <- realpos.K[kstart.A]
    end.A <- realpos.K[kend.A]
    chr.A <- chr.K[kstart.A]
    stopifnot(all(chr.K[kend.A] == chr.A))
    fmt <- "%s:%1d-%1d"
    coord.A <- sprintf(fmt, chr.A, start.A, end.A)
    nstart.A <- n.K[kstart.A]
    nend.A <- n.K[kend.A]
    width.A <- nend.A + 1 - nstart.A
    a.Z <- rep(seq(A), width.A)
    fn <- function(a) seq(from = nstart.A[a], to = nend.A[a])
    l.listA <- lapply(seq(A), fn)
    n.Z <- unlist(l.listA)
    region.N <- rep(NA_integer_, N)
    region.N[n.Z] <- a.Z
    levels <- seq(A)
    region.N <- factor(region.N, levels = levels)
    no_cpg.A <- c(table(region.N))
    REGIONSTAT <- function(field, fn) {
        x.N <- object[[field]]
        x.R <- tapply(x.N, region.N, fn)
        c(x.R)
    }
    fn_Stouffer <- function(x) pnorm(sum(qnorm(x))/sqrt(length(x)))
    fn_max <- function(x) x[which.max(abs(x))]
    results <- data.frame(coord = coord.A, no.cpgs = no_cpg.A, 
        minfdr = REGIONSTAT("fdr", min), Stouffer = REGIONSTAT("indfdr", 
            fn_Stouffer), maxbetafc = REGIONSTAT("betafc", fn_max), 
        meanbetafc = REGIONSTAT("betafc", mean), row.names = seq(A), 
        stringsAsFactors = FALSE)
    keep <- (results$no.cpgs >= min.cpgs)
    results <- results[keep, ]
    if (!is.null(betacutoff)) {
        keep <- (abs(results$meanbetafc) >= betacutoff)
        results <- results[keep, ]
    }
    o <- order(results$Stouffer, -results$no.cpgs)
    results <- results[o, ]
    message("Done!")
    output <- NULL
    output$input <- object
    output$results <- results
    output$cutoff <- pcutoff
    class(output) <- "dmrcate.output"
    output
}

dmrcoutput <- dmrcate(MA_DMRcate_list, lambda=1000, C=2, pcutoff = 0.01,min.cpgs = 2)
outcome_dmrcate <- dmrcoutput$results%>%
  write_rds("DMRCATE_USC_M_LOGIISTIC_VAL_DS_EWAS_MA_output.rds")

### comb-p 
usc <- results %>%
  dplyr::mutate(chrom=chr, end=pos+1, start=pos, p_val=bacon_P)%>%
  dplyr::filter(grepl('cg', cpg)) %>%
  dplyr::select(chrom, start, end, p_val)%>%
  na.omit()%>%
  arrange(chrom, start)

usc%>%
  write_tsv("comb_p_USC_SESAME_M_LOGISTIC_VAL_ALL_DS_EWAS_MA.txt")

```

```{bash run comb-p in USC dataset}

cd ~/DSALL/usc
#run comb-p and set settings according to preferences
comb-p pipeline -c 4 --seed 0.05 --dist 1000\
  -p comb_p_USC_M_LOGISTIC_VAL_DS_EWAS_MA_output\
  --region-filter-p 0.01 --region-filter-n 2 comb_p_USC_SESAME_M_LOGISTIC_VAL_ALL_DS_EWAS_MA.txt


cd ~/DSALL/usc
cat <(head -1 <(less comb_p_USC_M_LOGISTIC_VAL_DS_EWAS_MA_output.regions-p.bed.gz)) \
  <(tail -n +2 <(less comb_p_USC_M_LOGISTIC_VAL_DS_EWAS_MA_output.regions-p.bed.gz) | sort -k7,7g) \
  > comb_p_USC_M_LOGISTIC_VAL_DS_EWAS_MA_output.txt


```

```{r compare overlap of DMRcate and comb-p in USC datasset}

library(tidyverse)
library(AnnotationHub)
library(TxDb.Hsapiens.UCSC.hg19.knownGene)
library(GenomicRanges)
library(annotatr)
library(methyAnalysis)
setwd("~/DSALL/usc")

comb_p_output <- read_tsv("comb_p_USC_M_LOGISTIC_VAL_DS_EWAS_MA_output.txt")%>%
  dplyr::mutate(chrom=`#chrom`)%>%
  dplyr::select(-`#chrom`)%>%
  dplyr::select(chrom, everything())%>%
  dplyr::filter(z_sidak_p<0.01)
dmrcate_results <- read_rds("DMRCATE_USC_M_LOGIISTIC_VAL_DS_EWAS_MA_output.rds")%>%
  dplyr::filter(minfdr<0.01)%>%
  separate(coord, c("chrom","coord"), sep =":")%>%
  separate(coord, c("start","end"), sep = "-")%>%
  dplyr::mutate(start=as.numeric(start),end=as.numeric(end))
comb_p_GR <- makeGRangesFromDataFrame(comb_p_output, keep.extra.columns=T,
                         ignore.strand=T,
                         seqinfo=NULL,
                         seqnames.field=c("chrom"),
                         start.field="start",
                         end.field=c("end"),
                         strand.field="strand",
                         starts.in.df.are.0based=FALSE)
DMRCate_GR <- makeGRangesFromDataFrame(dmrcate_results, keep.extra.columns=T,
                         ignore.strand=T,
                         seqinfo=NULL,
                         seqnames.field=c("chrom"),
                         start.field="start",
                         end.field=c("end"),
                         strand.field="strand",
                         starts.in.df.are.0based=FALSE)
overlap_DMRcate_comb_p_GR <- GenomicRanges::intersect(DMRCate_GR,comb_p_GR, ignore.strand=T)
length(overlap_DMRcate_comb_p_GR)
#51

if (require(TxDb.Hsapiens.UCSC.hg19.knownGene)) {
overlap_DMRcate_comb_p_DF <-annotateDMRInfo(overlap_DMRcate_comb_p_GR, 'TxDb.Hsapiens.UCSC.hg19.knownGene')
}

#left join with original files to get relevant meta data columns
overlap_DMRcate_comb_p_DF <- data.frame(overlap_DMRcate_comb_p_DF$sigDMRInfo)%>%
  left_join(comb_p_output, by=c("seqnames"="chrom","start","end"))%>%
  left_join(dmrcate_results,by=c("seqnames"="chrom","start","end"))
geneRanges <-
    function(db, column="ENTREZID"){
    g <- genes(db, columns=column)
    col <- mcols(g)[[column]]
    genes <- granges(g)[rep(seq_along(g), elementLengths(col))]
    mcols(genes)[[column]] <- as.character(unlist(col))
    genes
}
splitColumnByOverlap <-
    function(query, subject, column="ENTREZID", ...){
    olaps <- GenomicRanges::findOverlaps(query, subject, ...)
    f1 <- factor(subjectHits(olaps),
                 levels=seq_len(subjectLength(olaps)))
    splitAsList(mcols(query)[[column]][queryHits(olaps)], f1)
    }
annotate_grange <- function(grange_object, annotation_grange_object, column_name=NULL){
    file <- splitColumnByOverlap(annotation_grange_object,grange_object, column = column_name)
    grange_object$name <- file
    return(grange_object)}
annotate_grange <- function(grange_object, annotation_grange_object, column_name=NULL){
    file <- splitColumnByOverlap(annotation_grange_object,grange_object, column = column_name)
    grange_object$name <- file
    return(grange_object)}
overlap_DMRcate_comb_p_GR <- annotate_grange(overlap_DMRcate_comb_p_GR,DMRCate_GR, column_name="minfdr")
overlap_DMRcate_comb_p_GR$DMRCate_fdr <- overlap_DMRcate_comb_p_GR$name
overlap_DMRcate_comb_p_GR <- annotate_grange(overlap_DMRcate_comb_p_GR,DMRCate_GR, column_name="meanbetafc")
overlap_DMRcate_comb_p_GR$DMRCate_meanbetafc <- overlap_DMRcate_comb_p_GR$name
overlap_DMRcate_comb_p_GR <- annotate_grange(overlap_DMRcate_comb_p_GR,DMRCate_GR, column_name="no.cpgs")
overlap_DMRcate_comb_p_GR$DMRcate_N_cpgs <- overlap_DMRcate_comb_p_GR$name
overlap_DMRcate_comb_p_GR <- annotate_grange(overlap_DMRcate_comb_p_GR,comb_p_GR, column_name="z_sidak_p")
overlap_DMRcate_comb_p_GR$combp_sidak <- overlap_DMRcate_comb_p_GR$name
overlap_DMRcate_comb_p_GR <- annotate_grange(overlap_DMRcate_comb_p_GR,comb_p_GR, column_name="n_probes")
overlap_DMRcate_comb_p_GR$combp_N_cpgs <- overlap_DMRcate_comb_p_GR$name
overlap_DMRcate_comb_p_GR$name <- NULL
overlap_DMRcate_comb_p_GR$Gene_Symbol <- overlap_DMRcate_comb_p_DF$GeneSymbol
overlap_DMRcate_comb_p_GR$Promoter <- overlap_DMRcate_comb_p_DF$PROMOTER
overlap_DMRcate_comb_p_final <- overlap_DMRcate_comb_p_GR%>%
  as.data.frame()%>%
  dplyr::mutate(chrom=seqnames)%>%
  dplyr::select(-seqnames)%>%
  dplyr::select(chrom, start,end,width,Gene_Symbol,Promoter,everything())
colnames(overlap_DMRcate_comb_p_final) = paste0("usc_",colnames(overlap_DMRcate_comb_p_final))
write.csv(overlap_DMRcate_comb_p_final,"compDMARoverlap_comb_p_USC_SESAME_M_LOGISTIC_VAL_ALL.csv")

```

##### Non-down results, not including

> correlation between arid5b methylation level and b cell proportion (not included in the paper)

```{r correlation between arid5b methylation level and b cell proportion}
setwd("~/DSALL/usc")
library(tidyverse)
library(data.table)

# extract ARID5B Methylation
clinical = fread('ds.all.covaraites.txt') 

# cg13344587 from subjects
methyl = read_rds("~/sets/set3/beta_noob_set3.rds")

uscCg = methyl["cg13344587",] %>% 
  as.data.frame() %>%
  rownames_to_column(var="beadPosition")
colnames(uscCg)[2] <- "cg13344587"

# merge dataset
corrBCG = clinical %>%
  inner_join(uscCg, by ="beadPosition") %>%
  mutate(cg13344587 = as.numeric(cg13344587))

## Correlation between B cell and methylation in all subjects
corrBCG%>%
  ggplot(aes(x=Bcell,y=cg13344587))+
  geom_point()+
  geom_smooth(method=lm)+
  theme_bw()+
  labs(x="Deconvoluted B cell proportions",y="Methylation of cg13344587")+
  ggtitle("Correlation between B cell proportions and ARID5B methylation")+
  theme(plot.title = element_text(hjust = 0.5))
ggsave("bcellARID5B.png",width = 8,height = 6)

cor.test(corrBCG$Bcell,corrBCG$cg13344587) 
# 	Pearson's product-moment correlation
# 
# data:  corrBCG$Bcell and corrBCG$cg13344587
# t = 4.5623, df = 341, p-value = 7.068e-06
# alternative hypothesis: true correlation is not equal to 0
# 95 percent confidence interval:
#  0.1374476 0.3371832
# sample estimates:
#      cor 
# 0.239852 

## Correlation between B cell and methylation in DS-nonALL
corrBCG1 = corrBCG %>%
  filter(CaCo==0)

corrBCG1%>%
  ggplot(aes(x=Bcell,y=cg13344587))+
  geom_point()+
  geom_smooth(method=lm)+
  theme_bw()+
  labs(x="Deconvoluted B cell proportions",y="Methylation of cg13344587")+
  ggtitle("Correlation between B cell proportions and ARID5B methylation in Down subjects")+
  theme(plot.title = element_text(hjust = 0.5))
ggsave("bcellARID5Bco.png",width = 8,height = 6)

cor.test(corrBCG1$Bcell,corrBCG1$cg13344587) 
# 	Pearson's product-moment correlation
# 
# data:  corrBCG1$Bcell and corrBCG1$cg13344587
# t = 3.178, df = 196, p-value = 0.001723
# alternative hypothesis: true correlation is not equal to 0
# 95 percent confidence interval:
#  0.08453532 0.35000498
# sample estimates:
#       cor 
# 0.2213674 
       
## Correlation between B cell and methylation in DS-ALL
corrBCG1 = corrBCG %>%
  filter(CaCo==1)

corrBCG1%>%
  ggplot(aes(x=Bcell,y=cg13344587))+
  geom_point()+
  geom_smooth(method=lm)+
  theme_bw()+
  labs(x="Deconvoluted B cell proportions",y="Methylation of cg13344587")+
  ggtitle("Correlation between B cell proportions and ARID5B methylation in Down ALL subjects")+
  theme(plot.title = element_text(hjust = 0.5))
ggsave("bcellARID5Bca.png",width = 8,height = 6)

cor.test(corrBCG1$Bcell,corrBCG1$cg13344587) 
# 	Pearson's product-moment correlation
# 
# data:  corrBCG1$Bcell and corrBCG1$cg13344587
# t = 3.7743, df = 143, p-value = 0.0002344
# alternative hypothesis: true correlation is not equal to 0
# 95 percent confidence interval:
#  0.1450971 0.4422959
# sample estimates:
#       cor 
# 0.3009873 

```

```{r mediation analysis to see if ARID5B mediates B cell proportions in ALL}

setwd("~/DSALL/usc")
library(tidyverse)
library(data.table)
library(mediation)
set.seed(1990)

clinical = fread('ds.all.covaraites.txt') 

methyl = fread("~/sets/set3/set3_Sesame_beta.csv") 

uscBeta = methyl %>%
  dplyr::select(-V1) %>%
  dplyr::filter(probeId=="cg13344587")%>%
  column_to_rownames(var="probeId") %>%
  t()%>%
  as.data.frame() %>%
  rownames_to_column(var="beadPosition")

cg = uscBeta%>%
  inner_join(clinical,by="beadPosition") %>%
  na.omit()

med.fit <- lm(cg13344587 ~ CaCo + sex  + plate + epi1 + 
                epi2 + epi3 + epi4 + epi5 + epi6, data = cg)
out.fit <- lm( Bcell ~ cg13344587 + CaCo + sex + plate + epi1 + 
                epi2 + epi3 + epi4 + epi5 + epi6, data = cg)
med.out <- mediate(med.fit, out.fit, treat = "CaCo", mediator = "cg13344587")
summary(med.out)

# Causal Mediation Analysis 
# 
# Quasi-Bayesian Confidence Intervals
# 
#                Estimate 95% CI Lower 95% CI Upper p-value    
# ACME           -0.00013     -0.00104         0.00    0.74    
# ADE             0.00534      0.00257         0.01  <2e-16 ***
# Total Effect    0.00521      0.00230         0.01  <2e-16 ***
# Prop. Mediated -0.02545     -0.28658         0.13    0.74    
# ---
# Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1
# 
# Sample Size Used: 343 
# 
# 
# Simulations: 1000 

med.fit.1 <- lm(cg13344587 ~ Bcell + sex  + plate + epi1 + 
                epi2 + epi3 + epi4 + epi5 + epi6, data = cg)
out.fit.1 <- glm( CaCo ~ cg13344587 + Bcell + sex + plate  + epi1 + 
                epi2 + epi3 + epi4 + epi5 + epi6, data = cg, family="binomial")

med.out.1 <- mediate(med.fit.1, out.fit.1, treat = "Bcell", 
                   mediator = "cg13344587")
summary(med.out.1)

# Causal Mediation Analysis 
# 
# Quasi-Bayesian Confidence Intervals
# 
#                           Estimate 95% CI Lower 95% CI Upper p-value    
# ACME (control)           -2.79e-01    -4.07e-01         0.21   0.124    
# ACME (treated)           -3.13e-04    -4.23e-06         0.00   0.074 .  
# ADE (control)             6.41e-01     5.81e-01         0.70  <2e-16 ***
# ADE (treated)             9.20e-01     4.12e-01         1.00  <2e-16 ***
# Total Effect              6.41e-01     5.80e-01         0.70  <2e-16 ***
# Prop. Mediated (control) -5.08e-01    -6.99e-01         0.33   0.124    
# Prop. Mediated (treated) -1.20e-14    -7.45e-06         0.00   0.074 .  
# ACME (average)           -1.40e-01    -2.04e-01         0.11   0.124    
# ADE (average)             7.81e-01     5.25e-01         0.84  <2e-16 ***
# Prop. Mediated (average) -2.54e-01    -3.50e-01         0.17   0.124    
# ---
# Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1
# 
# Sample Size Used: 343 
# 
# 
# Simulations: 1000 




```

```{r For the top USC DMRs, we can plot the mean beta values across the DMR CpGs in the BCM data. Perhaps start with the few DMRs (e.g. HOPX, SMIM24) in which several CpGs showed significance in the BCM data}

setwd("~/DSALL")
library(tidyverse)
library(data.table)

# read USC dmr results
uscDMR = read.csv("usc/compDMARoverlap_comb_p_USC_SESAME_ALL.csv", row.names=1) %>%
  mutate(dmrID = paste0(usc_chrom,"_",usc_start,"_",usc_end))%>%
  dplyr::select(-usc_chrom, -usc_start, -usc_end,usc_strand)

# read BCM CpGs results and locations
cpgsBaylor = read.csv("CpGswithUSCdmr.csv") %>%
  mutate(dmrID = paste0(chrom,"_",start,"_",end))%>%
  arrange(chrom,start,end)%>%
  select(dmrID,cpg_name,estimate_bcm,p_val_bcm,chrom,start,end)

# read BCM raw 
rawBaylor = fread("bcmExtract_cpgbeta.txt") %>%
  as.data.frame()

rawCase = rawBaylor %>% filter(phenotype==1) %>% colMeans
rawControl = rawBaylor %>% filter(phenotype==0) %>% colMeans

aves = data.frame(cpg_name = colnames(rawBaylor),caseM = rawCase, controlM = rawControl)
rownames(aves)=c()  

# a data frame icluding USC dmr results and UCB raw beta data

rawDMR = cpgsBaylor %>%
  left_join(aves,by="cpg_name") %>%
  left_join(uscDMR,by="dmrID")

chrIDFreq =  table(rawDMR$dmrID) %>% as.data.frame()
chrIDF1 = chrIDFreq %>% filter(Freq==1) %>% select(Var1)

## deal with DMRs where there is only 1 CpG from BCM first

dmr1time = rawDMR %>%
  arrange(p_val_bcm)%>%
  filter(dmrID %in% chrIDF1$Var1)
# write.csv(dmr1time,"1time.bcm.uscDMR.csv")

dmr1time%>%
  arrange(p_val_bcm)%>%
  select(cpg_name, caseM, controlM,p_val_bcm) %>%
  gather(key = caco, value = betaM,c("caseM","controlM"))%>%
  mutate(caco=as.factor(caco))%>%
  ggplot(aes(x=cpg_name,y=betaM))+
  geom_line(aes(colour=caco,group = caco))+
  scale_color_manual(values = c("darkred", "steelblue"))+
  theme_classic() +
  ggtitle("USC DMRs with only 1 BCM CpG")+
  xlab("CpGs, each CpG belongs to a different DMR") +
  ylab("Raw mean methylation beta levels") +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))+
  theme(plot.title = element_text(hjust = 0.5)) 
ggsave("rawBeta1CpG.png",width = 8,height = 5)


## for DMRs with multiple CPGs present from UCB
## There are 80 such DMRs
chrIDF1P = chrIDFreq %>% filter(Freq>1) 

sigs = rawDMR %>% filter(p_val_bcm<0.05) %>% select(cpg_name)


for(dmr in chrIDF1P$Var1){

  ##
#  dmr="chr1_2058230_2059086" 
  ##
  
  slice = rawDMR %>% filter(dmrID==dmr) %>%
    arrange(chrom, start, end)
  
  gene = slice$usc_Gene_Symbol[1]
  p_DMRcate = formatC(slice$usc_DMRCate_fdr[1], format = "e", digits = 4)
  b_DMRcate = round(slice$usc_DMRCate_meanbetafc[1],5)
  p_combp = slice$usc_combp_sidak[1]
  dmrID = slice$dmrID[1]
  
  sigsInd = which(slice$cpg_name %in% sigs$cpg_name)  
  colX = c(rep("black",nrow(slice)))
  colX[sigsInd] = "red"
  colX = as.factor(colX)
  
  slice1 = slice %>%
    select(caseM,controlM,cpg_name) %>%
    gather(key = caco, value = betaM,c("caseM","controlM"))%>%
    mutate(caco=as.factor(caco))
  
  slice1 %>%
    ggplot(aes(x=cpg_name,y=betaM))+
    geom_line(aes(colour=caco,group = caco))+
    scale_color_manual(values = c("darkred", "steelblue"))+
    theme_classic() +
    ggtitle(paste0("DMR in ",gene, "; DMRcate P value:",p_DMRcate,"; DMRcate FC:",b_DMRcate))+
    xlab(paste0("CpGs in ",dmrID," from Baylor dataset")) +
    ylab("Raw mean methylation beta levels") +
    theme(axis.text.x = element_text(angle = 30, hjust = 1,colour = colX))+
    theme(plot.title = element_text(hjust = 0.5)) 
  ggsave(paste0("rawB_",dmrID,".png"),width = 8,height = 5)

  }




```

> downs and non-downs ALL ewas analysis, lookup

```{r  downs and non-downs ALL ewas analysis, lookup}

setwd("~/DSALL/usc/keren")
library(tidyverse)
library(data.table)
library(methylGSA)

############################################################################################
# Set234 /scratch/lishaobo/downSyndrome/metafor_set234_anno.txt
## look up top DSs in Down ALLs

###############NOOB
nonAllTop = fread("metafor_set234_anno.txt") %>%
  as.data.frame()%>%
  mutate(cpg_name=probe, Estimate=coef.fe, `Pr(>|z|)` = p.fe) %>%
  arrange(`Pr(>|z|)`) %>%
  dplyr::select(cpg_name,Estimate,`Pr(>|z|)`,chr,pos,Islands_Name,
                Relation_to_Island,UCSC_RefGene_Name) %>% 
  dplyr::rename("nonall_beta"="Estimate", "noall_p"="Pr(>|z|)")

usc <- fread("~/DSALL/usc/ds.all.usc.ewas.anno.txt") %>%
  mutate(cpg_name=cpg,estimate=Beta,std_error=SE,p_val=P) %>%
  dplyr::select(cpg_name,estimate,std_error,p_val)%>%
  right_join(nonAllTop,by=c("cpg_name")) %>%
  filter(!is.na(estimate)) %>%
  arrange(noall_p) %>%
  head(100)

write.table(usc,"CpGLookUpNonAll234.txt",sep = "\t", row.names = FALSE, quote = FALSE)

## look up top DS-ALLs in Non Down ALLs
nonall = fread("metafor_set234_anno.txt") %>%
  as.data.frame()%>%
  mutate(cpg_name=probe, Estimate=coef.fe, `Pr(>|z|)` = p.fe) %>%
  dplyr::rename("nonall_beta"="Estimate", "noall_p"="Pr(>|z|)") %>%
  dplyr::select(cpg_name,nonall_beta,noall_p) 

usc <- fread("~/DSALL/usc/ds.all.usc.ewas.anno.txt") %>%
  mutate(cpg_name=cpg,estimate=Beta,std_error=SE,p_val=P) %>%
  dplyr::select(cpg_name,estimate,std_error,p_val,chr,pos,Relation_to_Island,UCSC_RefGene_Name,UCSC_RefGene_Group)%>%
  arrange(p_val)
usc1 = usc %>%  
  left_join(nonall,by=c("cpg_name")) %>%
  arrange(p_val) %>%
  filter(!is.na(nonall_beta)) %>%
  head(100)
write.table(usc1,"CpGLookUpDSAll234.txt",sep = "\t", row.names = FALSE, quote = FALSE)

###############SESAME
nonAllTop = fread("metafor_set234_anno.txt") %>%
  as.data.frame()%>%
  mutate(cpg_name=probe, Estimate=coef.fe, `Pr(>|z|)` = p.fe) %>%
  arrange(`Pr(>|z|)`) %>%
  dplyr::select(cpg_name,Estimate,`Pr(>|z|)`,chr,pos,Islands_Name,
                Relation_to_Island,UCSC_RefGene_Name) %>% 
  dplyr::rename("nonall_beta"="Estimate", "noall_p"="Pr(>|z|)")

usc <- fread("~/DSALL/usc/ds.all.usc.sesame.ewas.anno.txt") %>%
  mutate(cpg_name=cpg,estimate=Beta,std_error=SE,p_val=P) %>%
  dplyr::select(cpg_name,estimate,std_error,p_val)%>%
  right_join(nonAllTop,by=c("cpg_name")) %>%
  filter(!is.na(estimate)) %>%
  arrange(noall_p) %>%
  head(100)

write.table(usc,"CpGLookUpNonAll234.sesame.txt",sep = "\t", row.names = FALSE, quote = FALSE)

## look up top DS-ALLs in Non Down ALLs
nonall = fread("metafor_set234_anno.txt") %>%
  as.data.frame()%>%
  mutate(cpg_name=probe, Estimate=coef.fe, `Pr(>|z|)` = p.fe) %>%
  dplyr::rename("nonall_beta"="Estimate", "noall_p"="Pr(>|z|)") %>%
  dplyr::select(cpg_name,nonall_beta,noall_p) 

usc <- fread("~/DSALL/usc/ds.all.usc.sesame.ewas.anno.txt") %>%
  mutate(cpg_name=cpg,estimate=Beta,std_error=SE,p_val=P) %>%
  dplyr::select(cpg_name,estimate,std_error,p_val,chr,pos,Relation_to_Island,UCSC_RefGene_Name,UCSC_RefGene_Group)%>%
  arrange(p_val)
usc1 = usc %>%  
  left_join(nonall,by=c("cpg_name")) %>%
  arrange(p_val) %>%
  filter(!is.na(nonall_beta)) %>%
  head(100)
write.table(usc1,"CpGLookUpDSAll234.sesame.txt",sep = "\t", row.names = FALSE, quote = FALSE)


```

>> DMR analysis in Keren's non Down ALL EWAS (Sets 2,3,4) comb-P

```{r DMRs in the USC non DS ALL EWAS}

library(tidyverse)
library(data.table)
setwd("~/DSALL/usc/keren")

#prep for comb-p and write out
usc <- fread("metafor_set234_anno.txt") %>%
  mutate(chrom=chr, end=pos, start=pos, p_val=adjP)%>%
  dplyr::select(chrom, start, end, p_val)%>%
  drop_na()%>%
  arrange(chrom, start)

usc%>%
  write_tsv("comb_p_USC_ALL_nonDS_234_EWAS.txt")
```

```{bash run comb-p in non DS ALL EWAS on USC HPC}
cd /scratch/lishaobo/downSyndrome
#source python 2
#source /usr/usc/python/2.7.8/setup.sh
#run comb-p and set settings according to preferences
comb-p pipeline -c 4 --seed 0.05 --dist 1000 -p comb_p_USC_ALL_nonDS_234_EWAS_output --region-filter-p 0.01 --region-filter-n 2 comb_p_USC_ALL_nonDS_234_EWAS.txt
# (0 regions)
```

```{r cell proportions controlling for epi structures AND ARID5B cg13344587}

library(broom)
library(tidyverse)

setwd("~/DSALL/usc")

clinical <-  fread('ds.all.covaraites.txt') %>% 
  as.data.frame()


#methyl = read_rds("~/sets/set3/beta_noob_set3.rds")

methyl = fread("~/sets/set3/set3_Sesame_beta.csv") 

methyl_pick = methyl %>%
  dplyr::select(-V1)%>%
  dplyr::filter(probeId=="cg13344587")%>%
  column_to_rownames(var="probeId") %>%
  t()%>%
  as.data.frame() %>%
  rownames_to_column(var="beadPosition")

subs_cellproportion_0 <- clinical%>%
  inner_join(methyl_pick,by="beadPosition") 


cellP <- data.frame()
j = 1
for(i in  c("CD8T","CD4T","NK","Bcell","Mono","Gran","nRBC")){
  
  smokingDF <- subs_cellproportion_0

  lm1 <- lm( smokingDF[,i] ~ smokingDF$CaCo + smokingDF$cg13344587 + smokingDF$sex + smokingDF$plate + smokingDF$epi1 + smokingDF$epi2 + smokingDF$epi3+ smokingDF$epi4 + smokingDF$epi5 + smokingDF$epi6)%>%
  tidy() %>% as.data.frame()
  
  cellP[j , 1] <- i
  cellP[j , 2] <- lm1[2,2]
  cellP[j , 3] <- lm1[2,3]
  cellP[j , 4] <- lm1[2,4]
  cellP[j , 5] <- lm1[2,5]
  j = j+1
 
  }

colnames(cellP) <- c("cell_name","estimate_6epi","stderror_6epi","stat_6epi","pvalue_6epi")

write.csv(cellP,"usc_B_reg_lin_epi6_cg13344587.csv")

```

>  B cell proportion comparisons in nonDS ALL and controls (CCLS)

```{r B cell proportion comparison in Set 2 3 and 4 ALL control subjects, linear regression}

# linear regression

library(broom)
library(tidyverse)

setwd("~/DSALL/usc/keren")

# set 2
clinical <- read.csv('~/sets/set2/clinical_variables.csv') 
ID_cli <- clinical %>% 
  mutate(plate=paste0("2_",gsub("_.*$","",beadPosition)),
       set=2) %>%
  filter(Trisomy21.cnm==0)%>%
  dplyr::select(CaCo,beadPosition,plate,sex) %>% 
  na.omit() 

glint <- read.csv('~/sets/set2/glintControl.csv',row.names = 1)
cellp <- read.csv("~/sets/set2/idol_deconvolution.csv",row.names = 1)
subs_cellproportion_0 <- inner_join(ID_cli,cellp,by="beadPosition") %>%
  inner_join(glint,by="beadPosition")
cellP <- data.frame()
j = 1
for(i in  c("CD8T","CD4T","NK","Bcell","Mono","Gran","nRBC")){
  smokingDF <- subs_cellproportion_0
  lm1 <- lm( smokingDF[,i] ~ smokingDF$CaCo + smokingDF$sex + as.factor(smokingDF$plate) + smokingDF$epi1 + smokingDF$epi2 + smokingDF$epi3+ smokingDF$epi4 + smokingDF$epi5 + smokingDF$epi6)%>%
  tidy() %>% as.data.frame()
  cellP[j , 1] <- i
  cellP[j , 2] <- lm1[2,2]
  cellP[j , 3] <- lm1[2,3]
  cellP[j , 4] <- lm1[2,4]
  cellP[j , 5] <- lm1[2,5]
  j = j+1
  }
colnames(cellP) <- c("cell_name","estimate_6epi","stderror_6epi","stat_6epi","pvalue_6epi")

write.table(cellP,"set2_reg_lin_epi6.txt", 
            sep = "\t", row.names = FALSE, quote = FALSE)

# set 3 
clinical <- read.csv('~/sets/set3/clinical_variables.csv') 
ID_cli <- clinical %>% 
    mutate(plate=paste0("3_",gsub("_.*$","",beadPosition)),
           set=3) %>%
  dplyr::select(CaCo,beadPosition,plate,sex,Trisomy21) %>% 
  filter(Trisomy21==0) %>% 
  na.omit() 
glint <- read.csv('~/sets/set3/glintControl.csv',row.names = 1)
cellp <- read.csv("~/sets/set3/idol_deconvolution.csv",row.names = 1)
subs_cellproportion_0 <- inner_join(ID_cli,cellp,by="beadPosition") %>% 
  dplyr::select(-Trisomy21) %>%
  inner_join(glint,by="beadPosition")
cellP <- data.frame()
j = 1
for(i in  c("CD8T","CD4T","NK","Bcell","Mono","Gran","nRBC")){
  smokingDF <- subs_cellproportion_0
  lm1 <- lm( smokingDF[,i] ~ smokingDF$CaCo + smokingDF$sex + as.factor(smokingDF$plate) + smokingDF$epi1 + smokingDF$epi2 + smokingDF$epi3+ smokingDF$epi4 + smokingDF$epi5 + smokingDF$epi6)%>%
  tidy() %>% as.data.frame()
  cellP[j , 1] <- i
  cellP[j , 2] <- lm1[2,2]
  cellP[j , 3] <- lm1[2,3]
  cellP[j , 4] <- lm1[2,4]
  cellP[j , 5] <- lm1[2,5]
  j = j+1
  }
colnames(cellP) <- c("cell_name","estimate_6epi","stderror_6epi","stat_6epi","pvalue_6epi")

write.table(cellP,"set3_reg_lin_epi6.txt", 
            sep = "\t", row.names = FALSE, quote = FALSE)

# set 4
clinical <- read.csv('~/sets/set4/clinical_variables.csv') 
ID_cli <- clinical %>% 
  mutate(plate=paste0("4_",gsub("_.*$","",beadPosition)),
       set=4) %>%
  filter(Trisomy21.cnm==0,
         smp_type=="bg")%>%
  dplyr::select(CaCo,beadPosition,plate,sex,Trisomy21.cnm) %>% 
  filter(Trisomy21.cnm==0) %>% 
  na.omit() 
glint <- read.csv('~/sets/set4/glintControl.csv',row.names = 1)
cellp <- read.csv("~/sets/set4/idol_deconvolution.csv",row.names = 1)
subs_cellproportion_0 <- inner_join(ID_cli,cellp,by="beadPosition") %>% 
  dplyr::select(-Trisomy21.cnm) %>%
  inner_join(glint,by="beadPosition")
cellP <- data.frame()
j = 1
for(i in  c("CD8T","CD4T","NK","Bcell","Mono","Gran","nRBC")){
  smokingDF <- subs_cellproportion_0
  lm1 <- lm( smokingDF[,i] ~ smokingDF$CaCo + smokingDF$sex + as.factor(smokingDF$plate) + smokingDF$epi1 + smokingDF$epi2 + smokingDF$epi3+ smokingDF$epi4 + smokingDF$epi5 + smokingDF$epi6)%>%
  tidy() %>% as.data.frame()
  cellP[j , 1] <- i
  cellP[j , 2] <- lm1[2,2]
  cellP[j , 3] <- lm1[2,3]
  cellP[j , 4] <- lm1[2,4]
  cellP[j , 5] <- lm1[2,5]
  j = j+1
  }
colnames(cellP) <- c("cell_name","estimate_6epi","stderror_6epi","stat_6epi","pvalue_6epi")
write.table(cellP,"set4_reg_lin_epi6.txt", 
            sep = "\t", row.names = FALSE, quote = FALSE)


##########################################################################################
#B cell proportion comparison in Set 2 3 and 4 ALL control subjects, mixed effect}

# linear regression
library(broom)
library(tidyverse)
library(lme4)
library(lmerTest)

setwd("~/DSALL/usc/keren")

# set 2
clinical <- read.csv('~/sets/set2/clinical_variables.csv') 
ID_cli <- clinical %>% 
  mutate(plate=paste0("2_",gsub("_.*$","",beadPosition)),
       set=2) %>%
  filter(Trisomy21.cnm==0)%>%
  dplyr::select(CaCo,beadPosition,plate,sex,set) %>% 
  na.omit() 

glint <- read.csv('~/sets/set2/glintControl.csv',row.names = 1)
cellp <- read.csv("~/sets/set2/idol_deconvolution.csv",row.names = 1) %>%
  dplyr::select(-subjectId)
subs_cellproportion_2 <- inner_join(ID_cli,cellp,by="beadPosition") %>%
  inner_join(glint,by="beadPosition")

# set 3 
clinical <- read.csv('~/sets/set3/clinical_variables.csv') 
ID_cli <- clinical %>% 
    filter(Trisomy21==0) %>% 
    mutate(plate=paste0("3_",gsub("_.*$","",beadPosition)),
           set=3) %>%
  dplyr::select(CaCo,beadPosition,plate,sex,set) %>% 
  na.omit() 
glint <- read.csv('~/sets/set3/glintControl.csv',row.names = 1)
cellp <- read.csv("~/sets/set3/idol_deconvolution.csv",row.names = 1) %>%
  dplyr::select(-subjectId)
subs_cellproportion_3 <- inner_join(ID_cli,cellp,by="beadPosition") %>% 
  inner_join(glint,by="beadPosition")

# set 4
clinical <- read.csv('~/sets/set4/clinical_variables.csv') 
ID_cli <- clinical %>% 
  filter(Trisomy21.cnm==0,
         smp_type=="bg") %>% 
  mutate(plate=paste0("4_",gsub("_.*$","",beadPosition)),
       set=4) %>%
  dplyr::select(CaCo,beadPosition,plate,sex,set) %>% 
  na.omit() 
glint <- read.csv('~/sets/set4/glintControl.csv',row.names = 1)
cellp <- read.csv("~/sets/set4/idol_deconvolution.csv",row.names = 1)
subs_cellproportion_4 <- inner_join(ID_cli,cellp,by="beadPosition") %>% 
  inner_join(glint,by="beadPosition")

clinicals <- rbind(subs_cellproportion_2,subs_cellproportion_3,subs_cellproportion_4) %>%
  as.data.frame()%>% 
  mutate(set= as.numeric(as.factor(set)),
         plate=as.factor(plate)) %>%
  na.omit()

cells <-c("Bcell","CD4T","CD8T", "Gran", "Mono", "NK","nRBC")
cellMixEffect = list()
k=1

for(j in c(1:length(cells))){
    
  clinical1 <- clinicals %>% 
    dplyr::select(one_of(cells)[j] ,CaCo,sex,plate,epi1,epi2,epi3,epi4,epi5,epi6,set) %>% 
    na.omit()
  
    a <- lmer(clinical1[,1] ~ clinical1$CaCo + clinical1$sex + clinical1$plate +
              clinical1$epi1 + clinical1$epi2 + clinical1$epi3 + clinical1$epi4 +
              clinical1$epi5 + clinical1$epi6 + (1 | clinical1$set)) %>%
      summary()
    b <- a$coefficients[2,]  %>% t() %>% as.data.frame()
    b$cellType <- cells[j]
    colnames(b) <- c("Estimate","stdError","df","tValue","pValue","cellType")
    cellMixEffect[[k]] <- b 
    k <- k+1
    }
cellMixEffect1 <- data.frame(do.call(rbind, cellMixEffect)) 

write.csv(cellMixEffect1,"set234cellMixEffectCellRegression.csv")

```

```{bash meta-analysis of B cell proportion regression of ALL in set 3 and 4}

cd ~/DSALL/usc/keren
metal
MARKER cell_name
EFFECT estimate_6epi
PVALUE pvalue_6epi
SCHEME STDERR
STDERR stderror_6epi
VERBOSE OFF
PROCESS set2_reg_lin_epi6.txt
PROCESS set3_reg_lin_epi6.txt
PROCESS set4_reg_lin_epi6.txt
ANALYZE
exit
mv METAANALYSIS1.TBL usc_set234_cp_se_meta.TBL
```
