---
output: html_document
editor_options: 
  chunk_output_type: console
---

Project description: Investigating the associations between European ancestry proportions in Latino population and pilocytic astrocytoma risk (both globally and regionally)

This paper has been published: 
Shaobo Li et al, Localized variation in ancestral admixture identifies pilocytic astrocytoma risk loci among Latino children
DOI: 10.1371/journal.pgen.1010388

```{bash data preparation}

################################################################
## Testing Charleston's new files to find subjects with the most Amerindian ancestry
#/project/minhuic_62/datasets/public/gnomAD/r3.1/hgdp_1kg_subset is source file

cd /scratch/lishaobo/GWAS/glioma/populationRisk2
#cd /project/wiemels_494/sebastian/temp_scratch_backup/populationRisk2


#cp /project/wiemels_260/sebastian/gwas_glioma/populationRisk2/gnomad.genomes.v3.1.hgdp_1kg_subset.sample_meta.tsv .

for chr in {1..22}; do
  plink --vcf /project/minhuic_62/datasets/public/gnomAD/r3.1/hgdp_1kg_subset/gnomad.genomes.v3.1.hgdp_1kg_subset.chr$chr.vcf.bgz\
  --snps-only 'just-acgt'\
  --biallelic-only\
  --make-bed\
  --out hgdp_1kg_chr${chr}
done

find -name "hgdp_1kg_chr*.bim" > ForMerge_hgdp.list 
sed -i 's/.bim//g' ForMerge_hgdp.list

for chr in {1..22}; do
  awk -v OFS="\t" '{print $1, $1":"$4":"$5":"$6, $3, $4, $5, $6}' hgdp_1kg_chr${chr}.bim > hgdp_1kg_chr${chr}_norsid.bim
  mv hgdp_1kg_chr${chr}.bim hgdp_1kg_chr${chr}_bkp.bim
  mv hgdp_1kg_chr${chr}_norsid.bim hgdp_1kg_chr${chr}.bim
done

plink --merge-list ForMerge_hgdp.list\
  --make-bed\
  --out hgdp_1kg

plink2 --bfile hgdp_1kg\
  --maf 0.1\
  --indep-pairwise 50 10 0.1\
  --out hgdp_1kg
  
plink --bfile hgdp_1kg\
  --extract hgdp_1kg.prune.in\
  --make-bed\
  --out hgdp_1kg_prnd

## Cross validation to show the best K
for K in 1 2 3 4 5 6 7 8; do 
  admixture -s 19900528 --cv hgdp_1kg_prnd.bed $K -j200 | tee log${K}.hgdp_1kg.out
done

grep -h CV log?.hgdp_1kg.out
# CV error (K=1): 0.57475
# CV error (K=2): 0.54029
# CV error (K=3): 0.52242
# CV error (K=4): 0.51734
# CV error (K=5): 0.51386
# CV error (K=6): 0.51212
# CV error (K=7): 0.51135
# CV error (K=8): 0.51082
# CV error (K=9): 0.51041

admixture hgdp_1kg_prnd.bed 3
admixture hgdp_1kg_prnd.bed 4
admixture hgdp_1kg_prnd.bed 5
admixture hgdp_1kg_prnd.bed 6

## k=5 is good.
```

```{bash script Rerun k=5 with 10 diffrent random seeds and see the results}
#!/bin/bash
#SBATCH --time=12:00:00
#SBATCH --mail-type=END
#SBATCH --array=1-10
#SBATCH --output=adm5_%A_%a.out
#SBATCH --job-name=adm5
#SBATCH -c 20
#SBATCH --ntasks=1
#SBATCH --mem=40G

cd /scratch/lishaobo/GWAS/glioma/populationRisk2/adm5
mkdir $(($SLURM_ARRAY_TASK_ID+1))
cd $(($SLURM_ARRAY_TASK_ID+1))
admixture -s $(($SLURM_ARRAY_TASK_ID+10086)) -j200 ../../hgdp_1kg_prnd.bed 5
echo $SLURM_ARRAY_TASK_ID complete!

```

```{r global ancestry visualization for Charleston's}

library(tidyverse)
library(data.table)
library(forcats)
library(ggthemes)
library(patchwork)
#setwd("/scratch/lishaobo/GWAS/glioma/populationRisk2")
setwd("/project/wiemels_494/sebastian/temp_scratch_backup/populationRisk2")

inds = fread("/project/wiemels_260/sebastian/gwas_glioma/populationRisk2/gnomad.genomes.v3.1.hgdp_1kg_subset.sample_meta.tsv") %>%
  select(s, labeled_subpop,population_inference.pop)
subs <- read.table("hgdp_1kg_prnd.fam")%>%
  select(V1,V2) %>%
  rename("FID"="V1","IID" = "V2") %>%
  mutate(FIID = ifelse(FID==IID, IID, paste0(FID,"_",IID)))%>%
  left_join(inds, by=c("FIID" = "s")) 

# 3 components
tbl=read.table("hgdp_1kg_prnd.3.Q")
tbl_cmb <- cbind(subs,tbl) %>% as.data.frame() %>%
  arrange(population_inference.pop,labeled_subpop,desc(V3)) 
tbl_cmb%>%
  gather(key=ancestry, value=percentage,c("V1","V2","V3"))%>%
  ggplot(aes(factor(FIID), percentage, fill = factor(ancestry))) +
  geom_col(size = 0.1) +
  facet_grid(~fct_inorder(population_inference.pop), scales = "free", switch = "x", space = "free") +
  theme_minimal() + 
  labs(x = "Individuals", title = "Ancestry proportions, k=3", y = "Ancestry") +
  scale_y_continuous(expand = c(0, 0)) +
  scale_x_discrete(expand = expand_scale(add = 1)) +
  theme(
    plot.title = element_text(hjust = 0.5),
    panel.spacing.x = unit(0.1, "lines"),
    axis.text.x = element_blank(),
    panel.grid = element_blank(),
    strip.text.x = element_text(angle = 90, size=7)
  ) +
  scale_fill_gdocs(guide = FALSE)
ggsave("hgdp_1kg_prnd.3.png",width = 8,height = 6,dpi = 500)

tbl_cmb%>%
  gather(key=ancestry, value=percentage,c("V1","V2","V3"))%>%
  ggplot(aes(factor(FIID), percentage, fill = factor(ancestry))) +
  geom_col(size = 0.1) +
  facet_grid(~fct_inorder(labeled_subpop), scales = "free", switch = "x", space = "free") +
  theme_minimal() + 
  labs(x = "Individuals", title = "Ancestry proportions, k=3", y = "Ancestry") +
  scale_y_continuous(expand = c(0, 0)) +
  scale_x_discrete(expand = expand_scale(add = 1)) +
  theme(
    plot.title = element_text(hjust = 0.5),
    panel.spacing.x = unit(0.05, "lines"),
    axis.text.x = element_blank(),
    panel.grid = element_blank(),
    strip.text.x = element_text(angle = 90, size=5, margin=margin(t=1, r=1, b=5, l=1))
  ) +
  scale_fill_gdocs(guide = FALSE)
ggsave("hgdp_1kg_prnd.labeled_subpop.3.png",width = 25,height = 6,dpi = 500)

# 4 components
tbl=read.table("hgdp_1kg_prnd.4.Q")
tbl_cmb <- cbind(subs,tbl) %>% as.data.frame() %>%
  arrange(population_inference.pop,labeled_subpop) 
tbl_cmb%>%
  gather(key=ancestry, value=percentage,c("V1","V2","V3","V4"))%>%
  ggplot(aes(factor(FIID), percentage, fill = factor(ancestry))) +
  geom_col(size = 0.1) +
  facet_grid(~fct_inorder(population_inference.pop), scales = "free", switch = "x", space = "free") +
  theme_minimal() + 
  labs(x = "Individuals", title = "Ancestry proportions, k=4", y = "Ancestry") +
  scale_y_continuous(expand = c(0, 0)) +
  scale_x_discrete(expand = expand_scale(add = 1)) +
  theme(
    plot.title = element_text(hjust = 0.5),
    panel.spacing.x = unit(0.1, "lines"),
    axis.text.x = element_blank(),
    panel.grid = element_blank(),
    strip.text.x = element_text(angle = 90, size=7)
  ) +
  scale_fill_gdocs(guide = FALSE)
ggsave("hgdp_1kg_prnd.4.png",width = 8,height = 6,dpi = 500)

tbl_cmb%>%
  gather(key=ancestry, value=percentage,c("V1","V2","V3","V4"))%>%
  ggplot(aes(factor(FIID), percentage, fill = factor(ancestry))) +
  geom_col(size = 0.1) +
  facet_grid(~fct_inorder(labeled_subpop), scales = "free", switch = "x", space = "free") +
  theme_minimal() + 
  labs(x = "Individuals", title = "Ancestry proportions, k=4", y = "Ancestry") +
  scale_y_continuous(expand = c(0, 0)) +
  scale_x_discrete(expand = expand_scale(add = 1)) +
  theme(
    plot.title = element_text(hjust = 0.5),
    panel.spacing.x = unit(0.05, "lines"),
    axis.text.x = element_blank(),
    panel.grid = element_blank(),
    strip.text.x = element_text(angle = 90, size=5, margin=margin(t=1, r=1, b=5, l=1))
  ) +
  scale_fill_gdocs(guide = FALSE)
ggsave("hgdp_1kg_prnd.labeled_subpop.4.png",width = 25,height = 6,dpi = 500)

# 5 components
tbl=read.table("hgdp_1kg_prnd.5.Q")
tbl_cmb <- cbind(subs,tbl) %>% as.data.frame() %>%
  arrange(population_inference.pop,labeled_subpop) 
tbl_cmb%>%
  gather(key=ancestry, value=percentage,c("V1","V2","V3","V4","V5"))%>%
  ggplot(aes(factor(FIID), percentage, fill = factor(ancestry))) +
  geom_col(size = 0.1) +
  facet_grid(~fct_inorder(population_inference.pop), scales = "free", switch = "x", space = "free") +
  theme_minimal() + 
  labs(x = "Individuals", title = "Ancestry proportions, k=5", y = "Ancestry") +
  scale_y_continuous(expand = c(0, 0)) +
  scale_x_discrete(expand = expand_scale(add = 1)) +
  theme(
    plot.title = element_text(hjust = 0.5),
    panel.spacing.x = unit(0.1, "lines"),
    axis.text.x = element_blank(),
    panel.grid = element_blank(),
    strip.text.x = element_text(angle = 90, size=7)
  ) +
  scale_fill_gdocs(guide = FALSE)
ggsave("hgdp_1kg_prnd.5.png",width = 8,height = 6,dpi = 500)

tbl=read.table("hgdp_1kg_prnd.5.Q")
tbl_cmb <- cbind(subs,tbl) %>% as.data.frame() %>%
  arrange(population_inference.pop,labeled_subpop) 
tbl_cmb%>%
  gather(key=ancestry, value=percentage,c("V1","V2","V3","V4","V5"))%>%
  ggplot(aes(factor(FIID), percentage, fill = factor(ancestry))) +
  geom_col(size = 0.1) +
  facet_grid(~fct_inorder(labeled_subpop), scales = "free", switch = "x", space = "free") +
  theme_minimal() + 
  labs(x = "Individuals", title = "Ancestry proportions, k=5", y = "Ancestry") +
  scale_y_continuous(expand = c(0, 0)) +
  scale_x_discrete(expand = expand_scale(add = 1)) +
  theme(
    plot.title = element_text(hjust = 0.5),
    panel.spacing.x = unit(0.05, "lines"),
    axis.text.x = element_blank(),
    panel.grid = element_blank(),
    strip.text.x = element_text(angle = 90, size=5, margin=margin(t=1, r=1, b=5, l=1))
  ) +
  scale_fill_gdocs(guide = FALSE)
ggsave("hgdp_1kg_prnd.labeled_subpop.5.png",width =25,height = 6,dpi = 500)

# select the purest of white, black, amerindian, south asian and east asian ancestries for reference of rhabdo project
## eyeballing: V5 is AFR, V1 is eas, V3 is NFE, V4 is AMR and V2 is sas

afr = tbl_cmb %>%
  dplyr::filter(population_inference.pop == "afr") %>%
  arrange(desc(V5)) %>%
  head(90)

nfe = tbl_cmb %>%
  dplyr::filter(population_inference.pop == "nfe") %>%
  arrange(desc(V3)) %>%
  head(90)

amr = tbl_cmb %>%
  dplyr::filter(population_inference.pop == "amr") %>%
  arrange(desc(V4)) %>%
  head(90)

eas = tbl_cmb %>%
  dplyr::filter(population_inference.pop == "eas") %>%
  arrange(desc(V1)) %>%
  head(90)

sas = tbl_cmb %>%
  dplyr::filter(population_inference.pop == "sas") %>%
  arrange(desc(V2)) %>%
  head(90)

ref_panel = rbind(afr,nfe,amr,eas,sas) %>%
  as.data.frame() %>%
  dplyr::select(FID,IID,population_inference.pop)

write.table(ref_panel,"/project/lishaobo_902/sebastian/rhabdo/snp_dis/sample_id90.txt",
            quote=FALSE,
            row.names = FALSE,
            col.names = FALSE,
            sep='\t')

# 6 components
tbl=read.table("hgdp_1kg_prnd.6.Q")
tbl_cmb <- cbind(subs,tbl) %>% as.data.frame() %>%
  arrange(population_inference.pop,labeled_subpop,desc(V3)) 
tbl_cmb%>%
  gather(key=ancestry, value=percentage,c("V1","V2","V3","V4","V5","V6"))%>%
  ggplot(aes(factor(FIID), percentage, fill = factor(ancestry))) +
  geom_col(size = 0.1) +
  facet_grid(~fct_inorder(population_inference.pop), scales = "free", switch = "x", space = "free") +
  theme_minimal() + 
  labs(x = "Individuals", title = "Ancestry proportions, k=6", y = "Ancestry") +
  scale_y_continuous(expand = c(0, 0)) +
  scale_x_discrete(expand = expand_scale(add = 1)) +
  theme(
    plot.title = element_text(hjust = 0.5),
    panel.spacing.x = unit(0.1, "lines"),
    axis.text.x = element_blank(),
    panel.grid = element_blank(),
    strip.text.x = element_text(angle = 90, size=7)
  ) +
  scale_fill_gdocs(guide = FALSE)
ggsave("hgdp_1kg_prnd.6.png",width = 8,height = 6,dpi = 500)

tbl=read.table("hgdp_1kg_prnd.6.Q")
tbl_cmb <- cbind(subs,tbl) %>% as.data.frame() %>%
  arrange(population_inference.pop,labeled_subpop,desc(V3)) 
tbl_cmb%>%
  gather(key=ancestry, value=percentage,c("V1","V2","V3","V4","V5","V6"))%>%
  ggplot(aes(factor(FIID), percentage, fill = factor(ancestry))) +
  geom_col(size = 0.1) +
  facet_grid(~fct_inorder(labeled_subpop), scales = "free", switch = "x", space = "free") +
  theme_minimal() + 
  labs(x = "Individuals", title = "Ancestry proportions, k=6", y = "Ancestry") +
  scale_y_continuous(expand = c(0, 0)) +
  scale_x_discrete(expand = expand_scale(add = 1)) +
  theme(
    plot.title = element_text(hjust = 0.5),
    panel.spacing.x = unit(0.05, "lines"),
    axis.text.x = element_blank(),
    panel.grid = element_blank(),
    strip.text.x = element_text(angle = 90, size=5, margin=margin(t=1, r=1, b=5, l=1))
  ) +
  scale_fill_gdocs(guide = FALSE)
ggsave("hgdp_1kg_prnd.labeled_subpop.6.png",width =25,height = 6,dpi = 500)
```

```{r checking for charleston for a differnt way to choose subjects. }
# didn't end up using this since it doesn't change the results.
library(tidyverse)
library(data.table)
library(forcats)
library(ggthemes)
library(patchwork)
setwd("/scratch/lishaobo/GWAS/glioma/populationRisk2")


inds = fread("/project/wiemels_260/sebastian/gwas_glioma/populationRisk2/gnomad.genomes.v3.1.hgdp_1kg_subset.sample_meta.tsv") %>%
  select(s, labeled_subpop,population_inference.pop)
subs <- read.table("hgdp_1kg_prnd.fam")%>%
  select(V1,V2) %>%
  rename("FID"="V1","IID" = "V2") %>%
  mutate(FIID = ifelse(FID==IID, IID, paste0(FID,"_",IID)))%>%
  left_join(inds, by=c("FIID" = "s")) 

tbl=read.table("hgdp_1kg_prnd.5.Q")
tbl_cmb <- cbind(subs,tbl) %>% as.data.frame() %>%
  select(FID,IID,FIID,population_inference.pop,labeled_subpop,V1,V2,V3,V4,V5)

######################################################

inf.pop="afr"

tbl_cmb1 = tbl_cmb %>%
  filter(population_inference.pop==inf.pop) %>%
  filter(V5>0.9)

# check labeled_subpop = bedouin, mozabite, acb, asw, and pur and others
subP = "pur"
subP= c("bedouin", "mozabite", "acb", "asw", "pur")

subC=tbl_cmb1 %>% filter(!labeled_subpop%in%subP) 
summary(subC$V5)

subC %>%
  select(labeled_subpop,V5)%>%
  ggplot(aes(x=labeled_subpop,y=V5))+
  geom_boxplot(outlier.size=0.2)+
  theme_classic() +
  ggtitle(paste0("distribution of afr ancestry in",subP))
ggsave(paste0("distributionAfr.",subP,".png"),width =5,height = 5,dpi = 200)

tbl_cmb1%>%
  gather(key=ancestry, value=percentage,c("V1","V2","V3","V4","V5"))%>%
  ggplot(aes(factor(FIID), percentage, fill = factor(ancestry))) +
  geom_col(size = 0.1) +
  facet_grid(~fct_inorder(labeled_subpop), scales = "free", switch = "x", space = "free") +
  theme_minimal() + 
  labs(x = "Individuals", 
       title = paste0("Checking ancestry proportions ",inf.pop), 
       y = "Ancestry") +
  scale_y_continuous(expand = c(0, 0)) +
  scale_x_discrete(expand = expand_scale(add = 1)) +
  theme(
    plot.title = element_text(hjust = 0.5),
    panel.spacing.x = unit(0.5, "lines"),
    axis.text.x = element_blank(),
    panel.grid = element_blank(),
    strip.text.x = element_text(angle = 90, size=5, margin=margin(t=1, r=1, b=5, l=1))
  ) +
  scale_fill_gdocs(guide = FALSE)
ggsave(paste0(inf.pop,".check.cl.png"),width =15,height = 8,dpi = 500)

######################################################
inf.pop="eur"

tbl_cmb1 = tbl_cmb %>%
#  filter(population_inference.pop=="nfe" & labeled_subpop != "fin") 
  filter(V3>0.85)
table(tbl_cmb1$labeled_subpop)

subP = "russian"
subC=tbl_cmb1 %>% filter(!labeled_subpop%in%subP) 
summary(subC$V3)



tbl_cmb1 %>%
  select(labeled_subpop,V3)%>%
  ggplot(aes(x=labeled_subpop,y=V3))+
  geom_boxplot(outlier.size=0.2)+
  theme_classic() +
  ggtitle(paste0("distribution of eur ancestry in European subjects"))
ggsave(paste0("distributionEur.png"),width =15,height = 5,dpi = 200)


tbl_cmb1%>%
  gather(key=ancestry, value=percentage,c("V1","V2","V3","V4","V5"))%>%
  ggplot(aes(factor(FIID), percentage, fill = factor(ancestry))) +
  geom_col(size = 0.1) +
  facet_grid(~fct_inorder(labeled_subpop), scales = "free", switch = "x", space = "free") +
  theme_minimal() + 
  labs(x = "Individuals", title = "Checking a small group of ancestry proportions, k=5", y = "Ancestry") +
  scale_y_continuous(expand = c(0, 0)) +
  scale_x_discrete(expand = expand_scale(add = 1)) +
  theme(
    plot.title = element_text(hjust = 0.5),
    panel.spacing.x = unit(0.5, "lines"),
    axis.text.x = element_blank(),
    panel.grid = element_blank(),
    strip.text.x = element_text(angle = 90, size=5, margin=margin(t=1, r=1, b=5, l=1))
  ) +
  scale_fill_gdocs(guide = FALSE)
ggsave(paste0(inf.pop,".check.cl.png"),width =15,height = 8,dpi = 500)

inf.pop="amr"

tbl_cmb1 = tbl_cmb %>%
  filter(V4>0.85) 

tbl_cmb1%>%
  gather(key=ancestry, value=percentage,c("V1","V2","V3","V4","V5"))%>%
  ggplot(aes(factor(FIID), percentage, fill = factor(ancestry))) +
  geom_col(size = 0.1) +
  facet_grid(~fct_inorder(labeled_subpop), scales = "free", switch = "x", space = "free") +
  theme_minimal() + 
  labs(x = "Individuals", title = "Checking a small group of ancestry proportions, k=5", y = "Ancestry") +
  scale_y_continuous(expand = c(0, 0)) +
  scale_x_discrete(expand = expand_scale(add = 1)) +
  theme(
    plot.title = element_text(hjust = 0.5),
    panel.spacing.x = unit(0.5, "lines"),
    axis.text.x = element_blank(),
    panel.grid = element_blank(),
    strip.text.x = element_text(angle = 90, size=5, margin=margin(t=1, r=1, b=5, l=1))
  ) +
  scale_fill_gdocs(guide = FALSE)
ggsave(paste0(inf.pop,".check.cl.png"),width =15,height = 8,dpi = 500)


######################################################
## choose reference panles based on percentage
library(tidyverse)
library(data.table)
#setwd("/scratch/lishaobo/GWAS/glioma/populationRisk2")
setwd("/project/wiemels_494/sebastian/temp_scratch_backup/populationRisk2")

inds = fread("/project/wiemels_260/sebastian/gwas_glioma/populationRisk2/gnomad.genomes.v3.1.hgdp_1kg_subset.sample_meta.tsv") %>%
  select(s, labeled_subpop,population_inference.pop)
subs <- read.table("hgdp_1kg_prnd.fam")%>%
  select(V1,V2) %>%
  rename("FID"="V1","IID" = "V2") %>%
  mutate(FIID = ifelse(FID==IID, IID, paste0(FID,"_",IID)))%>%
  left_join(inds, by=c("FIID" = "s")) 

tbl=read.table("hgdp_1kg_prnd.5.Q")

tbl_cmb <- cbind(subs,tbl) %>% as.data.frame() %>%
  arrange(population_inference.pop,labeled_subpop) 

# V5 is African
# V4 is Native american
# V3 is Eur

tbl_cmb_amr= tbl_cmb %>%
  filter(V4>0.85) %>%
  select(FID,IID) #94 AMR reference subjects

tbl_cmb_afr = tbl_cmb %>%
  filter(population_inference.pop=="afr" & V5>0.85) %>%
  select(FID,IID) #816 AFR reference subjects

tbl_cmb_eur = tbl_cmb %>%
  filter(population_inference.pop=="nfe" & V3>0.85) %>%
  select(FID,IID) #633 EUR reference subjects

tbl_cmb_rnm <- tbl_cmb %>%
  rename("white" = "V3",
         "ameind" = "V4",
         "black" = "V5") %>%
  select(-V2,-V1)

write.table(tbl_cmb_rnm, "refAncestryProp.txt", sep="\t",quote=FALSE,row.names = FALSE)
write.table(tbl_cmb_eur,"EURpop.pr.txt", col.names = FALSE, row.names = FALSE, quote=FALSE)
write.table(tbl_cmb_afr,"AFRpop.pr.txt", col.names = FALSE, row.names = FALSE, quote=FALSE)
write.table(tbl_cmb_amr,"AMRpop.pr.txt", col.names = FALSE, row.names = FALSE, quote=FALSE)
```

```{r selecting subjects as the best reference panel, based on self-reported}

library(tidyverse)
library(data.table)
setwd("/scratch/lishaobo/GWAS/glioma/populationRisk2")

inds = fread("/project/wiemels_260/sebastian/gwas_glioma/populationRisk2/gnomad.genomes.v3.1.hgdp_1kg_subset.sample_meta.tsv") %>%
  select(s, labeled_subpop,population_inference.pop)
subs <- read.table("hgdp_1kg_prnd.fam")%>%
  select(V1,V2) %>%
  rename("FID"="V1","IID" = "V2") %>%
  mutate(FIID = ifelse(FID==IID, IID, paste0(FID,"_",IID)))%>%
  left_join(inds, by=c("FIID" = "s")) 

tbl=read.table("hgdp_1kg_prnd.5.Q")

tbl_cmb <- cbind(subs,tbl) %>% as.data.frame() %>%
  arrange(population_inference.pop,labeled_subpop) 

# V5 is African
# V4 is Native american
# V3 is Eur

tbl_cmb_amr= tbl_cmb %>%
  filter(V4>0.85) %>%
  select(FID,IID) #94 AMR reference subjects

tbl_cmb_afr = tbl_cmb %>%
  filter(population_inference.pop=="afr" & labeled_subpop != "acb" &  labeled_subpop != "asw") %>%
  select(FID,IID) #716 AFR reference subjects

tbl_cmb_eur = tbl_cmb %>%
  filter(population_inference.pop=="nfe" & labeled_subpop != "fin") %>%
  select(FID,IID) #671 EUR reference subjects

tbl_cmb_rnm <- tbl_cmb %>%
  rename("white" = "V3",
         "ameind" = "V4",
         "black" = "V5") %>%
  select(-V2,-V1)

write.table(tbl_cmb_rnm, "refAncestryProp.txt", sep="\t",quote=FALSE,row.names = FALSE)
write.table(tbl_cmb_eur,"EURpop.txt", col.names = FALSE, row.names = FALSE, quote=FALSE)
write.table(tbl_cmb_afr,"AFRpop.txt", col.names = FALSE, row.names = FALSE, quote=FALSE)
write.table(tbl_cmb_amr,"AMRpop.txt", col.names = FALSE, row.names = FALSE, quote=FALSE)
```

```{bash selecitng reference samples}

cd /scratch/lishaobo/GWAS/glioma/populationRisk2

cat EURpop.txt AFRpop.txt AMRpop.txt > ref.hgdp.1kg.subs.txt

plink --bfile hgdp_1kg \
  --keep ref.hgdp.1kg.subs.txt \
  --biallelic-only 'strict' \
  --snps-only 'just-acgt' \
  --make-bed \
  --out slc.panel.hgdp.1kg.subs
# double checked: this is also on Hg38

```

# Selecting subjects from batch 2 files policytic and astrocytoma subtype subjects
# For global ancestry comparison, use self-reported ancestry instead of imputed

```{r select subjects of the particular subtype and ancestry}
library(tidyverse)
library(data.table)
setwd("/scratch/lishaobo/GWAS/glioma/populationRisk2")

# select subjects with latino ancestries and pilocytic subtype
imp_race = read.csv("/scratch/lishaobo/GWAS/glioma/saxb4/self_reported_and_imputed_race.csv",row.names=1)%>%
  filter(self_race == "Hispanic" ) 

# select subjects with proper subtype and case/control info
glio <- fread("/project/wiemels_260/sebastian/gwas_glioma/saxb4/rawSampleAttributes.txt") %>%
  select(FID, IID,subjectId,phenotype, HISTO_T3) %>%
  dplyr::filter(phenotype ==1 | (HISTO_T3 == 9421) ) %>%
  inner_join(imp_race, by=c("FID","IID"))

write.table(glio, "/project/wiemels_260/sebastian/gwas_glioma/populationRisk2/poprisk2_pol_attributes.txt",sep="\t",quote=FALSE,row.names = FALSE)

glio1 = glio %>%
  select(FID,IID)
write.table(glio1, "polIDs.txt",sep="\t",quote=FALSE,row.names = FALSE)

##########################################################################################
# Select pilocytic european subjects for replication

imp_race = read.csv("/scratch/lishaobo/GWAS/glioma/saxb4/self_reported_and_imputed_race.csv",row.names=1)%>%
  filter(self_race == "White" ) 
# select subjects with proper subtype and case/control info
glio <- fread("/project/wiemels_260/sebastian/gwas_glioma/saxb4/rawSampleAttributes.txt") %>%
  select(FID, IID,subjectId,phenotype, HISTO_T3) %>%
  dplyr::filter(phenotype ==1 | (HISTO_T3 == 9421) ) %>%
  inner_join(imp_race, by=c("FID","IID"))
write.table(glio, "/project/wiemels_260/sebastian/gwas_glioma/populationRisk2/poprisk2_pol_attributes.wt.txt",sep="\t",quote=FALSE,row.names = FALSE)
glio1 = glio %>%
  select(FID,IID)
write.table(glio1, "polIDs.wt.txt",sep="\t",quote=FALSE,row.names = FALSE)

##########################################################################################
# Select astrocytoma latino subjects for replication
imp_race = read.csv("/scratch/lishaobo/GWAS/glioma/saxb4/self_reported_and_imputed_race.csv",row.names=1)%>%
  filter(self_race == "Hispanic" ) 

# select subjects with proper subtype and case/control info
glio <- fread("/project/wiemels_260/sebastian/gwas_glioma/saxb4/rawSampleAttributes.txt") %>%
  select(FID, IID,subjectId,phenotype, HISTO_T3) %>%
  dplyr::filter(phenotype ==1 | (HISTO_T3 >= 9400 & HISTO_T3 <= 9424  & HISTO_T3 != 9421) ) %>%
  inner_join(imp_race, by=c("FID","IID"))

write.table(glio, "/project/wiemels_260/sebastian/gwas_glioma/populationRisk2/poprisk2_ast_attributes.txt",sep="\t",quote=FALSE,row.names = FALSE)

glio1 = glio %>%
  select(FID,IID)
write.table(glio1, "astIDs.txt",sep="\t",quote=FALSE,row.names = FALSE)


```

```{bash merge with glioma dataset,merging pilocytic with reference}

# merging pilocytic with reference
#cd /scratch/lishaobo/GWAS/glioma/populationRisk2
cd /project/wiemels_494/sebastian/temp_scratch_backup/populationRisk2

## saxb4_incl_11 is pre-imputation
plink --bfile /scratch/lishaobo/GWAS/glioma/saxb4/saxb4_incl_11 \
  --keep polIDs.txt\
  --make-bed \
  --out pol.adm.subs
# 1486 people remaining
# 301 are cases and 1185 are controls
# 764144 SNPs

## Combine 1000 G data with glioma data
# choose overlap
# slc.panel.hgdp.1kg.subs
# pol.adm.subs

## delete duplicates with the same position
awk '{print $2, $1":"$4}' slc.panel.hgdp.1kg.subs.bim | sort -k2,2 | uniq -f1 -D  > slc.panel.hgdp.1kg.snps
awk '{print $1}' slc.panel.hgdp.1kg.snps > all.snps
sort -uk2 slc.panel.hgdp.1kg.snps | awk '{print $1}' > uniq.snps
awk 'NR==FNR{a[$0]=1;next}!a[$0]' uniq.snps all.snps > slc.panel.hgdp.1kg.dups

plink --bfile slc.panel.hgdp.1kg.subs \
  --exclude slc.panel.hgdp.1kg.dups \
  --make-bed \
  --out slc.panel.hgdp.1kg.nodup
awk '{print $1,"chr"$2,$3,$4,$5,$6}' slc.panel.hgdp.1kg.nodup.bim > slc.panel.hgdp.1kg.nodup.bim.bkup
mv slc.panel.hgdp.1kg.nodup.bim.bkup slc.panel.hgdp.1kg.nodup.bim

awk '{print $2, $1":"$4}' pol.adm.subs.bim | sort -k2,2 | uniq -f1 -D  > pol_adm.snps
awk '{print $1}' pol_adm.snps > all.snps
sort -uk2 pol_adm.snps | awk '{print $1}' > uniq.snps
awk 'NR==FNR{a[$0]=1;next}!a[$0]' uniq.snps all.snps > pol_adm.dups

plink --bfile pol.adm.subs \
  --exclude pol_adm.dups \
  --make-bed \
  --out pol_adm.nodup
awk '{print $1,"chr"$1":"$4":"$5":"$6,$3,$4,$5,$6}' pol_adm.nodup.bim > pol_adm.nodup.bim.bkup
mv pol_adm.nodup.bim.bkup pol_adm.nodup.bim
  
# make these two files include the same SNPs
awk '{print$2}' slc.panel.hgdp.1kg.nodup.bim > 1kg.SNPs
plink --bfile pol_adm.nodup \
  --extract 1kg.SNPs \
  --make-bed \
  --out pol_adm.merge1

awk '{print$2}' pol_adm.merge1.bim > pol_adm.merge1.SNPs
plink --bfile slc.panel.hgdp.1kg.nodup \
  --extract pol_adm.merge1.SNPs \
  --make-bed \
  --out hgdp.1kg.merge1
# 719803 SNPs in total that are overlapped
  
#nano glio.hgdp.1kg.mg
# pol_adm.merge1 
# hgdp.1kg.merge1

# Merge Glio and 1000G
plink --merge-list glio.hgdp.1kg.mg \
  --make-bed \
  --out glio.hgdp.1kg
# 719803 SNPs

# Calculation ancestry proportions again
admixture glio.hgdp.1kg.bed 5

#!/bin/bash
#SBATCH --time=12:00:00
#SBATCH --mail-type=END
#SBATCH --array=1-10
#SBATCH --output=adm5_%A_%a.out
#SBATCH --job-name=adm5
#SBATCH -c 20
#SBATCH --ntasks=1
#SBATCH --mem=40G

cd /scratch/lishaobo/GWAS/glioma/populationRisk2/adm5
mkdir $SLURM_ARRAY_TASK_ID
cd $SLURM_ARRAY_TASK_ID
admixture -s $(($SLURM_ARRAY_TASK_ID+10086)) -j200 ../../glio.hgdp.1kg.bed 5
echo $SLURM_ARRAY_TASK_ID complete!

##########################################################################################
# Select European subjects for replication in chr 6
cd /scratch/lishaobo/GWAS/glioma/populationRisk2

plink --bfile /scratch/lishaobo/GWAS/glioma/saxb4/saxb4_incl_11\
  --keep polIDs.wt.txt\
  --make-bed \
  --out pol.adm.subs.wt

```

```{bash merge with glioma dataset,merging astrocytoma with reference}

# using new reference panel from the block below: "rfmix using a different reference panel based on percentage only and validate if the results change"
cd /scratch/lishaobo/GWAS/glioma/populationRisk2

cat EURpop.pr.txt AFRpop.pr.txt AMRpop.pr.txt > ref.pr.hgdp.1kg.subs.txt

plink --bfile hgdp_1kg\
  --keep ref.pr.hgdp.1kg.subs.txt\
  --biallelic-only 'strict'\
  --snps-only 'just-acgt'\
  --make-bed\
  --out slc.panel.pr.hgdp.1kg.subs
# double checked: this is also on Hg38

# merging astrocytoma with reference
cd /scratch/lishaobo/GWAS/glioma/populationRisk2

plink --bfile /scratch/lishaobo/GWAS/glioma/saxb4/saxb4_incl_11\
  --keep astIDs.txt\
  --make-bed\
  --out ast.adm.subs
# 1431 people remaining
# 246 are cases and 1185 are controls

## Combine 1000 G data with glioma data
# choose overlap
# slc.panel.pr.hgdp.1kg.subs
# ast.adm.subs

## delete duplicates with the same position
awk '{print $2, $1":"$4}' slc.panel.pr.hgdp.1kg.subs.bim | sort -k2,2 | uniq -f1 -D  > slc.panel.pr.hgdp.1kg.snps
awk '{print $1}' slc.panel.pr.hgdp.1kg.snps > all.pr.snps
sort -uk2 slc.panel.pr.hgdp.1kg.snps | awk '{print $1}' > uniq.pr.snps
awk 'NR==FNR{a[$0]=1;next}!a[$0]' uniq.pr.snps all.pr.snps > slc.panel.pr.hgdp.1kg.dups
plink --bfile slc.panel.pr.hgdp.1kg.subs\
  --exclude slc.panel.pr.hgdp.1kg.dups\
  --make-bed\
  --out slc.panel.pr.hgdp.1kg.nodup
awk '{print $1,"chr"$2,$3,$4,$5,$6}' slc.panel.pr.hgdp.1kg.nodup.bim > slc.panel.pr.hgdp.1kg.nodup.bim.bkup
mv slc.panel.pr.hgdp.1kg.nodup.bim.bkup slc.panel.pr.hgdp.1kg.nodup.bim

awk '{print $2, $1":"$4}' ast.adm.subs.bim | sort -k2,2 | uniq -f1 -D  > ast_adm.snps
awk '{print $1}' ast_adm.snps > all.snps
sort -uk2 ast_adm.snps | awk '{print $1}' > uniq.snps
awk 'NR==FNR{a[$0]=1;next}!a[$0]' uniq.snps all.snps > ast_adm.dups

plink --bfile ast.adm.subs\
  --exclude ast_adm.dups\
  --make-bed\
  --out ast_adm.nodup
awk '{print $1,"chr"$1":"$4":"$5":"$6,$3,$4,$5,$6}' ast_adm.nodup.bim > ast_adm.nodup.bim.bkup
mv ast_adm.nodup.bim.bkup ast_adm.nodup.bim
  
# make these two files include the same SNPs
awk '{print$2}' slc.panel.pr.hgdp.1kg.nodup.bim > 1kg.SNPs
plink --bfile ast_adm.nodup \
  --extract 1kg.SNPs \
  --make-bed \
  --out ast_adm.merge1

awk '{print$2}' ast_adm.merge1.bim > ast_adm.merge1.SNPs
plink --bfile slc.panel.pr.hgdp.1kg.nodup\
  --extract ast_adm.merge1.SNPs\
  --make-bed\
  --out hgdp.1kg.merge1
# 719803 SNPs in total that are overlapped
  
#nano ast.hgdp.1kg.mg
# ast_adm.merge1 
# hgdp.1kg.merge1

# Merge Glio and 1000G
plink --merge-list ast.hgdp.1kg.mg\
  --make-bed\
  --out glio.ast.hgdp.1kg

# Calculation ancestry proportions again
admixture glio.ast.hgdp.1kg.bed 5
admixture glio.ast.hgdp.1kg.bed 3

#!/bin/bash
#SBATCH --time=12:00:00
#SBATCH --mail-type=END
#SBATCH --array=1-10
#SBATCH --output=adm5_%A_%a.out
#SBATCH --job-name=adm5
#SBATCH -c 20
#SBATCH --ntasks=1
#SBATCH --mem=40G

cd /scratch/lishaobo/GWAS/glioma/populationRisk2/adm5
# mkdir $SLURM_ARRAY_TASK_ID
cd $SLURM_ARRAY_TASK_ID
admixture -s $(($SLURM_ARRAY_TASK_ID+10086)) -j200 ../../glio.ast.hgdp.1kg.bed 5
echo $SLURM_ARRAY_TASK_ID complete!

#!/bin/bash
#SBATCH --time=12:00:00
#SBATCH --mail-type=END
#SBATCH --array=1-10
#SBATCH --output=adm3_%A_%a.out
#SBATCH --job-name=adm3
#SBATCH -c 20
#SBATCH --ntasks=1
#SBATCH --mem=40G

cd /scratch/lishaobo/GWAS/glioma/populationRisk2/adm5
# mkdir $SLURM_ARRAY_TASK_ID
cd $SLURM_ARRAY_TASK_ID
admixture -s $(($SLURM_ARRAY_TASK_ID+10086)) -j200 ../../glio.ast.hgdp.1kg.bed 3
echo $SLURM_ARRAY_TASK_ID complete!

```

# Global ancestry comparison

```{r Comparing global ancestry of latino case/controls in pol}
library(tidyverse)
setwd("/scratch/lishaobo/GWAS/glioma/populationRisk2")

######################I ran K=5 10 times. Plot them to compare
library(tidyverse)
library(data.table)
library(forcats)
library(ggthemes)
library(patchwork)
library(R.utils)

setwd("/scratch/lishaobo/GWAS/glioma/populationRisk2/adm5")

raw <- fread("/project/wiemels_260/sebastian/gwas_glioma/populationRisk2/poprisk2_pol_attributes.txt") %>%
  select(FID,IID, self_race, phenotype)

inds = fread("/project/wiemels_260/sebastian/gwas_glioma/populationRisk2/gnomad.genomes.v3.1.hgdp_1kg_subset.sample_meta.tsv") %>%
  select(s, labeled_subpop,population_inference.pop)
subhg <- read.table("../hgdp_1kg_prnd.fam")%>%
  select(V1,V2) %>%
  rename("FID"="V1","IID" = "V2") %>%
  mutate(FIID = ifelse(FID==IID, IID, paste0(FID,"_",IID)))%>%
  left_join(inds, by=c("FIID" = "s")) %>%
  mutate(self_race=population_inference.pop,
         phenotype=1) %>%
  select(FID,IID, self_race, phenotype)

raw1 = rbind(raw,subhg) %>%as.data.frame  

subs <- read.table("../glio.hgdp.1kg.fam")%>%
  select(V1,V2) %>%
  rename("FID"="V1","IID" = "V2") %>%
  left_join(raw1, by=c("FID","IID"))

for(i in 1:10){
tbl=read.table(gstring("${i}/glio.hgdp.1kg.5.Q"))
tbl_cmb <- cbind(subs,tbl) %>% as.data.frame() %>%
  mutate(fiid=paste0(FID,IID))%>%
  arrange(self_race) 
tbl_cmb%>%
  gather(key=ancestry, value=percentage,c("V1","V2","V3","V4","V5"))%>%
  ggplot(aes(factor(fiid), percentage, fill = factor(ancestry))) +
  geom_col(size = 0.1) +
  facet_grid(~fct_inorder(self_race), scales = "free", switch = "x", space = "free") +
  theme_minimal() + 
  labs(x = "Individuals", title = paste0("seed=",i," Ancestry proportions, k=5"), y = "Ancestry") +
  scale_y_continuous(expand = c(0, 0)) +
  scale_x_discrete(expand = expand_scale(add = 1)) +
  theme(
    plot.title = element_text(hjust = 0.5),
    panel.spacing.x = unit(0.1, "lines"),
    axis.text.x = element_blank(),
    panel.grid = element_blank(),
    strip.text.x = element_text(angle = 90, size=7)
  ) +
  scale_fill_gdocs(guide = FALSE)
ggsave(gstring("${i}_hgdp_1kg_prnd.5.png"),width = 8,height = 6,dpi = 500)
}

# the figures don't look "clean" as expected. K=3 and try again.
for(i in 1:10){
tbl=read.table(gstring("${i}/glio.hgdp.1kg.3.Q"))
tbl_cmb <- cbind(subs,tbl) %>% as.data.frame() %>%
  mutate(fiid=paste0(FID,IID))%>%
  arrange(self_race) 
tbl_cmb%>%
  gather(key=ancestry, value=percentage,c("V1","V2","V3"))%>%
  ggplot(aes(factor(fiid), percentage, fill = factor(ancestry))) +
  geom_col(size = 0.1) +
  facet_grid(~fct_inorder(self_race), scales = "free", switch = "x", space = "free") +
  theme_minimal() + 
  labs(x = "Individuals", title = paste0("seed=",i," Ancestry proportions, k=5"), y = "Ancestry") +
  scale_y_continuous(expand = c(0, 0)) +
  scale_x_discrete(expand = expand_scale(add = 1)) +
  theme(
    plot.title = element_text(hjust = 0.5),
    panel.spacing.x = unit(0.1, "lines"),
    axis.text.x = element_blank(),
    panel.grid = element_blank(),
    strip.text.x = element_text(angle = 90, size=7)
  ) +
  scale_fill_gdocs(guide = FALSE)
ggsave(gstring("${i}_hgdp_1kg_prnd.3.png"),width = 8,height = 6,dpi = 500)
}


####### For each table, assign the corresponding ancestry and compute an average
raw <- fread("/project/wiemels_260/sebastian/gwas_glioma/populationRisk2/poprisk2_pol_attributes.txt") %>%
  select(FID,IID, self_race, phenotype)
inds = fread("/project/wiemels_260/sebastian/gwas_glioma/populationRisk2/gnomad.genomes.v3.1.hgdp_1kg_subset.sample_meta.tsv") %>%
  select(s, labeled_subpop,population_inference.pop)
subhg <- read.table("../hgdp_1kg_prnd.fam")%>%
  select(V1,V2) %>%
  rename("FID"="V1","IID" = "V2") %>%
  mutate(FIID = ifelse(FID==IID, IID, paste0(FID,"_",IID)))%>%
  left_join(inds, by=c("FIID" = "s")) %>%
  mutate(self_race=population_inference.pop,
         phenotype=1) %>%
  select(FID,IID, self_race, phenotype)
raw1 = rbind(raw,subhg) %>%as.data.frame  
subs <- read.table("../glio.hgdp.1kg.fam")%>%
  select(V1,V2) %>%
  rename("FID"="V1","IID" = "V2") %>%
  left_join(raw1, by=c("FID","IID"))

# a figure of ancestry proportions after averaging, k =5

afrPro = list()
amrPro = list()
nfePro = list()

for(i in 1:10){
  tbl=read.table(gstring("${i}/glio.hgdp.1kg.5.Q"))
  tbl_cmb <- cbind(subs,tbl) %>% as.data.frame()
  
  afr = tbl_cmb %>% filter(self_race == "afr") %>% 
    select(V1,V2,V3)
  afrAve = which.max(colMeans(afr)) %>% names
  
  amr = tbl_cmb %>% filter(self_race == "amr") %>% 
    select(V1,V2,V3)
  amrAve = which.max(colMeans(amr)) %>% names
  
  nfe = tbl_cmb %>% filter(self_race == "nfe") %>% 
    select(V1,V2,V3)
  nfeAve = which.max(colMeans(nfe)) %>% names
  
  afrPro[[i]] = tbl_cmb[,afrAve]
  amrPro[[i]] = tbl_cmb[,amrAve]
  nfePro[[i]] = tbl_cmb[,nfeAve]
}

afrProDF = do.call(rbind,afrPro) %>% as.data.frame() %>%colMeans()
amrProDF = do.call(rbind,amrPro) %>% as.data.frame() %>%colMeans()
nfeProDF = do.call(rbind,nfePro) %>% as.data.frame() %>%colMeans()

tbl_ave = data.frame(afr=afrProDF,amr=amrProDF,nfe=nfeProDF)
tbl_ave_1 = cbind(subs,tbl_ave) %>% 
  as.data.frame() 
tbl_ave_1%>%
  mutate(fiid=paste0(FID,IID))%>%
  gather(key=ancestry, value=percentage,c("afr","amr","nfe"))%>%
  ggplot(aes(factor(fiid), percentage, fill = factor(ancestry))) +
  geom_col(size = 0.1) +
  facet_grid(~fct_inorder(self_race), scales = "free", switch = "x", space = "free") +
  theme_minimal() + 
  labs(x = "Individuals", title = paste0( "Average Ancestry proportions for eur, afr, amr"), y = "Ancestry") +
  scale_y_continuous(expand = c(0, 0)) +
  scale_x_discrete(expand = expand_scale(add = 1)) +
  theme(
    plot.title = element_text(hjust = 0.5),
    panel.spacing.x = unit(0.1, "lines"),
    axis.text.x = element_blank(),
    panel.grid = element_blank(),
    strip.text.x = element_text(angle = 90, size=7)
  ) +
  scale_fill_gdocs(guide = FALSE)
ggsave(gstring("ave_hgdp_1kg_prnd.5.png"),width = 8,height = 6,dpi = 500)


##################################################################
# a figure of ancestry proportions after averaging, k =3
library(tidyverse)
library(data.table)
library(forcats)
library(ggthemes)
library(patchwork)
library(R.utils)
#setwd("/scratch/lishaobo/GWAS/glioma/populationRisk2/adm5")
setwd("/project/lishaobo_902/sebastian/admixtureMapping/reviewers")

raw <- fread("/project/wiemels_260/sebastian/gwas_glioma/populationRisk2/poprisk2_pol_attributes.txt") %>%
  select(FID,IID, self_race, phenotype)
inds = fread("/project/wiemels_260/sebastian/gwas_glioma/populationRisk2/gnomad.genomes.v3.1.hgdp_1kg_subset.sample_meta.tsv") %>%
  select(s, labeled_subpop,population_inference.pop)
subhg <- read.table("/project/wiemels_494/sebastian/temp_scratch_backup/populationRisk2/hgdp_1kg_prnd.fam")%>%
  select(V1,V2) %>%
  rename("FID"="V1","IID" = "V2") %>%
  mutate(FIID = ifelse(FID==IID, IID, paste0(FID,"_",IID)))%>%
  left_join(inds, by=c("FIID" = "s")) %>%
  mutate(self_race=population_inference.pop,
         phenotype=1) %>%
  select(FID,IID, self_race, phenotype)
raw1 = rbind(raw,subhg) %>%as.data.frame  
subs <- read.table("/project/wiemels_494/sebastian/temp_scratch_backup/populationRisk2/glio.hgdp.1kg.fam")%>%
  select(V1,V2) %>%
  rename("FID"="V1","IID" = "V2") %>%
  left_join(raw1, by=c("FID","IID"))

afrPro = list()
amrPro = list()
nfePro = list()

for(i in 1:10){
  tbl=read.table(gstring("/project/wiemels_494/sebastian/temp_scratch_backup/populationRisk2/adm5/${i}/glio.hgdp.1kg.3.Q"))
  tbl_cmb <- cbind(subs,tbl) %>% as.data.frame()
  
  afr = tbl_cmb %>% filter(self_race == "afr") %>% 
    select(V1,V2,V3)
  afrAve = which.max(colMeans(afr)) %>% names
  
  amr = tbl_cmb %>% filter(self_race == "amr") %>% 
    select(V1,V2,V3)
  amrAve = which.max(colMeans(amr)) %>% names
  
  nfe = tbl_cmb %>% filter(self_race == "nfe") %>% 
    select(V1,V2,V3)
  nfeAve = which.max(colMeans(nfe)) %>% names
  
  afrPro[[i]] = tbl_cmb[,afrAve]
  amrPro[[i]] = tbl_cmb[,amrAve]
  nfePro[[i]] = tbl_cmb[,nfeAve]
}

afrProDF = do.call(rbind,afrPro) %>% as.data.frame() %>%colMeans()
amrProDF = do.call(rbind,amrPro) %>% as.data.frame() %>%colMeans()
nfeProDF = do.call(rbind,nfePro) %>% as.data.frame() %>%colMeans()

tbl_ave = data.frame(afr=afrProDF,amr=amrProDF,nfe=nfeProDF)
tbl_ave_1 = cbind(subs,tbl_ave) %>% 
  as.data.frame() 
rownames(tbl_ave_1) <- c()

f2AndF3Source = tbl_ave_1 %>%
  dplyr::select(-FID,-IID)
write.csv(f2AndF3Source,"f2AndF3Source.csv")

############Figure 2
tbl_ave_1%>%
  mutate(fiid=paste0(FID,IID))%>%
  gather(key=ancestry, value=percentage,c("afr","amr","nfe"))%>%
  mutate(ancestry.full = ifelse(self_race=="afr","AFR",
                                ifelse(self_race=="amr","AMR",
                                       ifelse(self_race=="nfe","EUR","Hispanic"))))%>%
  ggplot(aes(factor(fiid), percentage, fill = factor(ancestry))) +
  geom_col(size = 0.1) +
  facet_grid(~fct_inorder(ancestry.full), scales = "free", switch = "x", space = "free") +
  theme_minimal() + 
  labs(x = "Individuals", 
       title = paste0( "Estimated Ancestry proportions for Hipspanic case control subjects and\n reference samples"), 
       y = "Ancestry") +
  scale_y_continuous(expand = c(0, 0)) +
  scale_x_discrete(expand = expand_scale(add = 1)) +
  theme(
    plot.title = element_text(hjust = 0.5),
    panel.spacing.x = unit(0.1, "lines"),
    axis.text.x = element_blank(),
    panel.grid = element_blank(),
    strip.text.x = element_text(angle = 90, size=10)
  ) +
  scale_fill_gdocs(guide = FALSE)
ggsave(gstring("ave_hgdp_1kg_prnd.3.png"),width = 8,height = 6,dpi = 500)
write_rds(tbl_ave_1,"popPropLat.rds")

# Comparing white and amerindian ancestry in LAT subjects
his <- tbl_ave_1 %>% filter(self_race == "Hispanic") %>%
  mutate(CaCo = ifelse(phenotype==1,"control","case"))

write.table(his,"/project/wiemels_260/sebastian/gwas_glioma/populationRisk2/globalProportionsLat.txt",
            quote=FALSE, row.names = FALSE)

############################################################################################################
# Testing the distribution of white ancestries in latino subjects case/control
his = fread("/project/wiemels_260/sebastian/gwas_glioma/populationRisk2/globalProportionsLat.txt")
############Figure 3A
ggplot(his, aes(x=as.factor(CaCo), y=nfe, fill=as.factor(CaCo))) +
  geom_boxplot(outlier.size=0.2)+
  #ggtitle( "Distribution of European ancestry proportions in Hispanic \nPilocytic astrocytoma case control subjects" )+
  xlab("Case control status") +
  ylab("European proportions") +
  theme_classic() +
  theme(axis.text.x = element_text(size=13),
        axis.text.y = element_text(size=13),
        axis.title.x = element_text(size=15),
        axis.title.y = element_text(size=15),
        legend.title = element_text(size=14),
         legend.text = element_text(size=12))+
  geom_point(colour="darkblue",
             position=position_jitterdodge(),
             alpha=0.5,size=0.01) +  
  #theme(plot.title = element_text(hjust = 0.5)) +
  scale_fill_discrete(name="Disease status")
ggsave("3a_hisWhiteProportionsHGDP.png",width = 6 ,height = 6,dpi = 500)

median(his[his$CaCo=="control","nfe"])
#  0.51145
median(his[his$CaCo=="case","nfe"])
# 0.5501433
wilcox.test(his[his$CaCo=="control","nfe"],his[his$CaCo=="case","nfe"])
# 	Wilcoxon rank sum test with continuity correction
# 
# data:  his[his$CaCo == "control", "nfe"] and his[his$CaCo == "case", "nfe"]
# W = 158914, p-value = 0.003475
# alternative hypothesis: true location shift is not equal to 0

## Presenting the % is very underwhelming. Convert to a 5% increase and assign associated odds ratio. 
his1 = his %>%
  mutate(nfe_percen = (nfe*100)/5,
         status = factor(CaCo,levels=c("control","case")))
rg = glm(status~nfe_percen,family="binomial",data=his1)
summary(rg)
# Coefficients:
# Coefficients:
#             Estimate Std. Error z value Pr(>|z|)    
# (Intercept) -1.91703    0.21709  -8.830  < 2e-16 ***
# nfe_percen   0.05006    0.01868   2.679  0.00737 ** 

# 0.05006 (1.051334)  
# 0.05006-1.96*0.01868 = 0.0134472 (1.013538)
# 0.05006+1.96*0.01868 = 0.0866728 (1.09054)


# Testing the distribution of amr ancestries in latino subjects case/control
############Figure 3B
ggplot(his, aes(x=as.factor(CaCo), y=amr, fill=as.factor(CaCo))) +
  geom_boxplot(outlier.size=0.2)+
  theme_classic() +
  #ggtitle( "Distribution of Amerindian ancestry proportions in Hispanic \nPilocytic astrocytoma case control subjects" )+
  xlab("Case control status") +
  ylab("Amerindian Proportions") +
  theme(axis.text.x = element_text(size=13),
        axis.text.y = element_text(size=13),
        axis.title.x = element_text(size=15),
        axis.title.y = element_text(size=15),
        legend.title = element_text(size=14),
         legend.text = element_text(size=12))+
  geom_point(colour="darkblue",
             position=position_jitterdodge(),
             alpha=0.5,size=0.01) +  
  scale_fill_discrete(name="Disease status")
ggsave("3b_hisAmrProportionsHGDP.png",width = 6 ,height = 6,dpi = 500)

median(his[his$CaCo=="control","amr"])
# 0.4313664
median(his[his$CaCo=="case","amr"])
# 0.3975048
wilcox.test(his[his$CaCo=="control","amr"],his[his$CaCo=="case","amr"])
# 	Wilcoxon rank sum test with continuity correction
# 
# data:  his[his$CaCo == "control", "amr"] and his[his$CaCo == "case", "amr"]
# W = 199640, p-value = 0.001359
# alternative hypothesis: true location shift is not equal to 0

his = fread("/project/wiemels_260/sebastian/gwas_glioma/populationRisk2/globalProportionsLat.txt") %>%
  as.data.frame()
median(his[his$CaCo=="control","afr"])
# 0.0392897
median(his[his$CaCo=="case","afr"])
# 0.037594
wilcox.test(his[his$CaCo=="control","afr"],his[his$CaCo=="case","afr"])
# 
# 	Wilcoxon rank sum test with continuity correction
# 
# data:  his[his$CaCo == "control", "afr"] and his[his$CaCo == "case", "afr"]
# W = 186480, p-value = 0.2209
# alternative hypothesis: true location shift is not equal to 0

# Testing the distribution of white ancestries in different sites 
############Per kyle suggestion
glio2 = fread("/project/wiemels_260/sebastian/gwas_glioma/raw_data/B2SampleAttributes.txt") %>% 
  as.data.frame() %>%
  select(`Sample Filename`, SampleName,SubmittedPlate)%>%
  mutate(FID=0, SampleName=as.integer(SampleName), IID=paste0("0_",`Sample Filename`))%>%
  na.omit()%>% #delelte duplicates and inner controls
  dplyr::rename("subjectId" = "SampleName") 
glioS <- read.csv("/project/wiemels_260/sebastian/gwas_glioma/raw_data/likereallyraw/PedGlioma_4869_Full_Covariates_uncript.csv") %>% 
  as.data.frame() %>%
  dplyr::select(HISTO_T3,blindid,SITE_02) %>%
  inner_join(glio2,c("blindid"="subjectId")) %>%
  mutate(IID=gsub("0_","",IID))%>%
  dplyr::select(FID,IID,HISTO_T3,SITE_02)

his = fread("/project/wiemels_260/sebastian/gwas_glioma/populationRisk2/globalProportionsLat.txt") %>%
  as.data.frame() %>%
  inner_join(glioS,by=c("FID","IID")) %>%
  mutate(location = ifelse(SITE_02=="C723","optic",
                           ifelse(SITE_02!="C723"&SITE_02!="","non_optic",NA)))

his %>%
  filter(CaCo=="case",
         !is.na(SITE_02),
         SITE_02!="") %>%
  filter(SITE_02 %in% names(which(table(his$SITE_02)>10)))%>%
  ggplot(aes(x=SITE_02,y=nfe))+
  geom_boxplot()+
  theme_classic()
ggsave("ancestry.by.location.png", width = 8, height = 5)

his %>%
  filter(CaCo=="case",
         !is.na(SITE_02),
         SITE_02!="") %>%
  filter(SITE_02 %in% names(which(table(his$SITE_02)>10)))%>%
  ggplot(aes(x=location,y=nfe))+
  geom_boxplot()+
  theme_classic()
ggsave("ancestry.by.optic.png", width = 6, height = 5)

```

```{r Comparing global ancestry of latino case/controls in ast}

library(tidyverse)
#setwd("/scratch/lishaobo/GWAS/glioma/populationRisk2")

######################k=3 one time and plot, compare results

library(tidyverse)
library(data.table)
library(forcats)
library(ggthemes)
library(patchwork)
library(R.utils)

setwd("/project/lishaobo_902/admixtureMapping/reviewers")
#setwd("/scratch/lishaobo/GWAS/glioma/populationRisk2/adm5")

raw <- fread("/project/wiemels_260/sebastian/gwas_glioma/populationRisk2/poprisk2_ast_attributes.txt") %>%
  select(FID,IID, self_race, phenotype)

inds = fread("/project/wiemels_260/sebastian/gwas_glioma/populationRisk2/gnomad.genomes.v3.1.hgdp_1kg_subset.sample_meta.tsv") %>%
  select(s, labeled_subpop,population_inference.pop)
subhg <- read.table("../hgdp_1kg_prnd.fam")%>%
  select(V1,V2) %>%
  rename("FID"="V1","IID" = "V2") %>%
  mutate(FIID = ifelse(FID==IID, IID, paste0(FID,"_",IID)))%>%
  left_join(inds, by=c("FIID" = "s")) %>%
  mutate(self_race=population_inference.pop,
         phenotype=1) %>%
  select(FID,IID, self_race, phenotype)

raw1 = rbind(raw,subhg) %>%as.data.frame  

subs <- read.table("../glio.ast.hgdp.1kg.fam")%>%
  select(V1,V2) %>%
  rename("FID"="V1","IID" = "V2") %>%
  left_join(raw1, by=c("FID","IID"))

tbl=read.table(gstring("../glio.ast.hgdp.1kg.3.Q"))
tbl_cmb <- cbind(subs,tbl) %>% as.data.frame() %>%
  mutate(fiid=paste0(FID,IID))%>%
  arrange(self_race) 

tbl_cmb%>%
  gather(key=ancestry, value=percentage,c("V1","V2","V3"))%>%
  ggplot(aes(factor(fiid), percentage, fill = factor(ancestry))) +
  geom_col(size = 0.1) +
  facet_grid(~fct_inorder(self_race), scales = "free", switch = "x", space = "free") +
  theme_minimal() + 
  labs(x = "Individuals", title = paste0("Ast ancestry proportions, k=3"), y = "Ancestry") +
  scale_y_continuous(expand = c(0, 0)) +
  scale_x_discrete(expand = expand_scale(add = 1)) +
  theme(
    plot.title = element_text(hjust = 0.5),
    panel.spacing.x = unit(0.1, "lines"),
    axis.text.x = element_blank(),
    panel.grid = element_blank(),
    strip.text.x = element_text(angle = 90, size=7)
  ) +
  scale_fill_gdocs(guide = FALSE)
ggsave(gstring("ast.1kg.1time.3.png"),width = 8,height = 6,dpi = 500)

#####

tbl_ave = tbl_cmb %>%
  rename("afr"="V3",
         "nfe"="V2",
         "amr"="V1") %>% filter(self_race == "Hispanic") %>%
  mutate(CaCo = ifelse(phenotype==1,"control","case"))

write.table(tbl_ave,"/project/wiemels_260/sebastian/gwas_glioma/populationRisk2/globalProportionsLatAst.txt",
            quote=FALSE, row.names = FALSE)

# Testing the distribution of white ancestries in latino subjects case/control of ast
############Figure 3C
tbl_ave = fread("/project/wiemels_260/sebastian/gwas_glioma/populationRisk2/globalProportionsLatAst.txt")
ggplot(tbl_ave, aes(x=as.factor(CaCo), y=nfe, fill=as.factor(CaCo))) +
  geom_boxplot(outlier.size=0.2)+
  theme_classic() +
  #ggtitle( "Distribution of European ancestry proportions in Hispanic \nnon-pilocytic astrocytoma case control subjects" )+
  xlab("Case control status") +
  ylab("European Proportions") +
  theme(axis.text.x = element_text(size=13),
        axis.text.y = element_text(size=13),
        axis.title.x = element_text(size=15),
        axis.title.y = element_text(size=15),
        legend.title = element_text(size=14),
         legend.text = element_text(size=12))+
  geom_point(colour="darkblue",
             position=position_jitterdodge(),
             alpha=0.5,size=0.01) +  
  scale_fill_discrete(name="disease status")
ggsave("3c_hisAstWhiteProportionsHGDP.png",width = 6 ,height = 6,dpi = 500)

median(tbl_ave[tbl_ave$CaCo=="control","nfe"])
#  0.511812
median(tbl_ave[tbl_ave$CaCo=="case","nfe"])
# 0.525044
wilcox.test(tbl_ave[tbl_ave$CaCo=="control","nfe"],tbl_ave[tbl_ave$CaCo=="case","nfe"])
# 	Wilcoxon rank sum test with continuity correction
# 
# data:  tbl_ave[tbl_ave$CaCo == "control", "nfe"] and tbl_ave[tbl_ave$CaCo == "case", "nfe"]
# W = 138501, p-value = 0.2188
# alternative hypothesis: true location shift is not equal to 0
```

# Supplement table 1: demographic data on Eur + Lat

```{r}
library(tidyverse)
library(data.table)
setwd("/scratch/lishaobo/GWAS/glioma/populationRisk2")

# pilocytic and non-pilocytic astrocytoma

imp_race = read.csv("/scratch/lishaobo/GWAS/glioma/saxb4/self_reported_and_imputed_race.csv",row.names=1)%>%
  filter(self_race %in% c("White","Hispanic") ) 

glio <- fread("/project/wiemels_260/sebastian/gwas_glioma/saxb4/rawSampleAttributes.txt") %>%
  dplyr::filter((HISTO_T3 == 9421)|(HISTO_T3 >= 9400 & HISTO_T3 <= 9424) ) %>%
  select(FID, IID,subjectId) %>%
  inner_join(imp_race, by=c("FID","IID"))

# 772 is pol, 574 is ast

glio1 <- fread("/project/wiemels_260/sebastian/gwas_glioma/raw_data/likereallyraw/PedGlioma_4869_Full_Covariates_uncript.csv") %>%
  inner_join(glio, by=c("blindid"="subjectId"))%>%
  dplyr::select(SEX_CHILD,BTHWGT,GESTATION,self_race,AGE,SITE_02,DIFFERN2) 

dim(glio1)
# [1] 1346   8
table(glio1$SEX_CHILD)
#   1   2 
# 691 655 
summary(glio1$BTHWGT)
 # Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
 #     737    3140    3487    3452    3799    5840 
summary(glio1$GESTATION)
  # Min. 1st Qu.  Median    Mean 3rd Qu.    Max.    NA's 
  # 58.0   271.0   279.0   279.5   287.0   999.0      45 
table(glio1$self_race)
# Hispanic    White 
#   547      799 
summary(glio1$AGE)
 # Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
 #  0.000   3.000   6.000   7.263  11.000  19.000 
table(glio1$SITE_02)
#C692 C700 C701 C710 C711 C712 C713 C714 C715 C716 C717 C718 C719 C720 C723 C725 
#   1    1    1  156   58  114   45   15   62  369  187   78  120   70   54    8 
#C728 C753 
#   1    6 
table(glio1$DIFFERN2)
#   1   2   3   4   9 
# 167 222  14 179 764 



# non-pilocytic astrocytoma only
imp_race = read.csv("/scratch/lishaobo/GWAS/glioma/saxb4/self_reported_and_imputed_race.csv",row.names=1)%>%
  filter(self_race == "Hispanic") 
glio <- fread("/project/wiemels_260/sebastian/gwas_glioma/saxb4/rawSampleAttributes.txt") %>%
  dplyr::filter(phenotype==1| (HISTO_T3 >= 9400 & HISTO_T3 <= 9424  & HISTO_T3 != 9421)) %>%
  dplyr::select(FID, IID,subjectId,phenotype) %>%
  inner_join(imp_race, by=c("FID","IID"))
glio1 <- fread("/project/wiemels_260/sebastian/gwas_glioma/raw_data/likereallyraw/PedGlioma_4869_Full_Covariates_uncript.csv") %>%
  inner_join(glio, by=c("blindid"="subjectId"))%>%
  dplyr::select(SEX_CHILD,BTHWGT,GESTATION,self_race,AGE,SITE_02,DIFFERN2,phenotype,HISTO_T3) 

dim(glio1)
# 1076    8
summary(glio1$BTHWGT)
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
#    650    3062    3405    3368    3740    4933 
summary(glio1$GESTATION)
#Min. 1st Qu.  Median    Mean 3rd Qu.    Max.    NA's 
#64.0   268.0   277.0   276.4   285.0   637.0      33 
table(glio1$SEX_CHILD)
#   1   2 
# 691 655 
table(glio1$phenotype)
# 1   2 
#830 246
table(glio1$HISTO_T3)
#https://apps.who.int/iris/bitstream/handle/10665/96612/9789241548496_eng.pdf
##table 27
#9400 9401 9410 9411 9412 9413 9420 9424 
# 125   65    1    2    1   24   11   17 
summary(glio1$AGE)
 # Min. 1st Qu.  Median    Mean 3rd Qu.    Max.    NA's 
 #  0.000   3.000   6.000   7.045  11.000  19.000     830 
table(glio1$SITE_02)
#C714 C725 C753 C715 C723 C713 C720 C719 C711 C716 C718 C712 C710 C717      
#   3    3    3    9   10   13   13   15   16   17   21   39   40   44 
table(glio1$DIFFERN2)
#   1   2   3   4   9 
# 167 222  14 179 764 



```

# local ancestry comparison

```{bash Local ancestry regression in latino ancestry subjects}
#Local ancestry regression in latino ancestry subjects using RFmix

#cd /scratch/lishaobo/GWAS/glioma/populationRisk2/rgn
cd /project/wiemels_494/sebastian/temp_scratch_backup/populationRisk2/rgn


## Preparing files for the analysis
######################
# A phased VCF/BCF file containing "query" haplotypes which are to be analyzed

plink --bfile ../pol_adm.nodup\
  --recode vcf-iid bgz\
  --out lat.query
# 1486 subjects

## make glioma data conforming to that of reference data
for chr in {1..22};do
  java -jar /project/wiemels_260/sebastian/dependencies/beagle5/conform-gt.24May16.cee.jar\
  gt=lat.query.vcf.gz\
  ref=/project/wiemels_260/sebastian/gwas_1000G/GRCh38/ALL.chr$chr.shapeit2_integrated_snvindels_v2a_27022019.GRCh38.phased.vcf.gz\
  chrom=$chr\
  match=POS\
  out=$chr.ref.lat
done    

## converting ref vcf.gz file to bref3 files for faster phasing
for chr in {1..22};do
  java -jar /project/wiemels_260/sebastian/dependencies/beagle5/bref3.18May20.d20.jar\
    /project/wiemels_260/sebastian/gwas_1000G/GRCh38/ALL.chr$chr.shapeit2_integrated_snvindels_v2a_27022019.GRCh38.phased.vcf.gz > /project/wiemels_260/sebastian/gwas_1000G/GRCh38/ALL.chr$chr.bref3
done

## phase vcf files with beagle5
for chr in {1..22};do
  java -jar /project/wiemels_260/sebastian/dependencies/beagle5/beagle.18May20.d20.jar\
    gt=$chr.ref.lat.vcf.gz\
    ref=/project/wiemels_260/sebastian/gwas_1000G/GRCh38/ALL.chr$chr.bref3\
    out=$chr.query.phased\
    map=/project/wiemels_260/sebastian/dependencies/beagle5/plink.GRCh38.map/plink.chr$chr.GRCh38.map
done  

# count the number of SNPs per reviewer question
for chr in {1..22};do
  echo chr$chr
  bcftools stats $chr.query.phased.vcf.gz | grep "number of SNPs:"
done

#chr1:	5795045
#chr2:	6356815
#chr3:	5280536
#chr4:	5120664
#chr5:	4788374
#chr6:	4539754
#chr7:	4222931
#chr8:	4162377
#chr9:	3178999
#chr10:	3632297
#chr11:	3642067
#chr12:	3500318
#chr13:	2575087
#chr14:	2383125
#chr15:	2153932
#chr16:	2410531
#chr17:	2066684
#chr18:	2047353
#chr19:	1625698
#chr20:	1706442
#chr21:	976599
#chr22:	993880

######################
# A phased VCF/BCF file containing reference haplotypes (in any order)
# ref panel has too many SNPs. Only include overlaps with lat

awk '{print $2}' ../pol_adm.nodup.bim > lat.snps.list

plink --bfile ../slc.panel.hgdp.1kg.nodup\
  --extract lat.snps.list\
  --recode vcf bgz\
  --out panel.ref
# 1481 subjects

## make reference data conforming to that of reference data
for chr in {1..22};do
  java -jar /project/wiemels_260/sebastian/dependencies/beagle5/conform-gt.24May16.cee.jar\
  gt=panel.ref.vcf.gz\
  ref=/project/wiemels_260/sebastian/gwas_1000G/GRCh38/ALL.chr$chr.shapeit2_integrated_snvindels_v2a_27022019.GRCh38.phased.vcf.gz\
  chrom=$chr\
  match=POS\
  out=$chr.panel.ref
  
  tabix $chr.panel.ref.vcf.gz

done    

## phase vcf files with beagle5
for chr in {1..22};do
  java -Xmx100g -jar /project/wiemels_260/sebastian/dependencies/beagle5/beagle.18May20.d20.jar\
    gt=$chr.panel.ref.vcf.gz\
    ref=/project/wiemels_260/sebastian/gwas_1000G/GRCh38/ALL.chr$chr.bref3\
    out=$chr.panel.phased\
    map=/project/wiemels_260/sebastian/dependencies/beagle5/plink.GRCh38.map/plink.chr$chr.GRCh38.map

  tabix $chr.panel.phased.vcf.gz

done  

for chr in {1..22};do
  echo chr$chr
  bcftools stats $chr.panel.phased.vcf.gz | grep "number of SNPs:"
done


##############################################################################################################
# CCRLP European files for chr6,7,10,13 validation

cd /scratch/lishaobo/GWAS/glioma/populationRisk2

awk '{print $2, $1":"$4}' pol.adm.subs.wt.bim | sort -k2,2 | uniq -f1 -D  > pol_adm.wt.snps

plink --bfile pol.adm.subs.wt\
  --exclude pol_adm.wt.snps\
  --make-bed\
  --out pol.adm.subs.wt.nodup

cd /project/wiemels_494/sebastian/temp_scratch_backup/populationRisk2/rgn
#cd /scratch/lishaobo/GWAS/glioma/populationRisk2/rgn

plink --bfile /scratch/lishaobo/GWAS/glioma/populationRisk2/pol.adm.subs.wt.nodup\
  --recode vcf-iid bgz\
  --out eur.query
# 2040 subjects, 471 are cases and 1569 are controls

# chr=6
for chr in 6 7 10 13; do
  java -jar /project/wiemels_260/sebastian/dependencies/beagle5/conform-gt.24May16.cee.jar\
    gt=eur.query.vcf.gz\
    ref=/project/wiemels_260/sebastian/gwas_1000G/GRCh38/ALL.chr$chr.shapeit2_integrated_snvindels_v2a_27022019.GRCh38.phased.vcf.gz\
    chrom=$chr\
    match=POS\
    out=$chr.ref.eur
  
  ## phase vcf files with beagle5
    java -jar /project/wiemels_260/sebastian/dependencies/beagle5/beagle.18May20.d20.jar\
      gt=$chr.ref.eur.vcf.gz\
      ref=/project/wiemels_260/sebastian/gwas_1000G/GRCh38/ALL.chr$chr.bref3\
      out=$chr.query.phased.wt\
      map=/project/wiemels_260/sebastian/dependencies/beagle5/plink.GRCh38.map/plink.chr$chr.GRCh38.map
      
  tabix $chr.query.phased.wt.vcf.gz
done

for chr in {1..22};do
  echo chr$chr
  bcftools stats $chr.query.phased.wt.vcf.gz | grep "number of SNPs:"
done

```

```{bash try creating bcf files for rfmix running}

cd /project/wiemels_494/sebastian/temp_scratch_backup/populationRisk2/rgn

######################
# A reference sample map file matching reference samples to their respective reference populations
cat <(awk -v OFS='\t' '{print $1"_"$2,"eur"}' ../EURpop.txt)\
    <(awk -v OFS='\t' '{print $1"_"$2,"afr"}' ../AFRpop.txt)\
    <(awk -v OFS='\t' '{print $1"_"$2,"amr"}' ../AMRpop.txt) > samples.rgn.txt

######################
# Manipulating map file for rfmix
#chr=1
#awk -v OFS='\t' '{print $1,$4,$3}' /project/wiemels_260/sebastian/dependencies/beagle5/plink.GRCh38.map/plink.chr$chr.GRCh38.map > #/project/wiemels_260/sebastian/dependencies/rfmix/rfmix.GRCh38.map/rfmix.GRCh38.map
#for chr in {2..22};do
#  awk -v OFS='\t' '{print $1,$4,$3}' /project/wiemels_260/sebastian/dependencies/beagle5/plink.GRCh38.map/plink.chr$chr.GRCh38.map >> #/project/wiemels_260/sebastian/dependencies/rfmix/rfmix.GRCh38.map/rfmix.GRCh38.map
#done

```

```{bash script for rfmix, bcf file}
#!/bin/bash
#SBATCH --time=48:00:00
#SBATCH --mail-type=END
#SBATCH --job-name=rgnbcf
#SBATCH --array=1-22
#SBATCH --mem=160G

#sbatch /project/wiemels_260/sebastian/gwas_glioma/codes/temp3.sh

#cd /scratch/lishaobo/GWAS/glioma/populationRisk2/rgn
cd /project/wiemels_494/sebastian/temp_scratch_backup/populationRisk2/rgn

bcftools view -o ${SLURM_ARRAY_TASK_ID}.query.bcf -Ob\
  --threads 200\
  ${SLURM_ARRAY_TASK_ID}.query.phased.vcf.gz
bcftools index -f ${SLURM_ARRAY_TASK_ID}.query.bcf

bcftools view -o ${SLURM_ARRAY_TASK_ID}.panel.bcf -Ob\
  --threads 200\
  ${SLURM_ARRAY_TASK_ID}.panel.phased.vcf.gz
bcftools index -f ${SLURM_ARRAY_TASK_ID}.panel.bcf

rfmix\
    -f ${SLURM_ARRAY_TASK_ID}.query.bcf\
  	-r ${SLURM_ARRAY_TASK_ID}.panel.bcf\
  	-m samples.rgn.txt\
  	-g /project/wiemels_260/sebastian/dependencies/rfmix/rfmix.GRCh38.map/rfmix.GRCh38.map\
  	-o ${SLURM_ARRAY_TASK_ID}.bcf.rgn.lat\
	  --crf-weight=3\
  	--chromosome=${SLURM_ARRAY_TASK_ID}

# 21 Initial analysis - logl -510859337.0
# 22 Initial analysis - logl -510173966.5

```

```{r reviewer comment: comparing RFMix results with ADMIXTURE}

# The current gold standard program to infer global ancestry is RFMix. Please justify the use of ADMIXTURE, since RFMix was used to infer local ancestry.

# check the consistency of ADMIXURE and RFMix results

setwd("/project/lishaobo_902/sebastian/admixtureMapping/reviewers")
library(tidyverse)
library(data.table)
library(ggpmisc)

## ADMIXTURE results saved at
admi = fread("/project/wiemels_260/sebastian/gwas_glioma/populationRisk2/globalProportionsLat.txt")

## calculating RFMix results
weightF = function(a,i){
if(i==1){b = a*5795045/73159508}
if(i==2){b = a*6356815/73159508}
if(i==3){b = a*5280536/73159508}
if(i==4){b = a*5120664/73159508}
if(i==5){b = a*4788374/73159508}
if(i==6){b = a*4539754/73159508}
if(i==7){b = a*4222931/73159508}
if(i==8){b = a*4162377/73159508}
if(i==9){b = a*3178999/73159508}
if(i==10){b = a*3632297/73159508}
if(i==11){b = a*3642067/73159508}
if(i==12){b = a*3500318/73159508}
if(i==13){b = a*2575087/73159508}
if(i==14){b = a*2383125/73159508}
if(i==15){b = a*2153932/73159508}
if(i==16){b = a*2410531/73159508}
if(i==17){b = a*2066684/73159508}
if(i==18){b = a*2047353/73159508}
if(i==19){b = a*1625698/73159508}
if(i==20){b = a*1706442/73159508}
if(i==21){b = a*976599/73159508}
if(i==22){b = a*993880/73159508}
  return(b)
  }

rfLs = list()
for(i in 1:20){
 dfRF = fread(paste0("/project/wiemels_494/sebastian/temp_scratch_backup/populationRisk2/rgn/",i,".bcf.rgn.lat.rfmix.Q")) %>% 
   as.data.frame() %>%
   mutate(afr_w = weightF(afr,i),
          amr_w = weightF(amr,i),
          eur_w = weightF(eur,i))
 rfLs[[i]] = dfRF
}

for(i in 21:22){
  dfRF = fread(paste0("/project/wiemels_494/sebastian/temp_scratch_backup/populationRisk2/rgn/",i,".rgn.lat.rfmix.Q")) %>% 
    as.data.frame()%>%
   mutate(afr_w = weightF(afr,i),
          amr_w = weightF(amr,i),
          eur_w = weightF(eur,i))
 rfLs[[i]] = dfRF 
}

rfDFs = do.call(rbind,rfLs) %>% 
  as.data.frame()  %>%
  group_by(`#sample`) %>%
  summarise(afr_rfmix = sum(afr_w),
            amr_rfmix = sum(amr_w),
            eur_rfmix = sum(eur_w)) %>%
  as.data.frame() %>%
  inner_join(admi,by=c("#sample"="IID"))
  
rfDFs %>%
  ggplot(aes(x=afr_rfmix,y=afr))+
  stat_poly_line() +
  stat_poly_eq() +  
  geom_point()+
  #geom_abline(slope=1, intercept=0,color='red',alpha=0.3)+
  theme_classic() +
  xlab("African ancestry: RFMix") +
  ylab("African ancestry: ADMIXTURE") 
ggsave("afr.rfmix.admixture.comp.png",width = 5,height = 5)


rfDFs %>%
  ggplot(aes(x=amr_rfmix,y=amr))+
  stat_poly_line() +
  stat_poly_eq() +  
  geom_point()+
  #geom_abline(slope=1, intercept=0,color='red',alpha=0.3)+
  theme_classic() +
  xlab("Amerindian ancestry: RFMix") +
  ylab("Amerindian ancestry: ADMIXTURE") 
ggsave("amr.rfmix.admixture.comp.png",width = 5,height = 5)

rfDFs %>%
  ggplot(aes(x=eur_rfmix,y=nfe))+
  stat_poly_line() +
  stat_poly_eq() +  
  geom_point()+
  #geom_abline(slope=1, intercept=0,color='red',alpha=0.3)+
  theme_classic() +
  xlab("European ancestry: RFMix") +
  ylab("European ancestry: ADMIXTURE") 
ggsave("eur.rfmix.admixture.comp.png",width = 5,height = 5)

# comparing case and control ancestry differences using RFMix results
# Testing the distribution of white ancestries in latino subjects case/control
his = fread("/project/wiemels_260/sebastian/gwas_glioma/populationRisk2/globalProportionsLat.txt") %>%
  dplyr::select(IID) %>%
  inner_join(rfDFs,c("IID"="#sample"))
  
# 3A replication
ggplot(his, aes(x=as.factor(CaCo), y=eur_rfmix, fill=as.factor(CaCo))) +
  geom_boxplot(outlier.size=0.2)+
  xlab("Case control status") +
  ylab("European proportions") +
  theme_classic() +
  theme(axis.text.x = element_text(size=13),
        axis.text.y = element_text(size=13),
        axis.title.x = element_text(size=15),
        axis.title.y = element_text(size=15),
        legend.title = element_text(size=14),
         legend.text = element_text(size=12))+
  geom_point(colour="darkblue",
             position=position_jitterdodge(),
             alpha=0.5,size=0.01) +  
  scale_fill_discrete(name="Disease status")
#ggsave("3a_hisWhiteProportionsHGDP.png",width = 6 ,height = 6,dpi = 500)

median(his[his$CaCo=="control","eur_rfmix"]%>%as.matrix)
# 0.4221996
median(his[his$CaCo=="case","eur_rfmix"]%>%as.matrix)
# 0.4389163
wilcox.test(his[his$CaCo=="control","eur_rfmix"]%>%as.matrix,his[his$CaCo=="case","eur_rfmix"]%>%as.matrix)
# 	Wilcoxon rank sum test with continuity correction
# 
# data:  his[his$CaCo == "control", "eur_rfmix"] %>% as.matrix and his[his$CaCo == "case", "eur_rfmix"] %>% as.matrix
# W = 159319, p-value = 0.004218
# alternative hypothesis: true location shift is not equal to 0

# 3B replication
ggplot(his, aes(x=as.factor(CaCo), y=amr_rfmix, fill=as.factor(CaCo))) +
  geom_boxplot(outlier.size=0.2)+
  theme_classic() +
  xlab("Case control status") +
  ylab("Amerindian Proportions") +
  theme(axis.text.x = element_text(size=13),
        axis.text.y = element_text(size=13),
        axis.title.x = element_text(size=15),
        axis.title.y = element_text(size=15),
        legend.title = element_text(size=14),
         legend.text = element_text(size=12))+
  geom_point(colour="darkblue",
             position=position_jitterdodge(),
             alpha=0.5,size=0.01) +  
  scale_fill_discrete(name="Disease status")
#ggsave("3b_hisAmrProportionsHGDP.png",width = 6 ,height = 6,dpi = 500)

median(his[his$CaCo=="control","amr_rfmix"]%>%as.matrix)
# 0.4850451
median(his[his$CaCo=="case","amr_rfmix"]%>%as.matrix)
# 0.4696352
wilcox.test(his[his$CaCo=="control","amr_rfmix"]%>%as.matrix,his[his$CaCo=="case","amr_rfmix"]%>%as.matrix)
# 	Wilcoxon rank sum test with continuity correction
# 
# data:  his[his$CaCo == "control", "amr_rfmix"] %>% as.matrix and his[his$CaCo == "case", "amr_rfmix"] %>% as.matrix
# W = 199487, p-value = 0.001471
# alternative hypothesis: true location shift is not equal to 0

```


# rfmix using a different reference panel based on percentage only and validate if the results change

```{bash rfmix using a different reference panel based on percentage only and validate if the results change}
# select reference based on pure percentage .pr samples
cd /scratch/lishaobo/GWAS/glioma/populationRisk2

cat EURpop.pr.txt AFRpop.pr.txt AMRpop.pr.txt > ref.pr.hgdp.1kg.subs.txt

plink --bfile hgdp_1kg\
  --keep ref.pr.hgdp.1kg.subs.txt\
  --biallelic-only 'strict'\
  --snps-only 'just-acgt'\
  --make-bed\
  --out slc.panel.pr.hgdp.1kg.subs
# double checked: this is also on Hg38

cd /scratch/lishaobo/GWAS/glioma/populationRisk2

## delete duplicates with the same position
awk '{print $2, $1":"$4}' slc.panel.pr.hgdp.1kg.subs.bim | sort -k2,2 | uniq -f1 -D  > slc.panel.pr.hgdp.1kg.snps
awk '{print $1}' slc.panel.pr.hgdp.1kg.snps > all.snps
sort -uk2 slc.panel.pr.hgdp.1kg.snps | awk '{print $1}' > uniq.snps
awk 'NR==FNR{a[$0]=1;next}!a[$0]' uniq.snps all.snps > slc.panel.pr.hgdp.1kg.dups

plink --bfile slc.panel.pr.hgdp.1kg.subs \
  --exclude slc.panel.pr.hgdp.1kg.dups \
  --make-bed \
  --out slc.panel.pr.hgdp.1kg.nodup
awk '{print $1,"chr"$2,$3,$4,$5,$6}' slc.panel.pr.hgdp.1kg.nodup.bim > slc.panel.pr.hgdp.1kg.nodup.bim.bkup
mv slc.panel.pr.hgdp.1kg.nodup.bim.bkup slc.panel.pr.hgdp.1kg.nodup.bim

# make these two files include the same SNPs
awk '{print$2}' slc.panel.pr.hgdp.1kg.nodup.bim > 1kg.SNPs
plink --bfile pol_adm.nodup \
  --extract 1kg.SNPs \
  --make-bed \
  --out pol_adm.pr.merge1

awk '{print$2}' pol_adm.pr.merge1.bim > pol_adm.pr.merge1.SNPs
plink --bfile slc.panel.pr.hgdp.1kg.nodup \
  --extract pol_adm.pr.merge1.SNPs \
  --make-bed \
  --out hgdp.1kg.pr.merge1
# 719803 SNPs in total that are overlapped
  
#nano glio.hgdp.pr.1kg.mg
# pol_adm.pr.merge1 
# hgdp.1kg.pr.merge1

# Merge Glio and 1000G
plink --merge-list glio.hgdp.pr.1kg.mg \
  --make-bed \
  --out glio.hgdp.pr.1kg

#Local ancestry regression in latino ancestry subjects using RFmix

cd /scratch/lishaobo/GWAS/glioma/populationRisk2/rgn

## Preparing files for the analysis
######################
# A phased VCF/BCF file containing "query" haplotypes which are to be analyzed
head -100 ../pol_adm.nodup | awk '{print $1,$2}' > pol.adm.100.qry

for chr in 6 7 10 13; do

  # chr=1
  
  plink --bfile ../pol_adm.nodup\
    --keep pol.adm.100.qry\
    --chr $chr\
    --recode vcf-iid bgz\
    --out lat.query.pr
  # 1486 subjects
  
  ## make glioma data conforming to that of reference data
  java -jar /project/wiemels_260/sebastian/dependencies/beagle5/conform-gt.24May16.cee.jar\
    gt=lat.query.pr.vcf.gz\
    ref=/project/wiemels_260/sebastian/gwas_1000G/GRCh38/ALL.chr$chr.shapeit2_integrated_snvindels_v2a_27022019.GRCh38.phased.vcf.gz\
    chrom=$chr\
    match=POS\
    out=$chr.ref.lat.pr
  
  ## phase vcf files with beagle5
  java -jar /project/wiemels_260/sebastian/dependencies/beagle5/beagle.18May20.d20.jar\
   gt=$chr.ref.lat.pr.vcf.gz\
   ref=/project/wiemels_260/sebastian/gwas_1000G/GRCh38/ALL.chr$chr.bref3\
   out=$chr.query.pr.phased\
   map=/project/wiemels_260/sebastian/dependencies/beagle5/plink.GRCh38.map/plink.chr$chr.GRCh38.map
  
  ######################
  # A phased VCF/BCF file containing reference haplotypes (in any order)
  # ref panel has too many SNPs. Only include overlaps with lat
  
  awk '{print $2}' ../pol_adm.nodup.bim > lat.snps.list
  
  plink --bfile ../slc.panel.pr.hgdp.1kg.nodup\
    --extract lat.snps.list\
    --recode vcf bgz\
    --out panel.ref.pr
  
  java -jar /project/wiemels_260/sebastian/dependencies/beagle5/conform-gt.24May16.cee.jar\
    gt=panel.ref.pr.vcf.gz\
    ref=/project/wiemels_260/sebastian/gwas_1000G/GRCh38/ALL.chr$chr.shapeit2_integrated_snvindels_v2a_27022019.GRCh38.phased.vcf.gz\
    chrom=$chr\
    match=POS\
    out=$chr.panel.pr.ref
  
  java -Xmx100g -jar /project/wiemels_260/sebastian/dependencies/beagle5/beagle.18May20.d20.jar\
   gt=$chr.panel.pr.ref.vcf.gz\
   ref=/project/wiemels_260/sebastian/gwas_1000G/GRCh38/ALL.chr$chr.bref3\
   out=$chr.panel.pr.phased\
   map=/project/wiemels_260/sebastian/dependencies/beagle5/plink.GRCh38.map/plink.chr$chr.GRCh38.map
  
  tabix $chr.query.pr.phased.vcf.gz 
  tabix $chr.panel.pr.phased.vcf.gz 
done

######################
# A reference sample map file matching reference samples to their respective reference populations
cat <(awk -v OFS='\t' '{print $1"_"$2,"eur"}' ../EURpop.pr.txt)\
    <(awk -v OFS='\t' '{print $1"_"$2,"afr"}' ../AFRpop.pr.txt)\
    <(awk -v OFS='\t' '{print $1"_"$2,"amr"}' ../AMRpop.pr.txt) > samples.rgn.pr.txt


cd /scratch/lishaobo/GWAS/glioma/populationRisk2/rgn

sbatch /project/wiemels_260/sebastian/gwas_glioma/codes/temp.sh
sbatch --dependency=afterok:7410784 /project/wiemels_260/sebastian/gwas_glioma/codes/rfmix6.sh
sbatch --dependency=afterok:7410784 /project/wiemels_260/sebastian/gwas_glioma/codes/rfmix7.sh
sbatch --dependency=afterok:7410784 /project/wiemels_260/sebastian/gwas_glioma/codes/rfmix10.sh
sbatch --dependency=afterok:7410784 /project/wiemels_260/sebastian/gwas_glioma/codes/rfmix13.sh

# bash script for rfmix, bcf file}
#!/bin/bash
#SBATCH --time=168:00:00
#SBATCH --mail-type=END
#SBATCH --job-name=rgn1
#SBATCH --partition=largemem
#SBATCH --mem=800G

#sbatch /project/wiemels_260/sebastian/gwas_glioma/codes/rfmix1.sh

cd /scratch/lishaobo/GWAS/glioma/populationRisk2/rgn

chr=1

bcftools view -o $chr.query.pr.bcf -Ob\
  --threads 200\
  $chr.query.pr.phased.vcf.gz
bcftools index -f $chr.query.pr.bcf

bcftools view -o $chr.panel.pr.bcf -Ob\
  --threads 200\
  $chr.panel.pr.phased.vcf.gz
bcftools index -f $chr.panel.pr.bcf

rfmix\
    -f $chr.query.pr.bcf\
  	-r $chr.panel.pr.bcf\
  	-m samples.rgn.pr.txt\
  	-g /project/wiemels_260/sebastian/dependencies/rfmix/rfmix.GRCh38.map/rfmix.GRCh38.map\
  	-o $chr.bcf.rgn.pr.lat\
	  --crf-weight=3\
  	--chromosome=$chr


```

```{r compare if two reference panels yield a similar result (global)}

setwd("/scratch/lishaobo/GWAS/glioma/populationRisk2/rgn")
library(tidyverse)
library(data.table)

# 1.bcf.rgn.pr.lat.fb.tsv
# 1.bcf.rgn.pr.lat.rfmix.Q
# 
# 1.bcf.rgn.lat.fb.tsv
# 1.bcf.rgn.lat.rfmix.Q

# comparing global ancestry results
overall.1 = fread("1.bcf.rgn.lat.rfmix.Q") %>% 
  dplyr::rename("afr.old"="afr","amr.old"="amr","eur.old"="eur")
overall.2 = fread("1.bcf.rgn.pr.lat.rfmix.Q") %>% 
  dplyr::rename("afr.new"="afr","amr.new"="amr","eur.new"="eur") %>%
  dplyr::inner_join(overall.1, by="#sample")

overall.2 %>%
  ggplot(aes(x=afr.old,y=afr.new))+
  geom_point()+
  geom_smooth(method=lm)+
  theme_bw()+
  ggtitle("Comparing overal african ancestry in two reference panels")+
  theme(plot.title = element_text(hjust = 0.5))
ggsave("afrCompNewOld.png",width = 12,height = 8)


overall.2 %>%
  ggplot(aes(x=amr.old,y=amr.new))+
  geom_point()+
  geom_smooth(method=lm)+
  theme_bw()+
  ggtitle("Comparing overal amr ancestry in two reference panels")+
  theme(plot.title = element_text(hjust = 0.5))
ggsave("amrCompNewOld.png",width = 12,height = 8)


overall.2 %>%
  ggplot(aes(x=eur.old,y=eur.new))+
  geom_point()+
  geom_smooth(method=lm)+
  theme_bw()+
  ggtitle("Comparing overal eur ancestry in two reference panels")+
  theme(plot.title = element_text(hjust = 0.5))
ggsave("eurCompNewOld.png",width = 12,height = 8)

```

```{bash charleston's code for comparison}

cd /scratch/lishaobo/GWAS/glioma/populationRisk2/rgn

cp 1.bcf.rgn.pr.lat.fb.tsv ref.new
head -2 1.ref.new/bcf.rgn.pr.lat.fb.tsv > ref.new/hdr
head -2 1.bcf.rgn.lat.fb.tsv > ref.old/hdr

#a = fread("ref.old/hdr",skip="chromosome")
#b = fread("ref.new/hdr",skip="chromosome")
#ind = which(colnames(a)%in%colnames(b))
#cat(ind,sep=",")

cut -f 1-604 1.bcf.rgn.lat.fb.tsv > ref.old/1.bcf.rgn.lat.thinned.fb.tsv
head -2 ref.old/1.bcf.rgn.lat.thinned.fb.tsv > ref.old/hdr.thinned

#a = fread("ref.old/hdr.thinned",skip="chromosome")
#b = fread("ref.new/hdr",skip="chromosome")
#identical(a,b)

```

```{r compare_maps_improved.R}
#!env/bin/Rscript
# 
# This is a brief readme file about how to generate a 2-dimension contingency table to compare local ancestry.
# 
#  About the script
# Please reach out to the script: /home/rcf-proj/cc2/hanxiaos/compare_maps.R
# There is another version stored at ~/hanxiaos/181105_RFMix_NH. The only difference is whether the output is presented in percentage or not.
# 
# This script was initially designed to examine the consistency of local ancestry generated under different conditions. To be more specific, to check the generalization of conditions. For example, genetic map matters when calling local ancestry, as it defines the gene linkage of (a specific) ethnicity. Sometimes it is hard to say if a genetic map is suitable for a less well-studied population. Thus we have to call local ancestry by using different genetic maps and find out the most fitted one by testing the robustness. And vice versa, if a genetic map is proven to be robust enough, it could be applied to multiple scenarios.
# 
#  Basic concpets about the script
# The calculation is based on probabilities of haplotypes in ***.fb.tsv.
# For example: compare 2 maps
# 
# probabilities hap1:pop1 hap1:pop2  hap1:popn hap2:pop1 hap2:pop2  hap2:popn
#     map1      a11       a12        a1n       a21       a22        a2n
#     map2      b11       b12        b1n       b21       b22        b2n
# 
# It would generate a nn contingency-like table.
#                                  map1:
#                  pop1            pop2             popn
#       pop1 a11b11+a21b21 a12b11+a22b21  a1nb11+a2nb21
# map2: pop2 a11b12+a21b22 a12b12+a22b22  a1nb12+a2nb22
#       
#       popn a11b1n+a21b2n a12b1n+a22b2n  a1nb1n+a2nb2n
# 
# Each element presents the probability that the SNP is predicted as pop1 with map1, while as pop2 with map2.
# Therefore, compare_maps.R is designed to
# 1. get the dimension of ancestry population. The dimension should be equal and this is the basic assumption.
# 2. get the common SNPs.
# 3. get the common individuals.
# 4. calcualte the consistency.
# For more specific details, please read the comments in compare_maps.R.
# 
#  How to run the script
# First call local ancestry and store them in different folders according to generating conditions.
# Then submit the job of compare.sh to server. There are 5 external parameters to be specified. They can also be specified in compare.sh.
# sbatch compare.sh arg1 arg2 arg3 arg4 arg5.
# arg1 & arg2: the path where to find xx.fb.tsv.
# arg3 & arg4: the prefixed string of rownames and columnames as labels, eg: map1 in map1:pop1. The output file will add rownames and columnames as xx:popi automatically.
# arg5: the name of output files.
# Ps: Please do not store all tested xx.fb.tsv in one folder, they should be separately stored according to generating conditions. As currently the script cannot identify chromosomes of local ancestry, please do not store other untested xx.fb.tsv files in those folders, either. Please make sure that chromosomes of local ancestry are matched.
# PPs: There is actually a piece of code of getting common chromosomes in compare_maps.R, but it didn't work and was marked as comments.
# For more specific details, please read the instruction in compare.sh.
# 
#  Improvement
# Result interpretation: The result is a multi-dimensional McNemar test. Please check whether further dedcution or inference is required or not.
# Methodology: 1. debug the piece of code of getting common chromosomes 2. replace the for loop (accumulative function) with matrix multiplication. For each unit (per SNP per individual), assume AX=B, A = the matrix of parameters, X = the input file, ***.fb.tsv, B = contingency-like table. For each row in A, it is [,1,0,0,,1,0,0,]. Technically speaking, it would be more calculation-efficient.

# define an accumulative function that counts on something's amount
count<-function(a,b,d){
	interim1<-strsplit(a,b)
	interim2<-sapply(interim1,"[",d)
	interim3<-unique(interim2)
	return(interim3)
}

# define an accumulative function that generates chi-square-like table
cal<-function(a){
	d<-length(pop1) # get the length of admixed ancestry
	di<-length(a)/2 # get the number of local ancestry call personally in one map
	dindex<-seq(1,di,d) # index of local ancestry calls
	interim<-matrix(rep(0,d^2),ncol=d) # initialize the slot of chi-square table
	for(i in 1:(length(dindex)-1)){
		mat_a<-as.vector(a[dindex[i]:(dindex[i+1]-1)]) # get the local ancestry call in per haplotype in condition A
		mat_b<-as.vector(a[(dindex[i]+di):(dindex[i+1]-1+di)]) # same as above, but for condition B
		interim<-interim+mat_b %*% t(mat_a) # get the chi-square table per haplotype unit
		}
	interim<-as.vector(t(interim)) # transpose the chi-square table when converting into vectors
	return(interim)
}

# data files confirmation
args<-commandArgs(T)
temp1<-list.files(path=args[1],pattern="*.fb.tsv")
temp2<-list.files(path=args[2],pattern="*.fb.tsv")
#if(identical(length(temp1),length(temp2))==FALSE){
#	print("The amounts of the chromosomes are different. Please check it.")
#	q()
#}
#chr_stock1<-strsplit(temp1,"chr")
#chr1<-c(NULL)
#for(i in 1:length(chr_stock1))
#	chr1[i]<-chr_stock1[[i]][2]
#chr_stock2<-strsplit(temp2,"chr")
#chr2<-c(NULL)
#for(i in 1:length(chr_stock2))
#	chr2[i]<-chr_stock2[[i]][2]
#if(identical(chr1,chr2)==FALSE){
#	print("The chromosomes are mismatched. Please check it.")
#	q()
#}

# calculation
n=0 # snp amount
library('data.table') # ensure under the enviroment of data.table package
for(i in 1:length(temp1)){
	hap_file1<-fread(paste(args[1],temp1[i],sep="/"),header=TRUE,sep="\t",check.names=FALSE)
	hap_file2<-fread(paste(args[2],temp2[i],sep="/"),header=TRUE,sep="\t",check.names=FALSE)
	hap_file1<-as.data.frame(hap_file1)
	hap_file2<-as.data.frame(hap_file2)
	hap_file1<-hap_file1[ ,-c(1,3,4)] # only retain SNP ID
	hap_file2<-hap_file2[ ,-c(1,3,4)] # only retain SNP ID
	if(identical(hap_file1[ ,1],hap_file2[ ,1])==FALSE){
		common_snps<-intersect(hap_file1[ ,1],hap_file2[ ,1])
		hap_file1<-hap_file1[hap_file1[ ,1] %in% common_snps, ]	
		hap_file2<-hap_file2[hap_file2[ ,1] %in% common_snps, ]
	}else{
		common_snps<-hap_file1[ ,1]
	} # extract common snps
	print(paste(temp1[i],"snps amount",length(common_snps),sep=" "))
	if(i==1){
		temp3<-colnames(hap_file1[-1])
		pop1<-count(temp3,":::",3) # determine the admixture population in reference panel 1
		temp4<-colnames(hap_file2[-1])
                pop2<-count(temp4,":::",3) # determine the admixture population in reference panel 2
		if(identical(length(pop1),length(pop2))==FALSE){
			print("The number of admixed population in two reference panels are different. Please check it.")
			q()
		}
		indi1<-count(temp3,":::",1)
		indi2<-count(temp4,":::",1)
		common_indi<-intersect(indi1,indi2)
		index1<-indi1 %in% common_indi
		index1<-unlist(lapply(index1,rep,length(pop1)))
		index1<-c(TRUE,index1)
		index2<-indi2 %in% common_indi
		index2<-unlist(lapply(index2,rep,length(pop2)))
		index2<-c(TRUE,index2) # define the indicator variable to label common individuals
		sum_snps_indi<-rep(0,length(pop1)^2)
		sum_snps_chr<-rep(0,length(pop1)^2)
	}
	if(identical(index1,rep(TRUE,length(hap_file1)))==FALSE)
		hap_file1<-hap_file1[ ,index1] # extract the common individuals in panel 1
	if(identical(index2,rep(TRUE,length(hap_file2)))==FALSE)
		hap_file2<-hap_file2[ ,index2] # extract the common individuals in panel 2
	hap_data<-merge(hap_file1,hap_file2,by="physical_position",all=TRUE)
	hap_data<-hap_data[-1]
	sum_snps_indi<-apply(hap_data,1,cal)
	sum_snps_chr<-sum_snps_chr+apply(sum_snps_indi,1,sum)
	n<-n+length(common_snps)
}
print(paste("snps amount",n,sep=" "))
chi_matrix<-matrix(sum_snps_chr,ncol=length(pop1),byrow=T)
#chi_frame<-as.data.frame(round(chi_matrix,0))
#colnames(chi_frame)<-paste(args[3],pop1,sep=":")
#rownames(chi_frame)<-paste(args[4],pop2,sep=":")
#write.table(chi_frame,file=paste(args[5],"txt",sep="."),sep=" ",row.names=TRUE,col.names=TRUE,quote=FALSE)
percent_matrix<-round(chi_matrix/sum(chi_matrix),4)
percent_frame<-as.data.frame(percent_matrix)
#print(percent_frame)
colnames(percent_frame)<-paste(args[3],pop1,sep=":")
rownames(percent_frame)<-paste(args[4],pop2,sep=":")
write.table(percent_frame,file=paste(args[5],"txt",sep="."),sep=" ",row.names=TRUE,col.names=TRUE,quote=FALSE)

```

```{bash compare.sh}
#!/bin/bash
#SBATCH --time=36:00:00
#SBATCH --mem-per-cpu=16gb
#SBATCH --cpus-per-task=4
#SBATCH --job-name=comparison

set -e 
set -u 
set -x
set -o pipefail 

#####################################################################################
cd /scratch/lishaobo/GWAS/glioma/populationRisk2/rgn

#sbatch --account=wiemels_260 /project/wiemels_260/sebastian/gwas_glioma/codes/temp.sh

# This is designed for pairwise comparision between local ancestries result.
# arg1/arg2: address of the archive of fb.tsv files names as "***chr*.fb.tsv", please enter the absolute path
# arg3/arg4: column name/row name of the chi-square-like table
# arg5: output file name of this comparison

/usr/usc/R/default/bin/Rscript compare_maps_improved.R\
 /scratch/lishaobo/GWAS/glioma/populationRisk2/rgn/ref.old\
 /scratch/lishaobo/GWAS/glioma/populationRisk2/rgn/ref.new\
 old.ref new.ref compareChr1Refs

```

# Regional admixture mapping using RFMIX outputs

```{bash}
cd /scratch/lishaobo/GWAS/glioma/populationRisk2/rgn

plink --bfile ../pol_adm.nodup\
  --maf 0.1\
  --indep-pairwise 50 10 0.1\
  --out glio.lat    

plink2 --bfile ../pol_adm.nodup\
  --extract glio.lat.prune.in\
  --pca 10\
  --out glio.lat.pca\
  --threads 200
  
```

```{r prepare regression ready file}

library(tidyverse)
library(data.table)
setwd("/scratch/lishaobo/GWAS/glioma/populationRisk2")

# ancestry genotype file

  raw = fread("rgn/22.bcf.rgn.lat.msp.tsv") 
  subs = colnames(raw)[-c(1:6)]
  subjects = gsub(".[01]$", "",subs) %>% unique()
  phenoFile = data.frame(sub =subjects)

for(chr in 1:22){

  raw = fread(paste0("rgn/",chr,".bcf.rgn.lat.msp.tsv"))
  
  pos = raw[,c(1,2)]
  raw1 = raw[,-c(1:6)] 
  
  # afr=0 amr=1 eur=2
  raw2 = (raw1==2) 
  
  tem_sum_all <-list()
  for(i in 1:(length(subjects))){
    tem_a <- raw2[,i]+raw2[,i+1]
    tem_sum_all[[i]] <- tem_a
  }
  b2_numP1 <- data.frame(do.call(cbind, tem_sum_all)) 
  colnames(b2_numP1) = subjects
  b2_numP2 = cbind(pos,b2_numP1) %>% as.data.frame() %>%
    mutate(ID = paste0(`#chm`,"_",spos)) %>%
    select(-`#chm`,-spos)%>%
    column_to_rownames(var="ID")

  write.table(b2_numP2,paste0("rgn/chr",chr,"_eurCount.txt"), row.names=TRUE,col.names = TRUE, quote=FALSE)
  
}

# phenotype file
phe = fread("/scratch/lishaobo/GWAS/glioma/populationRisk2/pol_adm.nodup.fam")%>%
  mutate(sub=V2, sex=V5, caco = (V6-1)) %>%
  right_join(phenoFile, by="sub") %>%
  select(sub, sex, caco)

global = fread("/project/wiemels_260/sebastian/gwas_glioma/populationRisk2/globalProportionsLat.txt") %>%
  mutate(sub=IID, eurGlo = nfe, afrGlo=afr) %>%
  select(sub,eurGlo,afrGlo)%>%
  right_join(phe,by="sub") 

pcs = fread("/scratch/lishaobo/GWAS/glioma/populationRisk2/rgn/glio.lat.pca.eigenvec") %>%
  mutate(sub=IID) %>%
  select(sub, PC1, PC2, PC3, PC4, PC5, PC6, PC7, PC8, PC9, PC10)%>%
  right_join(global,by="sub") 

write.table(pcs, "rgn/lat.rfmix.pheono.txt", row.names = FALSE, quote=FALSE)

# correlation between PCs and CaCo
glm(pcs$caco~pcs$PC1+ pcs$PC2 +  pcs$PC3 + pcs$PC4  + pcs$PC5 + pcs$PC6 + pcs$PC7 + pcs$PC8 + pcs$PC9 + pcs$PC10, family="binomial") %>% summary()

# Coefficients:
#             Estimate Std. Error z value Pr(>|z|)    
# (Intercept) -1.39049    0.06567 -21.173  < 2e-16 ***
# pcs$PC1      7.64530    2.50177   3.056  0.00224 ** 
# pcs$PC2      1.13063    2.31022   0.489  0.62455    
# pcs$PC3      1.92163    2.29105   0.839  0.40161    
# pcs$PC4      2.98781    2.49371   1.198  0.23086    
# pcs$PC5      0.94512    2.53252   0.373  0.70900    
# pcs$PC6     -1.70737    2.51477  -0.679  0.49718    
# pcs$PC7      1.44349    2.53491   0.569  0.56905    
# pcs$PC8      1.45771    2.52919   0.576  0.56438    
# pcs$PC9     -3.08744    2.51417  -1.228  0.21944    
# pcs$PC10    -3.40063    2.50916  -1.355  0.17533    
# Later PCs are not correlated to CaCo status.

```

```{r regression of regional ancestry}

library(tidyverse)
library(data.table)
#setwd("/scratch/lishaobo/GWAS/glioma/populationRisk2/rgn")
setwd("/project/wiemels_494/sebastian/temp_scratch_backup/populationRisk2/rgn")

pheno = fread("lat.rfmix.pheono.txt") %>%
  mutate(caco=as.factor(caco)) %>%
  mutate(sex=as.factor(sex))

for(chr in 1:22){

  eur = read.table(paste0("chr",chr,"_eurCount.txt"))

  eur1 = eur %>%
    t() %>%
    as.data.frame()
  
  regressors = names(eur1)
  
  reg_out <- lapply(regressors,function(snp){c(data.frame(loc=snp),summary(
    glm(pheno$caco ~ eur1[,snp]+ pheno$sex + pheno$eurGlo+pheno$afrGlo, family="binomial"))$coefficients[2,1:4])
  })
  reg_df <- bind_rows(reg_out) %>% as.data.frame()

  reg_out1 <- lapply(regressors,function(snp){c(data.frame(loc=snp),summary(
    glm(pheno$caco ~ eur1[,snp]+ pheno$sex + pheno$eurGlo+pheno$afrGlo+pheno$PC1+ pheno$PC2 +  pheno$PC3 + pheno$PC4  + pheno$PC5 + pheno$PC6 + pheno$PC7 + pheno$PC8
        + pheno$PC9 + pheno$PC10, family="binomial"))$coefficients[2,1:4])
  })
  reg_df1 <- bind_rows(reg_out1) %>% as.data.frame()

  
  reg_out2 <- lapply(regressors,function(snp){c(data.frame(loc=snp),summary(
    glm(pheno$caco ~ eur1[,snp]+ pheno$eurGlo+pheno$afrGlo+pheno$PC1+ pheno$PC2 +  pheno$PC3 + pheno$PC4  + pheno$PC5 + pheno$PC6 + pheno$PC7 + pheno$PC8
        + pheno$PC9 + pheno$PC10, family="binomial"))$coefficients[2,1:4])
  })
  reg_df2 <- bind_rows(reg_out2) %>% as.data.frame()

  write.table(reg_df, paste0("reg.chr",chr,".out"), row.names = FALSE, quote=FALSE)
  write.table(reg_df1, paste0("reg.chr",chr,".pcs.out"), row.names = FALSE, quote=FALSE)
  write.table(reg_df2, paste0("reg.chr",chr,".nosex.pcs.out"), row.names = FALSE, quote=FALSE)

}

# Significant regional hits (lat.pol.pcs.long)
# chr6:85502415-85536682
# 
# awk '$5<7.753946e-06{print}' lat.pol.pcs.long 
# loc       Estimate Std. Error z value Pr(>|z|)
# 6_85502415 0.475013970481185 0.105433436915282 4.50534464567316 6.62653041680636e-06
# 6_85504599 0.465424440641071 0.101671585362287 4.57772384469684 4.70062673881385e-06
# 6_85509612 0.465424440641071 0.101671585362287 4.57772384469684 4.70062673881385e-06
# 6_85510292 0.465424440641071 0.101671585362287 4.57772384469684 4.70062673881385e-06
# 6_85511117 0.460038378272966 0.101507561866789 4.53206017179966 5.84112050212542e-06
# 6_85511242 0.460038378272966 0.101507561866789 4.53206017179966 5.84112050212542e-06
# 6_85512844 0.460038378272966 0.101507561866789 4.53206017179966 5.84112050212542e-06
# 6_85517080 0.460038378272966 0.101507561866789 4.53206017179966 5.84112050212542e-06
# 6_85520120 0.460038378272966 0.101507561866789 4.53206017179966 5.84112050212542e-06
# 6_85525874 0.460038378272966 0.101507561866789 4.53206017179966 5.84112050212542e-06
# 6_85526470 0.460038378272966 0.101507561866789 4.53206017179966 5.84112050212542e-06
# 6_85528498 0.454292620075753 0.101484283911257 4.47648249134824 7.58828880702667e-06
# 6_85536682 0.454292620075753 0.101484283911257 4.47648249134824 7.58828880702667e-06

```

```{bash combine results}

#cd /scratch/lishaobo/GWAS/glioma/populationRisk2/rgn
cd /project/wiemels_494/sebastian/temp_scratch_backup/populationRisk2/rgn

head -1 reg.chr1.out > lat.pol.long
tail -n +2 -q reg.chr?.out >> lat.pol.long
tail -n +2 -q reg.chr??.out >> lat.pol.long

head -1 reg.chr1.pcs.out > lat.pol.pcs.long
tail -n +2 -q reg.chr?.pcs.out >> lat.pol.pcs.long
tail -n +2 -q reg.chr??.pcs.out >> lat.pol.pcs.long

head -1 reg.chr1.nosex.pcs.out > lat.pol.nosex.pcs.long
tail -n +2 -q reg.chr?.nosex.pcs.out >> lat.pol.nosex.pcs.long
tail -n +2 -q reg.chr??.nosex.pcs.out >> lat.pol.nosex.pcs.long

```

Determine a genome-wide significance level

```{bash convert rfmix output to gds file for later use}

cd /scratch/lishaobo/GWAS/glioma/populationRisk2/rgn

for chr in {1..22}; do
  for race in "afr" "amr" "eur"; do

    # soyoung's code  
    python3 /project/wiemels_260/sebastian/gwas_glioma/codes/fbtoped.py \
      $chr.bcf.rgn.lat $race 0.9
    plink --tfile $chr.bcf.rgn.lat.$race --make-bed --out  chr$chr.$race.rgn.lat 
    
  done
done  
    
```

```{r determining a genomewide significance level, didn't work}

library(tidyverse)
library(data.table)
library(STEAM)
setwd("/scratch/lishaobo/GWAS/glioma/populationRisk2/rgn")

# prepare a map file
map = fread("/project/wiemels_260/sebastian/dependencies/rfmix/rfmix.GRCh38.map/rfmix.GRCh38.map")
colnames(map) = c("chr","pos","cM")

# population proportion file
prop = fread("/project/wiemels_260/sebastian/gwas_glioma/populationRisk2/globalProportionsLat.txt") %>%
  select(afr,amr,nfe)

## estimate the number of generations since admixture 

# convert plink files to gds using SNPRelate package
library(SeqArray)
for(chr in 1:22){
  for(race in c("eur","afr","amr")){
    bed.fn = paste0("chr",chr,".",race,".rgn.lat.bed")
    fam.fn = paste0("chr",chr,".",race,".rgn.lat.fam")
    bim.fn = paste0("chr",chr,".",race,".rgn.lat.bim")
    seqBED2GDS(bed.fn, fam.fn, bim.fn, paste0("chr",chr,".",race,".gds"))
  }
}

# compute correlation
corr.list <- list()
for(chr in 1:22){
  
  chrom=chr
  map.chr <- subset(map, chr == chrom)
  k.map=floor(nrow(map.chr)/22500) # otherwise it's too big for the function
  if(k.map!=0){
    indexMap = seq(1,nrow(map.chr),k.map+1)
    map.chr = map.chr[indexMap, ]
  }
  
  afr.gds <- paste0('chr', chr, '.afr.gds')
  eur.gds <- paste0('chr', chr, '.eur.gds')
  nam.gds <- paste0('chr', chr, '.amr.gds')
  corr.list[[chr]] <- get_corr_chr(chrom = chr, map = map.chr, 
        pop1.gds = afr.gds, pop2.gds = eur.gds, pop3.gds = nam.gds)}
corr_K3 <- combine_corr_chr(corr.list)
# get generation estimate
gval = get_g(corr_K3)
# calculate threshold
set.seed(1) 
get_thresh_simstat(g = gval, map = map, props = prop, nreps = 50)

```

Tsz Fung's codes for p value estimation

```{bash prepare input files}

cd /scratch/lishaobo/GWAS/glioma/populationRisk2/rgn

# RFMIX_fb_input.txt
nano RFMIX_fb_input.txt

/scratch/lishaobo/GWAS/glioma/populationRisk2/rgn/1.bcf.rgn.lat.fb.tsv
/scratch/lishaobo/GWAS/glioma/populationRisk2/rgn/2.bcf.rgn.lat.fb.tsv
/scratch/lishaobo/GWAS/glioma/populationRisk2/rgn/3.bcf.rgn.lat.fb.tsv
/scratch/lishaobo/GWAS/glioma/populationRisk2/rgn/4.bcf.rgn.lat.fb.tsv
/scratch/lishaobo/GWAS/glioma/populationRisk2/rgn/5.bcf.rgn.lat.fb.tsv
/scratch/lishaobo/GWAS/glioma/populationRisk2/rgn/6.bcf.rgn.lat.fb.tsv
/scratch/lishaobo/GWAS/glioma/populationRisk2/rgn/7.bcf.rgn.lat.fb.tsv
/scratch/lishaobo/GWAS/glioma/populationRisk2/rgn/8.bcf.rgn.lat.fb.tsv
/scratch/lishaobo/GWAS/glioma/populationRisk2/rgn/9.bcf.rgn.lat.fb.tsv
/scratch/lishaobo/GWAS/glioma/populationRisk2/rgn/10.bcf.rgn.lat.fb.tsv
/scratch/lishaobo/GWAS/glioma/populationRisk2/rgn/11.bcf.rgn.lat.fb.tsv
/scratch/lishaobo/GWAS/glioma/populationRisk2/rgn/12.bcf.rgn.lat.fb.tsv
/scratch/lishaobo/GWAS/glioma/populationRisk2/rgn/13.bcf.rgn.lat.fb.tsv
/scratch/lishaobo/GWAS/glioma/populationRisk2/rgn/14.bcf.rgn.lat.fb.tsv
/scratch/lishaobo/GWAS/glioma/populationRisk2/rgn/15.bcf.rgn.lat.fb.tsv
/scratch/lishaobo/GWAS/glioma/populationRisk2/rgn/16.bcf.rgn.lat.fb.tsv
/scratch/lishaobo/GWAS/glioma/populationRisk2/rgn/17.bcf.rgn.lat.fb.tsv
/scratch/lishaobo/GWAS/glioma/populationRisk2/rgn/18.bcf.rgn.lat.fb.tsv
/scratch/lishaobo/GWAS/glioma/populationRisk2/rgn/19.bcf.rgn.lat.fb.tsv
/scratch/lishaobo/GWAS/glioma/populationRisk2/rgn/20.bcf.rgn.lat.fb.tsv
/scratch/lishaobo/GWAS/glioma/populationRisk2/rgn/21.bcf.rgn.lat.fb.tsv
/scratch/lishaobo/GWAS/glioma/populationRisk2/rgn/22.bcf.rgn.lat.fb.tsv

# lat_indv.txt
awk 'NR>2 {print $1}' 1.bcf.rgn.lat.rfmix.Q > lat_indv.txt
```

```{r COMBINED.RFMIX.Q calculation}

setwd("/scratch/lishaobo/GWAS/glioma/populationRisk2/rgn")
library(tidyverse)
library(data.table)

est_list = list()

for(i in 1:22){
  est_list[[i]] = fread(paste0(i,".bcf.rgn.lat.rfmix.Q"))
  }

est_df=do.call(rbind,est_list)%>%as.data.frame() %>% 
  group_by(`#sample`) %>%
  summarise(avg_afr = mean(afr),
            avg_eur=mean(eur),
            avg_amr=mean(amr)) %>%
  select(-`#sample`)%>%
  as.data.frame()

write.table(est_df,"COMBINED.RFMIX.Q",sep = "\t", row.names = FALSE, quote = FALSE)

```

```{bash}
#!/bin/bash
#SBATCH --time=48:00:00
#SBATCH --mail-type=BEGIN
#SBATCH --mail-type=END
#SBATCH --mem=80GB
#SBATCH --job-name=sigP_zf

set -e 
set -u 
set -x
set -o pipefail 

#####################################################################################
#sbatch /project/wiemels_260/sebastian/gwas_glioma/codes/temp-1.sh

#########################################################################################
cd /scratch/lishaobo/GWAS/glioma/populationRisk2/rgn

# 1. A text file of a list of .fb.tsv
# 2. A Global ancestry file
# 3. Number of ancestry
# 4. A text file of a list of individuals
Rscript /project/wiemels_260/sebastian/gwas_glioma/codes/steam_run.R RFMIX_fb_input.txt COMBINED.RFMIX.Q 3 lat_indv.txt  > STEAM_threshold.out

#Estimated number of generation: 12.2221403460037
#Estimated Threshold: 7.75394645248347e-06
#$threshold
#         95% 
#7.753946e-06 

#$ci
#        2.5%        97.5% 
#8.289502e-06 6.955202e-06 

```

```{r steam_run.R}
library(STEAM)
library(tidyverse)
library(data.table)
#setwd("/scratch/lishaobo/GWAS/glioma/populationRisk2/rgn")
#args=c("RFMIX_fb_input.txt", "COMBINED.RFMIX.Q",3,"lat_indv.txt")

# Arguments
# 1. local ancestry file .fb.tsv
# 2. global ancestry file average of 22 chromosomes
# 3. number of population in you .fb.tsv
# 4. list of individuals
args <- commandArgs(trailingOnly=T)
RFMIX_LOCAL <- args[1]   
RFMIX_GLOBAL <- args[2]
NUM_ANC_POP <- as.numeric(args[3])
SELECTED_SAMPLE_FILE <-args[4]
ANC_IDX <- 3
#        The index of the ancestry of your interest. 
#        e.g. if the header of RFMIX is (AFR EUR NA ASN) and you are interested in EUR, put 2 in this argument

cat(paste0("Local Ancestry file list: ", RFMIX_LOCAL), "\n")
cat(paste0("Global Ancestry file: ", RFMIX_GLOBAL), "\n")
cat(paste0("Number of ancestral Population: ", NUM_ANC_POP), "\n")
cat(paste0("Selected sample file: ", SELECTED_SAMPLE_FILE), "\n")
cat(paste0("Selected ancestry: ancestry ", ANC_IDX), "\n")
require(Rcpp)
# Loop thru all marker pairs and calculate correlations, return a vector of correlations of all marker pairs
cppFunction('
            NumericVector corrC(NumericVector SNP1, NumericVector SNP2, NumericMatrix anc_matrix1, NumericMatrix anc_matrix2){
            
            Environment stats("package:stats");
            
            Function corr=stats["cor"];
            
            int n = SNP1.size();
            NumericVector out (n);
            // 
            
            for(int i=0; i < n;i++){
              out[i] = as<double>(corr(anc_matrix1(SNP1[i]-1,_), anc_matrix2(SNP2[i]-1,_)));
            }
            
            
            return out;
            
            }'
)

get_corr_table2 <- function(loc_anc, num_anc, anc_idx){
    
# function that RFMIX local ancestry correlation matrix
# Argument
# --------
#    loc_anc: table (number of markers, 4 + num_anc*2*num_individual)
#        the local ancestry files read from RFMIX output .fb.tsv
#    num_anc: int
#        number of ancestry in the RFMIX output
#    anc_idx: int
#        The index of the ancestry of your interest. 
#        e.g. if the header of RFMIX is (AFR EUR NA ASN) and you are interested in EUR, put 2 in this argument
# Return
# ------
# Prepare correlation tables:
# {bin: double
#     distance between markers in cM, round down to the nearest 0.5-cM bin
#     e.g. 0.0, 0.5, 1.0, 1.5 
#  corr: double
#     pearson correlation
#  anc: character
#     ancesteries pairs
#     e.g. 1_1 for pop1 vs pop1, 2_1 for pop2 vs pop1
# }



  # marker downsampling by bin
  cat("Downsampling markers...\n")
  markers <- tibble("index"=1:dim(loc_anc)[1], "cM"=loc_anc[, 3, drop=T])
  #markers <- markers %>% sample_n(10000)# memory allocation not enough
  markers_pair <- as_tibble(t(combn(markers$index, 2))) 
  markers_pair <- markers_pair %>% mutate("bins"=(abs(markers[markers_pair$V1,2]-markers[markers_pair$V2,2])%/%0.5*0.5)[[1]]) # binning
  markers_pair_downsample <- markers_pair %>% group_by(bins) %>% sample_n(min(20, n())) %>% ungroup() # downsample
  
  # make loc_anc matrices
  cat("Converting RFMix2 output to matrix...\n")
  loc_anc_list <- list()
  num_col <- dim(loc_anc)[2]
  ### print
  print(loc_anc)
  print(num_col)
  ### print
  cat(paste("...", num_anc, "ancestries found...\n"))
    # add up two haplo
  loc_anc_list[[1]] <- data.matrix(loc_anc[, seq(4+anc_idx, num_col, num_anc*2)] + loc_anc[, seq(4+anc_idx+num_anc, num_col, num_anc*2)])
  loc_anc_list[[2]] <- 2-loc_anc_list[[1]]

  # calculate corr between selected marker pairs
  cat("Calculate correlations...\n")
  corr_tbl_list <- list()
    # loop thru all selected marker pairs within corrC
    corrC.res <- corrC(markers_pair_downsample$V1, markers_pair_downsample$V2, loc_anc_list[[1]], loc_anc_list[[1]])
    corr_tbl_list[[1]] <- markers_pair_downsample %>% mutate("corr"=corrC.res, "anc"="1_1")
    corr_tbl_list[[2]] <- markers_pair_downsample %>% mutate("corr"=-corrC.res, "anc"="1_2")
    corr_tbl_list[[3]] <- markers_pair_downsample %>% mutate("corr"=-corrC.res, "anc"="2_1")
    corr_tbl_list[[4]] <- markers_pair_downsample %>% mutate("corr"=corrC.res, "anc"="2_2")
    cat(paste0("...Finished ancestry ...\n"))
  corr_tbl <- bind_rows(corr_tbl_list) %>% select(bins, corr, anc) %>% `colnames<-`(c("bin", "corr", "anc"))
  
  return(corr_tbl)
}

############# Main Script starts here #############

# Read RFMIX output ----
# 1. correlation tables
# 2. genetic map
FILE_LIST <- read_table(RFMIX_LOCAL, col_names=F)$X1
print(FILE_LIST)
input_corr_list <- list()
genetic_map_list <- list()
for( i in 1:length(FILE_LIST)){
    # read thru all local ancestry files
    # make a list of corr_table and genetic map
    
    loc_anc <- fread(FILE_LIST[i])
    loc_anc = loc_anc[sample(1:nrow(loc_anc),10000),] %>% arrange(physical_position) %>% as.data.frame()
    #  loc_anc <- read_delim(FILE_LIST[i], skip = 1, col_names = T, delim = "\t")
    
    cat(paste0("Reading", FILE_LIST[i], "\n"))
    sample_name <- str_split(colnames(loc_anc)[-c(1:4)], ":::", simplify = T)[, 1]
	  selected_sample <- read_delim(SELECTED_SAMPLE_FILE, col_names = F, delim = "\t")$X1
	  loc_anc <- loc_anc[ , c(1:4, which(sample_name %in% selected_sample)+4)]
	  
    input_corr_list[[i]] <- get_corr_table2(loc_anc, NUM_ANC_POP, ANC_IDX)
    genetic_map_list[[i]] <- data.frame("cM"=loc_anc[, 3, drop=T], "chr"=loc_anc[, 1, drop=T])

    
    cat(paste0("Read", FILE_LIST[i], "\n"))
    print(loc_anc[, c(3:6)])
}

# Construct observed local ancestry correlation vs recombination fraction theta ----
# 1. combine correlation tables from all chr
input_corr.combined <- bind_rows(input_corr_list) %>% 
  group_by(bin, anc) %>% summarise(corr = mean(corr)) %>% ungroup() 
input_corr.combined.theta <- input_corr.combined %>% mutate(theta=0.5*(1-exp(-2*bin/100))) %>% select(theta, corr, anc)
write.csv(input_corr.combined.theta, file=paste0(RFMIX_LOCAL, "csv"))

# Genetic map ----
# 2. combine genetic map
genetic_map.combined <- bind_rows(genetic_map_list)

# Estimate g ----
num_gens <-  get_g(as.data.frame(input_corr.combined.theta))

# Read Global Ancestry ----
global_anc <- fread(RFMIX_GLOBAL)
global_anc <- global_anc[which(unique(sample_name) %in% selected_sample), ]
global_anc <- cbind(rowSums(global_anc[, c(1,3)]), global_anc[, 2])
rownames(global_anc) <- 1:dim(global_anc)[1]
colnames(global_anc) <- c("P0", "P1")

# Estimate Threshold
system.time({
  thres.res <- get_thresh_simstat(g=num_gens, map = genetic_map.combined, props = global_anc)
})

p_thres <- thres.res$threshold[[1]]

cat(paste0("Estimated number of generation: ", num_gens, "\n"))
cat(paste0("Estimated Threshold: ", p_thres, "\n"))
print(thres.res)

```

```{r visualization}

library(tidyverse)
library(QCEWAS)
library(qqman)
library(data.table)

setwd("/scratch/lishaobo/GWAS/glioma/populationRisk2/rgn")

############################################################  

filename = "lat.pol.pcs.long"

gwas_data <- fread("lat.pol.pcs.long",skip=1)
colnames(gwas_data) = c("ID", "Estimate","StdError","Zvalue","P")

gwas_data_05 = gwas_data %>%
  mutate(SNPID = ID) %>%
  separate(SNPID, c("CHROM","POS"))%>%
  mutate(CHROM = as.numeric(CHROM),
         POS = as.numeric(POS)) %>%
  arrange(CHROM,POS)

gwas_data_1 <- gwas_data_05[is.finite(gwas_data_05$P),  ] %>%
  filter(!is.na(P))

tiff(paste0(filename,".manhattan.tiff"), width = 10, height = 7, units = 'in', res = 300)
manhattan(gwas_data_1, chr = "CHROM", bp = "POS", p = "P", snp = "ID",
          genomewideline=-log10(7.75394645248347e-06),
          suggestiveline=FALSE)
dev.off()

############################################################  
# show a zoomed in plot of the chr6 peak from the original admixture analysis for comparing with fine mapping 
## (chr6 1mb)
filename = "lat.pol.pcs.long"

gwas_data <- fread("lat.pol.pcs.long",skip=1)
colnames(gwas_data) = c("ID", "Estimate","StdError","Zvalue","P")

gwas_data_05 = gwas_data %>%
  mutate(SNPID = ID) %>%
  separate(SNPID, c("CHROM","POS"))%>%
  mutate(CHROM = as.numeric(CHROM),
         POS = as.numeric(POS)) %>%
  arrange(CHROM,POS) %>%
  filter(CHROM==6 & POS>=85012844 & POS <=86012844)

gwas_data_1 <- gwas_data_05[is.finite(gwas_data_05$P),  ] %>%
  filter(!is.na(P))

gwas_data_1 %>%
  ggplot(aes(x=POS,y=-log10(P)))+
  geom_point(size=2, alpha=0.6)+
  theme_bw()+
  #ggtitle("Zoomed plot for admixture association analysis")+
  #xlab("Chromosome 6 positions")+
  geom_hline(yintercept = -log10(7.75394645248347e-06), colour="red",alpha=0.5)+
  scale_x_continuous(labels = function(x) format(x, scientific = FALSE))+
  #theme(plot.title = element_text(hjust = 0.5))
  theme(axis.text.x = element_text(size=13),
      axis.text.y = element_text(size=13),
      axis.title.x = element_text(size=15),
      axis.title.y = element_text(size=15))
  
ggsave(paste0(filename,"zoomed.chr6.1mb.manhattan.png"),width = 10, height = 7)

##(chr6 3mb)
gwas_data_05 = gwas_data %>%
  mutate(SNPID = ID) %>%
  separate(SNPID, c("CHROM","POS"))%>%
  mutate(CHROM = as.numeric(CHROM),
         POS = as.numeric(POS)) %>%
  arrange(CHROM,POS) %>%
  filter(CHROM==6 & POS>=84009612 & POS <=87009612)

gwas_data_1 <- gwas_data_05[is.finite(gwas_data_05$P),  ] %>%
  filter(!is.na(P))

gwas_data_1 %>%
  ggplot(aes(x=POS,y=-log10(P)))+
  geom_point(size=2, alpha=0.6)+
  theme_bw()+
  #ggtitle("Zoomed plot for admixture association analysis")+
  #xlab("Chromosome 6 positions")+
  geom_hline(yintercept = -log10(7.75394645248347e-06), colour="red",alpha=0.5)+
  scale_x_continuous(labels = function(x) format(x, scientific = FALSE))+
  #theme(plot.title = element_text(hjust = 0.5))
  theme(axis.text.x = element_text(size=13),
      axis.text.y = element_text(size=13),
      axis.title.x = element_text(size=15),
      axis.title.y = element_text(size=15))
  
ggsave(paste0(filename,"zoomed.chr6.3mb.manhattan.png"),width = 10, height = 7)

## (chr6 5mb)
gwas_data_05 = gwas_data %>%
  mutate(SNPID = ID) %>%
  separate(SNPID, c("CHROM","POS"))%>%
  mutate(CHROM = as.numeric(CHROM),
         POS = as.numeric(POS)) %>%
  arrange(CHROM,POS) %>%
  filter(CHROM==6 & POS>=83012844 & POS <=88012844)

gwas_data_1 <- gwas_data_05[is.finite(gwas_data_05$P),  ] %>%
  filter(!is.na(P))

gwas_data_1 %>%
  ggplot(aes(x=POS,y=-log10(P)))+
  geom_point(size=2, alpha=0.6)+
  theme_bw()+
  #ggtitle("Zoomed plot for admixture association analysis")+
  #xlab("Chromosome 6 positions")+
  geom_hline(yintercept = -log10(7.75394645248347e-06), colour="red",alpha=0.5)+
  scale_x_continuous(labels = function(x) format(x, scientific = FALSE))+
  #theme(plot.title = element_text(hjust = 0.5))
  theme(axis.text.x = element_text(size=13),
      axis.text.y = element_text(size=13),
      axis.title.x = element_text(size=15),
      axis.title.y = element_text(size=15))
  
ggsave(paste0(filename,"zoomed.chr6.5mb.manhattan.png"),width = 10, height = 7)

############################################################  
# show a zoomed in plot of the chr7 peak from the original admixture analysis for comparing with fine mapping 
## (chr7 1mb)
filename = "lat.pol.pcs.long"

gwas_data <- fread("lat.pol.pcs.long",skip=1)
colnames(gwas_data) = c("ID", "Estimate","StdError","Zvalue","P")

gwas_data_05 = gwas_data %>%
  mutate(SNPID = ID) %>%
  separate(SNPID, c("CHROM","POS"))%>%
  mutate(CHROM = as.numeric(CHROM),
         POS = as.numeric(POS)) %>%
  arrange(CHROM,POS) %>%
  filter(CHROM==7 & POS>=29897103 & POS <=30897103)

gwas_data_1 <- gwas_data_05[is.finite(gwas_data_05$P),  ] %>%
  filter(!is.na(P))

gwas_data_1 %>%
  ggplot(aes(x=POS,y=-log10(P)))+
  geom_point(size=2, alpha=0.6)+
  theme_bw()+
  #ggtitle("Zoomed plot for admixture association analysis")+
  #xlab("Chromosome 6 positions")+
  geom_hline(yintercept = -log10(7.75394645248347e-06), colour="red",alpha=0.5)+
  scale_x_continuous(labels = function(x) format(x, scientific = FALSE))+
  #theme(plot.title = element_text(hjust = 0.5))
  theme(axis.text.x = element_text(size=13),
      axis.text.y = element_text(size=13),
      axis.title.x = element_text(size=15),
      axis.title.y = element_text(size=15))
  
ggsave(paste0(filename,"zoomed.chr7.1mb.manhattan.png"),width = 10, height = 7)

##  (chr7 2mb)

gwas_data_05 = gwas_data %>%
  mutate(SNPID = ID) %>%
  separate(SNPID, c("CHROM","POS"))%>%
  mutate(CHROM = as.numeric(CHROM),
         POS = as.numeric(POS)) %>%
  arrange(CHROM,POS) %>%
  filter(CHROM==7 & POS>=29397103 & POS <=31397103)

gwas_data_1 <- gwas_data_05[is.finite(gwas_data_05$P),  ] %>%
  filter(!is.na(P))

gwas_data_1 %>%
  ggplot(aes(x=POS,y=-log10(P)))+
  geom_point(size=2, alpha=0.6)+
  theme_bw()+
  #ggtitle("Zoomed plot for admixture association analysis")+
  #xlab("Chromosome 6 positions")+
  geom_hline(yintercept = -log10(7.75394645248347e-06), colour="red",alpha=0.5)+
  scale_x_continuous(labels = function(x) format(x, scientific = FALSE))+
  #theme(plot.title = element_text(hjust = 0.5))
  theme(axis.text.x = element_text(size=13),
      axis.text.y = element_text(size=13),
      axis.title.x = element_text(size=15),
      axis.title.y = element_text(size=15))
  
ggsave(paste0(filename,"zoomed.chr7.2mb.manhattan.png"),width = 10, height = 7)

##  (chr7 5mb)
gwas_data_05 = gwas_data %>%
  mutate(SNPID = ID) %>%
  separate(SNPID, c("CHROM","POS"))%>%
  mutate(CHROM = as.numeric(CHROM),
         POS = as.numeric(POS)) %>%
  arrange(CHROM,POS) %>%
  filter(CHROM==7 & POS>=27897103 & POS <=32897103)

gwas_data_1 <- gwas_data_05[is.finite(gwas_data_05$P),  ] %>%
  filter(!is.na(P))

gwas_data_1 %>%
  ggplot(aes(x=POS,y=-log10(P)))+
  geom_point(size=2, alpha=0.6)+
  theme_bw()+
  #ggtitle("Zoomed plot for admixture association analysis")+
  #xlab("Chromosome 6 positions")+
  geom_hline(yintercept = -log10(7.75394645248347e-06), colour="red",alpha=0.5)+
  scale_x_continuous(labels = function(x) format(x, scientific = FALSE))+
  #theme(plot.title = element_text(hjust = 0.5))
  theme(axis.text.x = element_text(size=13),
      axis.text.y = element_text(size=13),
      axis.title.x = element_text(size=15),
      axis.title.y = element_text(size=15))
  
ggsave(paste0(filename,"zoomed.chr7.5mb.manhattan.png"),width = 10, height = 7)

############################################################  
# show a zoomed in plot of the chr10 peak from the original admixture analysis for comparing with fine mapping 
## (chr10 1mb)
filename = "lat.pol.pcs.long"

gwas_data <- fread("lat.pol.pcs.long",skip=1)
colnames(gwas_data) = c("ID", "Estimate","StdError","Zvalue","P")

gwas_data_05 = gwas_data %>%
  mutate(SNPID = ID) %>%
  separate(SNPID, c("CHROM","POS"))%>%
  mutate(CHROM = as.numeric(CHROM),
         POS = as.numeric(POS)) %>%
  arrange(CHROM,POS) %>%
  filter(CHROM==10 & POS>=28875488 & POS <=29875488)

gwas_data_1 <- gwas_data_05[is.finite(gwas_data_05$P),  ] %>%
  filter(!is.na(P))

gwas_data_1 %>%
  ggplot(aes(x=POS,y=-log10(P)))+
  geom_point(size=2, alpha=0.6)+
  theme_bw()+
  #ggtitle("Zoomed plot for admixture association analysis")+
  #xlab("Chromosome 6 positions")+
  geom_hline(yintercept = -log10(7.75394645248347e-06), colour="red",alpha=0.5)+
  scale_x_continuous(labels = function(x) format(x, scientific = FALSE))+
  #theme(plot.title = element_text(hjust = 0.5))
  theme(axis.text.x = element_text(size=13),
      axis.text.y = element_text(size=13),
      axis.title.x = element_text(size=15),
      axis.title.y = element_text(size=15))
  
ggsave(paste0(filename,"zoomed.chr10.1mb.manhattan.png"),width = 10, height = 7)

## (chr10 2mb)
gwas_data_05 = gwas_data %>%
  mutate(SNPID = ID) %>%
  separate(SNPID, c("CHROM","POS"))%>%
  mutate(CHROM = as.numeric(CHROM),
         POS = as.numeric(POS)) %>%
  arrange(CHROM,POS) %>%
  filter(CHROM==10 & POS>=28375488 & POS <=30375488)

gwas_data_1 <- gwas_data_05[is.finite(gwas_data_05$P),  ] %>%
  filter(!is.na(P))

gwas_data_1 %>%
  ggplot(aes(x=POS,y=-log10(P)))+
  geom_point(size=2, alpha=0.6)+
  theme_bw()+
  #ggtitle("Zoomed plot for admixture association analysis")+
  #xlab("Chromosome 6 positions")+
  geom_hline(yintercept = -log10(7.75394645248347e-06), colour="red",alpha=0.5)+
  scale_x_continuous(labels = function(x) format(x, scientific = FALSE))+
  #theme(plot.title = element_text(hjust = 0.5))
  theme(axis.text.x = element_text(size=13),
      axis.text.y = element_text(size=13),
      axis.title.x = element_text(size=15),
      axis.title.y = element_text(size=15))
  
ggsave(paste0(filename,"zoomed.chr10.2mb.manhattan.png"),width = 10, height = 7)

## (chr10 5mb)
gwas_data_05 = gwas_data %>%
  mutate(SNPID = ID) %>%
  separate(SNPID, c("CHROM","POS"))%>%
  mutate(CHROM = as.numeric(CHROM),
         POS = as.numeric(POS)) %>%
  arrange(CHROM,POS) %>%
  filter(CHROM==10 & POS>=26875488 & POS <=31875488)

gwas_data_1 <- gwas_data_05[is.finite(gwas_data_05$P),  ] %>%
  filter(!is.na(P))

gwas_data_1 %>%
  ggplot(aes(x=POS,y=-log10(P)))+
  geom_point(size=2, alpha=0.6)+
  theme_bw()+
  #ggtitle("Zoomed plot for admixture association analysis")+
  #xlab("Chromosome 6 positions")+
  geom_hline(yintercept = -log10(7.75394645248347e-06), colour="red",alpha=0.5)+
  scale_x_continuous(labels = function(x) format(x, scientific = FALSE))+
  #theme(plot.title = element_text(hjust = 0.5))
  theme(axis.text.x = element_text(size=13),
      axis.text.y = element_text(size=13),
      axis.title.x = element_text(size=15),
      axis.title.y = element_text(size=15))
  
ggsave(paste0(filename,"zoomed.chr10.5mb.manhattan.png"),width = 10, height = 7)

############################################################  
# show a zoomed in plot of the chr13 peak from the original admixture analysis for comparing with fine mapping 
## (chr13 1mb)

filename = "lat.pol.pcs.long"

gwas_data <- fread("lat.pol.pcs.long",skip=1)
colnames(gwas_data) = c("ID", "Estimate","StdError","Zvalue","P")

gwas_data_05 = gwas_data %>%
  mutate(SNPID = ID) %>%
  separate(SNPID, c("CHROM","POS"))%>%
  mutate(CHROM = as.numeric(CHROM),
         POS = as.numeric(POS)) %>%
  arrange(CHROM,POS) %>%
  filter(CHROM==13 & POS>=93945215 & POS <=94945215)

gwas_data_1 <- gwas_data_05[is.finite(gwas_data_05$P),  ] %>%
  filter(!is.na(P))

gwas_data_1 %>%
  ggplot(aes(x=POS,y=-log10(P)))+
  geom_point(size=2, alpha=0.6)+
  theme_bw()+
  #ggtitle("Zoomed plot for admixture association analysis")+
  #xlab("Chromosome 6 positions")+
  geom_hline(yintercept = -log10(7.75394645248347e-06), colour="red",alpha=0.5)+
  scale_x_continuous(labels = function(x) format(x, scientific = FALSE))+
  #theme(plot.title = element_text(hjust = 0.5))
  theme(axis.text.x = element_text(size=13),
      axis.text.y = element_text(size=13),
      axis.title.x = element_text(size=15),
      axis.title.y = element_text(size=15))
  
ggsave(paste0(filename,"zoomed.chr13.1mb.manhattan.png"),width = 10, height = 7)

## (chr13 2mb)
gwas_data_05 = gwas_data %>%
  mutate(SNPID = ID) %>%
  separate(SNPID, c("CHROM","POS"))%>%
  mutate(CHROM = as.numeric(CHROM),
         POS = as.numeric(POS)) %>%
  arrange(CHROM,POS) %>%
  filter(CHROM==13 & POS>=93445215 & POS <=95445215)

gwas_data_1 <- gwas_data_05[is.finite(gwas_data_05$P),  ] %>%
  filter(!is.na(P))

gwas_data_1 %>%
  ggplot(aes(x=POS,y=-log10(P)))+
  geom_point(size=2, alpha=0.6)+
  theme_bw()+
  #ggtitle("Zoomed plot for admixture association analysis")+
  #xlab("Chromosome 6 positions")+
  geom_hline(yintercept = -log10(7.75394645248347e-06), colour="red",alpha=0.5)+
  scale_x_continuous(labels = function(x) format(x, scientific = FALSE))+
  #theme(plot.title = element_text(hjust = 0.5))
  theme(axis.text.x = element_text(size=13),
      axis.text.y = element_text(size=13),
      axis.title.x = element_text(size=15),
      axis.title.y = element_text(size=15))
  
ggsave(paste0(filename,"zoomed.chr13.2mb.manhattan.png"),width = 10, height = 7)

## (chr13 5mb)
gwas_data_05 = gwas_data %>%
  mutate(SNPID = ID) %>%
  separate(SNPID, c("CHROM","POS"))%>%
  mutate(CHROM = as.numeric(CHROM),
         POS = as.numeric(POS)) %>%
  arrange(CHROM,POS) %>%
  filter(CHROM==13 & POS>=91945215 & POS <=96945215)

gwas_data_1 <- gwas_data_05[is.finite(gwas_data_05$P),  ] %>%
  filter(!is.na(P))

gwas_data_1 %>%
  ggplot(aes(x=POS,y=-log10(P)))+
  geom_point(size=2, alpha=0.6)+
  theme_bw()+
  #ggtitle("Zoomed plot for admixture association analysis")+
  #xlab("Chromosome 6 positions")+
  geom_hline(yintercept = -log10(7.75394645248347e-06), colour="red",alpha=0.5)+
  scale_x_continuous(labels = function(x) format(x, scientific = FALSE))+
  #theme(plot.title = element_text(hjust = 0.5))
  theme(axis.text.x = element_text(size=13),
      axis.text.y = element_text(size=13),
      axis.title.x = element_text(size=15),
      axis.title.y = element_text(size=15))
  
ggsave(paste0(filename,"zoomed.chr13.5mb.manhattan.png"),width = 10, height = 7)

############################################################  
# show specific regions for comparing purposes (lat locuszoom)
filename = "lat.pol.pcs.long"

gwas_data <- fread("lat.pol.pcs.long",skip=1)
colnames(gwas_data) = c("ID", "Estimate","StdError","Zvalue","P")

gwas_data_05 = gwas_data %>%
  mutate(SNPID = ID) %>%
  separate(SNPID, c("CHROM","POS"))%>%
  mutate(CHROM = as.numeric(CHROM),
         POS = as.numeric(POS)) %>%
  arrange(CHROM,POS) %>%
  filter(CHROM==6 & POS>=84600000 & POS <=85400000)

gwas_data_1 <- gwas_data_05[is.finite(gwas_data_05$P),  ] %>%
  filter(!is.na(P))

gwas_data_1 %>%
  ggplot(aes(x=POS,y=-log10(P)))+
  geom_point(size=2, alpha=0.6)+
  theme_bw()+
  ggtitle("Zoomed plot for admixture association analysis")+
  xlab("Chromosome 6 positions")+
  geom_hline(yintercept = -log10(7.75394645248347e-06), colour="red",alpha=0.5)+
  scale_x_continuous(labels = function(x) format(x, scientific = FALSE))+
  theme(plot.title = element_text(hjust = 0.5))
ggsave(paste0(filename,"zoomed.forlatgwas.manhattan.png"),width = 10, height = 7)

############################################################  
# show specific regions for comparing purposes (eur locuszoom)
filename = "lat.pol.pcs.long"

gwas_data <- fread("lat.pol.pcs.long",skip=1)
colnames(gwas_data) = c("ID", "Estimate","StdError","Zvalue","P")

gwas_data_05 = gwas_data %>%
  mutate(SNPID = ID) %>%
  separate(SNPID, c("CHROM","POS"))%>%
  mutate(CHROM = as.numeric(CHROM),
         POS = as.numeric(POS)) %>%
  arrange(CHROM,POS) %>%
  filter(CHROM==6 & POS>=85000000 & POS <=85800000)

gwas_data_1 <- gwas_data_05[is.finite(gwas_data_05$P),  ] %>%
  filter(!is.na(P))

gwas_data_1 %>%
  ggplot(aes(x=POS,y=-log10(P)))+
  geom_point(size=2, alpha=0.6)+
  theme_bw()+
  ggtitle("Zoomed plot for admixture association analysis")+
  xlab("Chromosome 6 positions")+
  geom_hline(yintercept = -log10(7.75394645248347e-06), colour="red",alpha=0.5)+
  scale_x_continuous(labels = function(x) format(x, scientific = FALSE))+
  theme(plot.title = element_text(hjust = 0.5))
ggsave(paste0(filename,"zoomed.foreurgwas.manhattan.png"),width = 10, height = 7)


############################################################  
# show specific regions for isafe comparison (1MB)
filename = "lat.pol.pcs.long"

gwas_data <- fread("lat.pol.pcs.long",skip=1)
colnames(gwas_data) = c("ID", "Estimate","StdError","Zvalue","P")

gwas_data_05 = gwas_data %>%
  mutate(SNPID = ID) %>%
  separate(SNPID, c("CHROM","POS"))%>%
  mutate(CHROM = as.numeric(CHROM),
         POS = as.numeric(POS)) %>%
  arrange(CHROM,POS) %>%
  filter(CHROM==6 & POS>=84566601 & POS <=86012844)

gwas_data_1 <- gwas_data_05[is.finite(gwas_data_05$P),  ] %>%
  filter(!is.na(P))

gwas_data_1 %>%
  ggplot(aes(x=POS,y=-log10(P)))+
  geom_point(size=2, alpha=0.6)+
  theme_bw()+
  #ggtitle("Zoomed plot for admixture association analysis")+
  xlab("Chromosome 6 positions")+
  geom_hline(yintercept = -log10(7.75394645248347e-06), colour="red",alpha=0.5)+
  scale_x_continuous(labels = function(x) format(x, scientific = FALSE))+
  #theme(plot.title = element_text(hjust = 0.5))
  theme(axis.text.x = element_text(size=13),
      axis.text.y = element_text(size=13),
      axis.title.x = element_text(size=15),
      axis.title.y = element_text(size=15))
  
ggsave(paste0(filename,"zoomed.forisafe.1mb.manhattan.png"),width = 10, height = 7)

############################################################  
# show specific regions for isafe comparison (5MB)
filename = "lat.pol.pcs.long"

gwas_data <- fread("lat.pol.pcs.long",skip=1)
colnames(gwas_data) = c("ID", "Estimate","StdError","Zvalue","P")

gwas_data_05 = gwas_data %>%
  mutate(SNPID = ID) %>%
  separate(SNPID, c("CHROM","POS"))%>%
  mutate(CHROM = as.numeric(CHROM),
         POS = as.numeric(POS)) %>%
  arrange(CHROM,POS) %>%
  filter(CHROM==6 & POS>=82789723 & POS <=87789722)

gwas_data_1 <- gwas_data_05[is.finite(gwas_data_05$P),  ] %>%
  filter(!is.na(P))

gwas_data_1 %>%
  ggplot(aes(x=POS,y=-log10(P)))+
  geom_point(size=2, alpha=0.6)+
  theme_bw()+
  #ggtitle("Zoomed plot for admixture association analysis")+
  xlab("Chromosome 6 positions")+
  geom_hline(yintercept = -log10(7.75394645248347e-06), colour="red",alpha=0.5)+
  scale_x_continuous(labels = function(x) format(x, scientific = FALSE))+
  #theme(plot.title = element_text(hjust = 0.5))
  theme(axis.text.x = element_text(size=13),
      axis.text.y = element_text(size=13),
      axis.title.x = element_text(size=15),
      axis.title.y = element_text(size=15))
  
ggsave(paste0(filename,"zoomed.forisafe.5mb.manhattan.png"),width = 10, height = 7)

############################################################  

filename = "lat.pol.long"
  
gwas_data <- fread("lat.pol.long",skip=1)
colnames(gwas_data) = c("ID", "Estimate","StdError","Zvalue","P")

gwas_data_05 = gwas_data %>%
  mutate(SNPID = ID) %>%
  separate(SNPID, c("CHROM","POS"))%>%
  mutate(CHROM = as.numeric(CHROM),
         POS = as.numeric(POS)) %>%
  arrange(CHROM,POS)

gwas_data_1 <- gwas_data_05[is.finite(gwas_data_05$P),  ] %>%
  filter(!is.na(P))

tiff(paste0(filename,".manhattan.tiff"), width = 10, height = 7, units = 'in', res = 300)
manhattan(gwas_data_1, chr = "CHROM", bp = "POS", p = "P", snp = "ID")
dev.off()




############################################################  
# plotting relationship between beta and p value

filename = "lat.pol.pcs.long"

gwas_data <- fread("lat.pol.pcs.long",skip=1)
colnames(gwas_data) = c("ID", "Estimate","StdError","Zvalue","P")

gwas_data_05 = gwas_data %>%
  mutate(SNPID = ID) %>%
  separate(SNPID, c("CHROM","POS"))%>%
  mutate(CHROM = as.numeric(CHROM),
         POS = as.numeric(POS)) %>%
  arrange(CHROM,POS) %>%
  filter(CHROM==6 & POS>=85019548 & POS <=86019548)

gwas_data_1 <- gwas_data_05[is.finite(gwas_data_05$P),  ] %>%
  filter(!is.na(P))

gwas_data_1 %>%
  ggplot(aes(x=Estimate,y=-log10(P)))+
  geom_point(size=2, alpha=0.6)+
  theme_bw()+
  theme(axis.text.x = element_text(size=13),
      axis.text.y = element_text(size=13),
      axis.title.x = element_text(size=15),
      axis.title.y = element_text(size=15))
ggsave("betaAndPInPeak_AM.png",width = 10, height = 7)


```

look up or regional hits in GWAS hits

```{r look up or regional hits in GWAS hits}

#setwd("/scratch/lishaobo/GWAS/glioma/populationRisk2/rgn")
setwd("/project/wiemels_494/sebastian/temp_scratch_backup/populationRisk2/rgn")
library(tidyverse)
library(data.table)

reg = fread("lat.pol.pcs.long",skip=1) %>%
  mutate(reg_P=V5,
         locID=V1,
         reg_Estimate=V2)%>%
  arrange(reg_P)%>%
  select(locID,reg_Estimate,reg_P)

reg.sig=reg %>% filter(reg_P<7.753946e-06)

#        locID reg_Estimate        reg_P
#1: 6_85504599    0.4560379 6.113021e-06
#2: 6_85509612    0.4560379 6.113021e-06
#3: 6_85510292    0.4560379 6.113021e-06
#4: 6_85511117    0.4502469 7.752537e-06
#5: 6_85511242    0.4502469 7.752537e-06
#6: 6_85512844    0.4502469 7.752537e-06

## also check out other peaks that could be significant
reg.sig.other=reg %>% 
  separate(locID,sep="_",c("chr","pos"))

reg.sig.7 = reg.sig.other %>%
  filter(chr==7) %>%
  arrange(reg_P)

# 1:   7 30397103   -0.4830828 1.643601e-05
# 2:   7 30406545   -0.4830828 1.643601e-05
# 3:   7 30413668   -0.4792003 1.694277e-05

reg.sig.10 = reg.sig.other %>%
  filter(chr==10) %>%
  arrange(reg_P)
#    chr      pos reg_Estimate        reg_P
# 1:  10 29375488   -0.3995560 1.548852e-05
# 2:  10 29376844   -0.3978194 1.724949e-05
# 3:  10 29365234   -0.3948557 1.980317e-05


reg.sig.13 = reg.sig.other %>%
  filter(chr==13) %>%
  arrange(reg_P)
#    chr      pos reg_Estimate        reg_P
# 1:  13 94445215   -0.4237807 1.210850e-05
# 2:  13 94451891   -0.4181232 1.483341e-05
# 3:  13 94441872   -0.4081355 2.248226e-05


## look up in european subjects only
gwas=fread("/scratch/lishaobo/GWAS/glioma/saxb4/eur.pol.02.glio.1kg.PHENO1.glm.logistic") %>%
  mutate(SNP=ID,gwas_P=P,gwas_Estimate=BETA)%>%
  separate(ID,c("chr","loc","ref","alt"),sep=":")%>%
  mutate(locID=paste0(chr,"_",loc)) %>%
  mutate(locID=gsub("chr","",locID)) %>%
  select(SNP, locID, gwas_P, gwas_Estimate)
inter = inner_join(reg.sig,gwas,by="locID")

#         locID reg_Estimate        reg_P               SNP   gwas_P
# 1: 6_85526470    0.4600384 5.841121e-06 chr6:85526470:C:A 0.217718
# 2: 6_85502415    0.4750140 6.626530e-06 chr6:85502415:C:A 0.214417
# 3: 6_85536682    0.4542926 7.588289e-06 chr6:85536682:A:G 0.789415
#    gwas_Estimate
# 1:     0.0963361
# 2:    -0.4757630
# 3:    -0.0752498
# 

## look up in eur+lat subjects
gwas=fread("/scratch/lishaobo/GWAS/glioma/saxb4/meta.pol.eula.02.glio.1kg.PHENO1.glm.logistic.TBL") %>%
  mutate(SNP=MarkerName,gwas_P=`P-value`,gwas_Estimate=Effect,gwas_direction=Direction)%>%
  separate(MarkerName,c("chr","loc","ref","alt"),sep=":")%>%
  mutate(locID=paste0(chr,"_",loc)) %>%
  mutate(locID=gsub("chr","",locID)) %>%
  select(SNP, locID, gwas_P, gwas_Estimate, gwas_direction)
  
inter = inner_join(reg.sig,gwas,by="locID")

#         locID reg_Estimate        reg_P               SNP gwas_P gwas_Estimate
# 1: 6_85526470    0.4600384 5.841121e-06 chr6:85526470:C:A 0.4537        0.0451
# 2: 6_85502415    0.4750140 6.626530e-06 chr6:85502415:C:A 0.2144       -0.4758
# 3: 6_85536682    0.4542926 7.588289e-06 chr6:85536682:A:G 0.5796        0.1145
#    gwas_direction
# 1:             +-
# 2:             -?
# 3:             -+

## look up in all races 
gwas=fread("/scratch/lishaobo/GWAS/glioma/saxb4/meta.pol.02.glio.1kg.PHENO1.glm.logistic.TBL") %>%
  mutate(SNP=MarkerName,gwas_P=`P-value`,gwas_Estimate=Effect,gwas_direction=Direction)%>%
  separate(MarkerName,c("chr","loc","ref","alt"),sep=":")%>%
  mutate(locID=paste0(chr,"_",loc)) %>%
  mutate(locID=gsub("chr","",locID)) %>%
  select(SNP, locID, gwas_P, gwas_Estimate, gwas_direction)
  
inter = inner_join(reg.sig,gwas,by="locID")

#        locID reg_Estimate        reg_P               SNP gwas_P gwas_Estimate
# 1: 6_85526470    0.4600384 5.841121e-06 chr6:85526470:C:A 0.5342        0.0353
# 2: 6_85502415    0.4750140 6.626530e-06 chr6:85502415:C:A 0.2144       -0.4758
# 3: 6_85536682    0.4542926 7.588289e-06 chr6:85536682:A:G 0.6693        0.0852
#    gwas_direction
# 1:           +---
# 2:           -???
# 3:           -+-?

```

chr6:85512844 is in the middle of regional hit

enrichment of CORSIVs

```{r}

setwd("/project/wiemels_494/sebastian/temp_scratch_backup/populationRisk2/rgn")
library(tidyverse)
library(data.table)
library(GenomicRanges)

reg = fread("lat.pol.pcs.long",skip=1) %>%
  mutate(reg_P=V5,
         locID=V1,
         reg_Estimate=V2)%>%
  arrange(reg_P)%>%
  dplyr::select(locID,reg_Estimate,reg_P) %>%
  mutate(regID = locID,
         reg_P = as.numeric(reg_P)) %>%
  separate(locID,c("chr","pos")) 
reg$corsiv = 0

#reading corsiv
## this file is confirmed to be hg38
corsivs = read.csv("/project/wiemels_260/sebastian/CORSIV/all_corsivs.csv") %>%
  as.data.frame() %>%
  mutate(coor38 = gsub("chr[0-9]:|chr[0-9][0-9]:","",UCSC.Coordinates,perl=TRUE)) %>%
  separate(coor38,c("start","end")) %>%
  dplyr::select(Chromosome,start,end)
# 26446 bins and 9926 corsivs

am_GR <- makeGRangesFromDataFrame(reg, keep.extra.columns=T,
                                      ignore.strand=T,
                                      seqinfo=NULL,
                                      seqnames.field="chr",
                                      start.field="pos",
                                      end.field="pos",
                                      strand.field="strand",
                                     starts.in.df.are.0based=FALSE)
corsiv_GR <- makeGRangesFromDataFrame(corsivs, keep.extra.columns=T,
                                       ignore.strand=T,
                                       seqinfo=NULL,
                                       seqnames.field="Chromosome",
                                       start.field="start",
                                       end.field="end",
                                       strand.field="strand",
                                       starts.in.df.are.0based=FALSE)
overlap <- subsetByOverlaps(am_GR,corsiv_GR) %>%
  as.data.frame()

reg$corsiv[reg$regID %in% overlap$regID] =1

#summary(reg$reg_P)

enrich = list()
i=1

for(cutoff in c(0.1,0.01,0.001,0.0001,0.00001,0.000001)){
  #cutoff = 0.001
  a1 = sum((reg$reg_P<cutoff) & (reg$corsiv==1))
  a2 = sum((reg$reg_P<cutoff) & (reg$corsiv==0))
  b1 = sum(reg$corsiv==1)
  b2 = sum(reg$corsiv==0)
  
  if(a1>0){
    ft = fisher.test(matrix(c(c(a1,b1),c(a2,b2)), nrow=2, byrow=T))
    p = ft$p.value 
    est = ft$estimate
  }else{
    p =NA
    est=NA
    }
  
  enrich[[i]] = data.frame(cutoff= cutoff,
                           sig.corvis=a1,
                           sig.not.corvis=a2,
                           all.corsiv = b1,
                           all.not.corsiv = b2,
                           enrich.or = unname(est),
                           enrich.p = p)
  i = i +1
}

enrich.df = do.call(rbind,enrich) %>% as.data.frame()
write.csv(enrich.df,"ad.mapping.enrichment.csv")

```

# Case-only analysis required by reviewer

Reviewer 1: Admixture mapping was originally described as a case-only analysis looking for regions where cases were significantly enriched with a specific local ancestry, compared to other regions in the genome. The case-only analysis can be more powerful than case-control analysis if basic assumptions are met, specifically that there are no other selective forces that would enrich local ancestry in both the controls and cases. Here do the significant or close to significant regions show a rise in local European ancestry among the controls as well as cases?

https://doi.org/10.1146/annurev-genom-082509-141523
Association of local chromosomal ancestry with the disease 
Extent of ancestry at each locus could also be compared to genome-wide average ancestry for a case-only study design
A statistical deviation in local ancestry away from the genome-wide average could be used to identify a peak region of interest.

```{r case-case analysis}

setwd("/project/lishaobo_902/sebastian/admixtureMapping/reviewers")
library(tidyverse)
library(data.table)


for(chr in c("chr6","chr7","chr10","chr13")){
  
# chr="chr7"
# hitCenter = 30397103

  if(chr=="chr6"){hitCenter=85512844}
  if(chr=="chr7"){hitCenter=30397103}
  if(chr=="chr10"){hitCenter=29375488}
  if(chr=="chr13"){hitCenter=94445215}
  
    
    pheno = fread("/project/wiemels_494/sebastian/temp_scratch_backup/populationRisk2/rgn/lat.rfmix.pheono.txt") %>%
      dplyr::select(sub,caco)
    
    controls = pheno %>% filter(caco==0)
    cases = pheno %>% filter(caco==1)
    
    res = fread(paste0("/project/wiemels_494/sebastian/temp_scratch_backup/populationRisk2/rgn/",chr,"_eurCount.txt")) %>%
      separate(V1,c("chr","pos")) %>%
      mutate(pos=as.numeric(pos)) %>%
      filter(pos<hitCenter+2000000 & pos>hitCenter-2000000) %>% # chr 6
      #filter(pos<hitCenter+100000000 & pos>hitCenter-100000000) %>% # chr 7
      dplyr::select(-chr) %>%
      column_to_rownames(var="pos")
    
    res_cases = res[,cases$sub] 
    res_controls = res[,controls$sub]
      
    res_cases_plot = data.frame(
      pos = rownames(res_cases),
      eurCopies = rowSums(res_cases)/(2*ncol(res_cases))) %>%
      mutate(pos=as.numeric(pos),
             group="case") 
    
    
    res_controls_plot = data.frame(
      pos = rownames(res_controls),
      eurCopies = rowSums(res_controls)/(2*ncol(res_controls)))%>%
      mutate(pos=as.numeric(pos),
             group="control")
    
    res_plot = rbind(res_cases_plot,res_controls_plot) %>% as.data.frame()
      
    require(scales)
    
    res_plot %>%
      ggplot(aes(x=pos,y=eurCopies,group=group))+
      geom_line(aes(color=group),alpha=0.5)+
      theme_bw()+
      scale_x_continuous(labels = comma)+
      labs(x=paste0(chr," positions"),y="European ancestry proportions")
    ggsave(paste0(chr,".comp.4mb.regional.png"),width = 8,height = 4)
    
    
}

```

```{r calculating case-case z test and plotting them with case-control results}

setwd("/project/lishaobo_902/sebastian/admixtureMapping/reviewers")
library(tidyverse)
library(data.table)
library(BSDA)
library(qqman)
library(QCEWAS)

# compute case-case p values

pheno = fread("/project/wiemels_494/sebastian/temp_scratch_backup/populationRisk2/rgn/lat.rfmix.pheono.txt") %>%
  filter(caco==1) %>%
  dplyr::select(sub) 

propAveDfLs = list()

for(chr in c(1:22)){
  #chr=22
  res = fread(paste0("/project/wiemels_494/sebastian/temp_scratch_backup/populationRisk2/rgn/chr",chr,"_eurCount.txt")) %>%
    as.data.frame() %>%
    column_to_rownames(var="V1")
  res_case = res[,pheno$sub] 
  propAve = rowSums(res_case)/(2*ncol(res_case)) 
  propAveDf = data.frame(propAve=propAve)
  propAveDfLs[[chr]] = propAveDf
}

propAveDfLsAll = do.call(rbind,propAveDfLs) %>% 
  as.data.frame() 

mu = mean(propAveDfLsAll$propAve)
sigma = sd(propAve)

propAveDfLsAll$z_score = (propAveDfLsAll$propAve - mu)/sigma
propAveDfLsAll$z_pval = pnorm(propAveDfLsAll$z_score, lower.tail=FALSE)
propAveDfLsAll$z_p_log = -log10(propAveDfLsAll$z_pval)

propAveDfLsAll = propAveDfLsAll %>%
  rownames_to_column(var="SNPID") %>%
  separate(SNPID, c("CHROM","POS"))%>%
  mutate(CHROM = as.numeric(CHROM),
         POS = as.numeric(POS)) 

write.table(propAveDfLsAll,"case_case_test_whole.txt",
            quote=FALSE,
            sep='\t',
            row.names = FALSE)

# case case whole genome plot
propAveDfLsAll = fread("case_case_test_whole.txt")

propAveDfLsAll_05 = propAveDfLsAll %>%
  mutate(CHROM = as.numeric(CHROM),
         POS = as.numeric(POS),
         P = z_pval,
         ID = paste0(CHROM,"_",POS)) %>%
  arrange(CHROM,POS)
propAveDfLsAll_1 <- propAveDfLsAll_05[is.finite(propAveDfLsAll_05$P),  ] %>%
  filter(!is.na(P))

tiff("case.case.manhattan.tiff", width = 10, height = 7, units = 'in', res = 300)
manhattan(propAveDfLsAll_1, chr = "CHROM", bp = "POS", p = "P", snp = "ID",
          #genomewideline=-log10(7.75394645248347e-06),
          suggestiveline=FALSE)
dev.off()

lambda <- P_lambda(propAveDfLsAll_1$P)
# 1.009299

tiff("case.case.qq.tiff", width = 10, height = 7, units = 'in', res = 300)
qq(propAveDfLsAll_1$P)
dev.off()

# read case control and case case p values

chr="6"
hitCenter = 85512844

gwas_data <- fread("/project/wiemels_494/sebastian/temp_scratch_backup/populationRisk2/rgn/lat.pol.pcs.long",skip=1)
colnames(gwas_data) = c("ID", "Estimate","StdError","Zvalue","P")

gwas_data_05 = gwas_data %>%
  mutate(SNPID = ID) %>%
  separate(SNPID, c("CHROM","POS"))%>%
  mutate(CHROM = as.numeric(CHROM),
         POS = as.numeric(POS),
         case_control_p_log = -log10(P)) %>%
  arrange(CHROM,POS) %>%
  filter(CHROM==chr & POS>=(hitCenter-2000000) & POS <=(hitCenter+2000000)) %>%
  filter(!is.na(case_control_p_log))

#propAveDfLsAll = fread("case_case_test_whole.txt")

propAveDfLsAll_05 = propAveDfLsAll %>%
  filter(CHROM==chr & POS>=(hitCenter-2000000) & POS <=(hitCenter+2000000)) %>%
  dplyr::rename("case_case_log_P" = "z_p_log") %>%
  filter(!is.na(case_case_log_P)) %>%
  inner_join(gwas_data_05,by=c("CHROM","POS"))

require(scales)

propAveDfLsAll_05 %>%
  dplyr::select(CHROM,POS,case_case_log_P,case_control_p_log) %>%
  mutate(case_case_log_P=as.numeric(case_case_log_P),
         case_control_p_log = as.numeric(case_control_p_log))%>%
  gather(key="tests", value=log_P, c("case_case_log_P","case_control_p_log")) %>%
  ggplot(aes(x=POS,y=log_P,group=tests))+
  geom_line(aes(color=tests),alpha=0.8)+
  theme_bw()+
  scale_x_continuous(labels = comma)+
  labs(x=paste0(chr," positions"),y="log P values")
ggsave(paste0("case.case.case.control.chr",chr,".png"),width = 6,height = 4)


# check control-control analysis

pheno = fread("/project/wiemels_494/sebastian/temp_scratch_backup/populationRisk2/rgn/lat.rfmix.pheono.txt") %>%
  filter(caco==0) %>%
  dplyr::select(sub) 
propAveDfLs = list()

for(chr in c(1:22)){
  res = fread(paste0("/project/wiemels_494/sebastian/temp_scratch_backup/populationRisk2/rgn/chr",chr,"_eurCount.txt")) %>%
    as.data.frame() %>%
    column_to_rownames(var="V1")
  res_case = res[,pheno$sub] 
  propAve = rowSums(res_case)/(2*ncol(res_case)) 
  propAveDf = data.frame(propAve=propAve)
  propAveDfLs[[chr]] = propAveDf
}
propAveDfLsAll = do.call(rbind,propAveDfLs) %>% 
  as.data.frame() 
mu = mean(propAveDfLsAll$propAve)
sigma = sd(propAve)
propAveDfLsAll$z_score = (propAveDfLsAll$propAve - mu)/sigma
propAveDfLsAll$z_pval = pnorm(propAveDfLsAll$z_score, lower.tail=FALSE)
propAveDfLsAll$z_p_log = -log10(propAveDfLsAll$z_pval)
propAveDfLsAll = propAveDfLsAll %>%
  rownames_to_column(var="SNPID") %>%
  separate(SNPID, c("CHROM","POS"))%>%
  mutate(CHROM = as.numeric(CHROM),
         POS = as.numeric(POS)) 

chr="6"
hitCenter = 85512844

gwas_data <- fread("/project/wiemels_494/sebastian/temp_scratch_backup/populationRisk2/rgn/lat.pol.pcs.long",skip=1)
colnames(gwas_data) = c("ID", "Estimate","StdError","Zvalue","P")
gwas_data_05 = gwas_data %>%
  mutate(SNPID = ID) %>%
  separate(SNPID, c("CHROM","POS"))%>%
  mutate(CHROM = as.numeric(CHROM),
         POS = as.numeric(POS),
         case_control_p_log = -log10(P)) %>%
  arrange(CHROM,POS) %>%
  filter(CHROM==chr & POS>=(hitCenter-2000000) & POS <=(hitCenter+2000000)) %>%
  filter(!is.na(case_control_p_log))
propAveDfLsAll_05 = propAveDfLsAll %>%
  filter(CHROM==chr & POS>=(hitCenter-2000000) & POS <=(hitCenter+2000000)) %>%
  dplyr::rename("control_control_log_P" = "z_p_log") %>%
  filter(!is.na(control_control_log_P)) %>%
  inner_join(gwas_data_05,by=c("CHROM","POS"))
require(scales)
propAveDfLsAll_05 %>%
  dplyr::select(CHROM,POS,control_control_log_P,case_control_p_log) %>%
  mutate(control_control_log_P=as.numeric(control_control_log_P),
         case_control_p_log = as.numeric(case_control_p_log))%>%
  gather(key="tests", value=log_P, c("control_control_log_P","case_control_p_log")) %>%
  ggplot(aes(x=POS,y=log_P,group=tests))+
  geom_line(aes(color=tests),alpha=0.8)+
  theme_bw()+
  scale_x_continuous(labels = comma)+
  labs(x=paste0(chr," positions"),y="log P values")
ggsave(paste0("control.control.case.control.chr",chr,".png"),width = 6,height = 4)

```

# LAMP-ANC analysis suggested by Joe

```{bash}

cd /project/lishaobo_902/sebastian/admixtureMapping/reviewers/lamp

awk 'NR>1 {print $1,$2}' /project/wiemels_260/sebastian/gwas_glioma/populationRisk2/globalProportionsLat.txt > his.ids
awk 'NR>1 {print $1,$2}' /project/wiemels_494/sebastian/temp_scratch_backup/populationRisk2/EURpop.txt > eur.ids
awk 'NR>1 {print $1,$2}' /project/wiemels_494/sebastian/temp_scratch_backup/populationRisk2/AFRpop.txt > afr.ids
awk 'NR>1 {print $1,$2}' /project/wiemels_494/sebastian/temp_scratch_backup/populationRisk2/AMRpop.txt > amr.ids
cat his.ids eur.ids afr.ids amr.ids > all.subs.ids


# combine ref&query subjects into a plink file
plink --bfile /project/wiemels_494/sebastian/temp_scratch_backup/populationRisk2/glio.hgdp.1kg\
  --keep-allele-order\
  --keep all.subs.ids\
  --make-bed\
  --out all.subs.plink

chr=6

plink --bfile all.subs.plink\
  --keep-allele-order\
  --chr $chr\
  --keep his.ids\
  --make-bed\
  --out his.chr$chr
  
plink --bfile all.subs.plink\
  --keep-allele-order\
  --chr $chr\
  --keep eur.ids\
  --make-bed\
  --out eur.chr$chr

plink --bfile all.subs.plink\
  --keep-allele-order\
  --chr $chr\
  --keep afr.ids\
  --make-bed\
  --out afr.chr$chr

plink --bfile all.subs.plink\
  --keep-allele-order\
  --chr $chr\
  --keep amr.ids\
  --make-bed\
  --out amr.chr$chr

awk '{print $5}' eur.chr$chr.bim  >  file1
awk '{print $5}' amr.chr$chr.bim  > file2
awk '{print $5}' afr.chr$chr.bim  > file3
diff -sq file2 file1

# convert plinks into 0,1,2

plink2 --bfile his.chr$chr\
  --export A \
  --out chr$chr.his.add
awk 'NR>1' chr$chr.his.add.raw | cut --output-delimiter=' ' -f7-  | sed 's/NA/-1/g' >  chr$chr.his.add.geno.f

#awk '
#{ 
#    for (i=1; i<=NF; i++)  {
#        a[NR,i] = $i
#    }
#}
#NF>p { p = NF }
#END {    
#    for(j=1; j<=p; j++) {
#        str=a[1,j]
#        for(i=2; i<=NR; i++){
#            str=str" "a[i,j];
#        }
#        print str
#    }
#}' chr$chr.his.add.geno.trans > chr$chr.his.add.geno


# create a position file
awk '{print $4}' his.chr$chr.bim > chr$chr.pos

# create an ancestral allele frequency file for each ancestry

plink --bfile eur.chr$chr\
  --freq\
  --out eur.chr$chr
awk 'NR>1{print $5}' eur.chr$chr.frq | sed 's/NA/0/g' > eur.chr$chr.aaf

plink --bfile amr.chr$chr\
  --freq\
  --out amr.chr$chr
awk 'NR>1{print $5}' amr.chr$chr.frq | sed 's/NA/0/g' > amr.chr$chr.aaf

plink --bfile afr.chr$chr\
  --freq\
  --out afr.chr$chr
awk 'NR>1{print $5}' afr.chr$chr.frq | sed 's/NA/0/g' > afr.chr$chr.aaf

# create a configure file
## with the help of https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6218714/
nano config.chr$chr.txt

populations=3 
generations=20
alpha=0.5,0.4,0.1
recombrate=1e8
genofile=chr6.his.add.geno.f
posfile=chr6.pos
pfile=eur.chr6.aaf,amr.chr6.aaf,afr.chr6.aaf

lamp config.chr$chr.txt
awk 'NR % 3 == 1' ancestry.txt > chr$chr.eur.ancestry.txt

```

```{r plot case-case results from LAMP-ANC}

setwd("/project/lishaobo_902/sebastian/admixtureMapping/reviewers")
library(tidyverse)
library(data.table)


#for(chr in c("chr6","chr7","chr10","chr13")){
  
 chr="chr6"
 hitCenter = 85512844

#  if(chr=="chr6"){hitCenter=85512844}
#  if(chr=="chr7"){hitCenter=30397103}
#  if(chr=="chr10"){hitCenter=29375488}
#  if(chr=="chr13"){hitCenter=94445215}
  
    pheno = fread("/project/wiemels_494/sebastian/temp_scratch_backup/populationRisk2/rgn/lat.rfmix.pheono.txt") %>%
      dplyr::select(sub,caco)
    
    controls = pheno %>% filter(caco==0)
    cases = pheno %>% filter(caco==1)
    
    posF = fread(paste0("lamp/",chr,".pos"))
    famF = fread(paste0("lamp/his.",chr,".fam"))
    res = fread(paste0("lamp/",chr,".eur.ancestry.txt")) %>%
      dplyr::select(-V1) %>%
      t %>%
      as.data.frame()
    res = res[1:nrow(posF),]
    rownames(res) = posF$V1
    colnames(res) = famF$V2
    
    res = res %>% 
      rownames_to_column(var="pos") %>%
      mutate(pos=as.numeric(pos)) %>%
      filter(pos<hitCenter+2000000 & pos>hitCenter-2000000) %>% # chr 6
      #filter(pos<hitCenter+100000000 & pos>hitCenter-100000000) %>% # chr 7
      column_to_rownames(var="pos")
    
    res_cases = res[,cases$sub] 
    res_controls = res[,controls$sub]
      
    res_cases_plot = data.frame(
      pos = rownames(res_cases),
      eurCopies = rowSums(res_cases)/(2*ncol(res_cases))) %>%
      mutate(pos=as.numeric(pos),
             group="case") 
    
    res_controls_plot = data.frame(
      pos = rownames(res_controls),
      eurCopies = rowSums(res_controls)/(2*ncol(res_controls)))%>%
      mutate(pos=as.numeric(pos),
             group="control")
    
    res_plot = rbind(res_cases_plot,res_controls_plot) %>% as.data.frame()
      
    require(scales)
    
    res_plot %>%
      ggplot(aes(x=pos,y=eurCopies,group=group))+
      geom_line(aes(color=group),alpha=0.5)+
      theme_bw()+
      scale_x_continuous(labels = comma)+
      labs(x=paste0(chr," positions"),y="European ancestry proportions")
    ggsave(paste0("lampanc.",chr,".comp.4mb.regional.png"),width = 8,height = 4)
#}

```


#Conditional analysis of regional admixture results

1. Take the region around the significant SNPs (SNP_sig) from the regional ancestry manhattan plot
Impute this region on TopMed, filter for R^2>0.6 and run the following model for each SNP (glm in plink), use permuation to decide P
caco ~ SNP + 10 PC
And draw Manhattan plot style and see how it looks

```{bash take out the region already imputed with BEAGLE}

cd /scratch/lishaobo/GWAS/glioma/populationRisk2

########################################################################
# Imputed on beagle
## 1Mb window, 2Mb window and 5Mb window

cd /scratch/lishaobo/GWAS/glioma/populationRisk2/conditional

# compute genetic PCs for these subjects
plink2 --bfile /scratch/lishaobo/GWAS/glioma/populationRisk2/pol.adm.subs\
  --maf 0.1\
  --indep-pairwise 50 10 0.1\
  --out reg.lat
plink2 --bfile /scratch/lishaobo/GWAS/glioma/populationRisk2/pol.adm.subs\
  --extract reg.lat.prune.in\
  --pca 10\
  --out reg.lat\
  --threads 200

##### chr 6 peak, 1mb
bcftools view ../rgn/6.query.phased.vcf.gz\
  -o 6.reg.raw.vcf\
  -Ov\
  -r 6:85012844-86012844\
  --threads 200

plink2 --vcf 6.reg.raw.vcf\
  dosage=DS\
  --maf 0.01\
  --make-bed\
  --out chr6.dose.beagle\
  --threads 200
  
##### chr 6 peak, 3mb (1+1 peak +1)
bcftools view ../rgn/6.query.phased.vcf.gz\
  -o 6.reg.raw.conditional.vcf\
  -Ov\
  -r 6:84009612-87009612\
  --threads 200

plink2 --vcf 6.reg.raw.conditional.vcf\
  dosage=DS\
  --maf 0.01\
  --make-bed\
  --out chr6.dose.beagle.conditional\
  --threads 200
    
##### chr 6 peak, 5mb
bcftools view ../rgn/6.query.phased.vcf.gz\
  -o 6.reg.raw.5mb.vcf\
  -Ov\
  -r 6:83012844-88012844\
  --threads 200

plink2 --vcf 6.reg.raw.5mb.vcf\
  dosage=DS\
  --maf 0.01\
  --make-bed\
  --out chr6.dose.beagle.5mb\
  --threads 200

########## chr6 long version, for locuszoom later (around the hit)

bcftools view ../rgn/6.query.phased.vcf.gz\
  -o 6.reg.raw.long.vcf\
  -Ov\
  -r 6:84566601-86012844\
  --threads 200

plink2 --vcf 6.reg.raw.long.vcf\
  dosage=DS\
  --maf 0.01\
  --make-bed\
  --out chr6.dose.long.beagle\
  --threads 200

##### chr 7 peak, 1mb
bcftools view ../rgn/7.query.phased.vcf.gz\
  -o 7.reg.raw.vcf\
  -Ov\
  -r 7:29897103-30897103\
  --threads 200

plink2 --vcf 7.reg.raw.vcf\
  dosage=DS\
  --maf 0.01\
  --make-bed\
  --out chr7.dose.beagle\
  --threads 200

##### chr 7 peak, 2mb
bcftools view ../rgn/7.query.phased.vcf.gz\
  -o 7.reg.raw.conditional.vcf\
  -Ov\
  -r 7:29397103-31397103\
  --threads 200

plink2 --vcf 7.reg.raw.conditional.vcf\
  dosage=DS\
  --maf 0.01\
  --make-bed\
  --out chr7.dose.beagle.conditional\
  --threads 200

##### chr 7 peak, 5mb
bcftools view ../rgn/7.query.phased.vcf.gz\
  -o 7.reg.raw.5mb.vcf\
  -Ov\
  -r 7:27897103-32897103\
  --threads 200

plink2 --vcf 7.reg.raw.5mb.vcf\
  dosage=DS\
  --maf 0.01\
  --make-bed\
  --out chr7.dose.beagle.5mb\
  --threads 200

##### chr 10 peak, 1mb
bcftools view ../rgn/10.query.phased.vcf.gz\
  -o 10.reg.raw.vcf\
  -Ov\
  -r 10:28875488-29875488\
  --threads 200

plink2 --vcf 10.reg.raw.vcf\
  dosage=DS\
  --maf 0.01\
  --make-bed\
  --out chr10.dose.beagle\
  --threads 200

##### chr 10 peak, 2mb
bcftools view ../rgn/10.query.phased.vcf.gz\
  -o 10.reg.raw.conditional.vcf\
  -Ov\
  -r 10:28375488-30375488\
  --threads 200

plink2 --vcf 10.reg.raw.conditional.vcf\
  dosage=DS\
  --maf 0.01\
  --make-bed\
  --out chr10.dose.beagle.conditional\
  --threads 200

##### chr 10 peak, 5mb
bcftools view ../rgn/10.query.phased.vcf.gz\
  -o 10.reg.raw.5mb.vcf\
  -Ov\
  -r 10:26875488-31875488\
  --threads 200

plink2 --vcf 10.reg.raw.5mb.vcf\
  dosage=DS\
  --maf 0.01\
  --make-bed\
  --out chr10.dose.beagle.5mb\
  --threads 200

##### chr 13 peak, 1mb
bcftools view ../rgn/13.query.phased.vcf.gz\
  -o 13.reg.raw.vcf\
  -Ov\
  -r 13:93945215-94945215\
  --threads 200

plink2 --vcf 13.reg.raw.vcf\
  dosage=DS\
  --maf 0.01\
  --make-bed\
  --out chr13.dose.beagle\
  --threads 200

##### chr 13 peak, 2mb
bcftools view ../rgn/13.query.phased.vcf.gz\
  -o 13.reg.raw.conditional.vcf\
  -Ov\
  -r 13:93445215-95445215\
  --threads 200

plink2 --vcf 13.reg.raw.conditional.vcf\
  dosage=DS\
  --maf 0.01\
  --make-bed\
  --out chr13.dose.beagle.conditional\
  --threads 200
  
##### chr 13 peak, 5mb
bcftools view ../rgn/13.query.phased.vcf.gz\
  -o 13.reg.raw.5mb.vcf\
  -Ov\
  -r 13:91945215-96945215\
  --threads 200

plink2 --vcf 13.reg.raw.5mb.vcf\
  dosage=DS\
  --maf 0.01\
  --make-bed\
  --out chr13.dose.beagle.5mb\
  --threads 200
  

########################################################################
# Imputed on beagle, european validation (ccrlp)
cd /scratch/lishaobo/GWAS/glioma/populationRisk2/conditional

# compute genetic PCs for these subjects

plink2 --bfile /scratch/lishaobo/GWAS/glioma/populationRisk2/pol.adm.subs.wt\
  --maf 0.1\
  --indep-pairwise 50 10 0.1\
  --out reg.wt
plink2 --bfile /scratch/lishaobo/GWAS/glioma/populationRisk2/pol.adm.subs.wt\
  --extract reg.wt.prune.in\
  --pca 10\
  --out reg.wt\
  --threads 200

##### chr 6 peak, 1mb
bcftools view ../rgn/6.query.phased.wt.vcf.gz\
  -o 6.reg.raw.wt.vcf\
  -Ov\
  -r 6:85012844-86012844\
  --threads 200

plink2 --vcf 6.reg.raw.wt.vcf\
  dosage=DS\
  --maf 0.01\
  --make-bed\
  --out chr6.dose.wt.beagle\
  --threads 200

##### chr 6 peak, 3mb
bcftools view ../rgn/6.query.phased.wt.vcf.gz\
  -o 6.reg.raw.wt.conditional.vcf\
  -Ov\
  -r 6:84009612-87009612\
  --threads 200

plink2 --vcf 6.reg.raw.wt.conditional.vcf\
  dosage=DS\
  --maf 0.01\
  --make-bed\
  --out chr6.dose.wt.beagle.conditional\
  --threads 200  

##### chr 6 peak, 5mb
bcftools view ../rgn/6.query.phased.wt.vcf.gz\
  -o 6.reg.raw.wt.5mb.vcf\
  -Ov\
  -r 6:83012844-88012844\
  --threads 200

plink2 --vcf 6.reg.raw.wt.5mb.vcf\
  dosage=DS\
  --maf 0.01\
  --make-bed\
  --out chr6.dose.wt.beagle.5mb\
  --threads 200

##### chr 7 peak, 1mb
bcftools view ../rgn/7.query.phased.wt.vcf.gz\
  -o 7.reg.raw.wt.vcf\
  -Ov\
  -r 7:29897103-30897103\
  --threads 200

plink2 --vcf 7.reg.raw.wt.vcf\
  dosage=DS\
  --maf 0.01\
  --make-bed\
  --out chr7.dose.wt.beagle\
  --threads 200

##### chr 7 peak, 2mb
bcftools view ../rgn/7.query.phased.wt.vcf.gz\
  -o 7.reg.raw.wt.conditional.vcf\
  -Ov\
  -r 7:29397103-31397103\
  --threads 200

plink2 --vcf 7.reg.raw.wt.conditional.vcf\
  dosage=DS\
  --maf 0.01\
  --make-bed\
  --out chr7.dose.wt.beagle.conditional\
  --threads 200

##### chr 7 peak, 5mb
bcftools view ../rgn/7.query.phased.wt.vcf.gz\
  -o 7.reg.raw.wt.5mb.vcf\
  -Ov\
  -r 7:27897103-32897103\
  --threads 200

plink2 --vcf 7.reg.raw.wt.5mb.vcf\
  dosage=DS\
  --maf 0.01\
  --make-bed\
  --out chr7.dose.wt.beagle.5mb\
  --threads 200

##### chr 10 peak, 1mb
bcftools view ../rgn/10.query.phased.wt.vcf.gz\
  -o 10.reg.raw.wt.vcf\
  -Ov\
  -r 10:28875488-29875488\
  --threads 200

plink2 --vcf 10.reg.raw.wt.vcf\
  dosage=DS\
  --maf 0.01\
  --make-bed\
  --out chr10.dose.wt.beagle\
  --threads 200

##### chr 10 peak, 2mb
bcftools view ../rgn/10.query.phased.wt.vcf.gz\
  -o 10.reg.raw.wt.conditional.vcf\
  -Ov\
  -r 10:28375488-30375488\
  --threads 200

plink2 --vcf 10.reg.raw.wt.conditional.vcf\
  dosage=DS\
  --maf 0.01\
  --make-bed\
  --out chr10.dose.wt.beagle.conditional\
  --threads 200
  
##### chr 10 peak, 5mb
bcftools view ../rgn/10.query.phased.wt.vcf.gz\
  -o 10.reg.raw.wt.5mb.vcf\
  -Ov\
  -r 10:26875488-31875488\
  --threads 200

plink2 --vcf 10.reg.raw.wt.5mb.vcf\
  dosage=DS\
  --maf 0.01\
  --make-bed\
  --out chr10.dose.wt.beagle.5mb\
  --threads 200

##### chr 13 peak, 1mb
bcftools view ../rgn/13.query.phased.wt.vcf.gz\
  -o 13.reg.raw.wt.vcf\
  -Ov\
  -r 13:93945215-94945215\
  --threads 200

plink2 --vcf 13.reg.raw.wt.vcf\
  dosage=DS\
  --maf 0.01\
  --make-bed\
  --out chr13.dose.wt.beagle\
  --threads 200

##### chr 13 peak, 2mb
bcftools view ../rgn/13.query.phased.wt.vcf.gz\
  -o 13.reg.raw.wt.conditional.vcf\
  -Ov\
  -r 13:93445215-95445215\
  --threads 200

plink2 --vcf 13.reg.raw.wt.conditional.vcf\
  dosage=DS\
  --maf 0.01\
  --make-bed\
  --out chr13.dose.wt.beagle.conditional\
  --threads 200

##### chr 13 peak, 5mb
bcftools view ../rgn/13.query.phased.wt.vcf.gz\
  -o 13.reg.raw.wt.5mb.vcf\
  -Ov\
  -r 13:91945215-96945215\
  --threads 200

plink2 --vcf 13.reg.raw.wt.5mb.vcf\
  dosage=DS\
  --maf 0.01\
  --make-bed\
  --out chr13.dose.wt.beagle.5mb\
  --threads 200

```

```{r update fam files for imputed files}

setwd("/scratch/lishaobo/GWAS/glioma/populationRisk2/conditional")
library(tidyverse)
library(data.table)

fam.raw=fread("/scratch/lishaobo/GWAS/glioma/populationRisk2/pol.adm.subs.fam")

fam.2 = fread("chr6.dose.beagle.fam")
identical(fam.raw$V2,fam.2$V2)
fam.2$V5=fam.raw$V5
fam.2$V6=fam.raw$V6

write.table(fam.2,"chr6.dose.beagle.fam",sep = "\t", col.names=FALSE, row.names = FALSE, quote = FALSE)

fam.2 = fread("chr6.dose.beagle.conditional.fam")
identical(fam.raw$V2,fam.2$V2)
fam.2$V5=fam.raw$V5
fam.2$V6=fam.raw$V6

write.table(fam.2,"chr6.dose.beagle.conditional.fam",sep = "\t", col.names=FALSE, row.names = FALSE, quote = FALSE)

fam.2 = fread("chr6.dose.beagle.5mb.fam")
identical(fam.raw$V2,fam.2$V2)
fam.2$V5=fam.raw$V5
fam.2$V6=fam.raw$V6

write.table(fam.2,"chr6.dose.beagle.5mb.fam",sep = "\t", col.names=FALSE, row.names = FALSE, quote = FALSE)

fam.2 = fread("chr7.dose.beagle.fam")
identical(fam.raw$V2,fam.2$V2)
fam.2$V5=fam.raw$V5
fam.2$V6=fam.raw$V6

write.table(fam.2,"chr7.dose.beagle.fam",sep = "\t", col.names=FALSE, row.names = FALSE, quote = FALSE)

fam.2 = fread("chr7.dose.beagle.conditional.fam")
identical(fam.raw$V2,fam.2$V2)
fam.2$V5=fam.raw$V5
fam.2$V6=fam.raw$V6

write.table(fam.2,"chr7.dose.beagle.conditional.fam",sep = "\t", col.names=FALSE, row.names = FALSE, quote = FALSE)

fam.2 = fread("chr7.dose.beagle.5mb.fam")
identical(fam.raw$V2,fam.2$V2)
fam.2$V5=fam.raw$V5
fam.2$V6=fam.raw$V6

write.table(fam.2,"chr7.dose.beagle.5mb.fam",sep = "\t", col.names=FALSE, row.names = FALSE, quote = FALSE)

fam.2 = fread("chr10.dose.beagle.fam")
identical(fam.raw$V2,fam.2$V2)
fam.2$V5=fam.raw$V5
fam.2$V6=fam.raw$V6

write.table(fam.2,"chr10.dose.beagle.fam",sep = "\t", col.names=FALSE, row.names = FALSE, quote = FALSE)

fam.2 = fread("chr10.dose.beagle.conditional.fam")
identical(fam.raw$V2,fam.2$V2)
fam.2$V5=fam.raw$V5
fam.2$V6=fam.raw$V6

write.table(fam.2,"chr10.dose.beagle.conditional.fam",sep = "\t", col.names=FALSE, row.names = FALSE, quote = FALSE)

fam.2 = fread("chr10.dose.beagle.5mb.fam")
identical(fam.raw$V2,fam.2$V2)
fam.2$V5=fam.raw$V5
fam.2$V6=fam.raw$V6

write.table(fam.2,"chr10.dose.beagle.5mb.fam",sep = "\t", col.names=FALSE, row.names = FALSE, quote = FALSE)

fam.2 = fread("chr13.dose.beagle.fam")
identical(fam.raw$V2,fam.2$V2)
fam.2$V5=fam.raw$V5
fam.2$V6=fam.raw$V6

write.table(fam.2,"chr13.dose.beagle.fam",sep = "\t", col.names=FALSE, row.names = FALSE, quote = FALSE)

fam.2 = fread("chr13.dose.beagle.conditional.fam")
identical(fam.raw$V2,fam.2$V2)
fam.2$V5=fam.raw$V5
fam.2$V6=fam.raw$V6

write.table(fam.2,"chr13.dose.beagle.conditional.fam",sep = "\t", col.names=FALSE, row.names = FALSE, quote = FALSE)

fam.2 = fread("chr13.dose.beagle.5mb.fam")
identical(fam.raw$V2,fam.2$V2)
fam.2$V5=fam.raw$V5
fam.2$V6=fam.raw$V6

write.table(fam.2,"chr13.dose.beagle.5mb.fam",sep = "\t", col.names=FALSE, row.names = FALSE, quote = FALSE)

####################################################################################
# European validation file (ccrlp)

setwd("/scratch/lishaobo/GWAS/glioma/populationRisk2/conditional")
library(tidyverse)
library(data.table)

fam.raw=fread("/scratch/lishaobo/GWAS/glioma/populationRisk2/pol.adm.subs.wt.nodup.fam")

## need to run this after phasing is done
fam.3 = fread("chr6.dose.wt.beagle.fam")
identical(fam.raw$V2,fam.3$V2)
fam.3$V5=fam.raw$V5
fam.3$V6=fam.raw$V6

write.table(fam.3,"chr6.dose.wt.beagle.fam",sep = "\t", col.names=FALSE, row.names = FALSE, quote = FALSE)

fam.3 = fread("chr6.dose.wt.beagle.conditional.fam")
identical(fam.raw$V2,fam.3$V2)
fam.3$V5=fam.raw$V5
fam.3$V6=fam.raw$V6

write.table(fam.3,"chr6.dose.wt.beagle.conditional.fam",sep = "\t", col.names=FALSE, row.names = FALSE, quote = FALSE)

fam.3 = fread("chr6.dose.wt.beagle.5mb.fam")
identical(fam.raw$V2,fam.3$V2)
fam.3$V5=fam.raw$V5
fam.3$V6=fam.raw$V6

write.table(fam.3,"chr6.dose.wt.beagle.5mb.fam",sep = "\t", col.names=FALSE, row.names = FALSE, quote = FALSE)

fam.3 = fread("chr7.dose.wt.beagle.fam")
identical(fam.raw$V2,fam.3$V2)
fam.3$V5=fam.raw$V5
fam.3$V6=fam.raw$V6

write.table(fam.3,"chr7.dose.wt.beagle.fam",sep = "\t", col.names=FALSE, row.names = FALSE, quote = FALSE)

fam.3 = fread("chr7.dose.wt.beagle.conditional.fam")
identical(fam.raw$V2,fam.3$V2)
fam.3$V5=fam.raw$V5
fam.3$V6=fam.raw$V6

write.table(fam.3,"chr7.dose.wt.beagle.conditional.fam",sep = "\t", col.names=FALSE, row.names = FALSE, quote = FALSE)

fam.3 = fread("chr7.dose.wt.beagle.5mb.fam")
identical(fam.raw$V2,fam.3$V2)
fam.3$V5=fam.raw$V5
fam.3$V6=fam.raw$V6

write.table(fam.3,"chr7.dose.wt.beagle.5mb.fam",sep = "\t", col.names=FALSE, row.names = FALSE, quote = FALSE)

fam.3 = fread("chr10.dose.wt.beagle.fam")
identical(fam.raw$V2,fam.3$V2)
fam.3$V5=fam.raw$V5
fam.3$V6=fam.raw$V6

write.table(fam.3,"chr10.dose.wt.beagle.fam",sep = "\t", col.names=FALSE, row.names = FALSE, quote = FALSE)

fam.3 = fread("chr10.dose.wt.beagle.conditional.fam")
identical(fam.raw$V2,fam.3$V2)
fam.3$V5=fam.raw$V5
fam.3$V6=fam.raw$V6

write.table(fam.3,"chr10.dose.wt.beagle.conditional.fam",sep = "\t", col.names=FALSE, row.names = FALSE, quote = FALSE)

fam.3 = fread("chr10.dose.wt.beagle.5mb.fam")
identical(fam.raw$V2,fam.3$V2)
fam.3$V5=fam.raw$V5
fam.3$V6=fam.raw$V6

write.table(fam.3,"chr10.dose.wt.beagle.5mb.fam",sep = "\t", col.names=FALSE, row.names = FALSE, quote = FALSE)

fam.3 = fread("chr13.dose.wt.beagle.fam")
identical(fam.raw$V2,fam.3$V2)
fam.3$V5=fam.raw$V5
fam.3$V6=fam.raw$V6

write.table(fam.3,"chr13.dose.wt.beagle.fam",sep = "\t", col.names=FALSE, row.names = FALSE, quote = FALSE)

fam.3 = fread("chr13.dose.wt.beagle.conditional.fam")
identical(fam.raw$V2,fam.3$V2)
fam.3$V5=fam.raw$V5
fam.3$V6=fam.raw$V6

write.table(fam.3,"chr13.dose.wt.beagle.conditional.fam",sep = "\t", col.names=FALSE, row.names = FALSE, quote = FALSE)

fam.3 = fread("chr13.dose.wt.beagle.5mb.fam")
identical(fam.raw$V2,fam.3$V2)
fam.3$V5=fam.raw$V5
fam.3$V6=fam.raw$V6

write.table(fam.3,"chr13.dose.wt.beagle.5mb.fam",sep = "\t", col.names=FALSE, row.names = FALSE, quote = FALSE)

```

```{bash gwas style regression for this region}

#cd /scratch/lishaobo/GWAS/glioma/populationRisk2/conditional

cd /project/wiemels_494/sebastian/temp_scratch_backup/populationRisk2/conditional
# imputed by beagle
## in Latinos

# 1mb region
for chr in 6 7 10 13; do

  awk -v OFS="\t" '{print $1,$1":"$4":"$6":"$5,$3,$4,$5,$6}' chr$chr.dose.beagle.bim > chr$chr.dose.beagle.bim.neo
  mv chr$chr.dose.beagle.bim.neo chr$chr.dose.beagle.bim
  
  plink2 --bfile chr$chr.dose.beagle\
    --covar reg.lat.eigenvec\
    --glm hide-covar firth-fallback cols=chrom,pos,ref,alt1,nobs,a1countcc,a1freq,a1freqcc,beta,se,p\
    --adjust\
    --out chr$chr.beagle\
    --threads 200

done

#chr6.beagle.PHENO1.glm.logistic.hybrid.adjusted
#chr7.beagle.PHENO1.glm.logistic.hybrid.adjusted
#chr10.beagle.PHENO1.glm.logistic.hybrid.adjusted
#chr13.beagle.PHENO1.glm.logistic.hybrid.adjusted

## used in the final paper
# 2mb region (conditional, chr6 is 3mb)
for chr in 6 7 10 13; do

  awk -v OFS="\t" '{print $1,$1":"$4":"$6":"$5,$3,$4,$5,$6}' chr$chr.dose.beagle.conditional.bim > chr$chr.dose.beagle.conditional.bim.neo
  mv chr$chr.dose.beagle.conditional.bim.neo chr$chr.dose.beagle.conditional.bim
  
  plink2 --bfile chr$chr.dose.beagle.conditional\
    --covar reg.lat.eigenvec\
    --glm hide-covar firth-fallback cols=chrom,pos,ref,alt1,nobs,a1countcc,a1freq,a1freqcc,beta,se,p\
    --out chr$chr.beagle.conditional\
    --threads 200

  plink2 --bfile chr$chr.dose.beagle.conditional\
    --covar reg.lat.eigenvec\
    --adjust\
    --glm hide-covar firth-fallback cols=chrom,pos,ref,alt1,nobs,a1countcc,a1freq,a1freqcc,beta,se,p\
    --out chr$chr.beagle.conditional\
    --threads 200

done

#chr6.beagle.conditional.PHENO1.glm.logistic.hybrid.adjusted
#chr7.beagle.conditional.PHENO1.glm.logistic.hybrid.adjusted
#chr10.beagle.conditional.PHENO1.glm.logistic.hybrid.adjusted
#chr13.beagle.conditional.PHENO1.glm.logistic.hybrid.adjusted

# list of SNPs identified

6:84058087:G:A
6:86961711:G:A
6:85439850:G:C
grep '6:85439850:G:C' chr6.beagle.conditional.PHENO1.glm.logistic.hybrid |\
  awk '{print $3,$6, $9}'
grep '6:85439850:G:C' chr6.wt.beagle.conditional.PHENO1.glm.logistic.hybrid |\
  awk '{print $3,$6, $9}'


7:31316499:A:G
7:31324183:C:A
7:30138553:A:G
grep '7:30138553:A:G' chr7.beagle.conditional.PHENO1.glm.logistic.hybrid |\
  awk '{print $3,$6, $9}'
grep '7:30138553:A:G' chr7.wt.beagle.conditional.PHENO1.glm.logistic.hybrid |\
  awk '{print $3,$6, $9}'

10:29099450:C:T
10:29090828:C:T
10:29094615:G:A
grep '10:29094615:G:A' chr10.beagle.conditional.PHENO1.glm.logistic.hybrid |\
  awk '{print $3,$6, $9}'
grep '10:29094615:G:A' chr10.wt.beagle.conditional.PHENO1.glm.logistic.hybrid |\
  awk '{print $3,$6, $9}'

13:93952876:A:G	
13:95250755:TTC:T
13:93930275:G:GT
grep '13:93930275:G:GT' chr13.beagle.conditional.PHENO1.glm.logistic.hybrid |\
  awk '{print $3,$6, $9}'
grep '13:93930275:G:GT' chr13.wt.beagle.conditional.PHENO1.glm.logistic.hybrid |\
  awk '{print $3,$6, $9}'


# 5mb region
for chr in 6 7 10 13; do

  awk -v OFS="\t" '{print $1,$1":"$4":"$6":"$5,$3,$4,$5,$6}' chr$chr.dose.beagle.5mb.bim > chr$chr.dose.beagle.5mb.bim.neo
  mv chr$chr.dose.beagle.5mb.bim.neo chr$chr.dose.beagle.5mb.bim
  
  plink2 --bfile chr$chr.dose.beagle.5mb\
    --covar reg.lat.eigenvec\
    --glm hide-covar firth-fallback cols=chrom,pos,ref,alt1,nobs,a1countcc,a1freq,a1freqcc,beta,se,p\
    --adjust\
    --out chr$chr.beagle.5mb\
    --threads 200
done

#chr6.beagle.5mb.PHENO1.glm.logistic.hybrid.adjusted
#chr7.beagle.5mb.PHENO1.glm.logistic.hybrid.adjusted
#chr10.beagle.5mb.PHENO1.glm.logistic.hybrid.adjusted
#chr13.beagle.5mb.PHENO1.glm.logistic.hybrid.adjusted

#### chr6 long regression (for zoomlocus)

  awk -v OFS="\t" '{print $1,$1":"$4":"$6":"$5,$3,$4,$5,$6}' chr6.dose.long.beagle.bim > chr6.dose.long.beagle.bim.neo
  mv chr6.dose.long.beagle.bim.neo chr6.dose.long.beagle.bim
  
  plink2 --bfile chr6.dose.long.beagle\
    --covar reg.lat.eigenvec\
    --glm hide-covar firth-fallback cols=chrom,pos,ref,alt1,nobs,a1countcc,a1freq,a1freqcc,beta,se,p\
    --out chr6.long.beagle\
    --threads 200

# imputed by topmed 
#plink2 --bfile chr6.dose.topmed\
#  --covar reg.lat.eigenvec\
#  --glm hide-covar cols=chrom,pos,ref,alt1,nobs,a1countcc,a1freq,a1freqcc,beta,se,p\
#  --out chr6.topmed\
#  --threads 200

####################################################################################
# European validation (CCRLP)
# imputed by beagle

# 1mb region 

for chr in 6 7 10 13; do

  awk -v OFS="\t" '{print $1,$1":"$4":"$6":"$5,$3,$4,$5,$6}' chr$chr.dose.wt.beagle.bim > chr$chr.dose.wt.beagle.bim.neo
  mv chr$chr.dose.wt.beagle.bim.neo chr$chr.dose.wt.beagle.bim
  
  plink2 --bfile chr$chr.dose.wt.beagle\
    --covar reg.wt.eigenvec\
    --glm hide-covar firth-fallback cols=chrom,pos,ref,alt1,nobs,a1countcc,a1freq,a1freqcc,beta,se,p\
    --out chr$chr.wt.beagle\
    --adjust\
    --threads 200
done  

#chr6.wt.beagle.PHENO1.glm.logistic.hybrid.adjusted
#chr7.wt.beagle.PHENO1.glm.logistic.hybrid.adjusted
#chr10.wt.beagle.PHENO1.glm.logistic.hybrid.adjusted
#chr13.wt.beagle.PHENO1.glm.logistic.hybrid.adjusted


# 2mb region (conditional, chr6 is 3mb)
for chr in 6 7 10 13; do

  awk -v OFS="\t" '{print $1,$1":"$4":"$6":"$5,$3,$4,$5,$6}' chr$chr.dose.wt.beagle.conditional.bim > chr$chr.dose.wt.beagle.conditional.bim.neo
  mv chr$chr.dose.wt.beagle.conditional.bim.neo chr$chr.dose.wt.beagle.conditional.bim
  
  plink2 --bfile chr$chr.dose.wt.beagle.conditional\
    --covar reg.wt.eigenvec\
    --glm hide-covar firth-fallback cols=chrom,pos,ref,alt1,nobs,a1countcc,a1freq,a1freqcc,beta,se,p\
    --out chr$chr.wt.beagle.conditional\
    --threads 200

  plink2 --bfile chr$chr.dose.wt.beagle.conditional\
    --covar reg.wt.eigenvec\
    --adjust\
    --glm hide-covar firth-fallback cols=chrom,pos,ref,alt1,nobs,a1countcc,a1freq,a1freqcc,beta,se,p\
    --out chr$chr.wt.beagle.conditional\
    --threads 200

done

#chr6.wt.beagle.conditional.PHENO1.glm.logistic.hybrid.adjusted
#chr7.wt.beagle.conditional.PHENO1.glm.logistic.hybrid.adjusted
#chr10.wt.beagle.conditional.PHENO1.glm.logistic.hybrid.adjusted
#chr13.wt.beagle.conditional.PHENO1.glm.logistic.hybrid.adjusted

# 5mb region 

for chr in 6 7 10 13; do

  awk -v OFS="\t" '{print $1,$1":"$4":"$6":"$5,$3,$4,$5,$6}' chr$chr.dose.wt.beagle.5mb.bim > chr$chr.dose.wt.beagle.5mb.bim.neo
  mv chr$chr.dose.wt.beagle.5mb.bim.neo chr$chr.dose.wt.beagle.5mb.bim
  
  plink2 --bfile chr$chr.dose.wt.beagle.5mb\
    --covar reg.wt.eigenvec\
    --glm hide-covar firth-fallback cols=chrom,pos,ref,alt1,nobs,a1countcc,a1freq,a1freqcc,beta,se,p\
    --out chr$chr.wt.beagle.5mb\
    --adjust\
    --threads 200

done  

#chr6.wt.beagle.5mb.PHENO1.glm.logistic.hybrid.adjusted
#chr7.wt.beagle.5mb.PHENO1.glm.logistic.hybrid.adjusted
#chr10.wt.beagle.5mb.PHENO1.glm.logistic.hybrid.adjusted
#chr13.wt.beagle.5mb.PHENO1.glm.logistic.hybrid.adjusted


####################################################################################
## determine number of effective tests
# Bonferroni correction for effective tests following adjustment for LD

cd /scratch/lishaobo/GWAS/glioma/populationRisk2/conditional

# in ccrlp latino
for chr in 6 7 10 13; do

  echo 1mb lat chr $chr
  
  plink2 --bfile chr$chr.dose.beagle\
    --maf 0.1\
    --indep-pairwise 50 10 0.1

  echo conditional lat chr $chr
  
  plink2 --bfile chr$chr.dose.beagle.conditional\
    --indep-pairwise 50 5 0.5

  echo 5mb lat chr $chr
  
  plink2 --bfile chr$chr.dose.beagle.5mb\
    --maf 0.1\
    --indep-pairwise 50 10 0.1
    
done


#1mb lat chr 6 2188/2233 variants removed. 45
#1mb lat chr 7 1159/1186 variants removed. 27
#1mb lat chr 10 2711/2782 variants removed. 71
#1mb lat chr 13 1878/1923 variants removed. 45

#conditional lat chr 6
#--indep-pairwise (1 compute thread): 8719/11071 variants removed. 2352
#conditional lat chr 7
#--indep-pairwise (1 compute thread): 5648/7419 variants removed. 1771
#conditional lat chr 10
#--indep-pairwise (1 compute thread): 7508/9592 variants removed. 2084
#conditional lat chr 13
#--indep-pairwise (1 compute thread): 6471/8379 variants removed. 1908

#5mb lat chr 6: 9097/9261 variants removed. 164
#5mb lat chr 7: 10142/10389 variants removed. 247
#5mb lat chr 10: 11125/11368 variants removed. 243
#5mb lat chr 13: 9845/10072 variants removed. 227


# in ccrlp european

for chr in 6 7 10 13; do

  echo 1mb white chr $chr
  
  plink2 --bfile chr$chr.dose.wt.beagle\
    --maf 0.1\
    --indep-pairwise 50 10 0.1

  echo conditional white chr $chr
  
  plink2 --bfile chr$chr.dose.wt.beagle.conditional\
    --indep-pairwise 50 5 0.5

  echo 5mb white chr $chr
  
  plink2 --bfile chr$chr.dose.wt.beagle.5mb\
    --maf 0.1\
    --indep-pairwise 50 10 0.1
    
done

#1mb white echo chr 6 2291/2335 variants removed. 44
#1mb white echo chr 7 1177/1208 variants removed. 31
#1mb white echo chr 10 2745/2821 variants removed. 76
#1mb white echo chr 13 1941/1991 variants removed. 50

#conditional white chr 6
#--indep-pairwise (1 compute thread): 8446/10840 variants removed. 2394
#conditional white chr 7
#--indep-pairwise (1 compute thread): 5472/7171 variants removed. 1699
#conditional white chr 10
#--indep-pairwise (1 compute thread): 6927/8907 variants removed. 1980
#conditional white chr 13
#--indep-pairwise (1 compute thread): 6081/7991 variants removed. 1910

#5mb white chr 6 8791/8957 variants removed. 166
#5mb white chr 7 10393/10645 variants removed. 252
#5mb white chr 10 10876/11118 variants removed. 242
#5mb white chr 13 9763/9957 variants removed. 194

```

```{bash meta-analysis of fine mapping results}

cd /scratch/lishaobo/GWAS/glioma/populationRisk2/conditional

#chr6.beagle.conditional.PHENO1.glm.logistic.hybrid
#chr7.beagle.conditional.PHENO1.glm.logistic.hybrid
#chr10.beagle.conditional.PHENO1.glm.logistic.hybrid
#chr13.beagle.conditional.PHENO1.glm.logistic.hybrid

#chr6.wt.beagle.conditional.PHENO1.glm.logistic.hybrid
#chr7.wt.beagle.conditional.PHENO1.glm.logistic.hybrid
#chr10.wt.beagle.conditional.PHENO1.glm.logistic.hybrid
#chr13.wt.beagle.conditional.PHENO1.glm.logistic.hybrid


####################################################################################
# meta-analysis of ccrlp white and latino from conditional analyses
metal
MARKER ID
EFFECT BETA
PVALUE P
SCHEME STDERR
STDERR SE
VERBOSE OFF
GENOMICCONTROL OFF
PROCESS chr6.beagle.conditional.PHENO1.glm.logistic.hybrid
PROCESS chr6.wt.beagle.conditional.PHENO1.glm.logistic.hybrid
ANALYZE HETEROGENEITY
QUIT
mv METAANALYSIS1.TBL meta.chr6.conditional.eurlat.PHENO1.glm.logistic.hybrid
mv METAANALYSIS1.TBL.info meta.chr6.conditional.eurlat.PHENO1.glm.logistic.hybrid.info

metal
MARKER ID
EFFECT BETA
PVALUE P
SCHEME STDERR
STDERR SE
VERBOSE OFF
GENOMICCONTROL OFF
PROCESS chr7.beagle.conditional.PHENO1.glm.logistic.hybrid
PROCESS chr7.wt.beagle.conditional.PHENO1.glm.logistic.hybrid
ANALYZE HETEROGENEITY
QUIT
mv METAANALYSIS1.TBL meta.chr7.conditional.eurlat.PHENO1.glm.logistic.hybrid
mv METAANALYSIS1.TBL.info meta.chr7.conditional.eurlat.PHENO1.glm.logistic.hybrid.info

metal
MARKER ID
EFFECT BETA
PVALUE P
SCHEME STDERR
STDERR SE
VERBOSE OFF
GENOMICCONTROL OFF
PROCESS chr10.beagle.conditional.PHENO1.glm.logistic.hybrid
PROCESS chr10.wt.beagle.conditional.PHENO1.glm.logistic.hybrid
ANALYZE HETEROGENEITY
QUIT
mv METAANALYSIS1.TBL meta.chr10.conditional.eurlat.PHENO1.glm.logistic.hybrid
mv METAANALYSIS1.TBL.info meta.chr10.conditional.eurlat.PHENO1.glm.logistic.hybrid.info

metal
MARKER ID
EFFECT BETA
PVALUE P
SCHEME STDERR
STDERR SE
VERBOSE OFF
GENOMICCONTROL OFF
PROCESS chr13.beagle.conditional.PHENO1.glm.logistic.hybrid
PROCESS chr13.wt.beagle.conditional.PHENO1.glm.logistic.hybrid
ANALYZE HETEROGENEITY
QUIT
mv METAANALYSIS1.TBL meta.chr13.conditional.eurlat.PHENO1.glm.logistic.hybrid
mv METAANALYSIS1.TBL.info meta.chr13.conditional.eurlat.PHENO1.glm.logistic.hybrid.info

```

```{r visualization}

library(tidyverse)
library(QCEWAS)
library(qqman)
library(data.table)

#setwd("/scratch/lishaobo/GWAS/glioma/populationRisk2/conditional")
setwd("/project/wiemels_494/sebastian/temp_scratch_backup/populationRisk2/conditional")


files = c("meta.chr6.conditional.eurlat.PHENO1.glm.logistic.hybrid",
          "meta.chr7.conditional.eurlat.PHENO1.glm.logistic.hybrid",
          "meta.chr10.conditional.eurlat.PHENO1.glm.logistic.hybrid",
          "meta.chr13.conditional.eurlat.PHENO1.glm.logistic.hybrid")

files = c("chr6.beagle.conditional.PHENO1.glm.logistic.hybrid",
          "chr7.beagle.conditional.PHENO1.glm.logistic.hybrid",
          "chr10.beagle.conditional.PHENO1.glm.logistic.hybrid",
          "chr13.beagle.conditional.PHENO1.glm.logistic.hybrid",
          "chr6.wt.beagle.conditional.PHENO1.glm.logistic.hybrid",
          "chr7.wt.beagle.conditional.PHENO1.glm.logistic.hybrid",
          "chr10.wt.beagle.conditional.PHENO1.glm.logistic.hybrid",
          "chr13.wt.beagle.conditional.PHENO1.glm.logistic.hybrid")

# files =c("chr6.beagle.PHENO1.glm.logistic.hybrid",
#          "chr7.beagle.PHENO1.glm.logistic.hybrid",
#          "chr10.beagle.PHENO1.glm.logistic.hybrid",
#          "chr13.beagle.PHENO1.glm.logistic.hybrid",
#          "chr6.beagle.5mb.PHENO1.glm.logistic.hybrid",
#          "chr7.beagle.5mb.PHENO1.glm.logistic.hybrid",
#          "chr10.beagle.5mb.PHENO1.glm.logistic.hybrid",
#          "chr13.beagle.5mb.PHENO1.glm.logistic.hybrid")
# 
# files =c("chr6.wt.beagle.PHENO1.glm.logistic.hybrid",
#          "chr7.wt.beagle.PHENO1.glm.logistic.hybrid",
#          "chr10.wt.beagle.PHENO1.glm.logistic.hybrid",
#          "chr13.wt.beagle.PHENO1.glm.logistic.hybrid",
#          "chr6.wt.beagle.5mb.PHENO1.glm.logistic.hybrid",
#          "chr7.wt.beagle.5mb.PHENO1.glm.logistic.hybrid",
#          "chr10.wt.beagle.5mb.PHENO1.glm.logistic.hybrid",
#          "chr13.wt.beagle.5mb.PHENO1.glm.logistic.hybrid")
# 

for(i in files){
  
  filename <- gsub("^.*/","",i)
  gwas_data <- fread(i) 
  if(length(grep("meta",i)) !=0 ){
    gwas_data_1 <- gwas_data %>%
      dplyr::mutate(ID=MarkerName, P=`P-value`)%>%
      separate(MarkerName, c("CHROM","POS","REF","ALT"))%>%
      dplyr::mutate(CHROM = as.numeric(gsub("chr","",as.character(CHROM))),BP = as.numeric(as.character(POS)))
  } else{
    gwas_data_1 <- gwas_data %>%
      na.omit()%>%
      dplyr::mutate(P= as.numeric(as.character(P)),CHROM = as.numeric(as.character(`#CHROM`)), BP = as.numeric(as.character(POS)))%>%
      arrange(P)
  }
  
  chr <- gwas_data_1$CHROM[1]
  
  
  if(grepl("conditional",i)){
    
    if(!grepl("wt",i)){
      if(chr==6) sig.lev = -log10(0.05/2352)
      if(chr==7) sig.lev = -log10(0.05/1771)
      if(chr==10) sig.lev = -log10(0.05/2084)
      if(chr==13) sig.lev = -log10(0.05/1908)
    }else{
      if(chr==6) sig.lev = -log10(0.05/2394)
      if(chr==7) sig.lev = -log10(0.05/1699)
      if(chr==10) sig.lev = -log10(0.05/1980)
      if(chr==13) sig.lev = -log10(0.05/1910)
      }}
   
    if(grepl("eurlat",i)){
      if(chr==6) sig.lev = -log10(0.05/(2352+2394))
      if(chr==7) sig.lev = -log10(0.05/(1771+1699))
      if(chr==10) sig.lev = -log10(0.05/(2084+1980))
      if(chr==13) sig.lev = -log10(0.05/(1908+1910))}
  
  
  # if(grepl("wt",i)){
  #   
  #       if(!grepl("5mb",i)){
  #       if(chr==6) sig.lev = -log10(0.05/44)
  #       if(chr==7) sig.lev = -log10(0.05/31)
  #       if(chr==10) sig.lev = -log10(0.05/76)
  #       if(chr==13) sig.lev = -log10(0.05/50)
  #     }else{
  #       if(chr==6) sig.lev = -log10(0.05/166)
  #       if(chr==7) sig.lev = -log10(0.05/252)
  #       if(chr==10) sig.lev = -log10(0.05/242)
  #       if(chr==13) sig.lev = -log10(0.05/194)
  #     }}
  # 
  #     
  # if(grepl("oncoA",i)){
  #    
  #     if(!grepl("5mb",i)){
  #       if(chr==6) sig.lev = -log10(0.05/49)
  #       if(chr==7) sig.lev = -log10(0.05/27)
  #       if(chr==10) sig.lev = -log10(0.05/69)
  #       if(chr==13) sig.lev = -log10(0.05/47)
  #     }else{
  #       if(chr==6) sig.lev = -log10(0.05/166)
  #       if(chr==7) sig.lev = -log10(0.05/241)
  #       if(chr==10) sig.lev = -log10(0.05/225)
  #       if(chr==13) sig.lev = -log10(0.05/167)
  #       }}
     
     

  gwas_data_1 %>%
    ggplot(aes(x=BP,y=-log10(P)))+
    geom_point(size=0.5)+
    theme_bw()+
    #ggtitle("Regional Manhattan plot for admixture hits")+
    xlab(paste0("Chromosome ", chr," positions"))+
    geom_hline(yintercept = sig.lev, colour="red",alpha=0.5)+
    #geom_hline(yintercept = sig.lev1, colour="blue",alpha=0.5)+
    scale_x_continuous(labels = function(x) format(x, scientific = FALSE))+
    theme(axis.text.x = element_text(size=13),
      axis.text.y = element_text(size=13),
      axis.title.x = element_text(size=15),
      axis.title.y = element_text(size=15),
      plot.margin = unit(c(0.1, 1, 0.1, 0.2), "cm"))
      #plot.title = element_text(hjust = 0.5))

  ggsave(paste0(filename,".regional.png"),width = 6,height = 4) 
  
}

#i ="meta.chr6.conditional.eurlat.PHENO1.glm.logistic.hybrid"
#i="meta.chr7.conditional.eurlat.PHENO1.glm.logistic.hybrid"
#i="meta.chr10.conditional.eurlat.PHENO1.glm.logistic.hybrid"
#i="meta.chr13.conditional.eurlat.PHENO1.glm.logistic.hybrid"

#FOR table 1
filename <- gsub("^.*/","",i)
gwas_data <- fread(i) 
gwas_data_res = gwas_data %>% 
  filter(!grepl('\\?',Direction))%>%
  arrange(`P-value`)
head(gwas_data_res)

# generating raw numerica file for publication (figure 5)

files = c("meta.chr6.conditional.eurlat.PHENO1.glm.logistic.hybrid",
          "meta.chr7.conditional.eurlat.PHENO1.glm.logistic.hybrid",
          "meta.chr10.conditional.eurlat.PHENO1.glm.logistic.hybrid",
          "meta.chr13.conditional.eurlat.PHENO1.glm.logistic.hybrid")

files1 = c("chr6.beagle.conditional.PHENO1.glm.logistic.hybrid",
          "chr7.beagle.conditional.PHENO1.glm.logistic.hybrid",
          "chr10.beagle.conditional.PHENO1.glm.logistic.hybrid",
          "chr13.beagle.conditional.PHENO1.glm.logistic.hybrid",
          "chr6.wt.beagle.conditional.PHENO1.glm.logistic.hybrid",
          "chr7.wt.beagle.conditional.PHENO1.glm.logistic.hybrid",
          "chr10.wt.beagle.conditional.PHENO1.glm.logistic.hybrid",
          "chr13.wt.beagle.conditional.PHENO1.glm.logistic.hybrid")

rawfilesLS = list()
ct = 1

for(i in files){
  
  # i = "meta.chr6.conditional.eurlat.PHENO1.glm.logistic.hybrid"
  a = fread(i) %>%
    mutate(SnpName = MarkerName,
           pval = `P-value`) %>%
    dplyr::select(SnpName,Effect,pval) %>%
    mutate(model = "meta")
  
  rawfilesLS[[ct]] = a
  ct = ct + 1
  
}


for(i in files1){
  
  # i = "chr6.beagle.conditional.PHENO1.glm.logistic.hybrid"
  
  if(grepl("wt",i)==1){
    model = "nlw"
  }else{
    model = "lat"
  }
  
  a = fread(i) %>%
    mutate(SnpName = ID,
           Effect = BETA,
           pval = P) %>%
    dplyr::select(SnpName,Effect,pval) %>%
    mutate(model = model)
  
  rawfilesLS[[ct]] = a
  ct = ct + 1
  
}

rawfiles = do.call(rbind,rawfilesLS) %>% as.data.frame()

write.table(rawfiles,"rawFileForFig5.txt",
            quote=FALSE,
            sep='\t',
            row.names = FALSE)
```

2. Test SNP_s3 one by one
caco ~ (EUR copies for each SNP_sig) + sex+ 10 PC + EUR global ancestry + AFR global ancestry + (each SNP_s3)
If beta for (EUR copies for each SNP_sig) is 0, that SNP_s3 and its LD explains the difference

```{bash extract potential interesting SNP dosage information}

cd /scratch/lishaobo/GWAS/glioma/populationRisk2/conditional

# chr6
nano chr6.con.candidates
6    86473448
6    85058986
6    85060762
6    85063751
6    85067601
6    86716371
6    86937201
6    86961711

# chr7
nano chr7.con.candidates
7   29415632
7   29472420
7   29519952
7   29545533
7   29592838
7   30373205
7   31078391
7   31316815
7   31391559

# chr10
nano chr10.con.candidates
10    28400571
10    28404092
10    28407495
10    28424513
10    28952520
10    28988612
10    29608502
10    29608947
10    29608998
10    30134977
10    30151491

# chr13
nano chr13.con.candidates
13    93952876
13    94697263
13    94689022
13    94701628
13    94823173
13    94896646
13    95250755
13    95402750
13    95413423
13    95414919

module load bcftools
for chr in 6 7 10 13; do
  bcftools view -R chr$chr.con.candidates\
    ../rgn/$chr.query.phased.vcf.gz -o\
    chr$chr.dosage.vcf.gz -Oz
 done

```

```{r Test SNP_s3 one by one}

library(tidyverse)
library(data.table)
library(vcfR)

setwd("/scratch/lishaobo/GWAS/glioma/populationRisk2/conditional")

pheno = fread("/scratch/lishaobo/GWAS/glioma/populationRisk2/rgn/lat.rfmix.pheono.txt") %>%
  mutate(caco=as.factor(caco)) %>%
  mutate(sex=as.factor(sex))

all.regi = fread("/scratch/lishaobo/GWAS/glioma/populationRisk2/rgn/lat.pol.pcs.long",skip=1)
colnames(all.regi) = c("ID", "Estimate","StdError","Zvalue","P")
all.regi = all.regi %>%
  mutate(ID_dup = ID) %>%
  separate(ID_dup, c("chr","pos"))

for(chrNum in c(6,7,10,13)){
  
  reg = read.vcfR(paste0("chr",chrNum,".dosage.vcf.gz"))
  eur = read.table(paste0("/scratch/lishaobo/GWAS/glioma/populationRisk2/rgn/chr",chrNum,"_eurCount.txt"))
  
  vcf_tidy <-vcfR2tidy(reg, dot_is_NA=TRUE, single_frame=TRUE)
  vcf_data <- vcf_tidy$dat %>% 
    as.data.frame() %>% 
    mutate(ID=paste0(CHROM,"_",POS),
           fullID=paste0(CHROM,"_",POS,"_",ALT,"_",REF))
  
  regi.hits = all.regi %>% 
    filter(chr==chrNum)%>%
    arrange(P)
  regi.hits = regi.hits[1:5,] %>% dplyr::select(ID)
  
  eur1 = eur %>% t() %>% as.data.frame()
  eur2 = eur1[,regi.hits$ID]
  regressors = names(eur2)
  
  # execution of regression
  reg_df_all = list()
  
  reg_out <- lapply(regressors,function(snp){c(data.frame(loc=snp, gene="none"),summary(
    glm(pheno$caco ~ eur2[,snp]+ pheno$sex + pheno$eurGlo+pheno$afrGlo+
          +pheno$PC1+ pheno$PC2+pheno$PC3 + pheno$PC4  + pheno$PC5 + pheno$PC6 + 
          pheno$PC7 + pheno$PC8+pheno$PC9 + pheno$PC10, family="binomial"))$coefficients[2,1:4])})
  reg_df <- bind_rows(reg_out) %>% as.data.frame()
  reg_df_all[[1]] <- reg_df
  
  j=1
  
  for(i in unique(vcf_data$fullID)){
    
    j=j+1
    
    gene_ds = vcf_data %>%
      filter(fullID==i) %>%
      mutate(sub=Indiv) %>%
      dplyr::select(sub,gt_DS) %>%
      right_join(pheno,by="sub")
    
    reg_out <- lapply(regressors,function(snp){c(data.frame(loc=snp,gene=i),summary(
      glm(gene_ds$caco ~ eur2[,snp]+ gene_ds$gt_DS + gene_ds$sex + gene_ds$eurGlo+gene_ds$afrGlo+
            +gene_ds$PC1+ gene_ds$PC2+gene_ds$PC3 + gene_ds$PC4  + gene_ds$PC5 + gene_ds$PC6 + 
            gene_ds$PC7 + gene_ds$PC8+gene_ds$PC9 + gene_ds$PC10,
          family="binomial"))$coefficients[2,1:4])})
    
    reg_df_1 <- bind_rows(reg_out) %>% as.data.frame()
    reg_df_all[[j]] <- reg_df_1
    
    }
    
  reg_all =  do.call(rbind,reg_df_all) %>% 
    as.data.frame() %>%
    arrange(loc)
    
  write.csv(reg_all, paste0("chr",chrNum,".con.begal.output.csv"))

}

########################################################################  
## topmed

## extract that region in bash
#cd /scratch/lishaobo/GWAS/glioma/populationRisk2/conditional
#zcat chr6.dose.vcf.gz | awk '$1 ~ /^#/' > chr6.test.vcf
#zcat chr6.dose.vcf.gz | awk '$2 < 86937201 && $2 > 85019206' >>  chr6.test.vcf

# read gwas results
## bonferroni P value is 0.002941176

# gwas_data <- fread("chr6.topmed.PHENO1.glm.logistic") %>%
#   filter(!P>0.002941176) %>%
#   rename("CHROM"="#CHROM") %>%
#   mutate(CHROM=as.character(CHROM))%>%
#   mutate(ID=paste0(CHROM,"_",POS))%>%
#   select(ID,CHROM,POS)
# 
# reg = read.vcfR("chr6.test.vcf")
# vcf_tidy <-vcfR2tidy(reg, dot_is_NA=TRUE, single_frame=TRUE)
# vcf_data <- vcf_tidy$dat %>% as.data.frame() %>%
#   mutate(CHROM=gsub("chr","",CHROM),
#          Indiv=gsub("^0_","",Indiv))%>%
#   select(CHROM, POS, Indiv, gt_DS)
# 
# vcf_data_1 = vcf_data %>%
#   right_join(gwas_data,by=c("CHROM","POS")) %>%
#   select(CHROM,POS,Indiv,gt_DS) %>%
#   mutate(ID=paste0(CHROM,"_",POS))%>%
#   na.omit()
# 
# # execution of regression
# reg_df_all = list()
# 
# reg_out <- lapply(regressors,function(snp){c(data.frame(loc=snp,gene="none"),summary(
#   glm(pheno$caco ~ eur2[,snp]+ pheno$sex + pheno$eurGlo+pheno$afrGlo+
#         +pheno$PC1+ pheno$PC2+pheno$PC3 + pheno$PC4  + pheno$PC5 + pheno$PC6 + 
#         pheno$PC7 + pheno$PC8+pheno$PC9 + pheno$PC10, family="binomial"))$coefficients[2,1:4])})
# reg_df <- bind_rows(reg_out) %>% as.data.frame()
# 
# reg_df_all[[1]] <- reg_df
# 
# j=1
# 
# for(i in unique(vcf_data_1$ID)){
#   
#   j=j+1
#   
#   gene_ds = vcf_data_1 %>%
#     filter(ID==i) %>%
#     mutate(sub=Indiv) %>%
#     select(sub,gt_DS) %>%
#     arrange(sub,desc(gt_DS)) %>% distinct(sub,.keep_all=TRUE)%>%
#     right_join(pheno,by="sub") 
#   
#   reg_out <- lapply(regressors,function(snp){c(data.frame(loc=snp,gene=i),summary(
#     glm(gene_ds$caco ~ eur2[,snp]+ gene_ds$gt_DS + gene_ds$sex + gene_ds$eurGlo+gene_ds$afrGlo+
#           +gene_ds$PC1+ gene_ds$PC2+gene_ds$PC3 + gene_ds$PC4  + gene_ds$PC5 + gene_ds$PC6 + 
#           gene_ds$PC7 + gene_ds$PC8+gene_ds$PC9 + gene_ds$PC10,
#         family="binomial"))$coefficients[2,1:4])})
#   
#   reg_df_1 <- bind_rows(reg_out) %>% as.data.frame()
#   reg_df_all[[j]] <- reg_df_1
#   
# }
# 
# reg_all =  do.call(rbind,reg_df_all) %>% 
#   as.data.frame() %>%
#   arrange(loc)
# 
# write.table(reg_all, "con.topmed.output.txt", row.names = FALSE, quote=FALSE)

```

```{bash extract SNP conditioned analysis MAF}

cd /scratch/lishaobo/GWAS/glioma/populationRisk2/conditional

for chr in 6 7 10 13; do

  plink2 --bfile chr$chr.dose.beagle.conditional\
    --freq\
    --out lat.2mb.chr$chr.maf

  plink2 --bfile chr$chr.dose.wt.beagle.conditional\
    --freq\
    --out wt.2mb.chr$chr.maf

done


grep ":85067601:" lat.2mb.chr6.maf.afreq
#6	6:85067601:TA:T	TA	T	0.459198	2794
grep ":85067601:" wt.2mb.chr6.maf.afreq
#6	6:85067601:TA:T	TA	T	0.395029	3782

grep ":86716371:" lat.2mb.chr6.maf.afreq
#6	6:86716371:T:G	T	G	0.0295164	2812
 grep ":86716371:" wt.2mb.chr6.maf.afreq
# NA

grep ":86961711:" lat.2mb.chr6.maf.afreq
#6	6:86961711:G:A	G	A	0.0101045	2870
 grep ":86961711:" wt.2mb.chr6.maf.afreq
# 6	6:86961711:G:A	G	A	0.024987	3842

grep ":30134977:" lat.2mb.chr10.maf.afreq
#10	10:30134977:G:GCA	G	GCA	0.331187	2174
 grep ":30134977:" wt.2mb.chr10.maf.afreq
# 10	10:30134977:G:GCA	G	GCA	0.335561	2992

grep ":93952876:" lat.2mb.chr13.maf.afreq
# 13	13:93952876:A:G	A	G	0.262537	2712
grep ":93952876:" wt.2mb.chr13.maf.afreq
# 13	13:93952876:A:G	A	G	0.156888	3818

grep ":94701628:" lat.2mb.chr13.maf.afreq
# 13	13:94701628:A:C	A	C	0.0972028	2860
grep ":94701628:" wt.2mb.chr13.maf.afreq
# 13	13:94701628:A:C	A	C	0.118661	3944

#lat.2mb.chr7.maf.afreq
#lat.2mb.chr10.maf.afreq
#lat.2mb.chr13.maf.afreq
#wt.2mb.chr6.maf.afreq
#wt.2mb.chr7.maf.afreq
#wt.2mb.chr10.maf.afreq
#wt.2mb.chr13.maf.afreq

```

```{bash check the LD of all the 13 SNPs that are significant}

cd /scratch/lishaobo/GWAS/glioma/populationRisk2/conditional

bcftools view ../rgn/6.query.phased.vcf.gz\
  -o 6.reg.raw.peak.vcf\
  -Ov\
  -r 6:85502414-85536683\
  --threads 200

# chr6:85512844

nano 13hits.list
6   85502415    85502415    1
6   85504599    85504599    2
6   85509612    85509612    3
6   85510292    85510292    4
6   85511117    85511117    5
6   85511242    85511242    6
6   85512844    85512844    7
6   85517080    85517080    8
6   85520120    85520120    9
6   85525874    85525874    10
6   85526470    85526470    11
6   85528498    85528498    12
6   85536682    85536682    13

plink2 --vcf 6.reg.raw.peak.vcf\
  --make-bed\
  --extract 'range' 13hits.list\
  --out chr6.ld.check\
  --threads 200

awk -v OFS="\t" '{print $1,$1":"$4":"$5":"$6,$3,$4,$5,$6}' chr6.ld.check.bim > chr6.ld.check.bim.neo 
mv chr6.ld.check.bim.neo chr6.ld.check.bim

plink --bfile chr6.ld.check\
  --r2\
  --out chr6.ld.check.lz

  --ld-snps 6:85512844:C:A\
  --r2\
  --out chr6.ld.check.lz



```

3  targeted region association analysis of additional cohort

```{bash targeted region association analysis}
####################################################################################
# European validation (OncoArray)
# imputed by Topmed

## new version with harmonized ref allele according to CCRLP lat (chr6,7,10,13 only)

#### make the reference allele same as CCRLP latino data
cd /scratch/lishaobo/GWAS/glioma/saxb4/rep

for chr in 6 7 10 13; do
  awk '{print$2,$5}' /scratch/lishaobo/GWAS/glioma/populationRisk2/conditional/chr$chr.dose.beagle.5mb.bim > ccrlp.ref.$chr.list.txt
done

cat ccrlp.ref.6.list.txt ccrlp.ref.7.list.txt ccrlp.ref.10.list.txt ccrlp.ref.13.list.txt > ccrlp.ref.list.txt

plink --bfile rep.oncoA.pol.c1\
  --reference-allele ccrlp.ref.list.txt\
  --make-bed\
  --out rep.oncoA.pol.c1.hmnz

awk '{print $1,$1":"$5":"$4":"$6,$3,$4,$5,$6}' rep.oncoA.pol.c1.hmnz.bim >rep.oncoA.pol.c1.hmnz.bim.1
mv rep.oncoA.pol.c1.chr$chr.hmnz.bim.1 rep.oncoA.pol.c1.chr$chr.hmnz.bim

## the positions below HAVE BEEN LIFTED OVER TO HG38
# 1mb
#nano range1mb.txt
6 84303126 85303126 1
7 29857487 30857487 2
10 28586559 29586559 3        
13 93292962 94292961 4

# 2mb
#nano range2mb.txt
6 83299893 86299894 1
7 29357487 31357489 2
10 28086559 30086559 3        
13 92792962 94792961 4


# 5mb
#nano range5mb.txt
6 82303127 87303126 1
7 27857484 32857491 2
10 26586559 31586560 3        
13 91292961 96292961 4

for chr in 6 7 10 13; do

  plink --bfile rep.oncoA.pol.c1.hmnz\
    --extract 'range' range1mb.txt\
    --chr $chr\
    --make-bed\
    --out rep.oncoA.pol.c1.chr$chr.hmnz.1mb
    
  plink2 --bfile rep.oncoA.pol.c1.chr$chr.hmnz.1mb\
    --covar rep.oncoA.pol.pca.eigenvec\
    --glm hide-covar firth-fallback cols=chrom,pos,ref,alt1,nobs,a1countcc,a1freq,a1freqcc,beta,se,p\
    --out rep.chr$chr.oncoA.pol.1mb\
    --threads 200

  plink --bfile rep.oncoA.pol.c1.hmnz\
    --extract 'range' range2mb.txt\
    --chr $chr\
    --make-bed\
    --out rep.oncoA.pol.c1.chr$chr.hmnz.2mb
    
  plink2 --bfile rep.oncoA.pol.c1.chr$chr.hmnz.2mb\
    --covar rep.oncoA.pol.pca.eigenvec\
    --glm hide-covar firth-fallback cols=chrom,pos,ref,alt1,nobs,a1countcc,a1freq,a1freqcc,beta,se,p\
    --out rep.chr$chr.oncoA.pol.2mb\
    --threads 200

  plink --bfile rep.oncoA.pol.c1.hmnz\
    --extract 'range' range5mb.txt\
    --chr $chr\
    --make-bed\
    --out rep.oncoA.pol.c1.chr$chr.hmnz.5mb
    
  plink2 --bfile rep.oncoA.pol.c1.chr$chr.hmnz.5mb\
    --covar rep.oncoA.pol.pca.eigenvec\
    --glm hide-covar firth-fallback cols=chrom,pos,ref,alt1,nobs,a1countcc,a1freq,a1freqcc,beta,se,p\
    --out rep.chr$chr.oncoA.pol.5mb\
    --threads 200

  cp rep.chr$chr.oncoA.pol.1mb.PHENO1.glm.logistic.hybrid /scratch/lishaobo/GWAS/glioma/populationRisk2/conditional
  cp rep.chr$chr.oncoA.pol.2mb.PHENO1.glm.logistic.hybrid /scratch/lishaobo/GWAS/glioma/populationRisk2/conditional
  cp rep.chr$chr.oncoA.pol.5mb.PHENO1.glm.logistic.hybrid /scratch/lishaobo/GWAS/glioma/populationRisk2/conditional

done

####################################################################################
# European validation (CCRLP, with additional control)

cd /scratch/lishaobo/GWAS/glioma/populationRisk2/conditional

cp /scratch/lishaobo/GWAS/glioma/saxb4/rep/range1mb.txt .
cp /scratch/lishaobo/GWAS/glioma/saxb4/rep/range2mb.txt .
cp /scratch/lishaobo/GWAS/glioma/saxb4/rep/range5mb.txt .

race="eur"
d=03
sub="pol"

for chr in 6 7 10 13; do

  plink --bfile /scratch/lishaobo/GWAS/glioma/saxb4/$race.$sub.$d.glio.1kg.c1\
    --extract 'range' range1mb.txt\
    --chr $chr\
    --make-bed\
    --out ccrlp.adtnctls.$race.$sub.$d.chr$chr.glio.1kg.1mb
    
  plink2 --bfile ccrlp.adtnctls.$race.$sub.$d.chr$chr.glio.1kg.1mb\
    --covar /scratch/lishaobo/GWAS/glioma/saxb4/$race.$sub.$d.glio.1kg.pca.eigenvec\
    --glm hide-covar firth-fallback cols=chrom,pos,ref,alt1,nobs,a1countcc,a1freq,a1freqcc,beta,se,p\
    --out ccrlp.adtnctls.$race.$sub.$d.chr$chr.glio.1kg.1mb\
    --threads 200

  plink --bfile /scratch/lishaobo/GWAS/glioma/saxb4/$race.$sub.$d.glio.1kg.c1\
    --extract 'range' range2mb.txt\
    --chr $chr\
    --make-bed\
    --out ccrlp.adtnctls.$race.$sub.$d.chr$chr.glio.1kg.2mb
    
  plink2 --bfile ccrlp.adtnctls.$race.$sub.$d.chr$chr.glio.1kg.2mb\
    --covar /scratch/lishaobo/GWAS/glioma/saxb4/$race.$sub.$d.glio.1kg.pca.eigenvec\
    --glm hide-covar firth-fallback cols=chrom,pos,ref,alt1,nobs,a1countcc,a1freq,a1freqcc,beta,se,p\
    --out ccrlp.adtnctls.$race.$sub.$d.chr$chr.glio.1kg.2mb\
    --threads 200

  plink --bfile /scratch/lishaobo/GWAS/glioma/saxb4/$race.$sub.$d.glio.1kg.c1\
    --extract 'range' range5mb.txt\
    --chr $chr\
    --make-bed\
    --out ccrlp.adtnctls.$race.$sub.$d.chr$chr.glio.1kg.5mb
    
  plink2 --bfile ccrlp.adtnctls.$race.$sub.$d.chr$chr.glio.1kg.5mb\
    --covar /scratch/lishaobo/GWAS/glioma/saxb4/$race.$sub.$d.glio.1kg.pca.eigenvec\
    --glm hide-covar firth-fallback cols=chrom,pos,ref,alt1,nobs,a1countcc,a1freq,a1freqcc,beta,se,p\
    --out ccrlp.adtnctls.$race.$sub.$d.chr$chr.glio.1kg.5mb\
    --threads 200
done

####################################################################################
# Latino results  (CCRLP, with additional control)

cd /scratch/lishaobo/GWAS/glioma/populationRisk2/conditional

race="lat"
d=03
sub="pol"

for chr in 6 7 10 13; do

  plink --bfile /scratch/lishaobo/GWAS/glioma/saxb4/$race.$sub.$d.glio.1kg.c1\
    --extract 'range' range1mb.txt\
    --chr $chr\
    --make-bed\
    --out ccrlp.adtnctls.$race.$sub.$d.chr$chr.glio.1kg.1mb
    
  plink2 --bfile ccrlp.adtnctls.$race.$sub.$d.chr$chr.glio.1kg.1mb\
    --covar /scratch/lishaobo/GWAS/glioma/saxb4/$race.$sub.$d.glio.1kg.pca.eigenvec\
    --glm hide-covar firth-fallback cols=chrom,pos,ref,alt1,nobs,a1countcc,a1freq,a1freqcc,beta,se,p\
    --out ccrlp.adtnctls.$race.$sub.$d.chr$chr.glio.1kg.1mb\
    --threads 200

  plink --bfile /scratch/lishaobo/GWAS/glioma/saxb4/$race.$sub.$d.glio.1kg.c1\
    --extract 'range' range2mb.txt\
    --chr $chr\
    --make-bed\
    --out ccrlp.adtnctls.$race.$sub.$d.chr$chr.glio.1kg.2mb
    
  plink2 --bfile ccrlp.adtnctls.$race.$sub.$d.chr$chr.glio.1kg.2mb\
    --covar /scratch/lishaobo/GWAS/glioma/saxb4/$race.$sub.$d.glio.1kg.pca.eigenvec\
    --glm hide-covar firth-fallback cols=chrom,pos,ref,alt1,nobs,a1countcc,a1freq,a1freqcc,beta,se,p\
    --out ccrlp.adtnctls.$race.$sub.$d.chr$chr.glio.1kg.2mb\
    --threads 200

  plink --bfile /scratch/lishaobo/GWAS/glioma/saxb4/$race.$sub.$d.glio.1kg.c1\
    --extract 'range' range5mb.txt\
    --chr $chr\
    --make-bed\
    --out ccrlp.adtnctls.$race.$sub.$d.chr$chr.glio.1kg.5mb
    
  plink2 --bfile ccrlp.adtnctls.$race.$sub.$d.chr$chr.glio.1kg.5mb\
    --covar /scratch/lishaobo/GWAS/glioma/saxb4/$race.$sub.$d.glio.1kg.pca.eigenvec\
    --glm hide-covar firth-fallback cols=chrom,pos,ref,alt1,nobs,a1countcc,a1freq,a1freqcc,beta,se,p\
    --out ccrlp.adtnctls.$race.$sub.$d.chr$chr.glio.1kg.5mb\
    --threads 200

done


####################################################################################
## determine number of effective tests
# Bonferroni correction for effective tests following adjustment for LD

cd /scratch/lishaobo/GWAS/glioma/populationRisk2/conditional

# in ccrlp latino with additional controls
race="lat"
d=03
sub="pol"

for chr in 6 7 10 13; do

  echo 1mb ccrlp.add.control chr $chr
  plink2 --bfile ccrlp.adtnctls.$race.$sub.$d.chr$chr.glio.1kg.1mb\
    --indep-pairwise 50 10 0.5

  echo 2mb ccrlp.add.control chr $chr
  plink2 --bfile ccrlp.adtnctls.$race.$sub.$d.chr$chr.glio.1kg.2mb\
    --indep-pairwise 50 10 0.5

  echo 5mb ccrlp.add.control chr $chr
  plink2 --bfile ccrlp.adtnctls.$race.$sub.$d.chr$chr.glio.1kg.5mb\
    --indep-pairwise 50 10 0.5
    
done

#2mb ccrlp.add.control chr 6
#--indep-pairwise (1 compute thread): 7805/9884 variants removed. 2079
#2mb ccrlp.add.control chr 7
#--indep-pairwise (1 compute thread): 5326/7022 variants removed. 1696
#2mb ccrlp.add.control chr 10
#--indep-pairwise (1 compute thread): 6873/8834 variants removed. 1961
#2mb ccrlp.add.control chr 13
#--indep-pairwise (1 compute thread): 5372/6828 variants removed. 1456


# in ccrlp european with additional controls
race="eur"
d=03
sub="pol"

for chr in 6 7 10 13; do

  echo 1mb ccrlp.add.control chr $chr
  plink2 --bfile ccrlp.adtnctls.$race.$sub.$d.chr$chr.glio.1kg.1mb\
    --indep-pairwise 50 10 0.5

  echo 2mb ccrlp.add.control chr $chr
  plink2 --bfile ccrlp.adtnctls.$race.$sub.$d.chr$chr.glio.1kg.2mb\
    --indep-pairwise 50 10 0.5

  echo 5mb ccrlp.add.control chr $chr
  plink2 --bfile ccrlp.adtnctls.$race.$sub.$d.chr$chr.glio.1kg.5mb\
    --indep-pairwise 50 10 0.5
    
done

#2mb ccrlp.add.control chr 6
#--indep-pairwise (1 compute thread): 7139/9182 variants removed. 2043
#2mb ccrlp.add.control chr 7
#--indep-pairwise (1 compute thread): 5105/6756 variants removed. 1651
#2mb ccrlp.add.control chr 10
#--indep-pairwise (1 compute thread): 6523/8423 variants removed. 1900
#2mb ccrlp.add.control chr 13
#--indep-pairwise (1 compute thread): 4962/6432 variants removed. 1470


# in oncoarray european

for chr in 6 7 10 13; do

  echo 1mb duke chr $chr
  plink2 --bfile /scratch/lishaobo/GWAS/glioma/saxb4/rep/rep.oncoA.pol.c1.chr$chr.hmnz.1mb\
    --indep-pairwise 50 10 0.5

  echo 2mb duke chr $chr
  plink2 --bfile /scratch/lishaobo/GWAS/glioma/saxb4/rep/rep.oncoA.pol.c1.chr$chr.hmnz.2mb\
    --indep-pairwise 50 10 0.5

  echo 5mb duke chr $chr
  plink2 --bfile /scratch/lishaobo/GWAS/glioma/saxb4/rep/rep.oncoA.pol.c1.chr$chr.hmnz.5mb\
    --indep-pairwise 50 10 0.5
    
done

#2mb duke chr 6
#--indep-pairwise (1 compute thread): 6606/8529 variants removed. 1923
#2mb duke chr 7
#--indep-pairwise (1 compute thread): 4684/6191 variants removed. 1507
#2mb duke chr 10
#--indep-pairwise (1 compute thread): 6085/7821 variants removed. 1736
#2mb duke chr 13
#--indep-pairwise (1 compute thread): 4598/5989 variants removed. 1391


####################################################################################
# meta-analysis of ccrlp white and latino (results not good)
metal
MARKER ID
EFFECT BETA
PVALUE P
SCHEME STDERR
STDERR SE
VERBOSE OFF
GENOMICCONTROL OFF
PROCESS chr6.beagle.PHENO1.glm.logistic.hybrid
PROCESS chr6.beagle.wt.PHENO1.glm.logistic
ANALYZE HETEROGENEITY
QUIT
mv METAANALYSIS1.TBL meta.chr6.beagle.eurlat.PHENO1.glm.logistic
mv METAANALYSIS1.TBL.info meta.chr6.beagle.eurlat.PHENO1.glm.logistic.TBL.info

grep -E -v "\?" meta.chr6.beagle.eurlat.PHENO1.glm.logistic >\
meta.flted.chr6.beagle.eurlat.PHENO1.glm.logistic

metal
MARKER ID
EFFECT BETA
PVALUE P
SCHEME STDERR
STDERR SE
VERBOSE OFF
GENOMICCONTROL OFF
PROCESS chr7.beagle.PHENO1.glm.logistic.hybrid
PROCESS chr7.beagle.wt.PHENO1.glm.logistic
ANALYZE HETEROGENEITY
QUIT
mv METAANALYSIS1.TBL meta.chr7.beagle.eurlat.PHENO1.glm.logistic
mv METAANALYSIS1.TBL.info meta.chr7.beagle.eurlat.PHENO1.glm.logistic.TBL.info

grep -E -v "\?" meta.chr7.beagle.eurlat.PHENO1.glm.logistic >\
meta.flted.chr7.beagle.eurlat.PHENO1.glm.logistic

metal
MARKER ID
EFFECT BETA
PVALUE P
SCHEME STDERR
STDERR SE
VERBOSE OFF
GENOMICCONTROL OFF
PROCESS chr10.beagle.PHENO1.glm.logistic.hybrid
PROCESS chr10.beagle.wt.PHENO1.glm.logistic
ANALYZE HETEROGENEITY
QUIT
mv METAANALYSIS1.TBL meta.chr10.beagle.eurlat.PHENO1.glm.logistic
mv METAANALYSIS1.TBL.info meta.chr10.beagle.eurlat.PHENO1.glm.logistic.TBL.info

grep -E -v "\?" meta.chr10.beagle.eurlat.PHENO1.glm.logistic >\
meta.flted.chr10.beagle.eurlat.PHENO1.glm.logistic

metal
MARKER ID
EFFECT BETA
PVALUE P
SCHEME STDERR
STDERR SE
VERBOSE OFF
GENOMICCONTROL OFF
PROCESS chr13.beagle.PHENO1.glm.logistic.hybrid
PROCESS chr13.beagle.wt.PHENO1.glm.logistic
ANALYZE HETEROGENEITY
QUIT
mv METAANALYSIS1.TBL meta.chr13.beagle.eurlat.PHENO1.glm.logistic
mv METAANALYSIS1.TBL.info meta.chr13.beagle.eurlat.PHENO1.glm.logistic.TBL.info

grep -E -v "\?" meta.chr13.beagle.eurlat.PHENO1.glm.logistic >\
meta.flted.chr13.beagle.eurlat.PHENO1.glm.logistic

####################################################################################
# meta-analysis of white and latino, white is from oncoArray

cd /scratch/lishaobo/GWAS/glioma/populationRisk2/conditional

metal
MARKER ID
EFFECT BETA
PVALUE P
SCHEME STDERR
STDERR SE
VERBOSE OFF
GENOMICCONTROL OFF
PROCESS chr6.beagle.PHENO1.glm.logistic.hybrid
PROCESS rep.chr6.oncoA.pol.PHENO1.glm.logistic.hybrid
ANALYZE HETEROGENEITY
QUIT
mv METAANALYSIS1.TBL meta.chr6.beagle.onco.eurlat.PHENO1.glm.logistic
mv METAANALYSIS1.TBL.info meta.chr6.beagle.onco.eurlat.PHENO1.glm.logistic.TBL.info

grep -E -v "\?" meta.chr6.beagle.onco.eurlat.PHENO1.glm.logistic >\
meta.flted.chr6.beagle.onco.eurlat.PHENO1.glm.logistic


metal
MARKER ID
EFFECT BETA
PVALUE P
SCHEME STDERR
STDERR SE
VERBOSE OFF
GENOMICCONTROL OFF
PROCESS chr7.beagle.PHENO1.glm.logistic.hybrid
PROCESS rep.chr7.oncoA.pol.PHENO1.glm.logistic.hybrid
ANALYZE HETEROGENEITY
QUIT
mv METAANALYSIS1.TBL meta.chr7.beagle.onco.eurlat.PHENO1.glm.logistic
mv METAANALYSIS1.TBL.info meta.chr7.beagle.onco.eurlat.PHENO1.glm.logistic.TBL.info

grep -E -v "\?" meta.chr7.beagle.onco.eurlat.PHENO1.glm.logistic >\
meta.flted.chr7.beagle.onco.eurlat.PHENO1.glm.logistic

metal
MARKER ID
EFFECT BETA
PVALUE P
SCHEME STDERR
STDERR SE
VERBOSE OFF
GENOMICCONTROL OFF
PROCESS chr10.beagle.PHENO1.glm.logistic.hybrid
PROCESS rep.chr10.oncoA.pol.PHENO1.glm.logistic.hybrid
ANALYZE HETEROGENEITY
QUIT
mv METAANALYSIS1.TBL meta.chr10.beagle.onco.eurlat.PHENO1.glm.logistic
mv METAANALYSIS1.TBL.info meta.chr10.beagle.onco.eurlat.PHENO1.glm.logistic.TBL.info

grep -E -v "\?" meta.chr10.beagle.onco.eurlat.PHENO1.glm.logistic >\
meta.flted.chr10.beagle.onco.eurlat.PHENO1.glm.logistic


metal
MARKER ID
EFFECT BETA
PVALUE P
SCHEME STDERR
STDERR SE
VERBOSE OFF
GENOMICCONTROL OFF
PROCESS chr13.beagle.PHENO1.glm.logistic.hybrid
PROCESS rep.chr13.oncoA.pol.PHENO1.glm.logistic.hybrid
ANALYZE HETEROGENEITY
QUIT
mv METAANALYSIS1.TBL meta.chr13.beagle.onco.eurlat.PHENO1.glm.logistic
mv METAANALYSIS1.TBL.info meta.chr13.beagle.onco.eurlat.PHENO1.glm.logistic.TBL.info

grep -E -v "\?" meta.chr13.beagle.onco.eurlat.PHENO1.glm.logistic >\
meta.flted.chr13.beagle.onco.eurlat.PHENO1.glm.logistic

```

```{r visualization}

library(tidyverse)
library(QCEWAS)
library(qqman)
library(data.table)

setwd("/scratch/lishaobo/GWAS/glioma/populationRisk2/conditional")

# files =c("rep.chr6.oncoA.pol.1mb.PHENO1.glm.logistic.hybrid",
#          "rep.chr7.oncoA.pol.1mb.PHENO1.glm.logistic.hybrid",
#          "rep.chr10.oncoA.pol.1mb.PHENO1.glm.logistic.hybrid",
#          "rep.chr13.oncoA.pol.1mb.PHENO1.glm.logistic.hybrid",
#          "rep.chr6.oncoA.pol.2mb.PHENO1.glm.logistic.hybrid",
#          "rep.chr7.oncoA.pol.2mb.PHENO1.glm.logistic.hybrid",
#          "rep.chr10.oncoA.pol.2mb.PHENO1.glm.logistic.hybrid",
#          "rep.chr13.oncoA.pol.2mb.PHENO1.glm.logistic.hybrid",
#           "rep.chr6.oncoA.pol.5mb.PHENO1.glm.logistic.hybrid",
#          "rep.chr7.oncoA.pol.5mb.PHENO1.glm.logistic.hybrid",
#          "rep.chr10.oncoA.pol.5mb.PHENO1.glm.logistic.hybrid",
#          "rep.chr13.oncoA.pol.5mb.PHENO1.glm.logistic.hybrid")

files =c("rep.chr6.oncoA.pol.2mb.PHENO1.glm.logistic.hybrid",
         "rep.chr7.oncoA.pol.2mb.PHENO1.glm.logistic.hybrid",
         "rep.chr10.oncoA.pol.2mb.PHENO1.glm.logistic.hybrid",
         "rep.chr13.oncoA.pol.2mb.PHENO1.glm.logistic.hybrid")

for(i in files){
  
  filename <- gsub("^.*/","",i)
  gwas_data <- fread(i) 
  if(length(grep("meta",i)) !=0 ){
    gwas_data_1 <- gwas_data %>%
      dplyr::mutate(ID=MarkerName, P=`P-value`)%>%
      separate(MarkerName, c("CHROM","POS","REF","ALT"))%>%
      dplyr::mutate(CHROM = as.numeric(gsub("chr","",as.character(CHROM))),BP = as.numeric(as.character(POS)))
  } else{
    gwas_data_1 <- gwas_data %>%
      na.omit()%>%
      dplyr::mutate(P= as.numeric(as.character(P)),CHROM = as.numeric(as.character(`#CHROM`)), BP = as.numeric(as.character(POS)))%>%
      arrange(P)
  }
  
  chr <- gwas_data_1$CHROM[1]
  

# if(grepl("1mb",i)){
#   if(chr==6) sig.lev = -log10(0.05/49)
#   if(chr==7) sig.lev = -log10(0.05/27)
#   if(chr==10) sig.lev = -log10(0.05/69)
#   if(chr==13) sig.lev = -log10(0.05/47)}
  
if(grepl("2mb",i)){
  if(chr==6) sig.lev = -log10(0.05/1923)
  if(chr==7) sig.lev = -log10(0.05/1507)
  if(chr==10) sig.lev = -log10(0.05/1736)
  if(chr==13) sig.lev = -log10(0.05/1391)}
  
# if(grepl("5mb",i)){
#   if(chr==6) sig.lev = -log10(0.05/49)
#   if(chr==7) sig.lev = -log10(0.05/27)
#   if(chr==10) sig.lev = -log10(0.05/69)
#   if(chr==13) sig.lev = -log10(0.05/47)}  
    
  gwas_data_1 %>%
    ggplot(aes(x=BP,y=-log10(P)))+
    geom_point(size=0.5)+
    theme_bw()+
    #ggtitle("Regional Manhattan plot for admixture hits")+
    xlab(paste0("Chromosome ", chr," positions"))+
    geom_hline(yintercept = sig.lev, colour="red",alpha=0.5)+
    #geom_hline(yintercept = sig.lev1, colour="blue",alpha=0.5)+
    scale_x_continuous(labels = function(x) format(x, scientific = FALSE))+
    theme(axis.text.x = element_text(size=13),
      axis.text.y = element_text(size=13),
      axis.title.x = element_text(size=15),
      axis.title.y = element_text(size=15),
      plot.margin = unit(c(0.1, 1, 0.1, 0.2), "cm"))
      #plot.title = element_text(hjust = 0.5))

  ggsave(paste0(filename,".regional.png"),width = 6,height = 4) 
  
}

# files =c("ccrlp.adtnctls.eur.pol.03.chr6.glio.1kg.1mb.PHENO1.glm.logistic.hybrid",
#          "ccrlp.adtnctls.eur.pol.03.chr7.glio.1kg.1mb.PHENO1.glm.logistic.hybrid",
#          "ccrlp.adtnctls.eur.pol.03.chr10.glio.1kg.1mb.PHENO1.glm.logistic.hybrid",
#          "ccrlp.adtnctls.eur.pol.03.chr13.glio.1kg.1mb.PHENO1.glm.logistic.hybrid",
#          "ccrlp.adtnctls.eur.pol.03.chr6.glio.1kg.2mb.PHENO1.glm.logistic.hybrid",
#          "ccrlp.adtnctls.eur.pol.03.chr7.glio.1kg.2mb.PHENO1.glm.logistic.hybrid",
#          "ccrlp.adtnctls.eur.pol.03.chr10.glio.1kg.2mb.PHENO1.glm.logistic.hybrid",
#          "ccrlp.adtnctls.eur.pol.03.chr13.glio.1kg.2mb.PHENO1.glm.logistic.hybrid",
#           "ccrlp.adtnctls.eur.pol.03.chr6.glio.1kg.5mb.PHENO1.glm.logistic.hybrid",
#          "ccrlp.adtnctls.eur.pol.03.chr7.glio.1kg.5mb.PHENO1.glm.logistic.hybrid",
#          "ccrlp.adtnctls.eur.pol.03.chr10.glio.1kg.5mb.PHENO1.glm.logistic.hybrid",
#          "ccrlp.adtnctls.eur.pol.03.chr13.glio.1kg.5mb.PHENO1.glm.logistic.hybrid",
#          "ccrlp.adtnctls.lat.pol.03.chr6.glio.1kg.1mb.PHENO1.glm.logistic.hybrid",
#          "ccrlp.adtnctls.lat.pol.03.chr7.glio.1kg.1mb.PHENO1.glm.logistic.hybrid",
#          "ccrlp.adtnctls.lat.pol.03.chr10.glio.1kg.1mb.PHENO1.glm.logistic.hybrid",
#          "ccrlp.adtnctls.lat.pol.03.chr13.glio.1kg.1mb.PHENO1.glm.logistic.hybrid",
#          "ccrlp.adtnctls.lat.pol.03.chr6.glio.1kg.2mb.PHENO1.glm.logistic.hybrid",
#          "ccrlp.adtnctls.lat.pol.03.chr7.glio.1kg.2mb.PHENO1.glm.logistic.hybrid",
#          "ccrlp.adtnctls.lat.pol.03.chr10.glio.1kg.2mb.PHENO1.glm.logistic.hybrid",
#          "ccrlp.adtnctls.lat.pol.03.chr13.glio.1kg.2mb.PHENO1.glm.logistic.hybrid",
#           "ccrlp.adtnctls.lat.pol.03.chr6.glio.1kg.5mb.PHENO1.glm.logistic.hybrid",
#          "ccrlp.adtnctls.lat.pol.03.chr7.glio.1kg.5mb.PHENO1.glm.logistic.hybrid",
#          "ccrlp.adtnctls.lat.pol.03.chr10.glio.1kg.5mb.PHENO1.glm.logistic.hybrid",
#          "ccrlp.adtnctls.lat.pol.03.chr13.glio.1kg.5mb.PHENO1.glm.logistic.hybrid")

files =c("ccrlp.adtnctls.eur.pol.03.chr6.glio.1kg.2mb.PHENO1.glm.logistic.hybrid",
         "ccrlp.adtnctls.eur.pol.03.chr7.glio.1kg.2mb.PHENO1.glm.logistic.hybrid",
         "ccrlp.adtnctls.eur.pol.03.chr10.glio.1kg.2mb.PHENO1.glm.logistic.hybrid",
         "ccrlp.adtnctls.eur.pol.03.chr13.glio.1kg.2mb.PHENO1.glm.logistic.hybrid",
         "ccrlp.adtnctls.lat.pol.03.chr6.glio.1kg.2mb.PHENO1.glm.logistic.hybrid",
         "ccrlp.adtnctls.lat.pol.03.chr7.glio.1kg.2mb.PHENO1.glm.logistic.hybrid",
         "ccrlp.adtnctls.lat.pol.03.chr10.glio.1kg.2mb.PHENO1.glm.logistic.hybrid",
         "ccrlp.adtnctls.lat.pol.03.chr13.glio.1kg.2mb.PHENO1.glm.logistic.hybrid")

for(i in files){
  
  filename <- gsub("^.*/","",i)
  gwas_data <- fread(i) 
  if(length(grep("meta",i)) !=0 ){
    gwas_data_1 <- gwas_data %>%
      dplyr::mutate(ID=MarkerName, P=`P-value`)%>%
      separate(MarkerName, c("CHROM","POS","REF","ALT"))%>%
      dplyr::mutate(CHROM = as.numeric(gsub("chr","",as.character(CHROM))),BP = as.numeric(as.character(POS)))
  } else{
    gwas_data_1 <- gwas_data %>%
      na.omit()%>%
      dplyr::mutate(P= as.numeric(as.character(P)),CHROM = as.numeric(as.character(`#CHROM`)), BP = as.numeric(as.character(POS)))%>%
      arrange(P)
  }
  
  chr <- gwas_data_1$CHROM[1]
  
  if(grepl("lat",i)){
    
    # if(grepl("1mb",i)){
    #   if(chr==6) sig.lev = -log10(0.05/44)
    #   if(chr==7) sig.lev = -log10(0.05/25)
    #   if(chr==10) sig.lev = -log10(0.05/68)
    #   if(chr==13) sig.lev = -log10(0.05/45)}

    if(grepl("2mb",i)){
      if(chr==6) sig.lev = -log10(0.05/2079)
      if(chr==7) sig.lev = -log10(0.05/1696)
      if(chr==10) sig.lev = -log10(0.05/1961)
      if(chr==13) sig.lev = -log10(0.05/1456)}

    # if(grepl("5mb",i)){
    #     if(chr==6) sig.lev = -log10(0.05/44)
    #     if(chr==7) sig.lev = -log10(0.05/25)
    #     if(chr==10) sig.lev = -log10(0.05/68)
    #     if(chr==13) sig.lev = -log10(0.05/45)}
    }
      

  if(grepl("eur",i)){
    
    # if(grepl("1mb",i)){
    #   if(chr==6) sig.lev = -log10(0.05/44)
    #   if(chr==7) sig.lev = -log10(0.05/25)
    #   if(chr==10) sig.lev = -log10(0.05/68)
    #   if(chr==13) sig.lev = -log10(0.05/45)}

    if(grepl("2mb",i)){
      if(chr==6) sig.lev = -log10(0.05/2043)
      if(chr==7) sig.lev = -log10(0.05/1651)
      if(chr==10) sig.lev = -log10(0.05/1900)
      if(chr==13) sig.lev = -log10(0.05/1470)}

    # if(grepl("5mb",i)){
    #     if(chr==6) sig.lev = -log10(0.05/44)
    #     if(chr==7) sig.lev = -log10(0.05/25)
    #     if(chr==10) sig.lev = -log10(0.05/68)
    #     if(chr==13) sig.lev = -log10(0.05/45)}
    }
        
  
  
  gwas_data_1 %>%
    ggplot(aes(x=BP,y=-log10(P)))+
    geom_point(size=0.5)+
    theme_bw()+
    #ggtitle("Regional Manhattan plot for admixture hits")+
    xlab(paste0("Chromosome ", chr," positions"))+
    geom_hline(yintercept = sig.lev, colour="red",alpha=0.5)+
    #geom_hline(yintercept = sig.lev1, colour="blue",alpha=0.5)+
    scale_x_continuous(labels = function(x) format(x, scientific = FALSE))+
    theme(axis.text.x = element_text(size=13),
      axis.text.y = element_text(size=13),
      axis.title.x = element_text(size=15),
      axis.title.y = element_text(size=15),
      plot.margin = unit(c(0.1, 1, 0.1, 0.2), "cm"))
      #plot.title = element_text(hjust = 0.5))

  ggsave(paste0(filename,".regional.png"),width = 6,height = 4) 
  
}

###########################################################################look up in oncoArray replication
library(tidyverse)
library(QCEWAS)
library(qqman)
library(data.table)

setwd("/scratch/lishaobo/GWAS/glioma/populationRisk2/conditional")

files =c("rep.chr6.oncoA.pol.PHENO1.glm.logistic.hybrid",
         "rep.chr7.oncoA.pol.PHENO1.glm.logistic.hybrid",
         "rep.chr10.oncoA.pol.PHENO1.glm.logistic.hybrid",
         "rep.chr13.oncoA.pol.PHENO1.glm.logistic.hybrid")




i = "rep.chr6.oncoA.pol.PHENO1.glm.logistic.hybrid"
gwas_data <- fread(i) 
filename <- gsub("^.*/","",i)

gwas_data_1 <- gwas_data %>%
  na.omit()%>%
  dplyr::mutate(P= as.numeric(as.character(P)),CHR = as.numeric(as.character(`#CHROM`)), BP = as.numeric(as.character(POS)))%>%
  filter(P<0.05/45)


sig.lev = -log10(0.05/45)

gwas_data_1 %>%
  ggplot(aes(x=BP,y=-log10(P)))+
  geom_point(size=1, alpha=0.5)+
  theme_bw()+
  ggtitle("Regional Manhattan plot in region of admixture signal\n replication dataset")+
  xlab("Chromosome 6 positions")+
  geom_hline(yintercept = sig.lev, colour="red",alpha=0.5)+
  #geom_hline(yintercept = sig.lev1, colour="blue",alpha=0.5)+
  scale_x_continuous(labels = function(x) format(x, scientific = FALSE))+
  theme(plot.title = element_text(hjust = 0.5))
ggsave(paste0(filename,".regional.png"),width = 6,height = 4) 
  

###########################################################################look up for replication
i = "meta.flted.chr6.beagle.onco.eurlat.PHENO1.glm.logistic"
gwas_data <- fread(i) 

if(length(grep("meta",i)) !=0 ){
  gwas_data_1 <- gwas_data %>%
    dplyr::mutate(ID=MarkerName, P=`P-value`)%>%
    separate(MarkerName, c("CHROM","POS","REF","ALT"))%>%
    dplyr::mutate(CHROM = as.numeric(gsub("chr","",as.character(CHROM))),BP = as.numeric(as.character(POS)))%>%
    arrange(P)
} else{
  gwas_data_1 <- gwas_data %>%
    na.omit()%>%
    dplyr::mutate(P= as.numeric(as.character(P)),CHR = as.numeric(as.character(`#CHROM`)), BP = as.numeric(as.character(POS)))%>%
    arrange(P)
}
  
sig.lev = -log10(0.05/45)

gwas_data_1 %>%
  ggplot(aes(x=BP,y=-log10(P)))+
  geom_point(size=1, alpha=0.5)+
  theme_bw()+
  ggtitle("Regional Manhattan plot in region of admixture signal\n meta-analysis")+
  xlab("Chromosome 6 positions")+
  geom_hline(yintercept = sig.lev, colour="red",alpha=0.5)+
  #geom_hline(yintercept = sig.lev1, colour="blue",alpha=0.5)+
  scale_x_continuous(labels = function(x) format(x, scientific = FALSE))+
  theme(plot.title = element_text(hjust = 0.5))
ggsave(paste0(filename,".regional.png"),width = 6,height = 4) 
  

# [1] "lambda value of chr6.beagle.PHENO1.glm.logistic is: 1.38229574987352"
# [1] "lambda value of chr6.beagle.wt.PHENO1.glm.logistic is: 1.0097199092206"
# [1] "lambda value of meta.flted.chr6.beagle.eurlat.PHENO1.glm.logistic is: 1.02352687491244"

# threshold for significance
## [1] 0.0003787879

## chr6:85439850
```

# locuszoom plots for regional admixture results

>> Latinos

chr6:85067601:T:TA

```{bash}

cd /scratch/lishaobo/GWAS/glioma/populationRisk2/rgn

cp /scratch/lishaobo/GWAS/glioma/populationRisk2/conditional/chr6.dose.long.beagle.??? .

grep "6:85067601" chr6.dose.long.beagle.bim
#	6	6:85067601:T:TA	0	85067601	T	TA

plink --bfile chr6.dose.long.beagle\
  --allow-no-sex\
  --ld-window-kb 500\
  --ld-snps 6:85067601:T:TA\
  --r2 dprime\
  --ld-window-r2 0\
  --ld-window 99999\
  --out 6.query.lz

```

```{python}

## python3 script
f=open('6.query.lz.ld')
f.readline()
o=open('6.query.lz.ld_lz','w')
o.write('snp1 snp2 dprime rsquare\n')
for lines in f:
 lines=lines.split()
 o.write(lines[2].split(':')[0]+':'+lines[2].split(':')[1]+' '+lines[5].split(':')[0]+':'+lines[5].split(':')[1]+' '+lines[7]+' '+lines[6]+'\n')

o.close()
```

```{r}

setwd("/scratch/lishaobo/GWAS/glioma/populationRisk2/rgn")
library(tidyverse)
library(data.table)

lat_lz = fread("/scratch/lishaobo/GWAS/glioma/populationRisk2/conditional/chr6.long.beagle.PHENO1.glm.logistic.hybrid") %>%
  mutate(MarkerName = paste0("chr",`#CHROM`,":",POS),
         `P-value`=P) %>%
  select(MarkerName,`P-value`) %>%
  na.omit()
write.table(lat_lz,"lat.reg.adm.logistic_lz", sep='\t', row.names = FALSE, quote=FALSE)
```

>> Europeans

chr6:85381489:C:G

```{bash}

cd /scratch/lishaobo/GWAS/glioma/populationRisk2/rgn

plink2 --bfile /scratch/lishaobo/GWAS/glioma/saxb4/rep/rep.oncoA.c1\
  --chr 6\
  --make-bed\
  --out 6.query.wt.lz

awk -v OFS='\t' '{print $1,$1":"$4":"$5":"$6,$3,$4,$5,$6}' 6.query.wt.lz.bim > 6.query.wt.lz.bim.neo
mv 6.query.wt.lz.bim.neo 6.query.wt.lz.bim

grep "6:85381489" 6.query.wt.lz.bim
#	6:85381489:C:G

plink --bfile 6.query.wt.lz\
  --allow-no-sex\
  --ld-window-kb 500\
  --ld-snps 6:85381489:C:G\
  --r2 dprime\
  --ld-window-r2 0\
  --ld-window 99999\
  --out 6.query.wt.lz

```

```{python}

## python3 script
f=open('6.query.wt.lz.ld')
f.readline()
o=open('6.query.wt.lz.ld_lz','w')
o.write('snp1 snp2 dprime rsquare\n')
for lines in f:
 lines=lines.split()
 o.write(lines[2].split(':')[0]+':'+lines[2].split(':')[1]+' '+lines[5].split(':')[0]+':'+lines[5].split(':')[1]+' '+lines[7]+' '+lines[6]+'\n')

o.close()
```

```{r}

setwd("/scratch/lishaobo/GWAS/glioma/populationRisk2/rgn")
library(tidyverse)
library(data.table)

Wt_lz = fread("/scratch/lishaobo/GWAS/glioma/populationRisk2/conditional/rep.chr6.oncoA.pol.PHENO1.glm.logistic.hybrid") %>%
  mutate(MarkerName = paste0("chr",`#CHROM`,":",POS),
         `P-value`=P) %>%
  select(MarkerName,`P-value`) %>%
  na.omit()
write.table(Wt_lz,"wt.reg.adm.logistic_lz", sep='\t', row.names = FALSE, quote=FALSE)
```

```{r make corresponding recombination rate}

setwd("/scratch/lishaobo/GWAS/glioma/populationRisk2/rgn")
library(tidyverse)
library(data.table)

all.rgs = fread("/project/wiemels_260/sebastian/dependencies/rfmix/rfmix.GRCh38.map/rfmix.GRCh38.map")
all.rgs$V4 = 0

for(i in 1:(nrow(all.rgs)-1)){
  all.rgs$V4[i] <- (all.rgs$V3[i+1]-all.rgs$V3[i])/(all.rgs$V2[i+1]-all.rgs$V2[i])*10^6
}

all.rgs.1 <- all.rgs %>%
  dplyr::select(V1,V2,V4,V3)

colnames(all.rgs.1) <- c("chr",	"pos",	"recomb",	"cm_pos")
write.table(all.rgs.1,"recomb.hg38.map.txt",sep='\t', row.names = FALSE, quote=FALSE)


#### faster version
setwd("/scratch/lishaobo/GWAS/glioma/populationRisk2/rgn")
library(tidyverse)
library(data.table)
library(foreach)

n.cores <- parallel::detectCores() - 1
print(n.cores)

my.cluster <- parallel::makeCluster(
  n.cores, 
  type = "FORK")
print(my.cluster)

doParallel::registerDoParallel(cl = my.cluster)
print(foreach::getDoParRegistered())
print(foreach::getDoParWorkers())

all.rgs = fread("/project/wiemels_260/sebastian/dependencies/rfmix/rfmix.GRCh38.map/rfmix.GRCh38.map")
all.rgs$V4 = 0

x <- foreach(
  i = 1:(nrow(all.rgs)-1), 
  .combine = 'c'
) %dopar% {
  (all.rgs$V3[i+1]-all.rgs$V3[i])/(all.rgs$V2[i+1]-all.rgs$V2[i])*10^6
}

print(class(x))
print(head(x))
#write_rds(x,"for.each.x.rds")

all.rgs$V4[1:(nrow(all.rgs)-1)] <- x

all.rgs.1 <- all.rgs %>%
  dplyr::select(V1,V2,V4,V3)

colnames(all.rgs.1) <- c("chr","pos","recomb","cm_pos")
write.table(all.rgs.1,"recomb.hg38.map.txt",sep='\t', row.names = FALSE, quote=FALSE)

```


```{bash}
#!/bin/bash
#SBATCH --time=20:00:00
#SBATCH --mail-type=END
#SBATCH --job-name=lz_rg
#SBATCH --mem=80G

set -e 
set -u 
set -x
set -o pipefail 

#nano /scratch/lishaobo/GWAS/glioma/populationRisk2/rgn/bash.sh
#sbatch /scratch/lishaobo/GWAS/glioma/populationRisk2/rgn/bash.sh

# adding recombination rate to hg38 datapoints

cd /scratch/lishaobo/GWAS/glioma/populationRisk2/rgn

module load python/2.7.16
python /project/wiemels_260/sebastian/dependencies/locuszoom/locuszoom/bin/dbmeister.py\
  --db /project/wiemels_260/sebastian/dependencies/locuszoom/locuszoom/data/database/locuszoom_hg38.db\
  --recomb_rate /scratch/lishaobo/GWAS/glioma/populationRisk2/rgn/recomb.hg38.map.txt 
  
cd /scratch/lishaobo/GWAS/glioma/populationRisk2/rgn

locuszoom --build hg38\
  --refsnp 6:85067601\
  --flank 500kb\
  --ld 6.query.lz.ld_lz\
  --metal lat.reg.adm.logistic_lz\
  showRecomb=T\
  recombOver=TRUE\
  --prefix chr6.lat\
  --plotonly

locuszoom --build hg38\
  --refsnp 6:85381489\
  --flank 500kb\
  --ld 6.query.wt.lz.ld_lz\
  --metal wt.reg.adm.logistic_lz\
  showRecomb=T\
  recombOver=TRUE\
  --prefix chr6.wt\
  --plotonly

```

#iSAFE analysis

```{bash}

cd /scratch/lishaobo/GWAS/glioma/populationRisk2/rgn
module load python/2.7.16

## download hg38 ancestral alleles
#cd /scratch/lishaobo/GWAS/glioma/populationRisk2/ancestralAllele
#wget https://ftp.ensembl.org/pub/current_fasta/ancestral_alleles/homo_sapiens_ancestor_GRCh38.tar.gz

## controls are whites (ancestors) (--vcf-cont)
## cases are hispanics (--input)

#tabix 6.query.phased.vcf.gz
#tabix 6.panel.phased.vcf.gz

awk -v OFS='\t' '{print "EUR",$1"_"$2}' ../EURpop.txt > isafe.eur.cons

# 1Mb version
python2.7 /project/wiemels_260/sebastian/dependencies/isafe/iSAFE-1.0.7/src/isafe.py\
  --input 6.query.phased.vcf.gz\
  --output isafe.chr6.adrp\
  --region 6:84566601-86012844\
  --AA /scratch/lishaobo/GWAS/glioma/populationRisk2/ancestralAllele/homo_sapiens_ancestor_GRCh38/homo_sapiens_ancestor_6.fa\
  --vcf-cont 6.panel.phased.vcf.gz\
  --sample-cont isafe.eur.cons

# 5Mb version
python2.7 /project/wiemels_260/sebastian/dependencies/isafe/iSAFE-1.0.7/src/isafe.py\
  --input 6.query.phased.vcf.gz\
  --output isafe.chr6.5mb.adrp\
  --region 6:82789723-87789722\
  --AA /scratch/lishaobo/GWAS/glioma/populationRisk2/ancestralAllele/homo_sapiens_ancestor_GRCh38/homo_sapiens_ancestor_6.fa\
  --vcf-cont 6.panel.phased.vcf.gz\
  --sample-cont isafe.eur.cons
```

```{r visualization of iSAFE results}

setwd("/scratch/lishaobo/GWAS/glioma/populationRisk2/rgn")
library(tidyverse)
library(data.table)

# 1Mb region
isafe.res = fread("isafe.chr6.adrp.iSAFE.out")

isafe.res %>%
  ggplot(aes(x=POS,y=iSAFE))+
  geom_point(size=1, alpha=0.5)+
  theme_bw()+
  #ggtitle("Selective sweep scan results in regional admixture peak")+
  xlab("Chromosome 6 positions")+
  ylab("iSAFE score")+
  scale_x_continuous(labels = function(x) format(x, scientific = FALSE))+
  #theme(plot.title = element_text(hjust = 0.5))
  theme(axis.text.x = element_text(size=13),
        axis.text.y = element_text(size=13),
        axis.title.x = element_text(size=15),
        axis.title.y = element_text(size=15))
ggsave("isafe.plot.chr6.1mb.png",width = 10, height = 7)
  

# 5Mb region region

isafe.res = fread("isafe.chr6.5mb.adrp.iSAFE.out")

isafe.res %>%
  ggplot(aes(x=POS,y=iSAFE))+
  geom_point(size=1, alpha=0.5)+
  theme_bw()+
  #ggtitle("Selective sweep scan results in regional admixture peak")+
  xlab("Chromosome 6 positions")+
  ylab("iSAFE score")+
  scale_x_continuous(labels = function(x) format(x, scientific = FALSE))+
  #theme(plot.title = element_text(hjust = 0.5))
  theme(axis.text.x = element_text(size=13),
        axis.text.y = element_text(size=13),
        axis.title.x = element_text(size=15),
        axis.title.y = element_text(size=15))
ggsave("isafe.plot.chr6.5mb.png",width = 10, height = 7)

## investigating the peak
isafe.res %>% arrange(desc(iSAFE)) %>% head
```












































# previous other codes

```{python convert rfmix output to plink file}

'''
Converts RFMix2 output tsv file (designed for fb) to plink tped,tfam files
Usage: python3 fbtoped.py [fb.tsv and output prefix] [pop_of_interest] [max threshold]
Next step to convert to binary: plink --tfile (prefix) --make-bed --out 
'''
__author__ = 'sjeon'
__year__ = '2021'
from sys import argv
thrshld=float(argv[3]) #probability under threshold is set to missing
pre=str(argv[1])
race=str(argv[2])

#make fam file
of=open(pre+'.'+race+'.tfam','w')
f=open(pre+'.fb.tsv')
pop=f.readline().split()[1:]
idx=pop.index(str(argv[2]))

h=f.readline().split()[4:] 
for i in range(0,int(len(h)),6):
	of.write(h[i].split(':')[0]+'\t'+h[i].split(':')[0]+'\t0\t0\t0\t0\n')	#FID/IID/Pat/Mat/Sex/Pheno
of.close()

#make genotype file
#plink can't treat half-missing call so set to missing if one haplotype is missing
o=open(pre+'.'+race+'.tped','w')
for lines in f:
	lines=lines.split()
	o.write(lines[0]+'\t')	#chr
	o.write(lines[0]+':'+lines[1]+'\t')	#variantID
	o.write(lines[2]+'\t')	#genetic_pos
	o.write(lines[1])	#phys_pos
	lines=lines[4:]
	for i in range(0,int(len(h)),6):
		o.write('\t')
		hap1=lines[i:i+3]	#probability set for this haplotype
		hap2=lines[i+3:i+6]
		ps1=[i > thrshld for i in [float(i) for i in hap1]]	#if any pass threshold
		ps2=[i > thrshld for i in [float(i) for i in hap2]]
		if True in ps1 and True in ps2:
			if ps1.index(True) == idx:
				o.write('1')
			else:
				o.write('2')
			o.write('\t')
			if ps2.index(True) == idx:
				o.write('1')
			else:
				o.write('2')
		else:
			o.write('0\t0')
	o.write('\n')
		
o.close()

```

```{bash previous other codes}

# Suggestions from 10/15 lab presentation 
## Better amerindian reference population: 1000G with higher amerindian ancestry
## Reclustering glioma subjects and redefining 'latino' subjects and white, black
## Regional regression on Amerindian ancestry
## Case-only regression?


############################################################################################################
# Choosing 1000G with higher amerindian ancestry

## Choosing 1000G with the following ancestry: black, white and amr
## Creating a black, white and amr 1000G subjects with RMD

cd /scratch/lishaobo/GWAS/glioma/populationRisk2

cp /scratch/lishaobo/GWAS/gwas_1000G/GRCh38/1kG_hg38_nors.??? .
cp /scratch/lishaobo/GWAS/gwas_1000G/GRCh38/integrated_call_samples_v3.20130502.ALL.panel .
cp /scratch/lishaobo/GWAS/glioma/populationRisk/glio.1kg.noamb.??? .
cp /scratch/lishaobo/GWAS/glioma/populationRisk/ancestryPropwithwhite.txt .


plink --bfile 1kG_hg38_nors \
  --keep refSubs.txt \
  --make-bed \
  --out 1kg.bwa

## Pruning data first to make the calculation faster

plink2 --bfile 1kg.bwa \
  --maf 0.1 \
  --indep-pairwise 50 10 0.1 \
  --out 1kg.bwa.prn
  
plink --bfile 1kg.bwa \
  --extract 1kg.bwa.prn.prune.in \
  --make-bed \
  --out 1kg.bwa.prnd

## Imputing proportions of ancestries with admixture
admixture 1kg.bwa.prnd.bed 3

## Plotting genetic proportions and choose AMR with the highest AMR component

############################################################################################################
## Reclustering glioma cases and controls for more accurately ancestry
# Use KNN to resassign ancestry
cd /scratch/lishaobo/GWAS/glioma/populationRisk2

### Extract subjects that are chosen as reference, and those who are in glioma
cp ../populationRisk/pol_adm.fam .
awk '{print $1,$2}' pol_adm.fam > glioSubs.txt
cat glioSubs.txt AFRpop.txt EURpop.txt AMRpop.txt > forMDSsubs.txt

plink --bfile glio.1kg.noamb \
  --keep forMDSsubs.txt \
  --make-bed \
  --out glio.1kg.mds

plink --bfile glio.1kg.mds --exclude ../codes/inversion_hg38.txt --range --indep-pairwise 50 5 0.2 --out 1kgglioindep
plink --bfile glio.1kg.mds --extract 1kgglioindep.prune.in --genome --out glio.1kg.mds.1
plink --bfile glio.1kg.mds --read-genome glio.1kg.mds.1.genome --cluster --mds-plot 10 --out glio.1kg.mds.1


############################################################################################################
##  Get ready for LAMP-LD
cd /scratch/lishaobo/GWAS/glioma/populationRisk2

# Break into each chromosome
for chr in {1..22}; do
  plink --bfile glio.1kg.mds \
    --allow-no-sex \
    --chr $chr \
    --make-bed \
    --out rgn/glio.1kg.mds.chr${chr}
done

### Prepare input file for LAIT.PL

#### A plink style ped and map file for LAT populations
## Required inputs are PLINK-formatted admixed .map and .ped file
## Reference or subset or reference and admixed SNPs
## Reference haplotypes for pops

### PLINK-formatted admixed .map and .ped file
for chr in {1..22}; do
  plink --bfile rgn/glio.1kg.mds.chr${chr} \
    --keep LAT_subs.txt \
    --recode \
    --out rgn/lat.chr${chr}.unphased
done      

### Reference or subset or reference and admixed SNPs
for chr in {1..22}; do
  awk '{print $2}' rgn/glio.1kg.mds.chr${chr}.bim > rgn/chr${chr}.lait.snps
done


## Reference haplotypes for pops
cd /scratch/lishaobo/GWAS/glioma/populationRisk2

for chr in {1..22}; do
  plink --bfile rgn/glio.1kg.mds.chr${chr} \
      --recode transpose \
      --keep-allele-order \
      --out rgn/chr"${chr}"

  echo -n "I ID" > rgn/beagleheader-file-"${chr}".txt
  cat rgn/chr"${chr}".tfam | \
     perl -ne 'chomp; @a=split(/\s+/,$_); print " $a[1] $a[1]"' >> rgn/beagleheader-file-"${chr}".txt
  echo "" >> rgn/beagleheader-file-"${chr}".txt

  cut -d " " -f1,2,5- rgn/chr"${chr}".tped > rgn/chr"${chr}"_temp
  sed -e "s/$chr/M/" rgn/chr"${chr}"_temp > rgn/chr"${chr}"_temp2
  cat rgn/beagleheader-file-"${chr}".txt rgn/chr"${chr}"_temp2 > rgn/chr"${chr}"_beagle.txt
done


for chr in {1..22}; do
  java -Xmx64G -jar /scratch/lishaobo/dependencies/beagle/beagle.jar \
    unphased=rgn/chr"${chr}"_beagle.txt \
    missing=0 \
    out=rgn/chr"${chr}"_phased_beagle
done
    
for chr in {1..22}; do
  gunzip rgn/chr"${chr}"_phased_beagle.chr"${chr}"_beagle.txt.phased.gz
  cut -d" " -f2- rgn/chr"${chr}"_phased_beagle.chr"${chr}"_beagle.txt.phased > rgn/chr"${chr}"_phased.txt
done

cd /scratch/lishaobo/GWAS/glioma/populationRisk2
for chr in {1..22}; do
  Rscript ../codes/beagle_up.R "rgn/chr${chr}_phased.txt" ${chr}
done

bin="/scratch/lishaobo/dependencies/lampld/LAMPLD-v1.0"
for chr in {1..22}; do
  perl $bin/lait.pl lamp-ld 3 \
    rgn/lat.chr${chr}.unphased.map \
    rgn/lat.chr${chr}.unphased.ped \
    rgn/chr${chr}.lait.snps \
    rgn/chr${chr}_eur_phased.hap \
    rgn/chr${chr}_afr_phased.hap \
    rgn/chr${chr}_amr_phased.hap \
    rgn/
  mv rgn/pop1.hap rgn/chr${chr}.eur.hap
  mv rgn/pop2.hap rgn/chr${chr}.afr.hap
  mv rgn/pop3.hap rgn/chr${chr}.amr.hap
  mv rgn/genofile.gen rgn/chr${chr}.lat.gen
  mv rgn/chr.pos rgn/chr${chr}.pos 
done  

################################################################
##LAMPLD analysis
cd /scratch/lishaobo/GWAS/glioma/populationRisk2

for chr in {1..22}; do
  /scratch/lishaobo/dependencies/lampld/LAMPLD-v1.0/bin/unolanc \
    200 \
    15 \
    rgn/chr${chr}.pos \
    rgn/chr${chr}.eur.hap\
    rgn/chr${chr}.amr.hap \
    rgn/chr${chr}.afr.hap \
    rgn/chr${chr}.lat.gen \
    rgn/chr${chr}.lat.lampld.out
done  

################################################################
## Downstream regression analysis
cd /scratch/lishaobo/GWAS/glioma/populationRisk2
bin="/scratch/lishaobo/dependencies/lampld/LAMPLD-v1.0"

for chr in {1..22}; do
  $bin/convertLAMPLDout.pl  rgn/chr${chr}.lat.lampld.out  rgn/lamp_dstm/chr${chr}.lat.lampld.long
done  

# For each individual, calculate the whole number of 1s in each SNP with r processing output files
# Prepare a covaraite data file, including Latino's caco status, sex and global european ancestry proportions with  in rmd
# Actual correlation analysis, in rmd

################################################################
## combining and plotting

cd /scratch/lishaobo/GWAS/glioma/populationRisk2

head -1 rgn/lamp_dstm/reg.chr1.out > lat.lampld.pol.long
tail -n +2 -q rgn/lamp_dstm/reg.chr*.out >> lat.lampld.pol.long
```

```{r Select white, black and amr subjects, for choosing the most proper AMR reference subjects}
library(tidyverse)
library(data.table)
setwd("/scratch/lishaobo/GWAS/glioma/populationRisk2")

source <- fread("/scratch/lishaobo/GWAS/glioma/codes/integrated_call_samples_v3.20130502.ALL.panel",fill = TRUE,header=TRUE) %>%
  select(1:3) %>%
  rename("IID" = "sample") %>%
  mutate(FID=0) %>%
  filter(super_pop %in% c("EUR", "AFR","AMR")) %>%
  rename(race = super_pop) %>%
  select(-pop) %>%
  select(FID,IID)

write.table(source,"refSubs.txt", col.names = FALSE, row.names = FALSE, quote=FALSE)
```

```{r Plotting genetic proportions and choose AMR with the highest AMR component}
# choose k=3 and plot self-reported latino and african americans and white
library(tidyverse)
library(data.table)
setwd("/scratch/lishaobo/GWAS/glioma/populationRisk2")

source <- fread("/scratch/lishaobo/GWAS/glioma/codes/integrated_call_samples_v3.20130502.ALL.panel",fill = TRUE,header=TRUE) %>%
  select(1:3) %>%
  rename("IID" = "sample") %>%
  mutate(FID=0) %>%
  filter(super_pop %in% c("EUR", "AFR","AMR")) %>%
  rename(race = super_pop) %>%
  select(-pop) %>%
  select(FID,IID,race)

subs <- read.table("1kg.bwa.prnd.fam")%>%
  select(V1,V2) %>%
  rename("FID"="V1","IID" = "V2") %>%
  left_join(source, by=c("FID","IID"))

tbl=read.table("1kg.bwa.prnd.3.Q")
tbl_cmb <- cbind(subs,tbl) %>% as.data.frame() %>%
  arrange(race) 

tbl_cmb_afr <- tbl_cmb %>% filter(race=="AFR") %>% arrange(desc(V2))
tbl_cmb_his <- tbl_cmb %>% filter(race=="AMR") %>% arrange(desc(V2))
tbl_cmb_wt <- tbl_cmb %>% filter(race=="EUR") %>% arrange(desc(V2))

tbl_cmb_updt <- rbind(tbl_cmb_afr,tbl_cmb_his,tbl_cmb_wt) %>%
  as.data.frame() %>%
  select(FID, IID, race, V1, V2, V3)

tiff("1000GgeneticComponents.tiff", width = 14, height = 7, units = 'in', res = 100)
barplot(t(as.matrix(tbl_cmb_updt[,-c(1:3)])), col=rainbow(3), xlab="Subjects", ylab="Ancestry", border=NA)
dev.off()

# From the figure, red is white (V1), blue is african (V3) and green is ameind (V2)
tbl_cmb_rnm <- tbl_cmb_updt %>%
  rename("white" = "V1",
         "ameind" = "V2",
         "black" = "V3")

# Choose AMRs with high proportion of amerindian ancestry
tbl_cmb_amr = tbl_cmb_rnm %>%
  filter(race=="AMR",ameind>0.5) %>%
  select(FID,IID) #115 subjects, choosing top 115 subjects from EUR and AFR too

tbl_cmb_eur = tbl_cmb_rnm %>%
  filter(race=="EUR") %>%
  arrange(desc(white)) %>%
  select(FID,IID)
tbl_cmb_eur = tbl_cmb_eur[1:115,]

tbl_cmb_afr = tbl_cmb_rnm %>%
  filter(race=="AFR") %>%
  arrange(desc(black)) %>%
  select(FID,IID)
tbl_cmb_afr = tbl_cmb_afr[1:115,]

write.table(tbl_cmb_rnm,"1000GAncestryProps.txt", row.names = FALSE, quote=FALSE)
write.table(tbl_cmb_amr,"AMRpop.txt", col.names = FALSE, row.names = FALSE, quote=FALSE)
write.table(tbl_cmb_eur,"EURpop.txt", col.names = FALSE, row.names = FALSE, quote=FALSE)
write.table(tbl_cmb_afr,"AFRpop.txt", col.names = FALSE, row.names = FALSE, quote=FALSE)
```

```{r Reclassify glioma subjects and determine their ancestry}
library(tidyverse)
library(data.table)
library(class)

setwd("/scratch/lishaobo/GWAS/glioma/populationRisk2")

dfOri = read.table("ancestryPropwithwhite.txt",header=TRUE) %>%
  arrange(desc(ameind))
tiff("gliosubsArrangebyAmeriRedisWhite.tiff", width = 14, height = 7, units = 'in', res = 100)
barplot(t(as.matrix(dfOri[,-c(1:4)])), col=rainbow(3), xlab="Subjects", ylab="Ancestry", border=NA)
dev.off()

# Use KNN to resassign ancestry
data<- read.table(file="glio.1kg.mds.1.mds",header=TRUE)%>%
    mutate(FID=as.character(FID))


sourceP1 <- fread("1000GAncestryProps.txt",header=TRUE) %>%
  select(FID,IID,race)%>%
  mutate(FID=as.character(FID))
sourceP2 <- read.csv("/scratch/lishaobo/GWAS/glioma/raw_data/all_6512subs_variable.csv") %>%
  select(vcfID) %>%
  mutate(FID=0,race="glio")%>%
  mutate(FID=as.character(FID))%>%
  rename("IID" = "vcfID") %>%
  select(FID,IID,race)

sourceRace = rbind(sourceP1,sourceP2) %>% as.data.frame() %>%
  inner_join(data,by=c("FID","IID")) %>%
  mutate(race=as.character(race))

shape_vector <- c(20, 20, 20,  4)
color_vector <- c("green1","mediumblue","hotpink","grey67")

tiff("assignGlioMDS1_2.tiff",units  = "cm", width = 30, height = 15, res=300)
sourceRace%>%
  ggplot(aes(C1, C2, group=race)) + 
  geom_point(aes(shape=race, color=race), size=2)+
  scale_shape_manual(values=shape_vector)+
  scale_color_manual(values=color_vector)+
  theme_bw()
dev.off()

tiff("assignGlioMDS1_3.tiff",units  = "cm", width = 30, height = 15, res=300)
sourceRace%>%
  ggplot(aes(C1, C3, group=race)) + 
  geom_point(aes(shape=race, color=race), size=2)+
  scale_shape_manual(values=shape_vector)+
  scale_color_manual(values=color_vector)+
  theme_bw()
dev.off()

tiff("assignGlioMDS1_4.tiff",units  = "cm", width = 30, height = 15, res=300)
sourceRace%>%
  ggplot(aes(C1, C4, group=race)) + 
  geom_point(aes(shape=race, color=race), size=2)+
  scale_shape_manual(values=shape_vector)+
  scale_color_manual(values=color_vector)+
  theme_bw()
dev.off()

tiff("assignGlioMDS1_5.tiff",units  = "cm", width = 30, height = 15, res=300)
sourceRace%>%
  ggplot(aes(C1, C5, group=race)) + 
  geom_point(aes(shape=race, color=race), size=2)+
  scale_shape_manual(values=shape_vector)+
  scale_color_manual(values=color_vector)+
  theme_bw()
dev.off()

#impute race information

own_not_ind <- which(!sourceRace$race == "glio")
own_ind <- which(sourceRace$race == "glio")

train <-sourceRace[own_not_ind, c("C1","C2","C3","C4","C5","C6","C7","C8","C9","C10")]
test <-sourceRace[own_ind, c("C1","C2","C3","C4","C5","C6","C7","C8","C9","C10")]
train_results <- as.character(sourceRace[own_not_ind, "race"])
test_results <- as.character(knn(train=train, test=test, cl=train_results, k=10))
test_results_1 <- paste0(test_results,"_imp")
  
datafile_imputed <- sourceRace %>% mutate(race= as.character(race))
datafile_imputed[own_ind,"race"] <- test_results_1 

shape_vector <- c(20, 2, 20, 2, 20, 2)
color_vector <- c("green1","green1","mediumblue","mediumblue","hotpink","hotpink")

tiff("GlioMDS1_2_imputed.tiff",units  = "cm", width = 30, height = 15, res=300)
datafile_imputed%>%
 ggplot(aes(C1, C2, group=race)) + 
 geom_point(aes(shape=race, color=race), size=2)+
 scale_shape_manual(values=shape_vector)+
 scale_color_manual(values=color_vector)+
 theme_bw()
dev.off()

tiff("GlioMDS1_3_imputed.tiff",units  = "cm", width = 30, height = 15, res=300)
datafile_imputed%>%
  ggplot(aes(C1, C3, group=race)) + 
  geom_point(aes(shape=race, color=race), size=2)+
  scale_shape_manual(values=shape_vector)+
  scale_color_manual(values=color_vector)+
  theme_bw()
dev.off()

tiff("GlioMDS1_4_imputed.tiff",units  = "cm", width = 30, height = 15, res=300)
datafile_imputed%>%
  ggplot(aes(C1, C4, group=race)) + 
  geom_point(aes(shape=race, color=race), size=2)+
  scale_shape_manual(values=shape_vector)+
  scale_color_manual(values=color_vector)+
  theme_bw()
dev.off()

tiff("GlioMDS1_5_imputed.tiff",units  = "cm", width = 30, height = 15, res=300)
datafile_imputed%>%
  ggplot(aes(C1, C5, group=race)) + 
  geom_point(aes(shape=race, color=race), size=2)+
  scale_shape_manual(values=shape_vector)+
  scale_color_manual(values=color_vector)+
  theme_bw()
dev.off()

# Checking concordance between self-reported and actual race
all_samples_variables <- read.csv("/scratch/lishaobo/GWAS/glioma/raw_data/all_6512subs_variable.csv",row.names = 1) %>%
  select(FID, vcfID, race) %>%
  dplyr::rename("reported_race"="race", "IID"="vcfID") %>%
  mutate(FID = as.character(FID))
datafile_imputed_2 <- inner_join(all_samples_variables,datafile_imputed,by=c("FID","IID")) %>%
  select(FID,IID,reported_race,race) %>%
  mutate(reported_race=as.character(reported_race),
         race=as.character(race))
table(datafile_imputed_2$reported_race,datafile_imputed_2$race)
  #          AFR_imp AMR_imp EUR_imp
  # Black        188       2      30
  # Hispanic       6     895     223
  # White          2      42    1547


write.table(datafile_imputed_2, "race_assign_subs.txt", row.names = FALSE, quote=FALSE)

datafile_imputed_lat <- datafile_imputed_2 %>%
  dplyr::filter(race=="AMR_imp")%>%
  dplyr::select(FID, IID) %>%
  mutate(FID = as.character(FID))
write.table(datafile_imputed_lat, "LAT_subs.txt", col.names = FALSE, row.names = FALSE, quote=FALSE)
```

```{r Plotting ancestry proportions again to see if re-assigning improves ancestry proportion}
library(tidyverse)
library(data.table)
library(class)

setwd("/scratch/lishaobo/GWAS/glioma/populationRisk2")

dfOri = read.table("ancestryPropwithwhite.txt",header=TRUE) 

neo_race = read.table("race_assign_subs.txt", header=TRUE) %>%
  mutate(raceImp = gsub("_imp","",race)) %>%
  select(FID,IID,raceImp) %>%
  inner_join(dfOri, by=c("FID","IID"))
  

neo_race_1 = neo_race %>%
  arrange(race,ameind)

tiff("gliosubsASIS.tiff", width = 14, height = 7, units = 'in', res = 100)
barplot(t(as.matrix(neo_race_1[,-c(1:5)])), col=rainbow(3), xlab="Subjects", ylab="Ancestry", border=NA)
dev.off()

neo_race_2 = neo_race %>%
  arrange(raceImp,ameind)

tiff("gliosubsASIMP.tiff", width = 14, height = 7, units = 'in', res = 100)
barplot(t(as.matrix(neo_race_2[,-c(1:5)])), col=rainbow(3), xlab="Subjects", ylab="Ancestry", border=NA)
dev.off()
```

```{r comparison of ameIndian ancestries in all subjects}
library(tidyverse)
setwd("/scratch/lishaobo/GWAS/glioma/populationRisk2")

tbl_cmb_rnm <- read.table("../populationRisk/ancestryPropwithwhite.txt", header=TRUE) %>%
  select(-race)

neo_race <- read.table("race_assign_subs.txt", header=TRUE) %>%
  mutate(race = gsub("_imp","",race)) %>%
  right_join(tbl_cmb_rnm,by=c("FID","IID")) %>%
  select(-reported_race)

blk <- neo_race %>% filter(race == "AFR") %>%
  select(ameind,CaCo) %>%
  mutate(CaCo = ifelse(CaCo=="Control",0,1))
his <- neo_race %>% filter(race == "AMR")%>%
  select(ameind,CaCo) %>%
  mutate(CaCo = ifelse(CaCo=="Control",0,1))
wt <- neo_race %>% filter(race == "EUR")%>%
  select(ameind,CaCo) %>%
  mutate(CaCo = ifelse(CaCo=="Control",0,1))


# Testing the distribution of ameind ancestries in black subjects case/control
median(blk[blk$CaCo==0,1])
# 0.013869
median(blk[blk$CaCo==1,1])
# 0.014739
ggplot(blk, aes(x=as.factor(CaCo), y=ameind, fill=as.factor(CaCo))) +
  geom_boxplot(outlier.size=0.2)+
  theme_classic() +
  ggtitle( "Distribution of amerindian ancestry proportions in reassigned Black subjects" )+
  xlab("Case control status") +
  ylab("amerindian Proportions") +
  theme(plot.title = element_text(hjust = 0.5)) +
  scale_fill_discrete(name="disease status",labels=c("0","1"))
ggsave("blckamerindianProportions3rd.png")
wilcox.test(blk[blk$CaCo==0,1],blk[blk$CaCo==1,1])
# 	Wilcoxon rank sum test with continuity correction
# 
# data:  blk[blk$CaCo == 0, 1] and blk[blk$CaCo == 1, 1]
# W = 3597.5, p-value = 0.7752
# alternative hypothesis: true location shift is not equal to 0

# Testing the distribution of ameind ancestries in hispanic subjects case/control
ggplot(his, aes(x=as.factor(CaCo), y=ameind, fill=as.factor(CaCo))) +
  geom_boxplot(outlier.size=0.2)+
  theme_classic() +
  ggtitle( "Distribution of amerindian ancestry proportions in reassigned Hispanic subjects" )+
  xlab("Case control status") +
  ylab("amerindian Proportions") +
  theme(plot.title = element_text(hjust = 0.5)) +
  scale_fill_discrete(name="disease status",labels=c("0","1"))
ggsave("hisamerindianProportions3rd.png")

median(his[his$CaCo==0,1])
# 0.5434975
median(his[his$CaCo==1,1])
#  0.498127
wilcox.test(his[his$CaCo==0,1],his[his$CaCo==1,1])
# 	Wilcoxon rank sum test with continuity correction
# 
# data:  his[his$CaCo == 0, 1] and his[his$CaCo == 1, 1]
# W = 94042, p-value = 0.006215
# alternative hypothesis: true location shift is not equal to 0


# Testing the distribution of ameind ancestries in white subjects case/control
ggplot(wt, aes(x=as.factor(CaCo), y=ameind, fill=as.factor(CaCo))) +
  geom_boxplot(outlier.size=0.2)+
  theme_classic() +
  ggtitle( "Distribution of amerindian ancestry proportions in reassigned White subjects" )+
  xlab("Case control status") +
  ylab("amerindian Proportions") +
  theme(plot.title = element_text(hjust = 0.5)) +
  scale_fill_discrete(name="disease status",labels=c("0","1"))
ggsave("WTamerindianProportions3rd.png")

median(wt[wt$CaCo==0,1])
# 0.018178
median(wt[wt$CaCo==1,1])
# 0.017489
wilcox.test(wt[wt$CaCo==0,1],wt[wt$CaCo==1,1])
#	Wilcoxon rank sum test with continuity correction
#
#data:  wt[wt$CaCo == 0, 1] and wt[wt$CaCo == 1, 1]
#W = 347842, p-value = 0.3482
#alternative hypothesis: true location shift is not equal to 0
```

```{r comparison of white ancestries in all subjects}
library(tidyverse)
setwd("/scratch/lishaobo/GWAS/glioma/populationRisk2")

tbl_cmb_rnm <- read.table("../populationRisk/ancestryPropwithwhite.txt", header=TRUE) %>%
  select(-race)

neo_race <- read.table("race_assign_subs.txt", header=TRUE) %>%
  mutate(race = gsub("_imp","",race)) %>%
  right_join(tbl_cmb_rnm,by=c("FID","IID")) %>%
  select(-reported_race)

blk <- neo_race %>% filter(race == "AFR") %>%
  select(white,CaCo) %>%
  mutate(CaCo = ifelse(CaCo=="Control",0,1))
his <- neo_race %>% filter(race == "AMR")%>%
  select(white,CaCo) %>%
  mutate(CaCo = ifelse(CaCo=="Control",0,1))
wt <- neo_race %>% filter(race == "EUR")%>%
  select(white,CaCo) %>%
  mutate(CaCo = ifelse(CaCo=="Control",0,1))


# Testing the distribution of white ancestries in black subjects case/control
median(blk[blk$CaCo==0,1])
# 0.145016
median(blk[blk$CaCo==1,1])
# 0.146047
ggplot(blk, aes(x=as.factor(CaCo), y=white, fill=as.factor(CaCo))) +
  geom_boxplot(outlier.size=0.2)+
  theme_classic() +
  ggtitle( "Distribution of white ancestry proportions in reassigned Black subjects" )+
  xlab("Case control status") +
  ylab("White Proportions") +
  theme(plot.title = element_text(hjust = 0.5)) +
  scale_fill_discrete(name="disease status",labels=c("0","1"))
ggsave("blckWhiteProportions3rd.png")

wilcox.test(blk[blk$CaCo==0,1],blk[blk$CaCo==1,1])
# 	Wilcoxon rank sum test with continuity correction
# 
# data:  blk[blk$CaCo == 0, 1] and blk[blk$CaCo == 1, 1]
# W = 3858, p-value = 0.6461
# alternative hypothesis: true location shift is not equal to 0

# Testing the distribution of white ancestries in hispanic subjects case/control
ggplot(his, aes(x=as.factor(CaCo), y=white, fill=as.factor(CaCo))) +
  geom_boxplot(outlier.size=0.2)+
  theme_classic() +
  ggtitle( "Distribution of white ancestry proportions in reassigned Hispanic subjects" )+
  xlab("Case control status") +
  ylab("White Proportions") +
  theme(plot.title = element_text(hjust = 0.5)) +
  scale_fill_discrete(name="disease status",labels=c("0","1"))
ggsave("hisWhiteProportions3rd.png")

median(his[his$CaCo==0,1])
#  0.3861135
median(his[his$CaCo==1,1])
# 0.415611
wilcox.test(his[his$CaCo==0,1],his[his$CaCo==1,1])
# 	Wilcoxon rank sum test with continuity correction
# 
# data:  his[his$CaCo == 0, 1] and his[his$CaCo == 1, 1]
# W = 75825, p-value = 0.02249
# alternative hypothesis: true location shift is not equal to 0

# Testing the distribution of white ancestries in white subjects case/control
ggplot(wt, aes(x=as.factor(CaCo), y=white, fill=as.factor(CaCo))) +
  geom_boxplot(outlier.size=0.2)+
  theme_classic() +
  ggtitle( "Distribution of white ancestry proportions in reassigned White subjects" )+
  xlab("Case control status") +
  ylab("White Proportions") +
  theme(plot.title = element_text(hjust = 0.5)) +
  scale_fill_discrete(name="disease status",labels=c("0","1"))
ggsave("WTWhiteProportions3rd.png")

median(wt[wt$CaCo==0,1])
# 0.962527
median(wt[wt$CaCo==1,1])
#  0.964242
wilcox.test(wt[wt$CaCo==0,1],wt[wt$CaCo==1,1])
# 	Wilcoxon rank sum test with continuity correction
# 
# data:  wt[wt$CaCo == 0, 1] and wt[wt$CaCo == 1, 1]
# W = 330142, p-value = 0.4133
# alternative hypothesis: true location shift is not equal to 0
```

module load gcc/8.3.0
module load openblas/0.3.8
module load r/4.0.0
module load r

```{r processing output files}
library(tidyverse)
library(data.table)
setwd("/scratch/lishaobo/GWAS/glioma/populationRisk2")

for(chr in 1:22){
  a <- fread(paste0("rgn/lamp_dstm/chr",chr,".lat.lampld.long"),header = FALSE) 
  
  print(nrow(a))

  
  pos <-  fread(paste0("rgn/chr",chr,".pos")) %>% as.data.frame() %>%
    mutate(V2 = paste0("chr",chr,":",V1))
  
  
  b <- apply(a,1,function(x)str_split(x,pattern=""))
  b1 <- data.frame(matrix(unlist(b), nrow=length(b), byrow=T))
  b2_1ornot <- b1==0 
  tem_sum_all <-list()
  for(i in 1:(nrow(b2_1ornot)/2)){
    tem_a <- b2_1ornot[c(2*i-1,2*i),]
    tem_sum <- colSums(tem_a)
    tem_sum_all[[i]] <- tem_sum
  }
  b2_numP1 <- data.frame(do.call(rbind, tem_sum_all)) %>% t()
  rownames(b2_numP1) <- pos$V2
  
  print(nrow(b2_numP1))
  
  write.table(b2_numP1,paste0("rgn/lamp_dstm/chr",chr,"_eurCount.txt"), row.names=TRUE,col.names = FALSE, quote=FALSE)
}
```

```{r covariate file preparation}
library(tidyverse)
library(data.table)
setwd("/scratch/lishaobo/GWAS/glioma/populationRisk2")

subs <- fread("LAT_subs.txt")
colnames(subs) <- c("FID","IID")
tbl_cmb_rnm <- read.table("ancestryPropwithwhite.txt", header=TRUE) %>% 
  rename(reportRace=race) %>%
  right_join(subs,by=c("FID","IID"))
  
source1 <- read.csv("/scratch/lishaobo/GWAS/glioma/raw_data/all_samples_variables.csv") %>%
  select(FID, vcfID, sex_raw, phenotype_raw) %>%
  rename(IID = vcfID)
source2 <- read.csv("/scratch/lishaobo/GWAS/glioma/output_files/self_reported_and_imputed_race.csv")  %>%
  select(FID, IID, race) %>%
  mutate(race = gsub("_imp","",race),
         race=ifelse(race=="AMR","LAT",race)) %>%
 rename(imputeRace=race) %>%
 inner_join(source1, by=c("FID","IID")) %>%
  mutate(source="glio")

tbl_lat_var = tbl_cmb_rnm %>% 
  left_join(source2,by=c("FID","IID")) %>%
  select(FID, IID, white,sex_raw,phenotype_raw)
  
write.table(tbl_lat_var, "rgn/lamp_dstm/lat.pheono.txt", row.names = FALSE, quote=FALSE)

```

```{r correlation analysis}
library(tidyverse)
library(data.table)
setwd("/scratch/lishaobo/GWAS/glioma/populationRisk2")

ids = fread("LAT_subs.txt")
colnames(ids) = c("FID","IID")

for(chr in 1:22){

  eur = fread(paste0("rgn/lamp_dstm/chr",chr,"_eurCount.txt"),header = FALSE)
  colnames(eur) = c("SNP",ids$IID)
  
  eur1 = eur %>%
    column_to_rownames(var="SNP") %>% 
    t() %>%
    as.data.frame()
  
  pheno = fread("rgn/lamp_dstm/lat.pheono.txt") %>%
    right_join(ids,by=c("FID","IID")) %>%
    mutate(caco=as.factor(ifelse(phenotype_raw==1,0,1))) %>%
    mutate(sex=as.factor(sex_raw))
  
  regressors = names(eur1)
  reg_out <- lapply(regressors,function(snp){c(data.frame(SNP_ID=snp),summary(
    glm(pheno$caco ~ eur1[,snp]+ pheno$sex + pheno$white, family="binomial"))$coefficients[2,1:4])
  })
  reg_df <- bind_rows(reg_out) %>% as.data.frame()
  
  write.table(reg_df, paste0("rgn/lamp_dstm/reg.chr",chr,".out"), row.names = FALSE, quote=FALSE)

}

```

# Combine output with the following code:
cd /scratch/lishaobo/GWAS/glioma/populationRisk2
head -1 rgn/lamp_dstm/reg.chr1.out > lat.lampld.pol.long
tail -n +2 -q rgn/lamp_dstm/reg.chr?.out >> lat.lampld.pol.long
tail -n +2 -q rgn/lamp_dstm/reg.chr??.out >> lat.lampld.pol.long

```{r visualization of local ancestry, Manhattan plot}
library(tidyverse)
library(QCEWAS)
library(qqman)
library(data.table)

setwd("/scratch/lishaobo/GWAS/glioma/populationRisk2")

filename = "lat.lampld.pol"
  
gwas_data <- fread("lat.lampld.pol.long",skip=1,header=FALSE)
colnames(gwas_data) = c("ID", "Estimate","StdError","Zvalue","P")

gwas_data_05 = gwas_data %>%
  mutate(SNPID = ID) %>%
  separate(SNPID, c("CHROM","POS"))%>%
  mutate(CHROM = as.numeric(gsub("chr","",CHROM))) %>%
  mutate(POS = as.numeric(POS)) %>%
  arrange(CHROM,POS)

gwas_data_1 <- gwas_data_05[is.finite(gwas_data_05$P),  ]

tiff(paste0(filename,".manhattan.tiff"), width = 10, height = 7, units = 'in', res = 300)
manhattan(gwas_data_1, chr = "CHROM", bp = "POS", p = "P", snp = "ID")
dev.off()
```

# Regression on Amerindian ancestry and see what happens

```{r processing output files}
library(tidyverse)
library(data.table)
setwd("/scratch/lishaobo/GWAS/glioma/populationRisk2")

for(chr in 1:22){
  a <- fread(paste0("rgn/lamp_dstm/chr",chr,".lat.lampld.long"),header = FALSE) 
  
  print(nrow(a))

  
  pos <-  fread(paste0("rgn/chr",chr,".pos")) %>% as.data.frame() %>%
    mutate(V2 = paste0("chr",chr,":",V1))
  
  
  b <- apply(a,1,function(x)str_split(x,pattern=""))
  b1 <- data.frame(matrix(unlist(b), nrow=length(b), byrow=T))
  b2_1ornot <- b1==2 
  tem_sum_all <-list()
  for(i in 1:(nrow(b2_1ornot)/2)){
    tem_a <- b2_1ornot[c(2*i-1,2*i),]
    tem_sum <- colSums(tem_a)
    tem_sum_all[[i]] <- tem_sum
  }
  b2_numP1 <- data.frame(do.call(rbind, tem_sum_all)) %>% t()
  rownames(b2_numP1) <- pos$V2
  
  print(nrow(b2_numP1))
  
  write.table(b2_numP1,paste0("rgn/lamp_dstm/chr",chr,"_amrCount.txt"), row.names=TRUE,col.names = FALSE, quote=FALSE)
}
```

```{r covariate file preparation}
library(tidyverse)
library(data.table)
setwd("/scratch/lishaobo/GWAS/glioma/populationRisk2")

subs <- fread("LAT_subs.txt")
colnames(subs) <- c("FID","IID")
tbl_cmb_rnm <- read.table("ancestryPropwithwhite.txt", header=TRUE) %>% 
  rename(reportRace=race) %>%
  right_join(subs,by=c("FID","IID"))
  
source1 <- read.csv("/scratch/lishaobo/GWAS/glioma/raw_data/all_samples_variables.csv") %>%
  select(FID, vcfID, sex_raw, phenotype_raw) %>%
  rename(IID = vcfID)
source2 <- read.csv("/scratch/lishaobo/GWAS/glioma/output_files/self_reported_and_imputed_race.csv")  %>%
  select(FID, IID, race) %>%
  mutate(race = gsub("_imp","",race),
         race=ifelse(race=="AMR","LAT",race)) %>%
 rename(imputeRace=race) %>%
 inner_join(source1, by=c("FID","IID")) %>%
  mutate(source="glio")

tbl_lat_var = tbl_cmb_rnm %>% 
  left_join(source2,by=c("FID","IID")) %>%
  select(FID, IID, ameind,sex_raw,phenotype_raw)
  
write.table(tbl_lat_var, "rgn/lamp_dstm/lat.amr.pheono.txt", row.names = FALSE, quote=FALSE)

```

```{r correlation analysis}
library(tidyverse)
library(data.table)
setwd("/scratch/lishaobo/GWAS/glioma/populationRisk2")

ids = fread("LAT_subs.txt")
colnames(ids) = c("FID","IID")

for(chr in 1:22){

  amr = fread(paste0("rgn/lamp_dstm/chr",chr,"_amrCount.txt"),header = FALSE)
  colnames(amr) = c("SNP",ids$IID)
  
  amr1 = amr %>%
    column_to_rownames(var="SNP") %>% 
    t() %>%
    as.data.frame()
  
  pheno = fread("rgn/lamp_dstm/lat.amr.pheono.txt") %>%
    right_join(ids,by=c("FID","IID")) %>%
    mutate(caco=as.factor(ifelse(phenotype_raw==1,0,1))) %>%
    mutate(sex=as.factor(sex_raw))
  
  regressors = names(amr1)
  reg_out <- lapply(regressors,function(snp){c(data.frame(SNP_ID=snp),summary(
    glm(pheno$caco ~ amr1[,snp]+ pheno$sex + pheno$ameind, family="binomial"))$coefficients[2,1:4])
  })
  reg_df <- bind_rows(reg_out) %>% as.data.frame()
  
  write.table(reg_df, paste0("rgn/lamp_dstm/reg.amr.chr",chr,".out"), row.names = FALSE, quote=FALSE)

}

```

# Combine output with the following code:
cd /scratch/lishaobo/GWAS/glioma/populationRisk2
head -1 rgn/lamp_dstm/reg.amr.chr1.out > lat.amr.lampld.pol.long
tail -n +2 -q rgn/lamp_dstm/reg.amr.chr?.out >> lat.amr.lampld.pol.long
tail -n +2 -q rgn/lamp_dstm/reg.amr.chr??.out >> lat.amr.lampld.pol.long

```{r visualization of local ancestry, Manhattan plot}
library(tidyverse)
library(QCEWAS)
library(qqman)
library(data.table)

setwd("/scratch/lishaobo/GWAS/glioma/populationRisk2")

filename = "lat.amr.lampld.pol"
  
gwas_data <- fread("lat.amr.lampld.pol.long",skip=1,header=FALSE)
colnames(gwas_data) = c("ID", "Estimate","StdError","Zvalue","P")

gwas_data_05 = gwas_data %>%
  mutate(SNPID = ID) %>%
  separate(SNPID, c("CHROM","POS"))%>%
  mutate(CHROM = as.numeric(gsub("chr","",CHROM))) %>%
  mutate(POS = as.numeric(POS)) %>%
  arrange(CHROM,POS)

gwas_data_1 <- gwas_data_05[is.finite(gwas_data_05$P),  ]

tiff(paste0(filename,".manhattan.tiff"), width = 10, height = 7, units = 'in', res = 300)
manhattan(gwas_data_1, chr = "CHROM", bp = "POS", p = "P", snp = "ID")
dev.off()
```

