---
editor_options: 
  markdown: 
    wrap: 72
---

Project description: A multi-cohort meta-analysis (US, Denmark, Sweden) of childhood glioma GWAS to identify risk alleles.

This paper is in submission, me as a co-first author.


Step 1: choosing subjects to include in the analysis

```{bash New samples QC}

cd /scratch/lishaobo/GWAS/glioma/saxb4
#cd /project/wiemels_494/sebastian/glioma_gwas

# cp /project/wiemels_494/publicRaws/PRO100497_USC497_SAX_b04/PLINK.zip .
## Manipulate text for plink to work

#awk ' !( ($1 >=1  && $1 <=22) || $1 ~/^\#/ || $1  ~/^X/ || $1  ~/^Y/ ) {next} 1'  /project/wiemels_494/publicRaws/PRO100497_USC497_SAX_b04/AxiomGT1.calls.vcf > AxiomGT1.calls.1.vcf
#grep ^X  AxiomGT1.calls.1.vcf | head
#grep ^X /project/wiemels_494/publicRaws/PRO100497_USC497_SAX_b04/AxiomGT1.calls.vcf | head

#  Convert to plink bfile
# plink --vcf AxiomGT1.calls.1.vcf \
#   --biallelic-only \
#   --snps-only 'just-acgt' \
#   --make-bed \
#   --out saxb4
# 809367 variants and 6948 people pass filters and QC.

plink --vcf /project/wiemels_494/publicRaws/PRO100497_USC497_SAX_b04/AxiomGT1.calls.vcf \
  --allow-extra-chr \
  --snps-only 'just-acgt' \
  --chr-set 35 \
  --biallelic-only \
  --chr 1-26 \
  --make-bed \
  --out saxb4
# 825111 variants and 6948 samples pass filters and QC.

# cp /scratch/lishaobo/GWAS/glioma/saxb4/saxb4.??? .

```

```{r pick subjects to add and subjects to swap}
#setwd("/scratch/lishaobo/GWAS/glioma/saxb4")
setwd("/project/wiemels_494/sebastian/glioma_gwas")
library(tidyverse)
library(data.table)
# load subjects from the 2st batch of sequencing
# batch2 = fread("saxb4.fam")%>%
#   dplyr::rename("FID"="V1", "IID"="V2")%>%
#   mutate(IID = paste0(FID,"_",IID), FID=0)
# write.table(batch2,"saxb4.fam", sep = "\t", row.names = FALSE, col.names=FALSE,quote = FALSE)
batch2 = fread("saxb4.fam")%>%
  dplyr::rename("FID"="V1", "IID"="V2")%>%
  select(FID,IID) 
# load all subjects
glio2 = fread("/project/wiemels_260/sebastian/gwas_glioma/raw_data/B2SampleAttributes.txt") %>% 
  as.data.frame() %>%
  select(`Sample Filename`, SampleName,SubmittedPlate)%>%
  mutate(FID=0, SampleName=as.integer(SampleName))%>%
  na.omit()%>% #delelte duplicates and inner controls
  dplyr::rename("IID"="Sample Filename", "subjectId" = "SampleName") %>%
  inner_join(batch2, by=c("FID","IID"))

# load twins
pgs2=fread("/project/wiemels_260/sebastian/gwas_glioma/raw_data/bacth2twins.txt")
glio2_1 = glio2 %>%
  filter(!subjectId %in% pgs2$V1) # delete twins

# identify repeated samples from first batch
dups = which(table(glio2_1$subjectId)>1) %>% names
badIID = glio2_1 %>%
  filter(subjectId %in% dups) %>%
  filter(!SubmittedPlate %in% c("9SR31606","9SR31607","9SR31608") ) %>%
  select(IID)
         
# delete repeated samples from the first batch
glio2_2 = glio2_1 %>%
  filter(!IID %in% badIID$IID)

# Prepare a file for plink --keep
glioPlinkKeep = glio2_2 %>% select(FID,IID)
#write.table(glioPlinkKeep,"includeB2.txt", sep = "\t", row.names = FALSE, col.names=FALSE,quote = FALSE)

```

6948 subjects in plink files 5 duplicates 153 affymetrix inner controls
172 twins and 2 duplicate twins (526 and 630) 33 repeated samples from
batch 1 6583 subjects in the end

```{bash Data preparation and missingness}

cd /scratch/lishaobo/GWAS/glioma/saxb4
#cd /project/wiemels_494/sebastian/glioma_gwas

plink --bfile /scratch/lishaobo/GWAS/glioma/saxb4/saxb4\
  --keep includeB2.txt \
  --make-bed\
  --out saxb4_incl
  
#825111 variants and 6583 people pass filters and QC.
```

```{r fill in the info for plink fam files}
setwd("/scratch/lishaobo/GWAS/glioma/saxb4")
#setwd("/project/wiemels_494/sebastian/glioma_gwas")
library(tidyverse)
library(data.table)

glio <- read.csv("/project/wiemels_260/sebastian/gwas_glioma/raw_data/likereallyraw/PedGlioma_4869_Full_Covariates_uncript.csv") %>% 
  as.data.frame() %>%
  mutate(sex = ifelse(Y_male==0,2,1)) %>%
  mutate(phenotype = ifelse(status=="case",2,1))%>%
  dplyr::rename("race"="Y_race", "subjectId" = "blindid") %>%
  mutate(source = "glio")
medulla <- read.csv("/project/wiemels_260/sebastian/gwas_glioma/raw_data/likereallyraw/medulla.csv") %>%
  as.data.frame() %>%
  mutate(phenotype = ifelse(CaCo == "Case", 2,1),
         sex = ifelse(SEX == "M", 1, 2) ) %>%
  mutate(subjectId = blindid)%>%#*10) %>% #sequencing machine adds a 0 in their fam IDs
  dplyr::rename( "race" = "RACE08") %>%
  mutate(source = "medulla")

glio_fam <- glio %>%
  select(subjectId,sex,phenotype,source,YRBIRTH,race,HISTO_T3)
Y_race <- sapply(glio_fam$race, function(x){
  if(x == 1){return("White")}
  if(x == 2){return("Black")}
  if(x == 3){return("Hispanic")}
  if(x == 4){return("Asian")}
  if(x == 5){return("Mixed")}
})
glio_fam$race=Y_race
  
medulla_fam <- medulla %>%
  select(subjectId,sex,phenotype,source,YRBIRTH,race) %>%
  mutate(HISTO_T3=NA)

glio_medulla = rbind(glio_fam,medulla_fam) %>% as.data.frame() %>%
  distinct(subjectId,.keep_all = TRUE)

glio2Fam = fread("saxb4_incl.fam") %>%
  dplyr::rename("FID"="V1", "IID"="V2")%>%  
  left_join(glio2_2, by=c("FID","IID")) %>% # glio2_2 is from the previous R bloc
#which(!glio2Fam$subjectId %in% glio_medulla$subjectId) %>% length()
  left_join(glio_medulla,by="subjectId")
glio2Fam_plink = glio2Fam %>%
  mutate(faID=0, moID=0) %>%
  select(FID,IID,faID,moID,sex,phenotype)

#backup original fam file
#glio2FamOri = fread("saxb4_incl.fam")
#write.table(glio2FamOri,"saxb4_incl.fam.backup", sep = "\t", row.names = FALSE, col.names=FALSE,quote = FALSE)
#write.table(glio2Fam,"/project/wiemels_260/sebastian/gwas_glioma/saxb4/rawSampleAttributes.txt", sep = "\t", row.names = FALSE, col.names=TRUE,quote = FALSE)
#write.table(glio2Fam_plink,"saxb4_incl.fam", sep = "\t", row.names = FALSE, col.names=FALSE,quote = FALSE)

```

Step 2: annotate different sub-groups for meta-analysis subtypes to
include

all glioma LGG (WHO I-II) HGG (WHO III-IV) LG astrocytoma (WHO I-II) HG
astrocytoma (WHO III-IV) PA PA Optic pathway Glioblastoma

```{r prepare subjects by subtypes for meta-analysis scheme}

#setwd("/scratch/lishaobo/GWAS/glioma/saxb4")
setwd("/project/wiemels_494/sebastian/glioma_gwas")

library(tidyverse)
library(data.table)

# reassigning some WHO grades based on Jon's suggestions 
glioS.who <- read.csv("/project/wiemels_260/sebastian/gwas_glioma/raw_data/likereallyraw/PedGlioma_4869_Full_Covariates_uncript.csv") %>% 
  as.data.frame() %>%
  mutate(WHO_jon =DIFFERN2) %>%
  dplyr::select(blindid,WHO_jon,SITE_02)

glioS = fread("/project/wiemels_260/sebastian/gwas_glioma/saxb4/rawSampleAttributes.txt") %>%
  as.data.frame() %>%
  left_join(glioS.who,by=c("subjectId"="blindid"))%>%
  filter(HISTO_T3 != 9391| is.na(HISTO_T3)) 


## details of Ron's suggestions to re-assign WHO grades

glioS$WHO_jon[which(glioS$HISTO_T3==9384)]=1

glioS$WHO_jon[which(glioS$HISTO_T3==9401)]

glioS$WHO_jon[which(glioS$HISTO_T3==9410)]
glioS$WHO_jon[which((glioS$HISTO_T3==9410) & (glioS$WHO_jon==9))]=1

glioS$WHO_jon[which(glioS$HISTO_T3==9411)]
glioS$WHO_jon[which((glioS$HISTO_T3==9411) & (glioS$WHO_jon==9))]=2

glioS$WHO_jon[which(glioS$HISTO_T3==9412)]
glioS$WHO_jon[which((glioS$HISTO_T3==9412) & (glioS$WHO_jon==9))]=1

glioS$WHO_jon[which(glioS$HISTO_T3==9413)]
glioS$WHO_jon[which((glioS$HISTO_T3==9413) & (glioS$WHO_jon==9))]=1

glioS$WHO_jon[which(glioS$HISTO_T3==9420)]
glioS$WHO_jon[which((glioS$HISTO_T3==9420) & (glioS$WHO_jon==9))]=2

glioS$WHO_jon[which(glioS$HISTO_T3==9421)]
glioS$WHO_jon[which((glioS$HISTO_T3==9421) & (glioS$WHO_jon==9))]=1

glioS$WHO_jon[which(glioS$HISTO_T3==9440)]
glioS$WHO_jon[which((glioS$HISTO_T3==9440) & (glioS$WHO_jon==9))]=4

glioS$WHO_jon[which(glioS$HISTO_T3==9441)]
glioS$WHO_jon[which((glioS$HISTO_T3==9441) & (glioS$WHO_jon==9))]=4

glioS$WHO_jon[which(glioS$HISTO_T3==9442)]
glioS$WHO_jon[which((glioS$HISTO_T3==9442) & (glioS$WHO_jon==9))]=4

glioS$WHO_jon[which(glioS$HISTO_T3==9444)]
glioS$WHO_jon[which((glioS$HISTO_T3==9444) & (glioS$WHO_jon==9))]=2

glioS$WHO_jon[which(glioS$HISTO_T3==9450)]
glioS$WHO_jon[which((glioS$HISTO_T3==9450) & (glioS$WHO_jon==9))]=2

glioS$WHO_jon[which(glioS$HISTO_T3==9451)]

# assigning subgroup labels for each subjects, with control being 9

# all glioma is just everyone so no biggie

# LGG (WHO I-II): lgg
## (HISTO_T3 >= 9400 & HISTO_T3 <= 9424)| HISTO_T3 %in% c(9440,9441,9442)) AND
## (WHO_jon %in% c(1,2))

# HGG (WHO III-IV): hgg
## (HISTO_T3 >= 9400 & HISTO_T3 <= 9424)| HISTO_T3 %in% c(9440,9441,9442)) AND
## (WHO_jon %in% c(3,4))

# LG astrocytoma (WHO I-II): lga
## (HISTO_T3 >= 9400 & HISTO_T3 <= 9424)| HISTO_T3 %in% c(9440,9441,9442)) AND
## (WHO_jon %in% c(1,2))

# HG astrocytoma (WHO III-IV): hga
## (HISTO_T3 >= 9400 & HISTO_T3 <= 9424)| HISTO_T3 %in% c(9440,9441,9442)) AND
## (WHO_jon %in% c(3,4))

# PA: pa
## (HISTO_T3 == 9421) 

# PA Optic pathway: pao
## (HISTO_T3 == 9421|SITE_02=="C723")

# Glioblastoma: glm
## (HISTO_T3 == 9440) 

# All astrocytoma: ast
#(HISTO_T3 >= 9400 & HISTO_T3 <= 9424)| HISTO_T3 %in% c(9440,9441,9442)


glioS1 = glioS %>%
  mutate(ast = ifelse((!is.na(HISTO_T3) & HISTO_T3 >= 9400 & HISTO_T3 <= 9424)| HISTO_T3 %in% c(9440,9441,9442),1,ifelse(phenotype==1,9,NA)))
table(glioS1$ast,useNA="always")
 

glioS1 = glioS %>%
  mutate(lgg = ifelse(
    WHO_jon %in% c(1,2),1,ifelse(phenotype==1,9,NA)),
    hgg = ifelse(
    WHO_jon %in% c(3,4),1,ifelse(phenotype==1,9,NA)),
    lga = ifelse(
      ((HISTO_T3 >= 9400 & HISTO_T3 <= 9424)| HISTO_T3 %in% c(9440,9441,9442)) & (WHO_jon %in% c(1,2)),1,
      ifelse(phenotype==1,9,NA)),
    hga = ifelse(
      ((HISTO_T3 >= 9400 & HISTO_T3 <= 9424)| HISTO_T3 %in% c(9440,9441,9442)) & (WHO_jon %in% c(3,4)),1,
                  ifelse(phenotype==1,9,NA)),
    ast = ifelse(
      (!is.na(HISTO_T3) & (HISTO_T3 >= 9400 & HISTO_T3 <= 9424))| HISTO_T3 %in% c(9440,9441,9442),1,
                 ifelse(phenotype==1,9,NA)),
    pa = ifelse(
     (!is.na(HISTO_T3))&HISTO_T3 == 9421,1,
                   ifelse(phenotype==1,9,NA)),
    pao = ifelse(
      (!is.na(HISTO_T3))&(HISTO_T3 == 9421|SITE_02=="C723"),1,
      ifelse(phenotype==1,9,NA)),
    glm = ifelse(
      (!is.na(HISTO_T3))&(HISTO_T3 == 9440),1,
      ifelse(phenotype==1,9,NA)),
    olg = ifelse(
      (!is.na(HISTO_T3))&(HISTO_T3 %in% c(9450,9451)),1,
      ifelse(phenotype==1,9,NA))
    ) %>%
  dplyr::select(FID,IID,subjectId,sex,phenotype,source,YRBIRTH,race,HISTO_T3,WHO_jon,SITE_02,
                lgg, hgg, lga, hga, ast,pa, pao, glm,olg) %>%
    mutate(neoID = paste0(FID,"_",IID),
           neoRace = ifelse(race %in% c("Asian","Asian_ME"),"asn",
                            ifelse(race%in%c("Black","Black_ME"),"blk",
                                   ifelse(race %in%c("White","White_ME"),"eur",
                                          ifelse(race=="Hispanic","lat",race)))))
  
# 6582   20
fwrite(glioS1, "/project/wiemels_494/sebastian/glioma_gwas/eur_meta_variables.txt")
#glioS1 = fread("/project/wiemels_494/sebastian/glioma_gwas/eur_meta_variables.txt")

glioS1.subjects.collab = glioS1 %>%
  dplyr::select(FID,IID)
write.table(glioS1.subjects.collab,"/project/wiemels_494/sebastian/glioma_gwas/glioS1.subjects.collab.txt",
            quote=FALSE,
            row.names = FALSE,
            col.names = FALSE,
            sep='\t')

glioS1.subjects.collab.white = glioS1 %>%
  filter(race%in%c("White","White_ME")) %>%
  dplyr::select(FID,IID)
write.table(glioS1.subjects.collab.white,"/project/wiemels_494/sebastian/glioma_gwas/glioS1.subjects.collab.white.txt",
            quote=FALSE,
            row.names = FALSE,
            col.names = FALSE,
            sep='\t')

glioS1.subjects.collab.his = glioS1 %>%
  filter(neoRace=="lat") %>%
  dplyr::select(FID,IID)
write.table(glioS1.subjects.collab.his,"/project/wiemels_494/sebastian/glioma_gwas/glioS1.subjects.collab.his.txt",
            quote=FALSE,
            row.names = FALSE,
            col.names = FALSE,
            sep='\t')

glioS1.subjects.collab.asn = glioS1 %>%
  filter(neoRace=="asn") %>%
  dplyr::select(FID,IID)
write.table(glioS1.subjects.collab.asn,"/project/wiemels_494/sebastian/glioma_gwas/glioS1.subjects.collab.asn.txt",
            quote=FALSE,
            row.names = FALSE,
            col.names = FALSE,
            sep='\t')

glioS1.subjects.collab.blk = glioS1 %>%
  filter(neoRace=="blk") %>%
  dplyr::select(FID,IID)
write.table(glioS1.subjects.collab.blk,"/project/wiemels_494/sebastian/glioma_gwas/glioS1.subjects.collab.blk.txt",
            quote=FALSE,
            row.names = FALSE,
            col.names = FALSE,
            sep='\t')

glioS1_w_control = glioS1 %>%
  filter(race%in%c("White","White_ME")) %>%
  filter(phenotype==1) %>%
  dplyr::select(FID,IID)
  

glioS1_w_case = glioS1 %>%
  filter(race%in%c("White","White_ME"))%>%
  filter(phenotype==2)%>%
  dplyr::select(FID,IID)
  
write.table(glioS1_w_control,"/project/wiemels_494/sebastian/glioma_gwas/white_control.txt", sep = "\t", row.names = FALSE, col.names=FALSE,quote = FALSE)
write.table(glioS1_w_case,"/project/wiemels_494/sebastian/glioma_gwas/white_case.txt", sep = "\t", row.names = FALSE, col.names=FALSE,quote = FALSE)


###
# outputting and tabulating for european collaborators

# a=fread("EUR_ids.txt") %>% select(V1,V2) %>%
#   inner_join(glioS,by=c("V2"="IID"))
# b=fread("LAT_ids.txt") %>% select(V1,V2) %>%
#   inner_join(glioS,by=c("V2"="IID"))
# c=fread("AFR_ids.txt") %>% select(V1,V2) %>%
#   inner_join(glioS,by=c("V2"="IID"))
# d=fread("EAS_ids.txt") %>% select(V1,V2) %>%
#   inner_join(glioS,by=c("V2"="IID"))
# 
# 
# enquery = glioS # for all subjects
# #enquery1 = enquery %>% filter(status=="case") %>% dplyr::select(`Sample Filename`,WHO_jon,blindid)
# #write.csv(enquery1,"glio.all.1215.csv")
# 
# enquery = a # for european subjects 
# enquery = b # for latino subjects 
# enquery = c # for african subjects 
# enquery = d # for east asian subjects 
# enquery = glioS %>% filter(AGE<=5)
# 
# ## all glioma
# enquery = enquery %>%
#   mutate(ageGroup=ifelse(AGE>=0&AGE<=4,"1_0-4",
#                          ifelse(AGE>=5&AGE<=9,"2_5-9",
#                                 ifelse(AGE>=10&AGE<=14,"3_10-14","4_others"))))
# all.glios = enquery %>% filter(status=="case") 
# table(all.glios$ageGroup)
# nrow(all.glios)
# 
# #DIFFERN2 (1988 - 2017)
# #GRADECLIN (2018 – )
# #https://www.naaccr.org/wp-content/uploads/2021/03/Grade-Manual_v-2.01.pdf (page 185, Grade 24)
# # SUMSTAGE (https://cspedia.usc.edu/Tumor-Diagnostic_Staging/SUMSTAGE.htm)
# #0  In situ
# #1  Localized (including micro-invasive)
# #2  Regional, extension only
# #3  Regional, nodes only
# #4  Regional, extension and nodes
# #5  Regional, NOS
# #7  Distant
# #8  Blank - not abstracted
# #9  Unknown
# 
# # tb1=table(enquery$DIFFERN2)
# # tb2=table(enquery$HISTO_T3)
# # tb3=table(enquery$SITE_02)
# # write.csv(tb1,"tb1.csv")
# # write.csv(tb2,"tb2.csv")
# # write.csv(tb3,"tb3.csv")
# 
# ## LGG (WHO I-II)
# ## HGG (WHO III-IV)
# # table(enquery$DIFFERN2,enquery$ageGroup)
# # table(enquery$DIFFERN2)
# 
# tbl_grade_age = table(enquery$WHO_jon,enquery$ageGroup)
# tbl_grade = table(enquery$WHO_jon)
# 
# tbl_grade_age["1",] + tbl_grade_age["2",]
# sum(tbl_grade["1"]+tbl_grade["2"])
# 
# tbl_grade_age["3",] + tbl_grade_age["4",]
# sum(tbl_grade["3"]+tbl_grade["4"])
# 
# ##  All astrocytoma	
# #some additional codes should be added, including subependymal giant cell astrocytoma (9384) and glioblastoma (9440), giant cell glioblastoma (9441) and gliosarcoma (9442) 
# 
# temp = enquery %>% 
#   dplyr::filter((HISTO_T3 >= 9400 & HISTO_T3 <= 9424)| HISTO_T3 %in% c(9440,9441,9442))
# table(temp$ageGroup)
# nrow(temp)
# 
#  
# ## LG  astrocytoma (WHOI-II)	
# temp = enquery %>%
#   dplyr::filter((HISTO_T3 >= 9400 & HISTO_T3 <= 9424)| HISTO_T3 %in% c(9440,9441,9442))%>%
#   dplyr::filter(WHO_jon %in% c(1,2))
# table(temp$ageGroup)
# nrow(temp)
# 
# ## HG astrocytoma (WHO III-IV)	
# temp = enquery %>%
#   dplyr::filter((HISTO_T3 >= 9400 & HISTO_T3 <= 9424)| HISTO_T3 %in% c(9440,9441,9442))%>%
#   dplyr::filter(WHO_jon %in% c(3,4))
# table(temp$ageGroup)
# nrow(temp)
# 
# ## Pilocytic astrocytoma	
# temp = enquery %>% 
#   dplyr::filter(HISTO_T3 == 9421) 
# table(temp$ageGroup)
# nrow(temp)
# 
# ## Pilocytic astrocytoma incl. optic pathway glioma	
# temp = enquery %>% 
#   dplyr::filter(HISTO_T3 == 9421|SITE_02=="C723")
# table(temp$ageGroup)
# nrow(temp)
# 
# ## Glioblastoma
# temp = enquery %>% dplyr::filter(HISTO_T3 == 9440) 
# table(temp$ageGroup)
# nrow(temp)

#row1 = enquery %>% dplyr::filter(HISTO_T3 == 9440)  %>% filter(DIFFERN2 %in% c(1,2,3)) %>% mutate(patho="Glioblastoma")
#row2 = enquery %>% dplyr::filter(HISTO_T3 == 9421)  %>% filter(DIFFERN2 %in% c(2,3,4)) %>% mutate(patho="Pilocytic")
#rows = rbind(row1,row2) %>% as.data.frame()
#write.csv(rows,"gliomas.classification.csv")


```

# GWAS QC according to agreed pipleline

```{bash QC, subjects of all ancestries}

#cd /scratch/lishaobo/GWAS/glioma/saxb4
cd /project/wiemels_494/sebastian/glioma_gwas

## sex-check is not implemented in PLINK2 yet. use PLINK1.9 for this. Version is PLINK v1.90b6.17 64-bit (28 Apr 2020) 

# select subjects that will be included in the joint analysis
plink --bfile saxb4_incl\
  --keep glioS1.subjects.collab.txt\
  --make-bed\
  --out /project/wiemels_494/sebastian/glioma_gwas/saxb4_incl_collab
#3275 cases and 3307 controls remaining after main filters.

# (1)	Delete SNPs with more than 20% missing. 
plink --bfile /project/wiemels_494/sebastian/glioma_gwas/saxb4_incl_collab \
  --geno 0.2 --make-bed --out /scratch1/lishaobo/temps/saxb4_incl_1 
# 636 variants removed due to missing genotype data (--geno).

# (2)	Delete individuals with more than 20% missing 
plink --bfile /scratch1/lishaobo/temps/saxb4_incl_1 --mind 0.2 --make-bed --out /scratch1/lishaobo/temps/saxb4_incl_2
# 0 people removed due to missing genotype data (--mind).

# (3)	Delete SNPs with more than 2% missing.
plink --bfile /scratch1/lishaobo/temps/saxb4_incl_2 --geno 0.02 --make-bed --out /scratch1/lishaobo/temps/saxb4_incl_3
# 10854 variants removed due to missing genotype data (--geno).

# (4)	Delete individuals with more than 2% missing.
plink --bfile /scratch1/lishaobo/temps/saxb4_incl_3 --mind 0.02 --make-bed --out /scratch1/lishaobo/temps/saxb4_incl_4
# 37 people removed due to missing genotype data (--mind).

# (5) Remove samples with sex-check errors/problems

sed -i 's/^24/X/' /scratch1/lishaobo/temps/saxb4_incl_4.bim
sed -i 's/^25/Y/' /scratch1/lishaobo/temps/saxb4_incl_4.bim
sed -i 's/^26/MT/' /scratch1/lishaobo/temps/saxb4_incl_4.bim

#cutoff at 0.5 for mixed females and 0.7 for males in USC mixed dataset (0.2 and 0.8 for European dataset)
plink --bfile /scratch1/lishaobo/temps/saxb4_incl_4 --check-sex  0.5 0.7
# --check-sex: 54340 Xchr and 0 Ychr variant(s) scanned, 31 problems detected.

# Generate plots to visualize the sex-check results.
Rscript --vanilla --no-save /project/wiemels_260/sebastian/gwas_glioma/codes/gender_check.R

grep "PROBLEM" plink.sexcheck | awk '{print$1,$2}'> sex_discrepancy.txt
# 31 subjects removed for sex discrepancy issues

plink --bfile /scratch1/lishaobo/temps/saxb4_incl_4 --remove sex_discrepancy.txt --make-bed --out /scratch1/lishaobo/temps/saxb4_incl_5
#813621 variants and 6514 people pass filters and QC.

# Generate a bfile with autosomal SNPs only
plink --bfile /scratch1/lishaobo/temps/saxb4_incl_5 --chr 1-22 --make-bed --out /scratch1/lishaobo/temps/saxb4_incl_6
# 757225 variants and 6514 people pass filters and QC.
  
# (6)	Selecting SNPs with HWE p-value below 1e-4 for controls, 1e-10 for cases
## for USC mixed dataset, use white subjects to filter 

plink --bfile /scratch1/lishaobo/temps/saxb4_incl_6 --hardy --keep white_control.txt
awk '{ if ($9 <0.0001) print $0 }' plink.hwe>plinkzoomhwe.hwe

# to visualize HWE distributions
# Rscript --no-save /project/wiemels_260/sebastian/gwas_glioma/codes/hwe.R

plink --bfile /scratch1/lishaobo/temps/saxb4_incl_6 --hwe 1e-4 --make-bed --out /scratch1/lishaobo/temps/hwe_glio_6_controls --keep white_control.txt
plink --bfile /scratch1/lishaobo/temps/saxb4_incl_6 --extract /scratch1/lishaobo/temps/hwe_glio_6_controls.bim --make-bed --out /scratch1/lishaobo/temps/saxb4_incl_6_1
plink --bfile /scratch1/lishaobo/temps/saxb4_incl_6_1 --hwe 1e-10 --make-bed --out /scratch1/lishaobo/temps/glio_6_1_cases --keep white_case.txt 
plink --bfile /scratch1/lishaobo/temps/saxb4_incl_6_1 --extract /scratch1/lishaobo/temps/glio_6_1_cases.bim --make-bed --out /scratch1/lishaobo/temps/saxb4_incl_7
#755534 variants and 6514 people pass filters and QC.

# (7)	Remove heterozygosity rate outliers: Remove samples not within +/- 0.20
plink --bfile /scratch1/lishaobo/temps/saxb4_incl_7\
  --exclude /project/wiemels_260/sebastian/gwas_glioma/codes/inversion_hg38.txt\
  --indep-pairwise 50 5 0.2\
  --out indepSNP

plink --bfile /scratch1/lishaobo/temps/saxb4_incl_7\
  --extract indepSNP.prune.in\
  --het\
  --out R_check

# Plot of heterozygosity rate distribution
Rscript --no-save /project/wiemels_260/sebastian/gwas_glioma/codes/check_heterozygosity_rate.R

awk 'NR>1 && (($5-$3)/$5<-0.2||($5-$3)/$5>0.2){print $0}' R_check.het > het_fail_ind.txt
# none will be removed from USC mixed dataset

# Remove heterozygosity rate outliers.
plink --bfile /scratch1/lishaobo/temps/saxb4_incl_7 --remove het_fail_ind.txt --make-bed --out /scratch1/lishaobo/temps/saxb4_incl_8

# (8)	Remove samples that are cryptically related: PiHat=> 0.2
plink --bfile /scratch1/lishaobo/temps/saxb4_incl_8 \
  --extract indepSNP.prune.in \
  --genome \
  --min 0.2 \
  --out pihat_min0.2

awk '$10>0.25{print $0}' pihat_min0.2.genome > pihat_min0.25.genome

# Generate a plot to assess the type of relationship.
#awk '{ if ($8 >0.9) print $0 }' pihat_min0.25.genome > zoom_pihat.genome
#Rscript --no-save /project/wiemels_260/sebastian/gwas_glioma/codes/Relatedness.R

# Remove one person from each pair
# awk 'NR>1{print $1,$2}' pihat_min0.25.genome > remove.related.subs.txt
#plink --bfile /scratch1/lishaobo/temps/saxb4_incl_8\
#  --make-bed\
#  --out /scratch1/lishaobo/temps/saxb4_incl_9\
#  --remove remove.related.subs.txt
# 755534 variants and 6366 people pass filters and QC.

# drop previously.dropped.ids.old.cluster.txt file since the cluster changed 
plink --bfile /scratch1/lishaobo/temps/saxb4_incl_8\
  --make-bed\
  --out /scratch1/lishaobo/temps/saxb4_incl_9\
  --remove previously.dropped.ids.old.cluster.txt
# 755534 variants and 6366 people pass filters and QC.

# (9)	case-control SNP missingness filtering (2%)
## doing another 2% missingness filtering in cases and controls separately, and drop SNPs with more than 2% missing in either cases or controls.
plink --bfile /scratch1/lishaobo/temps/saxb4_incl_9\
  --geno 0.02\
  --filter-cases\
  --make-bed\
  --out /scratch1/lishaobo/temps/saxb4_incl_case 
# 484 variants removed due to missing genotype data (--geno).
sort <(awk '{print $2}' /scratch1/lishaobo/temps/saxb4_incl_case.bim) <(awk '{print $2}' /scratch1/lishaobo/temps/saxb4_incl_9.bim) | uniq -u > dropped.cases

plink --bfile /scratch1/lishaobo/temps/saxb4_incl_9\
  --geno 0.02\
  --filter-controls\
  --make-bed\
  --out /scratch1/lishaobo/temps/saxb4_incl_control
# 491 variants removed due to missing genotype data (--geno).
sort <(awk '{print $2}' /scratch1/lishaobo/temps/saxb4_incl_control.bim) <(awk '{print $2}' /scratch1/lishaobo/temps/saxb4_incl_9.bim) | uniq -u > dropped.controls

sort <(cat dropped.controls dropped.cases) | uniq > dropped.cases.controls
# 969 dropped.cases.controls

plink --bfile /scratch1/lishaobo/temps/saxb4_incl_9\
  --exclude dropped.cases.controls\
  --make-bed\
  --out /scratch1/lishaobo/temps/saxb4_incl_10 
# 754565 variants and 6366 people pass filters and QC.

# (10)	 (DK only) Remove A/T and C/G snps 

# (11)	 Remove SNPs with low MAF
## for mixed cohort, MAF<0.01
## for European cohorts, MAF<0.05
plink --bfile /scratch1/lishaobo/temps/saxb4_incl_10 --maf 0.01 --make-bed --out /scratch1/lishaobo/temps/saxb4_incl_11
#268109 variants removed due to minor allele threshold(s)
#486456 variants and 6366 people pass filters and QC.

# (12)	(DK and SWE only, and maybe USC Europeans) Remove genetic outliers (samples identified as outliers in EIGENSOFT (PCA), default settings (+-6SD in 10 pcs, in any PC, repeat 5 times). 
plink --bfile saxb4_incl\
   --keep glioS1.subjects.collab.white.txt\
   --make-bed\
   --out /scratch1/lishaobo/temps/saxb4_incl_collab_wt

plink --bfile /scratch1/lishaobo/temps/saxb4_incl_collab_wt\
  --exclude /project/wiemels_260/sebastian/gwas_glioma/codes/inversion_hg38.txt\
  --range \
  --indep-pairwise 50 5 0.2 \
  --out indepSNP.wt

# Using a set of pruned SNPs
plink --bfile /scratch1/lishaobo/temps/saxb4_incl_collab_wt --extract indepSNP.wt.prune.in --genome --out /scratch1/lishaobo/temps/saxb4_incl_collab_wt
plink --bfile /scratch1/lishaobo/temps/saxb4_incl_collab_wt \
  --read-genome /scratch1/lishaobo/temps/saxb4_incl_collab_wt.genome \
  --cluster \
  --mds-plot 10 \
  --out saxb4_incl_collab_wt
# saxb4_incl_collab_wt.mds  
  
Rscript /project/wiemels_260/sebastian/gwas_glioma/codes/removeGeneticOutliers.R saxb4_incl_collab_wt.mds

# (12.5)	Try (12) for USC Asian-, African and Latino-Americans too
cd /project/wiemels_494/sebastian/glioma_gwas

for race in "his" "blk" "asn"; do

  plink --bfile /project/wiemels_494/sebastian/glioma_gwas/saxb4_incl\
     --keep glioS1.subjects.collab.$race.txt\
     --make-bed\
     --out /scratch2/lishaobo/temp/saxb4_incl_collab_$race
  plink --bfile /scratch2/lishaobo/temp/saxb4_incl_collab_$race\
    --exclude /project/wiemels_260/sebastian/gwas_glioma/codes/inversion_hg38.txt\
    --range \
    --indep-pairwise 50 5 0.2 \
    --out /scratch2/lishaobo/temp/indepSNP.$race
  plink --bfile /scratch2/lishaobo/temp/saxb4_incl_collab_$race\
    --extract /scratch2/lishaobo/temp/indepSNP.$race.prune.in \
    --genome \
    --out /scratch2/lishaobo/temp/saxb4_incl_collab_$race
  plink --bfile /scratch2/lishaobo/temp/saxb4_incl_collab_$race\
    --read-genome /scratch2/lishaobo/temp/saxb4_incl_collab_$race.genome \
    --cluster \
    --mds-plot 10 \
    --out saxb4_incl_collab_$race
  Rscript /project/wiemels_260/sebastian/gwas_glioma/codes/removeGeneticOutliers.R saxb4_incl_collab_$race.mds
done

```

```{r removeGeneticOutliers.R}
#setwd("/scratch/lishaobo/GWAS/glioma/saxb4")
library(tidyverse)
library(data.table)
mdsFile = commandArgs(trailingOnly=TRUE)[1]
#mdsFile = "saxb4_incl_collab_wt.mds"

mds = fread(mdsFile) %>% as.data.frame()
outliers = list()
k=1

for(i in 1:5){
  
 for(j in 1:10){
   
   pcName = paste0("C",j)
   pcs = mds[,pcName] %>% as.numeric()
   sd = sd(pcs)
   mean = mean(pcs)
   
   dropId = which(mds[,pcName] < mean-6*sd & mds[,pcName] > mean+6*sd) 
   if(length(dropId)!=0){
      outliers[[k]]<-  mds[dropId,c("FID","IID")]
      k = k+1
      mds = mds[-dropId,]
   }
   #print(length(dropId))
 } 
  
  print(paste0("i = ",i," interation"))
  print(nrow(mds))
  }

if(length(outliers)==0){
  print("No outlier needs to be removed")
}else{
  dropped = do.call(rbind,outliers) %>% as.data.frame() %>% distinct(IID,.keep_all=TRUE)
  filename = sub(".mds",".dropped.txt",mdsFile)
  write.table(dropped,filename,
              quote=FALSE,
              row.names = FALSE,
              col.names = FALSE,
              sep='\t')
  print(paste(nrow(dropped),"subjects needs to be dropped. IDs were written to",filename))
}

```

Imputation of SNPs on topmed server

```{bash imputation preparation}

cd /scratch/lishaobo/GWAS/glioma/saxb4
#cd /project/wiemels_494/sebastian/glioma_gwas

plink --bfile /scratch1/lishaobo/temps/saxb4_incl_11\
  --allow-no-sex \
  --recode vcf\
  --keep-allele-order \
  --out c/pre_impute

cd /scratch/lishaobo/GWAS/glioma/saxb4/topmed_imputation

awk '{ 
        if($0 !~ /^#/) 
            print "chr"$0;
        else if(match($0,/(##contig=<ID=)(.*)/,m))
            print m[1]"chr"m[2];
        else print $0 
      }' pre_impute.vcf > pre_impute_chr.vcf

#sort and tabix 
bcftools sort -T /scratch/lishaobo/tmp  pre_impute_chr.vcf -Oz -o pre_impute_chr_sorted.vcf.gz
tabix pre_impute_chr_sorted.vcf.gz

#split by chromomse
for chr in {1..22}; doutput
  bcftools view --regions chr"${chr}" pre_impute_chr_sorted.vcf.gz -o pre_impute_chr"${chr}".vcf.gz -Oz &
done

```

```{bash after imputation, extract data}

cd /scratch/lishaobo/GWAS/glioma/saxb4/topmed_imputed

module load unzip

for chr in {1..22}; do
  unzip -P "9n4eLaTiNCeZgC" chr_"${chr}".zip
  echo chr_"${chr}".zip unzip success
done  

############################################################
# Filtering SNPs after TOPMED imputation: r^2 > 0.6, CREDIT TO NICK'S SUGGESTIONS

cd /scratch/lishaobo/GWAS/glioma/saxb4/topmed_imputed

for chr in {1..22}; do

  awk '($7 >0.6) {print $1}' <(zcat chr"${chr}".info.gz) | sed '1d' > include.chr"${chr}".txt
  awk '($7 >0.9) {print $1}' <(zcat chr"${chr}".info.gz) | sed '1d' > pca.include.chr"${chr}".txt

  # Only include r^>0.6, MAF>0.01 and bi-alleles
  plink2 --vcf chr"${chr}".dose.vcf.gz \
    dosage=HDS \
    --max-alleles 2 \
    --min-alleles 2 \
    --extract include.chr"${chr}".txt\
    --maf 0.01 \
    --make-pgen \
    --out chr"${chr}"_imputed \
    --threads 200

done

```

```{bash liftover of SNPs}

cd /project/wiemels_494/sebastian/glioma_gwas/usc_cohort/topmed_imputed
#cd /scratch/lishaobo/GWAS/glioma/saxb4/topmed_imputed

for chr in {1..22}; do

  Rscript /project/wiemels_260/sebastian/gwas_glioma/codes/createBedForLiftOver.R chr$chr.info.gz
  #output file will be chr$chr.glio.hg38.txt
  
  liftOver chr$chr.glio.hg38.txt /project/wiemels_260/sebastian/dependencies/liftOver/hg38ToHg19.over.chain\
    chr$chr.glio.hg37.txt unlifted.chr$chr.glio.hg37.txt

  Rscript /project/wiemels_260/sebastian/gwas_glioma/codes/cleaningLiftedOverFile.R chr$chr.glio.hg37.txt chr$chr.info.gz
  # output file will be chr$chr.hg37.ref.alleles.txt

done


```

```{r createBedForLiftOver.R}

library(tidyverse)
library(data.table)
#setwd("/scratch/lishaobo/GWAS/glioma/saxb4/topmed_imputed")
#setwd("/project/wiemels_494/sebastian/glioma_gwas/usc_cohort/topmed_imputed")

infoFileName = commandArgs(trailingOnly=TRUE)[1]
#infoFileName = "chr22.info.gz"

infoFile = fread(infoFileName) %>%
  as.data.frame() %>%
  mutate(snp38ID = SNP,
         ref0=`REF(0)`,
         alt1 = `ALT(1)`) %>%
  separate(SNP,c("chr","pos","ref","alt")) %>%
  mutate(pos = as.numeric(pos),
         start=pos,
         end=pos+1,
         chr38=chr) %>%
  dplyr::select(chr,start,end,snp38ID,chr38)

infoFile$start = format(infoFile$start,scientific = FALSE)
infoFile$end = format(infoFile$end,scientific = FALSE)

liftFileName = gsub("info.gz","glio.hg38.txt",infoFileName)

write.table(infoFile,liftFileName,
            sep='\t',
            row.names = FALSE,
            col.names = FALSE,
            quote=FALSE)

```

```{r cleaningLiftedOverFile.R}

# this code will remove SNPs lifted to a different chromosome, and based on info file, add ref and alt alleles
# use it like this:   Rscript /project/wiemels_260/sebastian/gwas_glioma/codes/cleaningLiftedOverFile.R chr$chr.glio.hg37.txt chr$chr.info.gz
# output file will be chr$chr.hg37.ref.alleles.txt

library(tidyverse)
library(data.table)
#setwd("/project/wiemels_494/sebastian/glioma_gwas/usc_cohort/topmed_imputed")

liftedFileName = commandArgs(trailingOnly=TRUE)[1]
infoFileName = commandArgs(trailingOnly=TRUE)[2]

#liftedFileName = "chr22.glio.hg37.txt"
#infoFileName = "chr22.info.gz"

liftedFile = fread(liftedFileName) %>%
  as.data.frame() %>%
  filter(V1==V5) %>%
  dplyr::select(-V3,-V5)
colnames(liftedFile) = c("chr_hg19","pos_hg19","snp_id_in_hg38")
  
infoFile = fread(infoFileName) %>%
  as.data.frame() %>%
  inner_join(liftedFile,by=c("SNP"="snp_id_in_hg38")) %>%
  dplyr::rename("ref"="REF(0)",
                "alt"="ALT(1)",
                "snp_id_in_hg38"="SNP") %>%
  dplyr::select(chr_hg19,pos_hg19,ref,alt,snp_id_in_hg38)
  
refFileName = gsub("glio.hg37.txt","hg37.ref.alleles.txt",liftedFileName)

write.table(infoFile,refFileName,
            sep='\t',
            row.names = FALSE,
            quote=FALSE)

```

```{r Generate ID lists for GWAS regression executions}

setwd("/project/wiemels_494/sebastian/glioma_gwas")
library(tidyverse)
library(data.table)

glioS = fread("/project/wiemels_494/sebastian/glioma_gwas/eur_meta_variables.txt") 

glioS05 = fread("/project/wiemels_494/sebastian/glioma_gwas/usc_cohort/topmed_imputed/chr1_imputed.psam") %>%
  mutate(neoIID = sub("0_","",`#IID`))

glioS1 = glioS05 %>% 
  dplyr::select(neoIID) %>%
  inner_join(glioS,by=c("neoIID"="IID"))

allSubsEur =  glioS1 %>% filter(neoRace == "eur") %>% dplyr::select(neoID,sex,phenotype) %>% dplyr::rename("#IID"="neoID","SEX"="sex")
allSubsLat =  glioS1 %>% filter(neoRace == "lat") %>% dplyr::select(neoID,sex,phenotype) %>% dplyr::rename("#IID"="neoID","SEX"="sex")
allSubsAfr =  glioS1 %>% filter(neoRace == "blk") %>% dplyr::select(neoID,sex,phenotype) %>% dplyr::rename("#IID"="neoID","SEX"="sex")
allSubsAsn =  glioS1 %>% filter(neoRace == "asn") %>% dplyr::select(neoID,sex,phenotype) %>% dplyr::rename("#IID"="neoID","SEX"="sex")

lggSubsEur =  glioS1 %>% filter(neoRace == "eur",!is.na(lgg)) %>% dplyr::select(neoID,sex,phenotype) %>% dplyr::rename("#IID"="neoID","SEX"="sex")
lggSubsLat =  glioS1 %>% filter(neoRace == "lat",!is.na(lgg)) %>% dplyr::select(neoID,sex,phenotype) %>% dplyr::rename("#IID"="neoID","SEX"="sex")
lggSubsAfr =  glioS1 %>% filter(neoRace == "blk",!is.na(lgg)) %>% dplyr::select(neoID,sex,phenotype) %>% dplyr::rename("#IID"="neoID","SEX"="sex")
lggSubsAsn =  glioS1 %>% filter(neoRace == "asn",!is.na(lgg)) %>% dplyr::select(neoID,sex,phenotype) %>% dplyr::rename("#IID"="neoID","SEX"="sex")

hggSubsEur =  glioS1 %>% filter(neoRace == "eur",!is.na(hgg)) %>% dplyr::select(neoID,sex,phenotype) %>% dplyr::rename("#IID"="neoID","SEX"="sex")
hggSubsLat =  glioS1 %>% filter(neoRace == "lat",!is.na(hgg)) %>% dplyr::select(neoID,sex,phenotype) %>% dplyr::rename("#IID"="neoID","SEX"="sex")
hggSubsAfr =  glioS1 %>% filter(neoRace == "blk",!is.na(hgg)) %>% dplyr::select(neoID,sex,phenotype) %>% dplyr::rename("#IID"="neoID","SEX"="sex")
hggSubsAsn =  glioS1 %>% filter(neoRace == "asn",!is.na(hgg)) %>% dplyr::select(neoID,sex,phenotype) %>% dplyr::rename("#IID"="neoID","SEX"="sex")

lgaSubsEur =  glioS1 %>% filter(neoRace == "eur",!is.na(lga)) %>% dplyr::select(neoID,sex,phenotype) %>% dplyr::rename("#IID"="neoID","SEX"="sex")
lgaSubsLat =  glioS1 %>% filter(neoRace == "lat",!is.na(lga)) %>% dplyr::select(neoID,sex,phenotype) %>% dplyr::rename("#IID"="neoID","SEX"="sex")
lgaSubsAfr =  glioS1 %>% filter(neoRace == "blk",!is.na(lga)) %>% dplyr::select(neoID,sex,phenotype) %>% dplyr::rename("#IID"="neoID","SEX"="sex")
lgaSubsAsn =  glioS1 %>% filter(neoRace == "asn",!is.na(lga)) %>% dplyr::select(neoID,sex,phenotype) %>% dplyr::rename("#IID"="neoID","SEX"="sex")

hgaSubsEur =  glioS1 %>% filter(neoRace == "eur",!is.na(hga)) %>% dplyr::select(neoID,sex,phenotype) %>% dplyr::rename("#IID"="neoID","SEX"="sex")
hgaSubsLat =  glioS1 %>% filter(neoRace == "lat",!is.na(hga)) %>% dplyr::select(neoID,sex,phenotype) %>% dplyr::rename("#IID"="neoID","SEX"="sex")
hgaSubsAfr =  glioS1 %>% filter(neoRace == "blk",!is.na(hga)) %>% dplyr::select(neoID,sex,phenotype) %>% dplyr::rename("#IID"="neoID","SEX"="sex")
hgaSubsAsn =  glioS1 %>% filter(neoRace == "asn",!is.na(hga)) %>% dplyr::select(neoID,sex,phenotype) %>% dplyr::rename("#IID"="neoID","SEX"="sex")

paSubsEur =  glioS1 %>% filter(neoRace == "eur",!is.na(pa)) %>% dplyr::select(neoID,sex,phenotype) %>% dplyr::rename("#IID"="neoID","SEX"="sex")
paSubsLat =  glioS1 %>% filter(neoRace == "lat",!is.na(pa)) %>% dplyr::select(neoID,sex,phenotype) %>% dplyr::rename("#IID"="neoID","SEX"="sex")
paSubsAfr =  glioS1 %>% filter(neoRace == "blk",!is.na(pa)) %>% dplyr::select(neoID,sex,phenotype) %>% dplyr::rename("#IID"="neoID","SEX"="sex")
paSubsAsn =  glioS1 %>% filter(neoRace == "asn",!is.na(pa)) %>% dplyr::select(neoID,sex,phenotype) %>% dplyr::rename("#IID"="neoID","SEX"="sex")

paoSubsEur =  glioS1 %>% filter(neoRace == "eur",!is.na(pao)) %>% dplyr::select(neoID,sex,phenotype) %>% dplyr::rename("#IID"="neoID","SEX"="sex")
paoSubsLat =  glioS1 %>% filter(neoRace == "lat",!is.na(pao)) %>% dplyr::select(neoID,sex,phenotype) %>% dplyr::rename("#IID"="neoID","SEX"="sex")
paoSubsAfr =  glioS1 %>% filter(neoRace == "blk",!is.na(pao)) %>% dplyr::select(neoID,sex,phenotype) %>% dplyr::rename("#IID"="neoID","SEX"="sex")
paoSubsAsn =  glioS1 %>% filter(neoRace == "asn",!is.na(pao)) %>% dplyr::select(neoID,sex,phenotype) %>% dplyr::rename("#IID"="neoID","SEX"="sex")

glmSubsEur =  glioS1 %>% filter(neoRace == "eur",!is.na(glm)) %>% dplyr::select(neoID,sex,phenotype) %>% dplyr::rename("#IID"="neoID","SEX"="sex")
glmSubsLat =  glioS1 %>% filter(neoRace == "lat",!is.na(glm)) %>% dplyr::select(neoID,sex,phenotype) %>% dplyr::rename("#IID"="neoID","SEX"="sex")
glmSubsAfr =  glioS1 %>% filter(neoRace == "blk",!is.na(glm)) %>% dplyr::select(neoID,sex,phenotype) %>% dplyr::rename("#IID"="neoID","SEX"="sex")
glmSubsAsn =  glioS1 %>% filter(neoRace == "asn",!is.na(glm)) %>% dplyr::select(neoID,sex,phenotype) %>% dplyr::rename("#IID"="neoID","SEX"="sex")

astSubsEur =  glioS1 %>% filter(neoRace == "eur",!is.na(ast)) %>% dplyr::select(neoID,sex,phenotype) %>% dplyr::rename("#IID"="neoID","SEX"="sex")
astSubsLat =  glioS1 %>% filter(neoRace == "lat",!is.na(ast)) %>% dplyr::select(neoID,sex,phenotype) %>% dplyr::rename("#IID"="neoID","SEX"="sex")
astSubsAfr =  glioS1 %>% filter(neoRace == "blk",!is.na(ast)) %>% dplyr::select(neoID,sex,phenotype) %>% dplyr::rename("#IID"="neoID","SEX"="sex")
astSubsAsn =  glioS1 %>% filter(neoRace == "asn",!is.na(ast)) %>% dplyr::select(neoID,sex,phenotype) %>% dplyr::rename("#IID"="neoID","SEX"="sex")

olgSubsEur =  glioS1 %>% filter(neoRace == "eur",!is.na(olg)) %>% dplyr::select(neoID,sex,phenotype) %>% dplyr::rename("#IID"="neoID","SEX"="sex")
olgSubsLat =  glioS1 %>% filter(neoRace == "lat",!is.na(olg)) %>% dplyr::select(neoID,sex,phenotype) %>% dplyr::rename("#IID"="neoID","SEX"="sex")
olgSubsAfr =  glioS1 %>% filter(neoRace == "blk",!is.na(olg)) %>% dplyr::select(neoID,sex,phenotype) %>% dplyr::rename("#IID"="neoID","SEX"="sex")
olgSubsAsn =  glioS1 %>% filter(neoRace == "asn",!is.na(olg)) %>% dplyr::select(neoID,sex,phenotype) %>% dplyr::rename("#IID"="neoID","SEX"="sex")

write.table(allSubsEur,"all.Subs.eur.txt",row.names = FALSE,col.names = TRUE, quote=FALSE, sep='\t')                                
write.table(allSubsLat,"all.Subs.lat.txt",row.names = FALSE,col.names = TRUE, quote=FALSE, sep='\t')                                
write.table(allSubsAfr,"all.Subs.blk.txt",row.names = FALSE,col.names = TRUE, quote=FALSE, sep='\t')                                
write.table(allSubsAsn,"all.Subs.asn.txt",row.names = FALSE,col.names = TRUE, quote=FALSE, sep='\t') 

write.table(lggSubsEur,"lgg.Subs.eur.txt",row.names = FALSE,col.names = TRUE, quote=FALSE, sep='\t')                                
write.table(lggSubsLat,"lgg.Subs.lat.txt",row.names = FALSE,col.names = TRUE, quote=FALSE, sep='\t')                                
write.table(lggSubsAfr,"lgg.Subs.blk.txt",row.names = FALSE,col.names = TRUE, quote=FALSE, sep='\t')                                
write.table(lggSubsAsn,"lgg.Subs.asn.txt",row.names = FALSE,col.names = TRUE, quote=FALSE, sep='\t')

write.table(hggSubsEur,"hgg.Subs.eur.txt",row.names = FALSE,col.names = TRUE, quote=FALSE, sep='\t')                                
write.table(hggSubsLat,"hgg.Subs.lat.txt",row.names = FALSE,col.names = TRUE, quote=FALSE, sep='\t')                                
write.table(hggSubsAfr,"hgg.Subs.blk.txt",row.names = FALSE,col.names = TRUE, quote=FALSE, sep='\t')                                
write.table(hggSubsAsn,"hgg.Subs.asn.txt",row.names = FALSE,col.names = TRUE, quote=FALSE, sep='\t')

write.table(lgaSubsEur,"lga.Subs.eur.txt",row.names = FALSE,col.names = TRUE, quote=FALSE, sep='\t')                                
write.table(lgaSubsLat,"lga.Subs.lat.txt",row.names = FALSE,col.names = TRUE, quote=FALSE, sep='\t')                                
write.table(lgaSubsAfr,"lga.Subs.blk.txt",row.names = FALSE,col.names = TRUE, quote=FALSE, sep='\t')                                
write.table(lgaSubsAsn,"lga.Subs.asn.txt",row.names = FALSE,col.names = TRUE, quote=FALSE, sep='\t')

write.table(hgaSubsEur,"hga.Subs.eur.txt",row.names = FALSE,col.names = TRUE, quote=FALSE, sep='\t')                                
write.table(hgaSubsLat,"hga.Subs.lat.txt",row.names = FALSE,col.names = TRUE, quote=FALSE, sep='\t')                                
write.table(hgaSubsAfr,"hga.Subs.blk.txt",row.names = FALSE,col.names = TRUE, quote=FALSE, sep='\t')                                
write.table(hgaSubsAsn,"hga.Subs.asn.txt",row.names = FALSE,col.names = TRUE, quote=FALSE, sep='\t')

write.table(paSubsEur,"pa.Subs.eur.txt",row.names = FALSE,col.names = TRUE, quote=FALSE, sep='\t')                                
write.table(paSubsLat,"pa.Subs.lat.txt",row.names = FALSE,col.names = TRUE, quote=FALSE, sep='\t')                                
write.table(paSubsAfr,"pa.Subs.blk.txt",row.names = FALSE,col.names = TRUE, quote=FALSE, sep='\t')                                
write.table(paSubsAsn,"pa.Subs.asn.txt",row.names = FALSE,col.names = TRUE, quote=FALSE, sep='\t')

write.table(paoSubsEur,"pao.Subs.eur.txt",row.names = FALSE,col.names = TRUE, quote=FALSE, sep='\t')                                
write.table(paoSubsLat,"pao.Subs.lat.txt",row.names = FALSE,col.names = TRUE, quote=FALSE, sep='\t')                                
write.table(paoSubsAfr,"pao.Subs.blk.txt",row.names = FALSE,col.names = TRUE, quote=FALSE, sep='\t')                                
write.table(paoSubsAsn,"pao.Subs.asn.txt",row.names = FALSE,col.names = TRUE, quote=FALSE, sep='\t')

write.table(glmSubsEur,"glm.Subs.eur.txt",row.names = FALSE,col.names = TRUE, quote=FALSE, sep='\t')                                
write.table(glmSubsLat,"glm.Subs.lat.txt",row.names = FALSE,col.names = TRUE, quote=FALSE, sep='\t')                                
write.table(glmSubsAfr,"glm.Subs.blk.txt",row.names = FALSE,col.names = TRUE, quote=FALSE, sep='\t')                                
write.table(glmSubsAsn,"glm.Subs.asn.txt",row.names = FALSE,col.names = TRUE, quote=FALSE, sep='\t')

write.table(astSubsEur,"ast.Subs.eur.txt",row.names = FALSE,col.names = TRUE, quote=FALSE, sep='\t')                                
write.table(astSubsLat,"ast.Subs.lat.txt",row.names = FALSE,col.names = TRUE, quote=FALSE, sep='\t')                                
write.table(astSubsAfr,"ast.Subs.blk.txt",row.names = FALSE,col.names = TRUE, quote=FALSE, sep='\t')                                
write.table(astSubsAsn,"ast.Subs.asn.txt",row.names = FALSE,col.names = TRUE, quote=FALSE, sep='\t')

write.table(olgSubsEur,"olg.Subs.eur.txt",row.names = FALSE,col.names = TRUE, quote=FALSE, sep='\t')                                
write.table(olgSubsLat,"olg.Subs.lat.txt",row.names = FALSE,col.names = TRUE, quote=FALSE, sep='\t')                                
write.table(olgSubsAfr,"olg.Subs.blk.txt",row.names = FALSE,col.names = TRUE, quote=FALSE, sep='\t')                                
write.table(olgSubsAsn,"olg.Subs.asn.txt",row.names = FALSE,col.names = TRUE, quote=FALSE, sep='\t')


# report numbers for Jon
setwd("/project/wiemels_494/sebastian/glioma_gwas")
library(tidyverse)
library(data.table)

files = list.files(pattern='*.Subs.*.txt') %>% sort

for(file in files){
  a = fread(file)
  tbl = table(a$phenotype)
  print(file)
  print(paste0("number of controls ",tbl['1']))
  print(paste0("number of cases ",tbl['2']))
}

# [1] "all.Subs.asn.txt"
# [1] "number of controls 222"
# [1] "number of cases 226"
# [1] "all.Subs.blk.txt"
# [1] "number of controls 196"
# [1] "number of cases 194"
# [1] "all.Subs.eur.txt"
# [1] "number of controls 1559"
# [1] "number of cases 1554"
# [1] "all.Subs.lat.txt"
# [1] "number of controls 1177"
# [1] "number of cases 1176"
# [1] "ast.Subs.asn.txt"
# [1] "number of controls 222"
# [1] "number of cases 115"
# [1] "ast.Subs.blk.txt"
# [1] "number of controls 196"
# [1] "number of cases 103"
# [1] "ast.Subs.eur.txt"
# [1] "number of controls 1559"
# [1] "number of cases 841"
# [1] "ast.Subs.lat.txt"
# [1] "number of controls 1177"
# [1] "number of cases 589"
# [1] "glm.Subs.asn.txt"
# [1] "number of controls 222"
# [1] "number of cases 9"
# [1] "glm.Subs.blk.txt"
# [1] "number of controls 196"
# [1] "number of cases 4"
# [1] "glm.Subs.eur.txt"
# [1] "number of controls 1559"
# [1] "number of cases 46"
# [1] "glm.Subs.lat.txt"
# [1] "number of controls 1177"
# [1] "number of cases 40"
# [1] "hga.Subs.asn.txt"
# [1] "number of controls 222"
# [1] "number of cases 21"
# [1] "hga.Subs.blk.txt"
# [1] "number of controls 196"
# [1] "number of cases 21"
# [1] "hga.Subs.eur.txt"
# [1] "number of controls 1559"
# [1] "number of cases 152"
# [1] "hga.Subs.lat.txt"
# [1] "number of controls 1177"
# [1] "number of cases 128"
# [1] "hgg.Subs.asn.txt"
# [1] "number of controls 222"
# [1] "number of cases 26"
# [1] "hgg.Subs.blk.txt"
# [1] "number of controls 196"
# [1] "number of cases 26"
# [1] "hgg.Subs.eur.txt"
# [1] "number of controls 1559"
# [1] "number of cases 182"
# [1] "hgg.Subs.lat.txt"
# [1] "number of controls 1177"
# [1] "number of cases 151"
# [1] "lga.Subs.asn.txt"
# [1] "number of controls 222"
# [1] "number of cases 80"
# [1] "lga.Subs.blk.txt"
# [1] "number of controls 196"
# [1] "number of cases 66"
# [1] "lga.Subs.eur.txt"
# [1] "number of controls 1559"
# [1] "number of cases 607"
# [1] "lga.Subs.lat.txt"
# [1] "number of controls 1177"
# [1] "number of cases 380"
# [1] "lgg.Subs.asn.txt"
# [1] "number of controls 222"
# [1] "number of cases 90"
# [1] "lgg.Subs.blk.txt"
# [1] "number of controls 196"
# [1] "number of cases 81"
# [1] "lgg.Subs.eur.txt"
# [1] "number of controls 1559"
# [1] "number of cases 692"
# [1] "lgg.Subs.lat.txt"
# [1] "number of controls 1177"
# [1] "number of cases 438"
# [1] "olg.Subs.asn.txt"
# [1] "number of controls 222"
# [1] "number of cases 2"
# [1] "olg.Subs.blk.txt"
# [1] "number of controls 196"
# [1] "number of cases 6"
# [1] "olg.Subs.eur.txt"
# [1] "number of controls 1559"
# [1] "number of cases 47"
# [1] "olg.Subs.lat.txt"
# [1] "number of controls 1177"
# [1] "number of cases 28"
# [1] "pa.Subs.asn.txt"
# [1] "number of controls 222"
# [1] "number of cases 59"
# [1] "pa.Subs.blk.txt"
# [1] "number of controls 196"
# [1] "number of cases 56"
# [1] "pa.Subs.eur.txt"
# [1] "number of controls 1559"
# [1] "number of cases 468"
# [1] "pa.Subs.lat.txt"
# [1] "number of controls 1177"
# [1] "number of cases 302"
# [1] "pao.Subs.asn.txt"
# [1] "number of controls 222"
# [1] "number of cases 64"
# [1] "pao.Subs.blk.txt"
# [1] "number of controls 196"
# [1] "number of cases 60"
# [1] "pao.Subs.eur.txt"
# [1] "number of controls 1559"
# [1] "number of cases 522"
# [1] "pao.Subs.lat.txt"
# [1] "number of controls 1177"
# [1] "number of cases 337"
```

```{bash execution of GWAS}

#cd /scratch/lishaobo/GWAS/glioma/saxb4
cd /project/wiemels_494/sebastian/glioma_gwas/usc_cohort

#mkdir temps

# for all Glioma subjects

#subtype="lga"
# race='blk'

for subtype in "all" "lgg" "hgg" "lga" "hga" "pa" "pao" "glm" "ast" "olg"; do

  for race in "eur" "lat" "blk" "asn"; do

    ## create a file including a certain race
    
    for chr in {1..22}; do
    
      plink2 --pfile topmed_imputed/chr"${chr}"_imputed\
        --keep ../$subtype.Subs.$race.txt\
        --indiv-sort f ../$subtype.Subs.$race.txt\
        --make-pgen\
        --out $subtype.Subs.$race.chr$chr
      
      cp $subtype.Subs.$race.chr$chr.psam temps/$subtype.Subs.$race.chr$chr.psam.bkup 
      cp $subtype.Subs.$race.txt $subtype.Subs.$race.chr$chr.psam 
      
    done
    
    ## generate PCA for this particular cohort (by race and by subtype)  
    for chr in {1..22}; do
    
        plink2 --pfile $subtype.Subs.$race.chr$chr\
        --make-bed\
        --out temps/$subtype.Subs.$race.chr$chr
    
    done
      
    find ./temps/$subtype.Subs.$race.chr*.bim  > temps/forMerge.$subtype.$race.list 
    sed -i 's/.bim//g' temps/forMerge.$subtype.$race.list 
    
    plink --merge-list temps/forMerge.$subtype.$race.list\
      --indiv-sort 0 \
      --make-bed \
      --out temps/$subtype.Subs.$race
      
    plink2 --bfile temps/$subtype.Subs.$race\
      --maf 0.1 \
      --indep-pairwise 50 10 0.1 \
      --out temps/$subtype.Subs.$race.pruned
      
    plink2 --bfile temps/$subtype.Subs.$race\
      --extract temps/$subtype.Subs.$race.pruned.prune.in \
      --pca 10\
      --out $subtype.Subs.$race.pca\
      --threads 200
    
    # output PCAs that are significantly associated with case control status
    # R code takes 2 parameters, (1) PCA file name (2) plink b file fam file name from which PCA is calculated
    Rscript /project/wiemels_260/sebastian/gwas_glioma/codes/sigPCAs.R $subtype.Subs.$race.pca.eigenvec temps/$subtype.Subs.$race.fam
  done
done


# carry out association analysis
cd /project/wiemels_494/sebastian/glioma_gwas/usc_cohort

#subtype="lga"
#race="lat"

for subtype in "all" "lgg" "hgg" "lga" "hga" "pa" "pao" "glm" "ast" "olg"; do

  for race in "eur" "lat" "blk" "asn"; do

    FILE=/project/wiemels_494/sebastian/glioma_gwas/usc_cohort/$subtype.Subs.$race.pca.eigenvec.to.include

    if test -f "$FILE"; then
    
      for chr in {1..22}; do
        plink2 --pfile $subtype.Subs.$race.chr$chr\
          --covar $FILE\
          --threads 200 \
          --glm firth-fallback sex hide-covar omit-ref cols=chrom,pos,ref,alt1,nobs,a1countcc,a1freq,a1freqcc,beta,se,p,err,ci\
          --ci 0.95\
          --out temps/$subtype.Subs.$race.chr$chr
      done
    
    else
    
      for chr in {1..22}; do
        plink2 --pfile $subtype.Subs.$race.chr$chr\
          --threads 200 \
          --glm firth-fallback sex hide-covar omit-ref cols=chrom,pos,ref,alt1,nobs,a1countcc,a1freq,a1freqcc,beta,se,p,err,ci\
          --ci 0.95\
          --out temps/$subtype.Subs.$race.chr$chr
      done
    
    fi
    
    # Combining results from all chromosomes into 1 file
    head -1 temps/$subtype.Subs.$race.chr$chr.phenotype.glm.logistic.hybrid > $subtype.Subs.$race.phenotype.glm.logistic.hybrid
    tail -n +2 -q temps/$subtype.Subs.$race.chr*.phenotype.glm.logistic.hybrid >> $subtype.Subs.$race.phenotype.glm.logistic.hybrid
    

  done
done

# liftOver to Hg19 (not necessary for non-USA cohorts)

#cd /scratch/lishaobo/GWAS/glioma/saxb4
cd /project/wiemels_494/sebastian/glioma_gwas/usc_cohort
#subtype="all"
#race="eur"

for subtype in "all" "lgg" "hgg" "lga" "hga" "pa" "pao" "glm" "ast" "olg"; do

  for race in "eur" "lat" "blk" "asn"; do

    sort -k1,1n -k2,2n  $subtype.Subs.$race.phenotype.glm.logistic.hybrid |\
      awk  -v OFS="\t" 'NR>1 {print "chr"$1,$2,$2+1,$3}' > temps/$subtype.Subs.$race.hg38.txt
    
    liftOver temps/$subtype.Subs.$race.hg38.txt /project/wiemels_260/sebastian/dependencies/liftOver/hg38ToHg19.over.chain\
      temps/$subtype.Subs.$race.hg37.txt temps/unlifted.$subtype.Subs.$race.hg37.txt
    
    Rscript /project/wiemels_260/sebastian/gwas_glioma/codes/outputHg37Results.R \
      $subtype.Subs.$race.phenotype.glm.logistic.hybrid\
      temps/$subtype.Subs.$race.hg37.txt 

  done
done

#for subtype in "all" "lgg" "hgg" "lga" "hga" "pa" "pao" "glm" "ast" "olg"; do
#  for race in "eur" "lat" "blk" "asn"; do
#    cp /scratch/lishaobo/GWAS/glioma/saxb4/$subtype.Subs.$race.phenotype.glm.logistic.hybrid /project/wiemels_494/sebastian/glioma_gwas/usc_cohort
#    cp /scratch/lishaobo/GWAS/glioma/saxb4/hg19.$subtype.Subs.$race.phenotype.glm.logistic.hybrid #/project/wiemels_494/sebastian/glioma_gwas/usc_cohort
#    cp /scratch/lishaobo/GWAS/glioma/saxb4/$subtype.$race.hwe /project/wiemels_494/sebastian/glioma_gwas/usc_cohort
#  done
#done  
    
```

```{r sigPCAs.R}

# R code takes 2 parameters, (1) PCA file name (2) plink b file fam file name from which PCA is calculated
# Rscript /project/wiemels_260/sebastian/gwas_glioma/codes/sigPCAs.R all.Subs.eur.pca.eigenvec temps/all.Subs.eur.fam

library(tidyverse)
library(data.table)

#setwd("/scratch/lishaobo/GWAS/glioma/saxb4")
#pcaFile = "all.Subs.eur.pca.eigenvec"
#phenotypeFile = "temps/all.Subs.eur.fam"
#estithreshold = 0.5
pthreshold=0.05

pcaFile = commandArgs(trailingOnly=TRUE)[1]
phenotypeFile = commandArgs(trailingOnly=TRUE)[2]

pcsList = c("PC1","PC2","PC3","PC4","PC5","PC6","PC7","PC8","PC9","PC10")
pcas= fread(pcaFile) %>% as.data.frame()

phenotypes = fread(phenotypeFile) %>% as.data.frame() %>% dplyr::select(V2,V6)

combined = pcas %>%
  inner_join(phenotypes , by = c("IID"="V2")) %>%
  as.data.frame()

pcCorrs = c()
pcCorrsPval = c()
i=0

for(pc in pcsList){
  i = i +1
  pcCorrs[i] = cor.test(combined[,pc],combined$V6)$estimate
  pcCorrsPval[i] = cor.test(combined[,pc],combined$V6)$p.value
}

names(pcCorrsPval) = pcsList
names(pcCorrs) = pcsList
pcsToInclude = names(which(pcCorrsPval<pthreshold))

if(length(pcsToInclude) == 0){
  
  print("No need to include any PCs for this regression analysis. Can ignore --covar flag")

  }else{
    
  pcasInclude = pcas%>% 
    as.data.frame() %>%
    dplyr::select('#FID',IID,all_of(pcsToInclude))
  
  fileNameNeo = paste0(pcaFile,".to.include")
  print(paste0("Include ",length(pcsToInclude)," PCs in the regression model. File has been written to ",fileNameNeo ))
  
  write.table(pcasInclude,fileNameNeo,
              row.names = FALSE,
              quote = FALSE,
              sep='\t')
  
}


```

```{r outputHg37Results.R}

# script to lift over regression results from hg38 to hg19
# Rscript /project/wiemels_260/sebastian/gwas_glioma/codes/outputHg37Results.R \
#   pa.Subs.asn.phenotype.glm.logistic.hybrid\
#   temps/pa.Subs.asn.hg37.txt

#setwd("/project/wiemels_494/sebastian/glioma_gwas/usc_cohort")
#resultsName = "all.Subs.eur.phenotype.glm.logistic.hybrid"
#hg37FileName = "temps/all.Subs.eur.hg37.txt"

library(tidyverse)
library(data.table)

resultsName = commandArgs(trailingOnly=TRUE)[1]
hg37FileName = commandArgs(trailingOnly=TRUE)[2]

results = fread(resultsName) %>% as.data.frame()
hg37File = fread(hg37FileName) %>% as.data.frame()
colnames(hg37File) = c("chr_19","pos_19","pos_end_19","ID")

results.hg37 = results %>%
  inner_join(hg37File,by="ID") %>%
  separate(ID,c("chr_38","pos_38","id1_38","id2_38")) 

#results.hg37 %>% filter(ID %in% c("chr4:112087041:T:C","chr11:96557415:T:C"))
#results.hg37 %>% filter(POS %in% c("112087041","96557415"))

results.hg37.output = results.hg37 %>%
  mutate(ID_hg19 = paste0(chr_19,":",pos_19,":",id1_38,":",id2_38),
         `#CHROM`=chr_19,
         POS = pos_19) %>%
  filter(chr_19==chr_38) %>%
  dplyr::select(-chr_38,-pos_38,-id1_38,-id2_38,-chr_19,-pos_19,-pos_end_19) %>%
  dplyr::select(`#CHROM`,POS,ID_hg19,everything()) %>%
  filter(gsub("chr","",`#CHROM`)%in%c(1:22))%>%
  arrange(as.numeric(gsub("chr","",`#CHROM`)),POS)

#results.hg37.output %>% filter(ID_hg19 %in% c("chr4:112087041:T:C","chr11:96557415:T:C"))

table(results.hg37.output$`#CHROM`)  

resultsNameNeo = paste0("hg19.",resultsName)
write.table(results.hg37.output,resultsNameNeo,
            sep='\t',
            quote=FALSE,
            row.names = FALSE,
            col.names = TRUE)

```

```{bash generate outputs to share}

cd /project/wiemels_494/sebastian/glioma_gwas/usc_cohort

# 1 glm.logistic.hybrid for the following groups:
# ls hg19*glm.logistic.hybrid

#hg19.all.Subs.asn.phenotype.glm.logistic.hybrid
#hg19.all.Subs.blk.phenotype.glm.logistic.hybrid
#hg19.all.Subs.eur.phenotype.glm.logistic.hybrid
#hg19.all.Subs.lat.phenotype.glm.logistic.hybrid
#hg19.ast.Subs.asn.phenotype.glm.logistic.hybrid
#hg19.ast.Subs.blk.phenotype.glm.logistic.hybrid
#hg19.ast.Subs.eur.phenotype.glm.logistic.hybrid
#hg19.ast.Subs.lat.phenotype.glm.logistic.hybrid
#hg19.glm.Subs.asn.phenotype.glm.logistic.hybrid
#hg19.glm.Subs.blk.phenotype.glm.logistic.hybrid
#hg19.glm.Subs.eur.phenotype.glm.logistic.hybrid
#hg19.glm.Subs.lat.phenotype.glm.logistic.hybrid
#hg19.hga.Subs.asn.phenotype.glm.logistic.hybrid
#hg19.hga.Subs.blk.phenotype.glm.logistic.hybrid
#hg19.hga.Subs.eur.phenotype.glm.logistic.hybrid
#hg19.hga.Subs.lat.phenotype.glm.logistic.hybrid
#hg19.hgg.Subs.asn.phenotype.glm.logistic.hybrid
#hg19.hgg.Subs.blk.phenotype.glm.logistic.hybrid
#hg19.hgg.Subs.eur.phenotype.glm.logistic.hybrid
#hg19.hgg.Subs.lat.phenotype.glm.logistic.hybrid
#hg19.lga.Subs.asn.phenotype.glm.logistic.hybrid
#hg19.lga.Subs.blk.phenotype.glm.logistic.hybrid
#hg19.lga.Subs.eur.phenotype.glm.logistic.hybrid
#hg19.lga.Subs.lat.phenotype.glm.logistic.hybrid
#hg19.lgg.Subs.asn.phenotype.glm.logistic.hybrid
#hg19.lgg.Subs.blk.phenotype.glm.logistic.hybrid
#hg19.lgg.Subs.eur.phenotype.glm.logistic.hybrid
#hg19.lgg.Subs.lat.phenotype.glm.logistic.hybrid
#hg19.olg.Subs.asn.phenotype.glm.logistic.hybrid
#hg19.olg.Subs.blk.phenotype.glm.logistic.hybrid
#hg19.olg.Subs.eur.phenotype.glm.logistic.hybrid
#hg19.olg.Subs.lat.phenotype.glm.logistic.hybrid
#hg19.pao.Subs.asn.phenotype.glm.logistic.hybrid
#hg19.pao.Subs.blk.phenotype.glm.logistic.hybrid
#hg19.pao.Subs.eur.phenotype.glm.logistic.hybrid
#hg19.pao.Subs.lat.phenotype.glm.logistic.hybrid
#hg19.pa.Subs.asn.phenotype.glm.logistic.hybrid
#hg19.pa.Subs.blk.phenotype.glm.logistic.hybrid
#hg19.pa.Subs.eur.phenotype.glm.logistic.hybrid
#hg19.pa.Subs.lat.phenotype.glm.logistic.hybrid

# 2. Hardy for case and controls for each group.
for subtype in "all" "lgg" "hgg" "lga" "hga" "pa" "pao" "glm" "ast" "olg"; do
   for race in "eur" "lat" "blk" "asn"; do
      for chr in {1..22}; do
        plink2 --pfile $subtype.Subs.$race.chr$chr\
  	      --hardy\
  	      --out temps/$subtype.$race.chr$chr.hwe
      done
      
    head -1 temps/$subtype.$race.chr1.hwe.hardy > $subtype.$race.hwe
    tail -n +2 -q temps/$subtype.$race.chr*.hwe.hardy >> $subtype.$race.hwe
    
  done
done

for subtype in "all" "lgg" "hgg" "lga" "hga" "pa" "pao" "glm" "ast" "olg"; do
   for race in "eur" "lat" "blk" "asn"; do
      echo $subtype.$race.hwe
    done
done

#all.eur.hwe
#all.lat.hwe
#all.blk.hwe
#all.asn.hwe
#lgg.eur.hwe
#lgg.lat.hwe
#lgg.blk.hwe
#lgg.asn.hwe
#hgg.eur.hwe
#hgg.lat.hwe
#hgg.blk.hwe
#hgg.asn.hwe
#lga.eur.hwe
#lga.lat.hwe
#lga.blk.hwe
#lga.asn.hwe
#hga.eur.hwe
#hga.lat.hwe
#hga.blk.hwe
#hga.asn.hwe
#pa.eur.hwe
#pa.lat.hwe
#pa.blk.hwe
#pa.asn.hwe
#pao.eur.hwe
#pao.lat.hwe
#pao.blk.hwe
#pao.asn.hwe
#glm.eur.hwe
#glm.lat.hwe
#glm.blk.hwe
#glm.asn.hwe
#ast.eur.hwe
#ast.lat.hwe
#ast.blk.hwe
#ast.asn.hwe
#olg.eur.hwe
#olg.lat.hwe
#olg.blk.hwe
#olg.asn.hwe

# 3. Infoscores for all SNPs.
# /project/wiemels_494/sebastian/glioma_gwas/usc_cohort/topmed_imputedchr1.info.gz
# /project/wiemels_494/sebastian/glioma_gwas/usc_cohort/topmed_imputedchr2.info.gz
# /project/wiemels_494/sebastian/glioma_gwas/usc_cohort/topmed_imputedchr3.info.gz
# /project/wiemels_494/sebastian/glioma_gwas/usc_cohort/topmed_imputedchr4.info.gz
# /project/wiemels_494/sebastian/glioma_gwas/usc_cohort/topmed_imputedchr5.info.gz
# /project/wiemels_494/sebastian/glioma_gwas/usc_cohort/topmed_imputedchr6.info.gz
# /project/wiemels_494/sebastian/glioma_gwas/usc_cohort/topmed_imputedchr7.info.gz
# /project/wiemels_494/sebastian/glioma_gwas/usc_cohort/topmed_imputedchr8.info.gz
# /project/wiemels_494/sebastian/glioma_gwas/usc_cohort/topmed_imputedchr9.info.gz
# /project/wiemels_494/sebastian/glioma_gwas/usc_cohort/topmed_imputedchr10.info.gz
# /project/wiemels_494/sebastian/glioma_gwas/usc_cohort/topmed_imputedchr11.info.gz
# /project/wiemels_494/sebastian/glioma_gwas/usc_cohort/topmed_imputedchr12.info.gz
# /project/wiemels_494/sebastian/glioma_gwas/usc_cohort/topmed_imputedchr13.info.gz
# /project/wiemels_494/sebastian/glioma_gwas/usc_cohort/topmed_imputedchr14.info.gz
# /project/wiemels_494/sebastian/glioma_gwas/usc_cohort/topmed_imputedchr15.info.gz
# /project/wiemels_494/sebastian/glioma_gwas/usc_cohort/topmed_imputedchr16.info.gz
# /project/wiemels_494/sebastian/glioma_gwas/usc_cohort/topmed_imputedchr17.info.gz
# /project/wiemels_494/sebastian/glioma_gwas/usc_cohort/topmed_imputedchr18.info.gz
# /project/wiemels_494/sebastian/glioma_gwas/usc_cohort/topmed_imputedchr19.info.gz
# /project/wiemels_494/sebastian/glioma_gwas/usc_cohort/topmed_imputedchr20.info.gz
# /project/wiemels_494/sebastian/glioma_gwas/usc_cohort/topmed_imputedchr21.info.gz
# /project/wiemels_494/sebastian/glioma_gwas/usc_cohort/topmed_imputedchr22.info.gz


# 4. A list PCs used as covariates for each model.  
for subtype in "all" "lgg" "hgg" "lga" "hga" "pa" "pao" "glm" "ast" "olg"; do
   for race in "eur" "lat" "blk" "asn"; do
     FILE=$subtype.Subs.$race.pca.eigenvec.to.include
     echo $FILE
     head -1 $FILE
   done
 done

# head -1 all.Subs.eur.pca.eigenvec.to.include
# Significant PCs are PC1	PC6
# head -1 all.Subs.lat.pca.eigenvec.to.include
# Significant PCs are PC1
# head -1 all.Subs.blk.pca.eigenvec.to.include
# No PCs is significant
# head -1 all.Subs.asn.pca.eigenvec.to.include
# Significant PCs are PC5
# head -1 lgg.Subs.eur.pca.eigenvec.to.include
# Significant PCs are PC1	PC6
# head -1 lgg.Subs.lat.pca.eigenvec.to.include
# Significant PCs are PC1
# head -1 lgg.Subs.blk.pca.eigenvec.to.include
# No PCs is significant
# head -1 lgg.Subs.asn.pca.eigenvec.to.include
# Significant PCs are PC5
# head -1 hgg.Subs.eur.pca.eigenvec.to.include
# No PCs is significant
# head -1 hgg.Subs.lat.pca.eigenvec.to.include
# Significant PCs are PC7	PC10
# head -1 hgg.Subs.blk.pca.eigenvec.to.include
# Significant PCs are PC3	PC6
# head -1 hgg.Subs.asn.pca.eigenvec.to.include
# No PCs is significant
# head -1 lga.Subs.eur.pca.eigenvec.to.include
# Significant PCs are PC1	PC6
# head -1 lga.Subs.lat.pca.eigenvec.to.include
# Significant PCs are PC1
# head -1 lga.Subs.blk.pca.eigenvec.to.include
# No PCs is significant
# head -1 lga.Subs.asn.pca.eigenvec.to.include
# Significant PCs are PC5
# head -1 hga.Subs.eur.pca.eigenvec.to.include
# No PCs is significant
# head -1 hga.Subs.lat.pca.eigenvec.to.include
# Significant PCs are PC7	PC10
# head -1 hga.Subs.blk.pca.eigenvec.to.include
# No PCs is significant
# head -1 hga.Subs.asn.pca.eigenvec.to.include
# Significant PCs are PC9
# head -1 pa.Subs.eur.pca.eigenvec.to.include
# Significant PCs are PC1	PC6
# head -1 pa.Subs.lat.pca.eigenvec.to.include
# Significant PCs are PC1	PC9
# head -1 pa.Subs.blk.pca.eigenvec.to.include
# No PCs is significant
# head -1 pa.Subs.asn.pca.eigenvec.to.include
# Significant PCs are PC5	PC8
# head -1 pao.Subs.eur.pca.eigenvec.to.include
# Significant PCs are PC1	PC6
# head -1 pao.Subs.lat.pca.eigenvec.to.include
# Significant PCs are PC1
# head -1 pao.Subs.blk.pca.eigenvec.to.include
# No PCs is significant
# head -1 pao.Subs.asn.pca.eigenvec.to.include
# Significant PCs are PC5
# head -1 glm.Subs.eur.pca.eigenvec.to.include
# Significant PCs are PC10
# head -1 glm.Subs.lat.pca.eigenvec.to.include
# Significant PCs are PC7	PC10
# head -1 glm.Subs.blk.pca.eigenvec.to.include
# Significant PCs are PC1	PC6
# head -1 glm.Subs.asn.pca.eigenvec.to.include
# Significant PCs are PC9
# head -1 ast.Subs.eur.pca.eigenvec.to.include
# Significant PCs are PC1	PC4	PC6
# head -1 ast.Subs.lat.pca.eigenvec.to.include
# Significant PCs are PC1	PC9
# head -1 ast.Subs.blk.pca.eigenvec.to.include
# No PCs is significant
# head -1 ast.Subs.asn.pca.eigenvec.to.include
# Significant PCs are PC3
# head -1 olg.Subs.eur.pca.eigenvec.to.include
# No PCs is significant
# head -1 olg.Subs.lat.pca.eigenvec.to.include
# Significant PCs are PC8
# head -1 olg.Subs.blk.pca.eigenvec.to.include
# Significant PCs are PC6
# head -1 olg.Subs.asn.pca.eigenvec.to.include
# No PCs is significant

```

# meta-analysis of results from different cohorts

```{bash file descriptions}

Anna cohort files:

/project/wiemels_494/sebastian/glioma_gwas/anna_cohort/concat.all.var_3PCA.20220401.PHENO1.glm.logistic.hybrid
/project/wiemels_494/sebastian/glioma_gwas/anna_cohort/concat.lgg.var_0PCA.20220401.PHENO1.glm.logistic.hybrid
/project/wiemels_494/sebastian/glioma_gwas/anna_cohort/concat.astro.var_10PCA.20220401.PHENO1.glm.logistic.hybrid
/project/wiemels_494/sebastian/glioma_gwas/anna_cohort/concat.lg_astro.var_8PCA.20220401.PHENO1.glm.logistic.hybrid
/project/wiemels_494/sebastian/glioma_gwas/anna_cohort/concat.pilo_optic.var_0PCA.20220401.PHENO1.glm.logistic.hybrid
/project/wiemels_494/sebastian/glioma_gwas/anna_cohort/concat.pilo.var_0PCA.20220401.PHENO1.glm.logistic.hybrid
/project/wiemels_494/sebastian/glioma_gwas/anna_cohort/concat.hgg.var_1PCA.20220401.PHENO1.glm.logistic.hybrid
/project/wiemels_494/sebastian/glioma_gwas/anna_cohort/concat.hg_astro.var_3PCA.20220401.PHENO1.glm.logistic.hybrid
/project/wiemels_494/sebastian/glioma_gwas/anna_cohort/concat.gbm.var_0PCA.20220401.PHENO1.glm.logistic.hybrid
/project/wiemels_494/sebastian/glioma_gwas/anna_cohort/concat.oligo.var_0PCA.20220401.PHENO1.glm.logistic.hybrid

DK cohort files:

/project/wiemels_494/sebastian/glioma_gwas/dk_cohort/Results/Astrocytoma.PC5.PHENO1.glm.logistic.hybrid\
/project/wiemels_494/sebastian/glioma_gwas/dk_cohort/Results/LG_astrocytoma.PC5.PHENO1.glm.logistic.hybrid
/project/wiemels_494/sebastian/glioma_gwas/dk_cohort/Results/Glioblastoma.PC5.PHENO1.glm.logistic.hybrid\
/project/wiemels_494/sebastian/glioma_gwas/dk_cohort/Results/LGG.PC5.PHENO1.glm.logistic.hybrid
/project/wiemels_494/sebastian/glioma_gwas/dk_cohort/Results/Glioma.PC5.PHENO1.glm.logistic.hybrid\
/project/wiemels_494/sebastian/glioma_gwas/dk_cohort/Results/Oligodendroglioma.PC5.PHENO1.glm.logistic.hybrid
/project/wiemels_494/sebastian/glioma_gwas/dk_cohort/Results/HG_astrocytoma.PC5.PHENO1.glm.logistic.hybrid\
/project/wiemels_494/sebastian/glioma_gwas/dk_cohort/Results/Pilocytic_astrocytoma_incl_optic.PC5.PHENO1.glm.logistic.hybrid
/project/wiemels_494/sebastian/glioma_gwas/dk_cohort/Results/HGG.PC5.PHENO1.glm.logistic.hybrid\
/project/wiemels_494/sebastian/glioma_gwas/dk_cohort/Results/Pilocytic_astrocytoma.PC5.PHENO1.glm.logistic.hybrid

USC cohort files:
/project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.all.Subs.eur.phenotype.glm.logistic.hybrid
/project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.lgg.Subs.eur.phenotype.glm.logistic.hybrid
/project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.hgg.Subs.eur.phenotype.glm.logistic.hybrid
/project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.lga.Subs.eur.phenotype.glm.logistic.hybrid
/project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.hga.Subs.eur.phenotype.glm.logistic.hybrid
/project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.pa.Subs.eur.phenotype.glm.logistic.hybrid
/project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.pao.Subs.eur.phenotype.glm.logistic.hybrid
/project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.glm.Subs.eur.phenotype.glm.logistic.hybrid
/project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.ast.Subs.eur.phenotype.glm.logistic.hybrid
/project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.olg.Subs.eur.phenotype.glm.logistic.hybrid

/project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.all.Subs.lat.phenotype.glm.logistic.hybrid
/project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.lgg.Subs.lat.phenotype.glm.logistic.hybrid
/project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.hgg.Subs.lat.phenotype.glm.logistic.hybrid
/project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.lga.Subs.lat.phenotype.glm.logistic.hybrid
/project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.hga.Subs.lat.phenotype.glm.logistic.hybrid
/project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.pa.Subs.lat.phenotype.glm.logistic.hybrid
/project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.pao.Subs.lat.phenotype.glm.logistic.hybrid
/project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.glm.Subs.lat.phenotype.glm.logistic.hybrid
/project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.ast.Subs.lat.phenotype.glm.logistic.hybrid
/project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.olg.Subs.lat.phenotype.glm.logistic.hybrid

/project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.all.Subs.blk.phenotype.glm.logistic.hybrid
/project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.lgg.Subs.blk.phenotype.glm.logistic.hybrid
/project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.hgg.Subs.blk.phenotype.glm.logistic.hybrid
/project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.lga.Subs.blk.phenotype.glm.logistic.hybrid
/project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.hga.Subs.blk.phenotype.glm.logistic.hybrid
/project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.pa.Subs.blk.phenotype.glm.logistic.hybrid
/project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.pao.Subs.blk.phenotype.glm.logistic.hybrid
/project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.glm.Subs.blk.phenotype.glm.logistic.hybrid
/project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.ast.Subs.blk.phenotype.glm.logistic.hybrid
/project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.olg.Subs.blk.phenotype.glm.logistic.hybrid

/project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.all.Subs.asn.phenotype.glm.logistic.hybrid
/project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.lgg.Subs.asn.phenotype.glm.logistic.hybrid
/project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.hgg.Subs.asn.phenotype.glm.logistic.hybrid
/project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.lga.Subs.asn.phenotype.glm.logistic.hybrid
/project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.hga.Subs.asn.phenotype.glm.logistic.hybrid
/project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.pa.Subs.asn.phenotype.glm.logistic.hybrid
/project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.pao.Subs.asn.phenotype.glm.logistic.hybrid
/project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.glm.Subs.asn.phenotype.glm.logistic.hybrid
/project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.ast.Subs.asn.phenotype.glm.logistic.hybrid
/project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.olg.Subs.asn.phenotype.glm.logistic.hybrid

Separated by subtypes 1. all subjects
/project/wiemels_494/sebastian/glioma_gwas/anna_cohort/concat.all.var_3PCA.20220401.PHENO1.glm.logistic.hybrid
/project/wiemels_494/sebastian/glioma_gwas/dk_cohort/Results/Glioma.PC5.PHENO1.glm.logistic.hybrid
/project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.all.Subs.eur.phenotype.glm.logistic.hybrid

/project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.all.Subs.lat.phenotype.glm.logistic.hybrid
/project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.all.Subs.blk.phenotype.glm.logistic.hybrid
/project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.all.Subs.asn.phenotype.glm.logistic.hybrid

2.  low grade glioma
    /project/wiemels_494/sebastian/glioma_gwas/anna_cohort/concat.lgg.var_0PCA.20220401.PHENO1.glm.logistic.hybrid
    /project/wiemels_494/sebastian/glioma_gwas/dk_cohort/Results/LGG.PC5.PHENO1.glm.logistic.hybrid
    /project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.lgg.Subs.eur.phenotype.glm.logistic.hybrid

/project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.lgg.Subs.lat.phenotype.glm.logistic.hybrid
/project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.lgg.Subs.blk.phenotype.glm.logistic.hybrid
/project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.lgg.Subs.asn.phenotype.glm.logistic.hybrid

3.  high grade glioma
    /project/wiemels_494/sebastian/glioma_gwas/anna_cohort/concat.hgg.var_1PCA.20220401.PHENO1.glm.logistic.hybrid
    /project/wiemels_494/sebastian/glioma_gwas/dk_cohort/Results/HGG.PC5.PHENO1.glm.logistic.hybrid
    /project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.hgg.Subs.eur.phenotype.glm.logistic.hybrid

/project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.hgg.Subs.lat.phenotype.glm.logistic.hybrid
/project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.hgg.Subs.blk.phenotype.glm.logistic.hybrid
/project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.hgg.Subs.asn.phenotype.glm.logistic.hybrid

4.  astrocytoma
    /project/wiemels_494/sebastian/glioma_gwas/anna_cohort/concat.astro.var_10PCA.20220401.PHENO1.glm.logistic.hybrid
    /project/wiemels_494/sebastian/glioma_gwas/dk_cohort/Results/Astrocytoma.PC5.PHENO1.glm.logistic.hybrid\
    /project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.ast.Subs.eur.phenotype.glm.logistic.hybrid

/project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.ast.Subs.lat.phenotype.glm.logistic.hybrid
/project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.ast.Subs.blk.phenotype.glm.logistic.hybrid
/project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.ast.Subs.asn.phenotype.glm.logistic.hybrid

5.  low grade astrocytoma
    /project/wiemels_494/sebastian/glioma_gwas/anna_cohort/concat.lg_astro.var_8PCA.20220401.PHENO1.glm.logistic.hybrid
    /project/wiemels_494/sebastian/glioma_gwas/dk_cohort/Results/LG_astrocytoma.PC5.PHENO1.glm.logistic.hybrid
    /project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.lga.Subs.eur.phenotype.glm.logistic.hybrid

/project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.lga.Subs.lat.phenotype.glm.logistic.hybrid
/project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.lga.Subs.blk.phenotype.glm.logistic.hybrid
/project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.lga.Subs.asn.phenotype.glm.logistic.hybrid

6.  high grade astrocytoma
    /project/wiemels_494/sebastian/glioma_gwas/anna_cohort/concat.hg_astro.var_3PCA.20220401.PHENO1.glm.logistic.hybrid
    /project/wiemels_494/sebastian/glioma_gwas/dk_cohort/Results/HG_astrocytoma.PC5.PHENO1.glm.logistic.hybrid\
    /project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.hga.Subs.eur.phenotype.glm.logistic.hybrid

/project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.hga.Subs.lat.phenotype.glm.logistic.hybrid
/project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.hga.Subs.blk.phenotype.glm.logistic.hybrid
/project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.hga.Subs.asn.phenotype.glm.logistic.hybrid

7.  pilocytic astrocytoma
    /project/wiemels_494/sebastian/glioma_gwas/anna_cohort/concat.pilo.var_0PCA.20220401.PHENO1.glm.logistic.hybrid
    /project/wiemels_494/sebastian/glioma_gwas/dk_cohort/Results/Pilocytic_astrocytoma.PC5.PHENO1.glm.logistic.hybrid
    /project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.pa.Subs.eur.phenotype.glm.logistic.hybrid

/project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.pa.Subs.lat.phenotype.glm.logistic.hybrid
/project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.pa.Subs.blk.phenotype.glm.logistic.hybrid
/project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.pa.Subs.asn.phenotype.glm.logistic.hybrid

8.  pilocytic astrocytoma + optic pathway
    /project/wiemels_494/sebastian/glioma_gwas/anna_cohort/concat.pilo_optic.var_0PCA.20220401.PHENO1.glm.logistic.hybrid
    /project/wiemels_494/sebastian/glioma_gwas/dk_cohort/Results/Pilocytic_astrocytoma_incl_optic.PC5.PHENO1.glm.logistic.hybrid
    /project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.pao.Subs.eur.phenotype.glm.logistic.hybrid

/project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.pao.Subs.lat.phenotype.glm.logistic.hybrid
/project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.pao.Subs.blk.phenotype.glm.logistic.hybrid
/project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.pao.Subs.asn.phenotype.glm.logistic.hybrid

9.  glioblastoma (asterisk)
    /project/wiemels_494/sebastian/glioma_gwas/anna_cohort/concat.gbm.var_0PCA.20220401.PHENO1.glm.logistic.hybrid
    /project/wiemels_494/sebastian/glioma_gwas/dk_cohort/Results/Glioblastoma.PC5.PHENO1.glm.logistic.hybrid\
    /project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.glm.Subs.eur.phenotype.glm.logistic.hybrid

/project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.glm.Subs.lat.phenotype.glm.logistic.hybrid
/project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.glm.Subs.blk.phenotype.glm.logistic.hybrid
/project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.glm.Subs.asn.phenotype.glm.logistic.hybrid

10. Oligodendroglioma (asterisk)
    /project/wiemels_494/sebastian/glioma_gwas/anna_cohort/concat.oligo.var_0PCA.20220401.PHENO1.glm.logistic.hybrid
    /project/wiemels_494/sebastian/glioma_gwas/dk_cohort/Results/Oligodendroglioma.PC5.PHENO1.glm.logistic.hybrid
    /project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.olg.Subs.eur.phenotype.glm.logistic.hybrid

/project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.olg.Subs.lat.phenotype.glm.logistic.hybrid
/project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.olg.Subs.blk.phenotype.glm.logistic.hybrid
/project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.olg.Subs.asn.phenotype.glm.logistic.hybrid

```

```{bash preprocessing for QC}

cd /project/wiemels_494/sebastian/glioma_gwas

awk '{print $1,$2,$4,$5}' anna_cohort/concat.all.var_3PCA.20220401.PHENO1.glm.logistic.hybrid > anna.snp.temp
awk '{print $1,$2,$4,$5}' usc_cohort/hg19.all.Subs.eur.phenotype.glm.logistic.hybrid | sed 's/chr//g' > usc.snp.temp
awk '{print $1,$2,$4,$5}' dk_cohort/Results/Glioma.PC5.PHENO1.glm.logistic.hybrid > dk.snp.temp

```

```{r QC for metal, all reference the same?}

setwd("/project/wiemels_494/sebastian/glioma_gwas")
library(tidyverse)
library(data.table)

#all subjects
anna = fread("anna.snp.temp") %>%
  as.data.frame() %>%
  dplyr::rename("CHR"="#CHROM",
         "REF_anna"="REF",
         "ALT_anna"="ALT1")
usc = fread("usc.snp.temp")%>%
  dplyr::rename("CHR"="#CHROM",
         "REF_usc"="REF",
         "ALT_usc"="ALT1")
dk = fread("dk.snp.temp")%>%
  dplyr::rename("CHR"="#CHROM",
         "REF_dk"="REF",
         "ALT_dk"="ALT")

comm1 = usc %>%
  inner_join(anna,by=c("CHR","POS"))
comm1_1 = comm1 %>%
  filter(REF_usc == ALT_anna & ALT_usc == REF_anna  )
nrow(comm1_1)
# 1

comm2 = usc %>%
  inner_join(dk,by=c("CHR","POS"))
comm2_1 = comm2 %>%
  filter(REF_usc == ALT_dk & ALT_usc == REF_dk  )
nrow(comm2_1)
# 3368


#write.table(comm1_1,"usc.anna.disprepency.txt",sep='\t',row.names=FALSE,quote=FALSE)
#write.table(comm2_1,"usc.dk.disprepency.txt",sep='\t',row.names=FALSE,quote=FALSE)

#identical(comm$REF_anna,comm$REF_usc)
#identical(comm$REF_dk,comm$REF_usc)
#identical(comm$ALT_anna,comm$ALT_usc)
#identical(comm$ALT_sk,comm$ALT_usc)


```

Method 1: using METAL

```{bash meta-analysis all subjects}

cd /project/wiemels_494/sebastian/glioma_gwas/meta/processedFilesFromCohorts

########################################################################################################################

## ALL SUBJECTS

# modifying file headers
awk -v OFS='\t' 'NR>1  {print $1,$2, $1":"$2":"$4":"$5,$4,$5,$6,$9,$13,$14,$15,$16,$17}' /project/wiemels_494/sebastian/glioma_gwas/anna_cohort/concat.all.var_3PCA.20220401.PHENO1.glm.logistic.hybrid |\
  sed '1i CHROM'\\t'POS'\\t'ID'\\t'REF'\\t'ALT1'\\t'A1'\\t'A1_FREQ'\\t'BETA'\\t'SE'\\t'L95'\\t'U95'\\t'P'  > anna.all.glm.txt

awk -v OFS='\t' 'NR>1 {print $1,$2, $1":"$2":"$4":"$5,$4,$5,$6,$13,$19,$20,$21,$22,$24}' /project/wiemels_494/sebastian/glioma_gwas/dk_cohort/Results/Glioma.PC5.PHENO1.glm.logistic.hybrid|\
  sed '1i CHROM'\\t'POS'\\t'ID'\\t'REF'\\t'ALT1'\\t'A1'\\t'A1_FREQ'\\t'BETA'\\t'SE'\\t'L95'\\t'U95'\\t'P'  >  dk.all.glm.txt

sed 's/chr//g' /project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.all.Subs.eur.phenotype.glm.logistic.hybrid |\
awk -v OFS='\t' 'NR>1  {print $1,$2, $1":"$2":"$4":"$5,$4,$5,$6,$9,$13,$14,$15,$16,$17}' |\
  sed '1i CHROM'\\t'POS'\\t'ID'\\t'REF'\\t'ALT1'\\t'A1'\\t'A1_FREQ'\\t'BETA'\\t'SE'\\t'L95'\\t'U95'\\t'P' > usc.all.eur.glm.txt

sed 's/chr//g' /project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.all.Subs.lat.phenotype.glm.logistic.hybrid |\
awk -v OFS='\t' 'NR>1  {print $1,$2, $1":"$2":"$4":"$5,$4,$5,$6,$9,$13,$14,$15,$16,$17}' |\
  sed '1i CHROM'\\t'POS'\\t'ID'\\t'REF'\\t'ALT1'\\t'A1'\\t'A1_FREQ'\\t'BETA'\\t'SE'\\t'L95'\\t'U95'\\t'P' > usc.all.lat.glm.txt

sed 's/chr//g' /project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.all.Subs.blk.phenotype.glm.logistic.hybrid |\
awk -v OFS='\t' 'NR>1  {print $1,$2, $1":"$2":"$4":"$5,$4,$5,$6,$9,$13,$14,$15,$16,$17}' |\
  sed '1i CHROM'\\t'POS'\\t'ID'\\t'REF'\\t'ALT1'\\t'A1'\\t'A1_FREQ'\\t'BETA'\\t'SE'\\t'L95'\\t'U95'\\t'P' > usc.all.blk.glm.txt

sed 's/chr//g' /project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.all.Subs.asn.phenotype.glm.logistic.hybrid |\
awk -v OFS='\t' 'NR>1  {print $1,$2, $1":"$2":"$4":"$5,$4,$5,$6,$9,$13,$14,$15,$16,$17}' |\
  sed '1i CHROM'\\t'POS'\\t'ID'\\t'REF'\\t'ALT1'\\t'A1'\\t'A1_FREQ'\\t'BETA'\\t'SE'\\t'L95'\\t'U95'\\t'P' > usc.all.asn.glm.txt

#grep "146486211" /project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.all.Subs.eur.phenotype.glm.logistic.hybrid

cd /project/wiemels_494/sebastian/glioma_gwas/meta/METAL

## Europeans only

metal
MARKER ID
EFFECT BETA
PVALUE P
SCHEME STDERR
STDERR SE
VERBOSE OFF
GENOMICCONTROL OFF
PROCESS ../processedFilesFromCohorts/anna.all.glm.txt
PROCESS ../processedFilesFromCohorts/dk.all.glm.txt
PROCESS ../processedFilesFromCohorts/usc.all.eur.glm.txt
ANALYZE HETEROGENEITY
QUIT
mv METAANALYSIS1.TBL meta.eur.all.glioma.TBL
mv METAANALYSIS1.TBL.info meta.eur.all.glioma.TBL.info
  
## All races  

metal
MARKER ID
EFFECT BETA
PVALUE P
SCHEME STDERR
STDERR SE
VERBOSE OFF
GENOMICCONTROL OFF
PROCESS ../processedFilesFromCohorts/anna.all.glm.txt
PROCESS ../processedFilesFromCohorts/dk.all.glm.txt
PROCESS ../processedFilesFromCohorts/usc.all.eur.glm.txt
PROCESS ../processedFilesFromCohorts/usc.all.lat.glm.txt
PROCESS ../processedFilesFromCohorts/usc.all.blk.glm.txt
PROCESS ../processedFilesFromCohorts/usc.all.asn.glm.txt
ANALYZE HETEROGENEITY
QUIT
mv METAANALYSIS1.TBL meta.four.races.all.glioma.TBL
mv METAANALYSIS1.TBL.info meta.four.races.all.glioma.TBL.info


```

```{bash meta-analysis low grade glioma}

cd /project/wiemels_494/sebastian/glioma_gwas/meta/processedFilesFromCohorts

########################################################################################################################

# modifying file headers
awk -v OFS='\t' 'NR>1  {print $1,$2, $1":"$2":"$4":"$5,$4,$5,$6,$9,$13,$14,$15,$16,$17}' /project/wiemels_494/sebastian/glioma_gwas/anna_cohort/concat.lgg.var_0PCA.20220401.PHENO1.glm.logistic.hybrid |\
  sed '1i CHROM'\\t'POS'\\t'ID'\\t'REF'\\t'ALT1'\\t'A1'\\t'A1_FREQ'\\t'BETA'\\t'SE'\\t'L95'\\t'U95'\\t'P'  > anna.lgg.glm.txt

awk -v OFS='\t' 'NR>1 {print $1,$2, $1":"$2":"$4":"$5,$4,$5,$6,$13,$19,$20,$21,$22,$24}' /project/wiemels_494/sebastian/glioma_gwas/dk_cohort/Results/LGG.PC5.PHENO1.glm.logistic.hybrid|\
  sed '1i CHROM'\\t'POS'\\t'ID'\\t'REF'\\t'ALT1'\\t'A1'\\t'A1_FREQ'\\t'BETA'\\t'SE'\\t'L95'\\t'U95'\\t'P'  >  dk.lgg.glm.txt

sed 's/chr//g' /project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.lgg.Subs.eur.phenotype.glm.logistic.hybrid |\
awk -v OFS='\t' 'NR>1  {print $1,$2, $1":"$2":"$4":"$5,$4,$5,$6,$9,$13,$14,$15,$16,$17}' |\
  sed '1i CHROM'\\t'POS'\\t'ID'\\t'REF'\\t'ALT1'\\t'A1'\\t'A1_FREQ'\\t'BETA'\\t'SE'\\t'L95'\\t'U95'\\t'P' > usc.lgg.eur.glm.txt

sed 's/chr//g' /project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.lgg.Subs.lat.phenotype.glm.logistic.hybrid |\
awk -v OFS='\t' 'NR>1  {print $1,$2, $1":"$2":"$4":"$5,$4,$5,$6,$9,$13,$14,$15,$16,$17}' |\
  sed '1i CHROM'\\t'POS'\\t'ID'\\t'REF'\\t'ALT1'\\t'A1'\\t'A1_FREQ'\\t'BETA'\\t'SE'\\t'L95'\\t'U95'\\t'P' > usc.lgg.lat.glm.txt

sed 's/chr//g' /project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.lgg.Subs.blk.phenotype.glm.logistic.hybrid |\
awk -v OFS='\t' 'NR>1  {print $1,$2, $1":"$2":"$4":"$5,$4,$5,$6,$9,$13,$14,$15,$16,$17}' |\
  sed '1i CHROM'\\t'POS'\\t'ID'\\t'REF'\\t'ALT1'\\t'A1'\\t'A1_FREQ'\\t'BETA'\\t'SE'\\t'L95'\\t'U95'\\t'P' > usc.lgg.blk.glm.txt

sed 's/chr//g' /project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.lgg.Subs.asn.phenotype.glm.logistic.hybrid |\
awk -v OFS='\t' 'NR>1  {print $1,$2, $1":"$2":"$4":"$5,$4,$5,$6,$9,$13,$14,$15,$16,$17}' |\
  sed '1i CHROM'\\t'POS'\\t'ID'\\t'REF'\\t'ALT1'\\t'A1'\\t'A1_FREQ'\\t'BETA'\\t'SE'\\t'L95'\\t'U95'\\t'P' > usc.lgg.asn.glm.txt


cd /project/wiemels_494/sebastian/glioma_gwas/meta/METAL

## Europeans only

metal
MARKER ID
EFFECT BETA
PVALUE P
SCHEME STDERR
STDERR SE
VERBOSE OFF
GENOMICCONTROL OFF
PROCESS ../processedFilesFromCohorts/anna.lgg.glm.txt
PROCESS ../processedFilesFromCohorts/dk.lgg.glm.txt
PROCESS ../processedFilesFromCohorts/usc.lgg.eur.glm.txt
ANALYZE HETEROGENEITY
QUIT
mv METAANALYSIS1.TBL meta.eur.lgg.glioma.TBL
mv METAANALYSIS1.TBL.info meta.eur.lgg.glioma.TBL.info
  
## All races  

metal
MARKER ID
EFFECT BETA
PVALUE P
SCHEME STDERR
STDERR SE
VERBOSE OFF
GENOMICCONTROL OFF
PROCESS ../processedFilesFromCohorts/anna.lgg.glm.txt
PROCESS ../processedFilesFromCohorts/dk.lgg.glm.txt
PROCESS ../processedFilesFromCohorts/usc.lgg.eur.glm.txt
PROCESS ../processedFilesFromCohorts/usc.lgg.lat.glm.txt
PROCESS ../processedFilesFromCohorts/usc.lgg.blk.glm.txt
PROCESS ../processedFilesFromCohorts/usc.lgg.asn.glm.txt
ANALYZE HETEROGENEITY
QUIT
mv METAANALYSIS1.TBL meta.four.races.lgg.glioma.TBL
mv METAANALYSIS1.TBL.info meta.four.races.lgg.glioma.TBL.info


```

```{bash meta-analysis high grade glioma}

cd /project/wiemels_494/sebastian/glioma_gwas/meta/processedFilesFromCohorts

########################################################################################################################

# modifying file headers
awk -v OFS='\t' 'NR>1  {print $1,$2, $1":"$2":"$4":"$5,$4,$5,$6,$9,$13,$14,$15,$16,$17}' /project/wiemels_494/sebastian/glioma_gwas/anna_cohort/concat.hgg.var_1PCA.20220401.PHENO1.glm.logistic.hybrid |\
  sed '1i CHROM'\\t'POS'\\t'ID'\\t'REF'\\t'ALT1'\\t'A1'\\t'A1_FREQ'\\t'BETA'\\t'SE'\\t'L95'\\t'U95'\\t'P'  > anna.hgg.glm.txt

awk -v OFS='\t' 'NR>1 {print $1,$2, $1":"$2":"$4":"$5,$4,$5,$6,$13,$19,$20,$21,$22,$24}' /project/wiemels_494/sebastian/glioma_gwas/dk_cohort/Results/HGG.PC5.PHENO1.glm.logistic.hybrid|\
  sed '1i CHROM'\\t'POS'\\t'ID'\\t'REF'\\t'ALT1'\\t'A1'\\t'A1_FREQ'\\t'BETA'\\t'SE'\\t'L95'\\t'U95'\\t'P'  >  dk.hgg.glm.txt

sed 's/chr//g' /project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.hgg.Subs.eur.phenotype.glm.logistic.hybrid |\
awk -v OFS='\t' 'NR>1  {print $1,$2, $1":"$2":"$4":"$5,$4,$5,$6,$9,$13,$14,$15,$16,$17}' |\
  sed '1i CHROM'\\t'POS'\\t'ID'\\t'REF'\\t'ALT1'\\t'A1'\\t'A1_FREQ'\\t'BETA'\\t'SE'\\t'L95'\\t'U95'\\t'P' > usc.hgg.eur.glm.txt

sed 's/chr//g' /project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.hgg.Subs.lat.phenotype.glm.logistic.hybrid |\
awk -v OFS='\t' 'NR>1  {print $1,$2, $1":"$2":"$4":"$5,$4,$5,$6,$9,$13,$14,$15,$16,$17}' |\
  sed '1i CHROM'\\t'POS'\\t'ID'\\t'REF'\\t'ALT1'\\t'A1'\\t'A1_FREQ'\\t'BETA'\\t'SE'\\t'L95'\\t'U95'\\t'P' > usc.hgg.lat.glm.txt

sed 's/chr//g' /project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.hgg.Subs.blk.phenotype.glm.logistic.hybrid |\
awk -v OFS='\t' 'NR>1  {print $1,$2, $1":"$2":"$4":"$5,$4,$5,$6,$9,$13,$14,$15,$16,$17}' |\
  sed '1i CHROM'\\t'POS'\\t'ID'\\t'REF'\\t'ALT1'\\t'A1'\\t'A1_FREQ'\\t'BETA'\\t'SE'\\t'L95'\\t'U95'\\t'P' > usc.hgg.blk.glm.txt

sed 's/chr//g' /project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.hgg.Subs.asn.phenotype.glm.logistic.hybrid |\
awk -v OFS='\t' 'NR>1  {print $1,$2, $1":"$2":"$4":"$5,$4,$5,$6,$9,$13,$14,$15,$16,$17}' |\
  sed '1i CHROM'\\t'POS'\\t'ID'\\t'REF'\\t'ALT1'\\t'A1'\\t'A1_FREQ'\\t'BETA'\\t'SE'\\t'L95'\\t'U95'\\t'P' > usc.hgg.asn.glm.txt


cd /project/wiemels_494/sebastian/glioma_gwas/meta/METAL

## Europeans only

metal
MARKER ID
EFFECT BETA
PVALUE P
SCHEME STDERR
STDERR SE
VERBOSE OFF
GENOMICCONTROL OFF
PROCESS ../processedFilesFromCohorts/anna.hgg.glm.txt
PROCESS ../processedFilesFromCohorts/dk.hgg.glm.txt
PROCESS ../processedFilesFromCohorts/usc.hgg.eur.glm.txt
ANALYZE HETEROGENEITY
QUIT
mv METAANALYSIS1.TBL meta.eur.hgg.glioma.TBL
mv METAANALYSIS1.TBL.info meta.eur.hgg.glioma.TBL.info
  
## All races  

metal
MARKER ID
EFFECT BETA
PVALUE P
SCHEME STDERR
STDERR SE
VERBOSE OFF
GENOMICCONTROL OFF
PROCESS ../processedFilesFromCohorts/anna.hgg.glm.txt
PROCESS ../processedFilesFromCohorts/dk.hgg.glm.txt
PROCESS ../processedFilesFromCohorts/usc.hgg.eur.glm.txt
PROCESS ../processedFilesFromCohorts/usc.hgg.lat.glm.txt
PROCESS ../processedFilesFromCohorts/usc.hgg.blk.glm.txt
PROCESS ../processedFilesFromCohorts/usc.hgg.asn.glm.txt
ANALYZE HETEROGENEITY
QUIT
mv METAANALYSIS1.TBL meta.four.races.hgg.glioma.TBL
mv METAANALYSIS1.TBL.info meta.four.races.hgg.glioma.TBL.info


```

```{bash meta-analysis astrocytoma}

cd /project/wiemels_494/sebastian/glioma_gwas/meta/processedFilesFromCohorts

########################################################################################################################

# modifying file headers
awk -v OFS='\t' 'NR>1  {print $1,$2, $1":"$2":"$4":"$5,$4,$5,$6,$9,$13,$14,$15,$16,$17}' /project/wiemels_494/sebastian/glioma_gwas/anna_cohort/concat.astro.var_10PCA.20220401.PHENO1.glm.logistic.hybrid |\
  sed '1i CHROM'\\t'POS'\\t'ID'\\t'REF'\\t'ALT1'\\t'A1'\\t'A1_FREQ'\\t'BETA'\\t'SE'\\t'L95'\\t'U95'\\t'P'  > anna.ast.glm.txt

awk -v OFS='\t' 'NR>1 {print $1,$2, $1":"$2":"$4":"$5,$4,$5,$6,$13,$19,$20,$21,$22,$24}' /project/wiemels_494/sebastian/glioma_gwas/dk_cohort/Results/Astrocytoma.PC5.PHENO1.glm.logistic.hybrid|\
  sed '1i CHROM'\\t'POS'\\t'ID'\\t'REF'\\t'ALT1'\\t'A1'\\t'A1_FREQ'\\t'BETA'\\t'SE'\\t'L95'\\t'U95'\\t'P'  >  dk.ast.glm.txt

sed 's/chr//g' /project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.ast.Subs.eur.phenotype.glm.logistic.hybrid |\
awk -v OFS='\t' 'NR>1  {print $1,$2, $1":"$2":"$4":"$5,$4,$5,$6,$9,$13,$14,$15,$16,$17}' |\
  sed '1i CHROM'\\t'POS'\\t'ID'\\t'REF'\\t'ALT1'\\t'A1'\\t'A1_FREQ'\\t'BETA'\\t'SE'\\t'L95'\\t'U95'\\t'P' > usc.ast.eur.glm.txt

sed 's/chr//g' /project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.ast.Subs.lat.phenotype.glm.logistic.hybrid |\
awk -v OFS='\t' 'NR>1  {print $1,$2, $1":"$2":"$4":"$5,$4,$5,$6,$9,$13,$14,$15,$16,$17}' |\
  sed '1i CHROM'\\t'POS'\\t'ID'\\t'REF'\\t'ALT1'\\t'A1'\\t'A1_FREQ'\\t'BETA'\\t'SE'\\t'L95'\\t'U95'\\t'P' > usc.ast.lat.glm.txt

sed 's/chr//g' /project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.ast.Subs.blk.phenotype.glm.logistic.hybrid |\
awk -v OFS='\t' 'NR>1  {print $1,$2, $1":"$2":"$4":"$5,$4,$5,$6,$9,$13,$14,$15,$16,$17}' |\
  sed '1i CHROM'\\t'POS'\\t'ID'\\t'REF'\\t'ALT1'\\t'A1'\\t'A1_FREQ'\\t'BETA'\\t'SE'\\t'L95'\\t'U95'\\t'P' > usc.ast.blk.glm.txt

sed 's/chr//g' /project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.ast.Subs.asn.phenotype.glm.logistic.hybrid |\
awk -v OFS='\t' 'NR>1  {print $1,$2, $1":"$2":"$4":"$5,$4,$5,$6,$9,$13,$14,$15,$16,$17}' |\
  sed '1i CHROM'\\t'POS'\\t'ID'\\t'REF'\\t'ALT1'\\t'A1'\\t'A1_FREQ'\\t'BETA'\\t'SE'\\t'L95'\\t'U95'\\t'P' > usc.ast.asn.glm.txt


cd /project/wiemels_494/sebastian/glioma_gwas/meta/METAL

## Europeans only

metal
MARKER ID
EFFECT BETA
PVALUE P
SCHEME STDERR
STDERR SE
VERBOSE OFF
GENOMICCONTROL OFF
PROCESS ../processedFilesFromCohorts/anna.ast.glm.txt
PROCESS ../processedFilesFromCohorts/dk.ast.glm.txt
PROCESS ../processedFilesFromCohorts/usc.ast.eur.glm.txt
ANALYZE HETEROGENEITY
QUIT
mv METAANALYSIS1.TBL meta.eur.ast.glioma.TBL
mv METAANALYSIS1.TBL.info meta.eur.ast.glioma.TBL.info
  
## All races  

metal
MARKER ID
EFFECT BETA
PVALUE P
SCHEME STDERR
STDERR SE
VERBOSE OFF
GENOMICCONTROL OFF
PROCESS ../processedFilesFromCohorts/anna.ast.glm.txt
PROCESS ../processedFilesFromCohorts/dk.ast.glm.txt
PROCESS ../processedFilesFromCohorts/usc.ast.eur.glm.txt
PROCESS ../processedFilesFromCohorts/usc.ast.lat.glm.txt
PROCESS ../processedFilesFromCohorts/usc.ast.blk.glm.txt
PROCESS ../processedFilesFromCohorts/usc.ast.asn.glm.txt
ANALYZE HETEROGENEITY
QUIT
mv METAANALYSIS1.TBL meta.four.races.ast.glioma.TBL
mv METAANALYSIS1.TBL.info meta.four.races.ast.glioma.TBL.info

## non europeans

metal
MARKER ID
EFFECT BETA
PVALUE P
SCHEME STDERR
STDERR SE
VERBOSE OFF
GENOMICCONTROL OFF
PROCESS ../processedFilesFromCohorts/usc.ast.lat.glm.txt
PROCESS ../processedFilesFromCohorts/usc.ast.blk.glm.txt
PROCESS ../processedFilesFromCohorts/usc.ast.asn.glm.txt
ANALYZE HETEROGENEITY
QUIT
mv METAANALYSIS1.TBL meta.nonwhites.ast.glioma.TBL
mv METAANALYSIS1.TBL.info meta.nonwhites.ast.glioma.TBL.info

```

```{bash meta-analysis low grade astrocytoma}

cd /project/wiemels_494/sebastian/glioma_gwas/meta/processedFilesFromCohorts

########################################################################################################################

# modifying file headers
awk -v OFS='\t' 'NR>1  {print $1,$2, $1":"$2":"$4":"$5,$4,$5,$6,$9,$13,$14,$15,$16,$17}' /project/wiemels_494/sebastian/glioma_gwas/anna_cohort/concat.lg_astro.var_8PCA.20220401.PHENO1.glm.logistic.hybrid |\
  sed '1i CHROM'\\t'POS'\\t'ID'\\t'REF'\\t'ALT1'\\t'A1'\\t'A1_FREQ'\\t'BETA'\\t'SE'\\t'L95'\\t'U95'\\t'P'  > anna.lga.glm.txt

awk -v OFS='\t' 'NR>1 {print $1,$2, $1":"$2":"$4":"$5,$4,$5,$6,$13,$19,$20,$21,$22,$24}' /project/wiemels_494/sebastian/glioma_gwas/dk_cohort/Results/LG_astrocytoma.PC5.PHENO1.glm.logistic.hybrid|\
  sed '1i CHROM'\\t'POS'\\t'ID'\\t'REF'\\t'ALT1'\\t'A1'\\t'A1_FREQ'\\t'BETA'\\t'SE'\\t'L95'\\t'U95'\\t'P'  >  dk.lga.glm.txt

sed 's/chr//g' /project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.lga.Subs.eur.phenotype.glm.logistic.hybrid |\
awk -v OFS='\t' 'NR>1  {print $1,$2, $1":"$2":"$4":"$5,$4,$5,$6,$9,$13,$14,$15,$16,$17}' |\
  sed '1i CHROM'\\t'POS'\\t'ID'\\t'REF'\\t'ALT1'\\t'A1'\\t'A1_FREQ'\\t'BETA'\\t'SE'\\t'L95'\\t'U95'\\t'P' > usc.lga.eur.glm.txt

sed 's/chr//g' /project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.lga.Subs.lat.phenotype.glm.logistic.hybrid |\
awk -v OFS='\t' 'NR>1  {print $1,$2, $1":"$2":"$4":"$5,$4,$5,$6,$9,$13,$14,$15,$16,$17}' |\
  sed '1i CHROM'\\t'POS'\\t'ID'\\t'REF'\\t'ALT1'\\t'A1'\\t'A1_FREQ'\\t'BETA'\\t'SE'\\t'L95'\\t'U95'\\t'P' > usc.lga.lat.glm.txt

sed 's/chr//g' /project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.lga.Subs.blk.phenotype.glm.logistic.hybrid |\
awk -v OFS='\t' 'NR>1  {print $1,$2, $1":"$2":"$4":"$5,$4,$5,$6,$9,$13,$14,$15,$16,$17}' |\
  sed '1i CHROM'\\t'POS'\\t'ID'\\t'REF'\\t'ALT1'\\t'A1'\\t'A1_FREQ'\\t'BETA'\\t'SE'\\t'L95'\\t'U95'\\t'P' > usc.lga.blk.glm.txt

sed 's/chr//g' /project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.lga.Subs.asn.phenotype.glm.logistic.hybrid |\
awk -v OFS='\t' 'NR>1  {print $1,$2, $1":"$2":"$4":"$5,$4,$5,$6,$9,$13,$14,$15,$16,$17}' |\
  sed '1i CHROM'\\t'POS'\\t'ID'\\t'REF'\\t'ALT1'\\t'A1'\\t'A1_FREQ'\\t'BETA'\\t'SE'\\t'L95'\\t'U95'\\t'P' > usc.lga.asn.glm.txt


cd /project/wiemels_494/sebastian/glioma_gwas/meta/METAL

## Europeans only

metal
MARKER ID
EFFECT BETA
PVALUE P
SCHEME STDERR
STDERR SE
VERBOSE OFF
GENOMICCONTROL OFF
PROCESS ../processedFilesFromCohorts/anna.lga.glm.txt
PROCESS ../processedFilesFromCohorts/dk.lga.glm.txt
PROCESS ../processedFilesFromCohorts/usc.lga.eur.glm.txt
ANALYZE HETEROGENEITY
QUIT
mv METAANALYSIS1.TBL meta.eur.lga.glioma.TBL
mv METAANALYSIS1.TBL.info meta.eur.lga.glioma.TBL.info
  
## All races  

metal
MARKER ID
EFFECT BETA
PVALUE P
SCHEME STDERR
STDERR SE
VERBOSE OFF
GENOMICCONTROL OFF
PROCESS ../processedFilesFromCohorts/anna.lga.glm.txt
PROCESS ../processedFilesFromCohorts/dk.lga.glm.txt
PROCESS ../processedFilesFromCohorts/usc.lga.eur.glm.txt
PROCESS ../processedFilesFromCohorts/usc.lga.lat.glm.txt
PROCESS ../processedFilesFromCohorts/usc.lga.blk.glm.txt
PROCESS ../processedFilesFromCohorts/usc.lga.asn.glm.txt
ANALYZE HETEROGENEITY
QUIT
mv METAANALYSIS1.TBL meta.four.races.lga.glioma.TBL
mv METAANALYSIS1.TBL.info meta.four.races.lga.glioma.TBL.info


```

```{bash meta-analysis high grade astrocytoma}

cd /project/wiemels_494/sebastian/glioma_gwas/meta/processedFilesFromCohorts

########################################################################################################################

# modifying file headers
awk -v OFS='\t' 'NR>1  {print $1,$2, $1":"$2":"$4":"$5,$4,$5,$6,$9,$13,$14,$15,$16,$17}' /project/wiemels_494/sebastian/glioma_gwas/anna_cohort/concat.hg_astro.var_3PCA.20220401.PHENO1.glm.logistic.hybrid |\
  sed '1i CHROM'\\t'POS'\\t'ID'\\t'REF'\\t'ALT1'\\t'A1'\\t'A1_FREQ'\\t'BETA'\\t'SE'\\t'L95'\\t'U95'\\t'P'  > anna.hga.glm.txt

awk -v OFS='\t' 'NR>1 {print $1,$2, $1":"$2":"$4":"$5,$4,$5,$6,$13,$19,$20,$21,$22,$24}' /project/wiemels_494/sebastian/glioma_gwas/dk_cohort/Results/HG_astrocytoma.PC5.PHENO1.glm.logistic.hybrid|\
  sed '1i CHROM'\\t'POS'\\t'ID'\\t'REF'\\t'ALT1'\\t'A1'\\t'A1_FREQ'\\t'BETA'\\t'SE'\\t'L95'\\t'U95'\\t'P'  >  dk.hga.glm.txt

sed 's/chr//g' /project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.hga.Subs.eur.phenotype.glm.logistic.hybrid |\
awk -v OFS='\t' 'NR>1  {print $1,$2, $1":"$2":"$4":"$5,$4,$5,$6,$9,$13,$14,$15,$16,$17}' |\
  sed '1i CHROM'\\t'POS'\\t'ID'\\t'REF'\\t'ALT1'\\t'A1'\\t'A1_FREQ'\\t'BETA'\\t'SE'\\t'L95'\\t'U95'\\t'P' > usc.hga.eur.glm.txt

sed 's/chr//g' /project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.hga.Subs.lat.phenotype.glm.logistic.hybrid |\
awk -v OFS='\t' 'NR>1  {print $1,$2, $1":"$2":"$4":"$5,$4,$5,$6,$9,$13,$14,$15,$16,$17}' |\
  sed '1i CHROM'\\t'POS'\\t'ID'\\t'REF'\\t'ALT1'\\t'A1'\\t'A1_FREQ'\\t'BETA'\\t'SE'\\t'L95'\\t'U95'\\t'P' > usc.hga.lat.glm.txt

sed 's/chr//g' /project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.hga.Subs.blk.phenotype.glm.logistic.hybrid |\
awk -v OFS='\t' 'NR>1  {print $1,$2, $1":"$2":"$4":"$5,$4,$5,$6,$9,$13,$14,$15,$16,$17}' |\
  sed '1i CHROM'\\t'POS'\\t'ID'\\t'REF'\\t'ALT1'\\t'A1'\\t'A1_FREQ'\\t'BETA'\\t'SE'\\t'L95'\\t'U95'\\t'P' > usc.hga.blk.glm.txt

sed 's/chr//g' /project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.hga.Subs.asn.phenotype.glm.logistic.hybrid |\
awk -v OFS='\t' 'NR>1  {print $1,$2, $1":"$2":"$4":"$5,$4,$5,$6,$9,$13,$14,$15,$16,$17}' |\
  sed '1i CHROM'\\t'POS'\\t'ID'\\t'REF'\\t'ALT1'\\t'A1'\\t'A1_FREQ'\\t'BETA'\\t'SE'\\t'L95'\\t'U95'\\t'P' > usc.hga.asn.glm.txt


cd /project/wiemels_494/sebastian/glioma_gwas/meta/METAL

## Europeans only

metal
MARKER ID
EFFECT BETA
PVALUE P
SCHEME STDERR
STDERR SE
VERBOSE OFF
GENOMICCONTROL OFF
PROCESS ../processedFilesFromCohorts/anna.hga.glm.txt
PROCESS ../processedFilesFromCohorts/dk.hga.glm.txt
PROCESS ../processedFilesFromCohorts/usc.hga.eur.glm.txt
ANALYZE HETEROGENEITY
QUIT
mv METAANALYSIS1.TBL meta.eur.hga.glioma.TBL
mv METAANALYSIS1.TBL.info meta.eur.hga.glioma.TBL.info
  
## All races  

metal
MARKER ID
EFFECT BETA
PVALUE P
SCHEME STDERR
STDERR SE
VERBOSE OFF
GENOMICCONTROL OFF
PROCESS ../processedFilesFromCohorts/anna.hga.glm.txt
PROCESS ../processedFilesFromCohorts/dk.hga.glm.txt
PROCESS ../processedFilesFromCohorts/usc.hga.eur.glm.txt
PROCESS ../processedFilesFromCohorts/usc.hga.lat.glm.txt
PROCESS ../processedFilesFromCohorts/usc.hga.blk.glm.txt
PROCESS ../processedFilesFromCohorts/usc.hga.asn.glm.txt
ANALYZE HETEROGENEITY
QUIT
mv METAANALYSIS1.TBL meta.four.races.hga.glioma.TBL
mv METAANALYSIS1.TBL.info meta.four.races.hga.glioma.TBL.info


```

```{bash meta-analysis pilocytic astrocytoma}

cd /project/wiemels_494/sebastian/glioma_gwas/meta/processedFilesFromCohorts

########################################################################################################################

# modifying file headers
awk -v OFS='\t' 'NR>1  {print $1,$2, $1":"$2":"$4":"$5,$4,$5,$6,$9,$13,$14,$15,$16,$17}' /project/wiemels_494/sebastian/glioma_gwas/anna_cohort/concat.pilo.var_0PCA.20220401.PHENO1.glm.logistic.hybrid |\
  sed '1i CHROM'\\t'POS'\\t'ID'\\t'REF'\\t'ALT1'\\t'A1'\\t'A1_FREQ'\\t'BETA'\\t'SE'\\t'L95'\\t'U95'\\t'P'  > anna.pa.glm.txt

awk -v OFS='\t' 'NR>1 {print $1,$2, $1":"$2":"$4":"$5,$4,$5,$6,$13,$19,$20,$21,$22,$24}' /project/wiemels_494/sebastian/glioma_gwas/dk_cohort/Results/Pilocytic_astrocytoma.PC5.PHENO1.glm.logistic.hybrid |\
  sed '1i CHROM'\\t'POS'\\t'ID'\\t'REF'\\t'ALT1'\\t'A1'\\t'A1_FREQ'\\t'BETA'\\t'SE'\\t'L95'\\t'U95'\\t'P'  >  dk.pa.glm.txt

sed 's/chr//g' /project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.pa.Subs.eur.phenotype.glm.logistic.hybrid |\
awk -v OFS='\t' 'NR>1  {print $1,$2, $1":"$2":"$4":"$5,$4,$5,$6,$9,$13,$14,$15,$16,$17}' |\
  sed '1i CHROM'\\t'POS'\\t'ID'\\t'REF'\\t'ALT1'\\t'A1'\\t'A1_FREQ'\\t'BETA'\\t'SE'\\t'L95'\\t'U95'\\t'P' > usc.pa.eur.glm.txt

sed 's/chr//g' /project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.pa.Subs.lat.phenotype.glm.logistic.hybrid |\
awk -v OFS='\t' 'NR>1  {print $1,$2, $1":"$2":"$4":"$5,$4,$5,$6,$9,$13,$14,$15,$16,$17}' |\
  sed '1i CHROM'\\t'POS'\\t'ID'\\t'REF'\\t'ALT1'\\t'A1'\\t'A1_FREQ'\\t'BETA'\\t'SE'\\t'L95'\\t'U95'\\t'P' > usc.pa.lat.glm.txt

sed 's/chr//g' /project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.pa.Subs.blk.phenotype.glm.logistic.hybrid |\
awk -v OFS='\t' 'NR>1  {print $1,$2, $1":"$2":"$4":"$5,$4,$5,$6,$9,$13,$14,$15,$16,$17}' |\
  sed '1i CHROM'\\t'POS'\\t'ID'\\t'REF'\\t'ALT1'\\t'A1'\\t'A1_FREQ'\\t'BETA'\\t'SE'\\t'L95'\\t'U95'\\t'P' > usc.pa.blk.glm.txt

sed 's/chr//g' /project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.pa.Subs.asn.phenotype.glm.logistic.hybrid |\
awk -v OFS='\t' 'NR>1  {print $1,$2, $1":"$2":"$4":"$5,$4,$5,$6,$9,$13,$14,$15,$16,$17}' |\
  sed '1i CHROM'\\t'POS'\\t'ID'\\t'REF'\\t'ALT1'\\t'A1'\\t'A1_FREQ'\\t'BETA'\\t'SE'\\t'L95'\\t'U95'\\t'P' > usc.pa.asn.glm.txt


cd /project/wiemels_494/sebastian/glioma_gwas/meta/METAL

## Europeans only

metal
MARKER ID
EFFECT BETA
PVALUE P
SCHEME STDERR
STDERR SE
VERBOSE OFF
GENOMICCONTROL OFF
PROCESS ../processedFilesFromCohorts/anna.pa.glm.txt
PROCESS ../processedFilesFromCohorts/dk.pa.glm.txt
PROCESS ../processedFilesFromCohorts/usc.pa.eur.glm.txt
ANALYZE HETEROGENEITY
QUIT
mv METAANALYSIS1.TBL meta.eur.pa.glioma.TBL
mv METAANALYSIS1.TBL.info meta.eur.pa.glioma.TBL.info
  
## All races  

metal
MARKER ID
EFFECT BETA
PVALUE P
SCHEME STDERR
STDERR SE
VERBOSE OFF
GENOMICCONTROL OFF
PROCESS ../processedFilesFromCohorts/anna.pa.glm.txt
PROCESS ../processedFilesFromCohorts/dk.pa.glm.txt
PROCESS ../processedFilesFromCohorts/usc.pa.eur.glm.txt
PROCESS ../processedFilesFromCohorts/usc.pa.lat.glm.txt
PROCESS ../processedFilesFromCohorts/usc.pa.blk.glm.txt
PROCESS ../processedFilesFromCohorts/usc.pa.asn.glm.txt
ANALYZE HETEROGENEITY
QUIT
mv METAANALYSIS1.TBL meta.four.races.pa.glioma.TBL
mv METAANALYSIS1.TBL.info meta.four.races.pa.glioma.TBL.info


```

```{bash meta-analysis pilocytic astrocytoma + optic pathway }

cd /project/wiemels_494/sebastian/glioma_gwas/meta/processedFilesFromCohorts

########################################################################################################################

# modifying file headers
awk -v OFS='\t' 'NR>1  {print $1,$2, $1":"$2":"$4":"$5,$4,$5,$6,$9,$13,$14,$15,$16,$17}' /project/wiemels_494/sebastian/glioma_gwas/anna_cohort/concat.pilo_optic.var_0PCA.20220401.PHENO1.glm.logistic.hybrid |\
  sed '1i CHROM'\\t'POS'\\t'ID'\\t'REF'\\t'ALT1'\\t'A1'\\t'A1_FREQ'\\t'BETA'\\t'SE'\\t'L95'\\t'U95'\\t'P'  > anna.pao.glm.txt

awk -v OFS='\t' 'NR>1 {print $1,$2, $1":"$2":"$4":"$5,$4,$5,$6,$13,$19,$20,$21,$22,$24}' /project/wiemels_494/sebastian/glioma_gwas/dk_cohort/Results/Pilocytic_astrocytoma_incl_optic.PC5.PHENO1.glm.logistic.hybrid |\
  sed '1i CHROM'\\t'POS'\\t'ID'\\t'REF'\\t'ALT1'\\t'A1'\\t'A1_FREQ'\\t'BETA'\\t'SE'\\t'L95'\\t'U95'\\t'P'  >  dk.pao.glm.txt

sed 's/chr//g' /project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.pao.Subs.eur.phenotype.glm.logistic.hybrid |\
awk -v OFS='\t' 'NR>1  {print $1,$2, $1":"$2":"$4":"$5,$4,$5,$6,$9,$13,$14,$15,$16,$17}' |\
  sed '1i CHROM'\\t'POS'\\t'ID'\\t'REF'\\t'ALT1'\\t'A1'\\t'A1_FREQ'\\t'BETA'\\t'SE'\\t'L95'\\t'U95'\\t'P' > usc.pao.eur.glm.txt

sed 's/chr//g' /project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.pao.Subs.lat.phenotype.glm.logistic.hybrid |\
awk -v OFS='\t' 'NR>1  {print $1,$2, $1":"$2":"$4":"$5,$4,$5,$6,$9,$13,$14,$15,$16,$17}' |\
  sed '1i CHROM'\\t'POS'\\t'ID'\\t'REF'\\t'ALT1'\\t'A1'\\t'A1_FREQ'\\t'BETA'\\t'SE'\\t'L95'\\t'U95'\\t'P' > usc.pao.lat.glm.txt

sed 's/chr//g' /project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.pao.Subs.blk.phenotype.glm.logistic.hybrid |\
awk -v OFS='\t' 'NR>1  {print $1,$2, $1":"$2":"$4":"$5,$4,$5,$6,$9,$13,$14,$15,$16,$17}' |\
  sed '1i CHROM'\\t'POS'\\t'ID'\\t'REF'\\t'ALT1'\\t'A1'\\t'A1_FREQ'\\t'BETA'\\t'SE'\\t'L95'\\t'U95'\\t'P' > usc.pao.blk.glm.txt

sed 's/chr//g' /project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.pao.Subs.asn.phenotype.glm.logistic.hybrid |\
awk -v OFS='\t' 'NR>1  {print $1,$2, $1":"$2":"$4":"$5,$4,$5,$6,$9,$13,$14,$15,$16,$17}' |\
  sed '1i CHROM'\\t'POS'\\t'ID'\\t'REF'\\t'ALT1'\\t'A1'\\t'A1_FREQ'\\t'BETA'\\t'SE'\\t'L95'\\t'U95'\\t'P' > usc.pao.asn.glm.txt


cd /project/wiemels_494/sebastian/glioma_gwas/meta/METAL

## Europeans only

metal
MARKER ID
EFFECT BETA
PVALUE P
SCHEME STDERR
STDERR SE
VERBOSE OFF
GENOMICCONTROL OFF
PROCESS ../processedFilesFromCohorts/anna.pao.glm.txt
PROCESS ../processedFilesFromCohorts/dk.pao.glm.txt
PROCESS ../processedFilesFromCohorts/usc.pao.eur.glm.txt
ANALYZE HETEROGENEITY
QUIT
mv METAANALYSIS1.TBL meta.eur.pao.glioma.TBL
mv METAANALYSIS1.TBL.info meta.eur.pao.glioma.TBL.info
  
## All races  

metal
MARKER ID
EFFECT BETA
PVALUE P
SCHEME STDERR
STDERR SE
VERBOSE OFF
GENOMICCONTROL OFF
PROCESS ../processedFilesFromCohorts/anna.pao.glm.txt
PROCESS ../processedFilesFromCohorts/dk.pao.glm.txt
PROCESS ../processedFilesFromCohorts/usc.pao.eur.glm.txt
PROCESS ../processedFilesFromCohorts/usc.pao.lat.glm.txt
PROCESS ../processedFilesFromCohorts/usc.pao.blk.glm.txt
PROCESS ../processedFilesFromCohorts/usc.pao.asn.glm.txt
ANALYZE HETEROGENEITY
QUIT
mv METAANALYSIS1.TBL meta.four.races.pao.glioma.TBL
mv METAANALYSIS1.TBL.info meta.four.races.pao.glioma.TBL.info


```

```{bash meta-analysis glioblastoma}

cd /project/wiemels_494/sebastian/glioma_gwas/meta/processedFilesFromCohorts

########################################################################################################################

# modifying file headers
awk -v OFS='\t' 'NR>1  {print $1,$2, $1":"$2":"$4":"$5,$4,$5,$6,$9,$13,$14,$15,$16,$17}' /project/wiemels_494/sebastian/glioma_gwas/anna_cohort/concat.gbm.var_0PCA.20220401.PHENO1.glm.logistic.hybrid |\
  sed '1i CHROM'\\t'POS'\\t'ID'\\t'REF'\\t'ALT1'\\t'A1'\\t'A1_FREQ'\\t'BETA'\\t'SE'\\t'L95'\\t'U95'\\t'P'  > anna.glm.glm.txt

awk -v OFS='\t' 'NR>1 {print $1,$2, $1":"$2":"$4":"$5,$4,$5,$6,$13,$19,$20,$21,$22,$24}' /project/wiemels_494/sebastian/glioma_gwas/dk_cohort/Results/Glioblastoma.PC5.PHENO1.glm.logistic.hybrid |\
  sed '1i CHROM'\\t'POS'\\t'ID'\\t'REF'\\t'ALT1'\\t'A1'\\t'A1_FREQ'\\t'BETA'\\t'SE'\\t'L95'\\t'U95'\\t'P'  >  dk.glm.glm.txt

sed 's/chr//g' /project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.glm.Subs.eur.phenotype.glm.logistic.hybrid |\
awk -v OFS='\t' 'NR>1  {print $1,$2, $1":"$2":"$4":"$5,$4,$5,$6,$9,$13,$14,$15,$16,$17}' |\
  sed '1i CHROM'\\t'POS'\\t'ID'\\t'REF'\\t'ALT1'\\t'A1'\\t'A1_FREQ'\\t'BETA'\\t'SE'\\t'L95'\\t'U95'\\t'P' > usc.glm.eur.glm.txt

sed 's/chr//g' /project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.glm.Subs.lat.phenotype.glm.logistic.hybrid |\
awk -v OFS='\t' 'NR>1  {print $1,$2, $1":"$2":"$4":"$5,$4,$5,$6,$9,$13,$14,$15,$16,$17}' |\
  sed '1i CHROM'\\t'POS'\\t'ID'\\t'REF'\\t'ALT1'\\t'A1'\\t'A1_FREQ'\\t'BETA'\\t'SE'\\t'L95'\\t'U95'\\t'P' > usc.glm.lat.glm.txt

sed 's/chr//g' /project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.glm.Subs.blk.phenotype.glm.logistic.hybrid |\
awk -v OFS='\t' 'NR>1  {print $1,$2, $1":"$2":"$4":"$5,$4,$5,$6,$9,$13,$14,$15,$16,$17}' |\
  sed '1i CHROM'\\t'POS'\\t'ID'\\t'REF'\\t'ALT1'\\t'A1'\\t'A1_FREQ'\\t'BETA'\\t'SE'\\t'L95'\\t'U95'\\t'P' > usc.glm.blk.glm.txt

sed 's/chr//g' /project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.glm.Subs.asn.phenotype.glm.logistic.hybrid |\
awk -v OFS='\t' 'NR>1  {print $1,$2, $1":"$2":"$4":"$5,$4,$5,$6,$9,$13,$14,$15,$16,$17}' |\
  sed '1i CHROM'\\t'POS'\\t'ID'\\t'REF'\\t'ALT1'\\t'A1'\\t'A1_FREQ'\\t'BETA'\\t'SE'\\t'L95'\\t'U95'\\t'P' > usc.glm.asn.glm.txt


cd /project/wiemels_494/sebastian/glioma_gwas/meta/METAL

## Europeans only

metal
MARKER ID
EFFECT BETA
PVALUE P
SCHEME STDERR
STDERR SE
VERBOSE OFF
GENOMICCONTROL OFF
PROCESS ../processedFilesFromCohorts/anna.glm.glm.txt
PROCESS ../processedFilesFromCohorts/dk.glm.glm.txt
PROCESS ../processedFilesFromCohorts/usc.glm.eur.glm.txt
ANALYZE HETEROGENEITY
QUIT
mv METAANALYSIS1.TBL meta.eur.glm.glioma.TBL
mv METAANALYSIS1.TBL.info meta.eur.glm.glioma.TBL.info
  
## Europeans only (not including Anna)

metal
MARKER ID
EFFECT BETA
PVALUE P
SCHEME STDERR
STDERR SE
VERBOSE OFF
GENOMICCONTROL OFF
PROCESS ../processedFilesFromCohorts/dk.glm.glm.txt
PROCESS ../processedFilesFromCohorts/usc.glm.eur.glm.txt
ANALYZE HETEROGENEITY
QUIT
mv METAANALYSIS1.TBL meta.noanna.eur.glm.glioma.TBL
mv METAANALYSIS1.TBL.info meta.noanna.eur.glm.glioma.TBL.info
    
## All races  
metal
MARKER ID
EFFECT BETA
PVALUE P
SCHEME STDERR
STDERR SE
VERBOSE OFF
GENOMICCONTROL OFF
PROCESS ../processedFilesFromCohorts/anna.glm.glm.txt
PROCESS ../processedFilesFromCohorts/dk.glm.glm.txt
PROCESS ../processedFilesFromCohorts/usc.glm.eur.glm.txt
PROCESS ../processedFilesFromCohorts/usc.glm.lat.glm.txt
PROCESS ../processedFilesFromCohorts/usc.glm.blk.glm.txt
PROCESS ../processedFilesFromCohorts/usc.glm.asn.glm.txt
ANALYZE HETEROGENEITY
QUIT
mv METAANALYSIS1.TBL meta.four.races.glm.glioma.TBL
mv METAANALYSIS1.TBL.info meta.four.races.glm.glioma.TBL.info

## All races(not including anna)
metal
MARKER ID
EFFECT BETA
PVALUE P
SCHEME STDERR
STDERR SE
VERBOSE OFF
GENOMICCONTROL OFF
PROCESS ../processedFilesFromCohorts/dk.glm.glm.txt
PROCESS ../processedFilesFromCohorts/usc.glm.eur.glm.txt
PROCESS ../processedFilesFromCohorts/usc.glm.lat.glm.txt
PROCESS ../processedFilesFromCohorts/usc.glm.blk.glm.txt
PROCESS ../processedFilesFromCohorts/usc.glm.asn.glm.txt
ANALYZE HETEROGENEITY
QUIT
mv METAANALYSIS1.TBL meta.noanna.four.races.glm.glioma.TBL
mv METAANALYSIS1.TBL.info noanna.meta.four.races.glm.glioma.TBL.info

```

```{bash meta-analysis Oligodendroglioma}

cd /project/wiemels_494/sebastian/glioma_gwas/meta/processedFilesFromCohorts

########################################################################################################################

# modifying file headers
awk -v OFS='\t' 'NR>1  {print $1,$2, $1":"$2":"$4":"$5,$4,$5,$6,$9,$13,$14,$15,$16,$17}' /project/wiemels_494/sebastian/glioma_gwas/anna_cohort/concat.oligo.var_0PCA.20220401.PHENO1.glm.logistic.hybrid |\
  sed '1i CHROM'\\t'POS'\\t'ID'\\t'REF'\\t'ALT1'\\t'A1'\\t'A1_FREQ'\\t'BETA'\\t'SE'\\t'L95'\\t'U95'\\t'P'  > anna.olg.glm.txt

awk -v OFS='\t' 'NR>1 {print $1,$2, $1":"$2":"$4":"$5,$4,$5,$6,$13,$19,$20,$21,$22,$24}' /project/wiemels_494/sebastian/glioma_gwas/dk_cohort/Results/Oligodendroglioma.PC5.PHENO1.glm.logistic.hybrid |\
  sed '1i CHROM'\\t'POS'\\t'ID'\\t'REF'\\t'ALT1'\\t'A1'\\t'A1_FREQ'\\t'BETA'\\t'SE'\\t'L95'\\t'U95'\\t'P'  >  dk.olg.glm.txt

sed 's/chr//g' /project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.olg.Subs.eur.phenotype.glm.logistic.hybrid |\
awk -v OFS='\t' 'NR>1  {print $1,$2, $1":"$2":"$4":"$5,$4,$5,$6,$9,$13,$14,$15,$16,$17}' |\
  sed '1i CHROM'\\t'POS'\\t'ID'\\t'REF'\\t'ALT1'\\t'A1'\\t'A1_FREQ'\\t'BETA'\\t'SE'\\t'L95'\\t'U95'\\t'P' > usc.olg.eur.glm.txt

sed 's/chr//g' /project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.olg.Subs.lat.phenotype.glm.logistic.hybrid |\
awk -v OFS='\t' 'NR>1  {print $1,$2, $1":"$2":"$4":"$5,$4,$5,$6,$9,$13,$14,$15,$16,$17}' |\
  sed '1i CHROM'\\t'POS'\\t'ID'\\t'REF'\\t'ALT1'\\t'A1'\\t'A1_FREQ'\\t'BETA'\\t'SE'\\t'L95'\\t'U95'\\t'P' > usc.olg.lat.glm.txt

sed 's/chr//g' /project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.olg.Subs.blk.phenotype.glm.logistic.hybrid |\
awk -v OFS='\t' 'NR>1  {print $1,$2, $1":"$2":"$4":"$5,$4,$5,$6,$9,$13,$14,$15,$16,$17}' |\
  sed '1i CHROM'\\t'POS'\\t'ID'\\t'REF'\\t'ALT1'\\t'A1'\\t'A1_FREQ'\\t'BETA'\\t'SE'\\t'L95'\\t'U95'\\t'P' > usc.olg.blk.glm.txt

sed 's/chr//g' /project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.olg.Subs.asn.phenotype.glm.logistic.hybrid |\
awk -v OFS='\t' 'NR>1  {print $1,$2, $1":"$2":"$4":"$5,$4,$5,$6,$9,$13,$14,$15,$16,$17}' |\
  sed '1i CHROM'\\t'POS'\\t'ID'\\t'REF'\\t'ALT1'\\t'A1'\\t'A1_FREQ'\\t'BETA'\\t'SE'\\t'L95'\\t'U95'\\t'P' > usc.olg.asn.glm.txt


cd /project/wiemels_494/sebastian/glioma_gwas/meta/METAL

## Europeans only
metal
MARKER ID
EFFECT BETA
PVALUE P
SCHEME STDERR
STDERR SE
VERBOSE OFF
GENOMICCONTROL OFF
PROCESS ../processedFilesFromCohorts/anna.olg.glm.txt
PROCESS ../processedFilesFromCohorts/dk.olg.glm.txt
PROCESS ../processedFilesFromCohorts/usc.olg.eur.glm.txt
ANALYZE HETEROGENEITY
QUIT
mv METAANALYSIS1.TBL meta.eur.olg.glioma.TBL
mv METAANALYSIS1.TBL.info meta.eur.olg.glioma.TBL.info

## Europeans only (not including anna)
metal
MARKER ID
EFFECT BETA
PVALUE P
SCHEME STDERR
STDERR SE
VERBOSE OFF
GENOMICCONTROL OFF
PROCESS ../processedFilesFromCohorts/dk.olg.glm.txt
PROCESS ../processedFilesFromCohorts/usc.olg.eur.glm.txt
ANALYZE HETEROGENEITY
QUIT
mv METAANALYSIS1.TBL meta.noanna.eur.olg.glioma.TBL
mv METAANALYSIS1.TBL.info meta.noanna.eur.olg.glioma.TBL.info
  
## All races  
metal
MARKER ID
EFFECT BETA
PVALUE P
SCHEME STDERR
STDERR SE
VERBOSE OFF
GENOMICCONTROL OFF
PROCESS ../processedFilesFromCohorts/anna.olg.glm.txt
PROCESS ../processedFilesFromCohorts/dk.olg.glm.txt
PROCESS ../processedFilesFromCohorts/usc.olg.eur.glm.txt
PROCESS ../processedFilesFromCohorts/usc.olg.lat.glm.txt
PROCESS ../processedFilesFromCohorts/usc.olg.blk.glm.txt
PROCESS ../processedFilesFromCohorts/usc.olg.asn.glm.txt
ANALYZE HETEROGENEITY
QUIT
mv METAANALYSIS1.TBL meta.four.races.olg.glioma.TBL
mv METAANALYSIS1.TBL.info meta.four.races.olg.glioma.TBL.info

## All races (not including anna)
metal
MARKER ID
EFFECT BETA
PVALUE P
SCHEME STDERR
STDERR SE
VERBOSE OFF
GENOMICCONTROL OFF
PROCESS ../processedFilesFromCohorts/dk.olg.glm.txt
PROCESS ../processedFilesFromCohorts/usc.olg.eur.glm.txt
PROCESS ../processedFilesFromCohorts/usc.olg.lat.glm.txt
PROCESS ../processedFilesFromCohorts/usc.olg.blk.glm.txt
PROCESS ../processedFilesFromCohorts/usc.olg.asn.glm.txt
ANALYZE HETEROGENEITY
QUIT
mv METAANALYSIS1.TBL meta.noanna.four.races.olg.glioma.TBL
mv METAANALYSIS1.TBL.info meta.noanna.four.races.olg.glioma.TBL.info


```

Method 2: using metafor

```{r}

setwd("/project/wiemels_494/sebastian/glioma_gwas/meta/metafor")
library(tidyverse)
library(data.table)
library(metafor)
library(reshape)

fixed.effects.meta.analysis <- function(list.of.studies,data){
                              require(metafor)
                              coefs = data[,c("ID",paste0("coef.",list.of.studies))]
                              ses = data[,c("ID",paste0("se.",list.of.studies))]
                              require(reshape)
                              coefs = melt(coefs,id.vars="ID")
                              names(coefs) <- c("ID","study","coef")
                              ses = melt(ses,id.vars="ID")
                              data.long = cbind(coefs,ses[,"value"])
                              names(data.long)<-c("ID","study","coef","se")
                              res = split(data.long, f=data$ID)
                              res = lapply(res, function(x) rma.uni(slab=x$study,
                                                                    yi=x$coef,
                                                                    sei=x$se,
                                                                    method="ML", # mixed effect
                                                                    weighted=TRUE,
                                                                    control=list(maxiter=10000)))
                              res
                              }

extract.and.merge.meta.analysis <-function(meta.res,data){
                                  require(plyr)
                                  data.meta = ldply(lapply(meta.res, function(x)
                                    unlist(c(x$b[[1]],x[c("se","pval","k","QE","QEp","I2","H2")]))))
                                  colnames(data.meta)<-c("ID","coef.fe","se.fe","p.fe","study_num",
                                                         "q.fe","het.p.fe","i2.fe","h2.fe")
                                  data = merge(data,data.meta,by="ID",all.x=T)
                                  data
                                  }



for(subtype in c("all","lgg","hgg","ast","lga","hga","pa","glm","pao","olg")){

set.file1 = "usc"
file1 = fread(paste0("../processedFilesFromCohorts/usc.",subtype,".eur.glm.txt")) %>%
  filter(!is.na(P)) %>%
  mutate(BETA=as.numeric(BETA),SE=as.numeric(SE),P=as.numeric(P))%>%
  dplyr::select(ID,BETA,SE,P) %>%
  dplyr::rename( !!paste0("coef.", set.file1) := BETA,
                 !!paste0("se.", set.file1) := SE, 
                 !!paste0("p.", set.file1) := P)

set.file2 = "dk" 
file2 = fread(paste0("../processedFilesFromCohorts/dk.",subtype,".glm.txt")) %>%
  filter(!is.na(P)) %>%
  mutate(BETA=as.numeric(BETA),SE=as.numeric(SE),P=as.numeric(P))%>%
  dplyr::select(ID,BETA,SE,P) %>%
  dplyr::rename( !!paste0("coef.", set.file2) := BETA,
                 !!paste0("se.", set.file2) := SE, 
                 !!paste0("p.", set.file2) := P)


set.file3 = "anna" 
file3 = fread(paste0("../processedFilesFromCohorts/anna.",subtype,".glm.txt")) %>%
  filter(!is.na(P)) %>%
  mutate(BETA=as.numeric(BETA),SE=as.numeric(SE),P=as.numeric(P))%>%
  dplyr::select(ID,BETA,SE,P) %>%
  dplyr::rename( !!paste0("coef.", set.file3) := BETA,
                 !!paste0("se.", set.file3) := SE, 
                 !!paste0("p.", set.file3) := P)

dat <- file1 %>% 
  full_join(file2, by ="ID") %>%
  full_join(file3, by ="ID")  %>%
  data.frame()

studies <- c(set.file1,set.file2,set.file3)

meta.results <- fixed.effects.meta.analysis(list.of.studies = studies, data = dat)
dat <- extract.and.merge.meta.analysis(meta.res = meta.results, data = dat)

fwrite(dat, paste0("meta.eur.",subtype,".glioma.metafor.txt"))


set.file4 = "usc.lat" 
file4 = fread(paste0("../processedFilesFromCohorts/usc.",subtype,".lat.glm.txt")) %>%
  filter(!is.na(P)) %>%
  mutate(BETA=as.numeric(BETA),SE=as.numeric(SE),P=as.numeric(P))%>%
  dplyr::select(ID,BETA,SE,P) %>%
  dplyr::rename( !!paste0("coef.", set.file4) := BETA,
                 !!paste0("se.", set.file4) := SE, 
                 !!paste0("p.", set.file4) := P)

set.file5 = "usc.blk" 
file5 = fread(paste0("../processedFilesFromCohorts/usc.",subtype,".blk.glm.txt")) %>%
  filter(!is.na(P)) %>%
  mutate(BETA=as.numeric(BETA),SE=as.numeric(SE),P=as.numeric(P))%>%
  dplyr::select(ID,BETA,SE,P) %>%
  dplyr::rename( !!paste0("coef.", set.file5) := BETA,
                 !!paste0("se.", set.file5) := SE, 
                 !!paste0("p.", set.file5) := P)

set.file6 = "usc.asn" 
file6 = fread(paste0("../processedFilesFromCohorts/usc.",subtype,".asn.glm.txt")) %>%
  filter(!is.na(P)) %>%
  mutate(BETA=as.numeric(BETA),SE=as.numeric(SE),P=as.numeric(P))%>%
  dplyr::select(ID,BETA,SE,P) %>%
  dplyr::rename( !!paste0("coef.", set.file6) := BETA,
                 !!paste0("se.", set.file6) := SE, 
                 !!paste0("p.", set.file6) := P)

dat2 <- file1 %>% 
  full_join(file2, by ="ID") %>%
  full_join(file3, by ="ID")  %>%
  full_join(file4, by ="ID")  %>%
  full_join(file5, by ="ID")  %>%
  full_join(file6, by ="ID")  %>%
  data.frame() 

dat2 = dat2%>%distinct(ID,.keep_all=TRUE)

studies2 <- c(set.file1,set.file2,set.file3,set.file4,set.file5,set.file6)

meta.results2 <- fixed.effects.meta.analysis(list.of.studies = studies2, data = dat2)
dat2 <- extract.and.merge.meta.analysis(meta.res = meta.results2, data = dat2)

fwrite(dat2, paste0("meta.four.races.",subtype,".glioma.metafor.txt"))

}


for(subtype in c("olg","glm")){

  set.file1 = "usc"
  file1 = fread(paste0("../processedFilesFromCohorts/usc.",subtype,".eur.glm.txt")) %>%
    filter(!is.na(P)) %>%
    mutate(BETA=as.numeric(BETA),SE=as.numeric(SE),P=as.numeric(P))%>%
    dplyr::select(ID,BETA,SE,P) %>%
    dplyr::rename( !!paste0("coef.", set.file1) := BETA,
                   !!paste0("se.", set.file1) := SE, 
                   !!paste0("p.", set.file1) := P)
  
  set.file2 = "dk" 
  file2 = fread(paste0("../processedFilesFromCohorts/dk.",subtype,".glm.txt")) %>%
    filter(!is.na(P)) %>%
    mutate(BETA=as.numeric(BETA),SE=as.numeric(SE),P=as.numeric(P))%>%
    dplyr::select(ID,BETA,SE,P) %>%
    dplyr::rename( !!paste0("coef.", set.file2) := BETA,
                   !!paste0("se.", set.file2) := SE, 
                   !!paste0("p.", set.file2) := P)
  
  
  dat <- file1 %>% 
    full_join(file2, by ="ID") %>%
    data.frame()
  
  studies <- c(set.file1,set.file2)
  
  meta.results <- fixed.effects.meta.analysis(list.of.studies = studies, data = dat)
  dat <- extract.and.merge.meta.analysis(meta.res = meta.results, data = dat)
  
  fwrite(dat, paste0("meta.noanna.eur.",subtype,".glioma.metafor.txt"))
  
  
  set.file4 = "usc.lat" 
  file4 = fread(paste0("../processedFilesFromCohorts/usc.",subtype,".lat.glm.txt")) %>%
    filter(!is.na(P)) %>%
    mutate(BETA=as.numeric(BETA),SE=as.numeric(SE),P=as.numeric(P))%>%
    dplyr::select(ID,BETA,SE,P) %>%
    dplyr::rename( !!paste0("coef.", set.file4) := BETA,
                   !!paste0("se.", set.file4) := SE, 
                   !!paste0("p.", set.file4) := P)
  
  set.file5 = "usc.blk" 
  file5 = fread(paste0("../processedFilesFromCohorts/usc.",subtype,".blk.glm.txt")) %>%
    filter(!is.na(P)) %>%
    mutate(BETA=as.numeric(BETA),SE=as.numeric(SE),P=as.numeric(P))%>%
    dplyr::select(ID,BETA,SE,P) %>%
    dplyr::rename( !!paste0("coef.", set.file5) := BETA,
                   !!paste0("se.", set.file5) := SE, 
                   !!paste0("p.", set.file5) := P)
  
  set.file6 = "usc.asn" 
  file6 = fread(paste0("../processedFilesFromCohorts/usc.",subtype,".asn.glm.txt")) %>%
    filter(!is.na(P)) %>%
    mutate(BETA=as.numeric(BETA),SE=as.numeric(SE),P=as.numeric(P))%>%
    dplyr::select(ID,BETA,SE,P) %>%
    dplyr::rename( !!paste0("coef.", set.file6) := BETA,
                   !!paste0("se.", set.file6) := SE, 
                   !!paste0("p.", set.file6) := P)
  
  dat2 <- file1 %>% 
    full_join(file2, by ="ID") %>%
    full_join(file4, by ="ID")  %>%
    full_join(file5, by ="ID")  %>%
    full_join(file6, by ="ID")  %>%
    data.frame()
  
  studies2 <- c(set.file1,set.file2,set.file4,set.file5,set.file6)
  
  meta.results2 <- fixed.effects.meta.analysis(list.of.studies = studies2, data = dat2)
  dat2 <- extract.and.merge.meta.analysis(meta.res = meta.results2, data = dat2)
  
  fwrite(dat2, paste0("meta.noanna.four.races.",subtype,".glioma.metafor.txt"))
}

```

```{r visualization}

library(tidyverse)
library(QCEWAS)
library(qqman)
library(data.table)

# for METAL
setwd("/project/wiemels_494/sebastian/glioma_gwas/meta/METAL")

files =c("meta.eur.all.glioma.TBL",
        "meta.four.races.all.glioma.TBL",
        "meta.eur.lgg.glioma.TBL",
        "meta.four.races.lgg.glioma.TBL",
        "meta.eur.hgg.glioma.TBL",
        "meta.four.races.hgg.glioma.TBL",
        "meta.eur.ast.glioma.TBL",
        "meta.four.races.ast.glioma.TBL",
        "meta.eur.lga.glioma.TBL",
        "meta.four.races.lga.glioma.TBL",
        "meta.eur.hga.glioma.TBL",
        "meta.four.races.hga.glioma.TBL",
        "meta.eur.pa.glioma.TBL",
        "meta.four.races.pa.glioma.TBL",
        "meta.eur.pao.glioma.TBL",
        "meta.four.races.pao.glioma.TBL",
        "meta.eur.glm.glioma.TBL",
        "meta.noanna.eur.glm.glioma.TBL",
        "meta.four.races.glm.glioma.TBL",
        "meta.noanna.four.races.glm.glioma.TBL",
        "meta.eur.olg.glioma.TBL",
        "meta.noanna.eur.olg.glioma.TBL",
        "meta.four.races.olg.glioma.TBL",
        "meta.noanna.four.races.olg.glioma.TBL")


for(i in files){
  
  # i = "meta.non.eurs.ast.glioma.TBL"
  
  filename <- gsub("^.*/","",i)
  gwas_data <- fread(i) 
  
  gwas_data_1 <- gwas_data %>%
    dplyr::mutate(ID=MarkerName, P=`P-value`)%>%
    separate(MarkerName, c("CHROM","POS","REF","ALT"))%>%
    dplyr::mutate(CHROM = as.numeric(gsub("chr","",as.character(CHROM))),POS = as.numeric(as.character(POS)))

  gwas_data_1 <- gwas_data_1[is.finite(gwas_data_1$P),  ]
  
  lambda <- P_lambda(gwas_data_1$P)
  print(paste("lambda value of",filename,"is:",lambda))
  
  tiff(paste0(filename,".manhattan.tiff"), width = 10, height = 7, units = 'in', res = 300)
  manhattan(gwas_data_1, chr = "CHROM", bp = "POS", p = "P", snp = "ID")
  dev.off()
  
  tiff(paste0(filename,".qq.tiff"), width = 10, height = 7, units = 'in', res = 300)
  qq(gwas_data_1$P)
  dev.off()
  
}

# [1] "lambda value of meta.eur.all.glioma.TBL is: 0.999533524250844"
# [1] "lambda value of meta.four.races.all.glioma.TBL is: 0.994877463584719"
# [1] "lambda value of meta.eur.lgg.glioma.TBL is: 0.94962461666587"
# [1] "lambda value of meta.four.races.lgg.glioma.TBL is: 1.00186748706957"
# [1] "lambda value of meta.eur.hgg.glioma.TBL is: 0.868180745391023"
# [1] "lambda value of meta.four.races.hgg.glioma.TBL is: 0.954126957770219"
# [1] "lambda value of meta.eur.ast.glioma.TBL is: 0.986074356310157"
# [1] "lambda value of meta.four.races.ast.glioma.TBL is: 0.983306195744465"
# [1] "lambda value of meta.eur.lga.glioma.TBL is: 0.941112090599992"
# [1] "lambda value of meta.four.races.lga.glioma.TBL is: 0.990237175838761"
# [1] "lambda value of meta.eur.hga.glioma.TBL is: 0.835294841000264"
# [1] "lambda value of meta.four.races.hga.glioma.TBL is: 0.938435290964941"
# [1] "lambda value of meta.eur.pa.glioma.TBL is: 0.971832636653243"
# [1] "lambda value of meta.four.races.pa.glioma.TBL is: 0.991627609352945"
# [1] "lambda value of meta.eur.pao.glioma.TBL is: 0.972289713831546"
# [1] "lambda value of meta.four.races.pao.glioma.TBL is: 0.983767164684692"
# [1] "lambda value of meta.eur.glm.glioma.TBL is: 0.752430490064976"
# [1] "lambda value of meta.noanna.eur.glm.glioma.TBL is: 0.726343679176261"
# [1] "lambda value of meta.four.races.glm.glioma.TBL is: 0.956383857091319"
# [1] "lambda value of meta.noanna.four.races.glm.glioma.TBL is: 0.898929460700466"
# [1] "lambda value of meta.eur.olg.glioma.TBL is: 0.97549359747585"
# [1] "lambda value of meta.noanna.eur.olg.glioma.TBL is: 0.946930527000494"
# [1] "lambda value of meta.four.races.olg.glioma.TBL is: 1.30382917243537"
# [1] "lambda value of meta.noanna.four.races.olg.glioma.TBL is: 1.28284301407355"

# for METAL, make y axis the same

setwd("/project/wiemels_494/sebastian/glioma_gwas/meta/METAL")
library(tidyverse)
library(QCEWAS)
library(qqman)
library(data.table)

files =c("meta.four.races.all.glioma.TBL",
        "meta.four.races.ast.glioma.TBL",
        "meta.four.races.lga.glioma.TBL")

for(i in files){
  
  filename <- gsub("^.*/","",i)
  gwas_data <- fread(i) 
  
  gwas_data_1 <- gwas_data %>%
    dplyr::mutate(ID=MarkerName, P=`P-value`)%>%
    separate(MarkerName, c("CHROM","POS","REF","ALT"))%>%
    dplyr::mutate(CHROM = as.numeric(gsub("chr","",as.character(CHROM))),POS = as.numeric(as.character(POS)))

  gwas_data_1 <- gwas_data_1[is.finite(gwas_data_1$P),  ]
  
  tiff(paste0(filename,".manhattan.yaxis.tiff"), width = 10, height = 7, units = 'in', res = 300)
  manhattan(gwas_data_1, chr = "CHROM", bp = "POS", p = "P", snp = "ID", ylim=c(0,12))
  dev.off()
}


# for metafor
setwd("/project/wiemels_494/sebastian/glioma_gwas/meta/metafor")
files =  c("meta.eur.all.glioma.metafor.txt",
          "meta.eur.ast.glioma.metafor.txt",
          "meta.eur.glm.glioma.metafor.txt",
          "meta.eur.hga.glioma.metafor.txt",
          "meta.eur.hgg.glioma.metafor.txt",
          "meta.eur.lga.glioma.metafor.txt",
          "meta.eur.lgg.glioma.metafor.txt",
          "meta.eur.olg.glioma.metafor.txt",
          "meta.eur.pa.glioma.metafor.txt", 
          "meta.eur.pao.glioma.metafor.txt",
          "meta.four.races.all.glioma.metafor.txt",
          "meta.four.races.ast.glioma.metafor.txt",
          "meta.four.races.glm.glioma.metafor.txt",
          "meta.four.races.hga.glioma.metafor.txt",
          "meta.four.races.hgg.glioma.metafor.txt",
          "meta.four.races.lga.glioma.metafor.txt",
          "meta.four.races.lgg.glioma.metafor.txt",
          "meta.four.races.olg.glioma.metafor.txt",
          "meta.four.races.pa.glioma.metafor.txt", 
          "meta.four.races.pao.glioma.metafor.txt",
          "meta.noanna.eur.glm.glioma.metafor.txt",
          "meta.noanna.eur.olg.glioma.metafor.txt",
          "meta.noanna.four.races.glm.glioma.metafor.txt",
          "meta.noanna.four.races.olg.glioma.metafor.txt")

for(i in files){
  
  filename <- gsub("meta\\.","",i) %>% gsub("\\.metafor\\.txt","",.)
  
  gwas_data <- fread(i) 
  
  gwas_data_1 <- gwas_data %>%
    dplyr::mutate(MarkerName=ID, P=p.fe)%>%
    separate(MarkerName, c("CHROM","POS","REF","ALT"))%>%
    dplyr::mutate(CHROM = as.numeric(gsub("chr","",as.character(CHROM))),POS = as.numeric(as.character(POS)))

  gwas_data_1 <- gwas_data_1[is.finite(gwas_data_1$P),  ]
  
  lambda <- P_lambda(gwas_data_1$P)
  print(paste("lambda value of",filename,"is:",lambda))
  
  tiff(paste0(filename,".manhattan.tiff"), width = 10, height = 7, units = 'in', res = 300)
  manhattan(gwas_data_1, chr = "CHROM", bp = "POS", p = "P", snp = "ID")
  dev.off()
  
  tiff(paste0(filename,".qq.tiff"), width = 10, height = 7, units = 'in', res = 300)
  qq(gwas_data_1$P)
  dev.off()
  
}

```

> generating files for manuscript

```{bash annotation with annovar}

#download dbsnp database 
## annotate_variation.pl -downdb -webfrom annovar -buildver hg19 avsnp142 humandb/

# metal
cd /project/wiemels_494/sebastian/glioma_gwas/meta/METAL

for fileRes in \
    "meta.eur.all.glioma.TBL"\
    "meta.four.races.all.glioma.TBL"\
    "meta.eur.lgg.glioma.TBL"\
    "meta.four.races.lgg.glioma.TBL"\
    "meta.eur.hgg.glioma.TBL"\
    "meta.four.races.hgg.glioma.TBL"\
    "meta.eur.ast.glioma.TBL"\
    "meta.four.races.ast.glioma.TBL"\
    "meta.eur.lga.glioma.TBL"\
    "meta.four.races.lga.glioma.TBL"\
    "meta.eur.hga.glioma.TBL"\
    "meta.four.races.hga.glioma.TBL"\
    "meta.eur.pa.glioma.TBL"\
    "meta.four.races.pa.glioma.TBL"\
    "meta.eur.pao.glioma.TBL"\
    "meta.four.races.pao.glioma.TBL"\
    "meta.eur.glm.glioma.TBL"\
    "meta.noanna.eur.glm.glioma.TBL"\
    "meta.four.races.glm.glioma.TBL"\
    "meta.noanna.four.races.glm.glioma.TBL"\
    "meta.eur.olg.glioma.TBL"\
    "meta.noanna.eur.olg.glioma.TBL"\
    "meta.four.races.olg.glioma.TBL"\
    "meta.nonwhites.ast.glioma.TBL"\
    "meta.noanna.four.races.olg.glioma.TBL";\
do 
  fileName=$(basename $fileRes)
  
  awk -F":" '$1=$1' OFS="\t" $fileName | awk 'NR>1' |\
  awk '{print $1, $2, $2, $3, $4, $7}' | awk '$6<1e-05' > $fileName.input

  annotate_variation.pl -out anno -build hg19 $fileName.input \
  /project/wiemels_260/sebastian/dependencies/annovar/annovar/humandb/

  annotate_variation.pl $fileName.input -build hg19 -filter -dbtype avsnp142 \
  /project/wiemels_260/sebastian/dependencies/annovar/annovar/humandb/

  mv anno.variant_function anno.variant.$fileName
  mv anno.exonic_variant_function anno.exonic.$fileName
  mv anno.log anno.$fileName.log
done

#tail -n +1 *input > all.significant.variants.results
#tail -n +1 anno.variant* > sum.anno.variant.significant.results 
#tail -n +1 anno.exonic.* > sum.anno.exonic.significant.results

# metafor
cd /project/wiemels_494/sebastian/glioma_gwas/meta/metafor

for fileRes in \
    "meta.eur.all.glioma.metafor.txt"\
    "meta.eur.ast.glioma.metafor.txt"\
    "meta.eur.glm.glioma.metafor.txt"\
    "meta.eur.hga.glioma.metafor.txt"\
    "meta.eur.hgg.glioma.metafor.txt"\
    "meta.eur.lga.glioma.metafor.txt"\
    "meta.eur.lgg.glioma.metafor.txt"\
    "meta.eur.olg.glioma.metafor.txt"\
    "meta.eur.pa.glioma.metafor.txt"\
    "meta.eur.pao.glioma.metafor.txt";\
do 
  fileName=$(basename $fileRes)
  paste <(awk -F"," 'NR>1{print $1}' $fileName | awk -v FS=':' -v OFS='\t' '{print $1,$2,$2,$3,$4}')\
        <(  awk -v FS=',' -v OFS='\t' 'NR>1{print $13}' $fileName ) |  awk '$6<1e-05' > $fileName.metafor.input
        
  annotate_variation.pl -out anno -build hg19 $fileName.metafor.input \
  /project/wiemels_260/sebastian/dependencies/annovar/annovar/humandb/

  annotate_variation.pl $fileName.metafor.input -build hg19 -filter -dbtype avsnp142 \
  /project/wiemels_260/sebastian/dependencies/annovar/annovar/humandb/ 
  
  mv anno.variant_function anno.variant.$fileName.metafor
  mv anno.exonic_variant_function anno.exonic.$fileName.metafor
  mv anno.log anno.$fileName.metafor.log
done    
    
for fileRes in \
    "meta.four.races.all.glioma.metafor.txt"\
    "meta.four.races.ast.glioma.metafor.txt"\
    "meta.four.races.glm.glioma.metafor.txt"\
    "meta.four.races.hga.glioma.metafor.txt"\
    "meta.four.races.hgg.glioma.metafor.txt"\
    "meta.four.races.lga.glioma.metafor.txt"\
    "meta.four.races.lgg.glioma.metafor.txt"\
    "meta.four.races.olg.glioma.metafor.txt"\
    "meta.four.races.pa.glioma.metafor.txt"\
    "meta.four.races.pao.glioma.metafor.txt";\
do 
  fileName=$(basename $fileRes)
  paste <(awk -F"," 'NR>1{print $1}' $fileName | awk -v FS=':' -v OFS='\t' '{print $1,$2,$2,$3,$4}')\
        <(  awk -v FS=',' -v OFS='\t' 'NR>1{print $22}' $fileName ) |  awk '$6<1e-05' > $fileName.metafor.input
  annotate_variation.pl -out anno -build hg19 $fileName.metafor.input \
  /project/wiemels_260/sebastian/dependencies/annovar/annovar/humandb/

  annotate_variation.pl $fileName.metafor.input -build hg19 -filter -dbtype avsnp142 \
  /project/wiemels_260/sebastian/dependencies/annovar/annovar/humandb/ \
  
  mv anno.variant_function anno.variant.$fileName.metafor
  mv anno.exonic_variant_function anno.exonic.$fileName.metafor
  mv anno.log anno.$fileName.metafor.log
done    

    
for fileRes in \
    "meta.noanna.eur.glm.glioma.metafor.txt"\
    "meta.noanna.eur.olg.glioma.metafor.txt";\
do 
  fileName=$(basename $fileRes)
  paste <(awk -F"," 'NR>1{print $1}' $fileName | awk -v FS=':' -v OFS='\t' '{print $1,$2,$2,$3,$4}')\
        <(  awk -v FS=',' -v OFS='\t' 'NR>1{print $10}' $fileName ) |  awk '$6<1e-05' > $fileName.metafor.input
  annotate_variation.pl -out anno -build hg19 $fileName.metafor.input \
  /project/wiemels_260/sebastian/dependencies/annovar/annovar/humandb/

  annotate_variation.pl $fileName.metafor.input -build hg19 -filter -dbtype avsnp142 \
  /project/wiemels_260/sebastian/dependencies/annovar/annovar/humandb/ \
  
  mv anno.variant_function anno.variant.$fileName.metafor
  mv anno.exonic_variant_function anno.exonic.$fileName.metafor
  mv anno.log anno.$fileName.metafor.log
done    


for fileRes in \
    "meta.noanna.four.races.glm.glioma.metafor.txt"\
    "meta.noanna.four.races.olg.glioma.metafor.txt";\
do
  fileName=$(basename $fileRes)
  paste <(awk -F"," 'NR>1{print $1}' $fileName | awk -v FS=':' -v OFS='\t' '{print $1,$2,$2,$3,$4}')\
        <(  awk -v FS=',' -v OFS='\t' 'NR>1{print $19}' $fileName ) |  awk '$6<1e-05' > $fileName.metafor.input
  annotate_variation.pl -out anno -build hg19 $fileName.metafor.input \
  /project/wiemels_260/sebastian/dependencies/annovar/annovar/humandb/

  annotate_variation.pl $fileName.metafor.input -build hg19 -filter -dbtype avsnp142 \
  /project/wiemels_260/sebastian/dependencies/annovar/annovar/humandb/ \
  
  mv anno.variant_function anno.variant.$fileName.metafor
  mv anno.exonic_variant_function anno.exonic.$fileName.metafor
  mv anno.log anno.$fileName.metafor.log
done    

#tail -n +1 *input > all.significant.variants.metafor.results
#tail -n +1 anno.variant* > sum.anno.variant.significant.metafor.results 
#tail -n +1 anno.exonic.* > sum.anno.exonic.significant.metafor.results

```

```{bash extract MAF for cases and controls in each cohort}

cd /project/wiemels_494/sebastian/glioma_gwas/meta/processedFilesFromCohorts

awk -v OFS='\t' 'NR>1  {print $1,$2, $1":"$2":"$4":"$5,$4,$5,$6,$9,$10,$11}' /project/wiemels_494/sebastian/glioma_gwas/anna_cohort/concat.all.var_3PCA.20220401.PHENO1.glm.logistic.hybrid |\
  sed '1i CHROM'\\t'POS'\\t'ID'\\t'REF'\\t'ALT1'\\t'A1'\\t'A1_FREQ'\\t'A1_CASE_FREQ'\\t'A1_CTRL_FREQ'  > anna.all.freqs.glm.txt

awk -v OFS='\t' 'NR>1 {print $1,$2, $1":"$2":"$4":"$5,$4,$5,$6,$13,$14,$15}' /project/wiemels_494/sebastian/glioma_gwas/dk_cohort/Results/Glioma.PC5.PHENO1.glm.logistic.hybrid|\
  sed '1i CHROM'\\t'POS'\\t'ID'\\t'REF'\\t'ALT1'\\t'A1'\\t'A1_FREQ'\\t'A1_CASE_FREQ'\\t'A1_CTRL_FREQ'  >  dk.all.freqs.glm.txt

sed 's/chr//g' /project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.all.Subs.eur.phenotype.glm.logistic.hybrid |\
awk -v OFS='\t' 'NR>1  {print $1,$2, $1":"$2":"$4":"$5,$4,$5,$6,$9,$10,$11}' |\
  sed '1i CHROM'\\t'POS'\\t'ID'\\t'REF'\\t'ALT1'\\t'A1'\\t'A1_FREQ'\\t'A1_CASE_FREQ'\\t'A1_CTRL_FREQ' > usc.all.eur.freqs.glm.txt

sed 's/chr//g' /project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.all.Subs.lat.phenotype.glm.logistic.hybrid |\
awk -v OFS='\t' 'NR>1  {print $1,$2, $1":"$2":"$4":"$5,$4,$5,$6,$9,$10,$11}' |\
  sed '1i CHROM'\\t'POS'\\t'ID'\\t'REF'\\t'ALT1'\\t'A1'\\t'A1_FREQ'\\t'A1_CASE_FREQ'\\t'A1_CTRL_FREQ' > usc.all.lat.freqs.glm.txt

sed 's/chr//g' /project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.all.Subs.blk.phenotype.glm.logistic.hybrid |\
awk -v OFS='\t' 'NR>1  {print $1,$2, $1":"$2":"$4":"$5,$4,$5,$6,$9,$10,$11}' |\
  sed '1i CHROM'\\t'POS'\\t'ID'\\t'REF'\\t'ALT1'\\t'A1'\\t'A1_FREQ'\\t'A1_CASE_FREQ'\\t'A1_CTRL_FREQ' > usc.all.blk.freqs.glm.txt


sed 's/chr//g' /project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.all.Subs.asn.phenotype.glm.logistic.hybrid |\
awk -v OFS='\t' 'NR>1  {print $1,$2, $1":"$2":"$4":"$5,$4,$5,$6,$9,$10,$11}' |\
  sed '1i CHROM'\\t'POS'\\t'ID'\\t'REF'\\t'ALT1'\\t'A1'\\t'A1_FREQ'\\t'A1_CASE_FREQ'\\t'A1_CTRL_FREQ' > usc.all.asn.freqs.glm.txt

awk -v OFS='\t' 'NR>1  {print $1,$2, $1":"$2":"$4":"$5,$4,$5,$6,$9,$10,$11}' /project/wiemels_494/sebastian/glioma_gwas/anna_cohort/concat.lgg.var_0PCA.20220401.PHENO1.glm.logistic.hybrid |\
  sed '1i CHROM'\\t'POS'\\t'ID'\\t'REF'\\t'ALT1'\\t'A1'\\t'A1_FREQ'\\t'A1_CASE_FREQ'\\t'A1_CTRL_FREQ'  > anna.lgg.freqs.glm.txt

awk -v OFS='\t' 'NR>1 {print $1,$2, $1":"$2":"$4":"$5,$4,$5,$6,$13,$14,$15}' /project/wiemels_494/sebastian/glioma_gwas/dk_cohort/Results/LGG.PC5.PHENO1.glm.logistic.hybrid|\
  sed '1i CHROM'\\t'POS'\\t'ID'\\t'REF'\\t'ALT1'\\t'A1'\\t'A1_FREQ'\\t'A1_CASE_FREQ'\\t'A1_CTRL_FREQ'  >  dk.lgg.freqs.glm.txt

sed 's/chr//g' /project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.lgg.Subs.eur.phenotype.glm.logistic.hybrid |\
awk -v OFS='\t' 'NR>1  {print $1,$2, $1":"$2":"$4":"$5,$4,$5,$6,$9,$10,$11}' |\
  sed '1i CHROM'\\t'POS'\\t'ID'\\t'REF'\\t'ALT1'\\t'A1'\\t'A1_FREQ'\\t'A1_CASE_FREQ'\\t'A1_CTRL_FREQ' > usc.lgg.eur.freqs.glm.txt

sed 's/chr//g' /project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.lgg.Subs.lat.phenotype.glm.logistic.hybrid |\
awk -v OFS='\t' 'NR>1  {print $1,$2, $1":"$2":"$4":"$5,$4,$5,$6,$9,$10,$11}' |\
  sed '1i CHROM'\\t'POS'\\t'ID'\\t'REF'\\t'ALT1'\\t'A1'\\t'A1_FREQ'\\t'A1_CASE_FREQ'\\t'A1_CTRL_FREQ' > usc.lgg.lat.freqs.glm.txt

sed 's/chr//g' /project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.lgg.Subs.blk.phenotype.glm.logistic.hybrid |\
awk -v OFS='\t' 'NR>1  {print $1,$2, $1":"$2":"$4":"$5,$4,$5,$6,$9,$10,$11}' |\
  sed '1i CHROM'\\t'POS'\\t'ID'\\t'REF'\\t'ALT1'\\t'A1'\\t'A1_FREQ'\\t'A1_CASE_FREQ'\\t'A1_CTRL_FREQ' > usc.lgg.blk.freqs.glm.txt

sed 's/chr//g' /project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.lgg.Subs.asn.phenotype.glm.logistic.hybrid |\
awk -v OFS='\t' 'NR>1  {print $1,$2, $1":"$2":"$4":"$5,$4,$5,$6,$9,$10,$11}' |\
  sed '1i CHROM'\\t'POS'\\t'ID'\\t'REF'\\t'ALT1'\\t'A1'\\t'A1_FREQ'\\t'A1_CASE_FREQ'\\t'A1_CTRL_FREQ' > usc.lgg.asn.freqs.glm.txt

awk -v OFS='\t' 'NR>1  {print $1,$2, $1":"$2":"$4":"$5,$4,$5,$6,$9,$10,$11}' /project/wiemels_494/sebastian/glioma_gwas/anna_cohort/concat.hgg.var_1PCA.20220401.PHENO1.glm.logistic.hybrid |\
  sed '1i CHROM'\\t'POS'\\t'ID'\\t'REF'\\t'ALT1'\\t'A1'\\t'A1_FREQ'\\t'A1_CASE_FREQ'\\t'A1_CTRL_FREQ'  > anna.hgg.freqs.glm.txt

awk -v OFS='\t' 'NR>1 {print $1,$2, $1":"$2":"$4":"$5,$4,$5,$6,$13,$14,$15}' /project/wiemels_494/sebastian/glioma_gwas/dk_cohort/Results/HGG.PC5.PHENO1.glm.logistic.hybrid|\
  sed '1i CHROM'\\t'POS'\\t'ID'\\t'REF'\\t'ALT1'\\t'A1'\\t'A1_FREQ'\\t'A1_CASE_FREQ'\\t'A1_CTRL_FREQ'  >  dk.hgg.freqs.glm.txt

sed 's/chr//g' /project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.hgg.Subs.eur.phenotype.glm.logistic.hybrid |\
awk -v OFS='\t' 'NR>1  {print $1,$2, $1":"$2":"$4":"$5,$4,$5,$6,$9,$10,$11}' |\
  sed '1i CHROM'\\t'POS'\\t'ID'\\t'REF'\\t'ALT1'\\t'A1'\\t'A1_FREQ'\\t'A1_CASE_FREQ'\\t'A1_CTRL_FREQ' > usc.hgg.eur.freqs.glm.txt

sed 's/chr//g' /project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.hgg.Subs.lat.phenotype.glm.logistic.hybrid |\
awk -v OFS='\t' 'NR>1  {print $1,$2, $1":"$2":"$4":"$5,$4,$5,$6,$9,$10,$11}' |\
  sed '1i CHROM'\\t'POS'\\t'ID'\\t'REF'\\t'ALT1'\\t'A1'\\t'A1_FREQ'\\t'A1_CASE_FREQ'\\t'A1_CTRL_FREQ' > usc.hgg.lat.freqs.glm.txt

sed 's/chr//g' /project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.hgg.Subs.blk.phenotype.glm.logistic.hybrid |\
awk -v OFS='\t' 'NR>1  {print $1,$2, $1":"$2":"$4":"$5,$4,$5,$6,$9,$10,$11}' |\
  sed '1i CHROM'\\t'POS'\\t'ID'\\t'REF'\\t'ALT1'\\t'A1'\\t'A1_FREQ'\\t'A1_CASE_FREQ'\\t'A1_CTRL_FREQ' > usc.hgg.blk.freqs.glm.txt

sed 's/chr//g' /project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.hgg.Subs.asn.phenotype.glm.logistic.hybrid |\
awk -v OFS='\t' 'NR>1  {print $1,$2, $1":"$2":"$4":"$5,$4,$5,$6,$9,$10,$11}' |\
  sed '1i CHROM'\\t'POS'\\t'ID'\\t'REF'\\t'ALT1'\\t'A1'\\t'A1_FREQ'\\t'A1_CASE_FREQ'\\t'A1_CTRL_FREQ' > usc.hgg.asn.freqs.glm.txt

awk -v OFS='\t' 'NR>1  {print $1,$2, $1":"$2":"$4":"$5,$4,$5,$6,$9,$10,$11}' /project/wiemels_494/sebastian/glioma_gwas/anna_cohort/concat.astro.var_10PCA.20220401.PHENO1.glm.logistic.hybrid |\
  sed '1i CHROM'\\t'POS'\\t'ID'\\t'REF'\\t'ALT1'\\t'A1'\\t'A1_FREQ'\\t'A1_CASE_FREQ'\\t'A1_CTRL_FREQ'  > anna.ast.freqs.glm.txt

awk -v OFS='\t' 'NR>1 {print $1,$2, $1":"$2":"$4":"$5,$4,$5,$6,$13,$14,$15}' /project/wiemels_494/sebastian/glioma_gwas/dk_cohort/Results/Astrocytoma.PC5.PHENO1.glm.logistic.hybrid|\
  sed '1i CHROM'\\t'POS'\\t'ID'\\t'REF'\\t'ALT1'\\t'A1'\\t'A1_FREQ'\\t'A1_CASE_FREQ'\\t'A1_CTRL_FREQ'  >  dk.ast.freqs.glm.txt

sed 's/chr//g' /project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.ast.Subs.eur.phenotype.glm.logistic.hybrid |\
awk -v OFS='\t' 'NR>1  {print $1,$2, $1":"$2":"$4":"$5,$4,$5,$6,$9,$10,$11}' |\
  sed '1i CHROM'\\t'POS'\\t'ID'\\t'REF'\\t'ALT1'\\t'A1'\\t'A1_FREQ'\\t'A1_CASE_FREQ'\\t'A1_CTRL_FREQ' > usc.ast.eur.freqs.glm.txt

sed 's/chr//g' /project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.ast.Subs.lat.phenotype.glm.logistic.hybrid |\
awk -v OFS='\t' 'NR>1  {print $1,$2, $1":"$2":"$4":"$5,$4,$5,$6,$9,$10,$11}' |\
  sed '1i CHROM'\\t'POS'\\t'ID'\\t'REF'\\t'ALT1'\\t'A1'\\t'A1_FREQ'\\t'A1_CASE_FREQ'\\t'A1_CTRL_FREQ' > usc.ast.lat.freqs.glm.txt

sed 's/chr//g' /project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.ast.Subs.blk.phenotype.glm.logistic.hybrid |\
awk -v OFS='\t' 'NR>1  {print $1,$2, $1":"$2":"$4":"$5,$4,$5,$6,$9,$10,$11}' |\
  sed '1i CHROM'\\t'POS'\\t'ID'\\t'REF'\\t'ALT1'\\t'A1'\\t'A1_FREQ'\\t'A1_CASE_FREQ'\\t'A1_CTRL_FREQ' > usc.ast.blk.freqs.glm.txt

sed 's/chr//g' /project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.ast.Subs.asn.phenotype.glm.logistic.hybrid |\
awk -v OFS='\t' 'NR>1  {print $1,$2, $1":"$2":"$4":"$5,$4,$5,$6,$9,$10,$11}' |\
  sed '1i CHROM'\\t'POS'\\t'ID'\\t'REF'\\t'ALT1'\\t'A1'\\t'A1_FREQ'\\t'A1_CASE_FREQ'\\t'A1_CTRL_FREQ' > usc.ast.asn.freqs.glm.txt

awk -v OFS='\t' 'NR>1  {print $1,$2, $1":"$2":"$4":"$5,$4,$5,$6,$9,$10,$11}' /project/wiemels_494/sebastian/glioma_gwas/anna_cohort/concat.lg_astro.var_8PCA.20220401.PHENO1.glm.logistic.hybrid |\
  sed '1i CHROM'\\t'POS'\\t'ID'\\t'REF'\\t'ALT1'\\t'A1'\\t'A1_FREQ'\\t'A1_CASE_FREQ'\\t'A1_CTRL_FREQ'  > anna.lga.freqs.glm.txt

awk -v OFS='\t' 'NR>1 {print $1,$2, $1":"$2":"$4":"$5,$4,$5,$6,$13,$14,$15}' /project/wiemels_494/sebastian/glioma_gwas/dk_cohort/Results/LG_astrocytoma.PC5.PHENO1.glm.logistic.hybrid|\
  sed '1i CHROM'\\t'POS'\\t'ID'\\t'REF'\\t'ALT1'\\t'A1'\\t'A1_FREQ'\\t'A1_CASE_FREQ'\\t'A1_CTRL_FREQ'  >  dk.lga.freqs.glm.txt

sed 's/chr//g' /project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.lga.Subs.eur.phenotype.glm.logistic.hybrid |\
awk -v OFS='\t' 'NR>1  {print $1,$2, $1":"$2":"$4":"$5,$4,$5,$6,$9,$10,$11}' |\
  sed '1i CHROM'\\t'POS'\\t'ID'\\t'REF'\\t'ALT1'\\t'A1'\\t'A1_FREQ'\\t'A1_CASE_FREQ'\\t'A1_CTRL_FREQ' > usc.lga.eur.freqs.glm.txt

sed 's/chr//g' /project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.lga.Subs.lat.phenotype.glm.logistic.hybrid |\
awk -v OFS='\t' 'NR>1  {print $1,$2, $1":"$2":"$4":"$5,$4,$5,$6,$9,$10,$11}' |\
  sed '1i CHROM'\\t'POS'\\t'ID'\\t'REF'\\t'ALT1'\\t'A1'\\t'A1_FREQ'\\t'A1_CASE_FREQ'\\t'A1_CTRL_FREQ' > usc.lga.lat.freqs.glm.txt

sed 's/chr//g' /project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.lga.Subs.blk.phenotype.glm.logistic.hybrid |\
awk -v OFS='\t' 'NR>1  {print $1,$2, $1":"$2":"$4":"$5,$4,$5,$6,$9,$10,$11}' |\
  sed '1i CHROM'\\t'POS'\\t'ID'\\t'REF'\\t'ALT1'\\t'A1'\\t'A1_FREQ'\\t'A1_CASE_FREQ'\\t'A1_CTRL_FREQ' > usc.lga.blk.freqs.glm.txt

sed 's/chr//g' /project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.lga.Subs.asn.phenotype.glm.logistic.hybrid |\
awk -v OFS='\t' 'NR>1  {print $1,$2, $1":"$2":"$4":"$5,$4,$5,$6,$9,$10,$11}' |\
  sed '1i CHROM'\\t'POS'\\t'ID'\\t'REF'\\t'ALT1'\\t'A1'\\t'A1_FREQ'\\t'A1_CASE_FREQ'\\t'A1_CTRL_FREQ' > usc.lga.asn.freqs.glm.txt

awk -v OFS='\t' 'NR>1  {print $1,$2, $1":"$2":"$4":"$5,$4,$5,$6,$9,$10,$11}' /project/wiemels_494/sebastian/glioma_gwas/anna_cohort/concat.hg_astro.var_3PCA.20220401.PHENO1.glm.logistic.hybrid |\
  sed '1i CHROM'\\t'POS'\\t'ID'\\t'REF'\\t'ALT1'\\t'A1'\\t'A1_FREQ'\\t'A1_CASE_FREQ'\\t'A1_CTRL_FREQ'  > anna.hga.freqs.glm.txt

awk -v OFS='\t' 'NR>1 {print $1,$2, $1":"$2":"$4":"$5,$4,$5,$6,$13,$14,$15}' /project/wiemels_494/sebastian/glioma_gwas/dk_cohort/Results/HG_astrocytoma.PC5.PHENO1.glm.logistic.hybrid|\
  sed '1i CHROM'\\t'POS'\\t'ID'\\t'REF'\\t'ALT1'\\t'A1'\\t'A1_FREQ'\\t'A1_CASE_FREQ'\\t'A1_CTRL_FREQ'  >  dk.hga.freqs.glm.txt

sed 's/chr//g' /project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.hga.Subs.eur.phenotype.glm.logistic.hybrid |\
awk -v OFS='\t' 'NR>1  {print $1,$2, $1":"$2":"$4":"$5,$4,$5,$6,$9,$10,$11}' |\
  sed '1i CHROM'\\t'POS'\\t'ID'\\t'REF'\\t'ALT1'\\t'A1'\\t'A1_FREQ'\\t'A1_CASE_FREQ'\\t'A1_CTRL_FREQ' > usc.hga.eur.freqs.glm.txt

sed 's/chr//g' /project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.hga.Subs.lat.phenotype.glm.logistic.hybrid |\
awk -v OFS='\t' 'NR>1  {print $1,$2, $1":"$2":"$4":"$5,$4,$5,$6,$9,$10,$11}' |\
  sed '1i CHROM'\\t'POS'\\t'ID'\\t'REF'\\t'ALT1'\\t'A1'\\t'A1_FREQ'\\t'A1_CASE_FREQ'\\t'A1_CTRL_FREQ' > usc.hga.lat.freqs.glm.txt

sed 's/chr//g' /project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.hga.Subs.blk.phenotype.glm.logistic.hybrid |\
awk -v OFS='\t' 'NR>1  {print $1,$2, $1":"$2":"$4":"$5,$4,$5,$6,$9,$10,$11}' |\
  sed '1i CHROM'\\t'POS'\\t'ID'\\t'REF'\\t'ALT1'\\t'A1'\\t'A1_FREQ'\\t'A1_CASE_FREQ'\\t'A1_CTRL_FREQ' > usc.hga.blk.freqs.glm.txt

sed 's/chr//g' /project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.hga.Subs.asn.phenotype.glm.logistic.hybrid |\
awk -v OFS='\t' 'NR>1  {print $1,$2, $1":"$2":"$4":"$5,$4,$5,$6,$9,$10,$11}' |\
  sed '1i CHROM'\\t'POS'\\t'ID'\\t'REF'\\t'ALT1'\\t'A1'\\t'A1_FREQ'\\t'A1_CASE_FREQ'\\t'A1_CTRL_FREQ' > usc.hga.asn.freqs.glm.txt

awk -v OFS='\t' 'NR>1  {print $1,$2, $1":"$2":"$4":"$5,$4,$5,$6,$9,$10,$11}' /project/wiemels_494/sebastian/glioma_gwas/anna_cohort/concat.pilo.var_0PCA.20220401.PHENO1.glm.logistic.hybrid |\
  sed '1i CHROM'\\t'POS'\\t'ID'\\t'REF'\\t'ALT1'\\t'A1'\\t'A1_FREQ'\\t'A1_CASE_FREQ'\\t'A1_CTRL_FREQ'  > anna.pa.freqs.glm.txt

awk -v OFS='\t' 'NR>1 {print $1,$2, $1":"$2":"$4":"$5,$4,$5,$6,$13,$14,$15}' /project/wiemels_494/sebastian/glioma_gwas/dk_cohort/Results/Pilocytic_astrocytoma.PC5.PHENO1.glm.logistic.hybrid |\
  sed '1i CHROM'\\t'POS'\\t'ID'\\t'REF'\\t'ALT1'\\t'A1'\\t'A1_FREQ'\\t'A1_CASE_FREQ'\\t'A1_CTRL_FREQ'  >  dk.pa.freqs.glm.txt

sed 's/chr//g' /project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.pa.Subs.eur.phenotype.glm.logistic.hybrid |\
awk -v OFS='\t' 'NR>1  {print $1,$2, $1":"$2":"$4":"$5,$4,$5,$6,$9,$10,$11}' |\
  sed '1i CHROM'\\t'POS'\\t'ID'\\t'REF'\\t'ALT1'\\t'A1'\\t'A1_FREQ'\\t'A1_CASE_FREQ'\\t'A1_CTRL_FREQ' > usc.pa.eur.freqs.glm.txt

sed 's/chr//g' /project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.pa.Subs.lat.phenotype.glm.logistic.hybrid |\
awk -v OFS='\t' 'NR>1  {print $1,$2, $1":"$2":"$4":"$5,$4,$5,$6,$9,$10,$11}' |\
  sed '1i CHROM'\\t'POS'\\t'ID'\\t'REF'\\t'ALT1'\\t'A1'\\t'A1_FREQ'\\t'A1_CASE_FREQ'\\t'A1_CTRL_FREQ' > usc.pa.lat.freqs.glm.txt

sed 's/chr//g' /project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.pa.Subs.blk.phenotype.glm.logistic.hybrid |\
awk -v OFS='\t' 'NR>1  {print $1,$2, $1":"$2":"$4":"$5,$4,$5,$6,$9,$10,$11}' |\
  sed '1i CHROM'\\t'POS'\\t'ID'\\t'REF'\\t'ALT1'\\t'A1'\\t'A1_FREQ'\\t'A1_CASE_FREQ'\\t'A1_CTRL_FREQ' > usc.pa.blk.freqs.glm.txt

sed 's/chr//g' /project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.pa.Subs.asn.phenotype.glm.logistic.hybrid |\
awk -v OFS='\t' 'NR>1  {print $1,$2, $1":"$2":"$4":"$5,$4,$5,$6,$9,$10,$11}' |\
  sed '1i CHROM'\\t'POS'\\t'ID'\\t'REF'\\t'ALT1'\\t'A1'\\t'A1_FREQ'\\t'A1_CASE_FREQ'\\t'A1_CTRL_FREQ' > usc.pa.asn.freqs.glm.txt

awk -v OFS='\t' 'NR>1  {print $1,$2, $1":"$2":"$4":"$5,$4,$5,$6,$9,$10,$11}' /project/wiemels_494/sebastian/glioma_gwas/anna_cohort/concat.pilo_optic.var_0PCA.20220401.PHENO1.glm.logistic.hybrid |\
  sed '1i CHROM'\\t'POS'\\t'ID'\\t'REF'\\t'ALT1'\\t'A1'\\t'A1_FREQ'\\t'A1_CASE_FREQ'\\t'A1_CTRL_FREQ'  > anna.pao.freqs.glm.txt

awk -v OFS='\t' 'NR>1 {print $1,$2, $1":"$2":"$4":"$5,$4,$5,$6,$13,$14,$15}' /project/wiemels_494/sebastian/glioma_gwas/dk_cohort/Results/Pilocytic_astrocytoma_incl_optic.PC5.PHENO1.glm.logistic.hybrid |\
  sed '1i CHROM'\\t'POS'\\t'ID'\\t'REF'\\t'ALT1'\\t'A1'\\t'A1_FREQ'\\t'A1_CASE_FREQ'\\t'A1_CTRL_FREQ'  >  dk.pao.freqs.glm.txt

sed 's/chr//g' /project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.pao.Subs.eur.phenotype.glm.logistic.hybrid |\
awk -v OFS='\t' 'NR>1  {print $1,$2, $1":"$2":"$4":"$5,$4,$5,$6,$9,$10,$11}' |\
  sed '1i CHROM'\\t'POS'\\t'ID'\\t'REF'\\t'ALT1'\\t'A1'\\t'A1_FREQ'\\t'A1_CASE_FREQ'\\t'A1_CTRL_FREQ' > usc.pao.eur.freqs.glm.txt

sed 's/chr//g' /project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.pao.Subs.lat.phenotype.glm.logistic.hybrid |\
awk -v OFS='\t' 'NR>1  {print $1,$2, $1":"$2":"$4":"$5,$4,$5,$6,$9,$10,$11}' |\
  sed '1i CHROM'\\t'POS'\\t'ID'\\t'REF'\\t'ALT1'\\t'A1'\\t'A1_FREQ'\\t'A1_CASE_FREQ'\\t'A1_CTRL_FREQ' > usc.pao.lat.freqs.glm.txt

sed 's/chr//g' /project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.pao.Subs.blk.phenotype.glm.logistic.hybrid |\
awk -v OFS='\t' 'NR>1  {print $1,$2, $1":"$2":"$4":"$5,$4,$5,$6,$9,$10,$11}' |\
  sed '1i CHROM'\\t'POS'\\t'ID'\\t'REF'\\t'ALT1'\\t'A1'\\t'A1_FREQ'\\t'A1_CASE_FREQ'\\t'A1_CTRL_FREQ' > usc.pao.blk.freqs.glm.txt


sed 's/chr//g' /project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.pao.Subs.asn.phenotype.glm.logistic.hybrid |\
awk -v OFS='\t' 'NR>1  {print $1,$2, $1":"$2":"$4":"$5,$4,$5,$6,$9,$10,$11}' |\
  sed '1i CHROM'\\t'POS'\\t'ID'\\t'REF'\\t'ALT1'\\t'A1'\\t'A1_FREQ'\\t'A1_CASE_FREQ'\\t'A1_CTRL_FREQ' > usc.pao.asn.freqs.glm.txt

awk -v OFS='\t' 'NR>1  {print $1,$2, $1":"$2":"$4":"$5,$4,$5,$6,$9,$10,$11}' /project/wiemels_494/sebastian/glioma_gwas/anna_cohort/concat.gbm.var_0PCA.20220401.PHENO1.glm.logistic.hybrid |\
  sed '1i CHROM'\\t'POS'\\t'ID'\\t'REF'\\t'ALT1'\\t'A1'\\t'A1_FREQ'\\t'A1_CASE_FREQ'\\t'A1_CTRL_FREQ'  > anna.glm.freqs.glm.txt

awk -v OFS='\t' 'NR>1 {print $1,$2, $1":"$2":"$4":"$5,$4,$5,$6,$13,$14,$15}' /project/wiemels_494/sebastian/glioma_gwas/dk_cohort/Results/Glioblastoma.PC5.PHENO1.glm.logistic.hybrid |\
  sed '1i CHROM'\\t'POS'\\t'ID'\\t'REF'\\t'ALT1'\\t'A1'\\t'A1_FREQ'\\t'A1_CASE_FREQ'\\t'A1_CTRL_FREQ'  >  dk.glm.freqs.glm.txt

sed 's/chr//g' /project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.glm.Subs.eur.phenotype.glm.logistic.hybrid |\
awk -v OFS='\t' 'NR>1  {print $1,$2, $1":"$2":"$4":"$5,$4,$5,$6,$9,$10,$11}' |\
  sed '1i CHROM'\\t'POS'\\t'ID'\\t'REF'\\t'ALT1'\\t'A1'\\t'A1_FREQ'\\t'A1_CASE_FREQ'\\t'A1_CTRL_FREQ' > usc.glm.eur.freqs.glm.txt

sed 's/chr//g' /project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.glm.Subs.lat.phenotype.glm.logistic.hybrid |\
awk -v OFS='\t' 'NR>1  {print $1,$2, $1":"$2":"$4":"$5,$4,$5,$6,$9,$10,$11}' |\
  sed '1i CHROM'\\t'POS'\\t'ID'\\t'REF'\\t'ALT1'\\t'A1'\\t'A1_FREQ'\\t'A1_CASE_FREQ'\\t'A1_CTRL_FREQ' > usc.glm.lat.freqs.glm.txt

sed 's/chr//g' /project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.glm.Subs.blk.phenotype.glm.logistic.hybrid |\
awk -v OFS='\t' 'NR>1  {print $1,$2, $1":"$2":"$4":"$5,$4,$5,$6,$9,$10,$11}' |\
  sed '1i CHROM'\\t'POS'\\t'ID'\\t'REF'\\t'ALT1'\\t'A1'\\t'A1_FREQ'\\t'A1_CASE_FREQ'\\t'A1_CTRL_FREQ' > usc.glm.blk.freqs.glm.txt

sed 's/chr//g' /project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.glm.Subs.asn.phenotype.glm.logistic.hybrid |\
awk -v OFS='\t' 'NR>1  {print $1,$2, $1":"$2":"$4":"$5,$4,$5,$6,$9,$10,$11}' |\
  sed '1i CHROM'\\t'POS'\\t'ID'\\t'REF'\\t'ALT1'\\t'A1'\\t'A1_FREQ'\\t'A1_CASE_FREQ'\\t'A1_CTRL_FREQ' > usc.glm.asn.freqs.glm.txt

awk -v OFS='\t' 'NR>1  {print $1,$2, $1":"$2":"$4":"$5,$4,$5,$6,$9,$10,$11}' /project/wiemels_494/sebastian/glioma_gwas/anna_cohort/concat.oligo.var_0PCA.20220401.PHENO1.glm.logistic.hybrid |\
  sed '1i CHROM'\\t'POS'\\t'ID'\\t'REF'\\t'ALT1'\\t'A1'\\t'A1_FREQ'\\t'A1_CASE_FREQ'\\t'A1_CTRL_FREQ'  > anna.olg.freqs.glm.txt

awk -v OFS='\t' 'NR>1 {print $1,$2, $1":"$2":"$4":"$5,$4,$5,$6,$13,$14,$15}' /project/wiemels_494/sebastian/glioma_gwas/dk_cohort/Results/Oligodendroglioma.PC5.PHENO1.glm.logistic.hybrid |\
  sed '1i CHROM'\\t'POS'\\t'ID'\\t'REF'\\t'ALT1'\\t'A1'\\t'A1_FREQ'\\t'A1_CASE_FREQ'\\t'A1_CTRL_FREQ'  >  dk.olg.freqs.glm.txt

sed 's/chr//g' /project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.olg.Subs.eur.phenotype.glm.logistic.hybrid |\
awk -v OFS='\t' 'NR>1  {print $1,$2, $1":"$2":"$4":"$5,$4,$5,$6,$9,$10,$11}' |\
  sed '1i CHROM'\\t'POS'\\t'ID'\\t'REF'\\t'ALT1'\\t'A1'\\t'A1_FREQ'\\t'A1_CASE_FREQ'\\t'A1_CTRL_FREQ' > usc.olg.eur.freqs.glm.txt

sed 's/chr//g' /project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.olg.Subs.lat.phenotype.glm.logistic.hybrid |\
awk -v OFS='\t' 'NR>1  {print $1,$2, $1":"$2":"$4":"$5,$4,$5,$6,$9,$10,$11}' |\
  sed '1i CHROM'\\t'POS'\\t'ID'\\t'REF'\\t'ALT1'\\t'A1'\\t'A1_FREQ'\\t'A1_CASE_FREQ'\\t'A1_CTRL_FREQ' > usc.olg.lat.freqs.glm.txt

sed 's/chr//g' /project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.olg.Subs.blk.phenotype.glm.logistic.hybrid |\
awk -v OFS='\t' 'NR>1  {print $1,$2, $1":"$2":"$4":"$5,$4,$5,$6,$9,$10,$11}' |\
  sed '1i CHROM'\\t'POS'\\t'ID'\\t'REF'\\t'ALT1'\\t'A1'\\t'A1_FREQ'\\t'A1_CASE_FREQ'\\t'A1_CTRL_FREQ' > usc.olg.blk.freqs.glm.txt

sed 's/chr//g' /project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.olg.Subs.asn.phenotype.glm.logistic.hybrid |\
awk -v OFS='\t' 'NR>1  {print $1,$2, $1":"$2":"$4":"$5,$4,$5,$6,$9,$10,$11}' |\
  sed '1i CHROM'\\t'POS'\\t'ID'\\t'REF'\\t'ALT1'\\t'A1'\\t'A1_FREQ'\\t'A1_CASE_FREQ'\\t'A1_CTRL_FREQ' > usc.olg.asn.freqs.glm.txt

```

```{r computing odds ratio etc}

library(tidyverse)
library(data.table)

# METAL
setwd("/project/wiemels_494/sebastian/glioma_gwas/meta/METAL")
files =c("meta.eur.all.glioma.TBL",
        "meta.four.races.all.glioma.TBL",
        "meta.eur.lgg.glioma.TBL",
        "meta.four.races.lgg.glioma.TBL",
        "meta.eur.hgg.glioma.TBL",
        "meta.four.races.hgg.glioma.TBL",
        "meta.eur.ast.glioma.TBL",
        "meta.four.races.ast.glioma.TBL",
        "meta.eur.lga.glioma.TBL",
        "meta.four.races.lga.glioma.TBL",
        "meta.eur.hga.glioma.TBL",
        "meta.four.races.hga.glioma.TBL",
        "meta.eur.pa.glioma.TBL",
        "meta.four.races.pa.glioma.TBL",
        "meta.eur.pao.glioma.TBL",
        "meta.four.races.pao.glioma.TBL",
        "meta.eur.glm.glioma.TBL",
        "meta.noanna.eur.glm.glioma.TBL",
        "meta.four.races.glm.glioma.TBL",
        "meta.noanna.four.races.glm.glioma.TBL",
        "meta.eur.olg.glioma.TBL",
        "meta.noanna.eur.olg.glioma.TBL",
        "meta.four.races.olg.glioma.TBL",
        "meta.noanna.four.races.olg.glioma.TBL",
        "meta.nonwhites.ast.glioma.TBL")

for(file in files){
  
  #file="meta.nonwhites.ast.glioma.TBL"
  snpAnno = fread(paste0(file,".input.hg19_avsnp142_dropped"),header=FALSE)
  varAnno = fread(paste0("anno.variant.",file),header=FALSE) 
  exoAnno = fread(paste0("anno.exonic.",file),sep='\t',header=FALSE)
  
  # allele frequency information annotation
  ## eur, with anna
  if(grepl("eur",file)&(!grepl("noanna",file))){
    subtype = gsub("meta.eur.|.glioma.TBL","",file)
    cohort1 = fread(paste0("../processedFilesFromCohorts/usc.",subtype,".eur.freqs.glm.txt")) %>%
      mutate(CHROM=as.character(CHROM),POS=as.character(POS)) %>%
      dplyr::select(CHROM,POS,ID,REF,ALT1,A1_FREQ,A1_CASE_FREQ,A1_CTRL_FREQ) %>%
      dplyr::rename("usc_eur_A1_FREQ"="A1_FREQ", 
                    "usc_eur_A1_CASE_FREQ"="A1_CASE_FREQ",
                    "usc_eur_A1_CTRL_FREQ"="A1_CTRL_FREQ")
    cohort2 = fread(paste0("../processedFilesFromCohorts/anna.",subtype,".freqs.glm.txt"))%>%
      mutate(CHROM=as.character(CHROM),POS=as.character(POS)) %>%
      dplyr::select(CHROM,POS,ID,REF,ALT1,A1_FREQ,A1_CASE_FREQ,A1_CTRL_FREQ) %>%
      dplyr::rename("anno_eur_A1_FREQ"="A1_FREQ", 
                    "anno_eur_A1_CASE_FREQ"="A1_CASE_FREQ",
                    "anno_eur_A1_CTRL_FREQ"="A1_CTRL_FREQ")
    cohort3 = fread(paste0("../processedFilesFromCohorts/dk.",subtype,".freqs.glm.txt"))%>%
      mutate(CHROM=as.character(CHROM),POS=as.character(POS)) %>%
      dplyr::select(CHROM,POS,ID,REF,ALT1,A1_FREQ,A1_CASE_FREQ,A1_CTRL_FREQ) %>%
      dplyr::rename("dk_eur_A1_FREQ"="A1_FREQ", 
                    "dk_eur_A1_CASE_FREQ"="A1_CASE_FREQ",
                    "dk_eur_A1_CTRL_FREQ"="A1_CTRL_FREQ")
    cohorts_af = cohort1 %>%
      full_join(cohort2,by=c("CHROM","POS","ID","REF","ALT1"))%>%
      full_join(cohort3,by=c("CHROM","POS","ID","REF","ALT1"))%>%
      distinct(ID,.keep_all=TRUE)
  }

  ## all races, with anna
  if(grepl("four.races",file)&(!grepl("noanna",file))){
    subtype = gsub("meta.four.races.|.glioma.TBL","",file)
    cohort1 = fread(paste0("../processedFilesFromCohorts/usc.",subtype,".eur.freqs.glm.txt")) %>%
      mutate(CHROM=as.character(CHROM),POS=as.character(POS)) %>%
      dplyr::select(CHROM,POS,ID,REF,ALT1,A1_FREQ,A1_CASE_FREQ,A1_CTRL_FREQ) %>%
      dplyr::rename("usc_eur_A1_FREQ"="A1_FREQ", 
                    "usc_eur_A1_CASE_FREQ"="A1_CASE_FREQ",
                    "usc_eur_A1_CTRL_FREQ"="A1_CTRL_FREQ")
    cohort2 = fread(paste0("../processedFilesFromCohorts/anna.",subtype,".freqs.glm.txt"))%>%
      mutate(CHROM=as.character(CHROM),POS=as.character(POS)) %>%
      dplyr::select(CHROM,POS,ID,REF,ALT1,A1_FREQ,A1_CASE_FREQ,A1_CTRL_FREQ) %>%
      dplyr::rename("anno_eur_A1_FREQ"="A1_FREQ", 
                    "anno_eur_A1_CASE_FREQ"="A1_CASE_FREQ",
                    "anno_eur_A1_CTRL_FREQ"="A1_CTRL_FREQ")
    cohort3 = fread(paste0("../processedFilesFromCohorts/dk.",subtype,".freqs.glm.txt"))%>%
      mutate(CHROM=as.character(CHROM),POS=as.character(POS)) %>%
      dplyr::select(CHROM,POS,ID,REF,ALT1,A1_FREQ,A1_CASE_FREQ,A1_CTRL_FREQ) %>%
      dplyr::rename("dk_eur_A1_FREQ"="A1_FREQ", 
                    "dk_eur_A1_CASE_FREQ"="A1_CASE_FREQ",
                    "dk_eur_A1_CTRL_FREQ"="A1_CTRL_FREQ")
    cohort4 = fread(paste0("../processedFilesFromCohorts/usc.",subtype,".lat.freqs.glm.txt")) %>%
      mutate(CHROM=as.character(CHROM),POS=as.character(POS)) %>%
      dplyr::select(CHROM,POS,ID,REF,ALT1,A1_FREQ,A1_CASE_FREQ,A1_CTRL_FREQ) %>%
      dplyr::rename("usc_lat_A1_FREQ"="A1_FREQ", 
                    "usc_lat_A1_CASE_FREQ"="A1_CASE_FREQ",
                    "usc_lat_A1_CTRL_FREQ"="A1_CTRL_FREQ")
    cohort5 = fread(paste0("../processedFilesFromCohorts/usc.",subtype,".blk.freqs.glm.txt")) %>%
      mutate(CHROM=as.character(CHROM),POS=as.character(POS)) %>%
      dplyr::select(CHROM,POS,ID,REF,ALT1,A1_FREQ,A1_CASE_FREQ,A1_CTRL_FREQ) %>%
      dplyr::rename("usc_blk_A1_FREQ"="A1_FREQ", 
                    "usc_blk_A1_CASE_FREQ"="A1_CASE_FREQ",
                    "usc_blk_A1_CTRL_FREQ"="A1_CTRL_FREQ")
    cohort6 = fread(paste0("../processedFilesFromCohorts/usc.",subtype,".asn.freqs.glm.txt")) %>%
      mutate(CHROM=as.character(CHROM),POS=as.character(POS)) %>%
      dplyr::select(CHROM,POS,ID,REF,ALT1,A1_FREQ,A1_CASE_FREQ,A1_CTRL_FREQ) %>%
      dplyr::rename("usc_asn_A1_FREQ"="A1_FREQ", 
                    "usc_asn_A1_CASE_FREQ"="A1_CASE_FREQ",
                    "usc_asn_A1_CTRL_FREQ"="A1_CTRL_FREQ")
    cohorts_af = cohort1 %>%
      full_join(cohort2,by=c("CHROM","POS","ID","REF","ALT1"))%>%
      full_join(cohort3,by=c("CHROM","POS","ID","REF","ALT1"))%>%
      full_join(cohort4,by=c("CHROM","POS","ID","REF","ALT1"))%>%
      full_join(cohort5,by=c("CHROM","POS","ID","REF","ALT1"))%>%
      full_join(cohort6,by=c("CHROM","POS","ID","REF","ALT1"))%>%
      distinct(ID,.keep_all=TRUE)
  }

  ## eur, without anna
  if(grepl("eur",file)&(grepl("noanna",file))){
    subtype = gsub("meta.noanna.eur.|.glioma.TBL","",file)
    cohort1 = fread(paste0("../processedFilesFromCohorts/usc.",subtype,".eur.freqs.glm.txt")) %>%
      mutate(CHROM=as.character(CHROM),POS=as.character(POS)) %>%
      dplyr::select(CHROM,POS,ID,REF,ALT1,A1_FREQ,A1_CASE_FREQ,A1_CTRL_FREQ) %>%
      dplyr::rename("usc_eur_A1_FREQ"="A1_FREQ", 
                    "usc_eur_A1_CASE_FREQ"="A1_CASE_FREQ",
                    "usc_eur_A1_CTRL_FREQ"="A1_CTRL_FREQ")
    cohort3 = fread(paste0("../processedFilesFromCohorts/dk.",subtype,".freqs.glm.txt"))%>%
      mutate(CHROM=as.character(CHROM),POS=as.character(POS)) %>%
      dplyr::select(CHROM,POS,ID,REF,ALT1,A1_FREQ,A1_CASE_FREQ,A1_CTRL_FREQ) %>%
      dplyr::rename("dk_eur_A1_FREQ"="A1_FREQ", 
                    "dk_eur_A1_CASE_FREQ"="A1_CASE_FREQ",
                    "dk_eur_A1_CTRL_FREQ"="A1_CTRL_FREQ")
    cohorts_af = cohort1 %>%
      full_join(cohort3,by=c("CHROM","POS","ID","REF","ALT1"))%>%
      distinct(ID,.keep_all=TRUE)
  }

  ## all races, without anna
  if(grepl("four.races",file)&(grepl("noanna",file))){
    subtype = gsub("meta.noanna.four.races.|.glioma.TBL","",file)
    cohort1 = fread(paste0("../processedFilesFromCohorts/usc.",subtype,".eur.freqs.glm.txt")) %>%
      mutate(CHROM=as.character(CHROM),POS=as.character(POS)) %>%
      dplyr::select(CHROM,POS,ID,REF,ALT1,A1_FREQ,A1_CASE_FREQ,A1_CTRL_FREQ) %>%
      dplyr::rename("usc_eur_A1_FREQ"="A1_FREQ", 
                    "usc_eur_A1_CASE_FREQ"="A1_CASE_FREQ",
                    "usc_eur_A1_CTRL_FREQ"="A1_CTRL_FREQ")
    cohort3 = fread(paste0("../processedFilesFromCohorts/dk.",subtype,".freqs.glm.txt"))%>%
      mutate(CHROM=as.character(CHROM),POS=as.character(POS)) %>%
      dplyr::select(CHROM,POS,ID,REF,ALT1,A1_FREQ,A1_CASE_FREQ,A1_CTRL_FREQ) %>%
      dplyr::rename("dk_eur_A1_FREQ"="A1_FREQ", 
                    "dk_eur_A1_CASE_FREQ"="A1_CASE_FREQ",
                    "dk_eur_A1_CTRL_FREQ"="A1_CTRL_FREQ")
    cohort4 = fread(paste0("../processedFilesFromCohorts/usc.",subtype,".lat.freqs.glm.txt")) %>%
      mutate(CHROM=as.character(CHROM),POS=as.character(POS)) %>%
      dplyr::select(CHROM,POS,ID,REF,ALT1,A1_FREQ,A1_CASE_FREQ,A1_CTRL_FREQ) %>%
      dplyr::rename("usc_lat_A1_FREQ"="A1_FREQ", 
                    "usc_lat_A1_CASE_FREQ"="A1_CASE_FREQ",
                    "usc_lat_A1_CTRL_FREQ"="A1_CTRL_FREQ")
    cohort5 = fread(paste0("../processedFilesFromCohorts/usc.",subtype,".blk.freqs.glm.txt")) %>%
      mutate(CHROM=as.character(CHROM),POS=as.character(POS)) %>%
      dplyr::select(CHROM,POS,ID,REF,ALT1,A1_FREQ,A1_CASE_FREQ,A1_CTRL_FREQ) %>%
      dplyr::rename("usc_blk_A1_FREQ"="A1_FREQ", 
                    "usc_blk_A1_CASE_FREQ"="A1_CASE_FREQ",
                    "usc_blk_A1_CTRL_FREQ"="A1_CTRL_FREQ")
    cohort6 = fread(paste0("../processedFilesFromCohorts/usc.",subtype,".asn.freqs.glm.txt")) %>%
      mutate(CHROM=as.character(CHROM),POS=as.character(POS)) %>%
      dplyr::select(CHROM,POS,ID,REF,ALT1,A1_FREQ,A1_CASE_FREQ,A1_CTRL_FREQ) %>%
      dplyr::rename("usc_asn_A1_FREQ"="A1_FREQ", 
                    "usc_asn_A1_CASE_FREQ"="A1_CASE_FREQ",
                    "usc_asn_A1_CTRL_FREQ"="A1_CTRL_FREQ")
    cohorts_af = cohort1 %>%
      full_join(cohort3,by=c("CHROM","POS","ID","REF","ALT1"))%>%
      full_join(cohort4,by=c("CHROM","POS","ID","REF","ALT1"))%>%
      full_join(cohort5,by=c("CHROM","POS","ID","REF","ALT1"))%>%
      full_join(cohort6,by=c("CHROM","POS","ID","REF","ALT1"))%>%
      distinct(ID,.keep_all=TRUE)    
  }
  
  ## no europeans
  
  if(grepl("nonwhites",file)){
    
    subtype = gsub("meta.nonwhites.|.glioma.TBL","",file)
    cohort4 = fread(paste0("../processedFilesFromCohorts/usc.",subtype,".lat.freqs.glm.txt")) %>%
      mutate(CHROM=as.character(CHROM),POS=as.character(POS)) %>%
      dplyr::select(CHROM,POS,ID,REF,ALT1,A1_FREQ,A1_CASE_FREQ,A1_CTRL_FREQ) %>%
      dplyr::rename("usc_lat_A1_FREQ"="A1_FREQ", 
                    "usc_lat_A1_CASE_FREQ"="A1_CASE_FREQ",
                    "usc_lat_A1_CTRL_FREQ"="A1_CTRL_FREQ")
    cohort5 = fread(paste0("../processedFilesFromCohorts/usc.",subtype,".blk.freqs.glm.txt")) %>%
      mutate(CHROM=as.character(CHROM),POS=as.character(POS)) %>%
      dplyr::select(CHROM,POS,ID,REF,ALT1,A1_FREQ,A1_CASE_FREQ,A1_CTRL_FREQ) %>%
      dplyr::rename("usc_blk_A1_FREQ"="A1_FREQ", 
                    "usc_blk_A1_CASE_FREQ"="A1_CASE_FREQ",
                    "usc_blk_A1_CTRL_FREQ"="A1_CTRL_FREQ")
    cohort6 = fread(paste0("../processedFilesFromCohorts/usc.",subtype,".asn.freqs.glm.txt")) %>%
      mutate(CHROM=as.character(CHROM),POS=as.character(POS)) %>%
      dplyr::select(CHROM,POS,ID,REF,ALT1,A1_FREQ,A1_CASE_FREQ,A1_CTRL_FREQ) %>%
      dplyr::rename("usc_asn_A1_FREQ"="A1_FREQ", 
                    "usc_asn_A1_CASE_FREQ"="A1_CASE_FREQ",
                    "usc_asn_A1_CTRL_FREQ"="A1_CTRL_FREQ")
    cohorts_af = cohort4 %>%
      full_join(cohort5,by=c("CHROM","POS","ID","REF","ALT1"))%>%
      full_join(cohort6,by=c("CHROM","POS","ID","REF","ALT1"))%>%
      distinct(ID,.keep_all=TRUE)    
  }
  
        
  if(nrow(snpAnno)>0){
    snpAnno = snpAnno %>%
      as.data.frame() %>%
      separate(V3,c("chr","pos","pos1","ref","alt","p"),sep=" ")%>%
      dplyr::rename("rsid"="V2") %>%
      dplyr::select(chr,pos,rsid)
  }
  
  if(nrow(varAnno)>0){
    varAnno = varAnno %>%
      as.data.frame() %>%
      separate(V3,c("chr","pos","pos1","ref","alt","p"),sep=" ")%>%
      dplyr::select(V1,V2,chr,pos) %>%
      dplyr::rename("location"="V1",
                    "geneAnnotation"="V2")
    
  }
  
  if(nrow(exoAnno)>0){
    exoAnno = exoAnno %>%
      as.data.frame() %>%
      separate(V4,c("chr","pos","pos1","ref","alt","p"),sep=" ")%>%
      mutate(geneAnnotation = gsub(":.*$","",V3),
             location="exon") %>%
      dplyr::select(location,geneAnnotation,chr,pos) %>%
      mutate(chr=as.character(chr),
             pos=as.character(pos))
    varAnno = rbind(varAnno,exoAnno) %>% as.data.frame()
  }
  
  inputF = fread(paste0(file,".input")) %>%
    as.data.frame() %>%
    mutate(id=paste0(V1,":",V2))%>%
    distinct(id,.keep_all=TRUE) %>%
    dplyr::select(-id) %>%
    mutate(V1=as.character(V1),
           V2=as.character(V2))%>%
    dplyr::select(-V3,-V4,-V5,-V6)
  colnames(inputF) = c("chr","pos")
  
  allSNPs = fread(file) %>%
    as.data.frame() %>% 
    dplyr::select(MarkerName,Effect,StdErr,`P-value`,Direction) %>%
    mutate(snpid=MarkerName) %>%
    separate(snpid,c("chr","pos","ref","alt"))%>%
    inner_join(inputF,c("chr","pos"))
    
  cianno = allSNPs %>%
    mutate(or=exp(Effect),
           ul = exp(Effect+1.96*StdErr),
           ll = exp(Effect - 1.96*StdErr)) %>%
    mutate(X95CI = paste0(round(ll,3),"-", round(ul,3)))
  
  if(nrow(varAnno)>0){
    cianno = cianno %>%
      left_join(varAnno,by=c("chr","pos"))
  }
  
  if(nrow(snpAnno)>0){
    cianno = cianno %>%
      left_join(snpAnno,by=c("chr","pos"))
  }
  
  cianno = cianno %>%
    distinct(MarkerName,.keep_all=TRUE) %>%
    dplyr::select(chr,pos,MarkerName,rsid,ref,alt,Effect,or,ul,ll,X95CI,`P-value`,location,geneAnnotation,Direction) %>%
    arrange(chr,pos) %>%
    left_join(cohorts_af,c("chr"="CHROM","pos"="POS","MarkerName"="ID",
                           "ref"="REF","alt"="ALT1"))
  
  # filename = paste0("METAL.",gsub(".TBL",".jon.annotation.txt",file))
  # write.table(cianno,filename,
  #             row.names = FALSE,
  #             quote=FALSE,
  #             sep='\t')
  
  filename = paste0("METAL.",gsub(".TBL",".jon.annotation.csv",file))
  write.csv2(cianno,filename)

}

############################################
# ast, for eur look-up in non-eur population

setwd("/project/wiemels_494/sebastian/glioma_gwas/meta/METAL")

lookups = fread("METAL.meta.four.races.ast.glioma.jon.annotation.csv") %>%
  as.data.frame() %>%
  mutate(allRacesEffect = Effect,
         allRacesOr = or,
         allRacesUl = ul,
         allRacesLl = ll,
         allRacesX95CI = X95CI,
         allRacesP = `P-value`,
         allRacesDirection = Direction) %>%
  dplyr::select(chr,pos,MarkerName,rsid,ref,alt,location,geneAnnotation,
                allRacesEffect,allRacesOr,allRacesUl,allRacesLl,allRacesX95CI,allRacesP,allRacesDirection,
                usc_eur_A1_FREQ,       
                usc_eur_A1_CASE_FREQ, 
                usc_eur_A1_CTRL_FREQ,  
                anno_eur_A1_FREQ,      
                anno_eur_A1_CASE_FREQ,
                anno_eur_A1_CTRL_FREQ, 
                dk_eur_A1_FREQ,        
                dk_eur_A1_CASE_FREQ,  
                dk_eur_A1_CTRL_FREQ,   
                usc_lat_A1_FREQ,       
                usc_lat_A1_CASE_FREQ, 
                usc_lat_A1_CTRL_FREQ,  
                usc_blk_A1_FREQ,       
                usc_blk_A1_CASE_FREQ, 
                usc_blk_A1_CTRL_FREQ,  
                usc_asn_A1_FREQ,       
                usc_asn_A1_CASE_FREQ, 
                usc_asn_A1_CTRL_FREQ)

file="meta.nonwhites.ast.glioma.TBL"

allSNPs = fread(file) %>%
  as.data.frame() %>% 
  mutate(nonEurEffect = Effect,
         nonEurStdErr = StdErr,
         nonEurP = `P-value`,
         nonEurDirection = Direction) %>%
  dplyr::select(MarkerName,nonEurEffect,nonEurStdErr,nonEurP,nonEurDirection) %>%
  inner_join(lookups,by="MarkerName")

cianno = allSNPs %>%
  mutate(nonEurOr=exp(nonEurEffect),
         nonEurUl = exp(nonEurEffect+1.96*nonEurStdErr),
         nonEurLl = exp(nonEurEffect - 1.96*nonEurStdErr)) %>%
  mutate(nonEurX95CI = paste0(round(nonEurLl,3),"-", round(nonEurUl,3))) %>%
  mutate(chr = as.character(chr),
         pos = as.character(pos))
  
cianno = cianno %>%
  distinct(MarkerName,.keep_all=TRUE) %>%
  arrange(chr,pos) 

filename = paste0("METAL.",gsub(".TBL",".jon.annotation.lookup.csv",file))
write.csv2(cianno,filename)

################################################################################################
# metafor
library(tidyverse)
library(data.table)

setwd("/project/wiemels_494/sebastian/glioma_gwas/meta/metafor")

files =  c("meta.eur.all.glioma.metafor.txt",
          "meta.eur.ast.glioma.metafor.txt",
          "meta.eur.glm.glioma.metafor.txt",
          "meta.eur.hga.glioma.metafor.txt",
          "meta.eur.hgg.glioma.metafor.txt",
          "meta.eur.lga.glioma.metafor.txt",
          "meta.eur.lgg.glioma.metafor.txt",
          "meta.eur.olg.glioma.metafor.txt",
          "meta.eur.pa.glioma.metafor.txt", 
          "meta.eur.pao.glioma.metafor.txt",
          "meta.four.races.all.glioma.metafor.txt",
          "meta.four.races.ast.glioma.metafor.txt",
          "meta.four.races.glm.glioma.metafor.txt",
          "meta.four.races.hga.glioma.metafor.txt",
          "meta.four.races.hgg.glioma.metafor.txt",
          "meta.four.races.lga.glioma.metafor.txt",
          "meta.four.races.lgg.glioma.metafor.txt",
          "meta.four.races.olg.glioma.metafor.txt",
          "meta.four.races.pa.glioma.metafor.txt", 
          "meta.four.races.pao.glioma.metafor.txt",
          "meta.noanna.eur.glm.glioma.metafor.txt",
          "meta.noanna.eur.olg.glioma.metafor.txt",
          "meta.noanna.four.races.glm.glioma.metafor.txt",
          "meta.noanna.four.races.olg.glioma.metafor.txt")

for(file in files){
  
  #file="meta.four.races.pa.glioma.metafor.txt"
  snpAnno = fread(paste0(file,".metafor.input.hg19_avsnp142_dropped"),header=FALSE)
  varAnno = fread(paste0("anno.variant.",file,".metafor"),header=FALSE) 
  exoAnno = fread(paste0("anno.exonic.",file,".metafor"),sep='\t',header=FALSE)
  
  directionFile = fread(paste0("/project/wiemels_494/sebastian/glioma_gwas/meta/METAL/",gsub(".metafor.txt",".TBL",file))) %>%
    as.data.frame() %>%
    separate(MarkerName,c("chr","pos","ref","alt"))%>%
    dplyr::select(chr,pos,Direction) 

  # allele frequency information annotation
  ## eur, with anna
  if(grepl("eur",file)&(!grepl("noanna",file))){
    subtype = gsub("meta.eur.|.glioma.metafor.txt","",file)
    cohort1 = fread(paste0("../processedFilesFromCohorts/usc.",subtype,".eur.freqs.glm.txt")) %>%
      mutate(CHROM=as.character(CHROM),POS=as.character(POS)) %>%
      dplyr::select(CHROM,POS,ID,REF,ALT1,A1_FREQ,A1_CASE_FREQ,A1_CTRL_FREQ) %>%
      dplyr::rename("usc_eur_A1_FREQ"="A1_FREQ", 
                    "usc_eur_A1_CASE_FREQ"="A1_CASE_FREQ",
                    "usc_eur_A1_CTRL_FREQ"="A1_CTRL_FREQ")
    cohort2 = fread(paste0("../processedFilesFromCohorts/anna.",subtype,".freqs.glm.txt"))%>%
      mutate(CHROM=as.character(CHROM),POS=as.character(POS)) %>%
      dplyr::select(CHROM,POS,ID,REF,ALT1,A1_FREQ,A1_CASE_FREQ,A1_CTRL_FREQ) %>%
      dplyr::rename("anno_eur_A1_FREQ"="A1_FREQ", 
                    "anno_eur_A1_CASE_FREQ"="A1_CASE_FREQ",
                    "anno_eur_A1_CTRL_FREQ"="A1_CTRL_FREQ")
    cohort3 = fread(paste0("../processedFilesFromCohorts/dk.",subtype,".freqs.glm.txt"))%>%
      mutate(CHROM=as.character(CHROM),POS=as.character(POS)) %>%
      dplyr::select(CHROM,POS,ID,REF,ALT1,A1_FREQ,A1_CASE_FREQ,A1_CTRL_FREQ) %>%
      dplyr::rename("dk_eur_A1_FREQ"="A1_FREQ", 
                    "dk_eur_A1_CASE_FREQ"="A1_CASE_FREQ",
                    "dk_eur_A1_CTRL_FREQ"="A1_CTRL_FREQ")
    cohorts_af = cohort1 %>%
      full_join(cohort2,by=c("CHROM","POS","ID","REF","ALT1"))%>%
      full_join(cohort3,by=c("CHROM","POS","ID","REF","ALT1"))%>%
      distinct(ID,.keep_all=TRUE)
  }

  ## all races, with anna
  if(grepl("four.races",file)&(!grepl("noanna",file))){
    subtype = gsub("meta.four.races.|.glioma.metafor.txt","",file)
    cohort1 = fread(paste0("../processedFilesFromCohorts/usc.",subtype,".eur.freqs.glm.txt")) %>%
      mutate(CHROM=as.character(CHROM),POS=as.character(POS)) %>%
      dplyr::select(CHROM,POS,ID,REF,ALT1,A1_FREQ,A1_CASE_FREQ,A1_CTRL_FREQ) %>%
      dplyr::rename("usc_eur_A1_FREQ"="A1_FREQ", 
                    "usc_eur_A1_CASE_FREQ"="A1_CASE_FREQ",
                    "usc_eur_A1_CTRL_FREQ"="A1_CTRL_FREQ")
    cohort2 = fread(paste0("../processedFilesFromCohorts/anna.",subtype,".freqs.glm.txt"))%>%
      mutate(CHROM=as.character(CHROM),POS=as.character(POS)) %>%
      dplyr::select(CHROM,POS,ID,REF,ALT1,A1_FREQ,A1_CASE_FREQ,A1_CTRL_FREQ) %>%
      dplyr::rename("anno_eur_A1_FREQ"="A1_FREQ", 
                    "anno_eur_A1_CASE_FREQ"="A1_CASE_FREQ",
                    "anno_eur_A1_CTRL_FREQ"="A1_CTRL_FREQ")
    cohort3 = fread(paste0("../processedFilesFromCohorts/dk.",subtype,".freqs.glm.txt"))%>%
      mutate(CHROM=as.character(CHROM),POS=as.character(POS)) %>%
      dplyr::select(CHROM,POS,ID,REF,ALT1,A1_FREQ,A1_CASE_FREQ,A1_CTRL_FREQ) %>%
      dplyr::rename("dk_eur_A1_FREQ"="A1_FREQ", 
                    "dk_eur_A1_CASE_FREQ"="A1_CASE_FREQ",
                    "dk_eur_A1_CTRL_FREQ"="A1_CTRL_FREQ")
    cohort4 = fread(paste0("../processedFilesFromCohorts/usc.",subtype,".lat.freqs.glm.txt")) %>%
      mutate(CHROM=as.character(CHROM),POS=as.character(POS)) %>%
      dplyr::select(CHROM,POS,ID,REF,ALT1,A1_FREQ,A1_CASE_FREQ,A1_CTRL_FREQ) %>%
      dplyr::rename("usc_lat_A1_FREQ"="A1_FREQ", 
                    "usc_lat_A1_CASE_FREQ"="A1_CASE_FREQ",
                    "usc_lat_A1_CTRL_FREQ"="A1_CTRL_FREQ")
    cohort5 = fread(paste0("../processedFilesFromCohorts/usc.",subtype,".blk.freqs.glm.txt")) %>%
      mutate(CHROM=as.character(CHROM),POS=as.character(POS)) %>%
      dplyr::select(CHROM,POS,ID,REF,ALT1,A1_FREQ,A1_CASE_FREQ,A1_CTRL_FREQ) %>%
      dplyr::rename("usc_blk_A1_FREQ"="A1_FREQ", 
                    "usc_blk_A1_CASE_FREQ"="A1_CASE_FREQ",
                    "usc_blk_A1_CTRL_FREQ"="A1_CTRL_FREQ")
    cohort6 = fread(paste0("../processedFilesFromCohorts/usc.",subtype,".asn.freqs.glm.txt")) %>%
      mutate(CHROM=as.character(CHROM),POS=as.character(POS)) %>%
      dplyr::select(CHROM,POS,ID,REF,ALT1,A1_FREQ,A1_CASE_FREQ,A1_CTRL_FREQ) %>%
      dplyr::rename("usc_asn_A1_FREQ"="A1_FREQ", 
                    "usc_asn_A1_CASE_FREQ"="A1_CASE_FREQ",
                    "usc_asn_A1_CTRL_FREQ"="A1_CTRL_FREQ")
    cohorts_af = cohort1 %>%
      full_join(cohort2,by=c("CHROM","POS","ID","REF","ALT1"))%>%
      full_join(cohort3,by=c("CHROM","POS","ID","REF","ALT1"))%>%
      full_join(cohort4,by=c("CHROM","POS","ID","REF","ALT1"))%>%
      full_join(cohort5,by=c("CHROM","POS","ID","REF","ALT1"))%>%
      full_join(cohort6,by=c("CHROM","POS","ID","REF","ALT1"))%>%
      distinct(ID,.keep_all=TRUE)
  }

  ## eur, without anna
  if(grepl("eur",file)&(grepl("noanna",file))){
    subtype = gsub("meta.noanna.eur.|.glioma.metafor.txt","",file)
    cohort1 = fread(paste0("../processedFilesFromCohorts/usc.",subtype,".eur.freqs.glm.txt")) %>%
      mutate(CHROM=as.character(CHROM),POS=as.character(POS)) %>%
      dplyr::select(CHROM,POS,ID,REF,ALT1,A1_FREQ,A1_CASE_FREQ,A1_CTRL_FREQ) %>%
      dplyr::rename("usc_eur_A1_FREQ"="A1_FREQ", 
                    "usc_eur_A1_CASE_FREQ"="A1_CASE_FREQ",
                    "usc_eur_A1_CTRL_FREQ"="A1_CTRL_FREQ")
    cohort3 = fread(paste0("../processedFilesFromCohorts/dk.",subtype,".freqs.glm.txt"))%>%
      mutate(CHROM=as.character(CHROM),POS=as.character(POS)) %>%
      dplyr::select(CHROM,POS,ID,REF,ALT1,A1_FREQ,A1_CASE_FREQ,A1_CTRL_FREQ) %>%
      dplyr::rename("dk_eur_A1_FREQ"="A1_FREQ", 
                    "dk_eur_A1_CASE_FREQ"="A1_CASE_FREQ",
                    "dk_eur_A1_CTRL_FREQ"="A1_CTRL_FREQ")
    cohorts_af = cohort1 %>%
      full_join(cohort3,by=c("CHROM","POS","ID","REF","ALT1"))%>%
      distinct(ID,.keep_all=TRUE)
  }

  ## all races, without anna
  if(grepl("four.races",file)&(grepl("noanna",file))){
    subtype = gsub("meta.noanna.four.races.|.glioma.metafor.txt","",file)
    cohort1 = fread(paste0("../processedFilesFromCohorts/usc.",subtype,".eur.freqs.glm.txt")) %>%
      mutate(CHROM=as.character(CHROM),POS=as.character(POS)) %>%
      dplyr::select(CHROM,POS,ID,REF,ALT1,A1_FREQ,A1_CASE_FREQ,A1_CTRL_FREQ) %>%
      dplyr::rename("usc_eur_A1_FREQ"="A1_FREQ", 
                    "usc_eur_A1_CASE_FREQ"="A1_CASE_FREQ",
                    "usc_eur_A1_CTRL_FREQ"="A1_CTRL_FREQ")
    cohort3 = fread(paste0("../processedFilesFromCohorts/dk.",subtype,".freqs.glm.txt"))%>%
      mutate(CHROM=as.character(CHROM),POS=as.character(POS)) %>%
      dplyr::select(CHROM,POS,ID,REF,ALT1,A1_FREQ,A1_CASE_FREQ,A1_CTRL_FREQ) %>%
      dplyr::rename("dk_eur_A1_FREQ"="A1_FREQ", 
                    "dk_eur_A1_CASE_FREQ"="A1_CASE_FREQ",
                    "dk_eur_A1_CTRL_FREQ"="A1_CTRL_FREQ")
    cohort4 = fread(paste0("../processedFilesFromCohorts/usc.",subtype,".lat.freqs.glm.txt")) %>%
      mutate(CHROM=as.character(CHROM),POS=as.character(POS)) %>%
      dplyr::select(CHROM,POS,ID,REF,ALT1,A1_FREQ,A1_CASE_FREQ,A1_CTRL_FREQ) %>%
      dplyr::rename("usc_lat_A1_FREQ"="A1_FREQ", 
                    "usc_lat_A1_CASE_FREQ"="A1_CASE_FREQ",
                    "usc_lat_A1_CTRL_FREQ"="A1_CTRL_FREQ")
    cohort5 = fread(paste0("../processedFilesFromCohorts/usc.",subtype,".blk.freqs.glm.txt")) %>%
      mutate(CHROM=as.character(CHROM),POS=as.character(POS)) %>%
      dplyr::select(CHROM,POS,ID,REF,ALT1,A1_FREQ,A1_CASE_FREQ,A1_CTRL_FREQ) %>%
      dplyr::rename("usc_blk_A1_FREQ"="A1_FREQ", 
                    "usc_blk_A1_CASE_FREQ"="A1_CASE_FREQ",
                    "usc_blk_A1_CTRL_FREQ"="A1_CTRL_FREQ")
    cohort6 = fread(paste0("../processedFilesFromCohorts/usc.",subtype,".asn.freqs.glm.txt")) %>%
      mutate(CHROM=as.character(CHROM),POS=as.character(POS)) %>%
      dplyr::select(CHROM,POS,ID,REF,ALT1,A1_FREQ,A1_CASE_FREQ,A1_CTRL_FREQ) %>%
      dplyr::rename("usc_asn_A1_FREQ"="A1_FREQ", 
                    "usc_asn_A1_CASE_FREQ"="A1_CASE_FREQ",
                    "usc_asn_A1_CTRL_FREQ"="A1_CTRL_FREQ")
    cohorts_af = cohort1 %>%
      full_join(cohort3,by=c("CHROM","POS","ID","REF","ALT1"))%>%
      full_join(cohort4,by=c("CHROM","POS","ID","REF","ALT1"))%>%
      full_join(cohort5,by=c("CHROM","POS","ID","REF","ALT1"))%>%
      full_join(cohort6,by=c("CHROM","POS","ID","REF","ALT1"))%>%
      as.data.frame() %>%
      distinct(ID,.keep_all=TRUE)    
    
  }
     
      
  if(nrow(snpAnno)>0){
    snpAnno = snpAnno %>%
      as.data.frame() %>%
      dplyr::rename("rsid"="V2",
                    "chr"="V3",
                    "pos"="V4") %>%
      dplyr::select(chr,pos,rsid) %>%
      mutate(chr=as.character(chr),
             pos=as.character(pos))
      
  }
  
  if(nrow(varAnno)>0){
    varAnno = varAnno %>%
      as.data.frame() %>%
      dplyr::select(V1,V2,V3,V4) %>%
      dplyr::rename("location"="V1",
                    "geneAnnotation"="V2",
                    "chr"="V3",
                    "pos"="V4") %>%
      mutate(chr=as.character(chr),
             pos=as.character(pos))
    
  }
  
  if(nrow(exoAnno)>0){
    exoAnno = exoAnno %>%
      as.data.frame() %>%
      dplyr::rename("chr"="V4",
                    "pos"="V5") %>%
      mutate(geneAnnotation = gsub(":.*$","",V3),
             location="exon") %>%
      dplyr::select(location,geneAnnotation,chr,pos)%>%
      mutate(chr=as.character(chr),
             pos=as.character(pos))
    varAnno = rbind(varAnno,exoAnno) %>% as.data.frame()
  }
  
  inputF = fread(paste0(file,".metafor.input")) %>%
    as.data.frame() %>%
    mutate(id=paste0(V1,":",V2))%>%
    distinct(id,.keep_all=TRUE) %>%
    dplyr::select(-id) %>%
    mutate(V1=as.character(V1),
           V2=as.character(V2))%>%
    dplyr::select(-V3,-V4,-V5,-V6)
  colnames(inputF) = c("chr","pos")
  
  allSNPs = fread(file) %>%
    as.data.frame() %>% 
    mutate(MarkerName=ID,
           Effect=coef.fe,
           StdErr=se.fe,
           `P-value`=p.fe) %>%
    dplyr::select(MarkerName,Effect,StdErr,`P-value`) %>%
    mutate(snpid=MarkerName) %>%
    separate(snpid,c("chr","pos","ref","alt"))%>%
    inner_join(inputF,c("chr","pos"))
    
  cianno = allSNPs %>%
    distinct(MarkerName,.keep_all=TRUE) %>%
    mutate(or=exp(Effect),
           ul = exp(Effect+1.96*StdErr),
           ll = exp(Effect - 1.96*StdErr)) %>%
    mutate(X95CI = paste0(round(ll,3),"-", round(ul,3)))
  
  if(nrow(varAnno)>0){
    cianno = cianno %>%
      left_join(varAnno,by=c("chr","pos"))
  }
  
  if(nrow(snpAnno)>0){
    cianno = cianno %>%
      left_join(snpAnno,by=c("chr","pos"))
  }
  
  cianno = cianno %>%
    distinct(MarkerName,.keep_all=TRUE) %>%
    left_join(directionFile,by=c("chr","pos")) %>%
    dplyr::select(chr,pos,MarkerName,rsid,ref,alt,Effect,or,ul,ll,X95CI,`P-value`,location,geneAnnotation,Direction) %>%
    arrange(chr,pos)%>%
    left_join(cohorts_af,c("chr"="CHROM","pos"="POS","MarkerName"="ID",
                           "ref"="REF","alt"="ALT1"))
 
  # filename = paste0("metafor.",gsub(".metafor.txt",".jon.annotation.txt",file))
  # write.table(cianno,filename,
  #             row.names = FALSE,
  #             quote=FALSE,
  #             sep='\t')

  filename = paste0("metafor.",gsub(".metafor.txt",".jon.annotation.csv",file))
  write.csv2(cianno,filename)
  
}

```

```{r MAF files}

library(tidyverse)
library(data.table)

setwd("/project/wiemels_494/sebastian/glioma_gwas/meta")

for(subtype in c("all","lgg","hgg","ast","lga","hga","pa","pao","glm","olg")){
  
  #subtype="all"
  
  if(subtype=="all"){
    file_sigs = "/project/wiemels_494/sebastian/glioma_gwas/meta/METAL/METAL.meta.four.races.all.glioma.jon.annotation.csv"
    file_anna = "/project/wiemels_494/sebastian/glioma_gwas/anna_cohort/concat.all.var_3PCA.20220401.PHENO1.glm.logistic.hybrid"
    file_jon = "/project/wiemels_494/sebastian/glioma_gwas/dk_cohort/Results/Glioma.PC5.PHENO1.glm.logistic.hybrid"
    file_usc_eur = "/project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.all.Subs.eur.phenotype.glm.logistic.hybrid"
    file_usc_lat = "/project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.all.Subs.lat.phenotype.glm.logistic.hybrid"
    file_usc_blk = "/project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.all.Subs.blk.phenotype.glm.logistic.hybrid"
    file_usc_asn = "/project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.all.Subs.asn.phenotype.glm.logistic.hybrid"
  }
  
  
  if(subtype=="lgg"){
    file_sigs = "/project/wiemels_494/sebastian/glioma_gwas/meta/METAL/METAL.meta.four.races.lgg.glioma.jon.annotation.csv"
    file_anna = "/project/wiemels_494/sebastian/glioma_gwas/anna_cohort/concat.lgg.var_0PCA.20220401.PHENO1.glm.logistic.hybrid"
    file_jon = "/project/wiemels_494/sebastian/glioma_gwas/dk_cohort/Results/LGG.PC5.PHENO1.glm.logistic.hybrid"
    file_usc_eur = "/project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.lgg.Subs.eur.phenotype.glm.logistic.hybrid"
    file_usc_lat = "/project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.lgg.Subs.lat.phenotype.glm.logistic.hybrid"
    file_usc_blk = "/project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.lgg.Subs.blk.phenotype.glm.logistic.hybrid"
    file_usc_asn = "/project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.lgg.Subs.asn.phenotype.glm.logistic.hybrid"
  }
  
  if(subtype=="hgg"){
    file_sigs = "/project/wiemels_494/sebastian/glioma_gwas/meta/METAL/METAL.meta.four.races.hgg.glioma.jon.annotation.csv"
    file_anna = "/project/wiemels_494/sebastian/glioma_gwas/anna_cohort/concat.hgg.var_1PCA.20220401.PHENO1.glm.logistic.hybrid"
    file_jon = "/project/wiemels_494/sebastian/glioma_gwas/dk_cohort/Results/HGG.PC5.PHENO1.glm.logistic.hybrid"
    file_usc_eur = "/project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.hgg.Subs.eur.phenotype.glm.logistic.hybrid"
    file_usc_lat = "/project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.hgg.Subs.lat.phenotype.glm.logistic.hybrid"
    file_usc_blk = "/project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.hgg.Subs.blk.phenotype.glm.logistic.hybrid"
    file_usc_asn = "/project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.hgg.Subs.asn.phenotype.glm.logistic.hybrid"
  }
  
  if(subtype=="ast"){
    file_sigs = "/project/wiemels_494/sebastian/glioma_gwas/meta/METAL/METAL.meta.four.races.ast.glioma.jon.annotation.csv"
    file_anna = "/project/wiemels_494/sebastian/glioma_gwas/anna_cohort/concat.astro.var_10PCA.20220401.PHENO1.glm.logistic.hybrid"
    file_jon = "/project/wiemels_494/sebastian/glioma_gwas/dk_cohort/Results/Astrocytoma.PC5.PHENO1.glm.logistic.hybrid"
    file_usc_eur = "/project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.ast.Subs.eur.phenotype.glm.logistic.hybrid"
    file_usc_lat = "/project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.ast.Subs.lat.phenotype.glm.logistic.hybrid"
    file_usc_blk = "/project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.ast.Subs.blk.phenotype.glm.logistic.hybrid"
    file_usc_asn = "/project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.ast.Subs.asn.phenotype.glm.logistic.hybrid"
  }
  
  if(subtype=="lga"){
    file_sigs = "/project/wiemels_494/sebastian/glioma_gwas/meta/METAL/METAL.meta.four.races.lga.glioma.jon.annotation.csv"
    file_anna = "/project/wiemels_494/sebastian/glioma_gwas/anna_cohort/concat.lg_astro.var_8PCA.20220401.PHENO1.glm.logistic.hybrid"
    file_jon = "/project/wiemels_494/sebastian/glioma_gwas/dk_cohort/Results/LG_astrocytoma.PC5.PHENO1.glm.logistic.hybrid"
    file_usc_eur = "/project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.lga.Subs.eur.phenotype.glm.logistic.hybrid"
    file_usc_lat = "/project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.lga.Subs.lat.phenotype.glm.logistic.hybrid"
    file_usc_blk = "/project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.lga.Subs.blk.phenotype.glm.logistic.hybrid"
    file_usc_asn = "/project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.lga.Subs.asn.phenotype.glm.logistic.hybrid"
  }
  
  if(subtype=="hga"){
    file_sigs = "/project/wiemels_494/sebastian/glioma_gwas/meta/METAL/METAL.meta.four.races.hga.glioma.jon.annotation.csv"
    file_anna = "/project/wiemels_494/sebastian/glioma_gwas/anna_cohort/concat.hg_astro.var_3PCA.20220401.PHENO1.glm.logistic.hybrid"
    file_jon = "/project/wiemels_494/sebastian/glioma_gwas/dk_cohort/Results/HG_astrocytoma.PC5.PHENO1.glm.logistic.hybrid"
    file_usc_eur = "/project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.hga.Subs.eur.phenotype.glm.logistic.hybrid"
    file_usc_lat = "/project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.hga.Subs.lat.phenotype.glm.logistic.hybrid"
    file_usc_blk = "/project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.hga.Subs.blk.phenotype.glm.logistic.hybrid"
    file_usc_asn = "/project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.hga.Subs.asn.phenotype.glm.logistic.hybrid"
  }
  
  if(subtype=="pa"){
    file_sigs = "/project/wiemels_494/sebastian/glioma_gwas/meta/METAL/METAL.meta.four.races.pa.glioma.jon.annotation.csv"
    file_anna = "/project/wiemels_494/sebastian/glioma_gwas/anna_cohort/concat.pilo.var_0PCA.20220401.PHENO1.glm.logistic.hybrid"
    file_jon = "/project/wiemels_494/sebastian/glioma_gwas/dk_cohort/Results/Pilocytic_astrocytoma.PC5.PHENO1.glm.logistic.hybrid"
    file_usc_eur = "/project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.pa.Subs.eur.phenotype.glm.logistic.hybrid"
    file_usc_lat = "/project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.pa.Subs.lat.phenotype.glm.logistic.hybrid"
    file_usc_blk = "/project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.pa.Subs.blk.phenotype.glm.logistic.hybrid"
    file_usc_asn = "/project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.pa.Subs.asn.phenotype.glm.logistic.hybrid"
  }
  
  if(subtype=="pao"){
    file_sigs = "/project/wiemels_494/sebastian/glioma_gwas/meta/METAL/METAL.meta.four.races.pao.glioma.jon.annotation.csv"
    file_anna = "/project/wiemels_494/sebastian/glioma_gwas/anna_cohort/concat.pilo_optic.var_0PCA.20220401.PHENO1.glm.logistic.hybrid"
    file_jon = "/project/wiemels_494/sebastian/glioma_gwas/dk_cohort/Results/Pilocytic_astrocytoma_incl_optic.PC5.PHENO1.glm.logistic.hybrid"
    file_usc_eur = "/project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.pao.Subs.eur.phenotype.glm.logistic.hybrid"
    file_usc_lat = "/project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.pao.Subs.lat.phenotype.glm.logistic.hybrid"
    file_usc_blk = "/project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.pao.Subs.blk.phenotype.glm.logistic.hybrid"
    file_usc_asn = "/project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.pao.Subs.asn.phenotype.glm.logistic.hybrid"
  }
  
  if(subtype=="glm"){
    file_sigs = "/project/wiemels_494/sebastian/glioma_gwas/meta/METAL/METAL.meta.four.races.glm.glioma.jon.annotation.csv"
    file_anna = "/project/wiemels_494/sebastian/glioma_gwas/anna_cohort/concat.gbm.var_0PCA.20220401.PHENO1.glm.logistic.hybrid"
    file_jon = "/project/wiemels_494/sebastian/glioma_gwas/dk_cohort/Results/Glioblastoma.PC5.PHENO1.glm.logistic.hybrid"
    file_usc_eur = "/project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.glm.Subs.eur.phenotype.glm.logistic.hybrid"
    file_usc_lat = "/project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.glm.Subs.lat.phenotype.glm.logistic.hybrid"
    file_usc_blk = "/project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.glm.Subs.blk.phenotype.glm.logistic.hybrid"
    file_usc_asn = "/project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.glm.Subs.asn.phenotype.glm.logistic.hybrid"
  }
  
  if(subtype=="olg"){
    file_sigs = "/project/wiemels_494/sebastian/glioma_gwas/meta/METAL/METAL.meta.four.races.olg.glioma.jon.annotation.csv"
    file_anna = "/project/wiemels_494/sebastian/glioma_gwas/anna_cohort/concat.oligo.var_0PCA.20220401.PHENO1.glm.logistic.hybrid"
    file_jon = "/project/wiemels_494/sebastian/glioma_gwas/dk_cohort/Results/Oligodendroglioma.PC5.PHENO1.glm.logistic.hybrid"
    file_usc_eur = "/project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.olg.Subs.eur.phenotype.glm.logistic.hybrid"
    file_usc_lat = "/project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.olg.Subs.lat.phenotype.glm.logistic.hybrid"
    file_usc_blk = "/project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.olg.Subs.blk.phenotype.glm.logistic.hybrid"
    file_usc_asn = "/project/wiemels_494/sebastian/glioma_gwas/usc_cohort/hg19.olg.Subs.asn.phenotype.glm.logistic.hybrid"
  }
  
  
  anna = fread(file_anna) %>% as.data.frame() %>%
    mutate(id = paste0(`#CHROM`,":",POS,":",REF,":",ALT1),
           CASE_ALLELE_CT = A1_CASE_CT/A1_CASE_FREQ,
           CTRL_ALLELE_CT = A1_CTRL_CT/A1_CTRL_FREQ) %>%
    na.omit() %>%
    dplyr::rename("A1_CASE_CT_anna"="A1_CASE_CT",
                  "A1_CTRL_CT_anna"="A1_CTRL_CT",
                  "CASE_ALLELE_CT_anna"="CASE_ALLELE_CT",
                  "CTRL_ALLELE_CT_anna" = "CTRL_ALLELE_CT") %>%
    dplyr::select(id,A1_CASE_CT_anna,A1_CTRL_CT_anna,CASE_ALLELE_CT_anna,CTRL_ALLELE_CT_anna)
  
  jon = fread(file_jon) %>% as.data.frame()%>%
    mutate(id = paste0(`#CHROM`,":",POS,":",REF,":",ALT)) %>%
    na.omit() %>%
    dplyr::rename("A1_CASE_CT_jon"="A1_CASE_CT",
                  "A1_CTRL_CT_jon"="A1_CTRL_CT",
                  "CASE_ALLELE_CT_jon"="CASE_ALLELE_CT",
                  "CTRL_ALLELE_CT_jon" = "CTRL_ALLELE_CT") %>%
    dplyr::select(id,A1_CASE_CT_jon,A1_CTRL_CT_jon,CASE_ALLELE_CT_jon,CTRL_ALLELE_CT_jon)
  
  usc_eur = fread(file_usc_eur) %>% as.data.frame() %>%
    mutate(`#CHROM` = sub("chr","",`#CHROM`),
           id = paste0(`#CHROM`,":",POS,":",REF,":",ALT1),
           CASE_ALLELE_CT = A1_CASE_CT/A1_CASE_FREQ,
           CTRL_ALLELE_CT = A1_CTRL_CT/A1_CTRL_FREQ) %>%
    na.omit() %>%
    dplyr::rename("A1_CASE_CT_usc_eur"="A1_CASE_CT",
                  "A1_CTRL_CT_usc_eur"="A1_CTRL_CT",
                  "CASE_ALLELE_CT_usc_eur"="CASE_ALLELE_CT",
                  "CTRL_ALLELE_CT_usc_eur" = "CTRL_ALLELE_CT") %>%
    dplyr::select(id,A1_CASE_CT_usc_eur,A1_CTRL_CT_usc_eur,CASE_ALLELE_CT_usc_eur,CTRL_ALLELE_CT_usc_eur)
  
  usc_lat = fread(file_usc_lat) %>% as.data.frame()%>%
    mutate(`#CHROM` = sub("chr","",`#CHROM`),
           id = paste0(`#CHROM`,":",POS,":",REF,":",ALT1),
           CASE_ALLELE_CT = A1_CASE_CT/A1_CASE_FREQ,
           CTRL_ALLELE_CT = A1_CTRL_CT/A1_CTRL_FREQ) %>%
    na.omit() %>%
    dplyr::rename("A1_CASE_CT_usc_lat"="A1_CASE_CT",
                  "A1_CTRL_CT_usc_lat"="A1_CTRL_CT",
                  "CASE_ALLELE_CT_usc_lat"="CASE_ALLELE_CT",
                  "CTRL_ALLELE_CT_usc_lat" = "CTRL_ALLELE_CT") %>%
    dplyr::select(id,A1_CASE_CT_usc_lat,A1_CTRL_CT_usc_lat,CASE_ALLELE_CT_usc_lat,CTRL_ALLELE_CT_usc_lat)
  
  usc_blk = fread(file_usc_blk) %>% as.data.frame()%>%
    mutate(`#CHROM` = sub("chr","",`#CHROM`),
           id = paste0(`#CHROM`,":",POS,":",REF,":",ALT1),
           CASE_ALLELE_CT = A1_CASE_CT/A1_CASE_FREQ,
           CTRL_ALLELE_CT = A1_CTRL_CT/A1_CTRL_FREQ) %>%
    na.omit() %>%
    dplyr::rename("A1_CASE_CT_usc_blk"="A1_CASE_CT",
                  "A1_CTRL_CT_usc_blk"="A1_CTRL_CT",
                  "CASE_ALLELE_CT_usc_blk"="CASE_ALLELE_CT",
                  "CTRL_ALLELE_CT_usc_blk" = "CTRL_ALLELE_CT") %>%
    dplyr::select(id,A1_CASE_CT_usc_blk,A1_CTRL_CT_usc_blk,CASE_ALLELE_CT_usc_blk,CTRL_ALLELE_CT_usc_blk)
  
  usc_asn = fread(file_usc_asn) %>% as.data.frame()%>%
    mutate(`#CHROM` = sub("chr","",`#CHROM`),
           id = paste0(`#CHROM`,":",POS,":",REF,":",ALT1),
           CASE_ALLELE_CT = A1_CASE_CT/A1_CASE_FREQ,
           CTRL_ALLELE_CT = A1_CTRL_CT/A1_CTRL_FREQ) %>%
    na.omit() %>%
    dplyr::rename("A1_CASE_CT_usc_asn"="A1_CASE_CT",
                  "A1_CTRL_CT_usc_asn"="A1_CTRL_CT",
                  "CASE_ALLELE_CT_usc_asn"="CASE_ALLELE_CT",
                  "CTRL_ALLELE_CT_usc_asn" = "CTRL_ALLELE_CT") %>%
    dplyr::select(id,A1_CASE_CT_usc_asn,A1_CTRL_CT_usc_asn,CASE_ALLELE_CT_usc_asn,CTRL_ALLELE_CT_usc_asn)
  
  all_mighty = anna %>%
    full_join(jon,by="id")%>%
    full_join(usc_eur,by="id")%>%
    full_join(usc_lat,by="id")%>%
    full_join(usc_blk,by="id")%>%
    full_join(usc_asn,by="id")
  
  all_mighty[is.na(all_mighty)] = 0
  
  rm(anna,jon,usc_eur,usc_lat,usc_blk,usc_asn) 
  
  all_mighty_af = all_mighty %>%
    mutate(ct_cases = CASE_ALLELE_CT_anna+CASE_ALLELE_CT_jon+CASE_ALLELE_CT_usc_eur+
             CASE_ALLELE_CT_usc_lat+CASE_ALLELE_CT_usc_blk+CASE_ALLELE_CT_usc_asn,
           ct_cases_eur = CASE_ALLELE_CT_anna+CASE_ALLELE_CT_jon+CASE_ALLELE_CT_usc_eur,
           ct_cases_non_eur = CASE_ALLELE_CT_usc_lat+CASE_ALLELE_CT_usc_blk+CASE_ALLELE_CT_usc_asn,
           ct_controls = CTRL_ALLELE_CT_anna+CTRL_ALLELE_CT_jon+CTRL_ALLELE_CT_usc_eur+
             CTRL_ALLELE_CT_usc_lat+CTRL_ALLELE_CT_usc_blk+CTRL_ALLELE_CT_usc_asn,
           ct_controls_eur = CTRL_ALLELE_CT_anna+CTRL_ALLELE_CT_jon+CTRL_ALLELE_CT_usc_eur,
           ct_controls_non_eur = CTRL_ALLELE_CT_usc_lat+CTRL_ALLELE_CT_usc_blk+CTRL_ALLELE_CT_usc_asn,
           a1_cases = A1_CASE_CT_anna+A1_CASE_CT_jon+A1_CASE_CT_usc_eur+A1_CASE_CT_usc_lat+A1_CASE_CT_usc_blk+A1_CASE_CT_usc_asn,
           a1_cases_eur = A1_CASE_CT_anna+A1_CASE_CT_jon+A1_CASE_CT_usc_eur,
           a1_cases_non_eur=A1_CASE_CT_usc_lat+A1_CASE_CT_usc_blk+A1_CASE_CT_usc_asn,
           a1_controls = A1_CTRL_CT_anna+A1_CTRL_CT_jon+A1_CTRL_CT_usc_eur+A1_CTRL_CT_usc_lat+A1_CTRL_CT_usc_blk+A1_CTRL_CT_usc_asn,
           a1_controls_eur = A1_CTRL_CT_anna+A1_CTRL_CT_jon+A1_CTRL_CT_usc_eur,
           a1_controls_non_eur=A1_CTRL_CT_usc_lat+A1_CTRL_CT_usc_blk+A1_CTRL_CT_usc_asn) %>%
    mutate(all_cases_maf = a1_cases/ct_cases,
           all_controls_maf = a1_controls/ct_controls,
           eur_cases_maf = a1_cases_eur/ct_cases_eur,
           eur_controls_maf = a1_controls_eur/ct_controls_eur,
           non_eur_cases_maf = a1_cases_non_eur/ct_cases_non_eur,
           non_eur_controls_maf = a1_controls_non_eur/ct_controls_non_eur) %>%
    dplyr::select(id,all_cases_maf,all_controls_maf,eur_cases_maf,eur_controls_maf,non_eur_cases_maf,non_eur_controls_maf)
  
  rm(all_mighty) 
  
  sigs = read.csv2(file_sigs) %>%
    as.data.frame() %>%
    dplyr::select(chr,pos,MarkerName,rsid,ref,alt,geneAnnotation,location,X95CI,P.value,Direction) %>%
    inner_join(all_mighty_af,by=c("MarkerName"="id"))
  
  # sigs %>% filter(rsid=="rs573687")
  
  filename = paste0("maf.anno.0816.",subtype,".csv")
  write.csv2(sigs,filename)
  
  filename2 = paste0("maf.anno.all.snps.",subtype,".csv")
  write.csv2(all_mighty_af,filename2)

}


a = fread(paste0("maf.anno.0816.",subtype,".csv"))

```

# Locus Zoom plots 

// usc results 

```{bash}

cd /project/wiemels_494/sebastian/glioma_gwas/usc_cohort

# 9:22029547:T:C for both european, latino and all subjecs from the USC cohort
#grep '22029547' temps/all.Subs.eur.hg37.txt chr9:22029548:T:C
# its hg38 cordinate is chr9:22029548:T:C


idSNP='chr9:22029548:T:C'
subtype='ast'
chr=9

awk '($7 >0.6) {print $1}' <(zcat topmed_imputed/chr"${chr}".info.gz) | sed '1d' >\
topmed_imputed/include.chr"${chr}".txt

for race in "eur" "lat" "blk" "asn"; do

  plink2 --vcf topmed_imputed/chr"${chr}".dose.vcf.gz \
    dosage=HDS \
    --max-alleles 2 \
    --min-alleles 2 \
    --extract topmed_imputed/include.chr"${chr}".txt\
    --maf 0.01 \
    --keep ../$subtype.Subs.$race.txt\
    --indiv-sort f ../$subtype.Subs.$race.txt\
    --make-bed \
    --out temps/$subtype.Subs.$race.chr$chr\
    --threads 200

  plink --bfile temps/$subtype.Subs.$race.chr$chr\
    --allow-no-sex\
    --ld-window-kb 500\
    --ld-snps $idSNP\
    --r2 dprime\
    --ld-window-r2 0\
    --ld-window 99999\
    --out temps/$race.$subtype.chr$chr
    
done

# extract results
temps/$subtype.Subs.$race.chr$chr.phenotype.glm.logistic.hybrid


```

```{python3}

import os
os.chdir('/project/wiemels_494/sebastian/glioma_gwas/usc_cohort/temps')

subtype='ast'
chr='chr9'

races = ['eur','lat','blk','asn']

for race in races:
  
  #race='eur'
  
  ldname = '.'.join([race,subtype,chr,'ld'])
  lzname = '_'.join([ldname,'lz'])   
  f=open(ldname)
  #f=open('eur.ast.chr9.ld')
  f.readline()
  o=open(lzname,'w')
  o.write('snp1 snp2 dprime rsquare\n')
  for lines in f:
   lines=lines.split()
   o.write(lines[2].split(':')[0]+':'+lines[2].split(':')[1]+' '+lines[5].split(':')[0]+':'+lines[5].split(':')[1]+' '+lines[7]+' '+lines[6]+'\n')
  
  o.close()

```

```{r}

setwd("/project/wiemels_494/sebastian/glioma_gwas/usc_cohort/")
library(tidyverse)
library(data.table)

subtype='ast'
chr=9

for(race in c("eur","lat","blk","asn")){
  
  #race='eur'
  
  fileName = paste0("temps/",subtype,".Subs.",race,".chr",chr,".phenotype.glm.logistic.hybrid")
  
  placeholder = fread(fileName) %>%
    mutate(MarkerName = paste0("chr",`#CHROM`,":",POS),
           `P-value`=P) %>%
    select(MarkerName,`P-value`) %>%
    na.omit()
  
  write.table(placeholder,paste0(fileName,"_lz"), sep='\t', row.names = FALSE, quote=FALSE)
  
}

```

```{bash}

cd /project/wiemels_494/sebastian/glioma_gwas/usc_cohort

idSNP='chr9:22029548'
subtype='ast'
chr=9

module load gcc/9.2.0 python/2.7.16

for race in "eur" "lat" "blk" "asn"; do

  #race='eur'

  locuszoom --build hg38\
    --refsnp $idSNP\
    --flank 500kb\
    --ld temps/$race.$subtype.chr$chr.ld_lz\
    --metal temps/$subtype.Subs.$race.chr$chr.phenotype.glm.logistic.hybrid_lz\
    --prefix $subtype.chr$chr.$race\
    --plotonly
    
done

```

// meta-analysis results of all cohorts, using 1000D LDs

```{bash}

cd /project/wiemels_494/sebastian/glioma_gwas/meta/METAL

idSNP='9:22011642:G:A '
subtype='ast'
chr=9
pos=22011642

awk -v chr=9 -v pos=22011642 '$1==chr && $4>pos-500000 && $4<pos+500000 {print}' \
  /project/wiemels_494/sebastian/temp_scratch_backup/gwas_1000G/GRCh19/ALL.2of4intersection.20100804.genotypes_no_missing_IDs.bim >\
  1000g.snps.win.$chr.$pos.500kb.txt

```

```{r}

setwd("/project/wiemels_494/sebastian/glioma_gwas/meta/METAL")
library(tidyverse)
library(data.table)

chr=9
pos=22011642

meta_file = "meta.four.races.ast.glioma.TBL"
kg1_file = paste0("1000g.snps.win.",chr,".",pos,".500kb.txt")
  
meta = fread(meta_file) %>%
    mutate(ID= MarkerName) %>%
    separate(ID, c("CHROM","POS","REF","ALT"))%>%
    dplyr::mutate(MarkerName=paste0("chr",CHROM,":",POS), P=`P-value`)%>%
    dplyr::mutate(CHROM = as.numeric(gsub("chr","",as.character(CHROM))),POS = as.numeric(as.character(POS)))

kg1 = fread(kg1_file)

Snps_selected = meta %>%
  dplyr::select(-Allele1,-Allele2) %>%
  inner_join(kg1,by=c("CHROM"="V1","POS"="V4")) %>%
  distinct(POS,.keep_all=TRUE) %>%
  arrange(CHROM,POS)  %>%
  na.omit()

Snps_selected_ld = Snps_selected%>%
  dplyr::select(MarkerName,`P-value`) %>%
  na.omit()

ref_for_1kg = Snps_selected %>%
  dplyr::select(V2,REF)
  
write.table(Snps_selected_ld,paste0("temps/",meta_file,"_lz"), sep='\t', row.names = FALSE, quote=FALSE)
write.table(ref_for_1kg,"1KG.REF.ALLELE.txt", sep='\t', row.names = FALSE, col.names=FALSE, quote=FALSE)

```

```{bash}

cd /project/wiemels_494/sebastian/glioma_gwas/meta/METAL

plink2 --bfile /project/wiemels_494/sebastian/temp_scratch_backup/gwas_1000G/GRCh19/ALL.2of4intersection.20100804.genotypes_no_missing_IDs\
  --extract 1KG.REF.ALLELE.txt\
  --ref-allele 1KG.REF.ALLELE.txt\
  --make-bed\
  --out temps/1kg.allele.reg

cp temps/1kg.allele.reg.bim temps/1kg.allele.reg.bim.backup

awk -v OFS='\t' '{print $1,"chr"$1":"$4,$3,$4,$5,$6}' temps/1kg.allele.reg.bim.backup > temps/1kg.allele.reg.bim

idSNP='chr9:22011642'

plink --bfile temps/1kg.allele.reg\
  --allow-no-sex\
  --ld-window-kb 500\
  --ld-snps $idSNP\
  --r2 dprime\
  --ld-window-r2 0\
  --ld-window 99999\
  --out temps/1kg.allele.reg
  
```

```{python3}

import os
os.chdir('/project/wiemels_494/sebastian/glioma_gwas/meta/METAL/temps')

subtype='ast'
chr='chr9'

ldname = '1kg.allele.reg.ld'
lzname = '1kg.allele.reg.ld_lz'

f=open(ldname)
f.readline()
o=open(lzname,'w')
o.write('snp1 snp2 dprime rsquare\n')
for lines in f:
 lines=lines.split()
 o.write(lines[2].split(':')[0]+':'+lines[2].split(':')[1]+' '+lines[5].split(':')[0]+':'+lines[5].split(':')[1]+' '+lines[7]+' '+lines[6]+'\n')

o.close()

```

```{bash}

cd /project/wiemels_494/sebastian/glioma_gwas/meta/METAL

idSNP='chr9:22011642'
subtype='ast'
chr=9

module load gcc/8.3.0  openblas/0.3.8 python/2.7.16

locuszoom --build hg19\
  --refsnp $idSNP\
  --flank 500kb\
  --ld temps/1kg.allele.reg.ld_lz\
  --metal temps/meta.four.races.ast.glioma.TBL_lz\
  --prefix $subtype.chr$chr\
  --plotonly

```
