---
output: html_document
editor_options: 
  chunk_output_type: console
---

Project description: USC part of childhood leukemia GWAS meta-analysis, led by Childhood Cancer & Leukemia International Consortium (CLIC)

```{bash manuscript preparation}

# 1) locuszoom plot for three loci   
# 2) PRS using known SNPs vs. known+novel SNPs   
# 3) PC  for each population.

######################################################################################################
# 1) use locuszoom, generate a plot with 1Mb region around 6:135411228, 10:65020890, 10:70329064
# If you aren't familiar with it, locuszoom needs a LD file and association result file. 
# LD file needs to have snp pairs in column 1 and 2 in chr:pos format, d-prime in column3, r2 in column 4. 
# you can generate LD files from ccls files with plink. Because I had plink files for each chromosome, I generated files separately for each chr. Command I used is...

cd ~/GWAS/libby/soyoung

for RACE in "EUR" "LAT"; do
  cp /ccls/home/sli/GWAS/libby/ccls/plink/ccls."${RACE}".??? .
done

for RACE in "EUR" "LAT"; do
  cp /ccls/home/sli/GWAS/libby/ccls/ccls."${RACE}".glm.PHENO1.glm.logistic .
done


for RACE in "EUR" "LAT"; do
  plink --bfile ccls."${RACE}" \
    --allow-no-sex \
    --ld-window-kb 500 \
    --ld-snps 6:135411228:T:C 10:65020890:A:G 10:70329064:G:A \
    --r2 dprime \
    --ld-window-r2 0 \
    --ld-window 99999 \
    --out ccls."${RACE}" 
done


# then convert plink output to locuszoom input format
#PYTHON3 SCRIPT
f=open('ccls.EUR.ld')
f.readline()
o=open('ccls.EUR.ld_lz','w')
o.write('snp1 snp2 dprime rsquare\n')
for lines in f:
  lines=lines.split()
  o.write(lines[2].split(':')[0]+':'+lines[2].split(':')[1]+' '+lines[5].split(':')[0]+':'+lines[5].split(':')[1]+' '+lines[7]+' '+lines[6]+'\n')
o.close()

f=open('ccls.LAT.ld')
f.readline()
o=open('ccls.LAT.ld_lz','w')
o.write('snp1 snp2 dprime rsquare\n')
for lines in f:
  lines=lines.split()
  o.write(lines[2].split(':')[0]+':'+lines[2].split(':')[1]+' '+lines[5].split(':')[0]+':'+lines[5].split(':')[1]+' '+lines[7]+' '+lines[6]+'\n')
o.close()


for RACE in "EUR" "LAT"; do
  awk 'NR==1; $1==6 || $1==10 {print}' ccls."${RACE}".glm.PHENO1.glm.logistic > ccls."${RACE}".chr6.chr10.logistic
done

#Association test file needs to have MarkerName (in "chrX:pos" format like chr1:1111) in col1 and P-value in col2. 
#PYTHON3 SCRIPT for glm
f=open('ccls.EUR.chr6.chr10.logistic')
f.readline()
o=open('ccls.EUR.logistic_lz','w')
o.write('MarkerName\tP-value\n')
for lines in f:
  lines=lines.split()
  o.write('chr'+lines[0]+':'+lines[1]+'\t'+lines[-1]+'\n')
o.close()

f=open('ccls.LAT.chr6.chr10.logistic')
f.readline()
o=open('ccls.LAT.logistic_lz','w')
o.write('MarkerName\tP-value\n')
for lines in f:
  lines=lines.split()
  o.write('chr'+lines[0]+':'+lines[1]+'\t'+lines[-1]+'\n')
o.close()

# locuszoom plot command
locuszoom --build hg19 \
  --refsnp 6:135411228 \
  --flank 500kb \
  --ld ccls.EUR.ld_lz \
  --metal ccls.EUR.logistic_lz \
  --prefix MYB.eur \
  --plotonly

locuszoom --build hg19 \
  --refsnp 10:65020890 \
  --flank 500kb \
  --ld ccls.EUR.ld_lz \
  --metal ccls.EUR.logistic_lz \
  --prefix JMJD1C.eur \
  --plotonly
  
locuszoom --build hg19 \
  --refsnp 10:70329064 \
  --flank 500kb \
  --ld ccls.EUR.ld_lz \
  --metal ccls.EUR.logistic_lz \
  --prefix TET1.eur \
  --plotonly

locuszoom --build hg19 \
  --refsnp 6:135411228 \
  --flank 500kb \
  --ld ccls.LAT.ld_lz \
  --metal ccls.LAT.logistic_lz \
  --prefix MYB.lat \
  --plotonly

locuszoom --build hg19 \
  --refsnp 10:65020890 \
  --flank 500kb \
  --ld ccls.LAT.ld_lz \
  --metal ccls.LAT.logistic_lz \
  --prefix JMJD1C.lat \
  --plotonly
  
locuszoom --build hg19 \
  --refsnp 10:70329064 \
  --flank 500kb \
  --ld ccls.LAT.ld_lz \
  --metal ccls.LAT.logistic_lz \
  --prefix TET1.lat \
  --plotonly
  
############################################################################################################
##1.5 Locuszoom for meta-analysis
cd ~/GWAS/libby/soyoung

## Combine EUR and LAT
nano soyong.ccls.rep.list
ccls.LAT
ccls.EUR

plink --merge-list soyong.ccls.rep.list \
  --make-bed \
  --out ccls.meta

plink --bfile ccls.meta \
  --allow-no-sex \
  --ld-window-kb 500 \
  --ld-snps 6:135411228:T:C 10:65020890:A:G 10:70329064:G:A \
  --r2 dprime \
  --ld-window-r2 0 \
  --ld-window 99999 \
  --out ccls.meta

# then convert plink output to locuszoom input format
#PYTHON3 SCRIPT
f=open('ccls.meta.ld')
f.readline()
o=open('ccls.meta.ld_lz','w')
o.write('snp1 snp2 dprime rsquare\n')
for lines in f:
  lines=lines.split()
  o.write(lines[2].split(':')[0]+':'+lines[2].split(':')[1]+' '+lines[5].split(':')[0]+':'+lines[5].split(':')[1]+' '+lines[7]+' '+lines[6]+'\n')
o.close()


#Association test file needs to have MarkerName (in "chrX:pos" format like chr1:1111) in col1 and P-value in col2. 
cp /ccls/home/sli/GWAS/libby/ccls/ccls.meta.glm.logistic .

# R code
setwd("~/GWAS/libby/soyoung")
library(tidyverse)
library(data.table)
meta <- fread("ccls.meta.glm.logistic") %>%
  separate(MarkerName, c("CHROM","POS","REF","ALT")) 
meta1 <-  meta%>%
  filter(CHROM %in% c(6,10)) %>%
  mutate(CHROM=as.numeric(CHROM), POS=as.numeric(POS)) %>%
  arrange(CHROM,POS) %>%
  mutate(MarkerName = paste0("chr",CHROM,":",POS))%>%
  select(MarkerName, "P-value")
write.table(meta1, "ccls.meta.logistic_lz", sep='\t',row.names = FALSE, quote=FALSE)


# locuszoom plot command
locuszoom --build hg19 \
  --refsnp 6:135411228 \
  --flank 500kb \
  --ld ccls.meta.ld_lz \
  --metal ccls.meta.logistic_lz \
  --prefix MYB.meta \
  --plotonly

locuszoom --build hg19 \
  --refsnp 10:65020890 \
  --flank 500kb \
  --ld ccls.meta.ld_lz \
  --metal ccls.meta.logistic_lz \
  --prefix JMJD1C.meta \
  --plotonly
  
locuszoom --build hg19 \
  --refsnp 10:70329064 \
  --flank 500kb \
  --ld ccls.meta.ld_lz \
  --metal ccls.meta.logistic_lz \
  --prefix TET1.meta \
  --plotonly



############################################################################################################
#2) For PRS, I had vcf for each chromosome, so I used plink to generate a score output in each chr, 
# and combine them with python3 script to make a final score file. If you have this too, you can use this. I attached two files you would need here.
# bash script for plink PRS. You can use much smaller memory in bash and plink commands since you have smaller sample size. 

cd ~/GWAS/libby/soyoung

for RACE in "EUR" "LAT"; do
  plink --bfile ccls."${RACE}" \
    --allow-no-sex \
    --recode vcf \
    --keep-allele-order \
    --out ccls."${RACE}"
  bcftools sort ccls."${RACE}".vcf -Oz -o ccls."${RACE}".vcf.gz 
  tabix ccls."${RACE}".vcf.gz 
done

#split by chromomse
for RACE in "EUR" "LAT"; do
  for chr in 5 6 7 8 9 10 12 14 17 21; do
    bcftools view --regions $chr ccls."${RACE}".vcf.gz -o "${RACE}".chr$chr.vcf.gz -Oz &
  done
done


## prstemp.sh

#!/bin/bash
#SBATCH --time=50:00:00
#SBATCH --job-name=PRS
#SBATCH --array=5,6,7,8,9,10,12,14,17,21
#SBATCH --output=slurm.%x.%j.out
#SBATCH --error=slurm.%x.%j.err
#SBATCH --partition=LargeHost
#SBATCH --nodelist=ecryptp16-cloud1

cd ~/GWAS/libby/soyoung

for RACE in "EUR" "LAT"; do
  echo "${RACE}".chr${SLURM_ARRAY_TASK_ID}.vcf.gz
  
  plink2 --vcf "${RACE}".chr${SLURM_ARRAY_TASK_ID}.vcf.gz dosage=DS \
    --memory 35000 \
    --score METAL_known.TBL.txt 1 2 6 list-variants 'no-mean-imputation' \
    --out known_2nd/2nd_"${RACE}"_chr${SLURM_ARRAY_TASK_ID}
    
  plink2 --vcf "${RACE}".chr${SLURM_ARRAY_TASK_ID}.vcf.gz dosage=DS \
    --memory 35000 \
    --score METAL_wnovel.TBL.txt 1 2 6 list-variants 'no-mean-imputation' \
    --out wnovel_2nd/2nd_"${RACE}"_chr${SLURM_ARRAY_TASK_ID}
    
done

# Then, I go into each directory and run this python script to generate a big score files for each population in known & wnovel directory. 

cd ~/GWAS/libby/soyoung/wnovel_2nd
cd ~/GWAS/libby/soyoung/known_2nd

#PYTHON3 script
#You can change this pop list to whatever prefix you are using. 

pop=['EUR','LAT']

chr=[5,6,7,8,9,10,12,14,17,21]
for p in pop:
  nm={}   #NMISS_ALLELE_CT
  sumsc={}        #score as SUM of dosage*score
  avg={}  
  f=open('2nd_'+p+'_chr10.sscore')
  f.readline()
  for lines in f:
    lines=lines.split()
    nm[lines[0]]=[]
    sumsc[lines[0]]=[]
    avg[lines[0]]=[]
  for i in chr:
    i=str(i)
    try:
      f=open('2nd_'+p+'_chr'+i+'.sscore')
      f.readline()
      for lines in f:
        lines=lines.split()
        id=lines[0]
        nm[id].append(float(lines[1]))
        sumsc[id].append(float(lines[1])*float(lines[3]))
        avg[id].append(float(lines[3]))
    except FileNotFoundError:
      print('File Not Found: '+p+'_chr'+i+'.sscore')
  o=open('2nd_'+p+'.sscore','w')
  o.write('id\tALLELE_CT\tSCORE_AVG\tSCORE_SUM\n')
  for i in nm.keys():
    o.write(i+'\t'+str(sum(nm[i]))+'\t'+str(sum(avg[i])/22)+'\t'+str(sum(sumsc[i]))+'\n')
  o.close()

#You can send the outputs from this script. (which should be 4)

#known.EUR.sscore
#known.LAT.sscore
#novel.EUR.sscore
#novel.LAT.sscore

############################################################################################################
# 3) You can just send me the PC file you generated for each pop.

cd ~/GWAS/libby/soyoung

for RACE in "EUR" "LAT"; do
  cp /ccls/home/sli/GWAS/libby/ccls/plink/ccls."${RACE}".pruned.pca.eigenvec .
done

#ccls.EUR.pruned.pca.eigenvec
#ccls.LAT.pruned.pca.eigenvec


############################################################################################################
# Preparing data to send to SY
cd ~/GWAS/libby/soyoung

awk '{print $1}' METAL_wnovel.TBL > extract.SNP

# adding 
# 17:38066240:T:C
# 9:21993964:T:C
# 14:23592617:A:C
# 7:50459043:G:A
# 7:50355207:G:C
# to extract.SNP

cd ~/GWAS/libby/soyoung

###############################################################

for RACE in "EUR" "LAT"; do
  for chr in 5 6 7 8 9 10 12 14 17 21; do
    vcftools --gzvcf ~/GWAS/libby/ccls/hrc_imputed/chr$chr.dose.vcf.gz \
      --positions extract.SNP.txt \
      --recode --stdout | gzip -c > sy2.$RACE.chr$chr.vcf.gz
  done
done  

###############################################################

for RACE in "EUR" "LAT"; do
  cp /ccls/home/sli/GWAS/libby/ccls/plink/ccls."${RACE}".??? .
done

for RACE in "EUR" "LAT"; do
  plink --bfile ccls."${RACE}" \
    --extract extract.SNP \
    --make-bed \
    --out ccls.sy."${RACE}"
done

```

# reviewer's comments (from pipleline_prepare.sh and probably pipleline.Rmd)

```{bash}

cd ~/GWAS/libby/soyoung

cp /ccls/home/sli/GWAS/libby/ccls/plink/ccls.EUR.??? .
cp /ccls/home/sli/GWAS/libby/ccls/plink/ccls.LAT.??? .

```

```{r  prepare a variable file with case/control status, gender, subtype and race}
setwd("~/GWAS/libby/soyoung")
library(tidyverse)
library(data.table)

clinical = fread("/ccls/home/sli/GWAS/libby/clinical/sebastian_ccls.csv") %>%
  select(FID,IID,subtype)

eurSubs <- fread("ccls.EUR.fam") %>%
  left_join(clinical,by=c("V1"="FID","V2"="IID"))

table(eurSubs$V5)
table(eurSubs$V6)

eurSubs_b <- eurSubs %>%
  filter(V6==1 | subtype=="Bcell") %>%
  select(V1,V2)

write.table(eurSubs_b,"bcell.all.EUR.txt",
            sep="\t",
            row.names = FALSE,
            col.names = FALSE,
            quote = FALSE)

eurSubs_t <- eurSubs %>%
  filter(V6==1 | subtype=="Tcell") %>%
  select(V1,V2)

write.table(eurSubs_t,"tcell.all.EUR.txt",
            sep="\t",
            row.names = FALSE,
            col.names = FALSE,
            quote = FALSE)

latSubs <- fread("ccls.LAT.fam")%>%
  left_join(clinical,by=c("V1"="FID","V2"="IID"))

table(latSubs$V5)
table(latSubs$V6)

latSubs_b <- latSubs %>%
  filter(V6==1 | subtype=="Bcell") %>%
  select(V1,V2)

write.table(latSubs_b,"bcell.all.LAT.txt",
            sep="\t",
            row.names = FALSE,
            col.names = FALSE,
            quote = FALSE)

latSubs_t <- latSubs %>%
  filter(V6==1 | subtype=="Tcell") %>%
  select(V1,V2)

write.table(latSubs_t,"tcell.all.LAT.txt",
            sep="\t",
            row.names = FALSE,
            col.names = FALSE,
            quote = FALSE)

# prepare for covariate files
eurCov_b <- fread("/ccls/home/sli/GWAS/libby/ccls/plink/ccls.EUR.pruned.pca.eigenvec") %>%
  right_join(eurSubs_b,by=c("#FID"="V1","IID"="V2"))

write.table(eurCov_b,"ccls.bcell.EUR.pruned.pca.eigenvec",
            sep="\t",
            row.names = FALSE,
            quote = FALSE)

eurCov_t <- fread("/ccls/home/sli/GWAS/libby/ccls/plink/ccls.EUR.pruned.pca.eigenvec") %>%
  right_join(eurSubs_t,by=c("#FID"="V1","IID"="V2"))

write.table(eurCov_t,"ccls.tcell.EUR.pruned.pca.eigenvec",
            sep="\t",
            row.names = FALSE,
            quote = FALSE)

latCov_b <- fread("/ccls/home/sli/GWAS/libby/ccls/plink/ccls.LAT.pruned.pca.eigenvec") %>%
  right_join(latSubs_b,by=c("#FID"="V1","IID"="V2"))

write.table(latCov_b,"ccls.bcell.LAT.pruned.pca.eigenvec",
            sep="\t",
            row.names = FALSE,
            quote = FALSE)

latCov_t <- fread("/ccls/home/sli/GWAS/libby/ccls/plink/ccls.LAT.pruned.pca.eigenvec") %>%
  right_join(latSubs_t,by=c("#FID"="V1","IID"="V2"))

write.table(latCov_t,"ccls.tcell.LAT.pruned.pca.eigenvec",
            sep="\t",
            row.names = FALSE,
            quote = FALSE)

```


```{bash subtype stratified analysis}

cd ~/GWAS/libby/soyoung

# select the corresponding subjects

for subtype in "tcell" "bcell"; do
  for race in "EUR" "LAT"; do
    plink2 --bfile ccls.$race\
      --keep $subtype.all.$race.txt\
      --indiv-sort f $subtype.all.$race.txt\
      --chr 6,10\
      --make-bed\
      --out ccls.chr6.chr10.$race.$subtype

    plink2 --bfile ccls.chr6.chr10.$race.$subtype\
      --covar ccls.$subtype.$race.pruned.pca.eigenvec\
      --threads 200 \
      --glm hide-covar cols=chrom,pos,ref,alt1,nobs,a1countcc,a1freq,a1freqcc,beta,se,p\
      --out ccls.chr6.chr10.$subtype.$race
  done
done

```

SNPTEST

```{bash snptest they say...}

cd ~/GWAS/libby/soyoung
# mkdir snptest


awk '{print $1"_"$2}' bcell.all.EUR.txt > bcell.EUR.snpTest.txt
awk '{print $1"_"$2}' bcell.all.LAT.txt > bcell.LAT.snpTest.txt
awk '{print $1"_"$2}' tcell.all.EUR.txt > tcell.EUR.snpTest.txt
awk '{print $1"_"$2}' tcell.all.LAT.txt > tcell.LAT.snpTest.txt


cat ccls.EUR.fam ccls.LAT.fam > ccls.snptest.caco.f

for chr in {1..22};do
  awk '($7 < 0.6) {print $1}' <(zcat /ccls/home/sli/GWAS/libby/ccls/hrc_imputed/chr"${chr}".info.gz) |\
  sed '1d' > \
  /ccls/home/sli/GWAS/libby/ccls/hrc_imputed/exclude.06.chr"${chr}".txt;
done

for race in "EUR" "LAT"; do
  for sub in "tcell" "bcell"; do
  
    for chr in {1..22};do
      vcftools --gzvcf /ccls/home/sli/GWAS/libby/ccls/hrc_imputed/chr"${chr}".dose.vcf.gz\
        --maf 0.01\
        --keep $sub.$race.snpTest.txt\
        --exclude /ccls/home/sli/GWAS/libby/ccls/hrc_imputed/exclude.06.chr"${chr}".txt\
        --recode --stdout \
        | gzip -c \
        > snptest/chr"${chr}".$sub.$race.snptest.vcf.gz
    done
    
    bcftools query -l snptest/chr22.$sub.$race.snptest.vcf.gz > snptest/$sub.$race.samples.ids
  done
done

```

```{r SNPTEST MAF=0.01 phenotype file}

setwd("~/GWAS/libby/soyoung/snptest")
library(tidyverse)
library(data.table)

pheno <- fread("~/GWAS/libby/soyoung/ccls.snptest.caco.f") %>%
  dplyr::mutate(subjectID = paste0(V1,"_",V2),
                caco = ifelse(V6==1,0,1)) %>%
  dplyr::select(subjectID,caco) 

for(race in c("EUR","LAT")){
  for(sub in c("tcell","bcell")){
    
    def.list = fread(paste0(sub,".",race,".samples.ids"),header=FALSE) %>%
      rename("subjectID"="V1")
  
    PCs <- fread(paste0("~/GWAS/libby/soyoung/ccls.",sub,".",race,".pruned.pca.eigenvec")) %>%
      dplyr::mutate(subjectID = paste0(`#FID`,"_",IID)) %>%
      select(-'#FID',-IID) 
    
    samples_cov <- def.list %>%
      left_join(pheno,by="subjectID") %>%
      left_join(PCs,by="subjectID") 
    
    line2 <- c(0,"B",c(rep("C",10))) %>% t() %>% as.data.frame()
    colnames(line2) <- colnames(samples_cov)
    samples_cov_1 <- rbind(line2, samples_cov) %>% as.data.frame()
    
    write.table(samples_cov_1, paste0(sub,".",race,".vars.txt") , row.names = FALSE, quote=FALSE)

  }
}


```

```{bash SNPTEST MAF=0.01 run}

cd ~/GWAS/libby/soyoung/snptest

for race in "EUR" "LAT"; do
  for sub in "tcell" "bcell"; do
    for chr in {1..22};do
      snptest -data chr"${chr}".$sub.$race.snptest.vcf.gz\
        $sub.$race.vars.txt\
        -genotype_field GP\
        -method expected\
        -frequentist 1\
        -pheno caco\
        -cov_names PC1 PC2 PC3 PC4 PC5 PC6 PC7 PC8 PC9 PC10 \
        -hwe\
        -o $sub.$race.chr"${chr}".out
    done
    
    head -1 $sub.$race.chr1.out > overall.$sub.$race.snptest
    tail -n +2 -q $sub.$race.chr*.out >> overall.$sub.$race.snptest

  done
done



```



