---
output: html_document
editor_options: 
  chunk_output_type: console
---

Project description: Rhabdomyosarcoma GWAS in Latino subjects

# data preprocessing

```{bash preparing 3 datasets}

cd /ccls/home/sli/GWAS/libby/rmsLat
#mkdir tmp

# rms cases
## generated by "rms.Rmd"
cp /ccls/home/sli/GWAS/libby/rms/rms.gwas.??? .
## 679880 variants and 633 people pass filters and QC.
awk '{print $1":"$4}' rms.gwas.bim > rms.gwas.snps

# ccls controls (hg19)
cp /ccls/home/sli/GWAS/libby/ccls/ccls.gwas.??? .
plink --bfile ccls.gwas\
  --remove /ccls/home/sli/GWAS/libby/v3/ccls.cases.list\
  --make-bed\
  --out ccls.controls
## 649876 variants and 912 people pass filters and QC.
awk '{print $1":"$4}' ccls.controls.bim > ccls.controls.snps

# dbgap controls, batch 79737 (hg19)
## generated by "rmsdbGAP.Rmd"
cp /ccls/home/sli/GWAS/libby/dbGAP/79737/PhenoGenotypeFiles/RootStudyConsentSet_phs001367.AddHealth.v1.p1.c1.GRU-IRB-PUB-GSO/GenotypeFiles/matrix/ahgwas.??? .
cp /ccls/home/sli/GWAS/libby/dbGAP/79737/PhenoGenotypeFiles/RootStudyConsentSet_phs001367.AddHealth.v1.p1.c1.GRU-IRB-PUB-GSO/PhenotypeFiles/phs001367.v1.pht008248.v1.p1.c1.Add_Health_Subject_Phenotypes.GRU-IRB-PUB-GSO.txt .

plink --bfile ahgwas\
  --chr 1-22\
  --make-bed\
  --out ahgwas.0
  
# get rid of variants sharing the same position but don't start with "rs"
Rscript /ccls/home/sli/GWAS/libby/codes/procDup.R ahgwas.0

plink --bfile ahgwas.0\
  --exclude ahgwas.0.drop\
  --make-bed\
  --out ahgwas.1

mv ahgwas.1.bim ahgwas.1.bim.bkp
awk '{print $1, $1":"$4,$3,$4,$5,$6}' ahgwas.1.bim.bkp > ahgwas.1.bim

# dbgap controls, batch 82131
cp /ccls/proj_circle2_p3/dbgap_phs000810/82131/PhenoGenotypeFiles/ChildStudyConsentSet_phs000880.HCHS_SOL.v1.p1.c1.HMB-NPU/GenotypeFiles/phg000663.v1.HCHS_SOL_HispCommunityHS.genotype-calls-matrixfmt.c1/subject_level_filtered_PLINK_set/SOL_TOP_subject_level_filtered_c1.??? .

cp /ccls/proj_circle2_p3/dbgap_phs000810/82131/PhenoGenotypeFiles/RootStudyConsentSet_phs000810.HCHS_SOL.v1.p1.c1.HMB-NPU/PhenotypeFiles/phs000810.v1.pht004715.v1.p1.c1.HCHS_SOL_Cohort_Subject_Phenotypes.HMB-NPU.txt .

plink --bfile SOL_TOP_subject_level_filtered_c1\
  --chr 1-22\
  --make-bed\
  --out sol.c1.0
  
# get rid of variants sharing the same position but don't start with "rs"
Rscript /ccls/home/sli/GWAS/libby/codes/procDup.R sol.c1.0

plink --bfile sol.c1.0\
  --exclude sol.c1.0.drop\
  --make-bed\
  --out sol.c1

# 2427209 variants and 3441 people pass filters and QC.

# dbgap controls, batch 82129 (hg19)
cp /ccls/proj_circle2_p3/dbgap_phs000810/82129/PhenoGenotypeFiles/ChildStudyConsentSet_phs000880.HCHS_SOL.v1.p1.c2.HMB/GenotypeFiles/phg000663.v1.HCHS_SOL_HispCommunityHS.genotype-calls-matrixfmt.c2/subject_level_filtered_PLINK_set/SOL_TOP_subject_level_filtered_c2.??? .
8993

cp /ccls/proj_circle2_p3/dbgap_phs000810/82129/PhenoGenotypeFiles/RootStudyConsentSet_phs000810.HCHS_SOL.v1.p1.c2.HMB/PhenotypeFiles/phs000810.v1.pht004715.v1.p1.c2.HCHS_SOL_Cohort_Subject_Phenotypes.HMB.txt .

plink --bfile SOL_TOP_subject_level_filtered_c2\
  --chr 1-22\
  --exclude sol.c1.0.drop\
  --make-bed\
  --out sol.c2

# 2427209 variants and 8993 people pass filters and QC.

plink --bfile rms.gwas\
  --bmerge ccls.controls\
  --allow-no-sex\
  --merge-equal-pos\
  --out rms.gwas.ccls.co

mv rms.gwas.ccls.co.bim rms.gwas.ccls.co.bim.bkp
awk '{print $1, $1":"$4,$3,$4,$5,$6}' rms.gwas.ccls.co.bim.bkp > rms.gwas.ccls.co.bim

awk '{print $2}' rms.gwas.ccls.co.bim > rms.gwas.ccls.co.snp

plink --bfile sol.c1\
  --bmerge sol.c2\
  --allow-no-sex\
  --merge-equal-pos\
  --out sol

mv sol.bim sol.bim.bkp
awk '{print $1, $1":"$4,$3,$4,$5,$6}' sol.bim.bkp > sol.bim

# merging ccls control, rms case and sol controls

plink --bfile sol\
  --extract rms.gwas.ccls.co.snp\
  --make-bed\
  --out sol.mg

awk '{print $2}' sol.mg.bim > sol.mg.snp
  
plink --bfile rms.gwas.ccls.co\
  --extract sol.mg.snp\
  --make-bed\
  --out rms.ccls.mg

# set reference genome
awk '{print$2,$5}' rms.ccls.mg.bim > rms.ccls.ref.list.txt

plink --bfile sol.mg\
  --reference-allele rms.ccls.ref.list.txt\
  --make-bed\
  --out sol.mg.adj

# Resolve strand issues
awk '{print $2,$5,$6}' rms.ccls.mg.bim > rms.ccls.mg_tmp
awk '{print $2,$5,$6}' sol.mg.adj.bim > sol.mg.adj_tmp
sort rms.ccls.mg_tmp sol.mg.adj_tmp |\
  uniq -u | awk '{print $1}' |\
  sort -u > flip_list.txt

plink --bfile sol.mg.adj\
  --flip flip_list.txt\
  --reference-allele rms.ccls.ref.list.txt\
  --make-bed\
  --out sol.mg.corrected

awk '{print $2,$5,$6}' sol.mg.corrected.bim > sol.mg.corrected_tmp

sort rms.ccls.mg_tmp sol.mg.corrected_tmp |\
  uniq -u  | awk '{print $1}' |\
  sort -u  > SNPs_for_exlusion.txt

plink --bfile sol.mg.corrected\
  --exclude SNPs_for_exlusion.txt\
  --make-bed\
  --out sol.mg.1

plink --bfile sol.mg.1\
  --bmerge rms.ccls.mg\
  --merge-equal-pos\
  --out rms.ccls.sol

awk '{print $2}' rms.ccls.sol.bim > rms.ccls.sol.snp

# merging all other files with ahgwas

plink --bfile ahgwas.1\
  --extract rms.ccls.sol.snp\
  --make-bed\
  --out ahgwas.mg

awk '{print $2}' ahgwas.mg.bim > ahgwas.mg.snp
  
plink --bfile rms.ccls.sol\
  --extract ahgwas.mg.snp\
  --make-bed\
  --out rms.ccls.sol.mg

# set reference genome
awk '{print$2,$5}' rms.ccls.sol.mg.bim > rms.ccls.sol.ref.list.txt

plink --bfile ahgwas.mg\
  --reference-allele rms.ccls.sol.ref.list.txt\
  --make-bed\
  --out ahgwas.mg.adj

# Resolve strand issues
awk '{print $2,$5,$6}' rms.ccls.sol.mg.bim > rms.ccls.sol.mg_tmp
awk '{print $2,$5,$6}' ahgwas.mg.adj.bim > ahgwas.mg.adj_tmp
sort rms.ccls.sol.mg_tmp ahgwas.mg.adj_tmp |\
  uniq -u | awk '{print $1}' |\
  sort -u > flip_list.txt

plink --bfile ahgwas.mg.adj\
  --flip flip_list.txt\
  --reference-allele rms.ccls.sol.ref.list.txt\
  --make-bed\
  --out ahgwas.mg.corrected

awk '{print $2,$5,$6}' ahgwas.mg.corrected.bim > ahgwas.mg.corrected_tmp

sort rms.ccls.sol.mg_tmp ahgwas.mg.corrected_tmp |\
  uniq -u  | awk '{print $1}' |\
  sort -u  > SNPs_for_exlusion.txt

plink --bfile ahgwas.mg.corrected\
  --exclude SNPs_for_exlusion.txt\
  --make-bed\
  --out ahgwas.mg.1

plink --bfile ahgwas.mg.1\
  --bmerge rms.ccls.sol.mg\
  --merge-equal-pos\
  --out rms.ccls.sol.ahgwas

# 13974 of rms.ccls.sol.mg.fam are new, while 5 are present in the base dataset.
# 23948 people, 603096 variants

```

code for remove weird SNPs from dup file
```{r procDup.R}
# remove weird SNPs from dup file
# setwd("/ccls/home/sli/GWAS/libby/rmsLat")
# args=c("ahgwas.0")

args = commandArgs(trailingOnly=TRUE)

library(tidyverse)
library(data.table)
dupF = fread(paste0(args[1],".bim")) %>% 
  as.data.frame() %>%
  dplyr::mutate(posID=paste0(V1,":",V4))%>%
  dplyr::select(posID,V2)

dupPId = names(which(table(dupF$posID) > 1))

dupIdall = dupF %>% 
  dplyr::filter(posID%in%dupPId) %>%
  dplyr::select(V2)%>%
  dplyr::filter(!grepl("^rs",V2))

write.table(dupIdall,paste0(args[1],".drop"),
            quote=FALSE,
            sep="\t",
            row.names = FALSE,
            col.names = FALSE)

```

```{bash imputing ancestry}

cd /ccls/home/sli/GWAS/libby/rmsLat

cp /ccls/home/sli/GWAS/ref/phase3/1kG_MDS_P3_5.??? .
mv 1kG_MDS_P3_5.bim 1kG_MDS_P3_5.bim.bkp
awk '{print $1, $1":"$4,$3,$4,$5,$6}' 1kG_MDS_P3_5.bim.bkp > 1kG_MDS_P3_5.bim
awk '{print $2}' 1kG_MDS_P3_5.bim > 1kG_MDS_P3_5.snp

plink --bfile rms.ccls.sol.ahgwas\
  --extract 1kG_MDS_P3_5.snp\
  --make-bed\
  --out rms.ccls.sol.ahgwas.tmp1

awk '{print $2}' rms.ccls.sol.ahgwas.tmp1.bim > rms.ccls.sol.ahgwas.tmp1.snp
  
plink --bfile 1kG_MDS_P3_5\
  --extract rms.ccls.sol.ahgwas.tmp1.snp\
  --make-bed\
  --out 1kG_MDS_P3_5.tmp1

# set reference genome
awk '{print$2,$5}' 1kG_MDS_P3_5.tmp1.bim > 1kg.ref.list.txt

plink --bfile rms.ccls.sol.ahgwas.tmp1\
  --reference-allele 1kg.ref.list.txt\
  --make-bed\
  --out rms.ccls.sol.ahgwas.tmp2

# Resolve strand issues
awk '{print $2,$5,$6}' rms.ccls.sol.ahgwas.tmp2.bim > rms.ccls.sol.ahgwas.tmp2_tmp
awk '{print $2,$5,$6}' 1kG_MDS_P3_5.tmp1.bim > 1kG_MDS_P3_5.tmp1_tmp
sort rms.ccls.sol.ahgwas.tmp2_tmp 1kG_MDS_P3_5.tmp1_tmp |\
  uniq -u | awk '{print $1}' |\
  sort -u > flip_list.txt

plink --bfile rms.ccls.sol.ahgwas.tmp2\
  --flip flip_list.txt\
  --reference-allele 1kg.ref.list.txt\
  --make-bed\
  --out rms.ccls.sol.ahgwas.tmp3

awk '{print $2,$5,$6}' rms.ccls.sol.ahgwas.tmp3.bim > rms.ccls.sol.ahgwas.tmp3_tmp

sort rms.ccls.sol.ahgwas.tmp3_tmp 1kG_MDS_P3_5.tmp1_tmp |\
  uniq -u  | awk '{print $1}' |\
  sort -u  > SNPs_for_exlusion.txt

plink --bfile rms.ccls.sol.ahgwas.tmp3\
  --exclude SNPs_for_exlusion.txt\
  --make-bed\
  --out rms.ccls.sol.ahgwas.tmp4

plink --bfile rms.ccls.sol.ahgwas.tmp4\
  --bmerge 1kG_MDS_P3_5.tmp1\
  --merge-equal-pos\
  --out 1kg.mds.rms.ccls.sol.ahgwas

plink --bfile 1kg.mds.rms.ccls.sol.ahgwas\
  --exclude ~/GWAS/ref/codes/inversion_hg37.txt\
  --range\
  --indep-pairwise 50 5 0.2\
  --out indepSNP.1000g
  
# Using a set of pruned SNPs
plink --bfile 1kg.mds.rms.ccls.sol.ahgwas\
  --extract indepSNP.1000g.prune.in\
  --genome\
  --out 1kg.mds.rms.ccls.sol.ahgwas_1

plink --bfile 1kg.mds.rms.ccls.sol.ahgwas\
  --read-genome 1kg.mds.rms.ccls.sol.ahgwas_1.genome\
  --cluster\
  --mds-plot 10\
  --out 1kg.mds.rms.ccls.sol.ahgwas

# Create a racefile of your own data.
awk '{print $1,$2,"OWN"}' 1kg.mds.rms.ccls.sol.ahgwas.fam > racefile_own.txt

# Concatenate racefiles
cat /ccls/home/sli/GWAS/ref/phase3/race_1kG_P3.txt racefile_own.txt | sed -e '1i\FID IID race' > racefile.txt

# prepare mds and fam files 
cp 1kg.mds.rms.ccls.sol.ahgwas.mds racefile.mds
cp 1kg.mds.rms.ccls.sol.ahgwas.fam racefile.fam

```


```{r clustering of samples, creating fam file}

setwd('/ccls/home/sli/GWAS/libby/rmsLat')

library(tidyverse)
library(class)
library(data.table)

data<- fread(file="racefile.mds",header=TRUE)
race<- fread(file="racefile.txt",header=TRUE) 

datafile <- left_join(data,race,by= c("IID","FID"))  %>%
  mutate(race = factor(race, levels =c("AFR", "AMR", "EAS", "EUR", "OWN") )) %>% na.omit() 

shape_vector <- c(20, 20, 20, 20, 4)
color_vector <- c("green1","mediumblue","hotpink","red","grey67")

png("MDS1_2.png",units  = "cm", width = 30, height = 15, res=300)
datafile%>%
  arrange(desc(race))%>%
  ggplot(aes(C1, C2, group=race)) + 
  geom_point(aes(shape=race, color=race), size=2)+
  scale_shape_manual(values=shape_vector)+
  scale_color_manual(values=color_vector)+
  theme_bw()
dev.off()

png("MDS1_3.png",units  = "cm", width = 30, height = 15, res=300)
datafile%>%
  arrange(desc(race))%>%
  ggplot(aes(C1, C3, group=race)) + 
  geom_point(aes(shape=race, color=race), size=2)+
  scale_shape_manual(values=shape_vector)+
  scale_color_manual(values=color_vector)+
  theme_bw()
dev.off()

png("MDS1_4.png",units  = "cm", width = 30, height = 15, res=300)
datafile%>%
  arrange(desc(race))%>%
  ggplot(aes(C1, C4, group=race)) + 
  geom_point(aes(shape=race, color=race), size=2)+
  scale_shape_manual(values=shape_vector)+
  scale_color_manual(values=color_vector)+
  theme_bw()
dev.off()

png("MDS1_5.png",units  = "cm", width = 30, height = 15, res=300)
datafile%>%
  arrange(desc(race))%>%
  ggplot(aes(C1, C5, group=race)) + 
  geom_point(aes(shape=race, color=race), size=2)+
  scale_shape_manual(values=shape_vector)+
  scale_color_manual(values=color_vector)+
  theme_bw()
dev.off()

#impute race information
own_not_ind <- which(!datafile$race == "OWN")
own_ind <- which(datafile$race == "OWN")

train <-datafile[own_not_ind, c("C1","C2","C3","C4","C5","C6","C7","C8","C9","C10")]
test <-datafile[own_ind, c("C1","C2","C3","C4","C5","C6","C7","C8","C9","C10")]
train_results <- datafile[own_not_ind, "race"]%>%as.data.frame()
train_results = as.character(train_results[,1])
set.seed(1990)
test_results <- as.character(knn(train=train, test=test, cl=train_results, k=10))
test_results_1 <- paste0(test_results,"_imp")

datafile_imputed <- datafile %>% mutate(race= as.character(race))
datafile_imputed[own_ind,"race"] <- test_results_1 

datafile_imputed_1 <- datafile_imputed %>%
  na.omit() %>%
  mutate(race = factor(race, levels =c("AFR","AFR_imp","AMR", "AMR_imp","EAS", "EAS_imp","EUR", "EUR_imp") ))

shape_vector <- c(20, 2, 20, 2, 20, 2, 20, 2, 20, 2)
color_vector <- c("green1","green1","mediumblue","mediumblue","hotpink","hotpink","coral","coral")

png("MDS1_2_imputed.png",units  = "cm", width = 30, height = 15, res=300)
datafile_imputed_1%>%
  ggplot(aes(C1, C2, group=race)) + 
  geom_point(aes(shape=race, color=race), size=2)+
  scale_shape_manual(values=shape_vector)+
  scale_color_manual(values=color_vector)+
  theme_bw()
dev.off()

png("MDS1_3_imputed.png",units  = "cm", width = 30, height = 15, res=300)
datafile_imputed_1%>%
  ggplot(aes(C1, C3, group=race)) + 
  geom_point(aes(shape=race, color=race), size=2)+
  scale_shape_manual(values=shape_vector)+
  scale_color_manual(values=color_vector)+
  theme_bw()
dev.off()

png("MDS1_4_imputed.png",units  = "cm", width = 30, height = 15, res=300)
datafile_imputed_1%>%
  ggplot(aes(C1, C4, group=race)) + 
  geom_point(aes(shape=race, color=race), size=2)+
  scale_shape_manual(values=shape_vector)+
  scale_color_manual(values=color_vector)+
  theme_bw()
dev.off()

png("MDS1_5_imputed.png",units  = "cm", width = 30, height = 15, res=300)
datafile_imputed_1%>%
  ggplot(aes(C1, C5, group=race)) + 
  geom_point(aes(shape=race, color=race), size=2)+
  scale_shape_manual(values=shape_vector)+
  scale_color_manual(values=color_vector)+
  theme_bw()
dev.off()

# create a file with all clinical information
imputed_race = datafile_imputed_1 %>% 
  select(FID,IID,race) %>% 
  dplyr::rename("race_imp"="race")

total_ids = fread("rms.ccls.sol.ahgwas.fam") %>% 
  select(V1,V2) %>% 
  mutate(V1=as.character(V1), V2=as.character(V2)) %>%
  dplyr::rename("FID"="V1","IID"="V2")

## ahgwas clinical info
ahgwas_0 = fread("phs001367.v1.pht008248.v1.p1.c1.Add_Health_Subject_Phenotypes.GRU-IRB-PUB-GSO.txt",
                 skip="dbGaP_Subject_ID")%>%
  mutate(IID=as.character(GID),
         sex=GENDER,
         age=AGE)%>%
  select(IID,sex,RACE,age) %>%
  left_join(total_ids,by="IID")
White <- which(ahgwas_0$RACE == 1)
Black <- which(ahgwas_0$RACE == 2)
Hispanic <- which(ahgwas_0$RACE == 5)
Asian <- which(ahgwas_0$RACE == 4)
NativeAmerican <- which(ahgwas_0$RACE == 3)
ahgwas_0$race <- NA
ahgwas_0$race[White] <- "White"
ahgwas_0$race[Black] <- "Black"
ahgwas_0$race[Hispanic] <- "Hispanic"
ahgwas_0$race[Asian] <- "Asian"
ahgwas_0$race[NativeAmerican] <- "NativeAmerican"
ahgwas_1 = ahgwas_0 %>% select(FID,IID,race,sex,age) %>% mutate(caco=0)

## ccls clinical info
cases = fread("/ccls/home/sli/GWAS/libby/v3/ccls.cases.list")
ccls_fam = fread("ccls.gwas.fam") %>% select(V1,V2,V5) %>%
  anti_join(cases,by=c("V1","V2"))
colnames(ccls_fam) = c("FID","IID","sex")
ccls_0 <- fread("/ccls/home/sli/GWAS/libby/clinical/sebastian_ccls_yrace.csv") %>%
  select(FID,IID,Y_race) %>%
  inner_join(ccls_fam,by=c("FID","IID"))
White <- which(ccls_0$Y_race == 1)
Black <- which(ccls_0$Y_race == 2)
Hispanic <- which(ccls_0$Y_race == 3)
Asian <- which(ccls_0$Y_race == 4)
Unknown <- which(ccls_0$Y_race == 5)
ccls_0$race <- NA
ccls_0$race[White] <- "White"
ccls_0$race[Black] <- "Black"
ccls_0$race[Hispanic] <- "Hispanic"
ccls_0$race[Asian] <- "Asian"
ccls_0$race[Unknown] <- "Unknown"
ccls_1 = ccls_0 %>% 
  select(FID,IID,race,sex) %>% 
  mutate(caco=0,age=09) %>%  # 09 is a placeholder to show they are ok
  filter(!(FID%in%ahgwas_1$FID & IID%in%ahgwas_1$IID)) # remove samples that are also in the ahgwas
  
## rms clinical info
rms_fam = fread("rms.gwas.fam") %>% select(V1,V2,V5)
colnames(rms_fam) = c("FID","IID","sex")
rms_0 <- fread("/ccls/home/sli/GWAS/libby/clinical/sebastian_rms.csv") %>%
  select(FID, IID, Y_race) %>%
  inner_join(rms_fam,by=c("FID","IID"))
White <- which(rms_0$Y_race == 1)
Black <- which(rms_0$Y_race == 2)
Hispanic <- which(rms_0$Y_race == 3)
Asian <- which(rms_0$Y_race == 4)
Unknown <- which(rms_0$Y_race == 5)
rms_0$race <- NA
rms_0$race[White] <- "White"
rms_0$race[Black] <- "Black"
rms_0$race[Hispanic] <- "Hispanic"
rms_0$race[Asian] <- "Asian"
rms_0$race[Unknown] <- "Unknown"
rms_1 = rms_0 %>% 
  select(FID,IID,race,sex) %>% 
  mutate(caco=1,age=09) 


## dbgap 82131 clinical info
d82131.0 = fread("phs000810.v1.pht004715.v1.p1.c1.HCHS_SOL_Cohort_Subject_Phenotypes.HMB-NPU.txt") %>% as.data.frame()
colnames(d82131.0)[1:ncol(d82131.0)-1] <- colnames(d82131.0)[2:ncol(d82131.0)]
d82131.1 = d82131.0[,-ncol(d82131.0)] %>%
  mutate(race=REGION,
         caco=0,
         age=AGE,
         sex=ifelse(GENDER=="M",1,
                    ifelse(GENDER=="F",2,0)),
         FID=SUBJECT_ID,
         IID=SUBJECT_ID) %>%
  select(FID,IID,race,sex,caco,age) 

## dbgap 82129 clinical info
d82129.0 = fread("phs000810.v1.pht004715.v1.p1.c2.HCHS_SOL_Cohort_Subject_Phenotypes.HMB.txt") %>% as.data.frame()
colnames(d82129.0)[1:ncol(d82129.0)-1] <- colnames(d82129.0)[2:ncol(d82129.0)]
d82129.1 = d82129.0[,-ncol(d82129.0)] %>%
  mutate(race=REGION,
         caco=0,
         age=AGE,
         sex=ifelse(GENDER=="M",1,
                    ifelse(GENDER=="F",2,0)),
         FID=SUBJECT_ID,
         IID=SUBJECT_ID) %>%
  select(FID,IID,race,sex,caco,age) 

# combine files
ahgwas_ccls_rms = rbind(ahgwas_1,ccls_1,rms_1,d82131.1,d82129.1) %>%  
  as.data.frame() %>%
  #inner_join(total_ids,by=c("FID","IID")) %>%
  inner_join(imputed_race,by=c("FID","IID")) %>%
  mutate(race_imp=as.character(race_imp))
ahgwas_ccls_rms$race[which(ahgwas_ccls_rms$race=="")] <- "Unknown"

table(ahgwas_ccls_rms$race_imp,ahgwas_ccls_rms$race)
  #         Asian Black Central_American Cuban Dominican Hispanic Mexican
  # AFR_imp     1  2107               45   153       544       53       1
  # AMR_imp    89    43             1207   921       587     1925    4131
  # EAS_imp   509     0                1     0         0       16       3
  # EUR_imp    12     1               13   915         1      225      12
  #        
  #         NativeAmerican Puerto_Rican South_American Unknown White
  # AFR_imp              2          140             10     119    19
  # AMR_imp             28         1841            741    1002   139
  # EAS_imp              0            0              1       7     2
  # EUR_imp             17            6             22      95  6237

table(ahgwas_ccls_rms$race_imp)
# AFR_imp AMR_imp EAS_imp EUR_imp 
# 3197   12656     539    7556 
   
write.table(ahgwas_ccls_rms,"allsubsRaceCaCoAge.txt", col.names = TRUE, row.names = FALSE, quote=FALSE)

################################################

ahgwas_ccls_rms= fread("allsubsRaceCaCoAge.txt") %>%
  mutate(caco = caco+1)

eur.nci.id = fread("/ccls/home/sli/GWAS/libby/v3/NCI.EUR.ids",
                   header=FALSE) %>%
  left_join(ahgwas_ccls_rms,c("V1"="FID","V2"="IID")) %>%
  filter(race_imp=="AMR_imp")

nrow(eur.nci.id)
# 38
## drop these 38 subjects

table(eur.nci.id$caco)
#  1  2 
# 23 15 

#LATINO ID
fam_file <- ahgwas_ccls_rms %>%
  anti_join(eur.nci.id,c("FID"="V1","IID"="V2")) %>%
  dplyr::filter(race_imp=="AMR_imp",
                sex %in% c(1,2))%>%
  mutate(Fa=0, Mo=0) %>%
  select(FID,IID,Fa,Mo,sex,caco) %>%
  mutate(FID = as.character(FID),
         IID = as.character(IID))

id_file = fam_file %>% select(FID,IID)
write.table(fam_file, "lat.ahgwas.ccls.rms.fam", col.names = FALSE, row.names = FALSE, quote=FALSE)
write.table(id_file, "lat.ids.txt", col.names = FALSE, row.names = FALSE, quote=FALSE)

con_id = fam_file %>%
  filter(caco==1) %>%
  select(FID,IID)
write.table(con_id, "lat.con.ids.txt", col.names = FALSE, row.names = FALSE, quote=FALSE)
  
```

```{bash extract samples and preprocess}


cd /ccls/home/sli/GWAS/libby/rmsLat

plink --bfile rms.ccls.sol.ahgwas\
  --keep lat.ids.txt\
  --indiv-sort f lat.ids.txt\
  --make-bed\
  --out rms.lat.0
# 603096 variants and 12617 people pass filters and QC.

# Missingness
plink --bfile rms.lat.0 --geno 0.2 --make-bed --out tmp/rms.lat.1
plink --bfile tmp/rms.lat.1 --mind 0.2 --make-bed --out tmp/rms.lat.2
plink --bfile tmp/rms.lat.2 --geno 0.05 --make-bed --out tmp/rms.lat.3
plink --bfile tmp/rms.lat.3 --mind 0.05 --make-bed --out tmp/rms.lat.4

# HW, cut-off is set to be 1e-7 since these are latinos
plink --bfile tmp/rms.lat.4\
  --hwe 1e-8\
  --make-bed\
  --keep lat.con.ids.txt\
  --out tmp/rms.lat.4.1
# 19332 variants removed due to Hardy-Weinberg exact test

plink --bfile tmp/rms.lat.4\
  --extract tmp/rms.lat.4.1.bim\
  --make-bed\
  --out tmp/rms.lat.5
# 607049 variants and 9124 people pass filters and QC.

# MAF 
plink --bfile tmp/rms.lat.5 --maf 0.01 --make-bed --out tmp/rms.lat.6
# 3789 variants removed due to minor allele threshold(s)

# Heterozygosity rate
plink --bfile tmp/rms.lat.6\
  --exclude /ccls/home/sli/GWAS/ref/codes/inversion_hg37.txt\
  --indep-pairwise 50 5 0.2\
  --out tmp/indepSNP
  
plink --bfile tmp/rms.lat.6\
  --extract tmp/indepSNP.prune.in\
  --het\
  --out R_check
  
# Plot of the heterozygosity rate distribution
Rscript --no-save ~/GWAS/ref/codes/check_heterozygosity_rate.R
Rscript --no-save ~/GWAS/ref/codes/heterozygosity_outliers_list.R
sed 's/"// g' fail-het-qc.txt | awk '{print$1, $2}'> het_fail_ind.txt

# Remove heterozygosity rate outliers.
plink --bfile tmp/rms.lat.6 --remove het_fail_ind.txt --make-bed\
  --out tmp/rms.lat.7
# 142 people removed
# 567881 variants and 12464 people pass filters and QC.

# relatedness
# Check for relationships between individuals with a pihat > 0.2.
plink --bfile tmp/rms.lat.7\
  --extract tmp/indepSNP.prune.in\
  --genome\
  --min 0.2\
  --out pihat_min0.2
grep -v "UN|PI_HAT" pihat_min0.2.genome > pihat_min0.2.related
awk 'NR!=1{print $1,$2}' pihat_min0.2.related > related.fam.txt
# Generate a plot to assess the type of relationship.
# Rscript --no-save ~/GWAS/ref/codes/Relatedness.R

plink --bfile tmp/rms.lat.7\
  -make-bed\
  --out rms.lat.8\
  --remove related.fam.txt
# 567881 variants and 12360 people pass filters and QC

##############################no c1
awk '{print $1,$2}' sol.c1.fam > sol.c1.subs

plink --bfile rms.lat.8\
  --remove sol.c1.subs\
  --make-bed\
  --out rms.lat.noc1
# 567881 variants and 9465 people pass filters and QC.
##############################

```

>  do case control gwas without control control GWAS

# imputing on topmed server

```{bash imputing on topmed server}

cd /ccls/home/sli/GWAS/libby/rmsLat

plink --bfile rms.lat.8\
  --allow-no-sex\
  --recode vcf bgz\
  --keep-allele-order\
  --out topmed_pre/rms.lat.preimp
  
bcftools sort topmed_pre/rms.lat.preimp.vcf.gz -Oz -o topmed_pre/rms.lat.sorted.vcf.gz
tabix topmed_pre/rms.lat.sorted.vcf.gz

#split by chromomse
for chr in {1..22}; do
  bcftools view --regions "${chr}" topmed_pre/rms.lat.sorted.vcf.gz\
  -o topmed_pre/rms.lat."${chr}".vcf.gz -Oz &
done

cd /ccls/home/sli/GWAS/libby/rmsLat/topmed_post
for chr in {1..20}; do
  unzip -P 'O.nDGMjgA8vp8' chr_"${chr}".zip
  echo chr_"${chr}".zip unzip success
done  
unzip -P 'O.nDGMjgA8vp8' chr_22.zip

##############################no c1
cd /ccls/home/sli/GWAS/libby/rmsLat

plink --bfile rms.lat.noc1\
  --recode vcf bgz\
  --keep-allele-order\
  --out topmed_pre/rms.lat.noc1

bcftools sort topmed_pre/rms.lat.noc1.vcf.gz -Oz -o topmed_pre/rms.lat.noc1.sorted.vcf.gz
tabix topmed_pre/rms.lat.noc1.sorted.vcf.gz

#split by chromomse
for chr in {1..22}; do
  bcftools view --regions "${chr}" topmed_pre/rms.lat.noc1.sorted.vcf.gz\
  -o topmed_pre/rms.lat."${chr}".noc1.vcf.gz -Oz &
done

cd /ccls/home/sli/GWAS/libby/rmsLat/topmed_post/noc1

for chr in {1..22}; do
  unzip -P 'ugzAHKldA27ECb' chr_"${chr}".zip
  echo chr_"${chr}".zip unzip success
done  
##############################


```

# run caco analysis (without doing control control GWAS)

including c1

```{bash run caco analysis}

cd /ccls/home/sli/GWAS/libby/rmsLat


for chr in {1..22};do
  awk '($7 >0.6) {print $1}' <(zcat topmed_post/chr"${chr}".info.gz) | sed '1d' > \
      topmed_post/include.06.chr"${chr}".txt
done

# Convert imputed data to plink files
cd /ccls/home/sli/GWAS/libby/rmsLat

RACE="lat"

for chr in {1..22};do
  plink2 --vcf topmed_post/chr"${chr}".dose.vcf.gz\
    dosage=HDS \
    --snps-only 'just-acgt'\
    --extract topmed_post/include.06.chr"${chr}".txt\
    --maf 0.01\
    --make-bed\
    --out topmed_post/$RACE.chr"${chr}".iptd\
    --threads 200
done

find ./topmed_post -name "$RACE.chr*.iptd.bim" > ./topmed_post/merge.$RACE.list 
sed -i 's/.bim//g' ./topmed_post/merge.$RACE.list

plink --merge-list topmed_post/merge.$RACE.list\
  --indiv-sort 0\
  --make-bed\
  --out rms.co.$RACE
  
```

```{r creating $RACE.ahgwas.ccls.rms.fam}

library(tidyverse)
library(data.table)
setwd("/ccls/home/sli/GWAS/libby/rmsLat")

allFam = fread("lat.ahgwas.ccls.rms.fam") %>%
  mutate(IID=paste0(V1,"_",V2)) %>%
  select(IID,V3,V4,V5,V6)
noc1Fam = fread(paste0("rms.co.",RACE,".fam")) %>%
  select(V1,V2) %>%
  left_join(allFam,by=c("V2"="IID"))

write.table(noc1Fam, "rms.co.lat.fam.updt", col.names = FALSE, row.names = FALSE, quote=FALSE)

```

```{bash GWAS trial run for RMS CACO with plink}

#################### with c1
cd /ccls/home/sli/GWAS/libby/rmsLat
RACE="lat"

mv rms.co.$RACE.fam rms.co.$RACE.fam.bkp
mv rms.co.$RACE.fam.updt rms.co.$RACE.fam

# HWE among controls using keep if
plink2 --bfile rms.co.$RACE\
  --hwe midp 1e-4 \
  --keep-if PHENO1 == control\
  --make-pgen\
  --out tmp/rms.co."${RACE}"_1\
  --threads 200

# Make pgen again using remaining variants from pvar file 
plink2 --bfile rms.co.$RACE\
  --make-pgen\
  --extract tmp/rms.co."${RACE}"_1.pvar\
  --out tmp/rms.co."${RACE}"_2\
  --threads 200

# Make bed
plink2 --pfile tmp/rms.co."${RACE}"_2\
  --make-bed \
  --out rms.control."${RACE}"

#make bed for PCA analysis
# Create pruned SNPs for later use as controlling variables in regression
plink2 --bfile rms.control."${RACE}"\
  --make-bed \
  --out tmp/rms.co."${RACE}".pca

plink2 --bfile tmp/rms.co."${RACE}".pca\
  --maf 0.1\
  --indep-pairwise 50 10 0.1 \
  --out tmp/rms.co."${RACE}".pruned

plink2 --bfile rms.control."${RACE}"\
  --extract tmp/rms.co."${RACE}".pruned.prune.in\
  --pca 10\
  --out rms.control."${RACE}".pruned.pca\
  --threads 200

# Regression
plink2 --bfile rms.control."${RACE}"\
  --covar rms.control."${RACE}".pruned.pca.eigenvec\
  --threads 200\
  --glm firth-fallback hide-covar cols=chrom,pos,ref,alt1,nobs,a1countcc,a1freq,a1freqcc,beta,se,p,err\
  --out rms.co."${RACE}"

```

>> Run analysis based on subtypes

```{r generating IDs for different subtypes}

setwd('/ccls/home/sli/GWAS/libby/rmsLat')

library(tidyverse)
library(data.table)

rms_0 <- fread("/ccls/home/sli/GWAS/libby/clinical/sebastian_rms.csv")
em = rms_0 %>% filter(subtype=="embryonal") %>% 
    dplyr::mutate(id = paste0(FID,"_",IID)) %>%
    select(id)

alv = rms_0 %>% filter(subtype=="alveolar") %>% 
    dplyr::mutate(id = paste0(FID,"_",IID)) %>%
    select(id)

RACE= "lat"
allSubs = fread(paste0("rms.co.",RACE,".fam")) %>%
    select(V1,V2,V6)

emOnly = allSubs %>%
    dplyr::filter(V2 %in% em$id | V6==1) %>%
    select(V1,V2)

alvOnly = allSubs %>%
    dplyr::filter(V2 %in% alv$id | V6==1) %>%
    select(V1,V2)     

write.table(emOnly, paste0("subtype.em.full.",RACE,".ids.txt"), 
    col.names = FALSE, row.names = FALSE, quote=FALSE)

write.table(alvOnly, paste0("subtype.alv.full.",RACE,".ids.txt"), 
    col.names = FALSE, row.names = FALSE, quote=FALSE)

```

```{bash run caco analysis in based on different subtypes}

cd /ccls/home/sli/GWAS/libby/rmsLat

# Select the proper subjects for subtypes
RACE="lat"
plink --bfile rms.co.$RACE\
  --keep subtype.em.full.$RACE.ids.txt\
  --make-bed\
  --out rms.em.co.$RACE

plink --bfile rms.co.$RACE\
  --keep subtype.alv.full.$RACE.ids.txt\
  --make-bed\
  --out rms.alv.co.$RACE

RACE="lat"
for sub in "em" "alv"; do
  # HWE among controls using keep if
  plink2 --bfile rms.$sub.co.$RACE\
    --hwe midp 1e-4 \
    --keep-if PHENO1 == control\
    --make-pgen \
    --out tmp/rms.$sub.co."${RACE}"_1\
    --threads 200

  # Make pgen again using remaining variants from pvar file 
  plink2 --bfile rms.$sub.co.$RACE\
    --make-pgen\
    --extract tmp/rms.$sub.co."${RACE}"_1.pvar\
    --out tmp/rms.$sub.co."${RACE}"_2\
    --threads 200

  # Make bed
  plink2 --pfile tmp/rms.$sub.co."${RACE}"_2\
    --make-bed\
    --out rms.$sub.control."${RACE}"

  #make bed for PCA analysis
  # Create pruned SNPs for later use as controlling variables in regression
  plink2 --bfile rms.$sub.control."${RACE}"\
    --make-bed \
    --out tmp/rms.$sub.co."${RACE}".pca
  
  plink2 --bfile tmp/rms.$sub.co."${RACE}".pca\
    --maf 0.1\
    --indep-pairwise 50 10 0.1 \
    --out tmp/rms.$sub.co."${RACE}".pruned
  
  plink2 --bfile rms.$sub.control."${RACE}"\
    --extract tmp/rms.$sub.co."${RACE}".pruned.prune.in\
    --pca 10\
    --out rms.$sub.control."${RACE}".pruned.pca\
    --threads 200
  
  # Regression
  plink2 --bfile rms.$sub.control."${RACE}"\
    --covar rms.$sub.control."${RACE}".pruned.pca.eigenvec\
    --threads 200\
    --glm firth-fallback hide-covar cols=chrom,pos,ref,alt1,nobs,a1countcc,a1freq,a1freqcc,beta,se,p,err\
    --out rms.$sub.co."${RACE}"
done

```

```{r visualization}

setwd("/ccls/home/sli/GWAS/libby/rmsLat")

library(tidyverse)
library(QCEWAS)
library(qqman)
library(data.table)

files =c("rms.co.lat.PHENO1.glm.logistic.hybrid",
         "rms.em.co.lat.PHENO1.glm.logistic.hybrid",
         "rms.alv.co.lat.PHENO1.glm.logistic.hybrid")

for(i in files){

  filename <- gsub("^.*/","",i)
  gwas_data <- fread(i) %>% na.omit()

  if(length(grep("meta",i)) !=0 ){
    gwas_data_1 <- gwas_data %>%
      dplyr::mutate(ID=MarkerName, P=`P-value`)%>%
      separate(MarkerName, c("CHROM","POS","REF","ALT"))%>%
      dplyr::mutate(CHROM = as.numeric(gsub("chr","",as.character(CHROM))),POS = as.numeric(as.character(POS)))
  } else{
    gwas_data_1 <- gwas_data %>%
      dplyr::mutate(P= as.numeric(as.character(P)),CHROM = as.numeric(as.character(`#CHROM`)), POS = as.numeric(as.character(POS)))%>%
      arrange(P)
  }

  gwas_data_1 <- gwas_data_1[is.finite(gwas_data_1$P),  ]

  lambda <- P_lambda(gwas_data_1$P)
  print(paste("lambda value of",filename,"is:",lambda))

  png(paste0(filename,".manhattan.png"), width = 10, height = 7, units = 'in', res = 300)
  manhattan(gwas_data_1, chr = "CHROM", bp = "POS", p = "P", snp = "ID")
  dev.off()

  png(paste0(filename,".qq.png"), width = 10, height = 7, units = 'in', res = 300)
  qq(gwas_data_1$P)
  dev.off()
}

```

Excluding c1

```{bash run caco analysis}


cd /ccls/home/sli/GWAS/libby/rmsLat


for chr in {1..22};do
  awk '($7 >0.6) {print $1}' <(zcat topmed_post/noc1/chr"${chr}".info.gz) | sed '1d' > \
      topmed_post/noc1/include.06.chr"${chr}".txt
done

# Convert imputed data to plink files
cd /ccls/home/sli/GWAS/libby/rmsLat

RACE="lat"

for chr in {1..22};do
  plink2 --vcf topmed_post/noc1/chr"${chr}".dose.vcf.gz\
    dosage=HDS \
    --snps-only 'just-acgt'\
    --extract topmed_post/noc1/include.06.chr"${chr}".txt\
    --maf 0.01\
    --make-bed\
    --out topmed_post/noc1/$RACE.chr"${chr}".iptd.noc1\
    --threads 200
done

find ./topmed_post/noc1 -name "$RACE.chr*.iptd.noc1.bim" > ./topmed_post/noc1/merge.$RACE.list 
sed -i 's/.bim//g' ./topmed_post/noc1/merge.$RACE.list

plink --merge-list topmed_post/noc1/merge.$RACE.list\
  --indiv-sort 0\
  --make-bed\
  --out rms.co.noc1.$RACE
  
```

```{r creating $RACE.ahgwas.ccls.rms.noc1.fam}

library(tidyverse)
library(data.table)
setwd("/ccls/home/sli/GWAS/libby/rmsLat")

allFam = fread("lat.ahgwas.ccls.rms.fam") %>%
  mutate(IID=paste0(V1,"_",V2)) %>%
  select(IID,V3,V4,V5,V6)
noc1Fam = fread("rms.co.noc1.lat.fam") %>%
  select(V1,V2) %>%
  left_join(allFam,by=c("V2"="IID"))

write.table(noc1Fam, "rms.co.noc1.lat.fam.updt", col.names = FALSE, row.names = FALSE, quote=FALSE)

```

```{bash GWAS trial run for RMS CACO with plink}

#################### without c1
cd /ccls/home/sli/GWAS/libby/rmsLat
RACE="lat"

mv rms.co.noc1.$RACE.fam rms.co.noc1.$RACE.fam.bkp
mv rms.co.noc1.$RACE.fam.updt rms.co.noc1.$RACE.fam

# HWE among controls using keep if
plink2 --bfile rms.co.noc1.$RACE\
  --hwe midp 1e-4 \
  --keep-if PHENO1 == control\
  --make-pgen\
  --out tmp/rms.co.noc1."${RACE}"_1\
  --threads 200

# Make pgen again using remaining variants from pvar file 
plink2 --bfile rms.co.noc1.$RACE\
  --make-pgen\
  --extract tmp/rms.co.noc1."${RACE}"_1.pvar\
  --out tmp/rms.co.noc1."${RACE}"_2\
  --threads 200

# Make bed
plink2 --pfile tmp/rms.co.noc1."${RACE}"_2\
  --make-bed \
  --out rms.control.noc1."${RACE}"

#make bed for PCA analysis
# Create pruned SNPs for later use as controlling variables in regression
plink2 --bfile rms.control.noc1."${RACE}"\
  --make-bed \
  --out tmp/rms.co.noc1."${RACE}".pca

plink2 --bfile tmp/rms.co.noc1."${RACE}".pca\
  --maf 0.1\
  --indep-pairwise 50 10 0.1 \
  --out tmp/rms.co.noc1."${RACE}".pruned

plink2 --bfile rms.control.noc1."${RACE}"\
  --extract tmp/rms.co.noc1."${RACE}".pruned.prune.in\
  --pca 10\
  --out rms.control.noc1."${RACE}".pruned.pca\
  --threads 200

# Regression
plink2 --bfile rms.control.noc1."${RACE}"\
  --covar rms.control.noc1."${RACE}".pruned.pca.eigenvec\
  --threads 200\
  --glm firth-fallback hide-covar cols=chrom,pos,ref,alt1,nobs,a1countcc,a1freq,a1freqcc,beta,se,p,err\
  --out rms.co.noc1."${RACE}"

```

>> Run analysis based on subtypes

```{r generating IDs for different subtypes}

setwd('/ccls/home/sli/GWAS/libby/rmsLat')

library(tidyverse)
library(data.table)

rms_0 <- fread("/ccls/home/sli/GWAS/libby/clinical/sebastian_rms.csv")

em = rms_0 %>% filter(subtype=="embryonal") %>% 
    dplyr::mutate(id = paste0(FID,"_",IID)) %>%
    select(id)

alv = rms_0 %>% filter(subtype=="alveolar") %>% 
    dplyr::mutate(id = paste0(FID,"_",IID)) %>%
    select(id)

RACE= "lat"
allSubs = fread(paste0("rms.co.noc1.",RACE,".fam")) %>%
    select(V1,V2,V6)

emOnly = allSubs %>%
    dplyr::filter(V2 %in% em$id | V6==1) %>%
    select(V1,V2)     

alvOnly = allSubs %>%
    dplyr::filter(V2 %in% alv$id | V6==1) %>%
    select(V1,V2)       

write.table(emOnly, paste0("subtype.em.",RACE,".ids.txt"), 
    col.names = FALSE, row.names = FALSE, quote=FALSE)

write.table(alvOnly, paste0("subtype.alv.",RACE,".ids.txt"), 
    col.names = FALSE, row.names = FALSE, quote=FALSE)


```

```{bash run caco analysis in based on different subtypes}

cd /ccls/home/sli/GWAS/libby/rmsLat

# Select the proper subjects for subtypes
RACE="lat"
plink --bfile rms.co.noc1.$RACE\
  --keep subtype.em.$RACE.ids.txt\
  --make-bed\
  --out rms.em.co.noc1.$RACE

plink --bfile rms.co.noc1.$RACE\
  --keep subtype.alv.$RACE.ids.txt\
  --make-bed\
  --out rms.alv.co.noc1.$RACE

RACE="lat"
for sub in "em" "alv"; do
  # HWE among controls using keep if
  plink2 --bfile rms.$sub.co.noc1.$RACE\
    --hwe midp 1e-4 \
    --keep-if PHENO1 == control\
    --make-pgen \
    --out tmp/rms.$sub.co.noc1."${RACE}"_1\
    --threads 200

  # Make pgen again using remaining variants from pvar file 
  plink2 --bfile rms.$sub.co.noc1.$RACE\
    --make-pgen\
    --extract tmp/rms.$sub.co.noc1."${RACE}"_1.pvar\
    --out tmp/rms.$sub.co.noc1."${RACE}"_2\
    --threads 200

  # Make bed
  plink2 --pfile tmp/rms.$sub.co.noc1."${RACE}"_2\
    --make-bed\
    --out rms.$sub.control.noc1."${RACE}"

  #make bed for PCA analysis
  # Create pruned SNPs for later use as controlling variables in regression
  plink2 --bfile rms.$sub.control.noc1."${RACE}"\
    --make-bed \
    --out tmp/rms.$sub.co.noc1."${RACE}".pca
  
  plink2 --bfile tmp/rms.$sub.co.noc1."${RACE}".pca\
    --maf 0.1\
    --indep-pairwise 50 10 0.1 \
    --out tmp/rms.$sub.co.noc1."${RACE}".pruned
  
  plink2 --bfile rms.$sub.control.noc1."${RACE}"\
    --extract tmp/rms.$sub.co.noc1."${RACE}".pruned.prune.in\
    --pca 10\
    --out rms.$sub.control.noc1."${RACE}".pruned.pca\
    --threads 200
  
  # Regression
  plink2 --bfile rms.$sub.control.noc1."${RACE}"\
    --covar rms.$sub.control.noc1."${RACE}".pruned.pca.eigenvec\
    --threads 200\
    --glm firth-fallback hide-covar cols=chrom,pos,ref,alt1,nobs,a1countcc,a1freq,a1freqcc,beta,se,p,err\
    --out rms.$sub.co.noc1."${RACE}"
done


RACE="lat"
sub="em"
awk '$15<5E-8 {print}' rms.$sub.co.noc1."${RACE}".PHENO1.glm.logistic.hybrid > rms.$sub.co.noc1."${RACE}".sigs

RACE="lat"
sub="alv"
awk '$15<5E-8 {print}' rms.$sub.co.noc1."${RACE}".PHENO1.glm.logistic.hybrid > rms.$sub.co.noc1."${RACE}".sigs

```

```{r visualization}


setwd("/ccls/home/sli/GWAS/libby/rmsLat")

library(tidyverse)
library(QCEWAS)
library(qqman)
library(data.table)

files =c("rms.co.noc1.lat.PHENO1.glm.logistic.hybrid",
         "rms.em.co.noc1.lat.PHENO1.glm.logistic.hybrid",
         "rms.alv.co.noc1.lat.PHENO1.glm.logistic.hybrid")

for(i in files){

  filename <- gsub("^.*/","",i)
  gwas_data <- fread(i) %>% na.omit()

  if(length(grep("meta",i)) !=0 ){
    gwas_data_1 <- gwas_data %>%
      dplyr::mutate(ID=MarkerName, P=`P-value`)%>%
      separate(MarkerName, c("CHROM","POS","REF","ALT"))%>%
      dplyr::mutate(CHROM = as.numeric(gsub("chr","",as.character(CHROM))),POS = as.numeric(as.character(POS)))
  } else{
    gwas_data_1 <- gwas_data %>%
      dplyr::mutate(P= as.numeric(as.character(P)),CHROM = as.numeric(as.character(`#CHROM`)), POS = as.numeric(as.character(POS)))%>%
      arrange(P)
  }

  gwas_data_1 <- gwas_data_1[is.finite(gwas_data_1$P),  ]

  lambda <- P_lambda(gwas_data_1$P)
  print(paste("lambda value of",filename,"is:",lambda))

  png(paste0(filename,".manhattan.png"), width = 10, height = 7, units = 'in', res = 300)
  manhattan(gwas_data_1, chr = "CHROM", bp = "POS", p = "P", snp = "ID")
  dev.off()

  png(paste0(filename,".qq.png"), width = 10, height = 7, units = 'in', res = 300)
  qq(gwas_data_1$P)
  dev.off()
}

```

> control control GWAS and then do case control gwas

```{r create a control control fam file and subject file}

setwd('/ccls/home/sli/GWAS/libby/rmsLat')

library(tidyverse)
library(data.table)

total_ids = fread("rms.lat.noc1.fam") %>% 
  select(V1,V2) %>% 
  mutate(V1=as.character(V1), V2=as.character(V2)) %>%
  dplyr::rename("FID"="V1","IID"="V2")

## ahgwas clinical info
ahgwas_0 = fread("phs001367.v1.pht008248.v1.p1.c1.Add_Health_Subject_Phenotypes.GRU-IRB-PUB-GSO.txt",
                 skip="dbGaP_Subject_ID")%>%
  mutate(IID=as.character(GID)) %>%
  inner_join(total_ids,by="IID") %>%
  select(FID,IID) %>% 
  mutate(caco=1)

## ccls clinical info
cases = fread("/ccls/home/sli/GWAS/libby/v3/ccls.cases.list")
ccls_fam = fread("ccls.gwas.fam") %>% 
  select(V1,V2) %>%
  anti_join(cases,by=c("V1","V2"))
colnames(ccls_fam) = c("FID","IID")
ccls_0 <- fread("/ccls/home/sli/GWAS/libby/clinical/sebastian_ccls_yrace.csv") %>%
  select(FID,IID) %>%
  inner_join(ccls_fam,by=c("FID","IID")) %>% 
  inner_join(total_ids,by=c("FID","IID")) %>% 
  select(FID,IID) %>% 
  mutate(caco=2) %>%
  filter(!(FID%in%ahgwas_0$FID & IID%in%ahgwas_0$IID)) 
  
## dbgap 82129 clinical info
d82129.0 = fread("phs000810.v1.pht004715.v1.p1.c2.HCHS_SOL_Cohort_Subject_Phenotypes.HMB.txt") %>% 
  as.data.frame()
colnames(d82129.0)[1:ncol(d82129.0)-1] <- colnames(d82129.0)[2:ncol(d82129.0)]
d82129.1 = d82129.0[,-ncol(d82129.0)] %>%
  mutate(race=REGION,
         caco=1,
         FID=SUBJECT_ID,
         IID=SUBJECT_ID) %>%
  select(FID,IID,caco) 

# combine files
ahgwas_ccls_con = rbind(ahgwas_0,ccls_0,d82129.1) %>%  
  as.data.frame() 

total_fams = fread("rms.lat.noc1.fam") %>% 
  select(V1,V2,V3,V4,V5) %>%
  inner_join(ahgwas_ccls_con,by=c("V1"="FID","V2"="IID"))

# > table(total_fams$caco)
# 
#    1    2 
# 8680  504 

write.table(total_fams, "lat.cvc.fam.txt", 
            col.names = FALSE, 
            row.names = FALSE, 
            quote=FALSE)

total_ids_cvc = total_fams %>%
  select(V1,V2)
  
write.table(total_ids_cvc, "lat.cvc.ids.txt", 
            col.names = FALSE, 
            row.names = FALSE, 
            quote=FALSE)

```

```{bash cvc}

cd /ccls/home/sli/GWAS/libby/rmsLat

plink --bfile rms.lat.noc1\
  --keep lat.cvc.ids.txt\
  --indiv-sort f lat.cvc.ids.txt\
  --make-bed\
  --out lat.cvc

mv lat.cvc.fam lat.cvc.fam.bkp
mv lat.cvc.fam.txt lat.cvc.fam

plink2 --bfile lat.cvc\
  --maf 0.1\
  --indep-pairwise 50 10 0.1\
  --out tmp/lat.cvc.prd

plink2 --bfile lat.cvc\
  --extract tmp/lat.cvc.prd.prune.in\
  --pca 10\
  --out lat.cvc.pca\
  --threads 200

plink2 --bfile lat.cvc\
  --covar lat.cvc.pca.eigenvec\
  --glm firth-fallback hide-covar cols=chrom,pos,ref,alt1,nobs,a1countcc,a1freq,a1freqcc,beta,se,p\
  --out lat.cvc\
  --threads 200

awk '$15<1e-3 {print $3}' lat.cvc.PHENO1.glm.logistic.hybrid > lat.cvc.drop.SNPs.03
awk '$15<1e-2 {print $3}' lat.cvc.PHENO1.glm.logistic.hybrid > lat.cvc.drop.SNPs.02


wc -l lat.cvc.drop.SNPs.03
# 913
wc -l lat.cvc.drop.SNPs.02
# 12715

################################################
# setting the cvc cut off as 03
plink --bfile rms.lat.noc1\
  --exclude lat.cvc.drop.SNPs.03\
  --recode vcf bgz\
  --keep-allele-order\
  --out topmed_pre/lat.rms.cvc03
  
bcftools sort topmed_pre/lat.rms.cvc03.vcf.gz -Oz -o topmed_pre/lat.rms.cvc03.sorted.vcf.gz
tabix topmed_pre/lat.rms.cvc03.sorted.vcf.gz

#split by chromomse
for chr in {1..22}; do
  bcftools view --regions "${chr}" topmed_pre/lat.rms.cvc03.sorted.vcf.gz\
  -o topmed_pre/lat.rms.cvc03.chr"${chr}".vcf.gz -Oz &
done

cd /ccls/home/sli/GWAS/libby/rmsLat/topmed_post
for chr in {1..22}; do
  unzip -P 'o5mKOMjPvMsW7v' chr_"${chr}".zip
  echo chr_"${chr}".zip unzip success
done  

cd /ccls/home/sli/GWAS/libby/rmsLat

for chr in {1..22};do
  awk '($7 >0.6) {print $1}' <(zcat topmed_post/chr"${chr}".info.gz) | sed '1d' > \
      topmed_post/include.06.chr"${chr}".txt
done

RACE="lat"
for chr in {1..22};do
  plink2 --vcf topmed_post/chr"${chr}".dose.vcf.gz\
    dosage=HDS \
    --snps-only 'just-acgt'\
    --extract topmed_post/include.06.chr"${chr}".txt\
    --maf 0.01\
    --make-bed\
    --out topmed_post/$RACE.chr"${chr}".iptd.cvc\
    --threads 200
done

find ./topmed_post -name "$RACE.chr*.iptd.cvc.bim" > ./topmed_post/merge.$RACE.cvc.list 
sed -i 's/.bim//g' ./topmed_post/merge.$RACE.cvc.list

plink --merge-list topmed_post/merge.$RACE.cvc.list\
  --indiv-sort 0\
  --make-bed\
  --out rms.co.cvc.$RACE
  
# control-control GWAS after imputation
cd /ccls/home/sli/GWAS/libby/rmsLat
RACE="lat"

awk '{print 0,$1"_"$2}' $RACE.cvc.ids.txt > $RACE.cvc.ids.topmed.txt
awk '{print 0,$1"_"$2,$3,$4,$5,$6}' $RACE.cvc.fam > $RACE.cvc.topmed.fam

plink2 --bfile rms.co.cvc.$RACE\
  --keep $RACE.cvc.ids.topmed.txt\
  --indiv-sort f $RACE.cvc.ids.topmed.txt\
  --make-bed\
  --out $RACE.cvc.postimp

mv $RACE.cvc.postimp.fam $RACE.cvc.postimp.fam.bkp
cp $RACE.cvc.topmed.fam $RACE.cvc.postimp.fam
  
plink2 --bfile $RACE.cvc.postimp\
  --maf 0.1\
  --indep-pairwise 50 10 0.1\
  --out tmp/$RACE.cvc.postimp.prd

plink2 --bfile $RACE.cvc.postimp\
  --extract tmp/$RACE.cvc.postimp.prd.prune.in\
  --pca 10\
  --out $RACE.cvc.postimp.pca\
  --threads 200

plink2 --bfile $RACE.cvc.postimp\
  --covar $RACE.cvc.postimp.pca.eigenvec\
  --glm firth-fallback hide-covar cols=chrom,pos,ref,alt1,nobs,a1countcc,a1freq,a1freqcc,beta,se,p\
  --out $RACE.cvc.postimp\
  --threads 200

awk '$15<1e-5 {print $3}' $RACE.cvc.postimp.PHENO1.glm.logistic.hybrid > $RACE.cvc.postimp.drop.SNPs.05

plink2 --bfile rms.co.cvc.$RACE\
  --exclude $RACE.cvc.postimp.drop.SNPs.05\
  --make-bed\
  --out rms.co.cvc.$RACE.05
  
################################################
# setting the cvc cut off as 02
plink --bfile rms.lat.noc1\
  --exclude lat.cvc.drop.SNPs.02\
  --recode vcf bgz\
  --keep-allele-order\
  --out topmed_pre/lat.rms.cvc02
  
bcftools sort topmed_pre/lat.rms.cvc02.vcf.gz -Oz -o topmed_pre/lat.rms.cvc02.sorted.vcf.gz
tabix topmed_pre/lat.rms.cvc02.sorted.vcf.gz

#split by chromomse
for chr in {1..22}; do
  bcftools view --regions "${chr}" topmed_pre/lat.rms.cvc02.sorted.vcf.gz\
  -o topmed_pre/lat.rms.cvc02.chr"${chr}".vcf.gz -Oz &
done

cd /ccls/home/sli/GWAS/libby/rmsLat/topmed_post/02
for chr in {1..22}; do
  unzip -P 'W4mlMBhTkv2A7' chr_"${chr}".zip
  echo chr_"${chr}".zip unzip success
done  

cd /ccls/home/sli/GWAS/libby/rmsLat

for chr in {1..22};do
  awk '($7 >0.6) {print $1}' <(zcat topmed_post/02/chr"${chr}".info.gz) | sed '1d' > \
      topmed_post/02/include.06.chr"${chr}".txt
done

RACE="lat"
for chr in {1..22};do
  plink2 --vcf topmed_post/02/chr"${chr}".dose.vcf.gz\
    dosage=HDS \
    --snps-only 'just-acgt'\
    --extract topmed_post/02/include.06.chr"${chr}".txt\
    --maf 0.01\
    --make-bed\
    --out topmed_post/02/$RACE.chr"${chr}".iptd.02.cvc\
    --threads 200
done

find ./topmed_post/02 -name "$RACE.chr*.iptd.02.cvc.bim" > ./topmed_post/02/merge.$RACE.02.cvc.list 
sed -i 's/.bim//g' ./topmed_post/02/merge.$RACE.02.cvc.list

plink --merge-list topmed_post/02/merge.$RACE.02.cvc.list\
  --indiv-sort 0\
  --make-bed\
  --out rms.co.02.cvc.$RACE

# control-control GWAS after imputation
cd /ccls/home/sli/GWAS/libby/rmsLat
RACE="lat"

awk '{print 0,$1"_"$2}' $RACE.cvc.ids.txt > $RACE.cvc.ids.topmed.txt
awk '{print 0,$1"_"$2,$3,$4,$5,$6}' $RACE.cvc.fam > $RACE.cvc.topmed.fam

plink2 --bfile rms.co.02.cvc.$RACE\
  --keep $RACE.cvc.ids.topmed.txt\
  --indiv-sort f $RACE.cvc.ids.topmed.txt\
  --make-bed\
  --out $RACE.02.cvc.postimp

mv $RACE.02.cvc.postimp.fam $RACE.02.cvc.postimp.fam.bkp
cp $RACE.cvc.topmed.fam $RACE.02.cvc.postimp.fam
  
plink2 --bfile $RACE.02.cvc.postimp\
  --maf 0.1\
  --indep-pairwise 50 10 0.1\
  --out tmp/$RACE.02.cvc.postimp.prd

plink2 --bfile $RACE.02.cvc.postimp\
  --extract tmp/$RACE.02.cvc.postimp.prd.prune.in\
  --pca 10\
  --out $RACE.02.cvc.postimp.pca\
  --threads 200

plink2 --bfile $RACE.02.cvc.postimp\
  --covar $RACE.02.cvc.postimp.pca.eigenvec\
  --glm firth-fallback hide-covar cols=chrom,pos,ref,alt1,nobs,a1countcc,a1freq,a1freqcc,beta,se,p\
  --out $RACE.02.cvc.postimp\
  --threads 200

awk '$15<1e-5 {print $3}' $RACE.02.cvc.postimp.PHENO1.glm.logistic.hybrid > $RACE.02.cvc.postimp.drop.SNPs.05

plink2 --bfile rms.co.02.cvc.$RACE\
  --exclude $RACE.02.cvc.postimp.drop.SNPs.05\
  --make-bed\
  --out rms.co.02.cvc.$RACE.05
      
```

```{r creating $RACE.ahgwas.ccls.rms.cvc.fam}

library(tidyverse)
library(data.table)
setwd("/ccls/home/sli/GWAS/libby/rmsLat")

allFam = fread("lat.ahgwas.ccls.rms.fam") %>%
  mutate(IID=paste0(V1,"_",V2)) %>%
  select(IID,V3,V4,V5,V6)
CaCoCvCFam = fread("topmed_post/lat.chr1.iptd.cvc.fam") %>%
  select(V1,V2) %>%
  left_join(allFam,by=c("V2"="IID"))

write.table(CaCoCvCFam, "rms.co.cvc.lat.fam.updt", 
            col.names = FALSE, 
            row.names = FALSE, 
            quote=FALSE)

```

```{bash GWAS trial run for RMS CACO with plink}

################################################
# setting the cvc cut off as 03

cd /ccls/home/sli/GWAS/libby/rmsLat
RACE="lat"

mv rms.co.cvc.$RACE.05.fam rms.co.cvc.$RACE.05.fam.bkp
cp rms.co.cvc.$RACE.fam.updt rms.co.cvc.05.$RACE.fam

# HWE among controls using keep if
plink2 --bfile rms.co.cvc.$RACE.05\
  --hwe midp 1e-4 \
  --keep-if PHENO1 == control\
  --make-pgen\
  --out tmp/rms.co.cvc."${RACE}"_1\
  --threads 200

# Make pgen again using remaining variants from pvar file 
plink2 --bfile rms.co.cvc.$RACE.05\
  --make-pgen\
  --extract tmp/rms.co.cvc."${RACE}"_1.pvar\
  --out tmp/rms.co.cvc."${RACE}"_2\
  --threads 200

# Make bed
plink2 --pfile tmp/rms.co.cvc."${RACE}"_2\
  --make-bed\
  --out rms.control.cvc."${RACE}"

#make bed for PCA analysis
# Create pruned SNPs for later use as controlling variables in regression
plink2 --bfile rms.control.cvc."${RACE}"\
  --make-bed \
  --out tmp/rms.co.cvc."${RACE}".pca

plink2 --bfile tmp/rms.co.cvc."${RACE}".pca\
  --maf 0.1\
  --indep-pairwise 50 10 0.1 \
  --out tmp/rms.co.cvc."${RACE}".pruned

plink2 --bfile rms.control.cvc."${RACE}"\
  --extract tmp/rms.co.cvc."${RACE}".pruned.prune.in\
  --pca 10\
  --out rms.control.cvc."${RACE}".pruned.pca\
  --threads 200

# Regression
plink2 --bfile rms.control.cvc."${RACE}"\
  --covar rms.control.cvc."${RACE}".pruned.pca.eigenvec\
  --threads 200\
  --glm firth-fallback hide-covar cols=chrom,pos,ref,alt1,nobs,a1countcc,a1freq,a1freqcc,beta,se,p,err\
  --out rms.co.cvc."${RACE}"


################################################
# setting the cvc cut off as 02
cd /ccls/home/sli/GWAS/libby/rmsLat
RACE="lat"

mv rms.co.02.cvc.$RACE.05.fam rms.co.02.cvc.$RACE.05.fam.bkp
cp rms.co.cvc.$RACE.fam.updt rms.co.02.cvc.$RACE.05.fam

# HWE among controls using keep if
plink2 --bfile rms.co.02.cvc.$RACE.05\
  --hwe midp 1e-4 \
  --keep-if PHENO1 == control\
  --make-pgen\
  --out tmp/rms.co.02.cvc."${RACE}"_1\
  --threads 200

# Make pgen again using remaining variants from pvar file 
plink2 --bfile rms.co.02.cvc.$RACE.05\
  --make-pgen\
  --extract tmp/rms.co.02.cvc."${RACE}"_1.pvar\
  --out tmp/rms.co.02.cvc."${RACE}"_2\
  --threads 200

# Make bed
plink2 --pfile tmp/rms.co.02.cvc."${RACE}"_2\
  --make-bed\
  --out rms.control.02.cvc."${RACE}"

#make bed for PCA analysis
# Create pruned SNPs for later use as controlling variables in regression
plink2 --bfile rms.control.02.cvc."${RACE}"\
  --make-bed \
  --out tmp/rms.co.02.cvc."${RACE}".pca

plink2 --bfile tmp/rms.co.02.cvc."${RACE}".pca\
  --maf 0.1\
  --indep-pairwise 50 10 0.1 \
  --out tmp/rms.co.02.cvc."${RACE}".pruned

plink2 --bfile rms.control.02.cvc."${RACE}"\
  --extract tmp/rms.co.02.cvc."${RACE}".pruned.prune.in\
  --pca 10\
  --out rms.control.02.cvc."${RACE}".pruned.pca\
  --threads 200

# Regression
plink2 --bfile rms.control.02.cvc."${RACE}"\
  --covar rms.control.02.cvc."${RACE}".pruned.pca.eigenvec\
  --threads 200\
  --glm firth-fallback hide-covar cols=chrom,pos,ref,alt1,nobs,a1countcc,a1freq,a1freqcc,beta,se,p,err\
  --out rms.co.02.cvc."${RACE}"

```

>> Run analysis based on subtypes

```{r generating IDs for different subtypes}

setwd('/ccls/home/sli/GWAS/libby/rmsLat')

library(tidyverse)
library(data.table)

rms_0 <- fread("/ccls/home/sli/GWAS/libby/clinical/sebastian_rms.csv")

em = rms_0 %>% filter(subtype=="embryonal") %>% 
    dplyr::mutate(id = paste0(FID,"_",IID)) %>%
    select(id)

alv = rms_0 %>% filter(subtype=="alveolar") %>% 
    dplyr::mutate(id = paste0(FID,"_",IID)) %>%
    select(id)

RACE= "lat"
allSubs = fread("rms.co.cvc.lat.fam.updt") %>%
    select(V1,V2,V6)

emOnly = allSubs %>%
    dplyr::filter(V2 %in% em$id | V6==1) %>%
    select(V1,V2)     

alvOnly = allSubs %>%
    dplyr::filter(V2 %in% alv$id | V6==1) %>%
    select(V1,V2)       

write.table(emOnly, paste0("subtype.em.",RACE,".cvc.ids.txt"), 
    col.names = FALSE, row.names = FALSE, quote=FALSE)

write.table(alvOnly, paste0("subtype.alv.",RACE,".cvc.ids.txt"), 
    col.names = FALSE, row.names = FALSE, quote=FALSE)


```

```{bash run caco analysis in based on different subtypes}

cd /ccls/home/sli/GWAS/libby/rmsLat

################################################
# setting the cvc cut off as 03

# Select the proper subjects for subtypes
RACE="lat"
plink --bfile rms.co.cvc.$RACE.05\
  --keep subtype.em.$RACE.cvc.ids.txt\
  --make-bed\
  --out rms.em.co.cvc.$RACE

plink --bfile rms.co.cvc.$RACE.05\
  --keep subtype.alv.$RACE.cvc.ids.txt\
  --make-bed\
  --out rms.alv.co.cvc.$RACE

RACE="lat"
for sub in "em" "alv"; do
  # HWE among controls using keep if
  plink2 --bfile rms.$sub.co.cvc.$RACE\
    --hwe midp 1e-4 \
    --keep-if PHENO1 == control\
    --make-pgen \
    --out tmp/rms.$sub.co.cvc."${RACE}"_1\
    --threads 200

  # Make pgen again using remaining variants from pvar file 
  plink2 --bfile rms.$sub.co.cvc.$RACE\
    --make-pgen\
    --extract tmp/rms.$sub.co.cvc."${RACE}"_1.pvar\
    --out tmp/rms.$sub.co.cvc."${RACE}"_2\
    --threads 200

  # Make bed
  plink2 --pfile tmp/rms.$sub.co.cvc."${RACE}"_2\
    --make-bed\
    --out rms.$sub.control.cvc."${RACE}"

  #make bed for PCA analysis
  # Create pruned SNPs for later use as controlling variables in regression
  plink2 --bfile rms.$sub.control.cvc."${RACE}"\
    --make-bed \
    --out tmp/rms.$sub.co.cvc."${RACE}".pca
  
  plink2 --bfile tmp/rms.$sub.co.cvc."${RACE}".pca\
    --maf 0.1\
    --indep-pairwise 50 10 0.1 \
    --out tmp/rms.$sub.co.cvc."${RACE}".pruned
  
  plink2 --bfile rms.$sub.control.cvc."${RACE}"\
    --extract tmp/rms.$sub.co.cvc."${RACE}".pruned.prune.in\
    --pca 10\
    --out rms.$sub.control.cvc."${RACE}".pruned.pca\
    --threads 200
  
  # Regression
  plink2 --bfile rms.$sub.control.cvc."${RACE}"\
    --covar rms.$sub.control.cvc."${RACE}".pruned.pca.eigenvec\
    --threads 200\
    --glm firth-fallback hide-covar cols=chrom,pos,ref,alt1,nobs,a1countcc,a1freq,a1freqcc,beta,se,p,err\
    --out rms.$sub.co.cvc."${RACE}"
done


RACE="lat"
sub="em"
awk '$15<5E-8 {print}' rms.$sub.co.cvc."${RACE}".PHENO1.glm.logistic.hybrid > rms.$sub.co.cvc."${RACE}".sigs

sub="alv"
awk '$15<5E-8 {print}' rms.$sub.co.cvc."${RACE}".PHENO1.glm.logistic.hybrid > rms.$sub.co.cvc."${RACE}".sigs

################################################
# setting the cvc cut off as 02

# Select the proper subjects for subtypes
RACE="lat"
plink --bfile rms.co.02.cvc.$RACE.05\
  --keep subtype.em.$RACE.cvc.ids.txt\
  --make-bed\
  --out rms.em.co.02.cvc.$RACE

plink --bfile rms.co.02.cvc.$RACE.05\
  --keep subtype.alv.$RACE.cvc.ids.txt\
  --make-bed\
  --out rms.alv.co.02.cvc.$RACE

RACE="lat"
for sub in "em" "alv"; do
  # HWE among controls using keep if
  plink2 --bfile rms.$sub.co.02.cvc.$RACE\
    --hwe midp 1e-4 \
    --keep-if PHENO1 == control\
    --make-pgen \
    --out tmp/rms.$sub.co.02.cvc."${RACE}"_1\
    --threads 200

  # Make pgen again using remaining variants from pvar file 
  plink2 --bfile rms.$sub.co.02.cvc.$RACE\
    --make-pgen\
    --extract tmp/rms.$sub.co.02.cvc."${RACE}"_1.pvar\
    --out tmp/rms.$sub.co.02.cvc."${RACE}"_2\
    --threads 200

  # Make bed
  plink2 --pfile tmp/rms.$sub.co.02.cvc."${RACE}"_2\
    --make-bed\
    --out rms.$sub.control.02.cvc."${RACE}"

  #make bed for PCA analysis
  # Create pruned SNPs for later use as controlling variables in regression
  plink2 --bfile rms.$sub.control.02.cvc."${RACE}"\
    --make-bed \
    --out tmp/rms.$sub.co.02.cvc."${RACE}".pca
  
  plink2 --bfile tmp/rms.$sub.co.02.cvc."${RACE}".pca\
    --maf 0.1\
    --indep-pairwise 50 10 0.1 \
    --out tmp/rms.$sub.co.02.cvc."${RACE}".pruned
  
  plink2 --bfile rms.$sub.control.02.cvc."${RACE}"\
    --extract tmp/rms.$sub.co.02.cvc."${RACE}".pruned.prune.in\
    --pca 10\
    --out rms.$sub.control.02.cvc."${RACE}".pruned.pca\
    --threads 200
  
  # Regression
  plink2 --bfile rms.$sub.control.02.cvc."${RACE}"\
    --covar rms.$sub.control.02.cvc."${RACE}".pruned.pca.eigenvec\
    --threads 200\
    --glm firth-fallback hide-covar cols=chrom,pos,ref,alt1,nobs,a1countcc,a1freq,a1freqcc,beta,se,p,err\
    --out rms.$sub.co.02.cvc."${RACE}"
done


RACE="lat"
sub="em"
awk '$15<5E-8 {print}' rms.$sub.co.02.cvc."${RACE}".PHENO1.glm.logistic.hybrid > rms.$sub.co.02.cvc."${RACE}".sigs

sub="alv"
awk '$15<5E-8 {print}' rms.$sub.co.02.cvc."${RACE}".PHENO1.glm.logistic.hybrid > rms.$sub.co.02.cvc."${RACE}".sigs

```

```{r visualization}


setwd("/ccls/home/sli/GWAS/libby/rmsLat")

library(tidyverse)
library(QCEWAS)
library(qqman)
library(data.table)

files =c("lat.cvc.postimp.PHENO1.glm.logistic.hybrid",
          "lat.02.cvc.postimp.PHENO1.glm.logistic.hybrid",
          "rms.co.cvc.lat.PHENO1.glm.logistic.hybrid",
         "rms.em.co.cvc.lat.PHENO1.glm.logistic.hybrid",
         "rms.alv.co.cvc.lat.PHENO1.glm.logistic.hybrid",
         "rms.co.02.cvc.lat.PHENO1.glm.logistic.hybrid",
         "rms.em.co.02.cvc.lat.PHENO1.glm.logistic.hybrid",
         "rms.alv.co.02.cvc.lat.PHENO1.glm.logistic.hybrid")

for(i in files){

  filename <- gsub("^.*/","",i)
  gwas_data <- fread(i) %>% na.omit()

  if(length(grep("meta",i)) !=0 ){
    gwas_data_1 <- gwas_data %>%
      dplyr::mutate(ID=MarkerName, P=`P-value`)%>%
      separate(MarkerName, c("CHROM","POS","REF","ALT"))%>%
      dplyr::mutate(CHROM = as.numeric(gsub("chr","",as.character(CHROM))),POS = as.numeric(as.character(POS)))
  } else{
    gwas_data_1 <- gwas_data %>%
      dplyr::mutate(P= as.numeric(as.character(P)),CHROM = as.numeric(as.character(`#CHROM`)), POS = as.numeric(as.character(POS)))%>%
      arrange(P)
  }

  gwas_data_1 <- gwas_data_1[is.finite(gwas_data_1$P),  ]

  lambda <- P_lambda(gwas_data_1$P)
  print(paste("lambda value of",filename,"is:",lambda))

  png(paste0(filename,".manhattan.png"), width = 10, height = 7, units = 'in', res = 300)
  manhattan(gwas_data_1, chr = "CHROM", bp = "POS", p = "P", snp = "ID")
  dev.off()

  png(paste0(filename,".qq.png"), width = 10, height = 7, units = 'in', res = 300)
  qq(gwas_data_1$P)
  dev.off()
}

# [1] "lambda value of lat.cvc.postimp.PHENO1.glm.logistic.hybrid is: 1.09795511588471"
# [1] "lambda value of lat.02.cvc.postimp.PHENO1.glm.logistic.hybrid is: 1.09634137053153"
# [1] "lambda value of rms.co.cvc.lat.PHENO1.glm.logistic.hybrid is: 1.03137168593309"
# [1] "lambda value of rms.em.co.cvc.lat.PHENO1.glm.logistic.hybrid is: 1.02158970688775"
# [1] "lambda value of rms.alv.co.cvc.lat.PHENO1.glm.logistic.hybrid is: 0.871226196684855"
# [1] "lambda value of rms.co.02.cvc.lat.PHENO1.glm.logistic.hybrid is: 1.03082538680772"
# [1] "lambda value of rms.em.co.02.cvc.lat.PHENO1.glm.logistic.hybrid is: 1.02062094667661"
# [1] "lambda value of rms.alv.co.02.cvc.lat.PHENO1.glm.logistic.hybrid is: 0.872127318005937"

```

```{r regional look up}

setwd("/ccls/home/sli/GWAS/libby/rmsLat")
library(tidyverse)
library(data.table)

##########################################################################################
# em 03
lat = fread("rms.em.co.cvc.lat.PHENO1.glm.logistic.hybrid") %>%
  filter(P<5E-8) %>%
  select(ID, `#CHROM`, POS, P) %>%
  arrange(`#CHROM`,POS)

#                   ID #CHROM      POS           P
# 1:   chr4:6547843:A:C      4  6547843 2.86904e-08
# 2:  chr9:19792646:T:G      9 19792646 4.68351e-08
# 3:  chr9:19792692:G:C      9 19792692 2.51377e-08
# 4:  chr9:19809128:T:G      9 19809128 1.53412e-08
# 5:  chr9:19832435:T:C      9 19832435 9.10494e-09
# 6: chr13:30775886:C:T     13 30775886 3.22878e-08

eur = fread("/ccls/home/sli/GWAS/libby/rmsdbGAP/rms.em.co.cvc.eur.PHENO1.glm.logistic.hybrid")

## (1) 4:6547843
chr=4
pos1 = 6547843 - 50000
pos2 = 6547843 + 50000
eurLu = eur %>% filter(`#CHROM`==chr,
                       POS<=pos2,
                       POS>=pos1) %>%
  arrange(P) %>%
  select(ID,`#CHROM`,POS,P)
head(eurLu) 
#                  ID #CHROM     POS          P
# 1: chr4:6500775:C:T      4 6500775 0.00645976
# 2: chr4:6576760:A:G      4 6576760 0.00750156
# 3: chr4:6574416:A:G      4 6574416 0.00887600


## (2) chr9:19792646 - chr9:19832435
chr=9
pos1 = 19792646 - 50000
pos2 = 19832435 + 50000
eurLu = eur %>% filter(`#CHROM`==chr,
                       POS<=pos2,
                       POS>=pos1) %>%
  arrange(P) %>%
  select(ID,`#CHROM`,POS,P)
head(eurLu) 
#                   ID #CHROM      POS         P
# 1: chr9:19844276:T:C      9 19844276 0.0146866
# 2: chr9:19880743:G:A      9 19880743 0.0202602
# 3: chr9:19851570:C:A      9 19851570 0.0209531

## (3) chr13:30775886
chr=13
pos1 = 30775886 - 50000
pos2 = 30775886 + 50000
eurLu = eur %>% filter(`#CHROM`==chr,
                       POS<=pos2,
                       POS>=pos1) %>%
  arrange(P) %>%
  select(ID,`#CHROM`,POS,P)
head(eurLu) 
#                    ID #CHROM      POS          P
# 1: chr13:30768558:T:A     13 30768558 0.00704468
# 2: chr13:30747325:T:C     13 30747325 0.01777310
# 3: chr13:30797846:C:T     13 30797846 0.02480000

##########################################################################################
# em 02
lat = fread("rms.em.co.02.cvc.lat.PHENO1.glm.logistic.hybrid") %>%
  filter(P<5E-8) %>%
  select(ID, `#CHROM`, POS, P) %>%
  arrange(`#CHROM`,POS)

#                    ID #CHROM       POS           P
# 1:  chr9:19792646:T:G      9  19792646 4.72881e-08
# 2:  chr9:19792692:G:C      9  19792692 2.50919e-08
# 3:  chr9:19809128:T:G      9  19809128 1.81361e-08
# 4:  chr9:19832435:T:C      9  19832435 1.08762e-08
# 5: chr9:130193706:G:A      9 130193706 4.98037e-08
# 6: chr13:30775886:C:T     13  30775886 4.94555e-08


eur = fread("/ccls/home/sli/GWAS/libby/rmsdbGAP/rms.em.co.02.cvc.eur.PHENO1.glm.logistic.hybrid")

## (1) chr9:19792646 - chr9:19832435
chr=9
pos1 = 19792646 - 50000
pos2 = 19832435 + 50000
eurLu = eur %>% filter(`#CHROM`==chr,
                       POS<=pos2,
                       POS>=pos1) %>%
  arrange(P) %>%
  select(ID,`#CHROM`,POS,P)
head(eurLu) 
#                   ID #CHROM      POS         P
# 1: chr9:19844276:T:C      9 19844276 0.0143226
# 2: chr9:19880743:G:A      9 19880743 0.0199632
# 3: chr9:19851570:C:A      9 19851570 0.0204494
# 4: chr9:19859474:G:C      9 19859474 0.0217376

## (2) chr9:130193706
chr=9
pos1 = 130193706 - 50000
pos2 = 130193706 + 50000
eurLu = eur %>% filter(`#CHROM`==chr,
                       POS<=pos2,
                       POS>=pos1) %>%
  arrange(P) %>%
  select(ID,`#CHROM`,POS,P)
head(eurLu) 
#                    ID #CHROM       POS          P
# 1: chr9:130193207:G:C      9 130193207 0.00783951
# 2: chr9:130187327:G:A      9 130187327 0.01473290
# 3: chr9:130241525:A:G      9 130241525 0.01523760

## (3) chr13:30775886
chr=13
pos1 = 30775886 - 50000
pos2 = 30775886 + 50000
eurLu = eur %>% filter(`#CHROM`==chr,
                       POS<=pos2,
                       POS>=pos1) %>%
  arrange(P) %>%
  select(ID,`#CHROM`,POS,P)
head(eurLu) 
#                    ID #CHROM      POS          P
# 1: chr13:30768558:T:A     13 30768558 0.00481056
# 2: chr13:30747325:T:C     13 30747325 0.01472630
# 3: chr13:30797420:C:T     13 30797420 0.01948790


##########################################################################################
# alv 03
lat = fread("rms.alv.co.cvc.lat.PHENO1.glm.logistic.hybrid") %>%
  filter(P<5E-8) %>%
  select(ID, `#CHROM`, POS, P) %>%
  arrange(`#CHROM`,POS)

#                   ID #CHROM      POS           P
# 1: chr6:153265235:A:G      6 153265235 3.45715e-08
# 2:  chr9:68554620:G:A      9  68554620 2.21988e-08
# 3:  chr9:68556109:C:T      9  68556109 2.21988e-08
# 4:  chr9:68599121:A:G      9  68599121 2.62985e-08
# 5:  chr9:68603905:A:T      9  68603905 6.52365e-09
# 6: chr15:99581506:T:G     15  99581506 2.74707e-09
# 7: chr16:82819052:C:T     16  82819052 2.21415e-08
# 8: chr17:37747101:G:A     17  37747101 3.73115e-08
# 9: chr18:10660903:A:T     18  10660903 4.81396e-08

eur = fread("/ccls/home/sli/GWAS/libby/rmsdbGAP/rms.alv.co.cvc.eur.PHENO1.glm.logistic.hybrid")

## (1) chr6:153265235
chr=6
pos1 = 153265235 - 50000
pos2 = 153265235 + 50000
eurLu = eur %>% filter(`#CHROM`==chr,
                       POS<=pos2,
                       POS>=pos1) %>%
  arrange(P) %>%
  select(ID,`#CHROM`,POS,P)
head(eurLu) 
#                    ID #CHROM       POS         P
# 1: chr6:153290854:C:G      6 153290854 0.0246814
# 2: chr6:153254016:C:T      6 153254016 0.0272389
# 3: chr6:153294049:G:A      6 153294049 0.0279726

## (2) chr9:68554620 - chr9:68603905
chr=9
pos1 = 68554620 - 50000
pos2 = 68603905 + 50000
eurLu = eur %>% filter(`#CHROM`==chr,
                       POS<=pos2,
                       POS>=pos1) %>%
  arrange(P) %>%
  select(ID,`#CHROM`,POS,P)
head(eurLu) 
#                   ID #CHROM      POS          P
# 1: chr9:68532767:G:A      9 68532767 0.00231015
# 2: chr9:68516671:G:A      9 68516671 0.00372569
# 3: chr9:68614208:C:T      9 68614208 0.01499620

## (3) chr15:99581506
chr=15
pos1 = 99581506 - 50000
pos2 = 99581506 + 50000
eurLu = eur %>% filter(`#CHROM`==chr,
                       POS<=pos2,
                       POS>=pos1) %>%
  arrange(P) %>%
  select(ID,`#CHROM`,POS,P)
head(eurLu) 
#                    ID #CHROM      POS          P
# 1: chr15:99618239:T:C     15 99618239 0.00757446
# 2: chr15:99556179:T:C     15 99556179 0.00803949
# 3: chr15:99605019:C:A     15 99605019 0.00942698

## (4) chr16:82819052
chr=16
pos1 = 82819052 - 50000
pos2 = 82819052 + 50000
eurLu = eur %>% filter(`#CHROM`==chr,
                       POS<=pos2,
                       POS>=pos1) %>%
  arrange(P) %>%
  select(ID,`#CHROM`,POS,P)
head(eurLu) 
#                    ID #CHROM      POS          P
# 1: chr16:82855954:C:A     16 82855954 0.00463728
# 2: chr16:82856452:G:A     16 82856452 0.00494452
# 3: chr16:82856281:C:T     16 82856281 0.02083160

## (5) chr17:37747101
chr=17
pos1 = 37747101 - 50000
pos2 = 37747101 + 50000
eurLu = eur %>% filter(`#CHROM`==chr,
                       POS<=pos2,
                       POS>=pos1) %>%
  arrange(P) %>%
  select(ID,`#CHROM`,POS,P)
head(eurLu) 
#                    ID #CHROM      POS          P
# 1: chr17:37718929:A:C     17 37718929 0.00157581
# 2: chr17:37707301:G:A     17 37707301 0.00239383
# 3: chr17:37709840:C:T     17 37709840 0.00301829

## (3) chr18:10660903
chr=18
pos1 = 10660903 - 50000
pos2 = 10660903 + 50000
eurLu = eur %>% filter(`#CHROM`==chr,
                       POS<=pos2,
                       POS>=pos1) %>%
  arrange(P) %>%
  select(ID,`#CHROM`,POS,P)
head(eurLu) 
#                    ID #CHROM      POS          P
# 1: chr18:10691652:T:C     18 10691652 0.00239982
# 2: chr18:10692939:T:C     18 10692939 0.00348710
# 3: chr18:10648101:T:C     18 10648101 0.00735820

##########################################################################################
# alv 02
lat = fread("rms.alv.co.02.cvc.lat.PHENO1.glm.logistic.hybrid") %>%
  filter(P<5E-8) %>%
  select(ID, `#CHROM`, POS, P) %>%
  arrange(`#CHROM`,POS)
#                    ID #CHROM       POS           P
# 1:  chr2:62273850:G:T      2  62273850 4.21436e-08
# 2: chr6:153265235:A:G      6 153265235 3.64737e-08
# 3:  chr9:68554620:G:A      9  68554620 1.95614e-08
# 4:  chr9:68556109:C:T      9  68556109 1.95614e-08
# 5:  chr9:68599121:A:G      9  68599121 2.47232e-08
# 6:  chr9:68603905:A:T      9  68603905 6.09870e-09
# 7: chr15:99581506:T:G     15  99581506 1.55932e-10
# 8: chr16:82819052:C:T     16  82819052 2.40978e-08
# 9: chr17:37747101:G:A     17  37747101 4.04756e-08

eur = fread("/ccls/home/sli/GWAS/libby/rmsdbGAP/rms.alv.co.02.cvc.eur.PHENO1.glm.logistic.hybrid")

## (1) chr2:62273850
chr=2
pos1 = 62273850 - 50000
pos2 = 62273850 + 50000
eurLu = eur %>% filter(`#CHROM`==chr,
                       POS<=pos2,
                       POS>=pos1) %>%
  arrange(P) %>%
  select(ID,`#CHROM`,POS,P)
head(eurLu) 
#                    ID #CHROM       POS         P
# 1: chr6:153290854:C:G      6 153290854 0.0246814
# 2: chr6:153254016:C:T      6 153254016 0.0272389
# 3: chr6:153294049:G:A      6 153294049 0.0279726

## (2) chr6:153265235
chr=6
pos1 = 153265235 - 50000
pos2 = 153265235 + 50000
eurLu = eur %>% filter(`#CHROM`==chr,
                       POS<=pos2,
                       POS>=pos1) %>%
  arrange(P) %>%
  select(ID,`#CHROM`,POS,P)
head(eurLu) 
#                    ID #CHROM       POS         P
# 1: chr6:153290854:C:G      6 153290854 0.0246814
# 2: chr6:153254016:C:T      6 153254016 0.0272389
# 3: chr6:153294049:G:A      6 153294049 0.0279726

## (3) chr9:68554620 - chr9:68603905
chr=9
pos1 = 68554620 - 50000
pos2 = 68603905 + 50000
eurLu = eur %>% filter(`#CHROM`==chr,
                       POS<=pos2,
                       POS>=pos1) %>%
  arrange(P) %>%
  select(ID,`#CHROM`,POS,P)
head(eurLu) 
#                   ID #CHROM      POS          P
# 1: chr9:68532767:G:A      9 68532767 0.00227737
# 2: chr9:68516671:G:A      9 68516671 0.00371553
# 3: chr9:68614208:C:T      9 68614208 0.01481450

## (4) chr15:99581506
chr=15
pos1 = 99581506 - 50000
pos2 = 99581506 + 50000
eurLu = eur %>% filter(`#CHROM`==chr,
                       POS<=pos2,
                       POS>=pos1) %>%
  arrange(P) %>%
  select(ID,`#CHROM`,POS,P)
head(eurLu) 
#                    ID #CHROM      POS          P
# 1: chr15:99618239:T:C     15 99618239 0.00757446
# 2: chr15:99556179:T:C     15 99556179 0.00803949
# 3: chr15:99605019:C:A     15 99605019 0.00942698

## (4) chr16:82819052
chr=16
pos1 = 82819052 - 50000
pos2 = 82819052 + 50000
eurLu = eur %>% filter(`#CHROM`==chr,
                       POS<=pos2,
                       POS>=pos1) %>%
  arrange(P) %>%
  select(ID,`#CHROM`,POS,P)
head(eurLu) 
#                    ID #CHROM      POS          P
# 1: chr16:82855954:C:A     16 82855954 0.00463728
# 2: chr16:82856452:G:A     16 82856452 0.00494452
# 3: chr16:82856281:C:T     16 82856281 0.02083160

## (5) chr17:37747101
chr=17
pos1 = 37747101 - 50000
pos2 = 37747101 + 50000
eurLu = eur %>% filter(`#CHROM`==chr,
                       POS<=pos2,
                       POS>=pos1) %>%
  arrange(P) %>%
  select(ID,`#CHROM`,POS,P)
head(eurLu) 
#                    ID #CHROM      POS          P
# 1: chr17:37718929:A:C     17 37718929 0.00157581
# 2: chr17:37707301:G:A     17 37707301 0.00239383
# 3: chr17:37709840:C:T     17 37709840 0.00301829

```

```{bash meta of eur and lat}

cd /ccls/home/sli/GWAS/libby/rmsLat

# cut off 10-3
metal
MARKER ID
EFFECT BETA
PVALUE P
SCHEME STDERR
STDERR SE
ALLELELABELS REF ALT1
VERBOSE OFF
GENOMICCONTROL OFF
PROCESS /ccls/home/sli/GWAS/libby/rmsdbGAP/rms.co.cvc.eur.PHENO1.glm.logistic.hybrid
PROCESS rms.co.cvc.lat.PHENO1.glm.logistic.hybrid
ANALYZE HETEROGENEITY
QUIT
mv METAANALYSIS1.TBL eur.lat.meta.rms.co.glm.TBL

grep -v "\?" eur.lat.meta.rms.co.glm.TBL > flt.eur.lat.meta.rms.co.glm.TBL

metal
MARKER ID
EFFECT BETA
PVALUE P
SCHEME STDERR
STDERR SE
ALLELELABELS REF ALT1
VERBOSE OFF
GENOMICCONTROL OFF
PROCESS /ccls/home/sli/GWAS/libby/rmsdbGAP/rms.em.co.cvc.eur.PHENO1.glm.logistic.hybrid
PROCESS rms.em.co.cvc.lat.PHENO1.glm.logistic.hybrid
ANALYZE HETEROGENEITY
QUIT
mv METAANALYSIS1.TBL eur.lat.em.meta.rms.co.glm.TBL

grep -v "\?" eur.lat.em.meta.rms.co.glm.TBL > flt.eur.lat.em.meta.rms.co.glm.TBL

metal
MARKER ID
EFFECT BETA
PVALUE P
SCHEME STDERR
STDERR SE
ALLELELABELS REF ALT1
VERBOSE OFF
GENOMICCONTROL OFF
PROCESS /ccls/home/sli/GWAS/libby/rmsdbGAP/rms.alv.co.cvc.eur.PHENO1.glm.logistic.hybrid
PROCESS rms.alv.co.cvc.lat.PHENO1.glm.logistic.hybrid
ANALYZE HETEROGENEITY
QUIT
mv METAANALYSIS1.TBL eur.lat.alv.meta.rms.co.glm.TBL

grep -v "\?" eur.lat.alv.meta.rms.co.glm.TBL > flt.eur.lat.alv.meta.rms.co.glm.TBL

awk '$6<5E-8 {print}' flt.eur.lat.meta.rms.co.glm.TBL > eur.lat.meta.rms.co.sigs
awk '$6<5E-8 {print}' flt.eur.lat.em.meta.rms.co.glm.TBL > eur.lat.em.meta.rms.co.sigs
awk '$6<5E-8 {print}' flt.eur.lat.alv.meta.rms.co.glm.TBL > eur.lat.alv.meta.rms.co.sigs


# cut off 10-2
metal
MARKER ID
EFFECT BETA
PVALUE P
SCHEME STDERR
STDERR SE
ALLELELABELS REF ALT1
VERBOSE OFF
GENOMICCONTROL OFF
PROCESS /ccls/home/sli/GWAS/libby/rmsdbGAP/rms.co.02.cvc.eur.PHENO1.glm.logistic.hybrid
PROCESS rms.co.02.cvc.lat.PHENO1.glm.logistic.hybrid
ANALYZE HETEROGENEITY
QUIT
mv METAANALYSIS1.TBL eur.lat.meta.rms.02.co.glm.TBL

grep -v "\?" eur.lat.meta.rms.02.co.glm.TBL > flt.eur.lat.meta.rms.02.co.glm.TBL

metal
MARKER ID
EFFECT BETA
PVALUE P
SCHEME STDERR
STDERR SE
ALLELELABELS REF ALT1
VERBOSE OFF
GENOMICCONTROL OFF
PROCESS /ccls/home/sli/GWAS/libby/rmsdbGAP/rms.em.co.02.cvc.eur.PHENO1.glm.logistic.hybrid
PROCESS rms.em.co.02.cvc.lat.PHENO1.glm.logistic.hybrid
ANALYZE HETEROGENEITY
QUIT
mv METAANALYSIS1.TBL eur.lat.em.meta.rms.02.co.glm.TBL

grep -v "\?" eur.lat.em.meta.rms.02.co.glm.TBL > flt.eur.lat.em.meta.rms.02.co.glm.TBL

metal
MARKER ID
EFFECT BETA
PVALUE P
SCHEME STDERR
STDERR SE
ALLELELABELS REF ALT1
VERBOSE OFF
GENOMICCONTROL OFF
PROCESS /ccls/home/sli/GWAS/libby/rmsdbGAP/rms.alv.co.02.cvc.eur.PHENO1.glm.logistic.hybrid
PROCESS rms.alv.co.02.cvc.lat.PHENO1.glm.logistic.hybrid
ANALYZE HETEROGENEITY
QUIT
mv METAANALYSIS1.TBL eur.lat.alv.meta.rms.02.co.glm.TBL

grep -v "\?" eur.lat.alv.meta.rms.02.co.glm.TBL > flt.eur.lat.alv.meta.rms.02.co.glm.TBL

awk '$6<5E-8 {print}' flt.eur.lat.meta.rms.02.co.glm.TBL > eur.lat.meta.rms.02.co.sigs
awk '$6<5E-8 {print}' flt.eur.lat.em.meta.rms.02.co.glm.TBL > eur.lat.em.meta.rms.02.co.sigs
awk '$6<5E-8 {print}' flt.eur.lat.alv.meta.rms.02.co.glm.TBL > eur.lat.alv.meta.rms.02.co.sigs

```

```{r visualization}

setwd("/ccls/home/sli/GWAS/libby/rmsLat")

library(tidyverse)
library(QCEWAS)
library(qqman)
library(data.table)

files =c("flt.eur.lat.meta.rms.co.glm.TBL",
         "flt.eur.lat.em.meta.rms.co.glm.TBL",
         "flt.eur.lat.alv.meta.rms.co.glm.TBL",
         "flt.eur.lat.meta.rms.02.co.glm.TBL",
         "flt.eur.lat.em.meta.rms.02.co.glm.TBL",
         "flt.eur.lat.alv.meta.rms.02.co.glm.TBL")

for(i in files){

  filename <- gsub("^.*/","",i)
  gwas_data <- fread(i) %>% na.omit()

  if(length(grep("meta",i)) !=0 ){
    gwas_data_1 <- gwas_data %>%
      dplyr::mutate(ID=MarkerName, P=`P-value`)%>%
      separate(MarkerName, c("CHROM","POS","REF","ALT"))%>%
      dplyr::mutate(CHROM = as.numeric(gsub("chr","",as.character(CHROM))),POS = as.numeric(as.character(POS)))
  } else{
    gwas_data_1 <- gwas_data %>%
      dplyr::mutate(P= as.numeric(as.character(P)),CHROM = as.numeric(as.character(`#CHROM`)), POS = as.numeric(as.character(POS)))%>%
      arrange(P)
  }

  gwas_data_1 <- gwas_data_1[is.finite(gwas_data_1$P),  ]

  lambda <- P_lambda(gwas_data_1$P)
  print(paste("lambda value of",filename,"is:",lambda))

  png(paste0(filename,".manhattan.png"), width = 10, height = 7, units = 'in', res = 300)
  manhattan(gwas_data_1, chr = "CHROM", bp = "POS", p = "P", snp = "ID")
  dev.off()

  png(paste0(filename,".qq.png"), width = 10, height = 7, units = 'in', res = 300)
  qq(gwas_data_1$P)
  dev.off()
}

# [1] "lambda value of flt.eur.lat.meta.rms.co.glm.TBL is: 1.03448415408069"
# [1] "lambda value of flt.eur.lat.em.meta.rms.co.glm.TBL is: 1.02637704375394"
# [1] "lambda value of flt.eur.lat.alv.meta.rms.co.glm.TBL is: 1.01124305408092"


```

Target analysis for libby

```{bash}
cd /ccls/home/sli/GWAS/libby/rmsLat

cp /ccls/data_ccrlp_rms_gwas/IGF/GeneBoundaries_10kb.txt .

```

```{r extract and look up}

setwd("/ccls/home/sli/GWAS/libby/rmsLat")
library(tidyverse)
library(data.table)

a = fread("GeneBoundaries_10kb_hg38.txt") %>%
  as.data.frame()

## latino results
for(race in c("eur","lat")){

  if(race=="eur"){
    al.f = fread("~/GWAS/libby/rmsdbGAP/rms.co.02.cvc.eur.PHENO1.glm.logistic.hybrid") %>%
      dplyr::mutate(CHROM = as.numeric(`#CHROM`)) %>%
      dplyr::select(ID,CHROM,POS,REF,ALT1,BETA,SE,P)
    
    al.em.f = fread("~/GWAS/libby/rmsdbGAP/rms.em.co.02.cvc.eur.PHENO1.glm.logistic.hybrid")%>%
      dplyr::mutate(CHROM = as.numeric(`#CHROM`)) %>%
      dplyr::select(ID,CHROM,POS,REF,ALT1,BETA,SE,P)
    
    al.al.f = fread("~/GWAS/libby/rmsdbGAP/rms.alv.co.02.cvc.eur.PHENO1.glm.logistic.hybrid")%>%
      dplyr::mutate(CHROM = as.numeric(`#CHROM`)) %>%
      dplyr::select(ID,CHROM,POS,REF,ALT1,BETA,SE,P)}
  
  if(race=="lat"){
    al.f = fread("rms.co.02.cvc.lat.PHENO1.glm.logistic.hybrid") %>%
      dplyr::mutate(CHROM = as.numeric(`#CHROM`)) %>%
      dplyr::select(ID,CHROM,POS,REF,ALT1,BETA,SE,P)
    
    al.em.f = fread("rms.em.co.02.cvc.lat.PHENO1.glm.logistic.hybrid")%>%
      dplyr::mutate(CHROM = as.numeric(`#CHROM`)) %>%
      dplyr::select(ID,CHROM,POS,REF,ALT1,BETA,SE,P)
    
    al.al.f = fread("rms.alv.co.02.cvc.lat.PHENO1.glm.logistic.hybrid")%>%
      dplyr::mutate(CHROM = as.numeric(`#CHROM`)) %>%
      dplyr::select(ID,CHROM,POS,REF,ALT1,BETA,SE,P)}  
  
  for(i in 1:nrow(a)){
  
    chr= gsub("chr","",a[i,"V1"])
    start = a[i,"V2"]
    end = a[i,"V3"]
    gene = a[i,"V4"]
    
    ## in all cases
    al.f.select = al.f %>%
      dplyr::filter(CHROM==chr,
                    POS >= start,
                    POS <= end) %>%
      dplyr::mutate(gene = gene)
    write.csv(al.f.select,paste0("targeted/",race,".",gene,".all.lookup.txt"))
  
    sig.lev = -log10(0.05/nrow(al.f.select))
    al.f.select %>%
      ggplot(aes(x=POS,y=-log10(P)))+
      geom_point(size=1, alpha=0.5)+
      theme_bw()+
      ggtitle(paste0("Targeted look up of ",gene," \n meta-analysis"))+
      xlab(paste0("Chromosome ",chr, " positions"))+
      geom_hline(yintercept = sig.lev, colour="red",alpha=0.5)+
      scale_x_continuous(labels = function(x) format(x, scientific = FALSE))+
      theme(plot.title = element_text(hjust = 0.5))
    ggsave(paste0("targeted/",race,".",gene,".all.look.up.png"),width = 6,height = 4) 
     
    ## in em
    al.f.em.select = al.em.f %>%
      dplyr::filter(CHROM==chr,
                    POS >= start,
                    POS <= end) %>%
      dplyr::mutate(gene = gene)
    write.csv(al.f.em.select,paste0("targeted/",race,".",gene,".em.lookup.txt"))
  
    sig.lev = -log10(0.05/nrow(al.f.em.select))
    al.f.em.select %>%
      ggplot(aes(x=POS,y=-log10(P)))+
      geom_point(size=1, alpha=0.5)+
      theme_bw()+
      ggtitle(paste0("Targeted look up of ",gene," \n meta-analysis"))+
      xlab(paste0("Chromosome ",chr, " positions"))+
      geom_hline(yintercept = sig.lev, colour="red",alpha=0.5)+
      scale_x_continuous(labels = function(x) format(x, scientific = FALSE))+
      theme(plot.title = element_text(hjust = 0.5))
    ggsave(paste0("targeted/",race,".",gene,".em.look.up.png"),width = 6,height = 4) 
  
    ## in alv
    al.f.al.select = al.al.f %>%
      dplyr::filter(CHROM==chr,
                    POS >= start,
                    POS <= end) %>%
      dplyr::mutate(gene = gene)
    write.csv(al.f.al.select,paste0("targeted/",race,".",gene,".em.lookup.txt"))
  
    sig.lev = -log10(0.05/nrow(al.f.al.select))
    al.f.al.select %>%
      ggplot(aes(x=POS,y=-log10(P)))+
      geom_point(size=1, alpha=0.5)+
      theme_bw()+
      ggtitle(paste0("Targeted look up of ",gene," \n meta-analysis"))+
      xlab(paste0("Chromosome ",chr, " positions"))+
      geom_hline(yintercept = sig.lev, colour="red",alpha=0.5)+
      scale_x_continuous(labels = function(x) format(x, scientific = FALSE))+
      theme(plot.title = element_text(hjust = 0.5))
    ggsave(paste0("targeted/",race,".",gene,".al.look.up.png"),width = 6,height = 4) 
    
}}


```

the number of tests

```{r}
setwd("/ccls/home/sli/GWAS/libby/rmsLat")
library(tidyverse)
library(data.table)

a = fread("GeneBoundaries_10kb_hg38.txt") %>%
  as.data.frame()

lat.all = fread("/ccls/home/sli/GWAS/libby/rmsLat/rms.co.02.cvc.lat.05.bim")
lat.lk.ls = list()
for(i in 1:nrow(a)){
  chr= gsub("chr","",a[i,"V1"])
  start = a[i,"V2"]
  end = a[i,"V3"]
  lat.slc = lat.all %>%
    filter(V1==chr,
           V4 >= start,
           V4 <= end)
  lat.lk.ls[[i]] <- lat.slc$V2
  }

lat.lk.df = data.frame(V1=unlist(lat.lk.ls))
write.table(lat.lk.df,"lat.lk.df",
            sep="\t",
            quote=FALSE,
            row.names = FALSE,
            col.names = FALSE)


eur.all = fread("/ccls/home/sli/GWAS/libby/rmsdbGAP/rms.co.02.cvc.eur.05.bim")
eur.lk.ls = list()
for(i in 1:nrow(a)){
  chr= gsub("chr","",a[i,"V1"])
  start = a[i,"V2"]
  end = a[i,"V3"]
  eur.slc = eur.all %>%
    filter(V1==chr,
           V4 >= start,
           V4 <= end)
  eur.lk.ls[[i]] <- eur.slc$V2
  }

eur.lk.df = data.frame(V1=unlist(eur.lk.ls))
write.table(eur.lk.df,"eur.lk.df",
            sep="\t",
            quote=FALSE,
            row.names = FALSE,
            col.names = FALSE)


```

```{bash}
# the number of tests
cd /ccls/home/sli/GWAS/libby/rmsLat

## eur
plink --bfile /ccls/home/sli/GWAS/libby/rmsdbGAP/rms.co.02.cvc.eur.05\
  --extract eur.lk.df\
  --indep-pairwise 50 10 0.1\
  --out eur.lk.prd
# 3540 of 4153 variants removed
# 613

## lat
plink --bfile /ccls/home/sli/GWAS/libby/rmsLat/rms.co.02.cvc.lat.05\
  --extract lat.lk.df\
  --indep-pairwise 50 10 0.1\
  --out lat.lk.prd
# 4046 of 4693 variants removed
# 647

# ~630
```

