---
output: html_document
editor_options: 
  chunk_output_type: console
---

Project description: Finding somatic variants from childhood leukemia tumor samples, with matched normal tissues available.


```{bash GATK pipeline on wgs sequencing data}

# raw files are saved at 

cd /project/desmith_128/sebastian/smoking_all_wgs_raw/raw_data_fastq

# 38 germline and 35 tumors

ls | sed -E 's/TM_|GM_//g' | uniq

```

```{bash fastqc}
#!/bin/bash
#SBATCH --time=48:00:00
#SBATCH --mail-type=ALL
#SBATCH --array=1-98
#SBATCH --output=/scratch1/lishaobo/smoking_temp/codes/output/fastqc_%A_%a.out
#SBATCH --error=/scratch1/lishaobo/smoking_temp/codes/error/fastqc_%A_%a.errcd
#SBATCH --job-name=fastqc
#SBATCH -c 20
#SBATCH --ntasks=1
#SBATCH --mem=60G
#SBATCH --partition=main

cd /scratch1/lishaobo/smoking_temp
#sbatch /scratch1/lishaobo/smoking_temp/codes/fastqc.sh
module load gcc/11.3.0 fastqc/0.11.9

#fastq1=(`ls /project/desmith_128/sebastian/smoking_all_wgs_raw/raw_data_fastq/*/*_1.fq.gz | sed -n ${SLURM_ARRAY_TASK_ID}p`)
fastq1=(`ls /project/desmith_128/sebastian/smoking_all_wgs_raw/raw_data_fastq/*/*_1.fq.gz | sed -n ${SLURM_ARRAY_TASK_ID}p`)

fastq2=(`echo $fastq1 | sed 's/_1.fq.gz/_2.fq.gz/g'`)

fastqc --noextract $fastq1 $fastq2 \
  --outdir /scratch1/lishaobo/smoking_temp/fastqc\
  -t 20
  
```

paired-fastq-to-unmapped-bam

```{bash fastq2ubam}

#!/bin/bash
#SBATCH --time=48:00:00
#SBATCH --mail-type=ALL
#SBATCH --array=1-37
#SBATCH --output=/scratch1/lishaobo/smoking_temp/codes/output/fastq2ubam_%A_%a.out
#SBATCH --error=/scratch1/lishaobo/smoking_temp/codes/error/fastq2ubam_%A_%a.errcd
#SBATCH --job-name=fastq2ubam
#SBATCH -c 15
#SBATCH --ntasks=1
#SBATCH --mem=60G
#SBATCH --partition=main

set -e
set -x
set -u
set -o pipefail

cd /scratch1/lishaobo/smoking_temp
#sbatch /scratch1/lishaobo/smoking_temp/codes/fastq2ubam.sh

#fastq1=(`ls /project/desmith_128/sebastian/smoking_all_wgs_raw/raw_data_fastq/*/*_1.fq.gz | sed -n ${SLURM_ARRAY_TASK_ID}p`)
fastq1=(`ls /project/desmith_128/sebastian/smoking_all_wgs_raw/raw_data_fastq/*/*_1.fq.gz | sed -n ${SLURM_ARRAY_TASK_ID}p`)

fastq2=(`echo $fastq1 | sed 's/_1.fq.gz/_2.fq.gz/g'`)

base=$(basename ${fastq1} _1.fq.gz)
sample_name=(`echo $base | cut -d '_' -f1-2`)
library_name=(`echo $base | cut -d '_' -f3`)
flowcell=(`echo $base | cut -d '_' -f4`)
lane=${base: -1}
readgroup_name="$sample_name.$flowcell"."$lane"
platform_unit="$flowcell"."$lane"
platform_name=ILLUMINA

output_unmapped_bam="$base.unmapped.unsorted.bam"
out_dir="/scratch2/lishaobo/smoking_temp/unmapped.bam"

singularity exec \
  --bind /project,/scratch2,/scratch1 \
  /project/wiemels_260/sebastian/dependencies/gatk/gatk_latest.sif \
  gatk FastqToSam \
    --java-options "-Xmx85G"\
    --FASTQ $fastq1 \
    --FASTQ2 $fastq2 \
    --OUTPUT $out_dir/$output_unmapped_bam\
    --READ_GROUP_NAME $readgroup_name \
    --SAMPLE_NAME $sample_name \
    --LIBRARY_NAME $library_name \
    --PLATFORM_UNIT $platform_unit \
    --PLATFORM $platform_name\
    --TMP_DIR /scratch2/lishaobo/temp
    
```

```{bash sort unmapped bam}

#!/bin/bash
#SBATCH --time=48:00:00
#SBATCH --mail-type=ALL
#SBATCH --array=1-98
#SBATCH --output=/scratch1/lishaobo/smoking_temp/codes/output/sortubam_%A_%a.out
#SBATCH --error=/scratch1/lishaobo/smoking_temp/codes/error/sortubam_%A_%a.errcd
#SBATCH --job-name=sortubam
#SBATCH -c 15
#SBATCH --ntasks=1
#SBATCH --mem=89G
#SBATCH --partition=main

set -e
set -x
set -u
set -o pipefail

# remember to generate fille list before sending out script!
#!/bin/bash
#SBATCH --time=1:00:00
#SBATCH --job-name=middle
#SBATCH --mem=8G
#sbatch --dependency=afterok:10862838 /scratch1/lishaobo/smoking_temp/codes/middle1.sh
#cd /scratch1/lishaobo/smoking_temp
#ls /scratch2/lishaobo/smoking_temp/unmapped.bam/*.unmapped.unsorted.bam | wc -l
#ls /scratch2/lishaobo/smoking_temp/unmapped.bam/*.unmapped.unsorted.bam > /scratch1/lishaobo/smoking_temp/all.unmapped.unsorted.files.txt

cd /scratch1/lishaobo/smoking_temp
#sbatch --dependency=afterok:10864019 /scratch1/lishaobo/smoking_temp/codes/sortubam.sh

input_bam=(`cat /scratch1/lishaobo/smoking_temp/all.unmapped.unsorted.files.txt | sed -n ${SLURM_ARRAY_TASK_ID}p`)
base=$(basename ${input_bam} .unmapped.unsorted.bam)

sorted_bam_name="$base.unmapped.sorted.bam"
out_dir="/scratch2/lishaobo/smoking_temp/unmapped.bam"

singularity exec \
  --bind /project,/scratch2,/scratch1 \
  /project/wiemels_260/sebastian/dependencies/gatk/gatk_latest.sif \
  gatk SortSam \
    --java-options "-Xmx55G"\
    --INPUT $input_bam \
    --OUTPUT $out_dir/$sorted_bam_name\
    --SORT_ORDER queryname \
    --TMP_DIR /scratch2/lishaobo/temp

# to save space!
rm -rf $input_bam   

```

bwa alignment (mapping)

```{bash bwa germline samples}
#!/bin/bash
#SBATCH --time=48:00:00
#SBATCH --mail-type=ALL
#SBATCH --array=8,18-47
#SBATCH --output=/scratch1/lishaobo/smoking_temp/codes/output/bwa_gm_%A_%a.out
#SBATCH --error=/scratch1/lishaobo/smoking_temp/codes/error/bwa_gm_%A_%a.errcd
#SBATCH --job-name=bwa_gm
#SBATCH -c 20
#SBATCH --ntasks=1
#SBATCH --mem=89G
#SBATCH --partition=main

set -e
set -x
set -u
set -o pipefail

cd /scratch1/lishaobo/smoking_temp
#sbatch /scratch1/lishaobo/smoking_temp/codes/bwa_gm.sh

module load gcc/11.3.0 bwa/0.7.17

ref=/project/desmith_488/sebastian/gatk/refs/GATK_resource_bundle/resources_broad_hg38_v0_Homo_sapiens_assembly38.fasta    

#fastq1=(`ls /project/desmith_128/sebastian/smoking_all_wgs_raw/raw_data_fastq/*/GM*_1.fq.gz | sed -n ${SLURM_ARRAY_TASK_ID}p`)
fastq1=(`ls /project/desmith_128/sebastian/smoking_all_wgs_raw/raw_data_fastq/*/GM*_1.fq.gz | sed -n ${SLURM_ARRAY_TASK_ID}p`)

fastq2=(`echo $fastq1 | sed 's/_1.fq.gz/_2.fq.gz/g'`)

base=$(basename ${fastq1} _1.fq.gz)
sample_name=(`echo $base | cut -d '_' -f1-2`)
output_aligned_bam="$base.aligned.unsorted.bam"
out_dir="/scratch1/lishaobo/smoking_temp/aligned.bam"

bwa mem $ref \
  -t 20 \
  $fastq1 $fastq2 \
  | samtools view -S -b \
   > $out_dir/$output_aligned_bam

```

```{bash bwa tumor samples}

#!/bin/bash
#SBATCH --time=48:00:00
#SBATCH --mail-type=ALL
#SBATCH --array=1-51
#SBATCH --output=/scratch1/lishaobo/smoking_temp/codes/output/bwa_tm_%A_%a.out
#SBATCH --error=/scratch1/lishaobo/smoking_temp/codes/error/bwa_tm_%A_%a.errcd
#SBATCH --job-name=bwa_tm
#SBATCH -c 20
#SBATCH --ntasks=1
#SBATCH --mem=89G
#SBATCH --partition=main

set -e
set -x
set -u
set -o pipefail

cd /scratch1/lishaobo/smoking_temp
#sbatch /scratch1/lishaobo/smoking_temp/codes/bwa_tm.sh

module load gcc/11.3.0 bwa/0.7.17

ref=/project/desmith_488/sebastian/gatk/refs/GATK_resource_bundle/resources_broad_hg38_v0_Homo_sapiens_assembly38.fasta    

fastq1=(`ls /project/desmith_128/sebastian/smoking_all_wgs_raw/raw_data_fastq/*/TM*_1.fq.gz | sed -n ${SLURM_ARRAY_TASK_ID}p`)
#fastq1=(`ls /project/desmith_128/sebastian/smoking_all_wgs_raw/raw_data_fastq/*/TM*_1.fq.gz | sed -n ${SLURM_ARRAY_TASK_ID}p`)
fastq2=(`echo $fastq1 | sed 's/_1.fq.gz/_2.fq.gz/g'`)

base=$(basename ${fastq1} _1.fq.gz)
sample_name=(`echo $base | cut -d '_' -f1-2`)
output_aligned_bam="$base.aligned.unsorted.bam"
out_dir="/scratch1/lishaobo/smoking_temp/aligned.bam"

bwa mem $ref \
  -t 20 \
  $fastq1 $fastq2 \
  | samtools view -S -b \
   > $out_dir/$output_aligned_bam

```

```{bash sort aligned bam}

#!/bin/bash
#SBATCH --time=48:00:00
#SBATCH --mail-type=ALL
#SBATCH --array=1-98
#SBATCH --output=/scratch1/lishaobo/smoking_temp/codes/output/sortabam_%A_%a.out
#SBATCH --error=/scratch1/lishaobo/smoking_temp/codes/error/sortabam_%A_%a.errcd
#SBATCH --job-name=sortabam
#SBATCH -c 15
#SBATCH --ntasks=1
#SBATCH --mem=89G
#SBATCH --partition=main

set -e
set -x
set -u
set -o pipefail

# remember to generate fille list before sending out script!
#ls /scratch1/lishaobo/smoking_temp/aligned.bam/*.aligned.unsorted.bam | wc -l
#ls /scratch1/lishaobo/smoking_temp/aligned.bam/*.aligned.unsorted.bam > /scratch1/lishaobo/smoking_temp/all.aligned.unsorted.files.txt

cd /scratch1/lishaobo/smoking_temp
#sbatch /scratch1/lishaobo/smoking_temp/codes/sortabam.sh

input_bam=(`cat /scratch1/lishaobo/smoking_temp/all.aligned.unsorted.files.txt | sed -n ${SLURM_ARRAY_TASK_ID}p`)
base=$(basename ${input_bam} .aligned.unsorted.bam)

out_dir="/scratch1/lishaobo/smoking_temp/aligned.bam"
sorted_bam_name="$base.aligned.sorted.bam"

singularity exec \
  --bind /project,/scratch2,/scratch1 \
  /project/wiemels_260/sebastian/dependencies/gatk/gatk_latest.sif \
  gatk SortSam \
    --java-options "-Xmx85G"\
    --INPUT $input_bam \
    --OUTPUT $out_dir/$sorted_bam_name\
    --SORT_ORDER queryname \
    --TMP_DIR /scratch2/lishaobo/temp

# to save space!
rm -rf $input_bam


# GM_552232_CSFP210004989-1a_HJMVJDSX2_L1.aligned.sorted.bam is too small, find out why
# -rw-rw---- 1 lishaobo lishaobo 971M Sep  7 12:44 GM_552232_CSFP210004989-1a_HJMVJDSX2_L1.aligned.sorted.bam

#!/bin/bash
#SBATCH --time=24:00:00
#SBATCH --mail-type=ALL
#SBATCH --output=/scratch1/lishaobo/smoking_temp/codes/output/debug.out
#SBATCH --error=/scratch1/lishaobo/smoking_temp/codes/error/debug.errcd
#SBATCH --job-name=debug
#SBATCH --ntasks=1
#SBATCH --mem=16G
#SBATCH --partition=main
set -e
set -x
set -u
set -o pipefail
cd /scratch1/lishaobo/smoking_temp
#sbatch /scratch1/lishaobo/smoking_temp/codes/debug.sh
module load gcc/11.3.0 bwa/0.7.17
ref=/project/desmith_488/sebastian/gatk/refs/GATK_resource_bundle/resources_broad_hg38_v0_Homo_sapiens_assembly38.fasta
fastq1=/project/desmith_128/sebastian/smoking_all_wgs_raw/raw_data_fastq/GM_552232/GM_552232_CSFP210004989-1a_HJMVJDSX2_L1_1.fq.gz
fastq2=(`echo $fastq1 | sed 's/_1.fq.gz/_2.fq.gz/g'`)
base=$(basename ${fastq1} _1.fq.gz)
sample_name=(`echo $base | cut -d '_' -f1-2`)
output_aligned_bam="$base.aligned.unsorted.debug.bam"
out_dir="/scratch1/lishaobo/smoking_temp/aligned.bam"
bwa mem $ref \
  -t 20 \
  $fastq1 $fastq2 \
  | samtools view -S -b \
   > $out_dir/$output_aligned_bam
input_bam=/scratch1/lishaobo/smoking_temp/aligned.bam/GM_552232_CSFP210004989-1a_HJMVJDSX2_L1.aligned.unsorted.debug.bam
base=$(basename ${input_bam} .aligned.bam)
sorted_bam_name="$base.aligned.sorted.debug.bam"
singularity exec \
  --bind /project,/scratch2,/scratch1 \
  /project/wiemels_260/sebastian/dependencies/gatk/gatk_latest.sif \
  gatk SortSam \
    --java-options "-Xmx85G"\
    --INPUT $input_bam \
    --OUTPUT $out_dir/$sorted_bam_name\
    --SORT_ORDER queryname \
    --TMP_DIR /scratch2/lishaobo/temp
#everything looks OK!    
rm -rf $input_bam
rm -rf $out_dir/$sorted_bam_name
```

MergeBam

```{bash}

#!/bin/bash
#SBATCH --time=48:00:00
#SBATCH --mail-type=ALL
#SBATCH --array=1-98
#SBATCH --output=/scratch1/lishaobo/smoking_temp/codes/output/mergebam_%A_%a.out
#SBATCH --error=/scratch1/lishaobo/smoking_temp/codes/error/mergebam_%A_%a.errcd
#SBATCH --job-name=mergebam
#SBATCH -c 15
#SBATCH --ntasks=1
#SBATCH --mem=89G
#SBATCH --partition=main

set -e
set -x
set -u
set -o pipefail

# remember to generate file list before sending out script!
#!/bin/bash
#SBATCH --time=1:00:00
#SBATCH --job-name=middle
#SBATCH --mem=8G
#sbatch --dependency=afterok:10862838 /scratch1/lishaobo/smoking_temp/codes/middle1.sh
#cd /scratch1/lishaobo/smoking_temp
#ls /scratch2/lishaobo/smoking_temp/unmapped.bam/*.unmapped.sorted.bam | wc -l
#ls /scratch2/lishaobo/smoking_temp/unmapped.bam/*.unmapped.sorted.bam > /scratch1/lishaobo/smoking_temp/all.unmapped.sorted.files.txt

cd /scratch1/lishaobo/smoking_temp
#sbatch /scratch1/lishaobo/smoking_temp/codes/mergebam.sh

ref=/project/desmith_488/sebastian/gatk/refs/GATK_resource_bundle/resources_broad_hg38_v0_Homo_sapiens_assembly38.fasta    

unmapped_bam=(`cat /scratch1/lishaobo/smoking_temp/all.unmapped.sorted.files.txt | sed -n ${SLURM_ARRAY_TASK_ID}p`)
base=$(basename ${unmapped_bam} .unmapped.sorted.bam)

aligned_bam_name="$base.aligned.sorted.bam"
aligned_bam="/scratch1/lishaobo/smoking_temp/aligned.bam/${aligned_bam_name}"
output_bam="/scratch2/lishaobo/smoking_temp/mergebam/${base}.bam"

singularity exec \
  --bind /project,/scratch2,/scratch1 \
  /project/wiemels_260/sebastian/dependencies/gatk/gatk_latest.sif \
  gatk MergeBamAlignment \
  --java-options "-Xmx85G"\
  --VALIDATION_STRINGENCY SILENT \
  --EXPECTED_ORIENTATIONS FR \
  --ATTRIBUTES_TO_RETAIN X0 \
  --ALIGNED_BAM $aligned_bam \
  --UNMAPPED_BAM $unmapped_bam\
  --OUTPUT $output_bam \
  --REFERENCE_SEQUENCE $ref \
  --PAIRED_RUN true \
  --SORT_ORDER "unsorted" \
  --IS_BISULFITE_SEQUENCE false \
  --ALIGNED_READS_ONLY false \
  --CLIP_ADAPTERS false \
  --MAX_RECORDS_IN_RAM 2000000 \
  --ADD_MATE_CIGAR true \
  --MAX_INSERTIONS_OR_DELETIONS -1 \
  --PRIMARY_ALIGNMENT_STRATEGY MostDistant \
  --PROGRAM_RECORD_ID "bwamem" \
  --PROGRAM_GROUP_VERSION "0.7.12-r1039" \
  --PROGRAM_GROUP_COMMAND_LINE "bwa mem $ref -t 20" \
  --PROGRAM_GROUP_NAME "bwamem" \
  --UNMAPPED_READ_STRATEGY COPY_TO_TAG \
  --ALIGNER_PROPER_PAIR_FLAGS true \
  --UNMAP_CONTAMINANT_READS true \
  --TMP_DIR /scratch2/lishaobo/temp
  
# to save space!
rm -rf $aligned_bam 
rm -rf $unmapped_bam
  
```

```{bash validate merged bam}

#!/bin/bash
#SBATCH --time=48:00:00
#SBATCH --mail-type=ALL
#SBATCH --array=1-98
#SBATCH --output=/scratch1/lishaobo/smoking_temp/codes/output/validatembam_%A_%a.out
#SBATCH --error=/scratch1/lishaobo/smoking_temp/codes/error/validatembam_%A_%a.errcd
#SBATCH --job-name=validatembam
#SBATCH -c 20
#SBATCH --ntasks=1
#SBATCH --mem=50G
#SBATCH --partition=main

set -e
set -x
set -u
set -o pipefail

cd /scratch1/lishaobo/smoking_temp
#sbatch /scratch1/lishaobo/smoking_temp/codes/validatembam.sh

input_bam=(`ls /scratch2/lishaobo/smoking_temp/mergebam/*.bam | sed -n ${SLURM_ARRAY_TASK_ID}p`)

singularity exec \
  --bind /project,/scratch2,/scratch1 \
  /project/wiemels_260/sebastian/dependencies/gatk/gatk_latest.sif \
  gatk ValidateSamFile \
  --java-options "-Xmx250G"\
  --INPUT $input_bam \
  --MODE SUMMARY \
  --TMP_DIR /scratch2/lishaobo/temp
  
```

Mark Duplicates

```{bash MarkDuplicates}

#!/bin/bash
#SBATCH --time=48:00:00
#SBATCH --mail-type=ALL
#SBATCH --array=1-73
#SBATCH --output=/scratch1/lishaobo/smoking_temp/codes/output/MarkDuplicates_%A_%a.out
#SBATCH --error=/scratch1/lishaobo/smoking_temp/codes/error/MarkDuplicates_%A_%a.errcd
#SBATCH --job-name=MarkDuplicates
#SBATCH -c 20
#SBATCH --ntasks=1
#SBATCH --mem=120G
#SBATCH --partition=main

set -e
set -x
set -u
set -o pipefail

cd /scratch1/lishaobo/smoking_temp
#sbatch --dependency=afterok:10948124 /scratch1/lishaobo/smoking_temp/codes/MarkDuplicates.sh

sample_name=(`ls /project/desmith_128/sebastian/smoking_all_wgs_raw/raw_data_fastq |sed -n ${SLURM_ARRAY_TASK_ID}p`)

input_bam=`ls /scratch2/lishaobo/smoking_temp/mergebam/${sample_name}*.bam`
input_bam_prefix=`printf " --INPUT %s" $input_bam`
prefix=" --INPUT "
input_bam_list=${input_bam_prefix#"$prefix"}

output_bam="$sample_name.aligned.duplicates_marked.bam"
metrics_filename="$sample_name.duplicate_metrics"
out_dir="/scratch1/lishaobo/smoking_temp/MarkDuplicates"

singularity exec \
  --bind /project,/scratch1,/scratch2 \
  /project/wiemels_260/sebastian/dependencies/gatk/gatk_latest.sif \
  gatk MarkDuplicates \
  --java-options "-Xmx100G"\
  --INPUT $input_bam_list \
  --OUTPUT $out_dir/$output_bam\
  --METRICS_FILE $out_dir/$metrics_filename \
  --VALIDATION_STRINGENCY SILENT \
  --OPTICAL_DUPLICATE_PIXEL_DISTANCE 2500 \
  --ASSUME_SORT_ORDER "queryname" \
  --CREATE_MD5_FILE true \
  --TMP_DIR /scratch2/lishaobo/temp

```

SortAndFixTags

```{bash}

#!/bin/bash
#SBATCH --time=48:00:00
#SBATCH --mail-type=ALL
#SBATCH --array=1-73
#SBATCH --output=/scratch1/lishaobo/smoking_temp/codes/output/SortAndFixTags_%A_%a.out
#SBATCH --error=/scratch1/lishaobo/smoking_temp/codes/error/SortAndFixTags_%A_%a.errcd
#SBATCH --job-name=SortAndFixTags
#SBATCH -c 20
#SBATCH --mem=246G
#SBATCH --partition epyc-64
#SBATCH --ntasks=1

set -e
set -x
set -u
set -o pipefail

cd /scratch1/lishaobo/smoking_temp
#sbatch /scratch1/lishaobo/smoking_temp/codes/SortAndFixTags.sh

# remember to generate file list before sending out script!
#ls /scratch1/lishaobo/smoking_temp/MarkDuplicates/*.bam | wc -l
#ls /scratch1/lishaobo/smoking_temp/MarkDuplicates/*.bam  > /scratch1/lishaobo/smoking_temp/all.merged.files.txt

ref=/project/desmith_488/sebastian/gatk/refs/GATK_resource_bundle/resources_broad_hg38_v0_Homo_sapiens_assembly38.fasta    
mergedbam=(`cat /scratch1/lishaobo/smoking_temp/all.merged.files.txt | sed -n ${SLURM_ARRAY_TASK_ID}p`)
output_bam_basename=$(basename ${mergedbam} .aligned.duplicates_marked.bam)
out_dir="/scratch1/lishaobo/smoking_temp/MarkDuplicates"
out_bam="${out_dir}/${output_bam_basename}.marked.sorted.bam"

singularity exec \
  --bind /project,/scratch1,/scratch2 \
  /project/wiemels_260/sebastian/dependencies/gatk/gatk_latest.sif \
  gatk SortSam  \
  --java-options "-Xms230G"\
  --INPUT $mergedbam \
  --OUTPUT /dev/stdout \
  --SORT_ORDER "coordinate" \
  --CREATE_INDEX false \
  --CREATE_MD5_FILE false \
  --TMP_DIR /scratch2/lishaobo/temp \
    | \
singularity exec \
  --bind /project,/scratch1,/scratch2 \
  /project/wiemels_260/sebastian/dependencies/gatk/gatk_latest.sif \
  gatk SetNmMdAndUqTags \
  --java-options "-Xms230G"\
  --INPUT /dev/stdin \
  --OUTPUT $out_bam\
  --CREATE_INDEX true \
  --CREATE_MD5_FILE true \
  --REFERENCE_SEQUENCE $ref\
  --TMP_DIR /scratch2/lishaobo/temp

rm -rf $mergedbam

```

BaseRecalibrator

```{bash BaseRecalibrator}

#!/bin/bash
#SBATCH --time=48:00:00
#SBATCH --mail-type=ALL
#SBATCH --array=1-73
#SBATCH --output=/scratch1/lishaobo/smoking_temp/codes/output/BaseRecalibrator_%A_%a.out
#SBATCH --error=/scratch1/lishaobo/smoking_temp/codes/error/BaseRecalibrator_%A_%a.errcd
#SBATCH --job-name=BaseRecalibrator
#SBATCH -c 20
#SBATCH --mem=110G
#SBATCH --ntasks=1
#SBATCH --partition=main

set -e
set -x
set -u
set -o pipefail

cd /scratch1/lishaobo/smoking_temp
#sbatch /scratch1/lishaobo/smoking_temp/codes/BaseRecalibrator.sh

# remember to generate fille list before sending out script!
ls /scratch1/lishaobo/smoking_temp/MarkDuplicates/*.marked.sorted.bam | wc -l
ls /scratch1/lishaobo/smoking_temp/MarkDuplicates/*.marked.sorted.bam  > /scratch1/lishaobo/smoking_temp/all.marked.sorted.files.txt

ref=/project/desmith_488/sebastian/gatk/refs/GATK_resource_bundle/resources_broad_hg38_v0_Homo_sapiens_assembly38.fasta    
known_sites_snp=/project/desmith_488/sebastian/gatk/refs/GATK_resource_bundle/Homo_sapiens_assembly38.dbsnp138.vcf
known_sites_indel=/project/desmith_488/sebastian/gatk/refs/GATK_resource_bundle/resources_broad_hg38_v0_Homo_sapiens_assembly38.known_indels.vcf.gz
interval_list=/project/desmith_488/sebastian/gatk/refs/GATK_resource_bundle/resources_broad_hg38_v0_wgs_calling_regions.hg38.interval_list

input_bam=(`cat /scratch1/lishaobo/smoking_temp/all.marked.sorted.files.txt | sed -n ${SLURM_ARRAY_TASK_ID}p`)
output_bam_basename=$(basename ${input_bam} .marked.sorted.bam)
output_dir="/scratch1/lishaobo/smoking_temp/recalibration_report"

recalibration_report_output="${output_dir}/${output_bam_basename}.recal_data.csv"

output_dir_bam="/scratch2/lishaobo/smoking_temp/MarkDuplicates_re"
output_bam_name="$output_bam_basename.recalibrated.bam"
output_bam="${output_dir_bam}/${output_bam_name}"

singularity exec \
  --bind /project,/scratch1,/scratch2 \
  /project/wiemels_260/sebastian/dependencies/gatk/gatk_latest.sif \
  gatk BaseRecalibrator  \
  --java-options "-Xmx100G -Djava.io.tmpdir=/scratch2/lishaobo" \
  -R $ref \
  -I $input_bam \
  --use-original-qualities \
  -O $recalibration_report_output\
  --known-sites $known_sites_snp \
  --known-sites $known_sites_indel \
  -L $interval_list 

singularity exec \
  --bind /project,/scratch1,/scratch2 \
  /project/wiemels_260/sebastian/dependencies/gatk/gatk_latest.sif \
  gatk ApplyBQSR  \
  --java-options "-Xmx100G -Djava.io.tmpdir=/scratch2/lishaobo" \
  -R $ref \
  -I $input_bam \
  -O $output_bam \
  -L $interval_list \
  -bqsr $recalibration_report_output\
  --static-quantized-quals 10 \
  --static-quantized-quals 20 \
  --static-quantized-quals 30 \
  --add-output-sam-program-record \
  --create-output-bam-md5 \
  --use-original-qualities
    
```

```{bash AnalyzeCovariates}

#!/bin/bash
#SBATCH --time=48:00:00
#SBATCH --mail-type=ALL
#SBATCH --array=1-73
#SBATCH --output=/scratch1/lishaobo/smoking_temp/codes/output/AnalyzeCovariates_%A_%a.out
#SBATCH --error=/scratch1/lishaobo/smoking_temp/codes/error/AnalyzeCovariates_%A_%a.errcd
#SBATCH --job-name=AnalyzeCovariates
#SBATCH -c 20
#SBATCH --ntasks=1
#SBATCH --mem=80G
#SBATCH --partition=main

set -e
set -x
set -u
set -o pipefail

cd /scratch1/lishaobo/smoking_temp
#sbatch --dependency=afterok:xxxxxx /scratch1/lishaobo/smoking_temp/codes/AnalyzeCovariates.sh

ref=/project/desmith_488/sebastian/gatk/refs/GATK_resource_bundle/resources_broad_hg38_v0_Homo_sapiens_assembly38.fasta    
known_sites_snp=/project/desmith_488/sebastian/gatk/refs/GATK_resource_bundle/Homo_sapiens_assembly38.dbsnp138.vcf
known_sites_indel=/project/desmith_488/sebastian/gatk/refs/GATK_resource_bundle/resources_broad_hg38_v0_Homo_sapiens_assembly38.known_indels.vcf.gz
interval_list=/project/desmith_488/sebastian/gatk/refs/GATK_resource_bundle/resources_broad_hg38_v0_wgs_calling_regions.hg38.interval_list

input_bam=(`ls /scratch1/lishaobo/smoking_temp/all.marked.sorted.files.txt | sed -n ${SLURM_ARRAY_TASK_ID}p`)
output_bam_basename=$(basename ${input_bam} .marked.sorted.bam)

output_dir="/scratch1/lishaobo/smoking_temp/recalibration_report"
recalibration_report_pre="${output_dir}/${output_bam_basename}.recal_data.csv"
recalibration_report_post="${output_dir}/${output_bam_basename}.recal_data.post.csv"

output_dir_bam="/scratch2/lishaobo/smoking_temp/MarkDuplicates_re"
output_bam_name="${output_bam_basename}.recalibrated.bam"
output_bam="${output_dir_bam}/${output_bam_name}"

outcsv="${output_dir}/${output_bam_basename}.AnalyzeCovariates.csv"
outpdf="${output_dir}/${output_bam_basename}.AnalyzeCovariates.pdf"

singularity exec \
  --bind /project,/scratch1,/scratch2 \
  /project/wiemels_260/sebastian/dependencies/gatk/gatk_latest.sif \
  gatk BaseRecalibrator  \
  --java-options "-Xmx100G -Djava.io.tmpdir=/scratch2/lishaobo" \
  -R $ref \
  -I $output_bam \
  -O $recalibration_report_post \
  --known-sites $known_sites_snp \
  --known-sites $known_sites_indel \
  -L $interval_list 
  
singularity exec \
  --bind /project,/scratch1,/scratch2 \
  /project/wiemels_260/sebastian/dependencies/gatk/gatk_latest.sif \
  gatk AnalyzeCovariates  \
  --java-options "-Xmx100G -Djava.io.tmpdir=/scratch2/lishaobo" \
  --before $recalibration_report_pre \
  --after $recalibration_report_post \
  --csv $outcsv\
  --plots $outpdf

```

Somatic variant calling

```{bash scatter-gather}

#!/bin/bash
#SBATCH --time=48:00:00
#SBATCH --mail-type=ALL
#SBATCH --output=/scratch1/lishaobo/smoking_temp/codes/output/Mutect2_%A_%a.out
#SBATCH --error=/scratch1/lishaobo/smoking_temp/codes/error/Mutect2_%A_%a.errcd
#SBATCH --job-name=Mutect2
#SBATCH -c 20
#SBATCH --ntasks=1
#SBATCH --mem=120G
#SBATCH --partition main
#SBATCH --array=1-38

set -e
set -x
set -u
set -o pipefail

cd /scratch1/lishaobo/smoking_temp

ref=/project/desmith_488/sebastian/gatk/refs/GATK_resource_bundle/resources_broad_hg38_v0_Homo_sapiens_assembly38.fasta    
interval_list=/project/desmith_488/sebastian/gatk/refs/GATK_resource_bundle/resources_broad_hg38_v0_wgs_calling_regions.hg38.interval_list
gnomad=/project/desmith_488/kerenxu/somatic_hg38/af-only-gnomad.hg38.vcf.gz
pon=/project/desmith_488/kerenxu/somatic_hg38/1000g_pon.hg38.vcf.gz

input_dir_bam="/scratch2/lishaobo/smoking_temp/MarkDuplicates_re"

normal_bam=(`ls /scratch2/lishaobo/smoking_temp/MarkDuplicates_re/G*.recalibrated.bam | sed -n ${SLURM_ARRAY_TASK_ID}p`)
normal_bam_file=$input_dir_bam/$normal_bam

base=$(basename ${normal_bam} .recalibrated.bam)
prefix="GM_"
sample_name=${base#"$prefix"}
tumor_bam="TM_$sample_name.recalibrated.bam"
tumor_bam_file=$input_dir_bam/$tumor_bam

output_dir_bam="/scratch2/lishaobo/smoking_temp/somaticCalling"
output_vcf="${output_dir_bam}/sample${sample_name}.vcf.gz"
output_bam="${output_dir_bam}/sample${sample_name}.tumor_normal.bam"
output_tar_gz="${output_dir_bam}/sample${sample_name}.f1r2.tar.gz"

singularity exec \
  --bind /project,/scratch1,/scratch2 \
  /project/wiemels_260/sebastian/dependencies/gatk/gatk_latest.sif \
  --java-options "-Xmx110G" \
  Mutect2 \
  -R $ref \
  -I $tumor_bam_file \
  -I $normal_bam_file \
  -normal $base \
  --germline-resource $gnomad \
  -pon $pon \
  -L $interval_list \
  -O $output_vcf\
  --bam-output $output_bam \
  --f1r2-tar-gz $output_tar_gz \
  --tmp-dir /scratch2/lishaobo/temp

```

```{bash contamination}

#!/bin/bash
#SBATCH --time=48:00:00
#SBATCH --mail-type=END
#SBATCH --output=/scratch1/lishaobo/smoking_temp/codes/output/contamination_%A_%a.out
#SBATCH --error=/scratch1/lishaobo/smoking_temp/codes/error/contamination_%A_%a.errcd
#SBATCH --job-name=contamination
#SBATCH -c 20
#SBATCH --ntasks=1
#SBATCH --mem=120G
#SBATCH --partition main
#SBATCH --array=1-38

set -e
set -x
set -u
set -o pipefail


cd /scratch1/lishaobo/smoking_temp

ref=/project/desmith_488/sebastian/gatk/refs/GATK_resource_bundle/resources_broad_hg38_v0_Homo_sapiens_assembly38.fasta    
interval_list=/project/desmith_488/sebastian/gatk/refs/GATK_resource_bundle/resources_broad_hg38_v0_wgs_calling_regions.hg38.interval_list
gnomad=/project/desmith_488/kerenxu/somatic_hg38/af-only-gnomad.hg38.vcf.gz
pon=/project/desmith_488/kerenxu/somatic_hg38/1000g_pon.hg38.vcf.gz
variants_for_contamination=/project/desmith_488/kerenxu/somatic_hg38/small_exac_common_3.hg38.vcf.gz

normal_bam=(`ls /scratch2/lishaobo/smoking_temp/MarkDuplicates_re/G*.recalibrated.bam | sed -n ${SLURM_ARRAY_TASK_ID}p`)
normal_bam_file=$normal_bam

base=$(basename ${normal_bam} .recalibrated.bam)
prefix="GM_"
sample_name=${base#"$prefix"}
tumor_bam="TM_$sample_name.recalibrated.bam"
tumor_bam_file=$input_dir_bam/$tumor_bam

output_dir_bam="/scratch2/lishaobo/smoking_temp/somaticCalling"
output_vcf="${output_dir_bam}/sample${sample_name}.vcf.gz"
output_bam="${output_dir_bam}/sample${sample_name}.tumor_normal.bam"
output_tar_gz="${output_dir_bam}/sample${sample_name}.f1r2.tar.gz"

normal_pileups="${output_dir_bam}/${base}.pileups.table"
tumor_pileups="${output_dir_bam}/T${sample_name}.pileups.table"
contamination_table="${output_dir_bam}/sample${sample_name}.contamination.table"
segments_table="${output_dir_bam}/sample${sample_name}.segments.table"

singularity exec \
  --bind /project,/scratch1,/scratch2 \
  /project/wiemels_260/sebastian/dependencies/gatk/gatk_latest.sif \
  --java-options "-Xmx110G" \
  GetPileupSummaries \
  -R $ref \
  -I $normal_bam_file \
  -V $variants_for_contamination \
  -L $variants_for_contamination \
  -O $normal_pileups \
  --tmp-dir /scratch2/lishaobo/temp

singularity exec \
  --bind /project,/scratch1,/scratch2 \
  /project/wiemels_260/sebastian/dependencies/gatk/gatk_latest.sif \
  --java-options "-Xmx110G" \
  GetPileupSummaries \
  -R $ref \
  -I $tumor_bam_file \
  -V $variants_for_contamination \
  -L $variants_for_contamination \
  -O $tumor_pileups \
  --tmp-dir /scratch2/lishaobo/temp

singularity exec \
  --bind /project,/scratch1,/scratch2 \
  /project/wiemels_260/sebastian/dependencies/gatk/gatk_latest.sif \
  --java-options "-Xmx110G" \
  CalculateContamination \
  -I $tumor_pileups \
  -O $contamination_table \
  --tumor-segmentation $segments_table \
  -matched $normal_pileups \
  --tmp-dir /scratch2/lishaobo/temp
    
```

```{bash OrientationBias}

#!/bin/bash
#SBATCH --time=48:00:00
#SBATCH --mail-type=END
#SBATCH --output=/scratch1/lishaobo/smoking_temp/codes/output/OrientationBias_%A_%a.out
#SBATCH --error=/scratch1/lishaobo/smoking_temp/codes/error/OrientationBias_%A_%a.errcd
#SBATCH --job-name=OrientationBias
#SBATCH -c 20
#SBATCH --ntasks=1
#SBATCH --mem=120G
#SBATCH --partition main
#SBATCH --array=1-38

set -e
set -x
set -u
set -o pipefail

cd /scratch1/lishaobo/smoking_temp

input_dir_bam="/scratch2/lishaobo/smoking_temp/somaticCalling"

input_f1r2_tar_gz=(`ls $input_dir_bam/*.f1r2.tar.gz | sed -n ${SLURM_ARRAY_TASK_ID}p`)
base=$(basename ${input_f1r2_tar_gz} .f1r2.tar.gz)
output_f1r2_tar_gz="$input_dir_bam/$base.artifact-priors.tar.gz"

singularity exec \
  --bind /project,/scratch1,/scratch2 \
  /project/wiemels_260/sebastian/dependencies/gatk/gatk_latest.sif \
  --java-options "-Xmx110G" \
  LearnReadOrientationModel \
  -I $input_f1r2_tar_gz \
  -O $output_f1r2_tar_gz \
  --tmp-dir /scratch2/lishaobo/temp

```

```{bash FilterMutectCalls}

#!/bin/bash
#SBATCH --time=48:00:00
#SBATCH --mail-type=END
#SBATCH --output=/scratch1/lishaobo/smoking_temp/codes/output/FilterMutectCalls_%A_%a.out
#SBATCH --error=/scratch1/lishaobo/smoking_temp/codes/error/FilterMutectCalls_%A_%a.errcd
#SBATCH --job-name=FilterMutectCalls
#SBATCH -c 20
#SBATCH --ntasks=1
#SBATCH --mem=120G
#SBATCH --partition main
#SBATCH --array=1-38

set -e
set -x
set -u
set -o pipefail

cd /scratch1/lishaobo/smoking_temp

ref=/project/desmith_488/sebastian/gatk/refs/GATK_resource_bundle/resources_broad_hg38_v0_Homo_sapiens_assembly38.fasta    


in_dir_bam="/scratch2/lishaobo/smoking_temp/somaticCalling"

unfiltered_vcf=(`ls $in_dir_bam/*.vcf.gz | sed -n ${SLURM_ARRAY_TASK_ID}p`)
base=$(basename ${unfiltered_vcf} .vcf.gz)
prefix="sample"
sample_name=${base#"$prefix"}

contamination_table="${in_dir_bam}/${base}.contamination.table"
maf_segments="${in_dir_bam}/${base}.segments.table"
artifact_priors_tar_gz="${in_dir_bam}/${base}.artifact-priors.tar.gz"
mutect_stats="${in_dir_bam}/${base}.vcf.gz.stats"

out_dir_bam="/scratch2/lishaobo/smoking_temp/somaticCalling"
output_vcf="${out_dir_bam}/${base}.filtering.vcf.gz"
filtering_stats="${out_dir_bam}/$base.filtering.vcf.gz.stats"

singularity exec \
  --bind /project,/scratch1,/scratch2 \
  /project/wiemels_260/sebastian/dependencies/gatk/gatk_latest.sif \
  --java-options "-Xmx110G" \
  FilterMutectCalls \
  -V $unfiltered_vcf \
  -R $ref \
  -O $output_vcf \
  --contamination-table $contamination_table \
  --tumor-segmentation $maf_segments \
  --ob-priors $artifact_priors_tar_gz \
  -stats $mutect_stats \
  --filtering-stats $filtering_stats \
  --tmp-dir /scratch2/lishaobo/temp
            
```

```{bash bcffilter}

#!/bin/bash
#SBATCH --time=01:00:00
#SBATCH --mail-type=ALL
#SBATCH --output=/scratch1/lishaobo/smoking_temp/codes/output/bcffilter_%A_%a.out
#SBATCH --error=/scratch1/lishaobo/smoking_temp/codes/error/bcffilter_%A_%a.errcd
#SBATCH --job-name=bcffilter
#SBATCH -c 15
#SBATCH --ntasks=1
#SBATCH --mem=59G
#SBATCH --partition=main

set -e
set -x
set -u
set -o pipefail

cd /scratch2/lishaobo/smoking_temp/somaticCalling

for i in {1..38}; do

  input_vcf=(`ls *.filtering.vcf.gz | sed -n ${i}p`)
  base=$(basename ${input_vcf} .filtering.vcf.gz)
  prefix="sample"
  sample_name=${base#"$prefix"}
  normal_sample_name="GM_$sample_name"
  output_vcf=${base}.bcftools.filtered.tumoronly.vcf.gz
  
  bcftools view $input_vcf \
    -Ou -i 'FILTER=="PASS"' | bcftools view -Ou -i 'FMT/DP[0]>=10' |\
    bcftools view -Ou -i 'FMT/DP[1]>=14' | bcftools view -Ou -i '(FORMAT/AD[*:1]/FORMAT/DP)>=0.1' |\
      bcftools view -s ^$normal_sample_name -o ../BCFtools.filter.tumoronly/$output_vcf -O z --threads 15
done 

cd /scratch2/lishaobo/smoking_temp/BCFtools.filter.tumoronly

for i in {1..38}; do

  input_vcf=(`ls *.vcf.gz | sed -n ${i}p`)
  tabix ${input_vcf}
  output_vcf=$(basename ${input_vcf} .bcftools.filtered.tumoronly.vcf.gz).bcftools.filtered.tumoronly.chr1to22.vcf.gz
  
  bcftools view $input_vcf -Ou \
  --regions chr1,chr2,chr3,chr4,chr5,chr6,chr7,chr8,chr9,chr10,chr11,chr12,chr13,chr14,chr15,chr16,chr17,chr18,chr19,chr20,chr21,chr22 \
  -o ../BCFtools.filter.tumoronly.autosomal/$output_vcf -O z --threads 15
done

```


