---
title: "lmr"
output: html_document
editor_options: 
  chunk_output_type: console
---

Project description: GWAS on the associations between lymphocyte/monocyte ratio and SNPs

```{r create a master file for all subject's information}
library(tidyverse)
library(data.table)
library(RNOmni)
setwd("/ccls/home/sli/GWAS/LMR")

## Set 1 fam files and IDs

sub1 <- read.csv("/ccls/home/sli/sets/set1/clinical_variables.csv") %>%
  select(beadPosition,subjectId) 
phenotype1 <- read.csv("~/sets/set1/idol_deconvolution.csv")  %>%
  as.data.frame() %>%
  dplyr::mutate(LMR = (CD8T+CD4T+Bcell)/Mono) %>%
  dplyr::mutate(LMRint = RNOmni::rankNormal(LMR)) %>%
  select(beadPosition,LMRint) %>%
  inner_join(sub1, by = "beadPosition")  %>%
  select(-beadPosition)   

set1.eur = fread("/ccls/home/sli/GWAS/bcell/ccls.set1.eur.fam")%>%
  left_join(phenotype1,by=c("V2"="subjectId")) %>%
  select(V1, V2, V3, V4, V5, LMRint)

set1.lat = fread("/ccls/home/sli/GWAS/bcell/ccls.set1.lat.fam")%>%
  left_join(phenotype1,by=c("V2"="subjectId")) %>%
  select(V1, V2, V3, V4, V5, LMRint)

write.table(gen1.eur, "set1.eur.fam", sep='\t', 
            quote = FALSE, row.names = FALSE, col.names = FALSE)
write.table(gen1.lat, "set1.lat.fam", sep='\t', 
            quote = FALSE, row.names = FALSE, col.names = FALSE)

## Set 2 fam files and IDs

sub2 <- read.csv("/ccls/home/sli/sets/set2/clinical_variables.csv") %>%
  select(beadPosition,subjectId)
phenotype2 <- read.csv("~/sets/set2/idol_deconvolution.csv")  %>%
  as.data.frame() %>%
  dplyr::mutate(LMR = (CD8T+CD4T+Bcell)/Mono) %>%
  dplyr::mutate(LMRint = RNOmni::rankNormal(LMR)) %>%
  select(beadPosition,LMRint) %>%
  inner_join(sub2, by = "beadPosition")  %>%
  select(-beadPosition)   

set2.eur = fread("/ccls/home/sli/GWAS/bcell/ccls.set2.eur.fam")%>%
  left_join(phenotype2,by=c("V2"="subjectId")) %>%
  select(V1, V2, V3, V4, V5, LMRint)

set2.lat = fread("/ccls/home/sli/GWAS/bcell/ccls.set2.lat.fam")%>%
  left_join(phenotype2,by=c("V2"="subjectId")) %>%
  select(V1, V2, V3, V4, V5, LMRint)

write.table(set2.eur, "set2.eur.fam", sep='\t', 
            quote = FALSE, row.names = FALSE, col.names = FALSE)
write.table(set2.lat, "set2.lat.fam", sep='\t', 
            quote = FALSE, row.names = FALSE, col.names = FALSE)


## Set 3 fam files and IDs

sub3 <- read.csv("/ccls/home/sli/sets/set3/clinical_variables.csv") %>%
  filter(Trisomy21==0) %>%
  select(beadPosition,subjectId)
phenotype3 <- read.csv("~/sets/set3/idol_deconvolution.csv")  %>%
  as.data.frame() %>%
  dplyr::mutate(LMR = (CD8T+CD4T+Bcell)/Mono) %>%
  dplyr::mutate(LMRint = RNOmni::rankNormal(LMR)) %>%
  select(beadPosition,LMRint) %>%
  inner_join(sub3, by = "beadPosition")  %>%
  select(-beadPosition)   

set3.eur = fread("/ccls/home/sli/GWAS/bcell/ccls.set3.eur.fam")%>%
  left_join(phenotype3,by=c("V2"="subjectId")) %>%
  select(V1, V2, V3, V4, V5, LMRint)

set3.lat = fread("/ccls/home/sli/GWAS/bcell/ccls.set3.lat.fam")%>%
  left_join(phenotype3,by=c("V2"="subjectId")) %>%
  select(V1, V2, V3, V4, V5, LMRint)
  
write.table(gen3.eur, "set3.eur.fam", sep='\t', 
            quote = FALSE, row.names = FALSE, col.names = FALSE)
write.table(gen3.lat, "set3.lat.fam", sep='\t', 
            quote = FALSE, row.names = FALSE, col.names = FALSE)

## Set 4 fam files and IDs

sub4 <- read.csv("/ccls/home/sli/sets/set4/clinical_variables.csv") %>%
  filter(smp_type == "bg") %>%
  distinct(subjectId,.keep_all=TRUE) %>%
  select(beadPosition,subjectId)

phenotype4 <- read.csv("~/sets/set4/idol_deconvolution.csv")  %>%
  as.data.frame() %>%
  dplyr::mutate(LMR = (CD8T+CD4T+Bcell)/Mono) %>%
  dplyr::mutate(LMRint = RNOmni::rankNormal(LMR)) %>%
  select(beadPosition,LMRint) %>%
  inner_join(sub4, by = "beadPosition")  %>%
  select(-beadPosition)   

set4.eur = fread("/ccls/home/sli/GWAS/bcell/ccls.set4.eur.fam")%>%
  left_join(phenotype4,by=c("V2"="subjectId")) %>%
  select(V1, V2, V3, V4, V5, LMRint)

set4.lat = fread("/ccls/home/sli/GWAS/bcell/ccls.set4.lat.fam")%>%
  left_join(phenotype4,by=c("V2"="subjectId")) %>%
  select(V1, V2, V3, V4, V5, LMRint)


write.table(gen4.eur, "set4.eur.fam", sep='\t', 
            quote = FALSE, row.names = FALSE, col.names = FALSE)
write.table(gen4.lat, "set4.lat.fam", sep='\t', 
            quote = FALSE, row.names = FALSE, col.names = FALSE)

```

```{r visualizing of LMR distribution}

library(tidyverse)
library(data.table)

setwd("/ccls/home/sli/GWAS/LMR")
  
for (SET in c("set1","set2","set3","set4")){
  for (RACE in c("eur","lat")){

  bcell <- read.table(paste0(SET,".",RACE,".fam")) 
  
  bcell %>%
    select(V6) %>%
    na.omit() %>%
    ggplot(aes(x=V6)) +
    geom_density() +
    theme_classic() +
    ggtitle(paste0("Distribution of LMR in ", SET, " of " ,RACE))+
    xlab("LMR") +
    theme(plot.title = element_text(hjust = 0.5),
          axis.text.x = element_blank()) 
  ggsave(paste0("LMR.DST",SET,RACE,".png"),width = 5,height = 5)}
  
  }

```

```{bash LMR analysis}

cd /ccls/home/sli/GWAS/LMR

for set in "set1" "set2" "set3" "set4"; do
  for race in "eur" "lat"; do
    plink2 --bfile /ccls/home/sli/GWAS/bcell/ccls.$set.$race\
    --fam $set.$race.fam\
    --covar /ccls/home/sli/GWAS/bcell/temp/ccls."${set}"."${race}".pca.eigenvec\
    --threads 200\
    --glm hide-covar cols=chrom,pos,ref,alt1,nobs,a1countcc,a1freq,a1freqcc,beta,se,p\
    --out ccls."${set}"."${race}".LMR
  done
done

# step 10.1.2 Meta-analysis
cd /ccls/home/sli/GWAS/LMR

  metal
  MARKER ID
  EFFECT BETA
  PVALUE P
  SCHEME STDERR
  STDERR SE
  ALLELELABELS REF ALT1
  VERBOSE OFF
  GENOMICCONTROL OFF
  PROCESS ccls.set1.eur.LMR.PHENO1.glm.linear
  PROCESS ccls.set2.eur.LMR.PHENO1.glm.linear
  PROCESS ccls.set3.eur.LMR.PHENO1.glm.linear
  PROCESS ccls.set4.eur.LMR.PHENO1.glm.linear
  PROCESS ccls.set1.lat.LMR.PHENO1.glm.linear
  PROCESS ccls.set2.lat.LMR.PHENO1.glm.linear
  PROCESS ccls.set3.lat.LMR.PHENO1.glm.linear
  PROCESS ccls.set4.lat.LMR.PHENO1.glm.linear
  ANALYZE HETEROGENEITY
  QUIT
  
  mv METAANALYSIS1.TBL meta.ccls.LMR.TBL
  mv METAANALYSIS1.TBL.info ccls.LMR.TBL.info

```  

```{r visualization}
setwd("/ccls/home/sli/GWAS/LMR")

library(tidyverse)
library(QCEWAS)
library(qqman)
library(data.table)

files =c("ccls.set1.eur.LMR.PHENO1.glm.linear",
         "ccls.set2.eur.LMR.PHENO1.glm.linear",
         "ccls.set3.eur.LMR.PHENO1.glm.linear",
         "ccls.set4.eur.LMR.PHENO1.glm.linear",
         "ccls.set1.lat.LMR.PHENO1.glm.linear",
         "ccls.set2.lat.LMR.PHENO1.glm.linear",
         "ccls.set3.lat.LMR.PHENO1.glm.linear",
         "ccls.set4.lat.LMR.PHENO1.glm.linear",
         "meta.ccls.LMR.TBL")

for(i in files){
  
  filename <- gsub("^.*/","",i)
  gwas_data <- fread(i) %>% na.omit()
  
  if(length(grep("meta",i)) !=0 ){
    gwas_data_1 <- gwas_data %>%
      dplyr::mutate(ID=MarkerName, P=`P-value`)%>%
      separate(MarkerName, c("CHROM","POS","REF","ALT"))%>%
      dplyr::mutate(CHROM = as.numeric(gsub("chr","",as.character(CHROM))),POS = as.numeric(as.character(POS)))
  } else{
    gwas_data_1 <- gwas_data %>%
      dplyr::mutate(P= as.numeric(as.character(P)),CHROM = as.numeric(as.character(`#CHROM`)), POS = as.numeric(as.character(POS)))%>%
      arrange(P)
  }
  
  gwas_data_1 <- gwas_data_1[is.finite(gwas_data_1$P),  ]
  
  lambda <- P_lambda(gwas_data_1$P)
  print(paste("lambda value of",filename,"is:",lambda))
  
  png(paste0(filename,".manhattan.png"), width = 10, height = 7, units = 'in', res = 300)
  manhattan(gwas_data_1, chr = "CHROM", bp = "POS", p = "P", snp = "ID")
  dev.off()
  
  png(paste0(filename,".qq.png"), width = 10, height = 7, units = 'in', res = 300)
  qq(gwas_data_1$P)
  dev.off()
}

```

```{r loop up key SNPs}

setwd("/ccls/home/sli/GWAS/LMR")
library(tidyverse)
library(data.table)

var="LMR"

raw = fread(paste0("meta.ccls.",var,".TBL"))%>%
  dplyr::mutate(ID=MarkerName, P=`P-value`)%>%
  separate(MarkerName, c("CHROM","POS","REF","ALT")) 

raw1 = raw %>%
  mutate(loc = paste0(CHROM,":",POS))

lookup = fread("/ccls/home/sli/GWAS/bcell/rawfiles/lookup.meta.snps", header=FALSE) %>%
  inner_join(raw1, by=c("V1"="loc"))

lookup = lookup %>%
  arrange(P)

dic = fread("/ccls/home/sli/GWAS/bcell/rawfiles/meta.lookup.cellP.txt") 
colnames(dic) = c("trait","matID","chr","pos","gene")
dic1 = dic %>% 
  dplyr::select(matID,gene,trait)

metalu = lookup%>%
  select(ID,CHROM,POS,REF,ALT,Effect,P,Direction) %>%
  mutate(matID = paste0(CHROM,":",POS)) %>%
  inner_join(dic1, by="matID") %>%
  arrange(P)
  
write.csv(metalu, paste0(var,".meta.lookup.gene.csv"))

metalu = read.csv(paste0(var,".meta.lookup.gene.csv"))
mean(metalu$P<0.05)

```

