---
output: html_document
editor_options: 
  chunk_output_type: console
---

Project description: Rhabdomyosarcoma GWAS in Latino subjects, with external dbGAP controls to boost power

```{bash}

cd /ccls/home/sli/GWAS/libby/rmsdbGAP

# generated by lat_rms.Rmd, a file for all subjects
cp /ccls/home/sli/GWAS/libby/rmsLat/rms.ccls.sol.ahgwas.??? .
cp /ccls/home/sli/GWAS/libby/rmsLat/sol.c1.subs .
awk '{print $1,$2}' /ccls/home/sli/GWAS/libby/rmsLat/rms.gwas.ccls.co.fam > ccls.rms.co.subs
 
 
plink --bfile rms.ccls.sol.ahgwas\
  --remove sol.c1.subs\
  --make-bed\
  --out rms.ccls.sol.ahgwas.noc1

# 603096 variants and 20507 people pass filters and QC.
```

```{r choosing european subjects. CCLS based on NCI score; others based on imputation}

setwd("/ccls/home/sli/GWAS/libby/rmsdbGAP")
library(tidyverse)
library(data.table)

all.c1.c2.subs = fread("rms.ccls.sol.ahgwas.noc1.fam") %>% 
  mutate(FID=V1,IID=V2) %>%
  select(FID,IID)

ccls = fread("ccls.rms.co.subs",header=FALSE)
allSubs = fread("/ccls/home/sli/GWAS/libby/rmsLat/allsubsRaceCaCoAge.txt")%>%
  mutate(caco = caco+1)

eurSubs= fread("/ccls/home/sli/GWAS/libby/rmsLat/allsubsRaceCaCoAge.txt") %>%
  mutate(caco = caco+1) %>%
  filter(race_imp=="EUR_imp") %>%
  anti_join(ccls,c("FID"="V1","IID"="V2")) %>%
  mutate(cvc_caco=1)

snpWeight <- read.table("/ccls/home/sli/GWAS/libby/v3/rms.ccls.merged.snpweight", header=TRUE) %>%
  as.data.frame() %>%
  filter(CEU>0.8) %>%
  select(FID, IID)%>%
  inner_join(allSubs,by=c("FID","IID"))%>%
  mutate(cvc_caco=2)


allEurs = rbind(eurSubs,snpWeight) %>% 
  as.data.frame() %>%
  inner_join(all.c1.c2.subs,by=c("FID","IID"))
table(allEurs$caco)  

write.table(allEurs, "all.eur.info.txt", quote=FALSE,row.names = FALSE)

eur_fam = allEurs %>%
  mutate(V3=0,V4=0) %>%
  select(FID,IID,V3,V4,sex,caco)
write.table(eur_fam, "eur.rms.fam", col.names = FALSE, row.names = FALSE, quote=FALSE)

eur_id = eur_fam %>% select(FID,IID)
write.table(eur_id, "eur.ids.txt", col.names = FALSE, row.names = FALSE, quote=FALSE)

```

```{bash QC for GWAS}

cd /ccls/home/sli/GWAS/libby/rmsdbGAP

plink --bfile rms.ccls.sol.ahgwas.noc1\
  --keep eur.ids.txt\
  --indiv-sort f eur.ids.txt\
  --make-bed\
  --out rms.eur.0

mv rms.eur.0.fam rms.eur.0.fam.bkp
mv eur.rms.fam rms.eur.0.fam

# Missingness
plink --bfile rms.eur.0 --geno 0.2 --make-bed --out tmp/rms.eur.1
plink --bfile tmp/rms.eur.1 --mind 0.2 --make-bed --out tmp/rms.eur.2
plink --bfile tmp/rms.eur.2 --geno 0.05 --make-bed --out tmp/rms.eur.3
plink --bfile tmp/rms.eur.3 --mind 0.05 --make-bed --out tmp/rms.eur.4

# HW
plink2 --bfile tmp/rms.eur.4\
  --hwe 1e-4\
  --make-bed\
  --keep-if PHENO1 == control\
  --out tmp/rms.eur.4.1
# --hwe: 668 variants removed due to Hardy-Weinberg exact test (founders only).

plink --bfile tmp/rms.eur.4\
  --extract tmp/rms.eur.4.1.bim\
  --make-bed\
  --out tmp/rms.eur.5
# 588861 variants and 7303 people pass filters and QC.

# MAF 
plink --bfile tmp/rms.eur.5 --maf 0.01 --make-bed --out tmp/rms.eur.6
# 36624 variants removed due to minor allele threshold(s)

# Heterozygosity rate
plink --bfile tmp/rms.eur.6\
  --exclude /ccls/home/sli/GWAS/ref/codes/inversion_hg37.txt\
  --indep-pairwise 50 5 0.2\
  --out tmp/indepSNP
  
plink --bfile tmp/rms.eur.6\
  --extract tmp/indepSNP.prune.in\
  --het\
  --out R_check
  
# Plot of the heterozygosity rate distribution
Rscript --no-save ~/GWAS/ref/codes/check_heterozygosity_rate.R
Rscript --no-save ~/GWAS/ref/codes/heterozygosity_outliers_list.R
sed 's/"// g' fail-het-qc.txt | awk '{print$1, $2}'> het_fail_ind.txt

# Remove heterozygosity rate outliers.
plink --bfile tmp/rms.eur.6 --remove het_fail_ind.txt --make-bed\
  --out tmp/rms.eur.7
# 115 people removed
# 552237 variants and 7189 people pass filters and QC.

# relatedness
# Check for relationships between individuals with a pihat > 0.2.
plink --bfile tmp/rms.eur.7\
  --extract tmp/indepSNP.prune.in\
  --genome\
  --min 0.2\
  --out pihat_min0.2

grep -v "UN|PI_HAT" pihat_min0.2.genome > pihat_min0.2.related
awk 'NR!=1{print $1,$2}' pihat_min0.2.related > related.fam.txt
# Generate a plot to assess the type of relationship.
# Rscript --no-save ~/GWAS/ref/codes/Relatedness.R

plink --bfile tmp/rms.eur.7\
  -make-bed\
  --out rms.eur.8\
  --remove related.fam.txt


```

```{r file for control control GWAS}

setwd("/ccls/home/sli/GWAS/libby/rmsdbGAP")
library(tidyverse)
library(data.table)


allEurs = fread("all.eur.info.txt")
allSams = fread("rms.eur.8.fam") %>%
  left_join(allEurs,by=c("V1"="FID","V2"="IID")) %>%
  filter(caco==1) %>%
  select(V1,V2,V3,V4,V5,cvc_caco)


write.table(allSams, "eur.cvc.fam.txt", col.names = FALSE, row.names = FALSE, quote=FALSE)

```

```{bash cvc}

cd /ccls/home/sli/GWAS/libby/rmsdbGAP

plink --bfile rms.eur.8\
  --keep eur.cvc.fam.txt\
  --indiv-sort f eur.cvc.fam.txt\
  --make-bed\
  --out eur.cvc

mv eur.cvc.fam eur.cvc.fam.bkp
cp eur.cvc.fam.txt eur.cvc.fam

plink2 --bfile eur.cvc\
  --maf 0.1\
  --indep-pairwise 50 10 0.1\
  --out tmp/eur.cvc.prd

plink2 --bfile eur.cvc\
  --extract tmp/eur.cvc.prd.prune.in\
  --pca 10\
  --out eur.cvc.pca\
  --threads 200

plink2 --bfile eur.cvc\
  --covar eur.cvc.pca.eigenvec\
  --glm firth-fallback hide-covar cols=chrom,pos,ref,alt1,nobs,a1countcc,a1freq,a1freqcc,beta,se,p\
  --out eur.cvc\
  --threads 200

awk '$15<1e-3 {print $3}' eur.cvc.PHENO1.glm.logistic.hybrid > eur.cvc.drop.SNPs.03
awk '$15<1e-2 {print $3}' eur.cvc.PHENO1.glm.logistic.hybrid > eur.cvc.drop.SNPs.02


wc -l eur.cvc.drop.SNPs.03
# 685
wc -l eur.cvc.drop.SNPs.02
# 6299

#####################################################################
# setting the cvc cut off as 03

cd /ccls/home/sli/GWAS/libby/rmsdbGAP

plink --bfile rms.eur.8\
  --exclude eur.cvc.drop.SNPs.03\
  --recode vcf bgz\
  --keep-allele-order\
  --out topmed_pre/eur.rms.cvc03
  
bcftools sort topmed_pre/eur.rms.cvc03.vcf.gz -Oz -o topmed_pre/eur.rms.cvc03.sorted.vcf.gz
tabix topmed_pre/eur.rms.cvc03.sorted.vcf.gz

#split by chromomse
for chr in {1..22}; do
  bcftools view --regions "${chr}" topmed_pre/eur.rms.cvc03.sorted.vcf.gz\
  -o topmed_pre/eur.rms.cvc03.chr"${chr}".vcf.gz -Oz &
done


cd /ccls/home/sli/GWAS/libby/rmsdbGAP/topmed_post
for chr in {1..22}; do
  unzip -P 'yuDoBLq0D2edlZ' chr_"${chr}".zip
  echo chr_"${chr}".zip unzip success
done  


cd /ccls/home/sli/GWAS/libby/rmsdbGAP

for chr in {1..22};do
  awk '($7 >0.6) {print $1}' <(zcat topmed_post/chr"${chr}".info.gz) | sed '1d' > \
      topmed_post/include.06.chr"${chr}".txt
done

RACE="eur"
for chr in {1..22};do
  plink2 --vcf topmed_post/chr"${chr}".dose.vcf.gz\
    dosage=HDS \
    --snps-only 'just-acgt'\
    --extract topmed_post/include.06.chr"${chr}".txt\
    --maf 0.01\
    --make-bed\
    --out topmed_post/$RACE.chr"${chr}".iptd.cvc\
    --threads 200
done

find ./topmed_post -name "$RACE.chr*.iptd.cvc.bim" > ./topmed_post/merge.$RACE.cvc.list 
sed -i 's/.bim//g' ./topmed_post/merge.$RACE.cvc.list

plink --merge-list topmed_post/merge.$RACE.cvc.list\
  --indiv-sort 0\
  --make-bed\
  --out rms.co.cvc.$RACE

# control-control GWAS after imputation
cd /ccls/home/sli/GWAS/libby/rmsdbGAP
RACE="eur"
awk '{print 0,$1"_"$2,$3,$4,$5,$6}' $RACE.cvc.fam.txt > $RACE.cvc.topmed.fam.txt

plink2 --bfile rms.co.cvc.$RACE\
  --keep $RACE.cvc.topmed.fam.txt\
  --indiv-sort f $RACE.cvc.topmed.fam.txt\
  --make-bed\
  --out $RACE.cvc.postimp

mv $RACE.cvc.postimp.fam $RACE.cvc.postimp.fam.bkp
cp $RACE.cvc.topmed.fam.txt $RACE.cvc.postimp.fam
  
plink2 --bfile $RACE.cvc.postimp\
  --maf 0.1\
  --indep-pairwise 50 10 0.1\
  --out tmp/$RACE.cvc.postimp.prd

plink2 --bfile $RACE.cvc.postimp\
  --extract tmp/$RACE.cvc.postimp.prd.prune.in\
  --pca 10\
  --out $RACE.cvc.postimp.pca\
  --threads 200

plink2 --bfile $RACE.cvc.postimp\
  --covar $RACE.cvc.postimp.pca.eigenvec\
  --glm firth-fallback hide-covar cols=chrom,pos,ref,alt1,nobs,a1countcc,a1freq,a1freqcc,beta,se,p\
  --out $RACE.cvc.postimp\
  --threads 200

awk '$15<1e-5 {print $3}' $RACE.cvc.postimp.PHENO1.glm.logistic.hybrid > $RACE.cvc.postimp.drop.SNPs.05

plink2 --bfile rms.co.cvc.$RACE\
  --exclude $RACE.cvc.postimp.drop.SNPs.05\
  --make-bed\
  --out rms.co.cvc.$RACE.05
  
#####################################################################
# setting the cvc cut off as 02

cd /ccls/home/sli/GWAS/libby/rmsdbGAP

plink --bfile rms.eur.8\
  --exclude eur.cvc.drop.SNPs.02\
  --recode vcf bgz\
  --keep-allele-order\
  --out topmed_pre/eur.rms.cvc02
  
bcftools sort topmed_pre/eur.rms.cvc02.vcf.gz -Oz -o topmed_pre/eur.rms.cvc02.sorted.vcf.gz
tabix topmed_pre/eur.rms.cvc02.sorted.vcf.gz

#split by chromomse
for chr in {1..22}; do
  bcftools view --regions "${chr}" topmed_pre/eur.rms.cvc02.sorted.vcf.gz\
  -o topmed_pre/eur.rms.cvc02.chr"${chr}".vcf.gz -Oz &
done

cd /ccls/home/sli/GWAS/libby/rmsdbGAP/topmed_post/02
for chr in {1..22}; do
  unzip -P 'nMu4aTYMuXZtc2' chr_"${chr}".zip
  echo chr_"${chr}".zip unzip success
done  


cd /ccls/home/sli/GWAS/libby/rmsdbGAP

for chr in {1..22};do
  awk '($7 >0.6) {print $1}' <(zcat topmed_post/02/chr"${chr}".info.gz) | sed '1d' > \
      topmed_post/02/include.06.chr"${chr}".txt
done

RACE="eur"
for chr in {1..22};do
  plink2 --vcf topmed_post/02/chr"${chr}".dose.vcf.gz\
    dosage=HDS \
    --snps-only 'just-acgt'\
    --extract topmed_post/02/include.06.chr"${chr}".txt\
    --maf 0.01\
    --make-bed\
    --out topmed_post/02/$RACE.chr"${chr}".iptd.02.cvc\
    --threads 200
done

find ./topmed_post/02 -name "$RACE.chr*.iptd.02.cvc.bim" > ./topmed_post/02/merge.$RACE.02.cvc.list 
sed -i 's/.bim//g' ./topmed_post/02/merge.$RACE.02.cvc.list

plink --merge-list topmed_post/02/merge.$RACE.02.cvc.list\
  --indiv-sort 0\
  --make-bed\
  --out rms.co.02.cvc.$RACE

# control-control GWAS after imputation
cd /ccls/home/sli/GWAS/libby/rmsdbGAP
RACE="eur"
awk '{print 0,$1"_"$2,$3,$4,$5,$6}' $RACE.cvc.fam.txt > $RACE.cvc.topmed.fam.txt

plink2 --bfile rms.co.02.cvc.$RACE\
  --keep $RACE.cvc.topmed.fam.txt\
  --indiv-sort f $RACE.cvc.topmed.fam.txt\
  --make-bed\
  --out $RACE.02.cvc.postimp

mv $RACE.02.cvc.postimp.fam $RACE.02.cvc.postimp.fam.bkp
cp $RACE.cvc.topmed.fam.txt $RACE.02.cvc.postimp.fam
  
plink2 --bfile $RACE.02.cvc.postimp\
  --maf 0.1\
  --indep-pairwise 50 10 0.1\
  --out tmp/$RACE.02.cvc.postimp.prd

plink2 --bfile $RACE.02.cvc.postimp\
  --extract tmp/$RACE.02.cvc.postimp.prd.prune.in\
  --pca 10\
  --out $RACE.02.cvc.postimp.pca\
  --threads 200

plink2 --bfile $RACE.02.cvc.postimp\
  --covar $RACE.02.cvc.postimp.pca.eigenvec\
  --glm firth-fallback hide-covar cols=chrom,pos,ref,alt1,nobs,a1countcc,a1freq,a1freqcc,beta,se,p\
  --out $RACE.02.cvc.postimp\
  --threads 200

awk '$15<1e-5 {print $3}' $RACE.02.cvc.postimp.PHENO1.glm.logistic.hybrid > $RACE.02.cvc.postimp.drop.SNPs.05

plink2 --bfile rms.co.02.cvc.$RACE\
  --exclude $RACE.cvc.postimp.drop.SNPs.05\
  --make-bed\
  --out rms.co.02.cvc.$RACE.05
  


```

```{r creating $RACE.ahgwas.ccls.rms.cvc.fam}

library(tidyverse)
library(data.table)
setwd("/ccls/home/sli/GWAS/libby/rmsdbGAP")

allFam = fread("all.eur.info.txt") %>%
  mutate(IID=paste0(FID,"_",IID),
         V3=0,V4=0) %>%
  select(IID,V3,V4,sex,caco)
CaCoCvCFam = fread("topmed_post/eur.chr1.iptd.cvc.fam") %>%
  select(V1,V2) %>%
  left_join(allFam,by=c("V2"="IID"))

write.table(CaCoCvCFam, "rms.co.cvc.eur.fam.updt", 
            col.names = FALSE, 
            row.names = FALSE, 
            quote=FALSE)

```

```{bash GWAS trial run for RMS CACO with plink}

#####################################################################
# setting the cvc cut off as 03

cd /ccls/home/sli/GWAS/libby/rmsdbGAP
RACE="eur"

mv rms.co.cvc.$RACE.05.fam rms.co.cvc.$RACE.05.fam.bkp
cp rms.co.cvc.$RACE.fam.updt rms.co.cvc.$RACE.05.fam

# HWE among controls using keep if
plink2 --bfile rms.co.cvc.$RACE.05\
  --hwe midp 1e-4 \
  --keep-if PHENO1 == control\
  --make-pgen\
  --out tmp/rms.co.cvc."${RACE}"_1\
  --threads 200

# Make pgen again using remaining variants from pvar file 
plink2 --bfile rms.co.cvc.$RACE.05\
  --make-pgen\
  --extract tmp/rms.co.cvc."${RACE}"_1.pvar\
  --out tmp/rms.co.cvc."${RACE}"_2\
  --threads 200

# Make bed
plink2 --pfile tmp/rms.co.cvc."${RACE}"_2\
  --make-bed\
  --out rms.control.cvc."${RACE}"

#make bed for PCA analysis
# Create pruned SNPs for later use as controlling variables in regression
plink2 --bfile rms.control.cvc."${RACE}"\
  --make-bed \
  --out tmp/rms.co.cvc."${RACE}".pca

plink2 --bfile tmp/rms.co.cvc."${RACE}".pca\
  --maf 0.1\
  --indep-pairwise 50 10 0.1 \
  --out tmp/rms.co.cvc."${RACE}".pruned

plink2 --bfile rms.control.cvc."${RACE}"\
  --extract tmp/rms.co.cvc."${RACE}".pruned.prune.in\
  --pca 10\
  --out rms.control.cvc."${RACE}".pruned.pca\
  --threads 200

# Regression
plink2 --bfile rms.control.cvc."${RACE}"\
  --covar rms.control.cvc."${RACE}".pruned.pca.eigenvec\
  --threads 200\
  --glm firth-fallback hide-covar cols=chrom,pos,ref,alt1,nobs,a1countcc,a1freq,a1freqcc,beta,se,p,err\
  --out rms.co.cvc."${RACE}"

#####################################################################
# setting the cvc cut off as 02

cd /ccls/home/sli/GWAS/libby/rmsdbGAP
RACE="eur"

mv rms.co.02.cvc.$RACE.05.fam rms.co.02.cvc.$RACE.05.fam.bkp
cp rms.co.cvc.$RACE.fam.updt rms.co.02.cvc.$RACE.05.fam

# HWE among controls using keep if
plink2 --bfile rms.co.02.cvc.$RACE.05\
  --hwe midp 1e-4 \
  --keep-if PHENO1 == control\
  --make-pgen\
  --out tmp/rms.co.02.cvc."${RACE}"_1\
  --threads 200

# Make pgen again using remaining variants from pvar file 
plink2 --bfile rms.co.02.cvc.$RACE.05\
  --make-pgen\
  --extract tmp/rms.co.02.cvc."${RACE}"_1.pvar\
  --out tmp/rms.co.02.cvc."${RACE}"_2\
  --threads 200

# Make bed
plink2 --pfile tmp/rms.co.02.cvc."${RACE}"_2\
  --make-bed\
  --out rms.control.02.cvc."${RACE}"

#make bed for PCA analysis
# Create pruned SNPs for later use as controlling variables in regression
plink2 --bfile rms.control.02.cvc."${RACE}"\
  --make-bed \
  --out tmp/rms.co.02.cvc."${RACE}".pca

plink2 --bfile tmp/rms.co.02.cvc."${RACE}".pca\
  --maf 0.1\
  --indep-pairwise 50 10 0.1 \
  --out tmp/rms.co.02.cvc."${RACE}".pruned

plink2 --bfile rms.control.02.cvc."${RACE}"\
  --extract tmp/rms.co.02.cvc."${RACE}".pruned.prune.in\
  --pca 10\
  --out rms.control.02.cvc."${RACE}".pruned.pca\
  --threads 200

# Regression
plink2 --bfile rms.control.02.cvc."${RACE}"\
  --covar rms.control.02.cvc."${RACE}".pruned.pca.eigenvec\
  --threads 200\
  --glm firth-fallback hide-covar cols=chrom,pos,ref,alt1,nobs,a1countcc,a1freq,a1freqcc,beta,se,p,err\
  --out rms.co.02.cvc."${RACE}"

```

>> Run analysis based on subtypes

```{r generating IDs for different subtypes}

setwd('/ccls/home/sli/GWAS/libby/rmsdbGAP')

library(tidyverse)
library(data.table)

rms_0 <- fread("/ccls/home/sli/GWAS/libby/clinical/sebastian_rms.csv")

em = rms_0 %>% filter(subtype=="embryonal") %>% 
    dplyr::mutate(id = paste0(FID,"_",IID)) %>%
    select(id)

alv = rms_0 %>% filter(subtype=="alveolar") %>% 
    dplyr::mutate(id = paste0(FID,"_",IID)) %>%
    select(id)

RACE= "eur"
allSubs = fread("rms.co.cvc.eur.fam.updt") %>%
    select(V1,V2,V6)

emOnly = allSubs %>%
    dplyr::filter(V2 %in% em$id | V6==1) %>%
    select(V1,V2)     

alvOnly = allSubs %>%
    dplyr::filter(V2 %in% alv$id | V6==1) %>%
    select(V1,V2)       

write.table(emOnly, paste0("subtype.em.",RACE,".cvc.ids.txt"), 
    col.names = FALSE, row.names = FALSE, quote=FALSE)

write.table(alvOnly, paste0("subtype.alv.",RACE,".cvc.ids.txt"), 
    col.names = FALSE, row.names = FALSE, quote=FALSE)


```

```{bash run caco analysis in based on different subtypes}

#####################################################################
# setting the cvc cut off as 03

cd /ccls/home/sli/GWAS/libby/rmsdbGAP

# Select the proper subjects for subtypes
RACE="eur"
plink --bfile rms.co.cvc.$RACE.05\
  --keep subtype.em.$RACE.cvc.ids.txt\
  --make-bed\
  --out rms.em.co.cvc.$RACE

plink --bfile rms.co.cvc.$RACE.05\
  --keep subtype.alv.$RACE.cvc.ids.txt\
  --make-bed\
  --out rms.alv.co.cvc.$RACE

RACE="eur"
for sub in "em" "alv"; do
  # HWE among controls using keep if
  plink2 --bfile rms.$sub.co.cvc.$RACE\
    --hwe midp 1e-4 \
    --keep-if PHENO1 == control\
    --make-pgen \
    --out tmp/rms.$sub.co.cvc."${RACE}"_1\
    --threads 200

  # Make pgen again using remaining variants from pvar file 
  plink2 --bfile rms.$sub.co.cvc.$RACE\
    --make-pgen\
    --extract tmp/rms.$sub.co.cvc."${RACE}"_1.pvar\
    --out tmp/rms.$sub.co.cvc."${RACE}"_2\
    --threads 200

  # Make bed
  plink2 --pfile tmp/rms.$sub.co.cvc."${RACE}"_2\
    --make-bed\
    --out rms.$sub.control.cvc."${RACE}"

  #make bed for PCA analysis
  # Create pruned SNPs for later use as controlling variables in regression
  plink2 --bfile rms.$sub.control.cvc."${RACE}"\
    --make-bed \
    --out tmp/rms.$sub.co.cvc."${RACE}".pca
  
  plink2 --bfile tmp/rms.$sub.co.cvc."${RACE}".pca\
    --maf 0.1\
    --indep-pairwise 50 10 0.1 \
    --out tmp/rms.$sub.co.cvc."${RACE}".pruned
  
  plink2 --bfile rms.$sub.control.cvc."${RACE}"\
    --extract tmp/rms.$sub.co.cvc."${RACE}".pruned.prune.in\
    --pca 10\
    --out rms.$sub.control.cvc."${RACE}".pruned.pca\
    --threads 200
  
  # Regression
  plink2 --bfile rms.$sub.control.cvc."${RACE}"\
    --covar rms.$sub.control.cvc."${RACE}".pruned.pca.eigenvec\
    --threads 200\
    --glm firth-fallback hide-covar cols=chrom,pos,ref,alt1,nobs,a1countcc,a1freq,a1freqcc,beta,se,p,err\
    --out rms.$sub.co.cvc."${RACE}"
done


RACE="eur"
sub="em"
awk '$15<5E-8 {print}' rms.$sub.co.cvc."${RACE}".PHENO1.glm.logistic.hybrid > rms.$sub.co.cvc."${RACE}".sigs

sub="alv"
awk '$15<5E-8 {print}' rms.$sub.co.cvc."${RACE}".PHENO1.glm.logistic.hybrid > rms.$sub.co.cvc."${RACE}".sigs

#####################################################################
# setting the cvc cut off as 02

cd /ccls/home/sli/GWAS/libby/rmsdbGAP

# Select the proper subjects for subtypes
RACE="eur"
plink --bfile rms.co.02.cvc.$RACE.05\
  --keep subtype.em.$RACE.cvc.ids.txt\
  --make-bed\
  --out rms.em.co.02.cvc.$RACE

plink --bfile rms.co.02.cvc.$RACE.05\
  --keep subtype.alv.$RACE.cvc.ids.txt\
  --make-bed\
  --out rms.alv.co.02.cvc.$RACE

RACE="eur"
for sub in "em" "alv"; do
  # HWE among controls using keep if
  plink2 --bfile rms.$sub.co.02.cvc.$RACE\
    --hwe midp 1e-4 \
    --keep-if PHENO1 == control\
    --make-pgen \
    --out tmp/rms.$sub.co.02.cvc."${RACE}"_1\
    --threads 200

  # Make pgen again using remaining variants from pvar file 
  plink2 --bfile rms.$sub.co.02.cvc.$RACE\
    --make-pgen\
    --extract tmp/rms.$sub.co.02.cvc."${RACE}"_1.pvar\
    --out tmp/rms.$sub.co.02.cvc."${RACE}"_2\
    --threads 200

  # Make bed
  plink2 --pfile tmp/rms.$sub.co.02.cvc."${RACE}"_2\
    --make-bed\
    --out rms.$sub.control.02.cvc."${RACE}"

  #make bed for PCA analysis
  # Create pruned SNPs for later use as controlling variables in regression
  plink2 --bfile rms.$sub.control.02.cvc."${RACE}"\
    --make-bed \
    --out tmp/rms.$sub.co.02.cvc."${RACE}".pca
  
  plink2 --bfile tmp/rms.$sub.co.02.cvc."${RACE}".pca\
    --maf 0.1\
    --indep-pairwise 50 10 0.1 \
    --out tmp/rms.$sub.co.02.cvc."${RACE}".pruned
  
  plink2 --bfile rms.$sub.control.02.cvc."${RACE}"\
    --extract tmp/rms.$sub.co.02.cvc."${RACE}".pruned.prune.in\
    --pca 10\
    --out rms.$sub.control.02.cvc."${RACE}".pruned.pca\
    --threads 200
  
  # Regression
  plink2 --bfile rms.$sub.control.02.cvc."${RACE}"\
    --covar rms.$sub.control.02.cvc."${RACE}".pruned.pca.eigenvec\
    --threads 200\
    --glm firth-fallback hide-covar cols=chrom,pos,ref,alt1,nobs,a1countcc,a1freq,a1freqcc,beta,se,p,err\
    --out rms.$sub.co.02.cvc."${RACE}"
done


RACE="eur"
sub="em"
awk '$15<5E-8 {print}' rms.$sub.co.02.cvc."${RACE}".PHENO1.glm.logistic.hybrid > rms.$sub.co.02.cvc."${RACE}".sigs

sub="alv"
awk '$15<5E-8 {print}' rms.$sub.co.02.cvc."${RACE}".PHENO1.glm.logistic.hybrid > rms.$sub.co.02.cvc."${RACE}".sigs

```

```{r visualization}

setwd("/ccls/home/sli/GWAS/libby/rmsdbGAP")

library(tidyverse)
library(QCEWAS)
library(qqman)
library(data.table)

files =c("eur.cvc.postimp.PHENO1.glm.logistic.hybrid",
         "eur.02.cvc.postimp.PHENO1.glm.logistic.hybrid",
         "rms.co.cvc.eur.PHENO1.glm.logistic.hybrid",
         "rms.em.co.cvc.eur.PHENO1.glm.logistic.hybrid",
         "rms.alv.co.cvc.eur.PHENO1.glm.logistic.hybrid",
         "rms.co.02.cvc.eur.PHENO1.glm.logistic.hybrid",
         "rms.em.co.02.cvc.eur.PHENO1.glm.logistic.hybrid",
         "rms.alv.co.02.cvc.eur.PHENO1.glm.logistic.hybrid")

for(i in files){

  filename <- gsub("^.*/","",i)
  gwas_data <- fread(i) %>% na.omit()

  if(length(grep("meta",i)) !=0 ){
    gwas_data_1 <- gwas_data %>%
      dplyr::mutate(ID=MarkerName, P=`P-value`)%>%
      separate(MarkerName, c("CHROM","POS","REF","ALT"))%>%
      dplyr::mutate(CHROM = as.numeric(gsub("chr","",as.character(CHROM))),POS = as.numeric(as.character(POS)))
  } else{
    gwas_data_1 <- gwas_data %>%
      dplyr::mutate(P= as.numeric(as.character(P)),CHROM = as.numeric(as.character(`#CHROM`)), POS = as.numeric(as.character(POS)))%>%
      arrange(P)
  }

  gwas_data_1 <- gwas_data_1[is.finite(gwas_data_1$P),  ]

  lambda <- P_lambda(gwas_data_1$P)
  print(paste("lambda value of",filename,"is:",lambda))

  png(paste0(filename,".manhattan.png"), width = 10, height = 7, units = 'in', res = 300)
  manhattan(gwas_data_1, chr = "CHROM", bp = "POS", p = "P", snp = "ID")
  dev.off()

  png(paste0(filename,".qq.png"), width = 10, height = 7, units = 'in', res = 300)
  qq(gwas_data_1$P)
  dev.off()
}
# [1] "lambda value of eur.cvc.postimp.PHENO1.glm.logistic.hybrid is: 1.03855051511659"
# [1] "lambda value of eur.02.cvc.postimp.PHENO1.glm.logistic.hybrid is: 1.0381188956971"
# [1] "lambda value of rms.co.cvc.eur.PHENO1.glm.logistic.hybrid is: 1.0222035242047"
# [1] "lambda value of rms.em.co.cvc.eur.PHENO1.glm.logistic.hybrid is: 1.01455371758917"
# [1] "lambda value of rms.alv.co.cvc.eur.PHENO1.glm.logistic.hybrid is: 1.02865660359657"
# [1] "lambda value of rms.co.02.cvc.eur.PHENO1.glm.logistic.hybrid is: 1.02318050068121"
# [1] "lambda value of rms.em.co.02.cvc.eur.PHENO1.glm.logistic.hybrid is: 1.01431327544427"
# [1] "lambda value of rms.alv.co.02.cvc.eur.PHENO1.glm.logistic.hybrid is: 1.02853753685421"

```

# imputing on michigan server

```{bash imputing on michigan server}

cd /ccls/home/sli/GWAS/libby/rmsdbGAP

for race in "eur" "lat" "afr" "eas"; do
  plink --bfile rms.ccls.ahgwas\
    --keep $race.ids.txt\
    --indiv-sort f $race.ids.txt\
    --make-bed\
    --out $race.RCA
done    

# Impute by ancestry. Do european first for libby
for race in "eur" "lat" "afr" "eas"; do
  plink --bfile $race.RCA\
    --allow-no-sex\
    --recode vcf bgz\
    --keep-allele-order\
    --out michigan/$race.RCA.preimp
    
  #sort and tabix 
  bcftools sort michigan/$race.RCA.preimp.vcf.gz -Oz -o michigan/$race.RCA.sorted.vcf.gz
  tabix michigan/$race.RCA.sorted.vcf.gz
  #split by chromomse
  for chr in {1..22}; do
    bcftools view --regions "${chr}" michigan/$race.RCA.sorted.vcf.gz -o michigan/$race.RCA."${chr}".vcf.gz -Oz &
  done
done


```

# run caco analysis

```{bash run caco analysis in EUR}
cd /ccls/home/sli/GWAS/libby/rmsdbGAP

# Choose two sets of files 0.6 and 0.9
for chr in {1..22};do
  awk '($7 >0.6) {print $1}' <(zcat michigan_imputed/chr"${chr}".info.gz) | sed '1d' > \
      michigan_imputed/include.06.chr"${chr}".txt
  awk '($7 >0.9) {print $1}' <(zcat michigan_imputed/chr"${chr}".info.gz) | sed '1d' > \
      michigan_imputed/include.09.chr"${chr}".txt
done

# Convert imputed data to plink files
cd /ccls/home/sli/GWAS/libby/rmsdbGAP
RACE="eur"
for chr in {1..22};do
  plink2 --vcf michigan_imputed/chr"${chr}".dose.vcf.gz\
    dosage=HDS \
    --snps-only 'just-acgt' \
    --max-alleles 2 \
    --min-alleles 2 \
    --extract michigan_imputed/include.06.chr"${chr}".txt\
    --maf 0.01 \
    --make-bed\
    --out michigan_imputed/$RACE.chr"${chr}".iptd\
    --threads 200
done

find ./michigan_imputed -name "$RACE.chr*.iptd.bim" > ./michigan_imputed/merge.$RACE.list 
sed -i 's/.bim//g' ./michigan_imputed/merge.$RACE.list

plink --merge-list michigan_imputed/merge.$RACE.list\
  --indiv-sort 0\
  --make-bed\
  --out rms.co.$RACE
  
####################
mv rms.co.$RACE.fam rms.co.$RACE.fam.backup
awk '{print 0,$1"_"$2,$3,$4,$5,$6}' $RACE.ahgwas.ccls.rms.fam > rms.co.$RACE.fam

cd /ccls/home/sli/GWAS/libby/rmsdbGAP

for RACE in "eur"; do
  # HWE among controls using keep if
  plink2 --bfile rms.co.$RACE\
    --hwe midp 1e-4 \
    --keep-if PHENO1 == control \
    --make-pgen \
    --out tmp/rms.co."${RACE}"_1 \
    --threads 200

  # Make pgen again using remaining variants from pvar file 
  plink2 --bfile rms.co.$RACE\
    --make-pgen\
    --extract tmp/rms.co."${RACE}"_1.pvar \
    --out tmp/rms.co."${RACE}"_2\
    --threads 200

  # Make bed
  plink2 --pfile tmp/rms.co."${RACE}"_2\
    --make-bed \
    --out rms.control."${RACE}"

  #make bed for PCA analysis
  # Create pruned SNPs for later use as controlling variables in regression
  plink2 --bfile rms.control."${RACE}"\
    --make-bed \
    --out tmp/rms.co."${RACE}".pca
  
  plink2 --bfile tmp/rms.co."${RACE}".pca\
    --maf 0.1\
    --indep-pairwise 50 10 0.1 \
    --out tmp/rms.co."${RACE}".pruned
  
  plink2 --bfile rms.control."${RACE}"\
    --extract tmp/rms.co."${RACE}".pruned.prune.in\
    --pca 10\
    --out rms.control."${RACE}".pruned.pca\
    --threads 200
  
    # Regression
  plink2 --bfile rms.control."${RACE}"\
    --covar rms.control."${RACE}".pruned.pca.eigenvec\
    --threads 200\
    --glm hide-covar cols=chrom,pos,ref,alt1,nobs,a1countcc,a1freq,a1freqcc,beta,se,p\
    --out rms.co."${RACE}"
done

```

```{bash GWAS trial run for RMS CACO with SNPTEST}

cd /ccls/home/sli/GWAS/libby/rmsdbGAP

for RACE in "eur"; do
  for chr in {1..22};do
    awk '($7 < 0.3) {print $1}' <(zcat michigan_imputed/chr"${chr}".info.gz) | sed '1d' > \
        michigan_imputed/exclude.03.chr"${chr}".txt
  done
done

for RACE in "eur"; do
  for chr in {1..22};do
    vcftools --gzvcf michigan_imputed/chr"${chr}".dose.vcf.gz\
      --maf 0.01\
      --exclude michigan_imputed/exclude.03.chr"${chr}".txt\
      --recode --stdout | gzip -c \
      > snptest/$RACE.chr"${chr}".snptest.vcf.gz
  done
done

bcftools query -l snptest/$RACE.chr22.snptest.vcf.gz > snptest/samples.list.ids

for RACE in "eur"; do
  for chr in {1..22};do
    plink --vcf snptest/$RACE.chr"${chr}".snptest.vcf.gz \
    --double-id \
    --make-bed \
    --out snptest/plinkPCA/$RACE.chr"${chr}"
  done
done

RACE="eur"
find ./snptest/plinkPCA -name "$RACE.chr*.bim" > ./snptest/plinkPCA/merge.$RACE.snptest.list 
sed -i 's/.bim//g' snptest/plinkPCA/merge.$RACE.snptest.list

plink --merge-list snptest/plinkPCA/merge.$RACE.snptest.list \
  --make-bed \
  --out snptest/plinkPCA/$RACE.snptest.pca

plink2 --bfile snptest/plinkPCA/$RACE.snptest.pca \
  --maf 0.1 \
  --indep-pairwise 50 10 0.1 \
  --out snptest/plinkPCA/$RACE.snptest

plink2 --bfile snptest/plinkPCA/$RACE.snptest.pca \
  --extract snptest/plinkPCA/$RACE.snptest.prune.in\
  --pca 10 \
  --out snptest/$RACE.snptest\
  --threads 200
```

```{r generate covariate data}

setwd("/ccls/home/sli/GWAS/libby/rmsdbGAP/snptest")
library(tidyverse)
library(data.table)

pheno = fread("/ccls/home/sli/GWAS/libby/rmsdbGAP/eur.ahgwas.ccls.rms.fam") %>%
  mutate(subjectID=paste0(V1,"_",V2),caco=V6)%>%
  select(subjectID,caco)

PCs <- fread("eur.snptest.eigenvec") %>%
  select(-1) %>%
  rename( subjectID = IID )

samples <- read.table("samples.list.ids")
colnames(samples) <- "subjectID"

samples_cov <- samples %>%
  left_join(pheno,by="subjectID") %>%
  left_join(PCs,by="subjectID") %>%
  rename(ID = subjectID) %>%
  mutate(caco = ifelse(caco==1,0,1))
line2 <- c(0,"B",rep("C",10))
samples_cov_1 <- rbind(line2, samples_cov) %>% as.data.frame()

samples_cov_2 <- samples %>%
  left_join(pheno,by="subjectID") %>%
  left_join(PCs,by="subjectID") %>%
  rename(ID = subjectID) %>%
  mutate(caco = ifelse(caco==1,0,1))
line2_2 <- c(0,"B",rep("C",10))
samples_cov_2_1 <- rbind(line2_2, samples_cov_2) %>% as.data.frame()

write.table(samples_cov_1, "eur.snptest.vars.txt" , row.names = FALSE, quote=FALSE)

```

```{bash snptest of rms cases and controls }

cd /ccls/home/sli/GWAS/libby/rmsdbGAP

for RACE in "eur"; do
  for chr in {1..22};do
    snptest -data snptest/$RACE.chr"${chr}".snptest.vcf.gz \
      snptest/$RACE.snptest.vars.txt \
      -genotype_field GP \
      -method expected \
      -frequentist 1 \
      -pheno caco \
      -cov_names PC1 PC2 PC3 PC4 PC5 PC6 PC7 PC8 PC9 PC10 \
      -hwe \
      -o snptest/$RACE.chr"${chr}".out
    done
done

RACE="eur"
head -1 snptest/$RACE.chr1.out > snptest/overall.$RACE.snptest
tail -n +2 -q snptest/$RACE.chr*.out >> snptest/overall.$RACE.snptest

```

```{r visualization}

setwd("/ccls/home/sli/GWAS/libby/rmsdbGAP")

library(tidyverse)
library(QCEWAS)
library(qqman)
library(data.table)

files =c("rms.co.eur.PHENO1.glm.logistic")

for(i in files){

  filename <- gsub("^.*/","",i)
  gwas_data <- fread(i) %>% na.omit()

  if(length(grep("meta",i)) !=0 ){
    gwas_data_1 <- gwas_data %>%
      dplyr::mutate(ID=MarkerName, P=`P-value`)%>%
      separate(MarkerName, c("CHROM","POS","REF","ALT"))%>%
      dplyr::mutate(CHROM = as.numeric(gsub("chr","",as.character(CHROM))),POS = as.numeric(as.character(POS)))
  } else{
    gwas_data_1 <- gwas_data %>%
      dplyr::mutate(P= as.numeric(as.character(P)),CHROM = as.numeric(as.character(`#CHROM`)), POS = as.numeric(as.character(POS)))%>%
      arrange(P)
  }

  gwas_data_1 <- gwas_data_1[is.finite(gwas_data_1$P),  ]

  lambda <- P_lambda(gwas_data_1$P)
  print(paste("lambda value of",filename,"is:",lambda))

  png(paste0(filename,".manhattan.png"), width = 10, height = 7, units = 'in', res = 300)
  manhattan(gwas_data_1, chr = "CHROM", bp = "POS", p = "P", snp = "ID")
  dev.off()

  png(paste0(filename,".qq.png"), width = 10, height = 7, units = 'in', res = 300)
  qq(gwas_data_1$P)
  dev.off()
}


## visualization of snp.test results

setwd("/ccls/home/sli/GWAS/libby/rmsdbGAP/snptest")

files =c("snptest/overall.eur.snptest")

for(i in files){
  
  filename <- gsub("^.*/","",i)
  
  gwas <- fread(i, skip=13,fill=TRUE)
  gwas_1 <- gwas %>%
    select(rsid, chromosome, position, frequentist_add_pvalue) %>%
    mutate(chromosome = as.numeric(chromosome), position = as.numeric(position), frequentist_add_pvalue = as.numeric(frequentist_add_pvalue)) %>%
    na.omit()
  
  gwas_data_1 <- gwas_1[is.finite(gwas_1$frequentist_add_pvalue),  ]
  
  lambda <- P_lambda(gwas_data_1$frequentist_add_pvalue)
  print(paste("lambda value of",filename,"is:",lambda))
  
  png(paste0(filename,".manhattan.png"), width = 10, height = 7, units = 'in', res = 300)
  manhattan(gwas_data_1, chr = "chromosome", bp = "position", p = "frequentist_add_pvalue", snp = "rsid")
  dev.off()
  
  png(paste0(filename,".qq.png"), width = 10, height = 7, units = 'in', res = 300)
  qq(gwas_data_1$frequentist_add_pvalue)
  dev.off()

}


```

# Run dbGAP analysis based on subtypes

```{r generating IDs for different subtypes}

setwd('/ccls/home/sli/GWAS/libby/rmsdbGAP')

library(tidyverse)
library(class)
library(data.table)

rms_0 <- fread("/ccls/home/sli/GWAS/libby/clinical/sebastian_rms.csv")
em = rms_0 %>% filter(subtype=="embryonal") %>% 
    dplyr::mutate(id = paste0(FID,"_",IID)) %>%
    select(id)

alv = rms_0 %>% filter(subtype=="alveolar") %>% 
    dplyr::mutate(id = paste0(FID,"_",IID)) %>%
    select(id)

RACE= "eur"
allSubs = fread(paste0("rms.co.",RACE,".fam")) %>%
    select(V1,V2)

emOnly = allSubs %>%
    dplyr::filter(!V2 %in% alv$id)     

alvOnly = allSubs %>%
    dplyr::filter(!V2 %in% em$id)     

write.table(emOnly, paste0("subtype.em.",RACE,".ids.txt"), 
    col.names = FALSE, row.names = FALSE, quote=FALSE)

write.table(alvOnly, paste0("subtype.alv.",RACE,".ids.txt"), 
    col.names = FALSE, row.names = FALSE, quote=FALSE)

```

```{bash run caco analysis in based on different subtypes}

cd /ccls/home/sli/GWAS/libby/rmsdbGAP

# Select the proper subjects for subtypes
for RACE in "eur"; do

  plink --bfile rms.co.$RACE\
    --keep subtype.em.$RACE.ids.txt\
    --make-bed\
    --out rms.em.co.$RACE

  plink --bfile rms.co.$RACE\
    --keep subtype.alv.$RACE.ids.txt\
    --make-bed\
    --out rms.alv.co.$RACE
done

for RACE in "eur"; do
  for sub in "em" "alv"; do
    # HWE among controls using keep if
    plink2 --bfile rms.$sub.co.$RACE\
      --hwe midp 1e-4 \
      --keep-if PHENO1 == control \
      --make-pgen \
      --out tmp/rms.$sub.co."${RACE}"_1 \
      --threads 200
  
    # Make pgen again using remaining variants from pvar file 
    plink2 --bfile rms.$sub.co.$RACE\
      --make-pgen\
      --extract tmp/rms.$sub.co."${RACE}"_1.pvar \
      --out tmp/rms.$sub.co."${RACE}"_2\
      --threads 200
  
    # Make bed
    plink2 --pfile tmp/rms.$sub.co."${RACE}"_2\
      --make-bed \
      --out rms.$sub.control."${RACE}"
  
    #make bed for PCA analysis
    # Create pruned SNPs for later use as controlling variables in regression
    plink2 --bfile rms.$sub.control."${RACE}"\
      --make-bed \
      --out tmp/rms.$sub.co."${RACE}".pca
    
    plink2 --bfile tmp/rms.$sub.co."${RACE}".pca\
      --maf 0.1\
      --indep-pairwise 50 10 0.1 \
      --out tmp/rms.$sub.co."${RACE}".pruned
    
    plink2 --bfile rms.$sub.control."${RACE}"\
      --extract tmp/rms.$sub.co."${RACE}".pruned.prune.in\
      --pca 10\
      --out rms.$sub.control."${RACE}".pruned.pca\
      --threads 200
    
    # Regression
    plink2 --bfile rms.$sub.control."${RACE}"\
      --covar rms.$sub.control."${RACE}".pruned.pca.eigenvec\
      --threads 200\
      --glm hide-covar cols=chrom,pos,ref,alt1,nobs,a1countcc,a1freq,a1freqcc,beta,se,p\
      --out rms.$sub.co."${RACE}"
  done
done
```

```{r visualization}
setwd("/ccls/home/sli/GWAS/libby/rmsdbGAP")

library(tidyverse)
library(QCEWAS)
library(qqman)
library(data.table)

files =c("rms.em.co.eur.PHENO1.glm.logistic")
files =c("rms.alv.co.eur.PHENO1.glm.logistic")


for(i in files){

  filename <- gsub("^.*/","",i)
  gwas_data <- fread(i) %>% na.omit()

  if(length(grep("meta",i)) !=0 ){
    gwas_data_1 <- gwas_data %>%
      dplyr::mutate(ID=MarkerName, P=`P-value`)%>%
      separate(MarkerName, c("CHROM","POS","REF","ALT"))%>%
      dplyr::mutate(CHROM = as.numeric(gsub("chr","",as.character(CHROM))),POS = as.numeric(as.character(POS)))
  } else{
    gwas_data_1 <- gwas_data %>%
      dplyr::mutate(P= as.numeric(as.character(P)),CHROM = as.numeric(as.character(`#CHROM`)), POS = as.numeric(as.character(POS)))%>%
      arrange(P)
  }

  gwas_data_1 <- gwas_data_1[is.finite(gwas_data_1$P),  ]

  lambda <- P_lambda(gwas_data_1$P)
  print(paste("lambda value of",filename,"is:",lambda))

  png(paste0(filename,".manhattan.png"), width = 10, height = 7, units = 'in', res = 300)
  manhattan(gwas_data_1, chr = "CHROM", bp = "POS", p = "P", snp = "ID")
  dev.off()

  png(paste0(filename,".qq.png"), width = 10, height = 7, units = 'in', res = 300)
  qq(gwas_data_1$P)
  dev.off()
}

```

# imputing on topmed server

```{bash}

cd /ccls/home/sli/GWAS/libby/rmsdbGAP/topmed_imputed
for chr in {1..22}; do
  unzip -P '#jTIpSdmWH4uq0' chr_"${chr}".zip
  echo chr_"${chr}".zip unzip success
done  

```

```{bash run caco analysis in EUR, topmed}
cd /ccls/home/sli/GWAS/libby/rmsdbGAP

# Choose two sets of files 0.6 and 0.9
for chr in {1..22};do
  awk '($7 >0.6) {print $1}' <(zcat topmed_imputed/chr"${chr}".info.gz) | sed '1d' > \
      topmed_imputed/include.06.chr"${chr}".txt
  awk '($7 >0.9) {print $1}' <(zcat topmed_imputed/chr"${chr}".info.gz) | sed '1d' > \
      topmed_imputed/include.09.chr"${chr}".txt
done

# Convert imputed data to plink files
cd /ccls/home/sli/GWAS/libby/rmsdbGAP
RACE="eur"
for chr in {1..22};do
  plink2 --vcf topmed_imputed/chr"${chr}".dose.vcf.gz\
    dosage=HDS \
    --snps-only 'just-acgt' \
    --max-alleles 2 \
    --min-alleles 2 \
    --extract topmed_imputed/include.06.chr"${chr}".txt\
    --maf 0.01 \
    --make-bed\
    --out topmed_imputed/$RACE.chr"${chr}".iptd\
    --threads 200
done

find ./topmed_imputed -name "$RACE.chr*.iptd.bim" > ./topmed_imputed/merge.$RACE.list 
sed -i 's/.bim//g' ./topmed_imputed/merge.$RACE.list

plink --merge-list topmed_imputed/merge.$RACE.list\
  --indiv-sort 0\
  --make-bed\
  --out rms.co.38.$RACE
  
####################
mv rms.co.38.$RACE.fam rms.co.38.$RACE.fam.backup
awk '{print 0,$1"_"$2,$3,$4,$5,$6}' $RACE.ahgwas.ccls.38.rms.fam > rms.co.38.$RACE.fam

####################
cd /ccls/home/sli/GWAS/libby/rmsdbGAP

for RACE in "eur"; do
  # HWE among controls using keep if
  plink2 --bfile rms.co.38.$RACE\
    --hwe midp 1e-4 \
    --keep-if PHENO1 == control \
    --make-pgen \
    --out tmp/rms.co.38."${RACE}"_1 \
    --threads 200

  # Make pgen again using remaining variants from pvar file 
  plink2 --bfile rms.co.38.$RACE\
    --make-pgen\
    --extract tmp/rms.co.38."${RACE}"_1.pvar \
    --out tmp/rms.co.38."${RACE}"_2\
    --threads 200

  # Make bed
  plink2 --pfile tmp/rms.co.38."${RACE}"_2\
    --make-bed \
    --out rms.control.38."${RACE}"

  #make bed for PCA analysis
  # Create pruned SNPs for later use as controlling variables in regression
  plink2 --bfile rms.control.38."${RACE}"\
    --make-bed \
    --out tmp/rms.co.38."${RACE}".pca
  
  plink2 --bfile tmp/rms.co.38."${RACE}".pca\
    --maf 0.1\
    --indep-pairwise 50 10 0.1 \
    --out tmp/rms.co.38."${RACE}".pruned
  
  plink2 --bfile rms.control.38."${RACE}"\
    --extract tmp/rms.co.38."${RACE}".pruned.prune.in\
    --pca 10\
    --out rms.control.38."${RACE}".pruned.pca\
    --threads 200
  
  # Regression
  plink2 --bfile rms.control.38."${RACE}"\
    --covar rms.control.38."${RACE}".pruned.pca.eigenvec\
    --threads 200\
    --glm firth-fallback hide-covar cols=chrom,pos,ref,alt1,nobs,a1countcc,a1freq,a1freqcc,beta,se,p\
    --out rms.co.38."${RACE}"
done

```

subtype analysis

```{r generating IDs for different subtypes}

setwd('/ccls/home/sli/GWAS/libby/rmsdbGAP')

library(tidyverse)
library(class)
library(data.table)

rms_0 <- fread("/ccls/home/sli/GWAS/libby/clinical/sebastian_rms.csv")
em = rms_0 %>% filter(subtype=="embryonal") %>% 
    dplyr::mutate(id = paste0(FID,"_",IID)) %>%
    select(id)

alv = rms_0 %>% filter(subtype=="alveolar") %>% 
    dplyr::mutate(id = paste0(FID,"_",IID)) %>%
    select(id)

RACE= "eur"
allSubs = fread(paste0("rms.co.38.",RACE,".fam")) %>%
    select(V1,V2,V6)

emOnly = allSubs %>%
    dplyr::filter(V2 %in% em$id | V6==1)     

alvOnly = allSubs %>%
    dplyr::filter(V2 %in% alv$id | V6==1)        

write.table(emOnly, paste0("subtype.em.38.",RACE,".ids.txt"), 
    col.names = FALSE, row.names = FALSE, quote=FALSE)

write.table(alvOnly, paste0("subtype.alv.38.",RACE,".ids.txt"), 
    col.names = FALSE, row.names = FALSE, quote=FALSE)

```

```{bash run caco analysis in based on different subtypes}

cd /ccls/home/sli/GWAS/libby/rmsdbGAP

# Select the proper subjects for subtypes
for RACE in "eur"; do

  plink --bfile rms.co.38.$RACE\
    --keep subtype.em.38.$RACE.ids.txt\
    --make-bed\
    --out rms.em.co.38.$RACE

  plink --bfile rms.co.38.$RACE\
    --keep subtype.alv.38.$RACE.ids.txt\
    --make-bed\
    --out rms.alv.co.38.$RACE
done

for RACE in "eur"; do
  for sub in "em" "alv"; do
    plink2 --bfile rms.$sub.co.38.$RACE\
      --hwe midp 1e-4 \
      --keep-if PHENO1 == control \
      --make-pgen \
      --out tmp/rms.$sub.co.38."${RACE}"_1 \
      --threads 200
  
    # Make pgen again using remaining variants from pvar file 
    plink2 --bfile rms.$sub.co.38.$RACE\
      --make-pgen\
      --extract tmp/rms.$sub.co.38."${RACE}"_1.pvar \
      --out tmp/rms.$sub.co.38."${RACE}"_2\
      --threads 200
  
    # Make bed
    plink2 --pfile tmp/rms.$sub.co.38."${RACE}"_2\
      --make-bed \
      --out rms.$sub.control.38."${RACE}"
  
    #make bed for PCA analysis
    plink2 --bfile rms.$sub.control.38."${RACE}"\
      --make-bed \
      --out tmp/rms.$sub.co.38."${RACE}".pca
    
    plink2 --bfile tmp/rms.$sub.co.38."${RACE}".pca\
      --maf 0.1\
      --indep-pairwise 50 10 0.1 \
      --out tmp/rms.$sub.co.38."${RACE}".pruned
    
    plink2 --bfile rms.$sub.control.38."${RACE}"\
      --extract tmp/rms.$sub.co.38."${RACE}".pruned.prune.in\
      --pca 10\
      --out rms.$sub.control.38."${RACE}".pruned.pca\
      --threads 200
    
    # Regression
    plink2 --bfile rms.$sub.control.38."${RACE}"\
      --covar rms.$sub.control.38."${RACE}".pruned.pca.eigenvec\
      --threads 200\
      --glm firth-fallback hide-covar cols=chrom,pos,ref,alt1,nobs,a1countcc,a1freq,a1freqcc,beta,se,p\
      --out rms.$sub.co.38."${RACE}"
  done
done


RACE="eur"
awk '$15<5E-8 {print}' rms.co.38."${RACE}".PHENO1.glm.logistic.hybrid > rms.co.38."${RACE}".sigs

RACE="eur"
sub="em"
awk '$15<5E-8 {print}' rms.$sub.co.38."${RACE}".PHENO1.glm.logistic.hybrid > rms.$sub.co.38."${RACE}".sigs

RACE="eur"
sub="alv"
awk '$15<5E-8 {print}' rms.$sub.co.38."${RACE}".PHENO1.glm.logistic.hybrid > rms.$sub.co.38."${RACE}".sigs

```

```{r visualization}
setwd("/ccls/home/sli/GWAS/libby/rmsdbGAP")

library(tidyverse)
library(QCEWAS)
library(qqman)
library(data.table)

files =c("rms.co.38.eur.PHENO1.glm.logistic.hybrid",
          "rms.em.co.38.eur.PHENO1.glm.logistic.hybrid",
          "rms.alv.co.38.eur.PHENO1.glm.logistic.hybrid")


for(i in files){

  filename <- gsub("^.*/","",i)
  gwas_data <- fread(i) %>% na.omit()

  if(length(grep("meta",i)) !=0 ){
    gwas_data_1 <- gwas_data %>%
      dplyr::mutate(ID=MarkerName, P=`P-value`)%>%
      separate(MarkerName, c("CHROM","POS","REF","ALT"))%>%
      dplyr::mutate(CHROM = as.numeric(gsub("chr","",as.character(CHROM))),POS = as.numeric(as.character(POS)))
  } else{
    gwas_data_1 <- gwas_data %>%
      dplyr::mutate(P= as.numeric(as.character(P)),CHROM = as.numeric(as.character(`#CHROM`)), POS = as.numeric(as.character(POS)))%>%
      arrange(P)
  }

  gwas_data_1 <- gwas_data_1[is.finite(gwas_data_1$P),  ]

  lambda <- P_lambda(gwas_data_1$P)
  print(paste("lambda value of",filename,"is:",lambda))

  png(paste0(filename,".manhattan.png"), width = 10, height = 7, units = 'in', res = 300)
  manhattan(gwas_data_1, chr = "CHROM", bp = "POS", p = "P", snp = "ID")
  dev.off()

  png(paste0(filename,".qq.png"), width = 10, height = 7, units = 'in', res = 300)
  qq(gwas_data_1$P)
  dev.off()
}


```
